// inet-ui/esm5/src/sticky-bar/sticky-bar.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as ResizeSensor from 'resize-sensor';
var StickyBar = /** @class */ (function () {
    function StickyBar(options) {
        this.topSpace = 5;
        this.bottomSpace = 5;
        this.onScroll = this.onScroll.bind(this);
        this.init(options);
    }
    /**
     * @param {?} options
     * @return {?}
     */
    StickyBar.prototype.init = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        var _this = this;
        for (var k in options) {
            this[k] = options[k];
        }
        if (!this.bar) {
            console.error('[StickyBar] Missing scroller or bar element');
            return;
        }
        this.container = this.bar.parentElement;
        this.temp = document.createElement('div');
        this.temp.style.cssText = 'position: relative; overflow: hidden';
        this.container.insertBefore(this.temp, this.bar);
        // this.temp.appendChild(this.bar);
        this.sensor = new ResizeSensor(this.temp, (/**
         * @return {?}
         */
        function () {
            if (_this.preventResize) {
                _this.preventResize = false;
                return;
            }
            if (!_this.isFixed()) {
                return;
            }
            _this.clearFixed();
            _this.onScroll();
        }));
        (this.scroller || window).addEventListener('scroll', this.onScroll);
    };
    /**
     * @return {?}
     */
    StickyBar.prototype.onScroll = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var scrollerBounding;
        if (this.scroller) {
            scrollerBounding = this.scroller.getBoundingClientRect();
            scrollerBounding['scrollHeight'] = this.scroller.scrollHeight;
            scrollerBounding['scrollTop'] = this.scroller.scrollTop;
        }
        else {
            scrollerBounding = {
                scrollHeight: document.documentElement.scrollHeight,
                scrollTop: pageYOffset || scrollY,
                top: 0,
                height: innerHeight
            };
        }
        /** @type {?} */
        var barBounding = this.bar.getBoundingClientRect();
        /** @type {?} */
        var staticBounding = this.temp.getBoundingClientRect();
        /** @type {?} */
        var isCanScrollFixed = scrollerBounding.scrollHeight >
            scrollerBounding.scrollTop + staticBounding.top + barBounding.height + 50;
        if (!isCanScrollFixed) {
            if (this.isFixed()) {
                this.clearFixed();
            }
            return;
        }
        /** @type {?} */
        var barTop = this.isFixed() ? staticBounding.top : barBounding.top;
        /** @type {?} */
        var barAboveTop = barTop - this.topSpace <= scrollerBounding.top;
        /** @type {?} */
        var barAboveBottom = barTop + barBounding.height + this.bottomSpace <= scrollerBounding.top + scrollerBounding.height;
        if (barAboveTop && barAboveBottom) {
            if (!this.isFixed()) {
                /** @type {?} */
                var topFixed = scrollerBounding.top + this.topSpace;
                // content fixed's height greater than scroller's height
                /** @type {?} */
                var distance = barBounding.height + this.bottomSpace + this.topSpace - scrollerBounding.height;
                if (distance > 0) {
                    topFixed -= distance;
                }
                this.preventResize = true;
                this.resizeDimension();
                this.setCssBar({
                    top: topFixed + 'px',
                    position: 'fixed'
                });
                this.setFixed(true);
            }
        }
        else if (this.isFixed()) {
            /** @type {?} */
            var barAboveTopStatic = barTop <= scrollerBounding.top;
            /** @type {?} */
            var barAboveBottomStatic = barTop + barBounding.height <= scrollerBounding.top + scrollerBounding.height;
            if (barAboveTopStatic && barAboveBottomStatic) {
                return;
            }
            this.clearFixed();
        }
    };
    /**
     * @return {?}
     */
    StickyBar.prototype.destroy = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.sensor) {
            (this.scroller || window).removeEventListener('scroll', this.onScroll);
            this.resetBarCss();
            this.sensor.detach();
            this.temp.remove();
            Object.keys(this).forEach((/**
             * @param {?} k
             * @return {?}
             */
            function (k) { return delete _this[k]; }));
        }
    };
    /**
     * @private
     * @return {?}
     */
    StickyBar.prototype.resizeDimension = /**
     * @private
     * @return {?}
     */
    function () {
        this.bar.style.width = this.temp.clientWidth + 'px';
        this.temp.style.height = this.getHeightBar() + 'px';
    };
    /**
     * @private
     * @return {?}
     */
    StickyBar.prototype.getHeightBar = /**
     * @private
     * @return {?}
     */
    function () {
        return parseFloat(getComputedStyle(this.bar).height);
    };
    /**
     * @private
     * @param {?} fixed
     * @return {?}
     */
    StickyBar.prototype.setFixed = /**
     * @private
     * @param {?} fixed
     * @return {?}
     */
    function (fixed) {
        this.fixed = fixed;
    };
    /**
     * @private
     * @return {?}
     */
    StickyBar.prototype.isFixed = /**
     * @private
     * @return {?}
     */
    function () {
        return this.fixed;
    };
    /**
     * @private
     * @return {?}
     */
    StickyBar.prototype.clearFixed = /**
     * @private
     * @return {?}
     */
    function () {
        this.resetBarCss();
        this.temp.style.height = '';
        this.setFixed(false);
    };
    /**
     * @private
     * @return {?}
     */
    StickyBar.prototype.resetBarCss = /**
     * @private
     * @return {?}
     */
    function () {
        this.setCssBar({
            top: '',
            position: '',
            transform: '',
            width: ''
        });
    };
    /**
     * @private
     * @param {?} css
     * @return {?}
     */
    StickyBar.prototype.setCssBar = /**
     * @private
     * @param {?} css
     * @return {?}
     */
    function (css) {
        for (var k in css) {
            this.bar.style[k] = css[k];
        }
    };
    return StickyBar;
}());
export { StickyBar };
if (false) {
    /**
     * @type {?}
     * @private
     */
    StickyBar.prototype.temp;
    /**
     * @type {?}
     * @private
     */
    StickyBar.prototype.sensor;
    /**
     * @type {?}
     * @private
     */
    StickyBar.prototype.fixed;
    /**
     * @type {?}
     * @private
     */
    StickyBar.prototype.preventResize;
    /** @type {?} */
    StickyBar.prototype.container;
    /** @type {?} */
    StickyBar.prototype.scroller;
    /** @type {?} */
    StickyBar.prototype.bar;
    /** @type {?} */
    StickyBar.prototype.topSpace;
    /** @type {?} */
    StickyBar.prototype.bottomSpace;
}
/**
 * @record
 */
export function StickyBarOptions() { }
if (false) {
    /** @type {?} */
    StickyBarOptions.prototype.bar;
    /** @type {?|undefined} */
    StickyBarOptions.prototype.scroller;
    /** @type {?|undefined} */
    StickyBarOptions.prototype.container;
    /** @type {?|undefined} */
    StickyBarOptions.prototype.topSpace;
    /** @type {?|undefined} */
    StickyBarOptions.prototype.bottomSpace;
}
