// inet-ui/esm5/src/ng-select/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { newId } from './model/id';
export { ItemsList } from './model/items-list';
export { KeyCode } from './model/ng-select.types';
export { DefaultSelectionModelFactory, DefaultSelectionModel } from './model/selection-model';
export { NgDropdownPanelComponent } from './components/dropdown/ng-dropdown-panel.component';
export { NgOptionComponent } from './components/option/ng-option.component';
export { SELECTION_MODEL_FACTORY, NgSelectComponent } from './components/select/ng-select.component';
export { NgOptionHighlightDirective } from './directive/ng-option-highlight.directive';
export { NgOptionTemplateDirective, NgOptgroupTemplateDirective, NgLabelTemplateDirective, NgMultiLabelTemplateDirective, NgHeaderTemplateDirective, NgFooterTemplateDirective, NgNotFoundTemplateDirective, NgTypeToSearchTemplateDirective, NgLoadingTextTemplateDirective, NgTagTemplateDirective } from './directive/ng-templates.directive';
export { NgSelectConfig } from './service/config.service';
export { ConsoleService } from './service/console.service';
export { VirtualScrollService } from './service/virtual-scroll.service';
export { WindowService } from './service/window.service';
export { stripSpecialChars } from './utils/search-helper';
export { isDefined, isObject, isPromise, isFunction } from './utils/value-utils';
export { NgSelectModule } from './ng-select.module';
