// inet-chat/esm2015/src/DateFormatter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as _moment from 'moment';
/** @type {?} */
const moment = _moment;
export class DateFormatter {
    /**
     * @param {?} date
     */
    constructor(date) {
        this.date = new Date(date);
    }
    /**
     * @return {?}
     */
    formatFullYear() {
        return moment(this.date).format('DD/MM/YYYY, H:mm');
    }
    /**
     * @return {?}
     */
    formatYear() {
        return moment(this.date).format('DD/MM/YYYY');
    }
    /**
     * @return {?}
     */
    formatMonth() {
        return moment(this.date).format('DD/MM, H:mm');
    }
    /**
     * @return {?}
     */
    formatDate() {
        return moment(this.date).format('DD/MM, H:mm');
    }
    /**
     * @return {?}
     */
    formatTime() {
        return moment(this.date).format('H:mm');
    }
    /**
     * @return {?}
     */
    format() {
        if (!this.date.getTime()) {
            return '';
        }
        /** @type {?} */
        const now = new Date();
        /** @type {?} */
        const conditions = {
            onYear: now.getFullYear() === this.date.getFullYear(),
            onMonth: now.getMonth() === this.date.getMonth(),
            onDate: now.getDate() === this.date.getDate(),
        };
        if (conditions.onYear && conditions.onMonth && conditions.onDate) {
            return this.formatTime();
        }
        if (conditions.onYear && conditions.onMonth) {
            return this.formatDate();
        }
        if (conditions.onYear) {
            return this.formatMonth();
        }
        return this.formatYear();
    }
}
if (false) {
    /** @type {?} */
    DateFormatter.prototype.date;
}
