// inet-chat/esm2015/src/video-conference.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ErrorMessage, HttpClientService } from "inet-core";
export class VideoConferenceService {
    /**
     * @param {?} httpService
     */
    constructor(httpService) {
        this.httpService = httpService;
        this.domain = 'jitsi.inetcloud.vn';
        this.messageJoin = { app: 'iChat', conferenceID: '', sender: '', joined: '', inAccess: '' };
        return VideoConferenceService.instance = VideoConferenceService.instance || this;
    }
    /**
     * @param {?} parentNode
     * @param {?} roomName
     * @param {?=} audioOnly
     * @param {?=} height
     * @param {?=} callback
     * @return {?}
     */
    join(parentNode, roomName, audioOnly = false, height = 350, callback) {
        if (!JitsiMeetExternalAPI) {
            return;
        }
        /** @type {?} */
        let tokenId = '';
        this.httpService.postJSON(iNet.getPUrl('conference/authorized'), { roomID: roomName }).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            tokenId = data.webdata || '';
            this.dispose();
            /** @type {?} */
            const options = {
                roomName: roomName,
                width: '100%',
                height: '100%',
                parentNode: parentNode,
                jwt: tokenId
            };
            if (audioOnly) {
                options.height = '100%';
                options['configOverwrite'] = {
                    startAudioOnly: true,
                    startWithVideoMuted: true
                };
            }
            this.api = new JitsiMeetExternalAPI(this.domain, options);
            if (callback) {
                callback(this.api);
            }
            this.api.executeCommand('sendTones', {
                tones: '12345#',
                // The dial pad touch tones to play. For example, '12345#'.
                duration: 200,
                // Optional. The number of milliseconds each tone should play. The default is 200.
                pause: 200 // Optional. The number of milliseconds between each tone. The default is 200.
            });
            // this.api.executeCommand('displayName', iNet.displayName);
            // this.api.executeCommand('email', iNet.username);
            // this.api.executeCommand('setVideoQuality', 720);
        }));
        /*
        this.api.addEventListener('readyToClose',  ()=> {
            this.dispose();
        });

         */
    }
    /**
     * @param {?} parentNode
     * @param {?} userName
     * @param {?=} audioOnly
     * @param {?=} height
     * @param {?=} callback
     * @return {?}
     */
    callTo(parentNode, userName, audioOnly = false, height = 350, callback) {
        /** @type {?} */
        let conferenceID = '';
        this.httpService.postJSON(iNet.getPUrl('conference/end2end'), { callTo: userName }).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data.type !== ErrorMessage.TYPE) {
                conferenceID = data.conferenceID || '';
                if (!iNet.isEmpty(conferenceID)) {
                    if (!JitsiMeetExternalAPI) {
                        return;
                    }
                    /** @type {?} */
                    let tokenId = '';
                    /** @type {?} */
                    let messageJSON = { sender: iNet.username, conferenceID: conferenceID };
                    /** @type {?} */
                    let __messageStr = 'uri_message:video_conference:' + this.encodeBase64(JSON.stringify(messageJSON)) || '';
                    // this.chatMessageService.send(userName,__messageStr);
                    this.httpService.postJSON(iNet.getPUrl('conference/authorized'), { roomID: conferenceID }).subscribe((/**
                     * @param {?} data
                     * @return {?}
                     */
                    (data) => {
                        tokenId = data.webdata || '';
                        this.dispose();
                        /** @type {?} */
                        const options = {
                            roomName: conferenceID,
                            width: '100%',
                            height: '100%',
                            parentNode: parentNode,
                            jwt: tokenId
                        };
                        if (audioOnly) {
                            options.height = '100%';
                            options['configOverwrite'] = {
                                startAudioOnly: true,
                                startWithVideoMuted: true
                            };
                        }
                        this.api = new JitsiMeetExternalAPI(this.domain, options);
                        this.messageJoin.sender = iNet.username;
                        this.messageJoin.conferenceID = conferenceID;
                        if (callback) {
                            callback(this.api, __messageStr);
                        }
                        this.api.executeCommand('sendTones', {
                            tones: '12345#',
                            // The dial pad touch tones to play. For example, '12345#'.
                            duration: 200,
                            // Optional. The number of milliseconds each tone should play. The default is 200.
                            pause: 200 // Optional. The number of milliseconds between each tone. The default is 200.
                        });
                        // this.api.executeCommand('displayName', iNet.displayName);
                        // this.api.executeCommand('email', iNet.username);
                        // this.api.executeCommand('setVideoQuality', 720);
                    }));
                }
            }
        }));
        /*
        this.api.addEventListener('readyToClose',  ()=> {
            this.dispose();
        });

         */
    }
    /**
     * @return {?}
     */
    getApi() {
        return this.api;
    }
    /**
     * @return {?}
     */
    dispose() {
        if (this.api) {
            this.api.dispose();
            delete this.api;
        }
    }
    /**
     * @return {?}
     */
    getMessageJoin() {
        return this.messageJoin;
    }
    /**
     * @param {?} receiver
     * @return {?}
     */
    getRoomNameByReceiver(receiver) {
        return iNet.username.split('@')[0] + '_' + receiver.split('@')[0];
    }
    /**
     * @return {?}
     */
    isExistMeet() {
        return !!JitsiMeetExternalAPI;
    }
    /**
     * @param {?} str
     * @return {?}
     */
    encodeBase64(str) {
        return window.btoa(unescape(encodeURIComponent(str)));
    }
}
VideoConferenceService.instance = null;
VideoConferenceService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
VideoConferenceService.ctorParameters = () => [
    { type: HttpClientService }
];
if (false) {
    /** @type {?} */
    VideoConferenceService.instance;
    /** @type {?} */
    VideoConferenceService.prototype.api;
    /** @type {?} */
    VideoConferenceService.prototype.domain;
    /** @type {?} */
    VideoConferenceService.prototype.messageJoin;
    /**
     * @type {?}
     * @private
     */
    VideoConferenceService.prototype.httpService;
}
