// inet-ui/esm2015/src/common/dictionary.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from 'inet-core';
import { Observable } from 'rxjs/Rx';
import 'rxjs/add/operator/map';
import 'rxjs/add/observable/of';
import 'rxjs/add/operator/do';
import { shareReplay } from 'rxjs/operators';
export class DictionaryService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.url = {
            find_by_key: iNet.getPUrl('global/category/dicts/fbykeys'),
            get_sex: iNet.getPUrl('gl/contact/profile/sex')
        };
        this.$findObservableCache = {};
        this.store = {};
        this.inventory = {};
        return DictionaryService.instance = DictionaryService.instance || this;
    }
    /**
     * @param {?} keyStr
     * @return {?}
     */
    findByKeys(keyStr) {
        /** @type {?} */
        const keys = keyStr.split(';');
        if (keys.length === 1 && this.inventory.hasOwnProperty(keyStr)) { // single key
            return Observable.of({ data: this.inventory[keyStr] });
        }
        if (this.store[keyStr]) { // multiple keys
            return Observable.of({ data: this.store[keyStr] });
        }
        if (!this.$findObservableCache[keyStr]) {
            this.$findObservableCache[keyStr] = this.http.getJSON(this.url.find_by_key, { keys: keyStr })
                .do((/**
             * @param {?} res
             * @return {?}
             */
            res => this.store[keyStr] = res['data']))
                .do((/**
             * @param {?} res
             * @return {?}
             */
            res => Object.assign(this.inventory, res['data'])))
                .pipe(shareReplay(1));
        }
        return this.$findObservableCache[keyStr];
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    getSex(params) {
        return this.http.postJSON(this.url.get_sex, params);
    }
}
DictionaryService.instance = null;
DictionaryService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DictionaryService.ctorParameters = () => [
    { type: HttpClientService }
];
if (false) {
    /** @type {?} */
    DictionaryService.instance;
    /**
     * @type {?}
     * @private
     */
    DictionaryService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    DictionaryService.prototype.$findObservableCache;
    /**
     * @type {?}
     * @private
     */
    DictionaryService.prototype.store;
    /**
     * @type {?}
     * @private
     */
    DictionaryService.prototype.inventory;
    /**
     * @type {?}
     * @private
     */
    DictionaryService.prototype.http;
}
