// inet-ui/esm2015/src/common/location.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService, ResponseData } from 'inet-core';
import { Observable } from 'rxjs/Rx';
import 'rxjs/add/operator/map';
import 'rxjs/add/observable/of';
import 'rxjs/add/operator/do';
import { shareReplay } from 'rxjs/operators';
export class LocationService {
    // store district by province code
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.url = {
            // list_country: iNet.getPUrl('cloudapp/country/list'),
            // list_city: iNet.getPUrl('cloudapp/city/list'),
            // list_district: iNet.getPUrl('cloudapp/district/list'),
            list_country: iNet.getPUrl('gl/country/list'),
            list_city: iNet.getPUrl('gl/city/list'),
            list_district: iNet.getPUrl('gl/district/list')
        };
        this.$provinceCache = {};
        this.$districtCache = {};
        this.provinceData = {}; // store province by country code
        // store province by country code
        this.districtData = {}; // store district by province code
        return LocationService.instance = LocationService.instance || this;
    }
    /**
     * @private
     * @return {?}
     */
    getCountry() {
        return this.http.getJSON(this.url.list_country).do((/**
         * @param {?} res
         * @return {?}
         */
        res => this.countries = res['items']));
    }
    /**
     * @private
     * @param {?} countryCode
     * @return {?}
     */
    _getProvinceByCountryCode(countryCode) {
        return this.http.getJSON(this.url.list_city, { countryCode: countryCode })
            .do((/**
         * @param {?} res
         * @return {?}
         */
        res => this.provinceData[countryCode] = res['items']));
    }
    /**
     * @private
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    _getDistrictByCode(countryCode, provinceCode) {
        /** @type {?} */
        const key = countryCode + '_' + provinceCode;
        return this.http.getJSON(this.url.list_district, { countryCode: countryCode, cityCode: provinceCode })
            .do((/**
         * @param {?} res
         * @return {?}
         */
        res => this.districtData[key] = res['items']));
    }
    /**
     * @return {?}
     */
    listCountry() {
        if (this.countries) {
            /** @type {?} */
            const responseData = new ResponseData();
            responseData.items = this.countries;
            return Observable.of(responseData);
        }
        if (!this.$countryCache) {
            this.$countryCache = this.getCountry().pipe(shareReplay(1));
        }
        return this.$countryCache;
    }
    /*
         * @deprecated Use getProvinceByCountryCode(countryCode) instead
         */
    /**
     * @param {?} countryId
     * @return {?}
     */
    getProvinceByCountryId(countryId) {
        return this.http.getJSON(this.url.list_city, { country: countryId });
    }
    /*
         * @deprecated Use getDistrictByCode(countryCode, provinceCode) instead
         */
    /**
     * @param {?} uuid
     * @return {?}
     */
    getDistrictByProvinceId(uuid) {
        return this.http.getJSON(this.url.list_district, { city: uuid });
    }
    /**
     * @param {?} countryCode
     * @return {?}
     */
    getProvinceByCountryCode(countryCode) {
        if (this.provinceData[countryCode]) {
            /** @type {?} */
            const responseData = new ResponseData();
            responseData.items = this.provinceData[countryCode];
            return Observable.of(responseData);
        }
        if (!this.$provinceCache[countryCode]) {
            this.$provinceCache[countryCode] = this._getProvinceByCountryCode(countryCode).pipe(shareReplay(1));
        }
        return this.$provinceCache[countryCode];
    }
    /**
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    getDistrictByCode(countryCode, provinceCode) {
        /** @type {?} */
        const key = countryCode + '_' + provinceCode;
        if (this.districtData[key]) {
            /** @type {?} */
            const responseData = new ResponseData();
            responseData.items = this.districtData[key];
            return Observable.of(responseData);
        }
        if (!this.$districtCache[key]) {
            this.$districtCache[key] = this._getDistrictByCode(countryCode, provinceCode).pipe(shareReplay(1));
        }
        return this.$districtCache[key];
    }
}
LocationService.instance = null;
LocationService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LocationService.ctorParameters = () => [
    { type: HttpClientService }
];
if (false) {
    /** @type {?} */
    LocationService.instance;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.$countryCache;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.$provinceCache;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.$districtCache;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.countries;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.provinceData;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.districtData;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.http;
}
