// inet-ui/esm2015/src/common/notify-message.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from 'inet-core';
export class NotifyMessageService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.url = {
            list: iNet.getPUrl('social/activity/notify'),
            clear: iNet.getPUrl('social/activity/clearnotify'),
            load_notify: iNet.getPUrl('social/activity/loadnotify')
        };
        return NotifyMessageService.instance = NotifyMessageService.instance || this;
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    getMessages(params = {}) {
        return this.http.getJSON(this.url.list, params);
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    count(params) {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            this.http.getJSON(this.url.list, params).subscribe((/**
             * @param {?} v
             * @return {?}
             */
            (v) => {
                resolve(v);
            }), (/**
             * @param {?} err
             * @return {?}
             */
            (err) => {
                resolve();
            }));
        }));
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    clearAll(params = {}) {
        return this.http.postJSON(this.url.clear, params);
    }
    /**
     * @param {?} app
     * @param {?} activityId
     * @return {?}
     */
    loadNotify(app, activityId) {
        return this.http.getJSON(this.url.load_notify, { activity: activityId, application: app });
    }
    /**
     * @param {?} activityId
     * @return {?}
     */
    clearByActivityId(activityId) {
        return this.http.postJSON(this.url.clear, { activity: activityId });
    }
}
NotifyMessageService.instance = null;
NotifyMessageService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
NotifyMessageService.ctorParameters = () => [
    { type: HttpClientService }
];
if (false) {
    /** @type {?} */
    NotifyMessageService.instance;
    /**
     * @type {?}
     * @private
     */
    NotifyMessageService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    NotifyMessageService.prototype.http;
}
