// inet-social/esm5/src/social.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { APP_ACTIVITY, APP_COMMENT, PINNEWS, Policy, PolicyIcon, PolicyValueAll, SHARE_COMPANY, SHARE_FRIEND, SHARE_ME, SHARE_PUBLISH, SocialTag } from './model/Config';
import { CoreService, HtmlUtils, HttpClientService, WebSocketClientService } from 'inet-core';
import { Subject } from "rxjs";
import { FrontViewService, PhotoSwipe, FileFormatService } from "inet-ui";
var SocialService = /** @class */ (function () {
    function SocialService(httpClient, coreService, photoSwipe, socketService, fileFormatService, frontViewService) {
        var _this = this;
        this.httpClient = httpClient;
        this.coreService = coreService;
        this.photoSwipe = photoSwipe;
        this.socketService = socketService;
        this.fileFormatService = fileFormatService;
        this.frontViewService = frontViewService;
        this.viewDate = 'dd/MM/yyyy HH:mm';
        this.activityChange = new Subject();
        this.commentChange = new Subject();
        this._removeTags = [
            'link',
            'meta',
            'script',
            'style',
            'open-graph'
        ];
        // Receive activity notify real time
        this._socket = this.socketService.onMessage.subscribe((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            /** @type {?} */
            var app = message.application;
            /** @type {?} */
            var activityID = message.message && message.message['msgID'];
            if (app === APP_ACTIVITY) {
                _this.activityChange.next(activityID);
            }
            else if (app === APP_COMMENT) {
                _this.commentChange.next(activityID);
            }
        }));
    }
    /**
     * @return {?}
     */
    SocialService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._socket.unsubscribe();
    };
    /**
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.listGroups = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        return this.getJSON('social/member/category', null, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            callback(data && data.elements || [], err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationReceiveList = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/invitation', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationSendList = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/invitation/mylist', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationRejectList = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/invitation/rejectedlist', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationCreate = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postJSON('social/invitation/create', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationDelete = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postJSON('social/invitation/delete', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationReject = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postJSON('social/invitation/rejected', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationAccept = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postJSON('social/invitation/accepted', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.invitationSearch = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postJSON('social/member/search', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.listFriends = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/member/list', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            /** @type {?} */
            var members = data && data['items'] || [];
            members.total = data && data.total || 0;
            callback(members, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.groupMemberCreate = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postJSON('social/member/create', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.groupMemberDelete = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postJSON('social/member/delete', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.searchMember = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        this.coreService.searchFirmAccount(params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            /** @type {?} */
            var members = data && data['items'] || [];
            members.total = data && data.total || 0;
            callback(members, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.favoriteList = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.postJSON('social/activity/favorite', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            if (data && data['items']) {
                data['items'].forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    item._favorite = true;
                    _this.indexActivity(item);
                }));
            }
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityWall = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.getJSON('social/activity/wall', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            if (data) {
                if (data['items']) {
                    data['items'].forEach(_this.indexActivity.bind(_this));
                }
            }
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityLoad = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.getJSON('social/activity/load', params, (/**
         * @param {?} data
         * @param {?} error
         * @return {?}
         */
        function (data, error) {
            if (data) {
                _this.indexActivity(data);
            }
            callback(data, error);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityPost = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.postForm('social/activity/post', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            if (data) {
                _this.indexActivity(data);
            }
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityUpdate = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.postForm('social/activity/update', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            if (data) {
                _this.indexActivity(data);
            }
            callback(data, err);
        }));
    };
    /**
     *
     * @param params
     * {
     *    application: string
     *    contextID: string
     *    element: [{
     *      member: string
     *      display: string
     *      policy: "FRIEND"
     *    }]
     * }
     * @param callback
     *
     */
    /**
     *
     * @param {?} params
     * {
     *    application: string
     *    contextID: string
     *    element: [{
     *      member: string
     *      display: string
     *      policy: "FRIEND"
     *    }]
     * }
     * @param {?} callback
     *
     * @return {?}
     */
    SocialService.prototype.activityUpdateScope = /**
     *
     * @param {?} params
     * {
     *    application: string
     *    contextID: string
     *    element: [{
     *      member: string
     *      display: string
     *      policy: "FRIEND"
     *    }]
     * }
     * @param {?} callback
     *
     * @return {?}
     */
    function (params, callback) {
        return this.postForm('social/activity/scope', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityRemove = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/activity/delete', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityRemoveFile = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/activity/binaryremove', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityLikeList = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/activity/likelist', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityLike = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/activity/like', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.activityUnLike = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/activity/unlike', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.pollLoad = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/poll/load', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.pollResult = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/poll/result', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.pollVote = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/poll/vote', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.pollPost = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postForm('social/activity/pollpost', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.pollUpdate = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.postForm('social/poll/update', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.commentWall = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.getJSON('social/comment/wall', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            if (data && data['items']) {
                data['items'].forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return _this.indexActivity(item); }));
            }
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.commentPost = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.postForm('social/comment/post', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            if (data) {
                _this.indexActivity(data);
            }
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.commentUpdate = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        return this.postForm('social/comment/update', params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) {
            if (data) {
                _this.indexActivity(data);
            }
            callback(data, err);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.commentRemove = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/comment/delete', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.commentLike = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/comment/like', params, callback);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.commentUnLike = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.getJSON('social/comment/unlike', params, callback);
    };
    // Download / View file
    // Download / View file
    /**
     * @param {?} params
     * @return {?}
     */
    SocialService.prototype.activityDownLoadUrl = 
    // Download / View file
    /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return iNet.getPUrl('social/activity/download') + '?' + $.param(params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SocialService.prototype.activityViewUrl = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return iNet.getPUrl('social/activity/inlineview') + '?' + $.param(params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SocialService.prototype.activityViewOnlineUrl = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return iNet.getPUrl('unicorn/page/social/activity-viewer') + '?' + $.param(params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SocialService.prototype.commentDownLoadUrl = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return iNet.getPUrl('social/comment/download') + '?' + $.param(params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SocialService.prototype.commentViewUrl = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return iNet.getPUrl('social/comment/inlineview') + '?' + $.param(params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SocialService.prototype.commentViewOnlineUrl = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return iNet.getPUrl('unicorn/page/social/comment-viewer') + '?' + $.param(params);
    };
    /**
     * @param {?} application
     * @return {?}
     */
    SocialService.prototype.isCommentApp = /**
     * @param {?} application
     * @return {?}
     */
    function (application) {
        return application === APP_COMMENT;
    };
    /**
     * @param {?} application
     * @return {?}
     */
    SocialService.prototype.isActivityApp = /**
     * @param {?} application
     * @return {?}
     */
    function (application) {
        return application === APP_ACTIVITY;
    };
    /**
     * @param {?} attachment
     * @return {?}
     */
    SocialService.prototype.viewAttachment = /**
     * @param {?} attachment
     * @return {?}
     */
    function (attachment) {
        if ((iNet.hasOnlineEditor && attachment.pathViewOnline) || attachment.pathViewPdf) {
            this.frontViewService.viewInline(attachment.pathViewOnline || attachment.pathViewPdf);
        }
        else {
            window.open(attachment.pathDownLoad, '_blank');
        }
    };
    /**
     * @param {?} images
     * @param {?=} options
     * @return {?}
     */
    SocialService.prototype.viewImages = /**
     * @param {?} images
     * @param {?=} options
     * @return {?}
     */
    function (images, options) {
        this.photoSwipe.open(images, options);
    };
    /**
     * @param {?} attachments
     * @return {?}
     */
    SocialService.prototype.attachmentToFiles = /**
     * @param {?} attachments
     * @return {?}
     */
    function (attachments) {
        var _this = this;
        /** @type {?} */
        var files = [];
        attachments.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            files.push({
                name: item.file,
                type: item.mimetype,
                size: item.size,
                url: item.pathView,
                id: item.gridfsUUID,
                image: _this.isImage(item.mimetype)
            });
        }));
        return files;
    };
    /**
     * @private
     * @param {?} url
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.getJSON = /**
     * @private
     * @param {?} url
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (url, params, callback) {
        return this.httpClient.getJSON(iNet.getPUrl(url), params).subscribe(callback, (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            callback(null, err);
        }));
    };
    /**
     * @private
     * @param {?} url
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.postJSON = /**
     * @private
     * @param {?} url
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (url, params, callback) {
        return this.httpClient.postJSON(iNet.getPUrl(url), params).subscribe(callback, (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            callback(null, err);
        }));
    };
    /**
     * @private
     * @param {?} url
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SocialService.prototype.postForm = /**
     * @private
     * @param {?} url
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (url, params, callback) {
        return this.httpClient.post(iNet.getPUrl(url), this.buildFormData(params)).subscribe(callback, (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            callback(null, err);
        }));
    };
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    SocialService.prototype.buildFormData = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        if (!data) {
            return;
        }
        /** @type {?} */
        var form = new FormData();
        Object.keys(data).forEach((/**
         * @param {?} k
         * @return {?}
         */
        function (k) {
            if (k === 'openGraph') {
                return;
            }
            if (k === 'message') {
                /** @type {?} */
                var message = _this.secureMessageHtml(data[k]);
                // append open graph data on message
                if (data.openGraph) {
                    message += _this.createOpenGraphMessage(data.openGraph);
                }
                form.append(k, message);
            }
            else {
                form.append(k, data[k]);
            }
        }));
        return form;
    };
    /**
     * @private
     * @param {?} activity
     * @return {?}
     */
    SocialService.prototype.indexActivity = /**
     * @private
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        if (!activity || !activity.uuid) {
            return;
        }
        // Get full path
        if (activity.path) {
            activity.path = this.coreService.getSsoRedirect({
                url: activity.path,
                application: activity.application
            });
        }
        if (activity.mpath) {
            activity.mpath = this.coreService.getSsoRedirect({
                url: activity.mpath,
                application: activity.application
            });
        }
        activity.isCreator = activity.creator === iNet.username;
        activity._editable = activity.isCreator && !activity.path;
        activity._localNewsPin = activity.type === PINNEWS;
        this.formatMessageDisplay(activity);
        this._indexAttachments(activity);
        this.indexPolicies(activity);
    };
    /**
     * @private
     * @param {?} activity
     * @return {?}
     */
    SocialService.prototype._indexAttachments = /**
     * @private
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        var _this = this;
        activity.images = [];
        activity.files = [];
        activity.attachments.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            /** @type {?} */
            var params = {
                binary: item.uuid
            };
            /** @type {?} */
            var isComment = activity.hasOwnProperty('activityID');
            if (isComment) {
                params.activity = activity.activityID;
                params.comment = activity.uuid;
                item.pathView = _this.commentViewUrl(params);
                item.pathDownLoad = _this.commentDownLoadUrl(params);
            }
            else {
                params.activity = activity.uuid;
                item.pathView = _this.activityViewUrl(params);
                item.pathDownLoad = _this.activityDownLoadUrl(params);
            }
            if (_this.isImage(item.mimetype)) {
                activity.images.push(item);
            }
            else {
                if (_this._isSupportViewOnline(item.file)) {
                    if (isComment) {
                        item.pathViewOnline = _this.commentViewOnlineUrl(params);
                    }
                    else {
                        item.pathViewOnline = _this.activityViewOnlineUrl(params);
                    }
                }
                if (_this._isPdf(item.file)) {
                    item.pathViewPdf = item.pathView;
                }
                activity.files.push(item);
            }
        }));
    };
    /**
     * @private
     * @param {?} filename
     * @return {?}
     */
    SocialService.prototype._isSupportViewOnline = /**
     * @private
     * @param {?} filename
     * @return {?}
     */
    function (filename) {
        return this.fileFormatService.getEditFormats().indexOf(this._getExt(filename)) > -1;
    };
    /**
     * @private
     * @param {?} filename
     * @return {?}
     */
    SocialService.prototype._isPdf = /**
     * @private
     * @param {?} filename
     * @return {?}
     */
    function (filename) {
        return this._getExt(filename) === 'pdf';
    };
    /**
     * @private
     * @param {?} name
     * @return {?}
     */
    SocialService.prototype._getExt = /**
     * @private
     * @param {?} name
     * @return {?}
     */
    function (name) {
        /** @type {?} */
        var temp = name.split('.');
        if (temp.length > 1) {
            return temp[temp.length - 1].toLowerCase();
        }
        else {
            return '';
        }
    };
    /**
     * @private
     * @param {?} activity
     * @return {?}
     */
    SocialService.prototype.indexPolicies = /**
     * @private
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        var _this = this;
        activity._policies = [];
        if (activity.published) {
            // Publish
            activity._policies.push(SHARE_PUBLISH);
        }
        else {
            activity.policies.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.inherit) {
                    return;
                }
                /** @type {?} */
                var policy = {
                    member: item.category,
                    display: item.display || item.category,
                    policy: item.policy
                };
                activity._policies.push(policy);
            }));
            activity._policies = this.splitPolicies(activity._policies);
            // Add icon
            activity._policies.forEach((/**
             * @param {?} item
             * @param {?} index
             * @return {?}
             */
            function (item, index) {
                if (!item.hasOwnProperty('icon')) {
                    activity._policies[index] = {
                        id: item.member,
                        display: item.display,
                        icon: _this.getIconByType(item.policy),
                        value: [item]
                    };
                }
            }));
        }
        // Policies display
        if (activity._policies.length <= 4) {
            activity._policiesStr = '';
            activity._policies.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item === SHARE_ME) {
                    return;
                }
                activity._policiesStr += item.display + ', ';
            }));
            activity._policiesStr = activity._policiesStr.slice(0, -2);
        }
        else {
            activity._policiesStr = activity._policies[0].display;
            activity._policiesStr += ' và ' + (activity._policies.length - 2) + ' người khác';
        }
    };
    /**
     * @param {?} policy
     * @return {?}
     */
    SocialService.prototype.getIconByType = /**
     * @param {?} policy
     * @return {?}
     */
    function (policy) {
        switch (policy) {
            case Policy.Company:
                return PolicyIcon.Company;
            case Policy.Friend:
                return PolicyIcon.Friend;
            default:
                return '';
        }
    };
    /**
     * @private
     * @param {?} policies
     * @return {?}
     */
    SocialService.prototype.splitPolicies = /**
     * @private
     * @param {?} policies
     * @return {?}
     */
    function (policies) {
        if (policies.length > 0) {
            /** @type {?} */
            var allCompany = void 0;
            /** @type {?} */
            var allFriend = void 0;
            for (var i = 0; i < policies.length; i++) {
                /** @type {?} */
                var item = policies[i];
                if (this.isAllFriend(item)) {
                    allFriend = item;
                }
                else if (this.isAllCompany(item)) {
                    allCompany = item;
                }
                else if (this.isOnlyMe(item)) {
                    policies[i] = SHARE_ME;
                }
            }
            if (allFriend) {
                policies[policies.indexOf(allFriend)] = SHARE_FRIEND;
            }
            else if (allCompany) {
                policies[policies.indexOf(allCompany)] = SHARE_COMPANY;
            }
        }
        return policies;
    };
    /**
     * @private
     * @param {?} policy
     * @return {?}
     */
    SocialService.prototype.isAllCompany = /**
     * @private
     * @param {?} policy
     * @return {?}
     */
    function (policy) {
        return policy.display === Policy.Company &&
            policy.policy === Policy.Company;
    };
    /**
     * @private
     * @param {?} policy
     * @return {?}
     */
    SocialService.prototype.isAllFriend = /**
     * @private
     * @param {?} policy
     * @return {?}
     */
    function (policy) {
        return policy.member === PolicyValueAll &&
            policy.policy === Policy.Friend;
    };
    /**
     * @private
     * @param {?} policy
     * @return {?}
     */
    SocialService.prototype.isOnlyMe = /**
     * @private
     * @param {?} policy
     * @return {?}
     */
    function (policy) {
        return policy.member === iNet.username &&
            policy.policy === Policy.Individual;
    };
    /**
     * @private
     * @param {?} mimeType
     * @return {?}
     */
    SocialService.prototype.isImage = /**
     * @private
     * @param {?} mimeType
     * @return {?}
     */
    function (mimeType) {
        return mimeType.indexOf('image') > -1;
    };
    /**
     * Remove html tag on message
     */
    /**
     * Remove html tag on message
     * @private
     * @param {?} message
     * @return {?}
     */
    SocialService.prototype.secureMessageHtml = /**
     * Remove html tag on message
     * @private
     * @param {?} message
     * @return {?}
     */
    function (message) {
        /** @type {?} */
        var $el = $('<div>').html(message);
        $el.find(this._removeTags.join(',')).remove();
        return $el.text();
    };
    /**
     * Format message and open graph data to display
     */
    /**
     * Format message and open graph data to display
     * @private
     * @param {?} activity
     * @return {?}
     */
    SocialService.prototype.formatMessageDisplay = /**
     * Format message and open graph data to display
     * @private
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        // Get social data message
        /** @type {?} */
        var dataSocial = this.parseSocialDataMessage(activity);
        // Only get first item
        if (dataSocial[SocialTag.OpenGraph].length > 0) {
            activity._openGraph = dataSocial[SocialTag.OpenGraph][0];
        }
        if (dataSocial[SocialTag.LocalNews].length > 0) {
            activity._localNews = dataSocial[SocialTag.LocalNews][0];
            activity._localNews.url = activity.path;
        }
        // remove script and style
        /** @type {?} */
        var message = activity.message = this.secureMessageHtml(activity.message);
        // Beak line
        message = HtmlUtils.formatHtmlDisplay(message);
        // link
        message = message.replace(/(https?:\/\/[^\s]+)/g, (/**
         * @param {?} url
         * @return {?}
         */
        function (url) {
            return '<a href="' + url + '" target="_blank" rel="nofollow noopener">' + url + '</a>';
        }));
        activity._displayMessage = message;
    };
    /**
     * @param {?} openGraph
     * @return {?}
     */
    SocialService.prototype.createOpenGraphMessage = /**
     * @param {?} openGraph
     * @return {?}
     */
    function (openGraph) {
        /** @type {?} */
        var value = this.encodeBase64(JSON.stringify(this.convertLinks(openGraph)));
        return 'uri_message:link_preview:' + value;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SocialService.prototype.convertLinks = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var param = {};
        param['title'] = data['title'];
        param['image'] = data['image'];
        param['description'] = data['description'];
        param['url'] = data['url'];
        param['finalUrl'] = data['site_name'].toUpperCase();
        return param;
    };
    /**
     * @param {?} str
     * @return {?}
     */
    SocialService.prototype.encodeBase64 = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return window.btoa(unescape(encodeURIComponent(str)));
    };
    /**
     * @param {?} localNews
     * @return {?}
     */
    SocialService.prototype.createLocalNewsMessage = /**
     * @param {?} localNews
     * @return {?}
     */
    function (localNews) {
        return this._createTagMessage(SocialTag.LocalNews, localNews);
    };
    // parseSocialDataMessage(activity: SocialActivity | SocialComment): Object {
    //     let data = {};
    //     let $el = $('<div>').html(activity.message);
    //
    //     for (let tag in SocialTag) {
    //         let tagName = SocialTag[tag];
    //         data[tagName] = [];
    //         $el.find(tagName).each(function () {
    //             try {
    //                 data[tagName].push(JSON.parse(this.innerHTML));
    //             } catch (e) {
    //             }
    //
    //             // Remove tag after extract data
    //             this.remove();
    //         });
    //     }
    //
    //     // Apply message after extract data
    //     activity.message = $el.html();
    //
    //     return data;
    // }
    // parseSocialDataMessage(activity: SocialActivity | SocialComment): Object {
    //     let data = {};
    //     let $el = $('<div>').html(activity.message);
    //
    //     for (let tag in SocialTag) {
    //         let tagName = SocialTag[tag];
    //         data[tagName] = [];
    //         $el.find(tagName).each(function () {
    //             try {
    //                 data[tagName].push(JSON.parse(this.innerHTML));
    //             } catch (e) {
    //             }
    //
    //             // Remove tag after extract data
    //             this.remove();
    //         });
    //     }
    //
    //     // Apply message after extract data
    //     activity.message = $el.html();
    //
    //     return data;
    // }
    /**
     * @param {?} activity
     * @return {?}
     */
    SocialService.prototype.parseSocialDataMessage = 
    // parseSocialDataMessage(activity: SocialActivity | SocialComment): Object {
    //     let data = {};
    //     let $el = $('<div>').html(activity.message);
    //
    //     for (let tag in SocialTag) {
    //         let tagName = SocialTag[tag];
    //         data[tagName] = [];
    //         $el.find(tagName).each(function () {
    //             try {
    //                 data[tagName].push(JSON.parse(this.innerHTML));
    //             } catch (e) {
    //             }
    //
    //             // Remove tag after extract data
    //             this.remove();
    //         });
    //     }
    //
    //     // Apply message after extract data
    //     activity.message = $el.html();
    //
    //     return data;
    // }
    /**
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        /** @type {?} */
        var data = {};
        for (var tag in SocialTag) {
            /** @type {?} */
            var tagName = SocialTag[tag];
            data[tagName] = [];
        }
        /** @type {?} */
        var message = activity.message;
        /** @type {?} */
        var index = message.indexOf("uri_message");
        if (index > -1) {
            message = message.substring(index);
            /** @type {?} */
            var typeMessages = message.split(":");
            if (typeMessages.length === 3) {
                if (typeMessages[0] === 'uri_message' && typeMessages[1] === 'link_preview') {
                    try {
                        /** @type {?} */
                        var __messageLink = JSON.parse(this.decodeBase64(typeMessages[2]));
                        data[SocialTag.OpenGraph].push(__messageLink);
                    }
                    catch (e) {
                    }
                }
            }
            // Apply message after extract data
            activity.message = activity.message.substring(0, index);
        }
        return data;
    };
    /**
     * @param {?} str
     * @return {?}
     */
    SocialService.prototype.decodeBase64 = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return decodeURIComponent(escape(window.atob(str)));
    };
    /**
     * @private
     * @param {?} tagName
     * @param {?} data
     * @return {?}
     */
    SocialService.prototype._createTagMessage = /**
     * @private
     * @param {?} tagName
     * @param {?} data
     * @return {?}
     */
    function (tagName, data) {
        /** @type {?} */
        var tag = document.createElement(tagName);
        tag.innerText = JSON.stringify(data);
        tag.style.display = 'none';
        tag.hidden = true;
        return tag.outerHTML;
    };
    SocialService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SocialService.ctorParameters = function () { return [
        { type: HttpClientService },
        { type: CoreService },
        { type: PhotoSwipe },
        { type: WebSocketClientService },
        { type: FileFormatService },
        { type: FrontViewService }
    ]; };
    return SocialService;
}());
export { SocialService };
if (false) {
    /** @type {?} */
    SocialService.prototype.viewDate;
    /** @type {?} */
    SocialService.prototype.activityChange;
    /** @type {?} */
    SocialService.prototype.commentChange;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype._removeTags;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype._socket;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype.httpClient;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype.photoSwipe;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype.socketService;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype.fileFormatService;
    /**
     * @type {?}
     * @private
     */
    SocialService.prototype.frontViewService;
}
/**
 * @record
 */
export function FavoriteParams() { }
if (false) {
    /** @type {?} */
    FavoriteParams.prototype.activity;
    /** @type {?|undefined} */
    FavoriteParams.prototype.remove;
}
/**
 * @record
 */
export function BinaryParams() { }
if (false) {
    /** @type {?} */
    BinaryParams.prototype.binary;
    /** @type {?|undefined} */
    BinaryParams.prototype.activity;
    /** @type {?|undefined} */
    BinaryParams.prototype.comment;
}
