// inet-ui/esm5/src/notify/notify.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ViewChild } from '@angular/core';
import { HttpClientService } from "inet-core";
import { ToastrService } from "./toastr/toastr.service";
import { ToastContainerDirective } from "./toastr/toast.directive";
var NotifyService = /** @class */ (function () {
    function NotifyService(http, toastService) {
        this.http = http;
        this.toastService = toastService;
        this.url = {
            list: iNet.getPUrl('social/activity/notify')
        };
        this.toastService.overlayContainer = this.toastContainer;
    }
    /**
     * @return {?}
     */
    NotifyService.prototype.count = /**
     * @return {?}
     */
    function () {
        return this.http.getJSON(this.url.list);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    NotifyService.prototype.getMessage = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return this.http.getJSON(this.url.list, params);
    };
    /**
     * @param {?} msg
     * @param {?=} type
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyService.prototype.showMessage = /**
     * @param {?} msg
     * @param {?=} type
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (msg, type, title, config) {
        if (title === void 0) { title = 'Thông báo'; }
        if (config === void 0) { config = {}; }
        switch (type) {
            case 'info':
                this.toastService.info(msg, title, config);
                break;
            case 'error':
                this.toastService.error(msg, title, config);
                break;
            case 'warning':
                this.toastService.warning(msg, title, config);
                break;
            case 'success':
            default:
                this.toastService.success(msg, title, config);
        }
    };
    /**
     * @return {?}
     */
    NotifyService.prototype.getNotifyService = /**
     * @return {?}
     */
    function () {
        return this.toastService;
    };
    NotifyService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    NotifyService.ctorParameters = function () { return [
        { type: HttpClientService },
        { type: ToastrService }
    ]; };
    NotifyService.propDecorators = {
        toastContainer: [{ type: ViewChild, args: [ToastContainerDirective,] }]
    };
    return NotifyService;
}());
export { NotifyService };
if (false) {
    /** @type {?} */
    NotifyService.prototype.toastContainer;
    /**
     * @type {?}
     * @private
     */
    NotifyService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    NotifyService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    NotifyService.prototype.toastService;
}
