// inet-social/esm2015/src/comment/comment.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { SocialComment } from '../model/Comment';
import { SocialService } from '../social.service';
export class CommentComponent {
    /**
     * @param {?} socialService
     */
    constructor(socialService) {
        this.socialService = socialService;
        this.editable = true;
        this.onUpdated = new EventEmitter();
        this.onDeleted = new EventEmitter();
        this.viewDate = this.socialService.viewDate;
    }
    /**
     * @return {?}
     */
    commentCancelEdit() {
        this.editing = false;
    }
    /**
     * @return {?}
     */
    commentEdit() {
        this.editing = true;
    }
    /**
     * @param {?} params
     * @return {?}
     */
    commentUpdate(params) {
        this.socialService.commentUpdate(params, (/**
         * @param {?} comment
         * @return {?}
         */
        (comment) => {
            if (comment && comment.uuid) {
                this.commentCancelEdit();
                this.onUpdated.emit(comment);
            }
        }));
    }
    /**
     * @return {?}
     */
    commentDelete() {
        this.socialService.commentRemove({
            comment: this.comment.uuid,
            activity: this.comment.activityID
        }, (/**
         * @param {?} comment
         * @return {?}
         */
        (comment) => {
            if (comment && comment.uuid) {
                this.onDeleted.emit(comment);
            }
        }));
    }
    /**
     * @return {?}
     */
    commentLike() {
        /** @type {?} */
        const params = {
            activity: this.comment.activityID,
            comment: this.comment.uuid,
        };
        this.comment.likeStatus = !this.comment.likeStatus;
        if (this.comment.likeStatus) {
            this.comment.likeCount++;
            this.socialService.commentLike(params, (/**
             * @param {?} data
             * @return {?}
             */
            data => { }));
        }
        else {
            this.comment.likeCount--;
            this.socialService.commentUnLike(params, (/**
             * @param {?} data
             * @return {?}
             */
            data => { }));
        }
    }
}
CommentComponent.decorators = [
    { type: Component, args: [{
                selector: '[socialComment]',
                template: "<div class=\"social-comment\">\n\n    <img userAvatar [usercode]=\"comment.creator\" class=\"social-comment__avatar\">\n\n    <div *ngIf=\"!editing; else commentPost;\">\n        <div actionMore dropdown\n             *ngIf=\"comment._editable\"\n             [data]=\"comment\"\n             (onEdit)=\"commentEdit()\"\n             (onDelete)=\"commentDelete()\" class=\"social-action-more\"></div>\n        <div class=\"social-comment__content\">\n            <b class=\"social-comment__name\">{{ comment.fullname }} </b>\n            <span [innerHTML]=\"comment._displayMessage\"></span>\n            <div *ngIf=\"comment.attachments.length\" attachmentList [attachments]=\"comment.attachments\" class=\"social-block\"></div>\n\n            <div class=\"social-comment__action\">\n            <span [title]=\"comment.posted | date:viewDate\">\n                {{comment.posted | dateShortcut}}\n            </span>\n                <button class=\"social-flat-btn\" (click)=\"commentLike()\" [ngClass]=\"{'active': comment.likeStatus}\">\n                    <i class=\"fa fa-thumbs-up\"></i>\n                    <span *ngIf=\"comment.likeCount\">{{comment.likeCount}}</span>\n                </button>\n            </div>\n        </div>\n    </div>\n\n    <ng-template #commentPost>\n        <div commentPost\n             [commentEdit]=\"comment\"\n             (onPost)=\"commentUpdate($event)\"\n             (onCancel)=\"commentCancelEdit()\" class=\"social-comment__content\"></div>\n    </ng-template>\n\n</div>\n",
                styles: [".social-block{display:block;margin-top:10px}.social-action-more{float:right;margin:-5px -10px 0 0}.social-flat-btn{line-height:26px;text-align:center;color:rgba(0,0,0,.5);margin-right:15px;display:inline-block;outline:0!important;border:0;padding:0 7px;background:0 0}.social-flat-btn i{line-height:inherit;font-size:16px}.social-flat-btn span{margin-left:5px}.social-flat-btn.active,.social-flat-btn:hover{color:#2067b0}.social-comment{padding:7px 0;color:#212121;font-size:13px}.social-comment__avatar{width:30px;height:30px;border-radius:50%;margin-right:10px;float:left}.social-comment__content{overflow:hidden;line-height:16px}.social-comment__name{color:#2067b0}.social-comment__action{color:#6c757d;font-size:12px}.social-comment__more{cursor:pointer;color:#2067b0;margin-left:40px;padding-bottom:10px}"]
            }] }
];
/** @nocollapse */
CommentComponent.ctorParameters = () => [
    { type: SocialService }
];
CommentComponent.propDecorators = {
    comment: [{ type: Input }],
    editable: [{ type: Input }],
    onUpdated: [{ type: Output }],
    onDeleted: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CommentComponent.prototype.comment;
    /** @type {?} */
    CommentComponent.prototype.editable;
    /** @type {?} */
    CommentComponent.prototype.onUpdated;
    /** @type {?} */
    CommentComponent.prototype.onDeleted;
    /** @type {?} */
    CommentComponent.prototype.viewDate;
    /** @type {?} */
    CommentComponent.prototype.editing;
    /**
     * @type {?}
     * @private
     */
    CommentComponent.prototype.socialService;
}
