// inet-calendar/esm2015/src/model/CalendarEvent.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HtmlUtils } from "inet-core";
import { CalendarUtils } from "../CalendarUtils";
export class CalendarEvent {
    /**
     * @param {?} event
     */
    constructor(event) {
        this.repeatText = '';
        this.timeDisplay = '';
        this.event = event;
        this.subject = HtmlUtils.formatHtmlDisplay(event.subject);
        this.summary = HtmlUtils.formatHtmlDisplay(event.summary);
        if (event.members) {
            this.myAttendee = event.members.find((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                return item.role !== 'CREATOR' && item.username === iNet.username;
            }));
        }
        if (event.repeat) {
            this.repeatText = CalendarUtils.getRepeatText(event.repeat);
        }
        else {
            this.repeatText = '';
        }
        this.timeDisplay = CalendarUtils.displayDay(event.start);
        if (event.attribute._allday !== 'true') {
            this.timeDisplay += ` ⋅ ${CalendarUtils.displayTime(event.start)} - ${CalendarUtils.displayTime(event.end)}`;
        }
    }
}
if (false) {
    /** @type {?} */
    CalendarEvent.prototype.subject;
    /** @type {?} */
    CalendarEvent.prototype.summary;
    /** @type {?} */
    CalendarEvent.prototype.repeatText;
    /** @type {?} */
    CalendarEvent.prototype.timeDisplay;
    /** @type {?} */
    CalendarEvent.prototype.event;
    /** @type {?} */
    CalendarEvent.prototype.myAttendee;
}
