// inet-ui/esm2015/src/viewer/error-page/error-page.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { Location } from '@angular/common';
import { HttpClientService } from "inet-core";
import { ViewerService } from "../../file/viewer.service";
export class ErrorPageComponent {
    /**
     * @param {?} route
     * @param {?} http
     * @param {?} viewerService
     * @param {?} location
     */
    constructor(route, http, viewerService, location) {
        this.route = route;
        this.http = http;
        this.viewerService = viewerService;
        this.location = location;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.sub = this.route.params.subscribe((/**
         * @param {?} params
         * @return {?}
         */
        params => {
            this.docId = params.id;
        }));
    }
    /**
     * @return {?}
     */
    download() {
        this.viewerService.downloadById(this.docId);
    }
    /**
     * @return {?}
     */
    back() {
        this.location.back();
    }
}
ErrorPageComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-error-page',
                template: "<div class=\"error-page\">\n  <div class=\"error-content text-center\">\n    <div>\n      <div class=\"error-subtitle\" [innerHTML]=\"'COMMON.MODULE.VIEWER.UNSUPPORTED' | translate\"></div>\n      <button class=\"btn btn-primary\" (click)=\"download()\"> <i class=\"fa fa-download\"></i> {{'TOOLBAR.DOWNLOAD' | translate}}</button>\n      <button class=\"btn btn-danger ml-1\" (click)=\"back()\"> <i class=\"fa fa-remove\"></i> {{'TOOLBAR.CLOSE' | translate}}</button>\n    </div>\n  </div>\n</div>",
                styles: [".error-page{position:relative}.error-content{padding:40px 1rem}.error-subtitle{font-size:20px;font-weight:300;padding-bottom:10px}"]
            }] }
];
/** @nocollapse */
ErrorPageComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: HttpClientService },
    { type: ViewerService },
    { type: Location }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.sub;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.docId;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.viewerService;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.location;
}
