// inet-social/esm2015/src/embed/social-embed.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { SocialActivity } from "../model/Activity";
import { SocialService } from "../social.service";
export class SocialEmbedComponent {
    /**
     * @param {?} socialService
     */
    constructor(socialService) {
        this.socialService = socialService;
        this.allowComment = true;
        this.onLoad = new EventEmitter();
        this.onError = new EventEmitter();
        this.totalComments = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.activity) {
            this._loaded();
        }
        else {
            this._loadActivity();
        }
    }
    /**
     * @private
     * @return {?}
     */
    _loadActivity() {
        /** @type {?} */
        let params;
        if (this.activityID) {
            params = {
                activity: this.activityID
            };
        }
        else if (this.contextID && this.application) {
            params = {
                contextID: this.contextID,
                application: this.application
            };
        }
        if (params) {
            this.socialService.activityLoad(params, (/**
             * @param {?} activity
             * @param {?} err
             * @return {?}
             */
            (activity, err) => {
                if (err) {
                    this.onError.emit(err);
                }
                else {
                    if (activity.errors) {
                        this.onError.emit(activity);
                    }
                    else {
                        this.activity = activity;
                        this._loaded();
                    }
                }
            }));
        }
        else {
            this.onError.emit({
                message: 'NOT_FOUND'
            });
        }
    }
    /**
     * @private
     * @return {?}
     */
    _loaded() {
        this.onLoad.emit(this.activity);
    }
}
SocialEmbedComponent.decorators = [
    { type: Component, args: [{
                selector: 'social-discussion',
                template: "<div *ngIf=\"activity\" socialCommentList\n     [activity]=\"activity\"\n     [allowComment]=\"allowComment\"\n     (commentLoad)=\"totalComments = $event.total\"></div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
SocialEmbedComponent.ctorParameters = () => [
    { type: SocialService }
];
SocialEmbedComponent.propDecorators = {
    activityID: [{ type: Input }],
    contextID: [{ type: Input }],
    application: [{ type: Input }],
    activity: [{ type: Input }],
    allowComment: [{ type: Input }],
    onLoad: [{ type: Output }],
    onError: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    SocialEmbedComponent.prototype.activityID;
    /** @type {?} */
    SocialEmbedComponent.prototype.contextID;
    /** @type {?} */
    SocialEmbedComponent.prototype.application;
    /** @type {?} */
    SocialEmbedComponent.prototype.activity;
    /** @type {?} */
    SocialEmbedComponent.prototype.allowComment;
    /** @type {?} */
    SocialEmbedComponent.prototype.onLoad;
    /** @type {?} */
    SocialEmbedComponent.prototype.onError;
    /** @type {?} */
    SocialEmbedComponent.prototype.totalComments;
    /**
     * @type {?}
     * @private
     */
    SocialEmbedComponent.prototype.socialService;
}
