// inet-calendar/esm5/src/calendar-viewer/calendar-viewer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import * as moment_ from 'moment';
/** @type {?} */
var moment = moment_;
var CalendarViewerComponent = /** @class */ (function () {
    function CalendarViewerComponent() {
        this.dateFormat = 'MMMM/YYYY';
        this.locale = 'vi';
        this.firstDay = 1;
        this.limitEvent = 3;
        this.view = CalendarViewerMode.MONTH;
        this.onChange = new EventEmitter();
        this.onActive = new EventEmitter();
        this.dates = [];
        this.weekDates = [];
    }
    Object.defineProperty(CalendarViewerComponent.prototype, "date", {
        get: /**
         * @return {?}
         */
        function () {
            return new Date(this._date);
        },
        set: /**
         * @param {?} date
         * @return {?}
         */
        function (date) {
            /** @type {?} */
            var _date;
            if (!date || !date.getTime()) {
                _date = new Date();
            }
            else {
                _date = new Date(date);
            }
            _date.setHours(0, 0, 0, 0);
            if (this.view === CalendarViewerMode.MONTH) {
                // Set date start month
                _date.setDate(1);
            }
            // current date view
            if (this._date && _date.getTime() === this._date.getTime()) {
                return;
            }
            this._date = _date;
            this._render();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    CalendarViewerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this._date) {
            this.date = new Date();
        }
    };
    /**
     * @param {?} events
     * @param {?} date
     * @return {?}
     */
    CalendarViewerComponent.prototype.setEvents = /**
     * @param {?} events
     * @param {?} date
     * @return {?}
     */
    function (events, date) {
        /** @type {?} */
        var dateViewer = this._getDateViewer(date);
        if (dateViewer) {
            dateViewer.events = events;
        }
    };
    /**
     * @return {?}
     */
    CalendarViewerComponent.prototype.clearEvents = /**
     * @return {?}
     */
    function () {
        this.dates.forEach((/**
         * @param {?} dateViewer
         * @return {?}
         */
        function (dateViewer) { return dateViewer.events = null; }));
    };
    /**
     * @return {?}
     */
    CalendarViewerComponent.prototype.toDay = /**
     * @return {?}
     */
    function () {
        this.date = new Date();
    };
    /**
     * @return {?}
     */
    CalendarViewerComponent.prototype.prev = /**
     * @return {?}
     */
    function () {
        if (this.view === CalendarViewerMode.MONTH) {
            this._prevMonth();
        }
        else {
            this._prevWeek();
        }
        this._render();
    };
    /**
     * @return {?}
     */
    CalendarViewerComponent.prototype.next = /**
     * @return {?}
     */
    function () {
        if (this.view === CalendarViewerMode.MONTH) {
            this._nextMonth();
        }
        else {
            this._nextWeek();
        }
        this._render();
    };
    /**
     * @param {?} dateViewer
     * @return {?}
     */
    CalendarViewerComponent.prototype.activeDate = /**
     * @param {?} dateViewer
     * @return {?}
     */
    function (dateViewer) {
        this.dateActive = dateViewer;
        // Move to target month
        if (dateViewer.outOfRange) {
            this.date = dateViewer.date;
        }
        this.onActive.emit(dateViewer);
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalendarViewerComponent.prototype.focusDate = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var dateViewer = this._getDateViewer(date);
        if (dateViewer) {
            this.activeDate(dateViewer);
        }
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._nextMonth = /**
     * @private
     * @return {?}
     */
    function () {
        this._date.setMonth(this._date.getMonth() + 1);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._prevMonth = /**
     * @private
     * @return {?}
     */
    function () {
        this._date.setMonth(this._date.getMonth() - 1);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._nextWeek = /**
     * @private
     * @return {?}
     */
    function () {
        this._date.setDate(this._date.getDate() + 7);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._prevWeek = /**
     * @private
     * @return {?}
     */
    function () {
        this._date.setDate(this._date.getDate() - 7);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._render = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.view === CalendarViewerMode.MONTH) {
            this._renderMonth();
        }
        else {
            this._renderWeek();
        }
        this.onChange.emit(this.date);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._renderMonth = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var date = this._getDateStartWeek();
        this.dates = [];
        this.weekDates = [];
        while (this.weekDates.length <= 6) {
            this.weekDates.push(this._createWeekDate(date));
            // detect date is end month
            if (date.getMonth() !== this._date.getMonth()) {
                break;
            }
        }
        this._renderTitle();
        this._renderWeekTitle(this.weekDates[0]);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._renderWeek = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var date = this._getDateStartWeek();
        this.dates = [];
        this.weekDates = [this._createWeekDate(date)];
        this.weekDates[0].forEach((/**
         * @param {?} date
         * @return {?}
         */
        function (date) { return date.outOfRange = false; }));
        this._renderTitle();
        this._renderWeekTitle(this.weekDates[0]);
    };
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    CalendarViewerComponent.prototype._createWeekDate = /**
     * @private
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var weekDates = [];
        for (var i = 0; i < 7; i++) {
            /** @type {?} */
            var dateViewer = {
                date: new Date(date),
                time: date.getTime(),
                outOfRange: this._date.getMonth() !== date.getMonth(),
                today: this._isToday(date)
            };
            weekDates.push(dateViewer);
            this.dates.push(dateViewer);
            date.setDate(date.getDate() + 1);
        }
        return weekDates;
    };
    /**
     * @private
     * @param {?} weekDates
     * @return {?}
     */
    CalendarViewerComponent.prototype._renderWeekTitle = /**
     * @private
     * @param {?} weekDates
     * @return {?}
     */
    function (weekDates) {
        var _this = this;
        this.weekTitles = [];
        weekDates.forEach((/**
         * @param {?} viewerDate
         * @return {?}
         */
        function (viewerDate) {
            _this.weekTitles.push(moment(viewerDate.date).locale(_this.locale).format('dd'));
        }));
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._renderTitle = /**
     * @private
     * @return {?}
     */
    function () {
        this.dateTitle = moment(this._date).locale(this.locale).format(this.dateFormat);
    };
    /**
     * @private
     * @return {?}
     */
    CalendarViewerComponent.prototype._getDateStartWeek = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var date = new Date(this._date);
        /** @type {?} */
        var startWeekFromNow = date.getDay() - this.firstDay;
        if (date.getDay() < this.firstDay) {
            startWeekFromNow += 7;
        }
        // Set date start week
        date.setDate(date.getDate() - startWeekFromNow);
        return date;
    };
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    CalendarViewerComponent.prototype._isToday = /**
     * @private
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var now = new Date();
        return now.getDate() === date.getDate() && now.getMonth() === date.getMonth() && now.getFullYear() === date.getFullYear();
    };
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    CalendarViewerComponent.prototype._getDateViewer = /**
     * @private
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var _date = new Date(date);
        _date.setHours(0, 0, 0, 0);
        for (var i = 0; i < this.dates.length; i++) {
            if (this.dates[i].time === _date.getTime()) {
                return this.dates[i];
            }
        }
    };
    CalendarViewerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'calendar-viewer',
                    template: "<div class=\"calendar-viewer\">\n\n  <div role=\"row\" class=\"calendar-viewer__head\">\n    <i (click)=\"next()\" class=\"calendar-viewer__button fa fa-chevron-right\"></i>\n    <i (click)=\"toDay()\" class=\"calendar-viewer__button fa fa-circle\" style=\"font-size:10px\"></i>\n    <i (click)=\"prev()\" class=\"calendar-viewer__button fa fa-chevron-left\"></i>\n    <div class=\"calendar-viewer__title\">{{dateTitle}}</div>\n  </div>\n\n  <div role=\"grid\" class=\"calendar-viewer__body\">\n    <div role=\"row\" class=\"calendar-viewer__row\">\n      <div class=\"calendar-viewer__th\" *ngFor=\"let title of weekTitles\">{{title}}</div>\n    </div>\n    <div *ngFor=\"let week of weekDates\" role=\"row\" class=\"calendar-viewer__row\">\n      <div *ngFor=\"let dateViewer of week\"\n           [ngClass]=\"{'light-cell': dateViewer.outOfRange, 'today-cell': dateViewer.today}\" class=\"calendar-viewer__td\">\n        <div (click)=\"activeDate(dateViewer)\" class=\"calendar-viewer__date\" [ngClass]=\"{'active': dateViewer.time === dateActive?.time}\">\n          <span>{{dateViewer.date.getDate()}}</span>\n          <div *ngIf=\"dateViewer.events\" class=\"calendar-viewer__events\">\n            <div *ngFor=\"let event of dateViewer.events\" [ngStyle]=\"event.style\"></div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n\n</div>\n",
                    styles: [".calendar-viewer{padding:10px 0}.calendar-viewer__head{font-size:14px;padding:0 5px 0 15px;height:40px;color:#757575}.calendar-viewer__title{overflow:hidden;line-height:40px;font-weight:700;text-transform:uppercase}.calendar-viewer__button{float:right;margin:2px 0;text-align:center}.calendar-viewer__body{display:table;table-layout:fixed;width:100%;text-align:center;color:#212121;font-size:14px}.calendar-viewer__row{display:table-row}.calendar-viewer__th{display:table-cell;vertical-align:middle;height:24px;font-size:12px;color:#757575;font-weight:700}.calendar-viewer__td{display:table-cell;vertical-align:middle;height:44px}.calendar-viewer__td.light-cell{opacity:.2}.calendar-viewer__td.today-cell .calendar-viewer__button,.calendar-viewer__td.today-cell .calendar-viewer__date{color:#dc3545}.calendar-viewer__button,.calendar-viewer__date{height:36px;width:36px;line-height:36px;border-radius:50%;cursor:pointer;display:inline-block;position:relative}.calendar-viewer__button:hover,.calendar-viewer__date:hover{background-color:rgba(0,0,0,.1)}.active.calendar-viewer__button,.calendar-viewer__date.active{background-color:#438eb9!important;color:#fff!important}.active.calendar-viewer__button .calendar-viewer__events div,.calendar-viewer__date.active .calendar-viewer__events div{background-color:#fff!important}.calendar-viewer__events{position:absolute;left:5px;right:5px;bottom:4px;text-align:center;overflow:hidden;line-height:0;height:6px}.calendar-viewer__events div{width:4px;height:4px;border-radius:50%;margin:0 1px;background-color:#438eb9;display:inline-block}"]
                }] }
    ];
    /** @nocollapse */
    CalendarViewerComponent.ctorParameters = function () { return []; };
    CalendarViewerComponent.propDecorators = {
        dateFormat: [{ type: Input }],
        locale: [{ type: Input }],
        firstDay: [{ type: Input }],
        limitEvent: [{ type: Input }],
        view: [{ type: Input }],
        date: [{ type: Input }],
        onChange: [{ type: Output }],
        onActive: [{ type: Output }]
    };
    return CalendarViewerComponent;
}());
export { CalendarViewerComponent };
if (false) {
    /** @type {?} */
    CalendarViewerComponent.prototype.dateFormat;
    /** @type {?} */
    CalendarViewerComponent.prototype.locale;
    /** @type {?} */
    CalendarViewerComponent.prototype.firstDay;
    /** @type {?} */
    CalendarViewerComponent.prototype.limitEvent;
    /** @type {?} */
    CalendarViewerComponent.prototype.view;
    /** @type {?} */
    CalendarViewerComponent.prototype.onChange;
    /** @type {?} */
    CalendarViewerComponent.prototype.onActive;
    /** @type {?} */
    CalendarViewerComponent.prototype.dateTitle;
    /** @type {?} */
    CalendarViewerComponent.prototype.dates;
    /** @type {?} */
    CalendarViewerComponent.prototype.weekDates;
    /** @type {?} */
    CalendarViewerComponent.prototype.weekTitles;
    /** @type {?} */
    CalendarViewerComponent.prototype.dateActive;
    /**
     * @type {?}
     * @private
     */
    CalendarViewerComponent.prototype._date;
}
/**
 * @record
 */
export function CalendarViewerDate() { }
if (false) {
    /** @type {?} */
    CalendarViewerDate.prototype.date;
    /** @type {?} */
    CalendarViewerDate.prototype.time;
    /** @type {?} */
    CalendarViewerDate.prototype.outOfRange;
    /** @type {?} */
    CalendarViewerDate.prototype.today;
    /** @type {?|undefined} */
    CalendarViewerDate.prototype.events;
}
/** @enum {string} */
var CalendarViewerMode = {
    MONTH: 'month',
    WEEK: 'week',
};
export { CalendarViewerMode };
