// inet-calendar/esm5/src/remote-module/time-picker/time-picker-modal/time-picker-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { TimepickerComponent } from "ngx-bootstrap";
var TimePickerModalComponent = /** @class */ (function () {
    function TimePickerModalComponent(el) {
        this.el = el;
        this.arrowkeys = true;
        this.hourStep = 1;
        this.minuteStep = 15;
        this.mousewheel = true;
        this.readonlyInput = false;
        this.secondsStep = 10;
        this.showMinutes = true;
        this.showSpinners = true;
        this.onChange = new EventEmitter();
    }
    Object.defineProperty(TimePickerModalComponent.prototype, "time", {
        get: /**
         * @return {?}
         */
        function () {
            return this._timeDate;
        },
        set: /**
         * @param {?} date
         * @return {?}
         */
        function (date) {
            this._timeDate = date;
            this.timeStr = TimePickerModalComponent._formatDate(date);
            this.onChange.emit(this.timeStr);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TimePickerModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.$modal = $(this.el.nativeElement).detach().addClass('modal');
        this.$modalContent = this.$modal.find('.time-picker-modal');
        // Touched close
        this.$modal.on('mousedown click', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if (e.target.isSameNode(_this.$modal[0])) {
                _this.close();
            }
        }));
    };
    /**
     * @return {?}
     */
    TimePickerModalComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.$modal.remove();
    };
    /**
     * @return {?}
     */
    TimePickerModalComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.$modal.detach();
    };
    /**
     * @param {?} css
     * @return {?}
     */
    TimePickerModalComponent.prototype.open = /**
     * @param {?} css
     * @return {?}
     */
    function (css) {
        this.$modalContent.css(css);
        document.body.appendChild(this.$modal[0]);
    };
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    TimePickerModalComponent._formatDate = /**
     * @private
     * @param {?} date
     * @return {?}
     */
    function (date) {
        if (!date || !date.getTime()) {
            return '';
        }
        else {
            return ('0' + date.getHours()).slice(-2) + ':' + ('0' + date.getMinutes()).slice(-2);
        }
    };
    TimePickerModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'time-picker-modal',
                    template: "<div class=\"time-picker-modal\">\n    <timepicker #timePicker [(ngModel)]=\"time\"\n                [arrowkeys]=\"arrowkeys\"\n                [max]=\"max\"\n                [min]=\"min\"\n                [minuteStep]=\"minuteStep\"\n                [mousewheel]=\"mousewheel\"\n                [readonlyInput]=\"readonlyInput\"\n                [secondsStep]=\"secondsStep\"\n                [showMeridian]=\"showMeridian\"\n                [showMinutes]=\"showMinutes\"\n                [showSeconds]=\"showSeconds\"\n                [showSpinners]=\"showSpinners\">\n\n    </timepicker>\n</div>",
                    styles: [":host(.modal){display:block;z-index:1101}.time-picker-modal{position:absolute;border:1px solid rgba(0,0,0,.2);background:#fff;border-radius:4px;padding:10px}"]
                }] }
    ];
    /** @nocollapse */
    TimePickerModalComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    TimePickerModalComponent.propDecorators = {
        arrowkeys: [{ type: Input }],
        hourStep: [{ type: Input }],
        max: [{ type: Input }],
        min: [{ type: Input }],
        minuteStep: [{ type: Input }],
        mousewheel: [{ type: Input }],
        readonlyInput: [{ type: Input }],
        secondsStep: [{ type: Input }],
        showMeridian: [{ type: Input }],
        showMinutes: [{ type: Input }],
        showSeconds: [{ type: Input }],
        showSpinners: [{ type: Input }],
        time: [{ type: Input }],
        onChange: [{ type: Output }],
        timePicker: [{ type: ViewChild, args: ['timePicker',] }]
    };
    return TimePickerModalComponent;
}());
export { TimePickerModalComponent };
if (false) {
    /** @type {?} */
    TimePickerModalComponent.prototype.arrowkeys;
    /** @type {?} */
    TimePickerModalComponent.prototype.hourStep;
    /** @type {?} */
    TimePickerModalComponent.prototype.max;
    /** @type {?} */
    TimePickerModalComponent.prototype.min;
    /** @type {?} */
    TimePickerModalComponent.prototype.minuteStep;
    /** @type {?} */
    TimePickerModalComponent.prototype.mousewheel;
    /** @type {?} */
    TimePickerModalComponent.prototype.readonlyInput;
    /** @type {?} */
    TimePickerModalComponent.prototype.secondsStep;
    /** @type {?} */
    TimePickerModalComponent.prototype.showMeridian;
    /** @type {?} */
    TimePickerModalComponent.prototype.showMinutes;
    /** @type {?} */
    TimePickerModalComponent.prototype.showSeconds;
    /** @type {?} */
    TimePickerModalComponent.prototype.showSpinners;
    /** @type {?} */
    TimePickerModalComponent.prototype.onChange;
    /** @type {?} */
    TimePickerModalComponent.prototype.timePicker;
    /** @type {?} */
    TimePickerModalComponent.prototype.timeStr;
    /** @type {?} */
    TimePickerModalComponent.prototype.$modal;
    /** @type {?} */
    TimePickerModalComponent.prototype.$modalContent;
    /**
     * @type {?}
     * @private
     */
    TimePickerModalComponent.prototype._timeDate;
    /** @type {?} */
    TimePickerModalComponent.prototype.el;
}
