// inet-ui/esm5/src/autocomplete/autocomplete-list.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef } from '@angular/core';
var AutocompleteListDirective = /** @class */ (function () {
    function AutocompleteListDirective(el) {
        this.el = el;
    }
    /**
     * @return {?}
     */
    AutocompleteListDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.$el = $(this.el.nativeElement);
    };
    /**
     * @return {?}
     */
    AutocompleteListDirective.prototype.open = /**
     * @return {?}
     */
    function () {
        this.$el.show();
    };
    /**
     * @return {?}
     */
    AutocompleteListDirective.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.$el.hide();
        this.$el.find('.active').removeClass('active');
    };
    /**
     * @param {?} e
     * @return {?}
     */
    AutocompleteListDirective.prototype.onInput = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.open();
        switch (e.which) {
            // arrow up
            case 38: /** @type {?} */
            var step = -1;
            // arrow down
            case 40:
                /** @type {?} */
                var $items = this.$el.children();
                if ($items.length < 1) {
                    return;
                }
                /** @type {?} */
                var $active = void 0;
                /** @type {?} */
                var index = -1;
                for (var i = 0; i < $items.length; i++) {
                    $active = $items.eq(i);
                    if ($active.hasClass('active')) {
                        $active.removeClass('active');
                        index = i;
                        break;
                    }
                }
                step = step || 1;
                // Active at first or last
                if ((index === 0 && step === -1) || (index === $items.length - 1 && step === 1)) {
                    return;
                }
                index += step;
                if (index < 0) {
                    index = $items.length - 1;
                }
                this.moveToActive($items.eq(index).addClass('active'));
                break;
            // enter
            case 13:
                this.$el.find('.active').trigger('click');
                break;
        }
    };
    // Todo: scroll to active item
    // Todo: scroll to active item
    /**
     * @private
     * @param {?} $active
     * @return {?}
     */
    AutocompleteListDirective.prototype.moveToActive = 
    // Todo: scroll to active item
    /**
     * @private
     * @param {?} $active
     * @return {?}
     */
    function ($active) {
        // let scrollHeight = this.el.nativeElement.scrollHeight;
        // let height = this.el.nativeElement.clientHeight;
        // if (scrollHeight <= height) {
        //   return;
        // }
        // let scrollTop = this.el.nativeElement.scrollTop;
        // let offset = $active.offset();
        //
        // console.log(scrollTop, scrollHeight, height);
        // console.log(offset);
        // let distanceTop = offset.top - scrollTop;
        // let distanceBottom = offset.top - scrollTop + height + $active.outerHeight();
        // if (distanceTop < 0) {
        //   this.el.nativeElement.scrollTop = scrollTop - distanceTop;
        // } else if(distanceBottom > 0){
        //   this.el.nativeElement.scrollTop = scrollTop - distanceBottom;
        // }
    };
    AutocompleteListDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[appAutocompleteList]',
                    exportAs: 'appAutocompleteList'
                },] }
    ];
    /** @nocollapse */
    AutocompleteListDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    return AutocompleteListDirective;
}());
export { AutocompleteListDirective };
if (false) {
    /** @type {?} */
    AutocompleteListDirective.prototype.$el;
    /**
     * @type {?}
     * @private
     */
    AutocompleteListDirective.prototype.el;
}
