// inet-social/esm5/src/social-friend/invitation/friend-invitation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { SocialService } from '../../social.service';
var FriendInvitationComponent = /** @class */ (function () {
    function FriendInvitationComponent(socialService) {
        this.socialService = socialService;
        // Default group
        this.category = 'Friends';
    }
    /**
     * @return {?}
     */
    FriendInvitationComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._loadReceiveInvite();
        this._loadSendInvite();
    };
    /**
     * @param {?} member
     * @return {?}
     */
    FriendInvitationComponent.prototype.acceptInvite = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        this.socialService.invitationAccept({
            member: member.inviter,
            category: member.category
        }, (/**
         * @param {?} response
         * @param {?} error
         * @return {?}
         */
        function (response, error) {
        }));
        this.invitations.splice(this.invitations.indexOf(member), 1);
        this.noInvitation = this.invitations.length < 1;
    };
    /**
     * @param {?} member
     * @return {?}
     */
    FriendInvitationComponent.prototype.rejectInvite = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        this.socialService.invitationReject({
            member: member.inviter,
            category: member.category
        }, (/**
         * @param {?} response
         * @param {?} error
         * @return {?}
         */
        function (response, error) {
        }));
        this.invitations.splice(this.invitations.indexOf(member), 1);
        this.noInvitation = this.invitations.length < 1;
    };
    /**
     * @param {?} member
     * @return {?}
     */
    FriendInvitationComponent.prototype.removeInvite = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        this.socialService.invitationDelete({
            member: member.usercode,
            category: member.category
        }, (/**
         * @param {?} response
         * @param {?} error
         * @return {?}
         */
        function (response, error) {
        }));
        this.invitationsSend.splice(this.invitationsSend.indexOf(member), 1);
        this.noInvitationSend = this.invitationsSend.length < 1;
    };
    /**
     * @private
     * @return {?}
     */
    FriendInvitationComponent.prototype._loadReceiveInvite = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.socialService.invitationReceiveList({ pageSize: -1 }, (/**
         * @param {?} respone
         * @param {?} error
         * @return {?}
         */
        function (respone, error) {
            _this.invitations = respone && respone.items || [];
            _this.noInvitation = _this.invitations.length < 1;
        }));
    };
    /**
     * @private
     * @return {?}
     */
    FriendInvitationComponent.prototype._loadSendInvite = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.socialService.invitationSendList({ pageSize: -1 }, (/**
         * @param {?} respone
         * @param {?} error
         * @return {?}
         */
        function (respone, error) {
            _this.invitationsSend = respone && respone.items || [];
            _this.noInvitationSend = _this.invitationsSend.length < 1;
        }));
    };
    FriendInvitationComponent.decorators = [
        { type: Component, args: [{
                    template: "<div class=\"grid-title\">\n  L\u1EDDi m\u1EDDi k\u1EBFt b\u1EA1n\n</div>\n<div class=\"grid-container\">\n  <div class=\"grid-item\" *ngFor=\"let invite of invitations\">\n      <img userAvatar [usercode]=\"invite.usercode\" class=\"grid-item__icon\">\n      <div class=\"grid-item__content\">\n        <div class=\"grid-content__ellipsis\">{{invite.fullname}}</div>\n          <div class=\"grid-content__brief\">{{invite.usercode}}</div>\n      </div>\n      <button (click)=\"acceptInvite(invite)\"\n              class=\"grid-item__btn btn btn-sm btn-primary\"\n              title=\"Th\u00EAm b\u1EA1n\">\n          <i class=\"fa fa-check\"></i>\n      </button>\n      <button (click)=\"rejectInvite(invite)\"\n              class=\"grid-item__btn btn btn-sm btn-danger\"\n              title=\"T\u1EEB ch\u1ED1i\">\n          <i class=\"fa fa-trash\"></i>\n      </button>\n  </div>\n  <div class=\"grid-item grid-item__sm\" *ngIf=\"noInvitation\">\n      <i class=\"grid-item__icon fa fa-bell\"></i>\n      <div class=\"grid-item__content\">Ch\u01B0a c\u00F3 l\u1EDDi m\u1EDDi k\u1EBFt b\u1EA1n m\u1EDBi</div>\n  </div>\n</div>\n\n<div class=\"grid-title\">\n  L\u1EDDi m\u1EDDi \u0111\u00E3 g\u1EEDi\n</div>\n<div class=\"grid-container\">\n  <div class=\"grid-item\" *ngFor=\"let invite of invitationsSend\">\n      <img userAvatar [usercode]=\"invite.usercode\" class=\"grid-item__icon\">\n      <div class=\"grid-item__content\">\n        <div class=\"grid-content__ellipsis\">{{invite.username}}</div>\n        <div class=\"grid-content__brief\">{{invite.usercode}}</div>\n      </div>\n      <button (click)=\"removeInvite(invite)\" class=\"grid-item__btn btn btn-sm btn-danger\">\n          <i class=\"fa fa-trash\"></i>\n      </button>\n  </div>\n  <div class=\"grid-item grid-item__sm\" *ngIf=\"noInvitationSend\">\n      <i class=\"grid-item__icon fa fa-bell\"></i>\n      <div class=\"grid-item__content\">Ch\u01B0a g\u1EEDi l\u1EDDi m\u1EDDi k\u1EBFt b\u1EA1n n\u00E0o</div>\n  </div>\n</div>\n",
                    styles: [".grid-title{color:rgba(0,0,0,.54);font-size:14px;margin:8px;font-weight:700}.grid-container{padding-bottom:20px;overflow:hidden}.grid-item{float:left;width:50%;padding:0 15px;margin-right:-1px;margin-bottom:-1px;border:1px solid rgba(0,0,0,.05);cursor:pointer;display:flex;height:90px;align-items:center}.grid-item__sm{height:80px}.grid-item__icon{width:60px;height:60px;line-height:60px;text-align:center;color:rgba(0,0,0,.54);border-radius:50%;font-size:26px;background:rgba(0,0,0,.05)}.grid-item__btn{line-height:20px;margin-left:5px}.grid-item__content{flex-grow:1;padding:0 10px 0 15px;line-height:20px;overflow:hidden;font-size:16px}.grid-content__brief,.grid-content__ellipsis{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.grid-content__brief{font-size:14px;color:rgba(0,0,0,.4)}@media (max-width:767px){.grid-item{width:100%}.grid-item__icon{width:40px;height:40px;line-height:40px}}", ""]
                }] }
    ];
    /** @nocollapse */
    FriendInvitationComponent.ctorParameters = function () { return [
        { type: SocialService }
    ]; };
    return FriendInvitationComponent;
}());
export { FriendInvitationComponent };
if (false) {
    /** @type {?} */
    FriendInvitationComponent.prototype.category;
    /** @type {?} */
    FriendInvitationComponent.prototype.invitations;
    /** @type {?} */
    FriendInvitationComponent.prototype.noInvitation;
    /** @type {?} */
    FriendInvitationComponent.prototype.invitationsSend;
    /** @type {?} */
    FriendInvitationComponent.prototype.noInvitationSend;
    /**
     * @type {?}
     * @private
     */
    FriendInvitationComponent.prototype.socialService;
}
