// inet-chat/esm5/src/component/message-panel/message-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ChatMessageService } from "../../message.service";
import { MessageChatComponent } from '../message-chat/message-chat.component';
var MessagePanelComponent = /** @class */ (function () {
    function MessagePanelComponent(messageService) {
        var _this = this;
        this.messageService = messageService;
        this.audioCall = true;
        this.videoCall = true;
        this.closable = false; // Support close button on small screen
        // Support close button on small screen
        this.unreadMessageChange = new EventEmitter();
        this.chatOpen = false;
        this.readyToChat = false;
        this._socketSub = this.messageService.socketService.onStateChange.subscribe((/**
         * @return {?}
         */
        function () { return _this._stateChange(); }));
        this._messageSub = this.messageService.unreadMessageChange.subscribe((/**
         * @param {?} count
         * @return {?}
         */
        function (count) { return _this.unreadMessageChange.emit(count); }));
        this._stateChange();
    }
    /**
     * @private
     * @return {?}
     */
    MessagePanelComponent.prototype._stateChange = /**
     * @private
     * @return {?}
     */
    function () {
        this.readyToChat = this.messageService.socketService.connected;
        if (this.messageService.socketService.connected) {
            this.stateMessage = '';
        }
        else if (this.messageService.socketService.closed) {
            this.stateMessage = 'Kết nối không thành công';
        }
        else {
            this.stateMessage = 'Đang kết nối...';
        }
    };
    /**
     * @return {?}
     */
    MessagePanelComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._socketSub.unsubscribe();
        this._messageSub.unsubscribe();
    };
    /**
     * @param {?} contact
     * @return {?}
     */
    MessagePanelComponent.prototype.openVideoCall = /**
     * @param {?} contact
     * @return {?}
     */
    function (contact) {
    };
    /**
     * @param {?} contact
     * @return {?}
     */
    MessagePanelComponent.prototype.openAudioCall = /**
     * @param {?} contact
     * @return {?}
     */
    function (contact) {
    };
    /**
     * @param {?} contact
     * @param {?=} call
     * @return {?}
     */
    MessagePanelComponent.prototype.openChatWith = /**
     * @param {?} contact
     * @param {?=} call
     * @return {?}
     */
    function (contact, call) {
        var _this = this;
        this.roomName = null;
        if (this.contact !== contact) {
            this.contact = contact;
            contact.unread = 0; // mark read
        }
        // Delay for chat
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.chatOpen = true;
            if (call) {
                _this.messageChatComponent.messageJoinCall['joined'] = false;
                _this.messageChatComponent.joinRoom(true, contact.usercode);
            }
        }), 100);
    };
    /**
     * @return {?}
     */
    MessagePanelComponent.prototype.closeChat = /**
     * @return {?}
     */
    function () {
        this.chatOpen = false;
    };
    /**
     * @param {?} roomName
     * @return {?}
     */
    MessagePanelComponent.prototype.openRoom = /**
     * @param {?} roomName
     * @return {?}
     */
    function (roomName) {
        this.contact = null;
        this.roomName = roomName;
    };
    /**
     * @param {?} contact
     * @return {?}
     */
    MessagePanelComponent.prototype.callToContactSelected = /**
     * @param {?} contact
     * @return {?}
     */
    function (contact) {
        this.openChatWith(contact, true);
    };
    MessagePanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'message-panel',
                    template: "<div *ngIf=\"!readyToChat\" class=\"message-panel__status\">{{stateMessage}}</div>\n<div class=\"message-panel__container\">\n  <message-contact-list class=\"message-panel__side\" (contactSelected)=\"openChatWith($event)\" (roomSelected)=\"openRoom($event)\" (callTo)=\"callToContactSelected($event)\"></message-contact-list>\n  <div class=\"message-panel__body\" [ngClass]=\"{'open': chatOpen}\">\n    <div *ngIf=\"!contact && !roomName\" class=\"message-panel__welcome text-primary\">\n      <div style=\"font-size: 151px;\"><i class=\"fa fa-comments\" aria-hidden=\"true\"></i>\n      </div>\n      <div>Ch\u00E0o m\u1EEBng \u0111\u1EBFn v\u1EDBi <b>Calista!</b></div>\n    </div>\n    <message-chat *ngIf=\"contact\"\n                  [username]=\"contact.usercode\"\n                  [audioCall]=\"audioCall\"\n                  [videoCall]=\"videoCall\"\n                  [closable]=\"closable\"\n                  (onAudioCall)=\"openAudioCall($event)\"\n                  (onVideoCall)=\"openVideoCall($event)\"\n                  (onClose)=\"closeChat()\"></message-chat>\n    <room-chat *ngIf=\"roomName\" [roomName]=\"roomName\"></room-chat>\n  </div>\n</div>\n",
                    styles: [":host{display:flex;flex-direction:column;width:100%;height:100%;overflow:hidden;position:relative}.message-panel__status{text-align:center;width:100%;background:rgba(0,0,0,.8);color:#fff;font-size:16px;line-height:50px}.message-panel__container{width:100%;height:100%;position:relative;flex-grow:1}.message-panel__side{float:left;width:300px;background:#f1f1f4;border-right:1px solid #dedede}.message-panel__body{overflow:hidden;height:100%;position:relative;background:#fff;border-left:0}.message-panel__welcome{text-align:center;font-size:30px;position:absolute;padding:0 20px;width:100%;top:30%;margin-top:-20px}@media (max-width:768px){.message-panel__side{width:100%}.message-panel__body{position:absolute;right:0;bottom:0;left:0;width:100%;height:400px;border-left:1px;display:none}.message-panel__body.open{display:block}}"]
                }] }
    ];
    /** @nocollapse */
    MessagePanelComponent.ctorParameters = function () { return [
        { type: ChatMessageService }
    ]; };
    MessagePanelComponent.propDecorators = {
        audioCall: [{ type: Input }],
        videoCall: [{ type: Input }],
        closable: [{ type: Input }],
        unreadMessageChange: [{ type: Output }],
        messageChatComponent: [{ type: ViewChild, args: [MessageChatComponent,] }]
    };
    return MessagePanelComponent;
}());
export { MessagePanelComponent };
if (false) {
    /** @type {?} */
    MessagePanelComponent.prototype.audioCall;
    /** @type {?} */
    MessagePanelComponent.prototype.videoCall;
    /** @type {?} */
    MessagePanelComponent.prototype.closable;
    /** @type {?} */
    MessagePanelComponent.prototype.unreadMessageChange;
    /** @type {?} */
    MessagePanelComponent.prototype.messageChatComponent;
    /** @type {?} */
    MessagePanelComponent.prototype.contact;
    /** @type {?} */
    MessagePanelComponent.prototype.chatOpen;
    /** @type {?} */
    MessagePanelComponent.prototype.readyToChat;
    /** @type {?} */
    MessagePanelComponent.prototype.stateMessage;
    /** @type {?} */
    MessagePanelComponent.prototype.roomName;
    /**
     * @type {?}
     * @private
     */
    MessagePanelComponent.prototype._socketSub;
    /**
     * @type {?}
     * @private
     */
    MessagePanelComponent.prototype._messageSub;
    /**
     * @type {?}
     * @private
     */
    MessagePanelComponent.prototype.messageService;
}
