// inet-ui/esm2015/src/grid/components/pagination/pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, forwardRef, Input, TemplateRef } from '@angular/core';
import { DataTable } from '../table/table.component';
export class DataTablePagination {
    /**
     * @param {?} dataTable
     */
    constructor(dataTable) {
        this.dataTable = dataTable;
        this.show_range = false;
        this.show_limit = false;
        this.show_input = false;
        this.show_numbers = true;
        this.show_column_selector = false;
        this.autoHide = false;
        this.columnSelectorOpen = false;
    }
    /**
     * @return {?}
     */
    _closeSelector() {
        this.columnSelectorOpen = false;
    }
    /**
     * @return {?}
     */
    pageBack() {
        this.dataTable.offset -= Math.min(this.dataTable.limit, this.dataTable.offset);
    }
    /**
     * @return {?}
     */
    pageForward() {
        this.dataTable.offset += this.dataTable.limit;
    }
    /**
     * @return {?}
     */
    pageFirst() {
        this.dataTable.offset = 0;
    }
    /**
     * @return {?}
     */
    pageLast() {
        this.dataTable.offset = (this.maxPage - 1) * this.dataTable.limit;
    }
    /**
     * @return {?}
     */
    get maxPage() {
        return Math.ceil(this.dataTable.itemCount / this.dataTable.limit);
    }
    /**
     * @return {?}
     */
    get limit() {
        return this.dataTable.limit;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set limit(value) {
        this.dataTable.limit = Number((/** @type {?} */ (value))); // TODO better way to handle that value of number <input> is string?
    }
    /**
     * @return {?}
     */
    get page() {
        return this.dataTable.page;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set page(value) {
        this.dataTable.page = Number((/** @type {?} */ (value)));
    }
    /**
     * @param {?} number
     * @param {?} page
     * @return {?}
     */
    createPageRange(number, page) {
        /** @type {?} */
        const displayedPage = 3;
        /** @type {?} */
        let items = [];
        if (number > 1) {
            /** @type {?} */
            let maxPage = number;
            /** @type {?} */
            let minPage = 1;
            if (page === 1 && maxPage >= displayedPage) {
                maxPage = 3;
            }
            else if (page > 1 && maxPage > page + 1) {
                minPage = page - 1;
                maxPage = page + 1;
            }
            else if (page > 2 && maxPage > page) {
                minPage = page - 1;
                maxPage = page + 1;
            }
            else if (page > 2 && maxPage === page) {
                minPage = page - 2;
                maxPage = page;
            }
            for (let i = minPage; i <= maxPage; i++) {
                items.push(i);
            }
        }
        return items;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onChange(value) {
        value = Number((/** @type {?} */ (value)));
        /** @type {?} */
        const minPage = 1;
        /** @type {?} */
        const maxPage = this.maxPage;
        if (value < minPage) {
            value = minPage;
        }
        else if (value > maxPage) {
            value = maxPage;
        }
        this.page = value;
        return value;
    }
}
DataTablePagination.decorators = [
    { type: Component, args: [{
                moduleId: 'grid-module',
                selector: 'data-table-pagination',
                template: "<div class=\"pagination-box\">\n    <div class=\"pagination-range\"  *ngIf=\"!show_range\">\n        <ng-template [ngTemplateOutlet]=\"basicSearch\"></ng-template>\n    </div>\n    <div class=\"pagination-range input-group-sm\" *ngIf=\"show_range\">\n        <div class=\"input-group-prepend\">\n            <span class=\"input-group-text\">\n        {{dataTable.translations.paginationRange}}:\n        <span class=\"ml-1\" [textContent]=\"dataTable.itemCount>0 ? dataTable.offset + 1 : 0\"></span>\n        -\n        <span [textContent]=\"[dataTable.offset + dataTable.limit , dataTable.itemCount] | min\"></span>\n        /\n        <span [textContent]=\"dataTable.itemCount\"></span>\n        </span>\n        </div>\n    </div>\n\n    <div class=\"pagination-controllers\">\n        <div class=\"pagination-limit\" *ngIf=\"show_limit\">\n            <div class=\"input-group input-group-sm\">\n                <div class=\"input-group-prepend\">\n                    <span class=\"input-group-text\">{{dataTable.translations.paginationLimit}}:</span>\n                </div>\n                <input #limitInput type=\"number\" class=\"form-control\" min=\"1\" step=\"1\" [disabled]=\"maxPage === 0\"\n                       [ngModel]=\"limit\" (blur)=\"limit = limitInput.value\"\n                       (keyup.enter)=\"limit = limitInput.value\" (keyup.esc)=\"limitInput.value = limit\"/>\n            </div>\n        </div>\n        <div class=\"pagination-pages mr-1\" *ngIf=\"!autoHide || (autoHide && (maxPage > 0))\">\n            <button [disabled]=\"dataTable.offset <= 0\" (click)=\"pageFirst()\"\n                    class=\"btn btn-sm btn-default pagination-firstpage\"\n                    [title]=\"dataTable.translations.firstText\">\n                <i class=\"fa fa-step-backward\"></i>\n            </button>\n            <button [disabled]=\"dataTable.offset <= 0\" (click)=\"pageBack()\"\n                    class=\"btn btn-sm btn-default pagination-prevpage ml-1\"\n                    [title]=\"dataTable.translations.prevText\">\n                <i class=\"fa fa-chevron-left\"></i>\n            </button>\n            <div class=\"pagination-page ml-1\" *ngIf=\"show_input\">\n                <div class=\"input-group input-group-sm\">\n                    <input #pageInput type=\"number\" class=\"form-control\" min=\"1\" step=\"1\" max=\"{{maxPage}}\"\n                           [ngModel]=\"page\" (blur)=\"pageInput.value = onChange(pageInput.value)\"\n                           (keyup.enter)=\"pageInput.value = onChange(pageInput.value)\" [disabled]=\"maxPage === 0\"\n                           (keyup.esc)=\"pageInput.value = page\"/>\n                    <div class=\"input-group-append\">\n                        <span class=\"input-group-text\"><span>/</span><span\n                                [textContent]=\"dataTable.lastPage\"></span></span>\n                    </div>\n                </div>\n            </div>\n            <div class=\"pagination-page\" *ngIf=\"show_numbers\">\n                <button *ngFor=\"let i of createPageRange(maxPage,page)\"\n                        [disabled]=\"i == page\"\n                        (click)=\"page = i\"\n                        class=\"btn btn-sm btn-default ml-1\">{{ i }}\n                </button>\n            </div>\n            <button [disabled]=\"(dataTable.offset + dataTable.limit) >= dataTable.itemCount\" (click)=\"pageForward()\"\n                    class=\"btn btn-sm btn-default pagination-nextpage ml-1\" [title]=\"dataTable.translations.nextText\">\n                <i class=\"fa fa-chevron-right\"></i>\n            </button>\n            <button [disabled]=\"(dataTable.offset + dataTable.limit) >= dataTable.itemCount\" (click)=\"pageLast()\"\n                    class=\"btn btn-sm btn-default pagination-lastpage ml-1\" [title]=\"dataTable.translations.lastText\">\n                <i class=\"fa fa-step-forward\"></i>\n            </button>\n            <button type=\"button\" class=\"btn btn-sm btn-default pagination-reload ml-1\"\n                    (click)=\"dataTable.reloadItems()\"\n                    [attr.aria-label]=\"dataTable.translations.headerReload\"\n                    [title]=\"dataTable.translations.headerReload\">\n                <i class=\"fa fa-refresh\"></i>\n            </button>\n            <button *ngIf=\"show_column_selector\" type=\"button\" class=\"btn btn btn-sm btn-default column-selector-button ml-1\"\n                    [class.active]=\"columnSelectorOpen\"\n                    (click)=\"columnSelectorOpen = !columnSelectorOpen; $event.stopPropagation()\"\n                    [attr.aria-label]=\"dataTable.translations.headerColumnSelector\"\n                    [title]=\"dataTable.translations.headerColumnSelector\">\n                <i class=\"fa fa-list\"></i>\n            </button>\n            <div *ngIf=\"show_column_selector\" class=\"column-selector-wrapper\" (click)=\"$event.stopPropagation()\">\n                <div *ngIf=\"columnSelectorOpen\" class=\"column-selector-box panel panel-default\">\n                    <div *ngIf=\"dataTable.expandableRows\" class=\"column-selector-fixed-column checkbox\">\n                        <label>\n                            <input type=\"checkbox\" id=\"chk-expand-column-selector\" [(ngModel)]=\"dataTable.expandColumnVisible\"/>\n                            <label class=\"mb-0 lbl\" for=\"chk-expand-column-selector\"> {{dataTable.translations.expandColumn}}</label>\n                        </label>\n                    </div>\n                    <div *ngIf=\"dataTable.indexColumn\" class=\"column-selector-fixed-column checkbox\">\n                        <label>\n                            <input type=\"checkbox\" id=\"chk-index-column-selector\" [(ngModel)]=\"dataTable.indexColumnVisible\"/>\n                            <label class=\"mb-0 lbl\" for=\"chk-index-column-selector\"> {{dataTable.translations.indexColumn}}</label>\n                        </label>\n                    </div>\n                    <div *ngIf=\"dataTable.selectColumn\" class=\"column-selector-fixed-column checkbox\">\n                        <label>\n                            <input type=\"checkbox\" id=\"chk-select-column-selector\" [(ngModel)]=\"dataTable.selectColumnVisible\"/>\n                            <label class=\"mb-0 lbl\" for=\"chk-select-column-selector\"> {{dataTable.translations.selectColumn}}</label>\n                        </label>\n                    </div>\n                    <div *ngFor=\"let column of dataTable.columns;\" class=\"column-selector-column checkbox\">\n                        <label>\n                            <input type=\"checkbox\" [id]=\"'chk-'+ column.property + '-selector'\" [(ngModel)]=\"column.visible\"/>\n                            <label class=\"mb-0 lbl\" [for]=\"'chk-'+ column.property + '-selector'\">\n                                &nbsp;<span [textContent]=\"column.header || (column.property)\"></span>\n                            </label>\n                        </label>\n                    </div>\n                </div>\n            </div>\n        </div>\n    </div>\n    <ng-template *ngIf=\"!show_range\" [ngTemplateOutlet]=\"advanceSearch\"></ng-template>\n</div>\n",
                host: {
                    '(document:click)': '_closeSelector()'
                },
                styles: [".pagination-box{position:relative;display:block;min-height:38px;background-color:#e9ecef;border:1px solid #ced4da;padding:2px}.pagination-range{margin-left:3px;display:inline-block}.pagination-controllers{float:right}.pagination-controllers input{min-width:60px;text-align:right;max-width:80px}.pagination-limit{margin-right:10px;display:inline-table;width:180px;float:left}.pagination-pages{display:inline-block}.pagination-page{display:inline-table}.pagination-box button{outline:0!important}.column-selector-button,.pagination-firstpage,.pagination-lastpage,.pagination-nextpage,.pagination-prevpage,.pagination-reload{vertical-align:top}.column-selector-wrapper{position:relative}.column-selector-box{box-shadow:0 0 10px #d3d3d3;width:150px;padding:10px;position:absolute;right:0;top:1px;z-index:1060;background:#fff}.column-selector-box .checkbox{margin-bottom:4px}.column-selector-fixed-column{font-style:italic}"]
            }] }
];
/** @nocollapse */
DataTablePagination.ctorParameters = () => [
    { type: DataTable, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => DataTable)),] }] }
];
DataTablePagination.propDecorators = {
    show_range: [{ type: Input }],
    show_limit: [{ type: Input }],
    show_input: [{ type: Input }],
    show_numbers: [{ type: Input }],
    show_column_selector: [{ type: Input }],
    autoHide: [{ type: Input }],
    basicSearch: [{ type: Input }],
    advanceSearch: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DataTablePagination.prototype.show_range;
    /** @type {?} */
    DataTablePagination.prototype.show_limit;
    /** @type {?} */
    DataTablePagination.prototype.show_input;
    /** @type {?} */
    DataTablePagination.prototype.show_numbers;
    /** @type {?} */
    DataTablePagination.prototype.show_column_selector;
    /** @type {?} */
    DataTablePagination.prototype.autoHide;
    /** @type {?} */
    DataTablePagination.prototype.basicSearch;
    /** @type {?} */
    DataTablePagination.prototype.advanceSearch;
    /** @type {?} */
    DataTablePagination.prototype.columnSelectorOpen;
    /** @type {?} */
    DataTablePagination.prototype.dataTable;
}
