// inet-chat/esm5/src/component/search-input/search-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var SearchInputComponent = /** @class */ (function () {
    function SearchInputComponent() {
        this.placeholder = '';
        this.keyword = '';
        this.onSearch = new EventEmitter();
        this.onPressed = new EventEmitter();
        this.onChanged = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SearchInputComponent.prototype.search = /**
     * @return {?}
     */
    function () {
        this.onSearch.emit(this.keyword);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SearchInputComponent.prototype.keyPressed = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.onPressed.emit(e);
        if (e.key === 'Enter') {
            this.search();
        }
        this.onChanged.emit(this.keyword);
    };
    /**
     * @return {?}
     */
    SearchInputComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.keyword = '';
        this.onChanged.emit(this.keyword);
    };
    SearchInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'search-input',
                    template: "<div class=\"search-input__container\">\n  <input [(ngModel)]=\"keyword\" (keyup)=\"keyPressed($event)\" [placeholder]=\"placeholder\" class=\"form-control form-control-sm search-input__text\">\n  <i (click)=\"search()\" class=\"fa fa-search search-input__icon\" style=\"left:0\"></i>\n  <i *ngIf=\"keyword.length\" (click)=\"clear()\" class=\"fa fa-times search-input__icon\" style=\"right:0\"></i>\n</div>\n",
                    styles: [":host{display:block}.search-input__container{position:relative}.search-input__text{padding-left:30px;padding-right:30px}.search-input__icon{position:absolute;top:0;line-height:30px;width:30px;height:30px;text-align:center;color:#1c9dea;cursor:pointer}.search-input__icon:hover{color:#1c9dea}"]
                }] }
    ];
    SearchInputComponent.propDecorators = {
        placeholder: [{ type: Input }],
        keyword: [{ type: Input }],
        onSearch: [{ type: Output }],
        onPressed: [{ type: Output }],
        onChanged: [{ type: Output }]
    };
    return SearchInputComponent;
}());
export { SearchInputComponent };
if (false) {
    /** @type {?} */
    SearchInputComponent.prototype.placeholder;
    /** @type {?} */
    SearchInputComponent.prototype.keyword;
    /** @type {?} */
    SearchInputComponent.prototype.onSearch;
    /** @type {?} */
    SearchInputComponent.prototype.onPressed;
    /** @type {?} */
    SearchInputComponent.prototype.onChanged;
}
