// inet-social/esm2015/src/comment-post/comment-post.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output, Input, ViewChild, ElementRef } from '@angular/core';
import { SocialComment } from '../model/Comment';
import { AutoSizeDirective } from 'inet-core';
import { SocialService } from "../social.service";
import { ProgressComponent } from "../progress/progress.component";
export class CommentPostComponent {
    /**
     * @param {?} socialService
     */
    constructor(socialService) {
        this.socialService = socialService;
        this.placeholder = 'Nội dung thảo luận..';
        this.onCancel = new EventEmitter();
        this.onPost = new EventEmitter();
        this.message = '';
        this.files = [];
        this.removeFiles = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.commentEdit) {
            this.editComment(this.commentEdit);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    enterPost(event) {
        event.preventDefault();
        this.doPost();
    }
    /**
     * @return {?}
     */
    doPost() {
        /** @type {?} */
        const params = this.getData();
        if (!params.message) {
            return this.focusMessage();
        }
        this.onPost.emit(params);
        this.indicator.show();
    }
    /**
     * @return {?}
     */
    doCancel() {
        this.onCancel.emit(this.commentEdit);
        if (this.commentEdit) {
            this.resetForm();
        }
    }
    /**
     * @return {?}
     */
    getData() {
        /** @type {?} */
        let params = {
            message: this.message
        };
        for (let i = 0; i < this.files.length; i++) {
            if (!this.files[i].id) {
                params['file-' + i] = this.files[i];
            }
        }
        if (this.commentEdit) {
            params.activity = this.commentEdit.activityID;
            params.comment = this.commentEdit.uuid;
            params.gridfsUUID = this._getRemoveFiles();
        }
        return params;
    }
    /**
     * @param {?} comment
     * @return {?}
     */
    editComment(comment) {
        this.resetForm();
        this.commentEdit = comment;
        this.files = this.socialService.attachmentToFiles(comment.attachments);
        this.message = comment.message;
        this.focusMessage();
        // Trigger input to adjust size
        this._resizeInput();
    }
    /**
     * @return {?}
     */
    resetForm() {
        // Clear data
        this.message = '';
        this.files.length = 0;
        this.removeFiles.length = 0;
        this._resizeInput();
        if (this.indicator) {
            this.indicator.hide();
        }
    }
    /**
     * @return {?}
     */
    focusMessage() {
        this.messageEl.nativeElement.focus();
    }
    /**
     * @private
     * @return {?}
     */
    _getRemoveFiles() {
        return this.removeFiles.filter((/**
         * @param {?} file
         * @return {?}
         */
        file => file.id)).map((/**
         * @param {?} file
         * @return {?}
         */
        file => file.id)).join(',');
    }
    /**
     * @private
     * @return {?}
     */
    _resizeInput() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.messageAutoSize.adjust();
        }), 0);
    }
}
CommentPostComponent.decorators = [
    { type: Component, args: [{
                selector: '[commentPost]',
                template: "<div class=\"comment-input\">\n    <textarea class=\"comment-textarea\"\n              rows=\"1\"\n              [placeholder]=\"placeholder\"\n              #messageEl\n              appAutoSize\n              (keydown.esc)=\"doCancel()\"\n              (keydown.enter)=\"enterPost($event)\"\n              [(ngModel)]=\"message\"></textarea>\n    <i class=\"comment-file__icon fa fa-image\">\n        <input #fileEl type=\"file\" multiple class=\"file-post\">\n    </i>\n</div>\n\n<file-list class=\"social-block\" [files]=\"files\" [fileEl]=\"fileEl\" (onRemove)=\"removeFiles.push($event)\"></file-list>\n\n<div *ngIf=\"commentEdit\">\n    <button class=\"btn btn-primary btn-sm mr-1\" (click)=\"doPost()\">C\u1EACP NH\u1EACT</button>\n    <button class=\"btn btn-light btn-sm active\" (click)=\"doCancel()\">HU\u1EF6</button>\n</div>\n\n<progress-indicator></progress-indicator>\n",
                styles: [".social-block{display:block;margin-top:10px}:host{position:relative}.comment-input{position:relative;background:#e3e5e5;border-radius:4px;padding:7px 10px 3px;margin-bottom:10px}.comment-textarea{border:0;padding:0 40px 0 0;line-height:16px;margin:0;outline:0!important;width:100%;background:0 0}.comment-file__icon{position:absolute;right:0;top:0;width:30px;height:30px;line-height:30px;text-align:center;cursor:pointer}.file-post{position:absolute;-webkit-appearance:none;-moz-appearance:none;appearance:none;top:0;right:0;bottom:0;left:0;width:100%;height:100%;opacity:0;z-index:100}"]
            }] }
];
/** @nocollapse */
CommentPostComponent.ctorParameters = () => [
    { type: SocialService }
];
CommentPostComponent.propDecorators = {
    placeholder: [{ type: Input }],
    commentEdit: [{ type: Input }],
    onCancel: [{ type: Output }],
    onPost: [{ type: Output }],
    messageEl: [{ type: ViewChild, args: ['messageEl',] }],
    messageAutoSize: [{ type: ViewChild, args: [AutoSizeDirective,] }],
    indicator: [{ type: ViewChild, args: [ProgressComponent,] }]
};
if (false) {
    /** @type {?} */
    CommentPostComponent.prototype.placeholder;
    /** @type {?} */
    CommentPostComponent.prototype.commentEdit;
    /** @type {?} */
    CommentPostComponent.prototype.onCancel;
    /** @type {?} */
    CommentPostComponent.prototype.onPost;
    /** @type {?} */
    CommentPostComponent.prototype.messageEl;
    /** @type {?} */
    CommentPostComponent.prototype.messageAutoSize;
    /** @type {?} */
    CommentPostComponent.prototype.indicator;
    /** @type {?} */
    CommentPostComponent.prototype.message;
    /** @type {?} */
    CommentPostComponent.prototype.files;
    /** @type {?} */
    CommentPostComponent.prototype.removeFiles;
    /**
     * @type {?}
     * @private
     */
    CommentPostComponent.prototype.socialService;
}
