// inet-calendar/esm2015/src/utils/common/AjaxAPI.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export let AjaxAPI = {
    /**
     * @param {?} url
     * @param {?=} domain
     * @param {?=} firmPrefix
     * @return {?}
     */
    getUrl(url, domain, firmPrefix) {
        if (url.indexOf('://') > -1)
            return url;
        domain = domain || iNet.getUrl('{0}');
        url = domain.replace('{0}', url);
        // Change firm prefix
        if (firmPrefix) {
            /** @type {?} */
            let urls = url.split('/');
            urls[4] = firmPrefix;
            url = urls.join('/');
        }
        return url;
    },
    // Get url by subfirm
    /**
     * @param {?} url
     * @param {?=} domain
     * @param {?=} firmPrefix
     * @return {?}
     */
    getPUrl(url, domain, firmPrefix) {
        return this.getUrl(url, domain, firmPrefix || iNet.firmPrefix);
    },
    // Send request
    /**
     * @param {?} options
     * @return {?}
     */
    sendRequest(options) {
        return $.ajax(options);
    },
    // Post ajax by FormData
    /**
     * @param {?} options
     * @return {?}
     */
    postForm(options) {
        options = $.extend({}, options, {
            type: 'post',
            cache: false,
            contentType: false,
            processData: false
        });
        // build form data
        if (options.data && options.data.toString().indexOf('FormData') < 0) {
            /** @type {?} */
            let form = new FormData();
            for (let k in options.data) {
                form.append(k, options.data[k]);
            }
            options.data = form;
        }
        return this.sendRequest(options);
    }
};
// Detect on mobile
if (window['eNet']) {
    AjaxAPI.sendRequest = window['eNet'].ajaxSecure;
}
