// inet-social/esm5/src/activity-action-more/activity-action-more.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var ActivityActionMoreComponent = /** @class */ (function () {
    function ActivityActionMoreComponent() {
        this.onEdit = new EventEmitter();
        this.onDelete = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ActivityActionMoreComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.data) {
            return;
        }
        this.isActivity = !this.data.activityID;
        this.hasAction = this.isActivity && this.data._editable;
    };
    ActivityActionMoreComponent.decorators = [
        { type: Component, args: [{
                    selector: '[actionMore]',
                    template: "<div class=\"action-more\">\n  <i class=\"fa fa-ellipsis-h\" dropdownToggle></i>\n  <ul *dropdownMenu class=\"dropdown-menu dropdown-menu-right\" role=\"menu\">\n    <li *ngIf=\"data._editable\" role=\"menuitem\" (click)=\"onEdit.emit($event)\">\n      <a class=\"dropdown-item\">\n        <i class=\"action-more-icon fa fa-edit\"></i>\n        <span> Ch\u1EC9nh s\u1EEDa</span>\n      </a>\n    </li>\n    <li *ngIf=\"data._editable\" role=\"menuitem\" (click)=\"onDelete.emit($event)\">\n      <a class=\"dropdown-item\" style=\"color: #dc3545\">\n        <i class=\"action-more-icon fa fa-trash\"></i>\n        <span> X\u00F3a</span>\n      </a>\n    </li>\n  </ul>\n</div>\n",
                    styles: [".action-more{cursor:pointer;color:#6c757d;position:relative;width:30px;height:30px}.action-more ul{margin-top:-20px;font-size:13px;border:1px solid #eef2f4;box-shadow:0 0 34px 0 rgba(63,66,87,.1)}.action-more>i{width:30px;line-height:30px;text-align:center}.action-more-icon{width:20px}"]
                }] }
    ];
    /** @nocollapse */
    ActivityActionMoreComponent.ctorParameters = function () { return []; };
    ActivityActionMoreComponent.propDecorators = {
        data: [{ type: Input }],
        onEdit: [{ type: Output }],
        onDelete: [{ type: Output }]
    };
    return ActivityActionMoreComponent;
}());
export { ActivityActionMoreComponent };
if (false) {
    /** @type {?} */
    ActivityActionMoreComponent.prototype.data;
    /** @type {?} */
    ActivityActionMoreComponent.prototype.onEdit;
    /** @type {?} */
    ActivityActionMoreComponent.prototype.onDelete;
    /** @type {?} */
    ActivityActionMoreComponent.prototype.hasAction;
    /** @type {?} */
    ActivityActionMoreComponent.prototype.isActivity;
}
