// inet-social/esm5/src/social-friend/browse/friend-browse.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { SocialService } from '../../social.service';
var FriendBrowseComponent = /** @class */ (function () {
    function FriendBrowseComponent(socialService) {
        this.socialService = socialService;
        this.category = 'All friends';
        this.members = [];
        this.params = {
            pageSize: 20,
            pageNumber: 0
        };
        this.placeholder = 'Tìm kiếm bạn bè...';
    }
    /**
     * @return {?}
     */
    FriendBrowseComponent.prototype.clearSearchValue = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FriendBrowseComponent.prototype.onSearch = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._searchMember(value);
    };
    /**
     * @return {?}
     */
    FriendBrowseComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._searchMember('');
    };
    /**
     * @param {?} friend
     * @return {?}
     */
    FriendBrowseComponent.prototype.inviteFriend = /**
     * @param {?} friend
     * @return {?}
     */
    function (friend) {
        friend.added = true;
        this.socialService.invitationCreate({
            member: friend.username,
            memberName: friend.fullname,
            type: 'FRIEND',
            category: this.category
        }, (/**
         * @param {?} response
         * @param {?} error
         * @return {?}
         */
        function (response, error) {
        }));
    };
    /**
     * @private
     * @param {?} keyword
     * @return {?}
     */
    FriendBrowseComponent.prototype._searchMember = /**
     * @private
     * @param {?} keyword
     * @return {?}
     */
    function (keyword) {
        var _this = this;
        if (this.params.keyword === keyword) {
            return;
        }
        this.params.keyword = keyword;
        this.params.pageNumber = 0;
        clearTimeout(this._timer);
        this._timer = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.socialService.invitationSearch(_this.params, (/**
             * @param {?} respone
             * @param {?} error
             * @return {?}
             */
            function (respone, error) {
                if (_this.params.keyword != keyword) {
                    return;
                }
                /** @type {?} */
                var members = respone || [];
                /** @type {?} */
                var userCodes = [];
                _this.members = [];
                members.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    if (userCodes.indexOf(item.username) < 0) {
                        userCodes.push(item.username);
                        if (!item.fullname) {
                            item.fullname = item.username;
                        }
                        _this.members.push(item);
                    }
                }));
            }));
        }), 500);
    };
    FriendBrowseComponent.decorators = [
        { type: Component, args: [{
                    template: "<div class=\"grid-title\">\n    Nh\u1EEFng ng\u01B0\u1EDDi b\u1EA1n c\u00F3 th\u1EC3 bi\u1EBFt\n</div>\n<div class=\"grid-container\">\n    <div class=\"grid-item\" *ngFor=\"let member of members\">\n        <img userAvatar [usercode]=\"member.username\" class=\"grid-item__icon\">\n        <div class=\"grid-item__content\">\n            <div class=\"grid-content__ellipsis\">{{member.fullname}}</div>\n            <div class=\"grid-content__brief\">{{member.username}}</div>\n        </div>\n        <button *ngIf=\"!member.added;else added;\" (click)=\"inviteFriend(member)\" class=\"grid-item__btn btn btn-sm btn-primary\" title=\"Th\u00EAm b\u1EA1n\">\n            <i class=\"fa fa-user-plus\"></i>\n        </button>\n    </div>\n    <ng-template #added>\n        <button class=\"grid-item__btn btn btn-sm btn-success\">\n            <i class=\"fa fa-check\"></i>\n        </button>\n    </ng-template>\n</div>",
                    styles: [".grid-title{color:rgba(0,0,0,.54);font-size:14px;margin:8px;font-weight:700}.grid-container{padding-bottom:20px;overflow:hidden}.grid-item{float:left;width:50%;padding:0 15px;margin-right:-1px;margin-bottom:-1px;border:1px solid rgba(0,0,0,.05);cursor:pointer;display:flex;height:90px;align-items:center}.grid-item__sm{height:80px}.grid-item__icon{width:60px;height:60px;line-height:60px;text-align:center;color:rgba(0,0,0,.54);border-radius:50%;font-size:26px;background:rgba(0,0,0,.05)}.grid-item__btn{line-height:20px;margin-left:5px}.grid-item__content{flex-grow:1;padding:0 10px 0 15px;line-height:20px;overflow:hidden;font-size:16px}.grid-content__brief,.grid-content__ellipsis{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.grid-content__brief{font-size:14px;color:rgba(0,0,0,.4)}@media (max-width:767px){.grid-item{width:100%}.grid-item__icon{width:40px;height:40px;line-height:40px}}", ""]
                }] }
    ];
    /** @nocollapse */
    FriendBrowseComponent.ctorParameters = function () { return [
        { type: SocialService }
    ]; };
    return FriendBrowseComponent;
}());
export { FriendBrowseComponent };
if (false) {
    /** @type {?} */
    FriendBrowseComponent.prototype.category;
    /** @type {?} */
    FriendBrowseComponent.prototype.members;
    /** @type {?} */
    FriendBrowseComponent.prototype.params;
    /**
     * @type {?}
     * @private
     */
    FriendBrowseComponent.prototype._timer;
    /** @type {?} */
    FriendBrowseComponent.prototype.placeholder;
    /**
     * @type {?}
     * @private
     */
    FriendBrowseComponent.prototype.socialService;
}
