// inet-social/esm5/src/share-member/share-member.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewChild, Input } from '@angular/core';
import { SocialService } from '../social.service';
import { Policy, SHARE_COMPANY, SHARE_FRIEND, SHARE_ME, SHARE_PUBLISH } from '../model/Config';
import { AccentService } from 'inet-core';
var ShareMemberComponent = /** @class */ (function () {
    function ShareMemberComponent(socialService, accentService) {
        this.socialService = socialService;
        this.accentService = accentService;
        this.companyMode = true;
        this.companyCategories = [
            SHARE_COMPANY,
            SHARE_ME
        ];
        this.communityCategories = [
            SHARE_PUBLISH,
            SHARE_FRIEND,
            SHARE_ME
        ];
        this.groups = [];
        // All groups, members
        this.contacts = [];
        // Groups, members active
        this.members = [];
        this.friends = [];
        // Search data
        this.searchContacts = [];
        this.selectContacts = [];
        this.showMenu = false;
        this.policies = [];
        this.customActive = false;
        this._matchContacts = [];
        this._params = {
            pageSize: 10,
            pageNumber: 0,
            loading: false
        };
    }
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        document.body.appendChild(this.shareModal._element.nativeElement);
        this.onCloseMenu = this.onCloseMenu.bind(this);
        document.addEventListener('click', this.onCloseMenu);
        this.init();
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.activity) {
                _this.setPolicies(_this.activity._policies);
            }
            else {
                _this.setPolicies();
            }
        }), 100);
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        document.removeEventListener('click', this.onCloseMenu);
        this._modalShowObserver.unsubscribe();
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._modalShowObserver = this.shareModal.onShown.subscribe((/**
         * @return {?}
         */
        function () {
            _this._focusInputMember();
            _this.searchMember();
        }));
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.init = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.companyMode) {
            this.categories = this.companyCategories;
            // Load all users company
            /** @type {?} */
            var userAdded = [];
            this.socialService.searchMember({ pageSize: -1 }, (/**
             * @param {?} members
             * @return {?}
             */
            function (members) {
                members.forEach((/**
                 * @param {?} member
                 * @return {?}
                 */
                function (member) {
                    /** @type {?} */
                    var user = (/** @type {?} */ (_this._createSocialMember({
                        member: member.username,
                        display: member.fullname,
                        policy: Policy.Individual
                    })));
                    user.username = member.username;
                    _this.contacts.push(user);
                }));
                _this._sortMembers(_this.contacts);
            }));
        }
        else {
            this.categories = this.communityCategories;
            // Load all groups
            this.socialService.listGroups((/**
             * @param {?} groups
             * @return {?}
             */
            function (groups) {
                groups.forEach((/**
                 * @param {?} group
                 * @return {?}
                 */
                function (group) {
                    /** @type {?} */
                    var member = _this._createSocialMember({
                        display: group,
                        member: group,
                        policy: Policy.Friend
                    });
                    member.group = true;
                    _this.groups.push(member);
                    _this.contacts.push(member);
                }));
                _this._sortMembers(_this.groups);
                _this._sortMembers(_this.contacts);
            }));
            // Load all friends
            this.socialService.listFriends({ pageSize: -1 }, (/**
             * @param {?} members
             * @return {?}
             */
            function (members) {
                members.forEach((/**
                 * @param {?} member
                 * @return {?}
                 */
                function (member) {
                    /** @type {?} */
                    var user = _this._createSocialMember({
                        member: member.usercode,
                        display: member.fullname,
                        policy: Policy.Individual
                    });
                    user.username = member.usercode;
                    _this.friends.push(user);
                    _this.contacts.push(user);
                }));
                _this._sortMembers(_this.friends);
                _this._sortMembers(_this.contacts);
            }));
        }
        this.categories.forEach((/**
         * @param {?} category
         * @return {?}
         */
        function (category) { return category._index = _this._indexKeyToSearch(category.display); }));
        this.contacts = this.categories.slice();
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.setDefault = /**
     * @return {?}
     */
    function () {
        this._resetActive();
        this._resetModal();
        this.closeMenu();
    };
    /**
     * @param {?=} members
     * @return {?}
     */
    ShareMemberComponent.prototype.setPolicies = /**
     * @param {?=} members
     * @return {?}
     */
    function (members) {
        this.setDefault();
        if (!members || members.length < 1) {
            members = [this.categories[0]];
        }
        // Set active category or group
        if (members.length === 1 && (members[0].category || members[0].group)) {
            members[0].active = true;
        }
        else {
            this.customActive = true;
        }
        /** @type {?} */
        var displays = [];
        /** @type {?} */
        var policies = [];
        for (var i = 0; i < members.length; i++) {
            /** @type {?} */
            var member = members[i];
            displays.push(member.display);
            policies = policies.concat(member.value);
        }
        this.policies = policies;
        this.display = displays.join(', ');
        this.displayIcon = members[0].icon;
        this.members = members;
    };
    /**
     * @param {?} params
     * @return {?}
     */
    ShareMemberComponent.prototype.injectPolicies = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        if (this.policies.length < 1) {
            params.published = true;
            // params.element = JSON.stringify('[]');
        }
        else {
            params.element = JSON.stringify(this.policies);
        }
        return params;
    };
    /**
     * @param {?} member
     * @return {?}
     */
    ShareMemberComponent.prototype.selectMember = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        this.setPolicies([member]);
        this.closeMenu();
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.onCloseMenu = /**
     * @return {?}
     */
    function () {
        if (!this.isOpening()) {
            this.showMenu = false;
        }
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.openSelectCustom = /**
     * @return {?}
     */
    function () {
        this.selectContacts = this.members.slice();
        this.shareModal.show();
        this._updateSelectContacts();
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.closeMenu = /**
     * @return {?}
     */
    function () {
        this.showMenu = false;
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.openMenu = /**
     * @return {?}
     */
    function () {
        this.showMenu = true;
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.searchMember = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var value = this._indexKeyToSearch(this._getSearchKeyword());
        if (value) {
            this._matchContacts = this.contacts.filter((/**
             * @param {?} member
             * @return {?}
             */
            function (member) {
                return member._index.indexOf(value) > -1 ||
                    (typeof member.id === 'string' && member.id.indexOf(value) > -1);
            }));
        }
        else {
            this._matchContacts = this.contacts;
        }
        this._loadContacts(true);
    };
    /**
     * @param {?} member
     * @return {?}
     */
    ShareMemberComponent.prototype.addMember = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        member.selected = !member.selected;
        if (member.selected) {
            if (!this.hasSelectedMember(member)) {
                this.selectContacts.push(member);
            }
        }
        else {
            this.removeMember(member);
        }
    };
    /**
     * @param {?} member
     * @return {?}
     */
    ShareMemberComponent.prototype.removeMember = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        /** @type {?} */
        var index = this.getIndexMemberSelected(member);
        if (index > -1) {
            member.selected = false;
            this.selectContacts.splice(index, 1);
        }
    };
    /**
     * @param {?} member
     * @return {?}
     */
    ShareMemberComponent.prototype.hasSelectedMember = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        return this.getIndexMemberSelected(member) > -1;
    };
    /**
     * @param {?} member
     * @return {?}
     */
    ShareMemberComponent.prototype.getIndexMemberSelected = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        for (var i = 0; i < this.selectContacts.length; i++) {
            if (this.selectContacts[i].id === member.id) {
                return i;
            }
        }
        return -1;
    };
    /**
     * @return {?}
     */
    ShareMemberComponent.prototype.isOpening = /**
     * @return {?}
     */
    function () {
        return this.shareModal.isShown;
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ShareMemberComponent.prototype.cancelChanges = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.stopPropagation();
        this.shareModal.hide();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ShareMemberComponent.prototype.saveChanges = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        e.stopPropagation();
        this.setPolicies(this.selectContacts);
        this.shareModal.hide();
        this.closeMenu();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ShareMemberComponent.prototype.onScrollContact = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        /** @type {?} */
        var node = e.target;
        if (!this._params.loading && node.scrollHeight - (node.scrollTop + node.clientHeight) <= 40) {
            this._loadContacts();
        }
    };
    /**
     * @private
     * @return {?}
     */
    ShareMemberComponent.prototype._scrollTopModal = /**
     * @private
     * @return {?}
     */
    function () {
        this.contactList.nativeElement.scrollTop = 0;
    };
    /**
     * @private
     * @return {?}
     */
    ShareMemberComponent.prototype._updateSelectContacts = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.contacts.forEach((/**
         * @param {?} member
         * @return {?}
         */
        function (member) { return member.selected = _this.hasSelectedMember(member); }));
    };
    /**
     * @private
     * @return {?}
     */
    ShareMemberComponent.prototype._resetActive = /**
     * @private
     * @return {?}
     */
    function () {
        this.categories.forEach((/**
         * @param {?} cat
         * @return {?}
         */
        function (cat) { return cat.active = false; }));
        this.groups.forEach((/**
         * @param {?} group
         * @return {?}
         */
        function (group) { return group.active = false; }));
        this.customActive = false;
    };
    /**
     * @private
     * @return {?}
     */
    ShareMemberComponent.prototype._resetModal = /**
     * @private
     * @return {?}
     */
    function () {
        this._params.loading = false;
        this._params.pageNumber = 0;
        this._matchContacts = [];
        this.selectContacts = [];
        this.searchContacts = [];
        this._setSearchKeyword('');
    };
    /**
     * @private
     * @param {?=} init
     * @return {?}
     */
    ShareMemberComponent.prototype._loadContacts = /**
     * @private
     * @param {?=} init
     * @return {?}
     */
    function (init) {
        this._params.loading = true;
        if (init) {
            this._scrollTopModal();
            this._params.pageNumber = 0;
            this.searchContacts = [];
        }
        this.searchContacts = this.searchContacts.concat(this._matchContacts.slice(this._params.pageNumber * this._params.pageSize, (this._params.pageNumber + 1) * this._params.pageSize));
        this._params.pageNumber += 1;
        this._params.loading = !this._hasContacts();
    };
    /**
     * @private
     * @return {?}
     */
    ShareMemberComponent.prototype._hasContacts = /**
     * @private
     * @return {?}
     */
    function () {
        return this._matchContacts.length > this._params.pageNumber * this._params.pageSize;
    };
    /**
     * @private
     * @param {?} members
     * @return {?}
     */
    ShareMemberComponent.prototype._sortMembers = /**
     * @private
     * @param {?} members
     * @return {?}
     */
    function (members) {
        members.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) { return a.display.localeCompare(b.display); }));
    };
    /**
     * @private
     * @param {?} str
     * @return {?}
     */
    ShareMemberComponent.prototype._indexKeyToSearch = /**
     * @private
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return this.accentService.viToEn(str).toLowerCase();
    };
    /**
     * @private
     * @param {?} policyItem
     * @return {?}
     */
    ShareMemberComponent.prototype._createSocialMember = /**
     * @private
     * @param {?} policyItem
     * @return {?}
     */
    function (policyItem) {
        return {
            id: policyItem.member,
            display: policyItem.display,
            icon: this.socialService.getIconByType(policyItem.policy),
            value: [policyItem],
            _index: this._indexKeyToSearch(policyItem.display)
        };
    };
    /**
     * @private
     * @return {?}
     */
    ShareMemberComponent.prototype._focusInputMember = /**
     * @private
     * @return {?}
     */
    function () {
        this.inputMember.nativeElement.focus();
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    ShareMemberComponent.prototype._setSearchKeyword = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.inputMember.nativeElement.value = value;
    };
    /**
     * @private
     * @return {?}
     */
    ShareMemberComponent.prototype._getSearchKeyword = /**
     * @private
     * @return {?}
     */
    function () {
        return this.inputMember.nativeElement.value;
    };
    ShareMemberComponent.decorators = [
        { type: Component, args: [{
                    selector: '[socialShareMember]',
                    template: "<div class=\"group-share\">\n    <button (click)=\"$event.stopPropagation(); showMenu = !showMenu\" class=\"share-button btn btn-light active btn-sm\" tabindex=\"-1\" [title]=\"display\">\n        <div class=\"share-button_contain\">\n            <div class=\"share-button__padding\">\n                <i class=\"fa\" [ngClass]=\"displayIcon\"></i>\n            </div>\n            <div class=\"share-button__content\">\n                <div>{{display}}</div>\n            </div>\n            <div>\n                <i class=\"fa fa-caret-down\"></i>\n            </div>\n        </div>\n    </button>\n    <ul (click)=\"$event.stopPropagation()\" class=\"dropdown-menu share-menu\" [ngClass]=\"{'show': showMenu}\">\n        <li class=\"dropdown-header\">\n            Ai s\u1EBD nh\u00ECn th\u1EA5y n\u1ED9i dung n\u00E0y?\n        </li>\n\n        <li class=\"dropdown-item\" *ngFor=\"let category of categories\" (click)=\"selectMember(category)\">\n            <i class=\"fa {{category.icon}} group-share-icon\"></i> {{ category.display }}\n            <i [hidden]=\"!category.active\" class=\"fa fa-check category-check\"></i>\n        </li>\n\n        <li class=\"dropdown-item\" (click)=\"openSelectCustom()\">\n            <i class=\"fa fa-cog group-share-icon\"></i> T\u00F9y ch\u1ECDn...\n            <div *ngIf=\"customActive\">\n                <div class=\"custom-display\">{{display}}</div>\n                <i class=\"fa fa-check category-check\"></i>\n            </div>\n        </li>\n\n        <div *ngIf=\"groups.length\" class=\"seperate\"></div>\n        <li *ngFor=\"let group of groups\" class=\"dropdown-item\" (click)=\"selectMember(group)\">\n            <i class=\"fa group-share-icon\" [ngClass]=\"group.icon\"></i> {{group.display}}\n            <i [hidden]=\"!group.active\" class=\"fa fa-check category-check\"></i>\n        </li>\n    </ul>\n</div>\n\n<div class=\"modal fade\" bsModal #shareModal=\"bs-modal\" tabindex=\"-1\" role=\"dialog\">\n  <div class=\"modal-dialog\">\n    <div class=\"modal-content\">\n        <div class=\"modal-header\">\n            <h5 class=\"modal-title pull-left\">Chia s\u1EBB</h5>\n            <button type=\"button\" class=\"btn btn-primary btn-sm pull-right\" (click)=\"saveChanges($event)\" tabindex=\"-1\">\n                Xong\n            </button>\n        </div>\n        <div class=\"modal-body\">\n            <div class=\"auto-tag form-control contact-search\">\n                <div *ngFor=\"let member of selectContacts\" class=\"auto-tag-item\">\n                    <span>{{member.display}}</span>\n                    <i class=\"fa fa-times\" (click)=\"removeMember(member)\"></i>\n                </div>\n                <input #searchInput (keyup)=\"searchMember()\" type=\"text\" class=\"auto-tag-input\" placeholder=\"T\u00ECm ki\u1EBFm b\u1EA1n b\u00E8...\">\n            </div>\n\n            <div #contactList (scroll)=\"onScrollContact($event)\" class=\"contact-list\">\n                <div class=\"list-item list-avatar\" *ngFor=\"let member of searchContacts\" (click)=\"addMember(member)\">\n                    <img *ngIf=\"member.username; else icon\" userAvatar [usercode]=\"member.username\" class=\"list-item__image\">\n                    <ng-template #icon>\n                        <i class=\"list-item__icon fa\" [ngClass]=\"member.icon\"></i>\n                    </ng-template>\n                    <div class=\"list-item__content\">{{member.display}}</div>\n                    <label class=\"checkbox-container\">\n                        <input type=\"checkbox\" [checked]=\"member.selected\" disabled>\n                        <span class=\"checkmark\"></span>\n                    </label>\n                </div>\n            </div>\n        </div>\n    </div>\n  </div>\n</div>\n",
                    styles: [".auto-tag{display:flex;flex-wrap:wrap;height:auto;padding-bottom:0}.auto-tag .auto-tag-item{display:flex;align-items:center;border-radius:4px;background:#eee;padding:0 10px;white-space:nowrap;margin:0 5px 5px 0}.auto-tag .auto-tag-item i{cursor:pointer;padding:5px;margin-right:-5px}.auto-tag .auto-tag-input{border:0;outline:0}.checkbox-container{display:block;position:relative;cursor:pointer;font-size:22px;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;min-width:20px;min-height:20px;margin:0}.checkbox-container input{position:absolute;opacity:0;cursor:pointer}.checkbox-container .checkmark{position:absolute;top:0;left:0;height:20px;width:20px;background-color:#eee;border-radius:4px}.checkbox-container .checkmark:after{content:\"\";position:absolute;display:none;left:8px;top:4px;width:5px;height:10px;border:solid #fff;border-width:0 3px 3px 0;transform:rotate(45deg)}.checkbox-container:hover .checkmark{background-color:#ccc}.checkbox-container input:checked+.checkmark{background-color:#2196f3}.checkbox-container input:checked+.checkmark:after{display:block}.list-avatar{padding:5px 10px;overflow:hidden;cursor:pointer;margin-bottom:5px;display:flex;align-items:center}.list-item__icon,.list-item__image{width:32px;height:32px;border-radius:50%}.list-item__icon{line-height:32px;text-align:center;background:rgba(0,0,0,.05);color:rgba(0,0,0,.5)}.list-item__content{overflow:hidden;text-overflow:ellipsis;white-space:nowrap;padding:0 15px;line-height:32px;flex-grow:1}", ".share-menu{max-height:300px;max-width:300px;overflow:auto}.seperate{margin:5px 20px;border-top:1px solid #eee}.group-share{position:relative;display:inline-block}.group-share-icon{width:22px}.dropdown-header,.dropdown-item{position:relative;font-size:14px;cursor:pointer;padding-left:32px}.category-check{position:absolute;left:8px;top:7px}.share-button{max-width:300px}.share-button_contain{display:flex}.share-button__content,.share-button__padding{padding-right:5px}.share-button__content{overflow:hidden}.share-button__content div{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.custom-display{font-size:13px;color:rgba(0,0,0,.6);line-height:15px;white-space:normal}.contact-list{height:300px;overflow:auto;margin-top:10px}@media (max-width:480px){.share-button{max-width:220px}.contact-search{max-height:140px;overflow:auto}.contact-list{height:200px}}"]
                }] }
    ];
    /** @nocollapse */
    ShareMemberComponent.ctorParameters = function () { return [
        { type: SocialService },
        { type: AccentService }
    ]; };
    ShareMemberComponent.propDecorators = {
        shareModal: [{ type: ViewChild, args: ['shareModal',] }],
        inputMember: [{ type: ViewChild, args: ['searchInput',] }],
        contactList: [{ type: ViewChild, args: ['contactList',] }],
        companyMode: [{ type: Input }],
        activity: [{ type: Input }]
    };
    return ShareMemberComponent;
}());
export { ShareMemberComponent };
if (false) {
    /** @type {?} */
    ShareMemberComponent.prototype.shareModal;
    /** @type {?} */
    ShareMemberComponent.prototype.inputMember;
    /** @type {?} */
    ShareMemberComponent.prototype.contactList;
    /** @type {?} */
    ShareMemberComponent.prototype.companyMode;
    /** @type {?} */
    ShareMemberComponent.prototype.activity;
    /** @type {?} */
    ShareMemberComponent.prototype.categories;
    /** @type {?} */
    ShareMemberComponent.prototype.companyCategories;
    /** @type {?} */
    ShareMemberComponent.prototype.communityCategories;
    /** @type {?} */
    ShareMemberComponent.prototype.groups;
    /** @type {?} */
    ShareMemberComponent.prototype.contacts;
    /** @type {?} */
    ShareMemberComponent.prototype.members;
    /** @type {?} */
    ShareMemberComponent.prototype.friends;
    /** @type {?} */
    ShareMemberComponent.prototype.searchContacts;
    /** @type {?} */
    ShareMemberComponent.prototype.selectContacts;
    /** @type {?} */
    ShareMemberComponent.prototype.showMenu;
    /** @type {?} */
    ShareMemberComponent.prototype.displayIcon;
    /** @type {?} */
    ShareMemberComponent.prototype.display;
    /** @type {?} */
    ShareMemberComponent.prototype.policies;
    /** @type {?} */
    ShareMemberComponent.prototype.customActive;
    /**
     * @type {?}
     * @private
     */
    ShareMemberComponent.prototype._matchContacts;
    /**
     * @type {?}
     * @private
     */
    ShareMemberComponent.prototype._params;
    /**
     * @type {?}
     * @private
     */
    ShareMemberComponent.prototype._modalShowObserver;
    /**
     * @type {?}
     * @private
     */
    ShareMemberComponent.prototype.socialService;
    /**
     * @type {?}
     * @private
     */
    ShareMemberComponent.prototype.accentService;
}
