// inet-calendar/esm2015/src/form/calendar-form.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewChild } from '@angular/core';
import { CalendarService } from "../calendar.service";
import { CalendarDialogRepeatComponent } from "../dialog-repeat/dialog-repeat.component";
import { ActivatedRoute, Router } from "@angular/router";
import { NotificationService, HttpClientService } from 'inet-core';
import { CalendarDialogDelete } from "../dialog-delete/dialog-delete.component";
import { AttachmentListComponent } from "../attachment-list/attachment-list.component";
import * as moment_ from 'moment';
import { vnToLatin } from "../utils/common/Accent";
import { CalAttendeeType } from "../utils/model/CalAttendeeType";
import { xCalendar } from "../utils/xCalendar";
import { Observable } from "rxjs";
import { debounceTime, distinctUntilChanged } from "rxjs/operators";
import { CalendarUtils } from "../CalendarUtils";
/** @type {?} */
const moment = moment_;
export class CalendarFormComponent {
    /**
     * @param {?} route
     * @param {?} calendarService
     * @param {?} notify
     * @param {?} router
     * @param {?} http
     */
    constructor(route, calendarService, notify, router, http) {
        this.route = route;
        this.calendarService = calendarService;
        this.notify = notify;
        this.router = router;
        this.http = http;
        this.event = {};
        this.formData = {
            title: '',
            date: '',
            startTime: '',
            toTime: '',
            location: '',
            summary: '',
            state: '',
            notes: '',
            repeat: false,
            _allday: ''
        };
        this.attendees = [];
        this._attendees = [];
        this.attendeeSearch = new Observable((/**
         * @param {?} subscriber
         * @return {?}
         */
        subscriber => {
            /** @type {?} */
            let value = this.attendeeModal.trim();
            if (!value) {
                subscriber.next([]);
                return;
            }
            value = vnToLatin(value);
            subscriber.next(this._attendees.filter((/**
             * @param {?} item
             * @return {?}
             */
            (item) => item['_index'].indexOf(value) > -1)).slice(0, 10));
        })).pipe(debounceTime(200), distinctUntilChanged());
        this.locationSearch = new Observable((/**
         * @param {?} subscriber
         * @return {?}
         */
        subscriber => {
            /** @type {?} */
            let value = this.formData.location;
            if (!value) {
                subscriber.next([]);
                return;
            }
            this.calendarService.searchLocation(value, (/**
             * @param {?} items
             * @return {?}
             */
            (items) => subscriber.next(items.map((/**
             * @param {?} item
             * @return {?}
             */
            item => item.content)))));
        })).pipe(debounceTime(200), distinctUntilChanged());
        this.calendarService.ready((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            let params = this.route.snapshot.paramMap;
            if (params.get('id')) {
                this.calendarService.loadEvent({
                    element: params.get('id')
                }, (/**
                 * @param {?} event
                 * @return {?}
                 */
                (event) => {
                    if (!event) {
                        this.eventNotFound = true;
                        return;
                    }
                    this.event = event;
                    this.attendees = event.attendees;
                    if (event.members) {
                        this.event.myAttendee = event.members.find((/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => {
                            return item.role !== 'CREATOR' && item.username === iNet.username;
                        }));
                    }
                    this._initForm();
                }));
            }
            else {
                /** @type {?} */
                let event;
                try {
                    event = JSON.parse(atob(params.get('data')));
                    event.start = new Date(event.start);
                    event.end = new Date(event.end);
                }
                catch (e) {
                    event = {};
                    event.start = new Date();
                    event.start.setHours(event.start.getHours() + 1, 0, 0, 0);
                    event.end = new Date(event.start.getTime() + 60 * 60 * 1000);
                }
                this.event = event;
                this._initForm();
            }
        }));
        this.calendarService.loadAttendees((/**
         * @param {?} attendees
         * @return {?}
         */
        (attendees) => this._attendees = attendees));
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.titleInput.nativeElement.focus();
    }
    /**
     * @private
     * @return {?}
     */
    _initForm() {
        /** @type {?} */
        let values = {
            title: this.event.subject || '',
            date: this.event.start,
            startTime: moment(this.event.start).format('HH:mm'),
            toTime: moment(this.event.end).format('HH:mm'),
            location: this.event.location || '',
            summary: this.event.summary || '',
            _allday: this.event.attribute && this.event.attribute['allday'] === 'true' || false,
            state: 'NEEDS_ACTION',
            notes: '',
            repeat: false,
        };
        if (this.event.myAttendee) {
            values.state = this.event.myAttendee.state;
            values.notes = this.event.myAttendee.notes || '';
        }
        Object.assign(this.formData, values);
        if (this.event.repeat) {
            this.setRepeatData(this.event.repeat);
        }
    }
    /**
     * @return {?}
     */
    createEvent() {
        /** @type {?} */
        let data = this._getFormData();
        if (!data) {
            return;
        }
        this.http.showLoading('Đang xử lý...');
        if (data.element) {
            this.calendarService.updateEvent(data, (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                if (data) {
                    this._onSuccess();
                }
                else {
                    this.notify.showMessage('Cập nhật lịch không thành công', 'error');
                }
            }));
            this._updateMyState();
        }
        else {
            this.calendarService.createEvent(data, (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                if (data) {
                    this._onSuccess();
                }
                else {
                    this.notify.showMessage('Tạo lịch không thành công', 'error');
                }
            }));
        }
    }
    /**
     * @private
     * @return {?}
     */
    _onSuccess() {
        this.router.navigate(['/calendar']);
    }
    /**
     * @private
     * @return {?}
     */
    _getFormData() {
        if (!this._validate()) {
            return;
        }
        /** @type {?} */
        let data = this.formData;
        /** @type {?} */
        let params = {
            subject: data.title,
            summary: data.summary,
            location: data.location,
            startTime: this.getMinutes(data._allday ? '00:00' : data.startTime),
            toTime: this.getMinutes(data._allday ? '23:59' : data.toTime),
            _allday: data._allday
        };
        /** @type {?} */
        let date = moment(data.date);
        params.day = date.dayOfYear();
        params.year = date.year();
        if (this.event.uuid) {
            params.element = this.event.uuid;
        }
        if (data.repeat && this.event.repeat) {
            Object.assign(params, this.event.repeat);
            if (params.wkdays) {
                params.wkdays = params.wkdays.join(',');
            }
        }
        this.fileList.files.forEach((/**
         * @param {?} file
         * @param {?} index
         * @return {?}
         */
        (file, index) => {
            if (!file.id) {
                params['file' + index] = file;
            }
        }));
        if (params.location) {
            this.calendarService.saveLocation(params.location, (/**
             * @return {?}
             */
            () => { }));
        }
        params.attendee = this.attendees;
        xCalendar.modifyDataUpdate(params, this.event);
        return params;
    }
    /**
     * @private
     * @param {?} attendee
     * @return {?}
     */
    _hasAttendee(attendee) {
        for (let i = 0; i < this.attendees.length; i++) {
            if (this.attendees[i].username === attendee.username) {
                return true;
            }
        }
        return false;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    addAttendee(e) {
        /** @type {?} */
        let attendee = e.item;
        this.attendeeModal = '';
        if (!this._hasAttendee(attendee)) {
            attendee.type = CalAttendeeType.USER;
            this.attendees.push(attendee);
        }
    }
    /**
     * @private
     * @return {?}
     */
    _validate() {
        if (!this.formData.title.trim() || !this.formData.date) {
            return false;
        }
        if (!this.formData._allday) {
            if (this.getMinutes(this.formData.startTime) >= this.getMinutes(this.formData.toTime)) {
                this.timeError = true;
                return false;
            }
        }
        this.timeError = false;
        return true;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    setRepeatData(data) {
        if (!data) {
            this._clearRepeat();
            return;
        }
        this.event.repeat = data;
        this.repeatText = CalendarUtils.getRepeatText(data);
        this.formData.repeat = true;
    }
    /**
     * @param {?} checked
     * @return {?}
     */
    openRepeat(checked) {
        if (checked) {
            this.repeatModal.openRepeat({
                lstart: this.event.lstart
            });
        }
        this._clearRepeat();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onEditRepeat(e) {
        e.preventDefault();
        e.stopPropagation();
        this.repeatModal.openRepeat(this.event.repeat);
    }
    /**
     * @private
     * @return {?}
     */
    _clearRepeat() {
        this.repeatText = '';
        this.formData.repeat = false;
    }
    /**
     * @private
     * @return {?}
     */
    _updateMyState() {
        if (!this.event.myAttendee) {
            return;
        }
        this.calendarService.loadEvent({
            state: this.formData.state,
            notes: this.formData.notes,
            element: this.event.uuid
        }, (/**
         * @return {?}
         */
        function () { }));
    }
    /**
     * @return {?}
     */
    goBack() {
        history.back();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onRemoveEvent(event) {
        if (event) {
            event.preventDefault();
            event.stopPropagation();
        }
        this.viewDeleteModal.show();
    }
    /**
     * @private
     * @param {?} hourStr
     * @return {?}
     */
    getMinutes(hourStr) {
        /** @type {?} */
        let hours = hourStr.split(':');
        return Number(hours[0]) * 60 + Number(hours[1]);
    }
}
CalendarFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-calendar-form',
                template: "<div class=\"admin-layout\">\n    <div class=\"container-fluid nav-fixed-top cp-toolbar\">\n        <button (click)=\"goBack()\" class=\"btn btn-sm btn-primary mr-1 ng-star-inserted\" data-toggle=\"tooltip\" type=\"button\" title=\"Tr\u1EDF v\u1EC1\">\n            <i class=\"fa fa-arrow-left\"></i>\n        </button>\n        <button *ngIf=\"event.isCreator || !event.uuid\" (click)=\"createEvent()\" class=\"btn btn-sm btn-primary ng-star-inserted mr-1\" data-toggle=\"tooltip\" type=\"button\" title=\"L\u01B0u s\u1EF1 ki\u1EC7n\">\n            <i class=\"fa fa-save\"></i>\n        </button>\n        <button *ngIf=\"event.isCreator\" (click)=\"onRemoveEvent($event)\" class=\"btn btn-sm btn-danger ng-star-inserted\" data-toggle=\"tooltip\" type=\"button\" title=\"X\u00F3a s\u1EF1 ki\u1EC7n\">\n            <i class=\"fa fa-trash\"></i>\n        </button>\n    </div>\n    <div class=\"cp-content p-3\">\n        <div class=\"card\" *ngIf=\"eventNotFound\">\n            <div class=\"card-body text-center\">\n                Kh\u00F4ng t\u00ECm th\u1EA5y l\u1ECBch\n            </div>\n        </div>\n        <div [ngClass]=\"{'hide': eventNotFound}\" class=\"row\">\n                <div class=\"col-md-6\">\n                    <div class=\"form-group\">\n                        <label class=\"font-weight-bold\">Ti\u00EAu \u0111\u1EC1 (<span class=\"text-danger\">*</span>):</label>\n                        <input #titleInput type=\"text\" class=\"form-control\"\n                               [(ngModel)]=\"formData.title\"\n                               [ngClass]=\"{'is-invalid': titleInput.value.trim().length === 0}\">\n                    </div>\n                    <div class=\"form-group row\">\n                        <div class=\"col-6\" style=\"padding-right:5px\">\n                            <label class=\"font-weight-bold\">Ng\u00E0y s\u1EF1 ki\u1EC7n (<span class=\"text-danger\">*</span>):</label>\n                            <input type=\"text\" class=\"form-control\"\n                                   [(ngModel)]=\"formData.date\"\n                                   [ngClass]=\"{'is-invalid': formData.date.length === 0}\"\n                                   #dp=\"bsDatepicker\"\n                                   bsDatepicker\n                                   [bsConfig]=\"{dateInputFormat: 'DD/MM/YYYY'}\">\n                        </div>\n                        <div *ngIf=\"!allDay.checked\" class=\"col-3\" style=\"padding-right:5px;padding-left:5px\">\n                            <label class=\"font-weight-bold\">T\u1EEB gi\u1EDD (<span class=\"text-danger\">*</span>):</label>\n                            <input type=\"text\" class=\"form-control\"\n                                   [(ngModel)]=\"formData.startTime\"\n                                   [ngClass]=\"{'is-invalid': (timeError || formData.startTime.length === 0)}\"\n                                   timePicker\n                                   [timePickerModal]=\"timeStartModal\">\n                            <time-picker-modal #timeStartModal></time-picker-modal>\n                        </div>\n                        <div *ngIf=\"!allDay.checked\" class=\"col-3\" style=\"padding-left:5px\">\n                            <label class=\"font-weight-bold\">\u0110\u1EBFn gi\u1EDD (<span class=\"text-danger\">*</span>):</label>\n                            <input type=\"text\" class=\"form-control\"\n                                   [(ngModel)]=\"formData.toTime\"\n                                   [ngClass]=\"{'is-invalid': (timeError || formData.toTime.length === 0)}\"\n                                   timePicker\n                                   [timePickerModal]=\"timeEndModal\">\n                            <time-picker-modal #timeEndModal></time-picker-modal>\n                        </div>\n                    </div>\n                    <div class=\"form-group form-check\">\n                        <input #allDay [(ngModel)]=\"formData._allday\" type=\"checkbox\" id=\"_allday\" class=\"form-check-input\" value=\"true\"/>\n                        <label class=\"form-check-label font-weight-bold\" for=\"_allday\">C\u1EA3 ng\u00E0y</label>\n                    </div>\n                    <div class=\"form-group form-check\">\n                        <input (click)=\"openRepeat(repeat.checked)\"\n                               #repeat\n                               [(ngModel)]=\"formData.repeat\"\n                               class=\"form-check-input\"\n                               id=\"repeat\" type=\"checkbox\" value=\"true\"/>\n                        <label class=\"form-check-label font-weight-bold\" for=\"repeat\">L\u1EB7p l\u1ECBch</label>\n                        <span *ngIf=\"repeatText\">\n                            <b>:</b> {{repeatText}}\n                            <i (click)=\"onEditRepeat($event)\" class=\"fa fa-pencil\" style=\"padding: 5px;margin: -5px 0;cursor: pointer;\" title=\"Ch\u1EC9nh s\u1EEDa\"></i>\n                        </span>\n                    </div>\n                    <div class=\"form-group\">\n                        <button (click)=\"fileAttachment.click()\" title=\"T\u1EADp tin \u0111\u00EDnh k\u00E8m\" type=\"button\" class=\"btn btn-primary btn-sm rounded\">\n                            <i class=\"fa fa-cloud-upload pr-1\" aria-hidden=\"true\"></i>\n                            <span>\u0110\u00EDnh k\u00E8m</span>\n                            <input #fileAttachment type=\"file\" multiple style=\"display:none\">\n                        </button>\n                        <div style=\"margin-top: 10px\">\n                            <calendar-attachment-list\n                                    [calEvent]=\"event\"\n                                    [removable]=\"true\"\n                                    [fileEl]=\"fileAttachment\"></calendar-attachment-list>\n                        </div>\n                    </div>\n                    <div class=\"form-group\">\n                        <label class=\"font-weight-bold\">\u0110\u1ECBa \u0111i\u1EC3m:</label>\n                        <input type=\"text\" class=\"form-control\"\n                               [(ngModel)]=\"formData.location\"\n                               [typeahead]=\"locationSearch\">\n                    </div>\n                    <div class=\"form-group\">\n                        <label class=\"font-weight-bold\">M\u00F4 t\u1EA3:</label>\n                        <textarea (keydown.enter)=\"$event.stopPropagation()\" type=\"text\" class=\"form-control\"\n                                  [(ngModel)]=\"formData.summary\"></textarea>\n                    </div>\n                </div>\n                <div class=\"col-md-6\">\n                    <div [hidden]=\"!event.myAttendee\" class=\"form-group\">\n                        <label class=\"font-weight-bold\">B\u1EA1n c\u00F3 tham d\u1EF1 kh\u00F4ng?</label>\n                        <div class=\"input-group\">\n                            <select [(ngModel)]=\"formData.state\" class=\"attendee-select form-control\" style=\"margin-right:-3px\">\n                                <option value=\"ACCEPTED\">C\u00F3</option>\n                                <option value=\"DECLINED\">Kh\u00F4ng</option>\n                                <option value=\"NEEDS_ACTION\">C\u00F3 th\u1EC3</option>\n                            </select>\n                            <input [(ngModel)]=\"formData.notes\" type=\"text\" class=\"form-control\" placeholder=\"Ghi ch\u00FA\">\n                        </div>\n                    </div>\n                    <div class=\"form-group\">\n                        <label class=\"font-weight-bold\">Th\u00E0nh ph\u1EA7n: </label>\n                        <ng-template #attendeeTemplate let-attendee=\"item\">\n                            <div class=\"attendee-item\">\n                                <img userAvatar [usercode]=\"attendee.username\">\n                                <div>{{attendee.fullname}}</div>\n                            </div>\n                        </ng-template>\n                        <input type=\"text\" class=\"form-control\"\n                               #attendeeInput\n                               [(ngModel)]=\"attendeeModal\"\n                               [typeahead]=\"attendeeSearch\"\n                               [typeaheadItemTemplate]=\"attendeeTemplate\"\n                               (typeaheadOnSelect)=\"addAttendee($event)\"/>\n                    </div>\n                    <div class=\"form-group\">\n                        <calendar-attendee-list [attendees]=\"attendees\" [editable]=\"true\"></calendar-attendee-list>\n                    </div>\n                </div>\n            </div>\n    </div>\n    <calendar-repeat-modal #repeatModal (onRepeat)=\"setRepeatData($event)\"></calendar-repeat-modal>\n</div>\n<calendar-dialog-delete #dialogRemove [calEvent]=\"event\" (onDelete)=\"goBack()\" hidden></calendar-dialog-delete>\n",
                styles: [".attendee-select{background:#e9ecef;width:90px;padding-right:5px;padding-left:5px;flex:none}.attendee-list-container{overflow:auto;max-height:200px;box-shadow:rgba(0,0,0,.14) 0 8px 10px 1px,rgba(0,0,0,.12) 0 3px 14px 2px,rgba(0,0,0,.2) 0 5px 5px -3px;border-radius:4px;margin-top:5px}.form-check-input{opacity:1!important;margin-top:1px}.attendee-item{display:flex;margin:0 -10px;align-items:center}.attendee-item>img{width:30px;height:30px;border-radius:50%;margin-right:10px}"]
            }] }
];
/** @nocollapse */
CalendarFormComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: CalendarService },
    { type: NotificationService },
    { type: Router },
    { type: HttpClientService }
];
CalendarFormComponent.propDecorators = {
    repeatModal: [{ type: ViewChild, args: ['repeatModal',] }],
    viewDeleteModal: [{ type: ViewChild, args: ['dialogRemove',] }],
    titleInput: [{ type: ViewChild, args: ['titleInput',] }],
    attendeeInput: [{ type: ViewChild, args: ['attendeeInput',] }],
    fileList: [{ type: ViewChild, args: [AttachmentListComponent,] }]
};
if (false) {
    /** @type {?} */
    CalendarFormComponent.prototype.repeatModal;
    /** @type {?} */
    CalendarFormComponent.prototype.viewDeleteModal;
    /** @type {?} */
    CalendarFormComponent.prototype.titleInput;
    /** @type {?} */
    CalendarFormComponent.prototype.attendeeInput;
    /** @type {?} */
    CalendarFormComponent.prototype.fileList;
    /** @type {?} */
    CalendarFormComponent.prototype.event;
    /** @type {?} */
    CalendarFormComponent.prototype.repeatText;
    /** @type {?} */
    CalendarFormComponent.prototype.eventNotFound;
    /** @type {?} */
    CalendarFormComponent.prototype.timeError;
    /** @type {?} */
    CalendarFormComponent.prototype.attendeeModal;
    /** @type {?} */
    CalendarFormComponent.prototype.formData;
    /** @type {?} */
    CalendarFormComponent.prototype.attendees;
    /**
     * @type {?}
     * @private
     */
    CalendarFormComponent.prototype._attendees;
    /** @type {?} */
    CalendarFormComponent.prototype.attendeeSearch;
    /** @type {?} */
    CalendarFormComponent.prototype.locationSearch;
    /**
     * @type {?}
     * @private
     */
    CalendarFormComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    CalendarFormComponent.prototype.calendarService;
    /**
     * @type {?}
     * @private
     */
    CalendarFormComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    CalendarFormComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    CalendarFormComponent.prototype.http;
}
