// inet-calendar/esm2015/src/calendar-widget/calendar-widget.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input, ElementRef } from '@angular/core';
import { CalendarViewerComponent } from "../calendar-viewer/calendar-viewer.component";
import { xCalendar } from "../utils/xCalendar";
import { CalendarService } from "../calendar.service";
export class CalendarWidgetComponent {
    /**
     * @param {?} calendarService
     */
    constructor(calendarService) {
        this.calendarService = calendarService;
        this.locale = 'vi';
        this.dateFormat = 'dddd, D/M/YYYY';
        this.colors = [
            '#438EB9',
            '#3CA54A',
            '#FF9800',
            '#EF5B49',
            '#77901B',
            '#7D5392',
            '#808080'
        ];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.loadEvents();
    }
    /**
     * @param {?} date
     * @return {?}
     */
    changeMonth(date) {
        this._date = date;
        this.calendarViewer.clearEvents();
        this.loadEvents();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    activeDate(e) {
        this.activeDateViewer = e;
        this.activeDateTitle = xCalendar.formatDate(e.date, this.dateFormat);
    }
    /**
     * @return {?}
     */
    loadEvents() {
        if (!this._date) {
            return;
        }
        this.calendarService.loadMonthEvents(this._date, (/**
         * @param {?} events
         * @return {?}
         */
        (events) => {
            events.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                return a.day - b.day;
            }));
            this._groupEventByDate(events).forEach((/**
             * @param {?} group
             * @return {?}
             */
            (group) => {
                this.calendarViewer.setEvents(group.events, group.date);
            }));
            // Focus today
            if (!this.activeDateViewer) {
                this.calendarViewer.focusDate(new Date());
            }
        }));
    }
    /**
     * @param {?} event
     * @param {?} jsEvent
     * @param {?} targetEl
     * @return {?}
     */
    onViewEvent(event, jsEvent, targetEl) {
        this.calendarService.viewEvent(event, jsEvent, targetEl, (/**
         * @return {?}
         */
        () => this.loadEvents()));
    }
    /**
     * @private
     * @param {?} events
     * @return {?}
     */
    _groupEventByDate(events) {
        /** @type {?} */
        let groups = [];
        events.forEach((/**
         * @param {?} event
         * @return {?}
         */
        event => {
            /** @type {?} */
            let group;
            /** @type {?} */
            let date = new Date(event.from);
            date.setHours(0, 0, 0, 0);
            for (let i = 0; i < groups.length; i++) {
                if (groups[i].date.getTime() === date.getTime()) {
                    group = groups[i];
                    break;
                }
            }
            if (!group) {
                group = {
                    events: [],
                    date: date
                };
                groups.push(group);
            }
            this._initEventStatus(event);
            group.events.push(event);
        }));
        // sort event by time
        groups.forEach((/**
         * @param {?} group
         * @return {?}
         */
        (group) => {
            group.events.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            (a, b) => a.startTime - b.startTime));
        }));
        return groups;
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    _initEventStatus(event) {
        /** @type {?} */
        let now = new Date();
        event.progressing = event.start < now && event.end > now;
        event.ended = event.end < now;
        event.pending = event.start > now;
        if (event.progressing) {
            event.statusText = 'Đang diễn ra';
            event.statusColor = this.colors[0];
        }
        else if (event.pending) {
            event.statusText = 'Chưa diễn ra';
            event.statusColor = this.colors[5];
        }
        else {
            event.statusText = 'Kết thúc';
            event.statusColor = this.colors[6];
        }
        event.style = {
            'background-color': this.colors[event.mode]
        };
    }
}
CalendarWidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'calendar-widget',
                template: "<calendar-viewer #calendarViewer [locale]=\"locale\" (onChange)=\"changeMonth($event)\" (onActive)=\"activeDate($event)\"></calendar-viewer>\n<div class=\"calendar-container\" *ngIf=\"activeDateViewer?.events?.length\">\n    <div class=\"calendar-event__date\">{{activeDateTitle}}</div>\n    <div #viewToggle class=\"calendar-body\">\n        <div *ngFor=\"let event of activeDateViewer.events\" #targetEl (click)=\"onViewEvent(event, $event, targetEl)\"  class=\"calendar-event\">\n            <div  class=\"calendar-event__state\" [ngStyle]=\"event.style\"></div>\n            <div class=\"calendar-event__content\">\n                <div class=\"calendar-event__title\">\n                    {{event.subject}}\n                    <div class=\"calendar-event__progress\" [style.color]=\"event.statusColor\">{{event.statusText}}</div>\n                </div>\n                <div class=\"calendar-event__text\">\n                    <i class=\"fa fa-clock-o calendar-event__icon\"></i>\n                    {{event.hourStr}}\n                </div>\n                <div *ngIf=\"event.location\" class=\"calendar-event__text\">\n                    <i class=\"fa fa-map-marker calendar-event__icon\"></i>\n                    {{event.location}}\n                </div>\n            </div>\n        </div>\n    </div>\n</div>",
                styles: [".calendar-container{padding-bottom:15px}.calendar-event__date{font-size:12px;color:#757575;font-weight:700;margin:0 15px 5px;padding-top:15px;text-transform:uppercase;border-top:1px solid #ddd}.calendar-body{max-height:300px;overflow:hidden auto}.calendar-event{padding:5px 15px;color:#212121;cursor:pointer}.calendar-event:hover{background-color:rgba(0,0,0,.05)}.calendar-event__state{width:8px;height:8px;border-radius:50%;margin:5px 10px 0 0;float:left}.calendar-event__content{overflow:hidden}.calendar-event__title{font-size:14px;line-height:18px;margin-bottom:3px}.calendar-event__progress{font-size:11px;text-transform:uppercase;font-weight:700}.calendar-event__text{font-size:12px;line-height:18px;color:#757575}.calendar-event__icon{width:12px}"]
            }] }
];
/** @nocollapse */
CalendarWidgetComponent.ctorParameters = () => [
    { type: CalendarService }
];
CalendarWidgetComponent.propDecorators = {
    calendarViewer: [{ type: ViewChild, args: ['calendarViewer',] }],
    viewToggle: [{ type: ViewChild, args: ['viewToggle',] }],
    locale: [{ type: Input }],
    dateFormat: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CalendarWidgetComponent.prototype.calendarViewer;
    /** @type {?} */
    CalendarWidgetComponent.prototype.viewToggle;
    /** @type {?} */
    CalendarWidgetComponent.prototype.locale;
    /** @type {?} */
    CalendarWidgetComponent.prototype.dateFormat;
    /** @type {?} */
    CalendarWidgetComponent.prototype.activeDateViewer;
    /** @type {?} */
    CalendarWidgetComponent.prototype.activeDateTitle;
    /** @type {?} */
    CalendarWidgetComponent.prototype.colors;
    /**
     * @type {?}
     * @private
     */
    CalendarWidgetComponent.prototype._date;
    /**
     * @type {?}
     * @private
     */
    CalendarWidgetComponent.prototype.calendarService;
}
