// inet-ui/esm2015/src/web-token/lib/WebTokenSigner.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EXT_ACTION, EXT_CHROME_URL, EXT_FIREFOX_URL, EXT_HASH_ALG, NATIVE_APP_URL } from "./Constants";
import { CertUtils } from "./CertUtils";
export class WebTokenSigner {
    constructor() {
        this.isChrome = /Google Inc/.test(navigator.vendor);
        this.isFirefox = /Firefox/.test(navigator.userAgent);
        this.isWindows = navigator.platform.indexOf('Win') > -1;
        this.osSupported = this.isWindows;
        this.browserSupported = this.isChrome || this.isFirefox;
        this.appInstalled = false;
        this.appInfo = {
            version: null,
            nativeVersion: null,
            nativeLatestVersion: null
        };
        this.textOsUnSupport = 'Tính năng đang hỗ trợ cho máy tính dùng Windows';
        this.textBrowserUnSupport = 'Tính năng đang hỗ trợ cho Chrome và FireFox';
        this.textExtNotInstalled = `
    Cần cài đặt extension <a href="${this.isChrome ? EXT_CHROME_URL : EXT_FIREFOX_URL}" target="_blank">TokenSigner</a> trên trình duyệt Web
    `;
        this.textAppNotInstalled = `
    Cần cài đặt ứng dụng ký <a href="${NATIVE_APP_URL}" target="_blank">TokenSigner</a> trên máy tính
    `;
        this.title = "Chứng thư số";
        this.description = "Chọn chứng thư số để ký";
        this.textOk = 'Đồng ý';
        this.textCancel = 'Đóng';
        this.certNotFound = 'Không tìm thấy chứng thư số';
        this.textIssuer = 'Nhà cung cấp';
        this.textValid = 'Hiệu lực';
        this.zIndex = 1100;
        this._onEscapeKey = this._onEscapeKey.bind(this);
    }
    /**
     * @return {?}
     */
    get extInstalled() {
        return !!window['_ExtTokenBridge'];
    }
    ;
    /**
     * @param {?=} silent
     * @return {?}
     */
    checkReady(silent) {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            if (this.appInstalled) {
                return resolve(true);
            }
            if (!this.osSupported) {
                if (!silent) {
                    this._createCertDialog(null, this.textOsUnSupport);
                }
                return resolve(false);
            }
            if (!this.browserSupported) {
                if (!silent) {
                    this._createCertDialog(null, this.textBrowserUnSupport);
                }
                return resolve(false);
            }
            if (!this.extInstalled) {
                if (!silent) {
                    this._createCertDialog(null, this.textExtNotInstalled);
                }
                return resolve(false);
            }
            this._loadAppInfo().then((/**
             * @return {?}
             */
            () => {
                if (!this.appInstalled && !silent) {
                    this._createCertDialog(null, this.textAppNotInstalled);
                }
                resolve(this.appInstalled);
            }));
        }));
    }
    /**
     * @return {?}
     */
    checkNativeUpdate() {
        return this.appInstalled && this.appInfo.nativeVersion < this.appInfo.nativeLatestVersion;
    }
    /**
     * @return {?}
     */
    loadCerts() {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            this.checkReady(true).then((/**
             * @param {?} ready
             * @return {?}
             */
            (ready) => {
                if (ready) {
                    this._postExt(EXT_ACTION.GET_CERTS).then((/**
                     * @param {?} data
                     * @return {?}
                     */
                    (data) => {
                        /** @type {?} */
                        const certs = data && data.data || [];
                        certs.forEach((/**
                         * @param {?} cert
                         * @return {?}
                         */
                        cert => {
                            cert.subjectData = CertUtils.parseCertString(cert.subject);
                            cert.issuerData = CertUtils.parseCertString(cert.issuer);
                        }));
                        resolve(certs);
                    }));
                }
                else {
                    resolve([]);
                }
            }));
        }));
    }
    /**
     * @param {?} data
     * @return {?}
     */
    signHash(data) {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            if (data && (data.hash || data.hashs)) {
                this.checkReady().then((/**
                 * @param {?} ready
                 * @return {?}
                 */
                (ready) => {
                    if (ready) {
                        data.hashAlg = data.hashAlg || EXT_HASH_ALG.SHA1;
                        this._postExt(EXT_ACTION.SIGN_HASH, data).then(resolve);
                    }
                    else {
                        resolve(null);
                    }
                }));
            }
            else {
                resolve({
                    error: "HASH_MISSING"
                });
            }
        }));
    }
    /**
     * @param {?=} forceSelect
     * @return {?}
     */
    selectCert(forceSelect) {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            this.checkReady().then((/**
             * @param {?} ready
             * @return {?}
             */
            (ready) => {
                if (ready) {
                    this.loadCerts().then((/**
                     * @param {?} certs
                     * @return {?}
                     */
                    (certs) => {
                        if (certs.length > 1 || forceSelect) {
                            this._certCallback = resolve;
                            this._createCertDialog(certs);
                        }
                        else {
                            resolve(certs[0]);
                        }
                    }));
                }
                else {
                    resolve(null);
                }
            }));
        }));
    }
    /**
     * @param {?} dateStr
     * @return {?}
     */
    dateDisplay(dateStr) {
        /** @type {?} */
        const date = new Date(dateStr);
        return [date.getDate(), date.getMonth() + 1, date.getFullYear()].join('/');
    }
    /**
     * @private
     * @param {?} action
     * @param {?=} data
     * @return {?}
     */
    _postExt(action, data) {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            if (this.extInstalled) {
                window["_ExtTokenBridge"].postToExtension(action, data).then((/**
                 * @param {?} response
                 * @return {?}
                 */
                (response) => {
                    resolve(response);
                }));
            }
            else {
                resolve({
                    error: "EXTENSION_NOT_INSTALL"
                });
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    _loadAppInfo() {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            this._postExt(EXT_ACTION.GET_INFO).then((/**
             * @param {?} reply
             * @return {?}
             */
            (reply) => {
                if (reply && reply.data) {
                    Object.assign(this.appInfo, reply.data);
                }
                this.appInstalled = !!this.appInfo.nativeVersion;
                resolve();
            }));
        }));
    }
    /**
     * @private
     * @param {?} certs
     * @param {?=} message
     * @return {?}
     */
    _createCertDialog(certs, message) {
        if (this._dialog) {
            return;
        }
        this._dialog = document.createElement('div');
        this._dialog.classList.add('webtoken-dialog');
        this._dialog.style.zIndex = this.zIndex;
        this._certs = certs;
        /** @type {?} */
        let msg;
        if (message) {
            msg = '<div class="webtoken-dialog__message">' + message + '</div>';
        }
        else {
            msg = certs.length < 1 ? `<div class="webtoken-dialog__message">${this.certNotFound}</div>` : this._buildCertHtml(certs);
        }
        this._dialog.innerHTML = `
        <style>
        .webtoken-dialog {
            position: fixed;
            top: 0;
            right: 0;
            bottom: 0;
            left: 0;
            background: rgba(0, 0, 0, .3);
        }
        .webtoken-dialog__container {
            width: 400px;
            margin: auto;
            background: #fff;
            overflow: hidden;
            color: #333;
            position: absolute;
            transform: translate(-50%, -50%);
            top: 50%;
            left: 50%;
            border-radius: 4px;
        }
        .webtoken-dialog__header {
            padding: 10px 15px;
            border-bottom: 1px solid #ddd;
        }
        .webtoken-dialog__title {
            font-size: 20px;
            font-weight: bold;
            margin-bottom: 5px;
        }
        .webtoken-dialog__desc {
            font-size: 14px;
            line-height: 20px;
        }
        .webtoken-dialog__close {
            float: right;
            width: 30px;
            cursor: pointer;
            text-align: center;
            line-height: 30px;
            margin: -5px 0;
            font-size: 20px;
        }
        .webtoken-dialog__body {
            max-height: 250px;
            overflow: auto;
            font-size: 14px;
            line-height: 20px;
        }
        .webtoken-dialog__message {
            padding: 20px;            
        }
        .webtoken-cert__item {
            padding: 10px 20px;
            cursor: pointer;
            line-height: 20px;
        }
        .webtoken-cert__item.active {
            background: #ddd !important;
        }
        .webtoken-cert__item:hover {
            background: #f5f5f5;
        }
        .webtoken-cert__title {
            font-size: 18px;
            margin-bottom: 2px;
        }
        .webtoken-cert__text {
            font-size: 14px;
        }
        .webtoken-dialog__footer {
            padding: 10px 20px;
            border-top: 1px solid #ddd;
            text-align: right;
        }
        .webtoken-dialog__button {
            height: 30px;
            line-height: 30px;
            padding: 0 10px;
            min-width: 60px;
            margin-left: 10px;
            cursor: pointer;
            border-radius: 4px;
            font-weight: bold;
            background: #fff;
            border: 1px solid #ddd;
        }
        .webtoken-dialog__button:hover {
            opacity: 8;
        }
        .btn-ok {
            background: #007bff;
            border-color: #007bff;
            color: #fff;
        }
        </style>
        <div class="webtoken-dialog__container">
        <div class="webtoken-dialog__header">
        <div class="webtoken-dialog__close">×</div>
        <div class="webtoken-dialog__title">${this.title}</div>
        ${message ? '' : `<div class="webtoken-dialog__desc">${this.description}</div>`}
        </div>
        <div class="webtoken-dialog__body">${msg}</div>
        <div class="webtoken-dialog__footer">
        <button class="btn-close webtoken-dialog__button">${this.textCancel}</button>
        ${message ? '' : `<button class="btn-ok webtoken-dialog__button">${this.textOk}</button>`}
        </div>
        </div>
        `;
        /** @type {?} */
        const certEl = this._dialog.getElementsByClassName('webtoken-cert__item');
        // Close dialog
        this._dialog.addEventListener('click', (/**
         * @return {?}
         */
        () => this._closeDialog()));
        this._dialog.getElementsByClassName('webtoken-dialog__close')[0].addEventListener('click', (/**
         * @return {?}
         */
        () => this._closeDialog()));
        // Prevent close dialog
        this._dialog.getElementsByClassName('webtoken-dialog__container')[0].addEventListener('click', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            e.stopPropagation();
        }));
        this._dialog.getElementsByClassName('btn-close')[0].addEventListener('click', (/**
         * @return {?}
         */
        () => this._closeDialog()));
        if (certs && certs.length > 0) {
            // Ok select
            this._dialog.getElementsByClassName('btn-ok')[0].addEventListener('click', (/**
             * @return {?}
             */
            () => {
                if (this._certEl) {
                    /** @type {?} */
                    let index = Number(this._certEl.getAttribute('index'));
                    if (this._certs[index]) {
                        this._certCallback(this._certs[index]);
                    }
                }
                else {
                    this._certCallback(null);
                }
                this._closeDialog();
            }));
            // On click cert item
            for (let i = 0; i < certEl.length; i++) {
                certEl[i].addEventListener('click', (/**
                 * @param {?} e
                 * @return {?}
                 */
                (e) => this._onClickCertEl(e)));
            }
            // Select first cert
            certEl[0].click();
        }
        // Esc key close dialog
        document.addEventListener('keydown', this._onEscapeKey);
        document.body.appendChild(this._dialog);
    }
    /**
     * @private
     * @param {?} certs
     * @return {?}
     */
    _buildCertHtml(certs) {
        /** @type {?} */
        let html = '';
        certs.forEach((/**
         * @param {?} cert
         * @param {?} index
         * @return {?}
         */
        (cert, index) => {
            html += `
            <div class="webtoken-cert__item" index="${index}">
            <div class="webtoken-cert__title">${cert.subjectData.CN}</div>
            <div class="webtoken-cert__text">${this.textIssuer}: ${cert.issuerData.CN}</div>
            <div class="webtoken-cert__text">${this.textValid}: ${this.dateDisplay(cert.validFrom)} - ${this.dateDisplay(cert.validTo)}</div>
            </div>
            `;
        }));
        return html;
    }
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    _onClickCertEl(e) {
        /** @type {?} */
        let certEl = this._findCerEl(e.target);
        if (certEl.classList.contains('active')) {
            return;
        }
        if (this._certEl) {
            this._certEl.classList.remove('active');
        }
        certEl.classList.add('active');
        this._certEl = certEl;
    }
    /**
     * @private
     * @param {?} certEl
     * @return {?}
     */
    _findCerEl(certEl) {
        if (!certEl.classList.contains('webtoken-cert__item')) {
            return this._findCerEl(certEl.parentNode);
        }
        return certEl;
    }
    /**
     * @private
     * @return {?}
     */
    _closeDialog() {
        document.body.removeChild(this._dialog);
        document.removeEventListener('keydown', this._onEscapeKey);
        this._dialog = null;
        this._certEl = null;
        this._certCallback = null;
    }
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    _onEscapeKey(e) {
        if (e.which === 27) {
            this._closeDialog();
        }
    }
}
if (false) {
    /** @type {?} */
    WebTokenSigner.prototype.isChrome;
    /** @type {?} */
    WebTokenSigner.prototype.isFirefox;
    /** @type {?} */
    WebTokenSigner.prototype.isWindows;
    /** @type {?} */
    WebTokenSigner.prototype.osSupported;
    /** @type {?} */
    WebTokenSigner.prototype.browserSupported;
    /** @type {?} */
    WebTokenSigner.prototype.appInstalled;
    /** @type {?} */
    WebTokenSigner.prototype.appInfo;
    /** @type {?} */
    WebTokenSigner.prototype.textOsUnSupport;
    /** @type {?} */
    WebTokenSigner.prototype.textBrowserUnSupport;
    /** @type {?} */
    WebTokenSigner.prototype.textExtNotInstalled;
    /** @type {?} */
    WebTokenSigner.prototype.textAppNotInstalled;
    /** @type {?} */
    WebTokenSigner.prototype.title;
    /** @type {?} */
    WebTokenSigner.prototype.description;
    /** @type {?} */
    WebTokenSigner.prototype.textOk;
    /** @type {?} */
    WebTokenSigner.prototype.textCancel;
    /** @type {?} */
    WebTokenSigner.prototype.certNotFound;
    /** @type {?} */
    WebTokenSigner.prototype.textIssuer;
    /** @type {?} */
    WebTokenSigner.prototype.textValid;
    /** @type {?} */
    WebTokenSigner.prototype.zIndex;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._dialog;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._certEl;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._certs;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._certCallback;
    /* Skipping unhandled member: ;*/
}
