// inet-ui/esm5/src/list-view/list-view-row.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ListViewRow = /** @class */ (function () {
    function ListViewRow($implicit, index, count) {
        if (index === void 0) { index = 0; }
        if (count === void 0) { count = 0; }
        this.$implicit = $implicit;
        this.index = index;
        this.count = count;
        this.selected = false;
        this.active = false;
    }
    Object.defineProperty(ListViewRow.prototype, "first", {
        get: /**
         * @return {?}
         */
        function () { return this.index === 0; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListViewRow.prototype, "last", {
        get: /**
         * @return {?}
         */
        function () { return this.index === this.count - 1; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListViewRow.prototype, "even", {
        get: /**
         * @return {?}
         */
        function () { return this.index % 2 === 0; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListViewRow.prototype, "odd", {
        get: /**
         * @return {?}
         */
        function () { return !this.even; },
        enumerable: true,
        configurable: true
    });
    return ListViewRow;
}());
export { ListViewRow };
if (false) {
    /** @type {?} */
    ListViewRow.prototype.selected;
    /** @type {?} */
    ListViewRow.prototype.active;
    /** @type {?} */
    ListViewRow.prototype.$implicit;
    /** @type {?} */
    ListViewRow.prototype.index;
    /** @type {?} */
    ListViewRow.prototype.count;
}
