// inet-ui/esm2015/src/list-view/list-view-row.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ListViewRow {
    /**
     * @param {?} $implicit
     * @param {?=} index
     * @param {?=} count
     */
    constructor($implicit, index = 0, count = 0) {
        this.$implicit = $implicit;
        this.index = index;
        this.count = count;
        this.selected = false;
        this.active = false;
    }
    /**
     * @return {?}
     */
    get first() { return this.index === 0; }
    /**
     * @return {?}
     */
    get last() { return this.index === this.count - 1; }
    /**
     * @return {?}
     */
    get even() { return this.index % 2 === 0; }
    /**
     * @return {?}
     */
    get odd() { return !this.even; }
}
if (false) {
    /** @type {?} */
    ListViewRow.prototype.selected;
    /** @type {?} */
    ListViewRow.prototype.active;
    /** @type {?} */
    ListViewRow.prototype.$implicit;
    /** @type {?} */
    ListViewRow.prototype.index;
    /** @type {?} */
    ListViewRow.prototype.count;
}
