// inet-ui/esm2015/src/list-view/list-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, ContentChild, TemplateRef, } from '@angular/core';
import { ListViewRow } from "./list-view-row";
export class ListViewComponent {
    constructor() {
        this.listItems = [];
        this.selectedItems = [];
        this.activeItem = -1;
        this.items = [];
        this.onSelectionChanged = new EventEmitter();
        this.idProperty = 'uuid';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /*
        console.log('[listItems]', this.listItems);
        console.log('[template]', this.template);
        console.log('[toolbar]', this.toolbar);
        */
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const count = this.items.length;
        this.listItems = this.items.map((/**
         * @param {?} item
         * @param {?} index
         * @return {?}
         */
        (item, index) => {
            return new ListViewRow(item, index, count);
        }));
        this.autoSelectItem();
    }
    /**
     * @private
     * @return {?}
     */
    autoSelectItem() {
        if (this.autoSelect && this.listItems.length > 0) {
            this.selectedItems = [];
            this.selectItem(0); //active first item
        }
    }
    /**
     * @return {?}
     */
    requestFocus() {
        this.listBody.nativeElement.focus();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    handleKeyPress(event) {
        /** @type {?} */
        const keyCode = event.which;
        /** @type {?} */
        let row = this.activeItem < 0 ? {} : this.listItems[this.activeItem];
        row.active = false;
        /** @type {?} */
        const prevActiveItem = this.activeItem;
        switch (keyCode) {
            case 40:
                if (this.activeItem === this.items.length - 1) {
                    this.activeItem = 0;
                    break;
                }
                this.activeItem++;
                break;
            case 38:
            case 9:
                if (this.activeItem === 0) {
                    this.activeItem = this.items.length - 1;
                    break;
                }
                this.activeItem--;
                break;
            case 32:
                this.toggleItem(this.activeItem);
                break;
            default:
                return;
        }
        event.preventDefault();
        event.stopPropagation();
        if (this.activeItem == prevActiveItem) {
            return;
        }
        if (event.altKey) {
            this.updateItemState();
            return;
        }
        //const isSelected = row.selected;
        if (event.shiftKey) {
            row.selected = true;
        }
        else {
            row.selected = false;
            this.selectedItems = [];
        }
        this.selectItem(this.activeItem);
    }
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    handleClick(event, index) {
        if (event.metaKey || event.ctrlKey) {
            this.toggleItem(index);
        }
        else {
            this.selectedItems = [];
            this.selectItem(index);
        }
    }
    /**
     * @return {?}
     */
    updateItemState() {
        this.listItems.forEach((/**
         * @param {?} row
         * @param {?} index
         * @return {?}
         */
        (row, index) => {
            row.selected = this.selectedItems.indexOf(this.items[index]) >= 0;
            row.active = index == this.activeItem;
        }));
        if (this.activeItem >= 0) {
            this.scrollItemIntoView(this.activeItem);
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    scrollItemIntoView(index) {
        /** @type {?} */
        const activeItemEl = (/** @type {?} */ (((/** @type {?} */ (this.listBody.nativeElement))).children.item(index)));
        /** @type {?} */
        const scrollTop = this.listBody.nativeElement.scrollTop;
        /** @type {?} */
        const scrollEnd = scrollTop + this.listBody.nativeElement.offsetHeight;
        if (activeItemEl.offsetTop < scrollTop) {
            this.listBody.nativeElement.scrollTop = activeItemEl.offsetTop;
        }
        else if (activeItemEl.offsetTop + activeItemEl.offsetHeight > scrollEnd) {
            this.listBody.nativeElement.scrollTop = activeItemEl.offsetTop + activeItemEl.offsetHeight - this.listBody.nativeElement.offsetHeight;
        }
        else if (activeItemEl.offsetTop > scrollEnd) {
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    toggleItem(index) {
        if (this.selectedItems.indexOf(this.items[index]) >= 0) {
            this.deselectItem(index);
        }
        else {
            this.selectItem(index);
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    deselectItem(index) {
        this.selectedItems.splice(this.selectedItems.indexOf(this.items[index]), 1);
        this.updateItemState();
        this.onSelectionChanged.emit(this.selectedItems);
    }
    /**
     * @param {?} index
     * @return {?}
     */
    selectItem(index) {
        if (this.singleSelect) {
            this.selectedItems = [];
        }
        this.activeItem = index;
        // (<EmbeddedViewRef<ListViewRow>>this._vr.get(index)).rootNodes[0].scrollIntoView(false);
        if (this.selectedItems.indexOf(this.items[index]) < 0) {
            this.selectedItems.push(this.items[index]);
        }
        this.updateItemState();
        this.onSelectionChanged.emit(this.selectedItems);
    }
    /**
     * @return {?}
     */
    selectAll() {
        if (this.singleSelect) {
            this.autoSelectItem();
        }
        else {
            this.selectedItems = this.items.map((/**
             * @param {?} l
             * @return {?}
             */
            l => l));
            this.updateItemState();
        }
        this.onSelectionChanged.emit(this.selectedItems);
    }
    /**
     * @return {?}
     */
    clearSelection() {
        this.selectedItems = [];
        this.updateItemState();
        this.onSelectionChanged.emit(this.selectedItems);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    selectById(id) {
        //this.selectedItems = [];
        setTimeout((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            let index = this.items.findIndex((/**
             * @param {?} obj
             * @return {?}
             */
            obj => obj[this.getIdProperty()] == id));
            if (index > -1) {
                this.selectItem(index);
            }
        }), 10);
    }
    /**
     * @return {?}
     */
    getIdProperty() {
        return this.idProperty;
    }
}
ListViewComponent.decorators = [
    { type: Component, args: [{
                selector: 'list-view',
                template: "<div class=\"toolbar\">\n    <ng-template [ngTemplateOutlet]=\"toolbar\"></ng-template>\n</div>\n<div class=\"search-container\">\n    <ng-template [ngTemplateOutlet]=\"search\"></ng-template>\n</div>\n<div #listBody class=\"list-body content\" tabindex=\"0\" (keydown)=\"handleKeyPress($event)\">\n    <i class=\"list-item-body flex-column\" style=\"text-align: center;padding-top: 10px;\" *ngIf=\"listItems?.length < 1\">\n        {{'grid.emptyMsg' | translate}}\n    </i>\n    <a *ngFor=\"let item of listItems; let index = index\" class=\"list-item-body flex-column\"\n       (click)=\"handleClick($event, index)\" [attr.data-index]=\"index\">\n        <ng-template [ngTemplateOutlet]=\"template\" [ngTemplateOutletContext]=\"item\"></ng-template>\n    </a>\n</div>\n<div class=\"footer\">\n    <ng-template [ngTemplateOutlet]=\"footer\"></ng-template>\n</div>"
            }] }
];
/** @nocollapse */
ListViewComponent.ctorParameters = () => [];
ListViewComponent.propDecorators = {
    items: [{ type: Input }],
    onSelectionChanged: [{ type: Output }],
    listBody: [{ type: ViewChild, args: ["listBody",] }],
    toolbar: [{ type: ContentChild, args: ['toolbarTpl',] }],
    template: [{ type: ContentChild, args: ['contentTpl',] }],
    search: [{ type: ContentChild, args: ['searchTpl',] }],
    footer: [{ type: ContentChild, args: ['footerTpl',] }],
    idProperty: [{ type: Input, args: ['idProperty',] }],
    autoSelect: [{ type: Input, args: ['autoSelect',] }],
    singleSelect: [{ type: Input, args: ['singleSelect',] }]
};
if (false) {
    /** @type {?} */
    ListViewComponent.prototype.listItems;
    /** @type {?} */
    ListViewComponent.prototype.selectedItems;
    /** @type {?} */
    ListViewComponent.prototype.activeItem;
    /** @type {?} */
    ListViewComponent.prototype.items;
    /** @type {?} */
    ListViewComponent.prototype.onSelectionChanged;
    /** @type {?} */
    ListViewComponent.prototype.listBody;
    /** @type {?} */
    ListViewComponent.prototype.toolbar;
    /** @type {?} */
    ListViewComponent.prototype.template;
    /** @type {?} */
    ListViewComponent.prototype.search;
    /** @type {?} */
    ListViewComponent.prototype.footer;
    /** @type {?} */
    ListViewComponent.prototype.idProperty;
    /** @type {?} */
    ListViewComponent.prototype.autoSelect;
    /** @type {?} */
    ListViewComponent.prototype.singleSelect;
}
