// inet-ui/esm2015/src/user-profile/user-profile/user-profile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, ElementRef, Input, TemplateRef, ViewChild } from '@angular/core';
import { Contact } from "../../model/contact";
import { Address } from "../../model/address";
import { GlobalContactService } from "../global-contact.service";
import { ErrorMessage, NotificationService, UserProfileService } from "inet-core";
import { TranslateService } from "@ngx-translate/core";
import { CropperAvatarDialogComponent } from "../../image-cropper/cropper-avatar/cropper-avatar-dialog.component";
export class UserProfileComponent {
    /**
     * @param {?} contactService
     * @param {?} userProfileService
     * @param {?} notification
     * @param {?} translate
     */
    constructor(contactService, userProfileService, notification, translate) {
        this.contactService = contactService;
        this.userProfileService = userProfileService;
        this.notification = notification;
        this.translate = translate;
        this.profile = new Contact();
        this.defaultAvatarUrl = UserProfileService.DEFAULT_AVATAR_URL;
        this.address = new Address();
        this.isAvatarExist = false;
        this.translateSubscription = translate.get(['COMMON.MODULE.USER_PROFILE']).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            this.profileTranslations = res['COMMON.MODULE.USER_PROFILE'];
        }));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.profile.address) {
            this.profile.address = new Address();
        }
        this.isAvatarExist = !!(this.imageAvatarUrl); // Check avatar exist to delete
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.translateSubscription) {
            this.translateSubscription.unsubscribe();
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeFileAvatar($event) {
        /** @type {?} */
        const files = $event.target.files;
        if (files.length > 0) {
            /** @type {?} */
            const image = files[0];
            image['value'] = image;
            if (image) {
                /** @type {?} */
                const reader = new FileReader();
                reader.onload = (/**
                 * @param {?} e
                 * @return {?}
                 */
                (e) => {
                    image.url = e.target.result;
                    this.imageAvatarUrl = image.url;
                });
                reader.readAsDataURL(image);
            }
            this.cropperAvatarDialog.show($event);
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onSelectFile($event) {
        if (!this.upload) {
            return;
        }
        if (this.fileUpload && this.fileUpload.nativeElement) {
            this.fileUpload.nativeElement.click();
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    uploadImage($event) {
        if (this.fileUpload.nativeElement.value) {
            this.updateAvatar();
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    removeImage($event) {
        if (this.fileUpload.nativeElement.value) {
            this.fileUpload.nativeElement.value = '';
            if (this.isAvatarExist) {
                this.userProfileService.getAvatarUrl().then((/**
                 * @param {?} url
                 * @return {?}
                 */
                url => {
                    this.imageAvatarUrl = url;
                }));
            }
            else {
                this.imageAvatarUrl = this.defaultAvatarUrl;
            }
            //this.imageChangedEvent = $event;
        }
        // Delete avatar
        /*
        if (this.upload && this.isAvatarExist) {
            this.contactService.deleteAvatar().subscribe((data: any) => {
                const {AVATAR, IMAGE_REMOVE_SUCCESSFUL_MSG, IMAGE_REMOVE_ERROR_MSG} = this.profileTranslations;
                if (ErrorMessage.TYPE !== data.type) {
                    this.fileUpload.nativeElement.value = '';
                    this.notification.showMessage(IMAGE_REMOVE_SUCCESSFUL_MSG, 'success', AVATAR);
                } else {
                    this.notification.showMessage(IMAGE_REMOVE_ERROR_MSG, 'error', AVATAR);
                }
            });
        }
        */
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    updateBlankImage($event) {
        if ($event) {
            this.imageAvatarUrl = this.defaultAvatarUrl;
        }
    }
    /**
     * @private
     * @return {?}
     */
    updateAvatar() {
        /** @type {?} */
        const formData = new FormData(this.photoFrmElementRef.nativeElement);
        /** @type {?} */
        const name = this.fileUpload.nativeElement.name;
        formData.delete(name);
        //Upload Blob (cropped image) instead of file.
        formData.append(name, this.cropperAvatarDialog.getImageBlob());
        // Upload Avatar
        this.contactService.uploadPhoto(formData).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            const { AVATAR, IMAGE_UPDATE_SUCCESSFUL_MSG, IMAGE_UPDATE_ERROR_MSG } = this.profileTranslations;
            if (ErrorMessage.TYPE !== data.type) {
                this.fileUpload.nativeElement.value = '';
                this.userProfileService.increaseAvatarVersion();
                this.notification.showMessage(IMAGE_UPDATE_SUCCESSFUL_MSG, 'success', AVATAR);
            }
            else {
                this.notification.showMessage(IMAGE_UPDATE_ERROR_MSG, 'error', AVATAR);
            }
        }));
    }
    /**
     * @param {?} url
     * @return {?}
     */
    onCropped(url) {
        this.imageAvatarUrl = url;
    }
}
UserProfileComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-user-profile',
                template: "<div class=\"content-profile\">\n    <div class=\"row\">\n        <div class=\"col-lg-4\">\n            <div class=\"text-center card-box\">\n                <div class=\"member-card\">\n                    <div class=\"thumb-xl member-thumb mb-3 center-block\">\n                        <form #userPhotoFrm style=\"display: none\">\n                            <input #fileAvatar name=\"fileUpload\" (change)=\"changeFileAvatar($event)\" type=\"file\"\n                                   accept=\"image/*\"\n                                   class=\"form-control-file\">\n                        </form>\n                        <img [src]=\"imageAvatarUrl ? imageAvatarUrl: defaultAvatarUrl\" (error)=\"updateBlankImage($event)\"\n                             class=\"rounded-circle\" [ngClass]=\"{'empty-photo': !imageAvatarUrl, 'img-thumbnail': imageAvatarUrl}\"\n                             (click)=\"onSelectFile($event)\">\n                    </div>\n                    <div>\n                        <div *ngIf=\"upload\" class=\"m-t-10\">\n                            <button type=\"button\" class=\"btn btn-primary btn-sm\" (click)=\"fileAvatar.click()\">\n                                <i class=\"fa fa-picture-o\"></i> {{'COMMON.MODULE.USER_PROFILE.CHOSEN_IMAGE' | translate}}\n                            </button>\n                            <button *ngIf=\"fileUpload?.nativeElement?.value\" type=\"button\" class=\"btn btn-success btn-sm ml-1\" (click)=\"uploadImage($event)\">\n                                <i class=\"fa fa-upload\"></i> {{'COMMON.MODULE.USER_PROFILE.SAVE_IMAGE' | translate}}\n                            </button>\n                            <button *ngIf=\"fileUpload?.nativeElement?.value\" type=\"button\" (click)=\"removeImage($event)\" class=\"btn btn-danger btn-sm ml-1\">\n                                <i class=\"fa fa-trash\"></i> {{'COMMON.MODULE.USER_PROFILE.REMOVE_IMAGE' | translate}}\n                            </button>\n                        </div>\n                        <div class=\"mt-3\">\n                            <h5 *ngIf=\"profile.salutationType\">{{profile.salutationType}}. {{profile.firstName}}</h5>\n                            <h5 *ngIf=\"!profile.salutationType\">{{profile.fullName}}</h5>\n                            <p *ngIf=\"profile.userCode\" class=\"mt-0\">\n                                <i class=\"fa fa-user-circle\" aria-hidden=\"true\"></i>\n                                {{profile.userCode}}\n                            </p>\n                        </div>\n                    </div>\n                    <div class=\"text-left\">\n                        <div *ngIf=\"profile.salutationType\">\n                            <strong><i class=\"fa fa-id-card margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.FULL_NAME' | translate}}</strong>\n                            <p>{{profile.fullName}}</p>\n                        </div>\n                        <hr>\n                        <strong><i class=\"fa fa-phone margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.MOBILE' | translate}}</strong>\n                        <p class=\"text-muted\"><a href=\"tel:{{profile.mobilePhone}}\">{{profile.mobilePhone}}</a></p>\n                        <hr>\n                        <strong><i class=\"fa fa-envelope margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.EMAIL' | translate}}</strong>\n                        <p class=\"text-muted\"><a href=\"mailto:{{profile.primaryEmail}}\">{{profile.primaryEmail}}</a></p>\n                        <hr>\n                        <strong><i class=\"fa fa-map-marker margin-r-5\"></i> {{'COMMON.MODULE.ADDRESS.ADDRESS' | translate}}</strong>\n                        <p>{{profile.address?.address}}</p>\n                    </div>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-lg-8 p-0\">\n            <div class=\"info-box\">\n                <div class=\"card tab-style1\">\n                    <div>\n                        <ng-template [ngTemplateOutlet]=\"tabContent\"></ng-template>\n                        <tabset *ngIf=\"!tabContent\">\n                            <tab id=\"profile-general-tab\">\n                                <ng-template tabHeading>\n                                    <b><i class=\"fa fa-id-card margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.GENERAL_INFORMATION' | translate}}</b>\n                                </ng-template>\n                                <app-user-profile-info [profile]=\"profile\"></app-user-profile-info>\n                            </tab>\n                        </tabset>\n                    </div>\n                </div>\n            </div>\n        </div>\n    </div>\n    <app-cropper-avatar-dialog (onCropped)=\"onCropped($event)\" (onSelectFile)=\"onSelectFile($event)\"></app-cropper-avatar-dialog>\n</div>\n",
                styles: [".content-profile{min-height:250px;padding:0 15px 15px 0}.m-b-3{margin-bottom:30px!important}.box-profile{padding:10px}.img-circle{border-radius:50%}.m-b-2{margin-bottom:20px!important}.profile-username{font-size:21px;margin-top:5px}.info-box{display:block;min-height:90px;background:#fff;width:100%;box-shadow:1px 1px 20px #bababa;-o-box-shadow:1px 1px 20px #bababa;-webkit-box-shadow:1px 1px 20px #bababa;-moz-box-shadow:1px 1px 20px #bababa;border-radius:5px;height:100%;padding:15px}.margin-r-5{margin-right:5px}.card{position:relative;display:flex;flex-direction:column;background-color:#fff;border-radius:.25rem}.tab-style1{border:0}.card-box{padding:10px 20px;border-radius:5px;-moz-border-radius:5px;background-clip:padding-box;margin-bottom:20px;background-color:#fff;height:100%;box-shadow:1px 1px 20px #bababa;-o-box-shadow:1px 1px 20px #bababa;-webkit-box-shadow:1px 1px 20px #bababa;-moz-box-shadow:1px 1px 20px #bababa}.img-thumbnail{padding:0;width:150px;height:150px;cursor:pointer}.text-muted{color:#80898e!important}:host /deep/ .icon-div{width:42px;position:relative}:host /deep/ .icon-custom{position:absolute;top:50%;left:50%;margin-right:-50%;transform:translate(-50%,-50%)}:host /deep/ .upload-signature-photo{text-align:center;font-size:130px;border:2px dashed #acacac;background:#f3f3f3;width:260px;cursor:pointer}:host /deep/ .signature-thumbnail{cursor:pointer;max-width:100%;max-height:300px}.empty-photo{text-align:center;border:2px dashed #acacac;background:#f3f3f3;cursor:pointer;height:150px;width:150px}"]
            }] }
];
/** @nocollapse */
UserProfileComponent.ctorParameters = () => [
    { type: GlobalContactService },
    { type: UserProfileService },
    { type: NotificationService },
    { type: TranslateService }
];
UserProfileComponent.propDecorators = {
    tabContent: [{ type: ContentChild, args: ['tabContent',] }],
    photoFrmElementRef: [{ type: ViewChild, args: ['userPhotoFrm',] }],
    fileUpload: [{ type: ViewChild, args: ['fileAvatar',] }],
    cropperAvatarDialog: [{ type: ViewChild, args: [CropperAvatarDialogComponent,] }],
    profile: [{ type: Input }],
    imageAvatarUrl: [{ type: Input }],
    upload: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    UserProfileComponent.prototype.tabContent;
    /** @type {?} */
    UserProfileComponent.prototype.photoFrmElementRef;
    /** @type {?} */
    UserProfileComponent.prototype.fileUpload;
    /** @type {?} */
    UserProfileComponent.prototype.cropperAvatarDialog;
    /** @type {?} */
    UserProfileComponent.prototype.profile;
    /** @type {?} */
    UserProfileComponent.prototype.imageAvatarUrl;
    /** @type {?} */
    UserProfileComponent.prototype.upload;
    /** @type {?} */
    UserProfileComponent.prototype.defaultAvatarUrl;
    /** @type {?} */
    UserProfileComponent.prototype.address;
    /**
     * @type {?}
     * @private
     */
    UserProfileComponent.prototype.isAvatarExist;
    /**
     * @type {?}
     * @private
     */
    UserProfileComponent.prototype.profileTranslations;
    /**
     * @type {?}
     * @private
     */
    UserProfileComponent.prototype.translateSubscription;
    /**
     * @type {?}
     * @private
     */
    UserProfileComponent.prototype.contactService;
    /**
     * @type {?}
     * @private
     */
    UserProfileComponent.prototype.userProfileService;
    /**
     * @type {?}
     * @private
     */
    UserProfileComponent.prototype.notification;
    /**
     * @type {?}
     * @private
     */
    UserProfileComponent.prototype.translate;
}
