// inet-ui/esm2015/src/plugin-manager/cloud-plugin.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { PluginManagerService } from "./plugin-manager.service";
import { CoreService } from "inet-core";
export class CloudPluginComponent {
    /**
     * @param {?} coreService
     * @param {?} pluginManagerService
     */
    constructor(coreService, pluginManagerService) {
        this.coreService = coreService;
        this.pluginManagerService = pluginManagerService;
        this.params = {};
        this.autoHeight = false;
        this.load = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.pluginManagerService.register(this.pluginId, this);
        if (!iNet.isEmptyObject(this.params)) {
            // const __params = this.pluginManagerService.convertToHttpParams(this.params).toString();
            /** @type {?} */
            const __params = this.pluginManagerService.convertToHttpParams({
                application: this.params['application'],
                secrd: this.params['queryStr']
            }).toString();
            this.url = iNet.urlAppend(this.url, __params);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.pluginManagerService.destroyById(this.pluginId);
        // console.log('[CloudPluginComponent]--ngOnDestroy--', this.pluginId);
    }
    /**
     * @private
     * @return {?}
     */
    getContentWindow() {
        return this.pluginManagerService.getContentWindowById(this.pluginId);
    }
    /**
     * @return {?}
     */
    reload() {
        this.getContentWindow().then((/**
         * @param {?} contentWindow
         * @return {?}
         */
        contentWindow => {
            if (contentWindow && this.coreService.getEnvironment().production) {
                contentWindow['location'].reload(true);
            }
        }));
    }
    /**
     * @return {?}
     */
    resize() {
        // console.log('[resize]');
        if (this.autoHeight && this.coreService.getEnvironment().production) {
            this.getContentWindow().then((/**
             * @param {?} contentWindow
             * @return {?}
             */
            contentWindow => {
                if (contentWindow) {
                    /** @type {?} */
                    const height = contentWindow['document']['body']['offsetHeight'] || contentWindow['document']['body']['scrollHeight'];
                    if (height) {
                        document.getElementById(this.pluginId).style.height = `${height}px`;
                    }
                }
            }));
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onLoad($event) {
        if (document.getElementById(this.pluginId)) {
            this.getContentWindow().then((/**
             * @param {?} contentWindow
             * @return {?}
             */
            contentWindow => {
                if (contentWindow) {
                    this.resize();
                    // The plugin has finished loading
                    // console.log('[onLoad]--pluginId', this.pluginId, contentWindow);
                    // this.pluginManagerService.sendMessageTo({pluginId: this.pluginId, eventName: 'loaded'}, contentWindow);
                    setTimeout((/**
                     * @return {?}
                     */
                    () => {
                        this.pluginManagerService.sendMessageTo({ pluginId: this.pluginId, eventName: 'loaded' }, contentWindow);
                    }), 100);
                }
            }));
        }
    }
}
CloudPluginComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-cloud-plugin',
                template: `<iframe *ngIf="pluginId && url" id="{{pluginId}}" [src]="url | safe" (load)="onLoad($event)"></iframe>`,
                styles: [`
        iframe {
            width: 100%;
            height: 100%;
            border: none;
        }
    `]
            }] }
];
/** @nocollapse */
CloudPluginComponent.ctorParameters = () => [
    { type: CoreService },
    { type: PluginManagerService }
];
CloudPluginComponent.propDecorators = {
    pluginId: [{ type: Input }],
    url: [{ type: Input }],
    params: [{ type: Input }],
    autoHeight: [{ type: Input }],
    load: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CloudPluginComponent.prototype.pluginId;
    /** @type {?} */
    CloudPluginComponent.prototype.url;
    /** @type {?} */
    CloudPluginComponent.prototype.params;
    /** @type {?} */
    CloudPluginComponent.prototype.autoHeight;
    /** @type {?} */
    CloudPluginComponent.prototype.load;
    /**
     * @type {?}
     * @private
     */
    CloudPluginComponent.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    CloudPluginComponent.prototype.pluginManagerService;
}
