// inet-ui/esm2015/src/plugin-manager/plugin-manager.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CloudPluginComponent } from './cloud-plugin.component';
import { PluginManagerService } from './plugin-manager.service';
import { PluginToolbarDirective } from './plugin-toolbar.directive';
import { SafePipeModule } from "../pipes/safe-pipe.module";
import { CoreService, NotificationService } from "inet-core";
export class PluginManagerModule {
    /**
     * @param {?} pluginManagerService
     */
    constructor(pluginManagerService) {
        this.pluginManagerService = pluginManagerService;
        this.pluginManagerService.listen();
    }
}
PluginManagerModule.decorators = [
    { type: NgModule, args: [{
                declarations: [CloudPluginComponent, PluginToolbarDirective],
                imports: [
                    CommonModule,
                    SafePipeModule
                ],
                exports: [CloudPluginComponent, PluginToolbarDirective],
                providers: [PluginManagerService, CoreService, NotificationService]
            },] }
];
/** @nocollapse */
PluginManagerModule.ctorParameters = () => [
    { type: PluginManagerService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PluginManagerModule.prototype.pluginManagerService;
}
