// inet-ui/esm5/src/grid/components/table/table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ContentChildren, QueryList, TemplateRef, ContentChild, ViewChildren } from '@angular/core';
import { DataTableColumn } from '../column/column.component';
import { DataTableRow } from '../row/row.component';
import { defaultTranslations } from '../../types/default-translations.type';
import { drag } from '../../utils/drag';
import { TranslateService } from "@ngx-translate/core";
import { CloudTranslateService } from "../../../translate/cloud-translate.service";
var DataTable = /** @class */ (function () {
    function DataTable(translate) {
        this.translate = translate;
        this._items = [];
        this.header = false;
        this.pagination = true;
        this.pagination_range = true;
        this.pagination_limit = false;
        this.pagination_input = true;
        this.pagination_numbers = false;
        this.indexColumn = true;
        this.indexColumnHeader = '';
        this.selectColumn = false;
        this.multiSelect = true;
        this.substituteRows = false;
        this.expandableRows = false;
        this.translations = defaultTranslations;
        this.selectOnRowClick = false;
        this.autoReload = true;
        this.showReloading = true;
        this.showColumnSelector = false;
        this.autoHidePaging = false;
        this._sortAsc = true;
        this._offset = 0;
        this._limit = 10;
        // Reloading:
        this._reloading = false;
        this.reload = new EventEmitter();
        this._displayParams = (/** @type {?} */ ({})); // params of the last finished reload
        this._scheduledReload = null;
        // event handlers:
        this.rowClick = new EventEmitter();
        this.rowDoubleClick = new EventEmitter();
        this.headerClick = new EventEmitter();
        this.cellClick = new EventEmitter();
        this.selectedRows = [];
        this._selectAllCheckbox = false;
        this.selectionChange = new EventEmitter();
        this.expandClick = new EventEmitter();
        // column resizing:
        this._resizeInProgress = false;
        this.resizeLimit = 30;
    }
    Object.defineProperty(DataTable.prototype, "items", {
        get: /**
         * @return {?}
         */
        function () {
            return this._items;
        },
        set: /**
         * @param {?} items
         * @return {?}
         */
        function (items) {
            this.selectedRows = [];
            this._items = items;
            this._onReloadFinished();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTable.prototype, "sortBy", {
        get: /**
         * @return {?}
         */
        function () {
            return this._sortBy;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._sortBy = value;
            this._triggerReload();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTable.prototype, "sortAsc", {
        get: /**
         * @return {?}
         */
        function () {
            return this._sortAsc;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._sortAsc = value;
            this._triggerReload();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTable.prototype, "offset", {
        get: /**
         * @return {?}
         */
        function () {
            return this._offset;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._offset = value;
            this._triggerReload();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTable.prototype, "limit", {
        get: /**
         * @return {?}
         */
        function () {
            return this._limit;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._limit = value;
            this._triggerReload();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTable.prototype, "page", {
        // calculated property:
        get: 
        // calculated property:
        /**
         * @return {?}
         */
        function () {
            return Math.floor(this.offset / this.limit) + 1;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.offset = (value - 1) * this.limit;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataTable.prototype, "lastPage", {
        get: /**
         * @return {?}
         */
        function () {
            return Math.ceil(this.itemCount / this.limit);
        },
        enumerable: true,
        configurable: true
    });
    // setting multiple observable properties simultaneously
    // setting multiple observable properties simultaneously
    /**
     * @param {?} sortBy
     * @param {?} asc
     * @return {?}
     */
    DataTable.prototype.sort = 
    // setting multiple observable properties simultaneously
    /**
     * @param {?} sortBy
     * @param {?} asc
     * @return {?}
     */
    function (sortBy, asc) {
        this.sortBy = sortBy;
        this.sortAsc = asc;
    };
    /**
     * @return {?}
     */
    DataTable.prototype.firstPage = /**
     * @return {?}
     */
    function () {
        this.offset = 0;
    };
    // init
    // init
    /**
     * @return {?}
     */
    DataTable.prototype.ngOnInit = 
    // init
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.translateSubscription = this.translate.get([CloudTranslateService.GRID_KEY]).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            /** @type {?} */
            var __resources = res[CloudTranslateService.GRID_KEY];
            if (__resources.translations) {
                _this.translations = ((/** @type {?} */ (__resources.translations)));
            }
            else {
                _this.translations = defaultTranslations;
            }
            if (!_this.noDataMessage && __resources.emptyMsg) {
                _this.noDataMessage = __resources.emptyMsg;
            }
        }));
        this._initDefaultValues();
        this._initDefaultClickEvents();
        this._updateDisplayParams();
        if (this.autoReload && this._scheduledReload == null) {
            this.reloadItems();
        }
    };
    /**
     * @return {?}
     */
    DataTable.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.translateSubscription) {
            this.translateSubscription.unsubscribe();
        }
    };
    /**
     * @private
     * @return {?}
     */
    DataTable.prototype._initDefaultValues = /**
     * @private
     * @return {?}
     */
    function () {
        this.indexColumnVisible = this.indexColumn;
        this.selectColumnVisible = this.selectColumn;
        this.expandColumnVisible = this.expandableRows;
    };
    /**
     * @private
     * @return {?}
     */
    DataTable.prototype._initDefaultClickEvents = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.headerClick.subscribe((/**
         * @param {?} tableEvent
         * @return {?}
         */
        function (tableEvent) { return _this.sortColumn(tableEvent.column); }));
        if (this.selectOnRowClick) {
            this.rowClick.subscribe((/**
             * @param {?} tableEvent
             * @return {?}
             */
            function (tableEvent) { return tableEvent.row.selected = !tableEvent.row.selected; }));
        }
    };
    Object.defineProperty(DataTable.prototype, "reloading", {
        get: /**
         * @return {?}
         */
        function () {
            return this._reloading;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DataTable.prototype.reloadItems = /**
     * @return {?}
     */
    function () {
        this._reloading = true;
        this.reload.emit(this._getRemoteParameters());
    };
    /**
     * @private
     * @return {?}
     */
    DataTable.prototype._onReloadFinished = /**
     * @private
     * @return {?}
     */
    function () {
        this._updateDisplayParams();
        this._selectAllCheckbox = false;
        this._reloading = false;
    };
    Object.defineProperty(DataTable.prototype, "displayParams", {
        get: 
        // params of the last finished reload
        /**
         * @return {?}
         */
        function () {
            return this._displayParams;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DataTable.prototype._updateDisplayParams = /**
     * @return {?}
     */
    function () {
        this._displayParams = {
            sortBy: this.sortBy,
            sortAsc: this.sortAsc,
            offset: this.offset,
            limit: this.limit
        };
    };
    // for avoiding cascading reloads if multiple params are set at once:
    // for avoiding cascading reloads if multiple params are set at once:
    /**
     * @return {?}
     */
    DataTable.prototype._triggerReload = 
    // for avoiding cascading reloads if multiple params are set at once:
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._scheduledReload) {
            clearTimeout(this._scheduledReload);
        }
        this._scheduledReload = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.reloadItems();
        }));
    };
    /**
     * @param {?} row
     * @param {?} event
     * @return {?}
     */
    DataTable.prototype.rowClicked = /**
     * @param {?} row
     * @param {?} event
     * @return {?}
     */
    function (row, event) {
        this.rowClick.emit({ row: row, event: event });
    };
    /**
     * @param {?} row
     * @param {?} event
     * @return {?}
     */
    DataTable.prototype.rowDoubleClicked = /**
     * @param {?} row
     * @param {?} event
     * @return {?}
     */
    function (row, event) {
        this.rowDoubleClick.emit({ row: row, event: event });
    };
    /**
     * @param {?} column
     * @param {?} event
     * @return {?}
     */
    DataTable.prototype.headerClicked = /**
     * @param {?} column
     * @param {?} event
     * @return {?}
     */
    function (column, event) {
        if (!this._resizeInProgress) {
            event.preventDefault();
            event.stopPropagation();
            this.headerClick.emit({ column: column, event: event });
        }
        else {
            this._resizeInProgress = false; // this is because I can't prevent click from mousup of the drag end
        }
    };
    /**
     * @private
     * @param {?} column
     * @param {?} row
     * @param {?} event
     * @return {?}
     */
    DataTable.prototype.cellClicked = /**
     * @private
     * @param {?} column
     * @param {?} row
     * @param {?} event
     * @return {?}
     */
    function (column, row, event) {
        this.cellClick.emit({ row: row, column: column, event: event });
    };
    // functions:
    // functions:
    /**
     * @private
     * @return {?}
     */
    DataTable.prototype._getRemoteParameters = 
    // functions:
    /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var params = (/** @type {?} */ ({}));
        if (this.sortBy) {
            params.sortBy = this.sortBy;
            params.sortAsc = this.sortAsc;
        }
        if (this.pagination) {
            params.offset = this.offset;
            params.limit = this.limit;
        }
        return params;
    };
    /**
     * @private
     * @param {?} column
     * @return {?}
     */
    DataTable.prototype.sortColumn = /**
     * @private
     * @param {?} column
     * @return {?}
     */
    function (column) {
        if (column.sortable) {
            /** @type {?} */
            var ascending = this.sortBy === column.property ? !this.sortAsc : true;
            this.sort(column.property, ascending);
        }
    };
    Object.defineProperty(DataTable.prototype, "columnCount", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var count = 0;
            count += this.indexColumnVisible ? 1 : 0;
            count += this.selectColumnVisible ? 1 : 0;
            count += this.expandColumnVisible ? 1 : 0;
            this.columns.toArray().forEach((/**
             * @param {?} column
             * @return {?}
             */
            function (column) {
                count += column.visible ? 1 : 0;
            }));
            return count;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} item
     * @param {?} index
     * @param {?} row
     * @return {?}
     */
    DataTable.prototype.getRowColor = /**
     * @param {?} item
     * @param {?} index
     * @param {?} row
     * @return {?}
     */
    function (item, index, row) {
        if (this.rowColors !== undefined) {
            return ((/** @type {?} */ (this.rowColors)))(item, row, index);
        }
    };
    Object.defineProperty(DataTable.prototype, "selectAllCheckbox", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectAllCheckbox;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._selectAllCheckbox = value;
            this._onSelectAllChanged(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    DataTable.prototype._onSelectAllChanged = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.rows.toArray().forEach((/**
         * @param {?} row
         * @return {?}
         */
        function (row) { return row.selected = value; }));
    };
    /**
     * @param {?} selected
     * @param {?} row
     * @return {?}
     */
    DataTable.prototype.onRowSelectChanged = /**
     * @param {?} selected
     * @param {?} row
     * @return {?}
     */
    function (selected, row) {
        // maintain the selectedRow(s) view
        if (this.multiSelect) {
            /** @type {?} */
            var index = this.selectedRows.indexOf(row);
            if (row.selected && index < 0) {
                this.selectedRows.push(row);
            }
            else if (!row.selected && index >= 0) {
                this.selectedRows.splice(index, 1);
            }
        }
        else {
            if (row.selected) {
                this.selectedRow = row;
                this.selectedRows = [row];
                // unselect all other rows:
                this.rows.toArray().filter((/**
                 * @param {?} row_
                 * @return {?}
                 */
                function (row_) { return row_.selected; })).forEach((/**
                 * @param {?} row_
                 * @return {?}
                 */
                function (row_) {
                    if (row_ !== row) { // avoid endless loop
                        row_.selected = false;
                    }
                }));
            }
            else if (this.selectedRow === row) {
                this.selectedRow = undefined;
                this.selectedRows = [];
            }
        }
        //FIXME: prevent sticky execution
        this.selectionChange.emit({ selected: selected, row: row, selectedRows: this.selectedRows });
    };
    /**
     * @param {?} expanded
     * @param {?} row
     * @return {?}
     */
    DataTable.prototype.onExpandChanged = /**
     * @param {?} expanded
     * @param {?} row
     * @return {?}
     */
    function (expanded, row) {
        this.expandClick.emit({ expanded: expanded, row: row });
    };
    Object.defineProperty(DataTable.prototype, "substituteItems", {
        // other:
        get: 
        // other:
        /**
         * @return {?}
         */
        function () {
            return Array.from({ length: this.displayParams.limit - this.items.length });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @param {?} column
     * @param {?} columnElement
     * @return {?}
     */
    DataTable.prototype.resizeColumnStart = /**
     * @param {?} event
     * @param {?} column
     * @param {?} columnElement
     * @return {?}
     */
    function (event, column, columnElement) {
        var _this = this;
        this._resizeInProgress = true;
        drag(event, {
            move: (/**
             * @param {?} moveEvent
             * @param {?} dx
             * @return {?}
             */
            function (moveEvent, dx) {
                if (_this._isResizeInLimit(columnElement, dx)) {
                    column.width = columnElement.offsetWidth + dx;
                }
            }),
        });
    };
    /**
     * @private
     * @param {?} columnElement
     * @param {?} dx
     * @return {?}
     */
    DataTable.prototype._isResizeInLimit = /**
     * @private
     * @param {?} columnElement
     * @param {?} dx
     * @return {?}
     */
    function (columnElement, dx) {
        /* This is needed because CSS min-width didn't work on table-layout: fixed.
         Without the limits, resizing can make the next column disappear completely,
         and even increase the table width. The current implementation suffers from the fact,
         that offsetWidth sometimes contains out-of-date values. */
        if ((dx < 0 && (columnElement.offsetWidth + dx) <= this.resizeLimit) ||
            !columnElement.nextElementSibling || // resizing doesn't make sense for the last visible column
            (dx >= 0 && (((/** @type {?} */ (columnElement.nextElementSibling))).offsetWidth + dx) <= this.resizeLimit)) {
            return false;
        }
        return true;
    };
    DataTable.decorators = [
        { type: Component, args: [{
                    moduleId: 'grid-module',
                    selector: 'data-table',
                    template: "<div class=\"data-table-wrapper\">\n    <data-table-header *ngIf=\"header\"></data-table-header>\n\n    <data-table-pagination\n            *ngIf=\"pagination\"\n            [show_range]=\"pagination_range\"\n            [show_limit]=\"pagination_limit\"\n            [show_input]=\"pagination_input\"\n            [show_numbers]=\"pagination_numbers\"\n            [show_column_selector]=\"showColumnSelector\"\n            [autoHide]=\"autoHidePaging\"\n            [basicSearch]=\"basicSearch\"\n            [advanceSearch]=\"advanceSearch\">\n    </data-table-pagination>\n\n    <div class=\"data-table-box\">\n        <table class=\"table table-condensed table-bordered data-table\">\n            <thead>\n            <tr>\n                <th scope=\"col\" [hide]=\"!expandColumnVisible\" class=\"expand-column-header\">\n                <th scope=\"col\" [hide]=\"!indexColumnVisible\" class=\"index-column-header\">\n                    <span [textContent]=\"indexColumnHeader\"></span>\n                </th>\n                <th scope=\"col\" [hide]=\"!selectColumnVisible\" class=\"select-column-header\">\n                    <input [hide]=\"!multiSelect\" type=\"checkbox\" [(ngModel)]=\"selectAllCheckbox\" [attr.aria-label]=\"translations.selectAllRows\" />\n                    <label class=\"mb-0 lbl\"></label>\n                </th>\n                <th scope=\"col\" *ngFor=\"let column of columns\" #th [hide]=\"!column.visible\"\n                    (click)=\"headerClicked(column, $event)\"\n                    (keydown.enter)=\"headerClicked(column, $event)\" (keydown.space)=\"headerClicked(column, $event)\"\n\n                    [class.sortable]=\"column.sortable\" [class.resizable]=\"column.resizable\"\n\n                    [ngClass]=\"column.styleClassObject\" class=\"column-header\" [style.width]=\"column.width | px\"\n\n                    [attr.aria-sort]=\"column.sortable ? (column.property === sortBy ? (sortAsc ? 'ascending' : 'descending') : 'none') : null\"\n                    [attr.tabindex]=\"column.sortable ? '0' : null\">\n                    <span *ngIf=\"!column.headerTemplate\" [textContent]=\"column.header\"></span>\n                    <span *ngIf=\"column.headerTemplate\" [ngTemplateOutlet]=\"column.headerTemplate\" [ngTemplateOutletContext]=\"{column: column}\"></span>\n                    <span class=\"column-sort-icon\" *ngIf=\"column.sortable\">\n                            <span class=\"fa fa-sort column-sortable-icon\" [hide]=\"column.property === sortBy\"></span>\n                            <span [hide]=\"column.property !== sortBy\">\n                                 <span class=\"fa\" [ngClass]=\"{'fa-sort-asc': !sortAsc, 'fa-sort-desc': sortAsc}\"></span>\n                            </span>\n                    </span>\n                    <span *ngIf=\"column.resizable\" class=\"column-resize-handle\" (mousedown)=\"resizeColumnStart($event, column, th)\"></span>\n                </th>\n            </tr>\n            </thead>\n            <tbody *ngFor=\"let item of items; let index=index\" class=\"data-table-row-wrapper\"\n                   dataTableRow #row [item]=\"item\" [index]=\"index\"\n                   (selectedChange)=\"onRowSelectChanged($event, row)\"\n                   (expandedChange)=\"onExpandChanged($event, row)\">\n            </tbody>\n            <tbody *ngIf=\"itemCount === 0 && noDataMessage\">\n            <tr>\n                <td [attr.colspan]=\"columnCount\">{{ noDataMessage }}</td>\n            </tr>\n            </tbody>\n            <tbody class=\"substitute-rows\" *ngIf=\"pagination && substituteRows\">\n            <tr *ngFor=\"let item of substituteItems, let index = index\"\n                [class.row-odd]=\"(index + items.length) % 2 === 0\"\n                [class.row-even]=\"(index + items.length) % 2 === 1\"\n            >\n                <td [hide]=\"!expandColumnVisible\"></td>\n                <td [hide]=\"!indexColumnVisible\">&nbsp;</td>\n                <td [hide]=\"!selectColumnVisible\"></td>\n                <td *ngFor=\"let column of columns\" [hide]=\"!column.visible\">\n            </tr>\n            </tbody>\n        </table>\n\n        <div class=\"loading-cover\" *ngIf=\"showReloading && reloading\"></div>\n    </div>\n</div>\n",
                    styles: [":host /deep/ .data-table.table>tbody+tbody{border-top:none}:host /deep/ .data-table.table td{vertical-align:middle}:host /deep/ .data-table>tbody>tr>td,:host /deep/ .data-table>thead>tr>th{overflow:hidden}:host /deep/ .row-odd{background-color:#f6f6f6}.data-table .substitute-rows>tr:hover,:host /deep/ .data-table .data-table-row:hover{background-color:#ececec}.data-table{box-shadow:0 0 15px #ececec;table-layout:fixed}.column-header{position:relative}.expand-column-header{width:50px}.select-column-header{width:50px;text-align:center}.index-column-header{width:40px;text-align:center}.column-header.sortable{cursor:pointer}.column-header .column-sort-icon{float:right}.column-header.resizable .column-sort-icon{margin-right:8px}.column-header .column-sort-icon .column-sortable-icon{color:#d3d3d3}.column-header .column-resize-handle{position:absolute;top:0;right:0;margin:0;padding:0;width:8px;height:100%;cursor:col-resize}.data-table-box{position:relative}.loading-cover{position:absolute;width:100%;height:100%;background-color:rgba(255,255,255,.3);top:0}@-webkit-keyframes spinner{to{transform:rotate(360deg)}}@keyframes spinner{to{transform:rotate(360deg)}}.loading-cover:before{content:'';box-sizing:border-box;position:absolute;top:50%;left:50%;width:40px;height:40px;border-radius:50%;margin-top:-15px;margin-left:-15px;border:1px solid #ccc;border-top-color:#07d;-webkit-animation:.6s linear infinite spinner;animation:.6s linear infinite spinner}:host /deep/ .data-table.table input[type=checkbox],:host /deep/ .data-table.table input[type=radio]{opacity:0;position:absolute;z-index:12;width:18px;height:18px}:host /deep/ .data-table.table input[type=checkbox]+.lbl::before,:host /deep/ .data-table.table input[type=radio]+.lbl::before{font-family:FontAwesome;font-weight:400;font-size:11px;color:#32a3ce;content:\"\\a0\";display:inline-block;background-color:#fafafa;border:1px solid #ccc;box-shadow:0 1px 2px rgba(0,0,0,.05);border-radius:0;text-align:center;vertical-align:middle;height:13px;line-height:13px;min-width:13px;margin-right:1px}:host /deep/ .data-table.table input[type=checkbox]:active+.lbl::before,:host /deep/ .data-table.table input[type=checkbox]:checked:active+.lbl::before,:host /deep/ .data-table.table input[type=radio]:active+.lbl::before,:host /deep/ .data-table.table input[type=radio]:checked:active+.lbl::before{box-shadow:0 1px 2px rgba(0,0,0,.05),inset 0 1px 3px rgba(0,0,0,.1)}:host /deep/ .data-table.table input[type=checkbox]:checked+.lbl::before,:host /deep/ .data-table.table input[type=radio]:checked+.lbl::before{content:'\\f00c';display:inline-block;content:'\\f00c';background-color:#f5f8fc;border-color:#adb8c0;box-shadow:0 1px 2px rgba(0,0,0,.05),inset 0 -15px 10px -12px rgba(0,0,0,.05),inset 15px 10px -12px rgba(255,255,255,.1)}:host /deep/ .data-table.table input[type=checkbox]+.lbl:hover::before,:host /deep/ .data-table.table input[type=checkbox]:hover+.lbl::before,:host /deep/ .data-table.table input[type=radio]+.lbl:hover::before,:host /deep/ .data-table.table input[type=radio]:hover+.lbl::before{border-color:#ff893c}"]
                }] }
    ];
    /** @nocollapse */
    DataTable.ctorParameters = function () { return [
        { type: TranslateService }
    ]; };
    DataTable.propDecorators = {
        items: [{ type: Input }],
        itemCount: [{ type: Input }],
        columns: [{ type: ContentChildren, args: [DataTableColumn,] }],
        rows: [{ type: ViewChildren, args: [DataTableRow,] }],
        expandTemplate: [{ type: ContentChild, args: ['dataTableExpand',] }],
        headerTitle: [{ type: Input }],
        header: [{ type: Input }],
        pagination: [{ type: Input }],
        pagination_range: [{ type: Input }],
        pagination_limit: [{ type: Input }],
        pagination_input: [{ type: Input }],
        pagination_numbers: [{ type: Input }],
        indexColumn: [{ type: Input }],
        indexColumnHeader: [{ type: Input }],
        rowColors: [{ type: Input }],
        rowTooltip: [{ type: Input }],
        selectColumn: [{ type: Input }],
        multiSelect: [{ type: Input }],
        substituteRows: [{ type: Input }],
        expandableRows: [{ type: Input }],
        translations: [{ type: Input }],
        selectOnRowClick: [{ type: Input }],
        autoReload: [{ type: Input }],
        showReloading: [{ type: Input }],
        noDataMessage: [{ type: Input }],
        showColumnSelector: [{ type: Input }],
        autoHidePaging: [{ type: Input }],
        basicSearch: [{ type: ContentChild, args: ['basicSearch',] }],
        advanceSearch: [{ type: ContentChild, args: ['advanceSearch',] }],
        sortBy: [{ type: Input }],
        sortAsc: [{ type: Input }],
        offset: [{ type: Input }],
        limit: [{ type: Input }],
        page: [{ type: Input }],
        reload: [{ type: Output }],
        rowClick: [{ type: Output }],
        rowDoubleClick: [{ type: Output }],
        headerClick: [{ type: Output }],
        cellClick: [{ type: Output }],
        selectionChange: [{ type: Output }],
        expandClick: [{ type: Output }]
    };
    return DataTable;
}());
export { DataTable };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype._items;
    /** @type {?} */
    DataTable.prototype.itemCount;
    /** @type {?} */
    DataTable.prototype.columns;
    /** @type {?} */
    DataTable.prototype.rows;
    /** @type {?} */
    DataTable.prototype.expandTemplate;
    /** @type {?} */
    DataTable.prototype.headerTitle;
    /** @type {?} */
    DataTable.prototype.header;
    /** @type {?} */
    DataTable.prototype.pagination;
    /** @type {?} */
    DataTable.prototype.pagination_range;
    /** @type {?} */
    DataTable.prototype.pagination_limit;
    /** @type {?} */
    DataTable.prototype.pagination_input;
    /** @type {?} */
    DataTable.prototype.pagination_numbers;
    /** @type {?} */
    DataTable.prototype.indexColumn;
    /** @type {?} */
    DataTable.prototype.indexColumnHeader;
    /** @type {?} */
    DataTable.prototype.rowColors;
    /** @type {?} */
    DataTable.prototype.rowTooltip;
    /** @type {?} */
    DataTable.prototype.selectColumn;
    /** @type {?} */
    DataTable.prototype.multiSelect;
    /** @type {?} */
    DataTable.prototype.substituteRows;
    /** @type {?} */
    DataTable.prototype.expandableRows;
    /** @type {?} */
    DataTable.prototype.translations;
    /** @type {?} */
    DataTable.prototype.selectOnRowClick;
    /** @type {?} */
    DataTable.prototype.autoReload;
    /** @type {?} */
    DataTable.prototype.showReloading;
    /** @type {?} */
    DataTable.prototype.noDataMessage;
    /** @type {?} */
    DataTable.prototype.showColumnSelector;
    /** @type {?} */
    DataTable.prototype.autoHidePaging;
    /** @type {?} */
    DataTable.prototype.indexColumnVisible;
    /** @type {?} */
    DataTable.prototype.selectColumnVisible;
    /** @type {?} */
    DataTable.prototype.expandColumnVisible;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype._sortBy;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype._sortAsc;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype._offset;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype._limit;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype.translateSubscription;
    /** @type {?} */
    DataTable.prototype.basicSearch;
    /** @type {?} */
    DataTable.prototype.advanceSearch;
    /** @type {?} */
    DataTable.prototype._reloading;
    /** @type {?} */
    DataTable.prototype.reload;
    /** @type {?} */
    DataTable.prototype._displayParams;
    /** @type {?} */
    DataTable.prototype._scheduledReload;
    /** @type {?} */
    DataTable.prototype.rowClick;
    /** @type {?} */
    DataTable.prototype.rowDoubleClick;
    /** @type {?} */
    DataTable.prototype.headerClick;
    /** @type {?} */
    DataTable.prototype.cellClick;
    /** @type {?} */
    DataTable.prototype.selectedRow;
    /** @type {?} */
    DataTable.prototype.selectedRows;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype._selectAllCheckbox;
    /** @type {?} */
    DataTable.prototype.selectionChange;
    /** @type {?} */
    DataTable.prototype.expandClick;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype._resizeInProgress;
    /** @type {?} */
    DataTable.prototype.resizeLimit;
    /**
     * @type {?}
     * @private
     */
    DataTable.prototype.translate;
}
