// inet-core/esm2015/src/cache-storage.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class CacheStorageService {
    constructor() {
        // Time cache data. Default 5 minutes.
        this.timeCache = 5 * 60 * 1000;
        this.__storage = {};
        this.__queue = {};
    }
    /**
     * @param {?} name
     * @param {?} promise
     * @param {?} resovle
     * @return {?}
     */
    promiseQueue(name, promise, resovle) {
        if (this.__queue[name]) {
            this.__queue[name].push(resovle);
        }
        else {
            this.__queue[name] = [resovle];
            promise(name);
        }
    }
    /**
     * @param {?} name
     * @param {?} context
     * @param {?=} args
     * @return {?}
     */
    resolveQueue(name, context, args) {
        if (this.__queue[name]) {
            this.__queue[name].forEach((/**
             * @param {?} fn
             * @return {?}
             */
            fn => {
                try {
                    fn.apply(context, args);
                }
                catch (e) {
                }
            }));
            delete this.__queue[name];
        }
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getData(key) {
        /** @type {?} */
        const data = this.__storage[key];
        // Check data expired
        if (data && new Date().getTime() - data.time <= data.timeCache) {
            return data.value;
        }
    }
    /**
     * @param {?} key
     * @param {?} data
     * @param {?=} options
     * @return {?}
     */
    setData(key, data, options) {
        /** @type {?} */
        const timeCache = options && options.timeCache || this.timeCache;
        this.__storage[key] = (/** @type {?} */ ({
            time: new Date().getTime(),
            timeCache: timeCache,
            value: data
        }));
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeData(key) {
        delete this.__storage[key];
    }
    /**
     * @return {?}
     */
    clearData() {
        this.__storage = {};
    }
}
CacheStorageService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CacheStorageService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    CacheStorageService.prototype.timeCache;
    /**
     * @type {?}
     * @private
     */
    CacheStorageService.prototype.__storage;
    /**
     * @type {?}
     * @private
     */
    CacheStorageService.prototype.__queue;
}
/**
 * @record
 */
export function CacheStorageOptions() { }
if (false) {
    /** @type {?|undefined} */
    CacheStorageOptions.prototype.timeCache;
    /** @type {?|undefined} */
    CacheStorageOptions.prototype.resolveData;
}
/**
 * @record
 */
export function CacheStorageItem() { }
if (false) {
    /** @type {?} */
    CacheStorageItem.prototype.time;
    /** @type {?} */
    CacheStorageItem.prototype.timeCache;
    /** @type {?} */
    CacheStorageItem.prototype.value;
}
