// inet-core/esm2015/src/loading-indicator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class LoadingIndicatorService {
    constructor() {
        // Create a singleton service
        this.elementId = 'loading-indicator';
        return LoadingIndicatorService.instance = LoadingIndicatorService.instance || this;
    }
    /**
     * @param {?} elementId
     * @return {?}
     */
    setElementId(elementId) {
        this.elementId = elementId;
    }
    /**
     * @return {?}
     */
    getElementId() {
        return this.elementId;
    }
    /**
     * Get DOM of loading indicator
     * @private
     * @return {?}
     */
    getIndicator() {
        if (LoadingIndicatorService.loadingIndicator && LoadingIndicatorService.loadingIndicator.container) {
            return LoadingIndicatorService.loadingIndicator;
        }
        // If Loading indicator not exist
        /** @type {?} */
        const container = (/** @type {?} */ (document.getElementById(this.getElementId())));
        /** @type {?} */
        let loadingElement;
        /** @type {?} */
        let message;
        if (container) {
            /** @type {?} */
            const elements = container.getElementsByTagName('span');
            if (elements && elements.length > 0) {
                loadingElement = (/** @type {?} */ (container.getElementsByTagName('span')[0]));
                if (loadingElement && loadingElement.innerText) {
                    message = loadingElement.innerText;
                }
            }
        }
        // cache Loading indicator
        LoadingIndicatorService.loadingIndicator = new LoadingIndicator(container, loadingElement, message);
        return LoadingIndicatorService.loadingIndicator;
    }
    /**
     * @private
     * @return {?}
     */
    getLoadingIndicator() {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            setTimeout((/**
             * @return {?}
             */
            () => resolve(this.getIndicator())));
        }));
    }
    /**
     * @param {?=} message
     * @return {?}
     */
    showLoading(message) {
        this.getLoadingIndicator().then((/**
         * @param {?} loadingIndicator
         * @return {?}
         */
        loadingIndicator => {
            if (loadingIndicator && loadingIndicator.container) {
                if (loadingIndicator.element) {
                    loadingIndicator.element.innerText = message || loadingIndicator.message || loadingIndicator.container.title;
                }
                loadingIndicator.container.style.display = 'block';
            }
        }));
    }
    /**
     * @return {?}
     */
    hideLoading() {
        this.getLoadingIndicator().then((/**
         * @param {?} loadingIndicator
         * @return {?}
         */
        loadingIndicator => {
            if (loadingIndicator && loadingIndicator.container) {
                loadingIndicator.container.style.display = 'none';
            }
        }));
    }
    /**
     * @return {?}
     */
    complete() {
        this.hideLoading();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.complete();
    }
}
LoadingIndicatorService.instance = null; // Create a singleton service
LoadingIndicatorService.loadingIndicator = null;
LoadingIndicatorService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LoadingIndicatorService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    LoadingIndicatorService.instance;
    /**
     * @type {?}
     * @private
     */
    LoadingIndicatorService.loadingIndicator;
    /** @type {?} */
    LoadingIndicatorService.prototype.elementId;
}
export class LoadingIndicator {
    /**
     * @param {?} container
     * @param {?} el
     * @param {?=} msg
     */
    constructor(container, el, msg = '') {
        this.container = container;
        this.element = el;
        this.message = msg;
    }
}
if (false) {
    /** @type {?} */
    LoadingIndicator.prototype.container;
    /** @type {?} */
    LoadingIndicator.prototype.element;
    /** @type {?} */
    LoadingIndicator.prototype.message;
}
