// inet-ui/esm2015/src/file/viewer.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input } from '@angular/core';
import { ViewerService } from './viewer.service';
export class ViewerDirective {
    /**
     * @param {?} viewerService
     */
    constructor(viewerService) {
        this.viewerService = viewerService;
    }
    /**
     * @return {?}
     */
    onClick() {
        this.viewerService.setFiles(this.files);
        this.viewerService.open(this.ext, this.docId, this.extras);
    }
}
ViewerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[appViewer]'
            },] }
];
/** @nocollapse */
ViewerDirective.ctorParameters = () => [
    { type: ViewerService }
];
ViewerDirective.propDecorators = {
    ext: [{ type: Input, args: ['ext',] }],
    docId: [{ type: Input, args: ['docId',] }],
    files: [{ type: Input, args: ['files',] }],
    extras: [{ type: Input, args: ['extras',] }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /** @type {?} */
    ViewerDirective.prototype.ext;
    /** @type {?} */
    ViewerDirective.prototype.docId;
    /** @type {?} */
    ViewerDirective.prototype.files;
    /** @type {?} */
    ViewerDirective.prototype.extras;
    /**
     * @type {?}
     * @private
     */
    ViewerDirective.prototype.viewerService;
}
