// inet-ui/esm2015/src/file/viewer.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { HttpClientService } from "inet-core";
import { Subject } from "rxjs";
import { FileFormatService } from "./file-format.service";
export class ViewerService {
    /**
     * @param {?} router
     * @param {?} formatService
     * @param {?} http
     */
    constructor(router, formatService, http) {
        this.router = router;
        this.formatService = formatService;
        this.http = http;
        this.routerPath = 'viewer'; //Navigate with the router path
        //Navigate with the router path
        this.downloadUrl = iNet.getPUrl('docx/download');
        // Observable sources
        this.emitLoadSource = new Subject();
        // Observable when document loaded
        this.loadEmitted = this.emitLoadSource.asObservable();
        return ViewerService.instance = ViewerService.instance || this;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    setRouterPath(v) {
        this.routerPath = v;
    }
    /**
     * @return {?}
     */
    getRouterPath() {
        return this.routerPath;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    setDownloadUrl(v) {
        this.downloadUrl = v;
    }
    /**
     * @return {?}
     */
    getDownloadUrl() {
        return this.downloadUrl;
    }
    /**
     * @param {?} ext
     * @param {?} docId
     * @param {?=} extras
     * @return {?}
     */
    open(ext, docId, extras) {
        if (this.hasEdit(ext) || (this.hasView(ext) && this.hasViewerInBrowser())) {
            this.router.navigate([this.getRouterPath(), ext, docId], extras);
        }
        else { //error pages
            this.router.navigate([this.getRouterPath(), docId]);
        }
    }
    /**
     * @param {?} name
     * @return {?}
     */
    hasNameInPlugin(name) {
        /** @type {?} */
        let plugins = navigator.plugins || [];
        /** @type {?} */
        let plugin = {};
        for (let i = 0; i < plugins.length; i++) {
            plugin = plugins[i] || {};
            if ((plugin['name']).search(name) > -1) {
                return true;
            }
        }
        return false;
    }
    /**
     * @return {?}
     */
    hasViewerInBrowser() {
        if (iNet.isWebKit) { //Chrome or Safari
            if (iNet.isChrome) {
                return this.hasNameInPlugin('PDF Viewer');
            }
            else if (iNet.isSafari) {
                return (this.hasNameInPlugin('Adobe Acrobat') || this.hasNameInPlugin('PDF'));
            }
        }
        else if (iNet.isGecko) { //Firefox
            return this.hasNameInPlugin('Adobe Acrobat');
        }
        return false;
    }
    /**
     * @param {?} ext
     * @return {?}
     */
    hasEdit(ext) {
        return this.formatService.getEditFormats().indexOf(ext) > -1;
    }
    /**
     * @param {?} ext
     * @return {?}
     */
    hasView(ext) {
        return this.formatService.getViewFormats().indexOf(ext) > -1;
    }
    /**
     * @param {?} docId
     * @return {?}
     */
    downloadById(docId) {
        this.http.downloadFile(this.getDownloadUrl(), { docID: docId });
    }
    /**
     * @param {?} v
     * @return {?}
     */
    setFiles(v) {
        ViewerService.files = v;
    }
    /**
     * @return {?}
     */
    getFiles() {
        return ViewerService.files;
    }
    /**
     * Returns true if this url is viewer module, false otherwise
     * @param {?} url - the give URL
     * @return {?}
     */
    isViewerModuleByUrl(url) {
        return (url.indexOf('/' + this.getRouterPath()) === 0);
    }
    /**
     * Fire event when document loaded
     *
     * @param {?} $event
     * @param {?} viewer
     * @return {?}
     */
    sendEvent($event, viewer) {
        this.emitLoadSource.next({ $event, viewer });
    }
}
ViewerService.instance = null;
ViewerService.files = [];
ViewerService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ViewerService.ctorParameters = () => [
    { type: Router },
    { type: FileFormatService },
    { type: HttpClientService }
];
if (false) {
    /** @type {?} */
    ViewerService.instance;
    /**
     * @type {?}
     * @private
     */
    ViewerService.files;
    /**
     * @type {?}
     * @private
     */
    ViewerService.prototype.routerPath;
    /**
     * @type {?}
     * @private
     */
    ViewerService.prototype.downloadUrl;
    /**
     * @type {?}
     * @private
     */
    ViewerService.prototype.emitLoadSource;
    /** @type {?} */
    ViewerService.prototype.loadEmitted;
    /**
     * @type {?}
     * @private
     */
    ViewerService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    ViewerService.prototype.formatService;
    /**
     * @type {?}
     * @private
     */
    ViewerService.prototype.http;
}
