// inet-ui/esm5/src/select2/app-select2.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef } from '@angular/core';
import { Input, Output } from '@angular/core';
import { EventEmitter } from '@angular/core';
import { Directive } from '@angular/core';
var AppSelect2Directive = /** @class */ (function () {
    function AppSelect2Directive(_el) {
        this._el = _el;
        this.onChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    AppSelect2Directive.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.element = $(this._el.nativeElement).select2(this.options).change((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.onChange.emit(e.target.value);
        }));
        this.updateSelect();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AppSelect2Directive.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.options) {
            this.updateSelect();
        }
        if (changes.initValue) {
            this.updateValue();
        }
    };
    /**
     * @return {?}
     */
    AppSelect2Directive.prototype.updateSelect = /**
     * @return {?}
     */
    function () {
        if (this.element && this.options) {
            this.element.empty().select2(this.options);
            this.updateValue();
        }
    };
    /**
     * @return {?}
     */
    AppSelect2Directive.prototype.updateValue = /**
     * @return {?}
     */
    function () {
        if (this.element && this.initValue) {
            this.element.val(this.initValue).trigger('change');
        }
    };
    AppSelect2Directive.decorators = [
        { type: Directive, args: [{
                    selector: 'select[appSelect2]'
                },] }
    ];
    /** @nocollapse */
    AppSelect2Directive.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AppSelect2Directive.propDecorators = {
        initValue: [{ type: Input }],
        options: [{ type: Input }],
        onChange: [{ type: Output }]
    };
    return AppSelect2Directive;
}());
export { AppSelect2Directive };
if (false) {
    /** @type {?} */
    AppSelect2Directive.prototype.initValue;
    /** @type {?} */
    AppSelect2Directive.prototype.options;
    /** @type {?} */
    AppSelect2Directive.prototype.onChange;
    /** @type {?} */
    AppSelect2Directive.prototype.element;
    /**
     * @type {?}
     * @private
     */
    AppSelect2Directive.prototype._el;
}
