// inet-ui/esm5/src/select-user/select-user.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ModalSelectComponent } from "./modal-select/modal-select.component";
import { CoreService } from "inet-core";
var SelectUserComponent = /** @class */ (function () {
    function SelectUserComponent(coreService) {
        this.coreService = coreService;
        this.id = iNet.generateId();
        this.selectUsers = [];
        this.onAdd = new EventEmitter();
        this.isMouseEnter = false;
        this.isFocus = false;
        this.users = [];
    }
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.valid = /**
     * @return {?}
     */
    function () {
        return this.isValid && this.selectUsers.length > 0;
    };
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.clearValid = /**
     * @return {?}
     */
    function () {
        this.isFocus = false;
    };
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.validInput = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.valid()) {
                _this.isFocus = false;
            }
            else {
                _this.isFocus = true;
            }
        }), 250);
    };
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.searchUser = /**
     * @return {?}
     */
    function () {
        var _this = this;
        clearTimeout(this.timer);
        this.timer = setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var value = _this.getKeyword();
            if (value === _this.searchValue) {
                return;
            }
            if (!value) {
                _this.users.length = 0;
                _this.searchValue = value;
                return;
            }
            _this.coreService.searchFirmAccount({ keyword: value, pageSize: 10 }, (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                /** @type {?} */
                var users = data && data['items'] || [];
                _this.setUser(users);
            }));
        }), 300);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    SelectUserComponent.prototype.removeUser = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        for (var i = 0; i < this.selectUsers.length; i++) {
            if (this.selectUsers[i].username === item.username) {
                this.selectUsers.splice(i, 1);
                this.validInput();
                this.onAdd.emit(item);
                return;
            }
        }
    };
    /**
     * @private
     * @return {?}
     */
    SelectUserComponent.prototype.setDefault = /**
     * @private
     * @return {?}
     */
    function () {
        this.inputUser.nativeElement.value = '';
        this.searchValue = '';
        this.users = [];
        this.selectUsers = [];
    };
    /**
     * @param {?} item
     * @return {?}
     */
    SelectUserComponent.prototype.addUser = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var e_1, _a;
        if (this.isSingle) {
            this.selectUsers[0] = item;
        }
        else {
            try {
                for (var _b = tslib_1.__values(this.selectUsers), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var selectMember = _c.value;
                    if (selectMember.username === item.username) {
                        return;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            this.selectUsers.push(item);
        }
        this.inputUser.nativeElement.value = '';
        this.searchValue = '';
        this.focusInput();
        this.validInput();
    };
    /**
     * @param {?} e
     * @param {?} el
     * @return {?}
     */
    SelectUserComponent.prototype.clickUser = /**
     * @param {?} e
     * @param {?} el
     * @return {?}
     */
    function (e, el) {
        if (e.target.isSameNode(el)) {
            this.focusInput();
        }
    };
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.removeLastUser = /**
     * @return {?}
     */
    function () {
        this.selectUsers.splice(-1);
    };
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.openModal = /**
     * @return {?}
     */
    function () {
        this.modalSelect.show();
    };
    /**
     * @private
     * @param {?} user
     * @return {?}
     */
    SelectUserComponent.prototype.hasUser = /**
     * @private
     * @param {?} user
     * @return {?}
     */
    function (user) {
        var e_2, _a;
        try {
            for (var _b = tslib_1.__values(this.selectUsers), _c = _b.next(); !_c.done; _c = _b.next()) {
                var selectMember = _c.value;
                if (selectMember.username === user.username) {
                    return true;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return false;
    };
    /**
     * @private
     * @param {?} users
     * @return {?}
     */
    SelectUserComponent.prototype.setUser = /**
     * @private
     * @param {?} users
     * @return {?}
     */
    function (users) {
        var _this = this;
        this.users = users.filter((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return !_this.hasUser(item); }));
    };
    /**
     * @private
     * @return {?}
     */
    SelectUserComponent.prototype.focusInput = /**
     * @private
     * @return {?}
     */
    function () {
        this.inputUser.nativeElement.focus();
    };
    /**
     * @private
     * @return {?}
     */
    SelectUserComponent.prototype.getKeyword = /**
     * @private
     * @return {?}
     */
    function () {
        return this.inputUser.nativeElement.value.trim();
    };
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.onMouseEnter = /**
     * @return {?}
     */
    function () {
        this.isMouseEnter = true;
    };
    /**
     * @return {?}
     */
    SelectUserComponent.prototype.onMouseDown = /**
     * @return {?}
     */
    function () {
        this.isMouseEnter = false;
    };
    SelectUserComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-select-user',
                    template: "<div class=\"row\">\n    <div class=\"col-sm-12\">\n        <div class=\"border select-multiple-content\"\n             #shareUserContainer\n             (click)=\"clickUser($event, shareUserContainer)\"\n             [ngClass]=\"{'select-valid': isValid && isFocus && selectUsers?.length < 1}\"\n             (mouseenter)=\"onMouseEnter()\"\n             (mouseleave)=\"onMouseDown()\">\n            <span *ngIf=\"isValid && isFocus && isMouseEnter && selectUsers?.length < 1\" class=\"select-tooltip\">\n                B\u1EA1n c\u1EA7n ch\u1ECDn \u00EDt nh\u1EA5t m\u1ED9t ng\u01B0\u1EDDi.\n            </span>\n            <span *ngFor=\"let user of selectUsers\" class=\"select-multiple-item border\">\n                <img userAvatar [usercode]=\"user?.username\" width=\"28\" height=\"28\">\n                <span>{{user.fullname}}</span>\n                <span (click)=\"removeUser(user)\" class=\"badge badge-muted text-muted rounded-circle m-1\">\n                    <i class=\"fa fa-times\" style=\"cursor: pointer;\"></i>\n                </span>\n            </span>\n            <span class=\"autocomplete-container\">\n                <input type=\"text\" class=\"auto-tag-input\" style=\"width: 80px !important;\"\n                       #inputUser\n                       appAutocompleteInput\n                       [autoList]=\"autoList\"\n                       (input)=\"searchUser()\"\n                       (onBackSpaceRemove)=\"removeLastUser()\"\n                       (blur)=\"validInput()\">\n            </span>\n        </div>\n        <ul class=\"list-group autocomplete-list\"\n            appAutocompleteList\n            #autoList=\"appAutocompleteList\">\n            <li *ngFor=\"let item of users\" (click)=\"addUser(item)\" class=\"list-group-item auto-item-image rounded-0\">\n                <img userAvatar [usercode]=\"item?.username\">\n                <div>{{item.fullname}}</div>\n            </li>\n        </ul>\n        <i (click)=\"openModal()\" class=\"icon-user\" style=\"position: absolute;top: 0;right: 10px;padding: 16px;cursor: pointer;\"></i>\n    </div>\n    <app-modal-select\n        [selectUsers]=\"selectUsers\"\n        [isSingle]=\"isSingle\">\n    </app-modal-select>\n</div>",
                    styles: [".autocomplete-container{position:relative}.autocomplete-list{display:none;position:absolute;left:0;right:0;z-index:1;overflow:auto;max-height:500px}.autocomplete-list>li{padding:10px;cursor:pointer;border-width:0 1px}.autocomplete-list>li:first-child{border-top-width:1px}.autocomplete-list>li:last-child{border-bottom-width:1px}.auto-item-image i,.auto-item-image img{float:left;width:30px;height:30px;margin-right:10px}.auto-item-image i{text-align:center;line-height:30px;background:rgba(0,0,0,.05);color:rgba(0,0,0,.5)}.auto-item-image div{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.auto-tag{padding:2px 7px 7px;min-height:38px;height:auto}.auto-tag-input{border:0;padding:0;margin-top:5px;height:24px;line-height:22px}.auto-tag-input:focus{outline:0}.auto-tag-item{font-size:14px;line-height:24px;margin-right:5px;margin-top:5px;padding:0 5px;float:left;background:#e9ebee;border:1px solid rgba(0,0,0,.2);border-radius:2px}.auto-tag-item i{float:right;margin-right:-5px;height:24px;width:24px;line-height:24px;text-align:center;cursor:pointer}", ".select-multiple-content{width:100%;min-height:46px;padding:3px}.select-multiple-item{color:#343434;background:#e0e0e0;margin:4px;display:inline-block;font-size:.875em}.list-group-item img{height:34px!important;width:34px!important}.badge-muted{color:#e0e0e0!important;background:#9e9e9e}.auto-tag-input{margin-top:0!important;height:37px!important}.auto-item-image>img{margin-top:-5px}.autocomplete-list{padding:0 15px;overflow:hidden}.select-valid{border:1px solid red!important}.select-tooltip{position:absolute;bottom:150%;margin-bottom:-15px;padding:7px;width:50%;left:25%;border-radius:3px;background-color:#dc3545;color:#fff;content:attr(data-tooltip);text-align:center;font-size:14px;line-height:1.2}"]
                }] }
    ];
    /** @nocollapse */
    SelectUserComponent.ctorParameters = function () { return [
        { type: CoreService }
    ]; };
    SelectUserComponent.propDecorators = {
        id: [{ type: Input, args: ['id',] }],
        selectUsers: [{ type: Input, args: ['selectUsers',] }],
        isSingle: [{ type: Input, args: ['isSingle',] }],
        isValid: [{ type: Input, args: ['isValid',] }],
        onAdd: [{ type: Output }],
        modalSelect: [{ type: ViewChild, args: [ModalSelectComponent,] }],
        inputUser: [{ type: ViewChild, args: ['inputUser',] }]
    };
    return SelectUserComponent;
}());
export { SelectUserComponent };
if (false) {
    /** @type {?} */
    SelectUserComponent.prototype.id;
    /** @type {?} */
    SelectUserComponent.prototype.selectUsers;
    /** @type {?} */
    SelectUserComponent.prototype.isSingle;
    /** @type {?} */
    SelectUserComponent.prototype.isValid;
    /** @type {?} */
    SelectUserComponent.prototype.onAdd;
    /** @type {?} */
    SelectUserComponent.prototype.modalSelect;
    /** @type {?} */
    SelectUserComponent.prototype.inputUser;
    /** @type {?} */
    SelectUserComponent.prototype.isMouseEnter;
    /** @type {?} */
    SelectUserComponent.prototype.isFocus;
    /** @type {?} */
    SelectUserComponent.prototype.users;
    /**
     * @type {?}
     * @private
     */
    SelectUserComponent.prototype.timer;
    /**
     * @type {?}
     * @private
     */
    SelectUserComponent.prototype.searchValue;
    /**
     * @type {?}
     * @private
     */
    SelectUserComponent.prototype.coreService;
}
