// inet-calendar/esm5/src/CalendarUtils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as moment_ from "moment";
/** @type {?} */
var moment = moment_;
moment.locale('vi');
var CalendarUtils = /** @class */ (function () {
    function CalendarUtils() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    CalendarUtils.getRepeatText = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var display = '';
        for (var i = 0; i < CalendarUtils.repeatTypes.length; i++) {
            if (CalendarUtils.repeatTypes[i].value === data.rrmode) {
                display = CalendarUtils.repeatTypes[i].label;
            }
        }
        if (data.rrmode === 'WEEKLY') {
            display += ' - ' + CalendarUtils._getTextRepeatWeek(data.wkdays);
        }
        return display;
    };
    /**
     * @param {?} wkdays
     * @return {?}
     */
    CalendarUtils._getTextRepeatWeek = /**
     * @param {?} wkdays
     * @return {?}
     */
    function (wkdays) {
        if (wkdays) {
            return CalendarUtils.weekDays.filter((/**
             * @param {?} day
             * @return {?}
             */
            function (day) {
                return wkdays.indexOf(day.value) > -1;
            })).map((/**
             * @param {?} day
             * @return {?}
             */
            function (day) {
                return day.title;
            })).join(', ');
        }
        return '';
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalendarUtils.displayDay = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var str = moment(date).format('dddd, d MMMM');
        if (new Date().getFullYear() !== date.getFullYear()) {
            str += ' năm ' + date.getFullYear();
        }
        return str.slice(0, 1).toUpperCase() + str.slice(1);
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalendarUtils.displayDate = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return moment(date).format('[Ngày] LL');
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalendarUtils.displayWeek = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return moment(date).format('[Tuần] w [năm] Y');
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalendarUtils.displayMonth = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return moment(date).format('[Tháng] MM [năm] Y');
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalendarUtils.displayTime = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return moment(date).format('HH:mm');
    };
    /**
     * @param {?} element
     * @param {?} e
     * @param {?=} padding
     * @return {?}
     */
    CalendarUtils.computePosByPoint = /**
     * @param {?} element
     * @param {?} e
     * @param {?=} padding
     * @return {?}
     */
    function (element, e, padding) {
        /** @type {?} */
        var bounds = element.getBoundingClientRect();
        /** @type {?} */
        var halfHeight = bounds.height / 2;
        /** @type {?} */
        var halfWidth = bounds.width / 2;
        /** @type {?} */
        var direction;
        /** @type {?} */
        var pos = {
            left: e.x - halfWidth,
            top: e.y - halfHeight
        };
        padding = padding || 5;
        // Position out of width
        if (innerWidth < pos.left + bounds.width + padding) {
            pos.left = innerWidth - padding - bounds.width;
        }
        else if (pos.left < padding) {
            pos.left = padding;
        }
        // Position out of height
        if (innerHeight < pos.top + bounds.height + padding) {
            pos.top = innerHeight - padding - bounds.height;
        }
        else if (pos.top < padding) {
            pos.top = padding;
        }
        // Detect best position
        /** @type {?} */
        var deltaTop = e.pageY;
        /** @type {?} */
        var deltaBottom = innerHeight - e.pageX;
        /** @type {?} */
        var deltaLeft = e.pageX;
        /** @type {?} */
        var deltaRight = innerWidth - e.pageY;
        /** @type {?} */
        var deltaMax = Math.max(deltaTop, deltaRight, deltaBottom, deltaLeft);
        switch (deltaMax) {
            case deltaTop:
                direction = 'top';
                pos.top -= halfHeight + padding;
                break;
            case deltaBottom:
                direction = 'bottom';
                pos.top += halfHeight + padding;
                break;
            case deltaLeft:
                direction = 'left';
                pos.left -= halfWidth + padding;
                break;
            case deltaRight:
                direction = 'right';
                pos.left += halfWidth + padding;
                break;
        }
        return {
            direction: direction,
            position: pos,
            elementBounds: bounds
        };
    };
    /**
     * @param {?} element
     * @param {?} target
     * @param {?} jsEvent
     * @param {?=} padding
     * @return {?}
     */
    CalendarUtils.computePosByTargetEl = /**
     * @param {?} element
     * @param {?} target
     * @param {?} jsEvent
     * @param {?=} padding
     * @return {?}
     */
    function (element, target, jsEvent, padding) {
        /** @type {?} */
        var elementBounds = element.getBoundingClientRect();
        /** @type {?} */
        var targetBounds = target.getBoundingClientRect();
        /** @type {?} */
        var pos = {};
        /** @type {?} */
        var direction = {};
        padding = padding || 10;
        /** @type {?} */
        var deltaRect = {
            top: targetBounds.top,
            right: innerWidth - targetBounds.right,
            bottom: innerHeight - targetBounds.bottom,
            left: targetBounds.left
        };
        /** @type {?} */
        var maxX = Math.max(deltaRect.left, deltaRect.right);
        /** @type {?} */
        var maxY = Math.max(deltaRect.top, deltaRect.bottom);
        // Get click point when left and right not enough modal
        if (elementBounds.width > maxX) {
            deltaRect.left = jsEvent.pageX;
            deltaRect.right = innerWidth - jsEvent.pageX;
        }
        // Get click point when top and bottom not enough modal
        if (elementBounds.height > maxY) {
            deltaRect.top = jsEvent.pageY;
            deltaRect.bottom = innerHeight - jsEvent.pageY;
        }
        // X direction
        if (deltaRect.left > deltaRect.right) {
            direction.left = true;
            pos.left = deltaRect.left - elementBounds.width;
        }
        else {
            direction.right = true;
            pos.left = innerWidth - deltaRect.right;
        }
        // Y direction
        if (deltaRect.top > deltaRect.bottom) {
            direction.top = true;
            pos.top = deltaRect.top - elementBounds.height;
        }
        else {
            direction.bottom = true;
            pos.top = innerHeight - deltaRect.bottom;
        }
        // Best dimension to display
        if (maxX * elementBounds.width > maxY * elementBounds.height) {
            // By x
            /** @type {?} */
            var targetHeight = innerHeight - deltaRect.bottom - deltaRect.top;
            if (direction.top) {
                pos.top += targetHeight;
            }
            else {
                pos.top -= targetHeight;
            }
        }
        else {
            // By y
            /** @type {?} */
            var targetWidth = innerWidth - deltaRect.left - deltaRect.right;
            if (direction.left) {
                pos.left += targetWidth;
            }
            else {
                pos.left -= targetWidth;
            }
        }
        // Out of width
        if (pos.left < padding) {
            pos.left = padding;
        }
        else if (pos.left + elementBounds.width > innerWidth) {
            pos.left = innerWidth - elementBounds.width - padding;
        }
        // Out of height
        if (pos.top < padding) {
            pos.top = padding;
        }
        else if (pos.top + elementBounds.height > innerHeight) {
            pos.top = innerHeight - elementBounds.height - padding;
        }
        // compute with document scroll
        pos.top += document.scrollingElement.scrollTop;
        pos.left += document.scrollingElement.scrollLeft;
        return pos;
    };
    CalendarUtils.repeatTypes = [
        { label: 'Hàng ngày', value: 'DAILY' },
        { label: 'Hàng tuần', value: 'WEEKLY' },
        { label: 'Hàng tháng', value: 'MONTHLY' }
    ];
    CalendarUtils.weekDays = [
        { label: 'T2', title: 'Thứ hai', value: 'MO' },
        { label: 'T3', title: 'Thứ ba', value: 'TU' },
        { label: 'T4', title: 'Thứ tư', value: 'WE' },
        { label: 'T5', title: 'Thứ năm', value: 'TH' },
        { label: 'T6', title: 'Thứ sáu', value: 'FR' },
        { label: 'T7', title: 'Thứ bảy', value: 'SA' },
        { label: 'CN', title: 'Chủ nhật', value: 'SU' }
    ];
    return CalendarUtils;
}());
export { CalendarUtils };
if (false) {
    /** @type {?} */
    CalendarUtils.repeatTypes;
    /** @type {?} */
    CalendarUtils.weekDays;
}
