// inet-calendar/esm2015/src/attachment-list/attachment-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild } from '@angular/core';
import { xCalendar } from "../utils/xCalendar";
import { CalendarService } from "../calendar.service";
import { BsModalService } from "ngx-bootstrap";
import { FileListComponent } from "../remote-module/filelist/list/file-list.component";
export class AttachmentListComponent {
    /**
     * @param {?} calendarService
     * @param {?} modalService
     */
    constructor(calendarService, modalService) {
        this.calendarService = calendarService;
        this.modalService = modalService;
        this.files = [];
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.calEvent) {
            this._attachmentToFiles();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.modalRef) {
            this.modalRef.hide();
        }
    }
    /**
     * @param {?} file
     * @param {?} event
     * @return {?}
     */
    viewAttachment(file, event) {
        window.open(xCalendar.getFileViewUrl(file.data, null, event));
    }
    /**
     * @param {?} file
     * @param {?} template
     * @return {?}
     */
    removeAttachment(file, template) {
        if (file.id) {
            this._file = file;
            this.confirmMessage = 'Tệp tin <b>"' + file.name + '"</b> sẽ bị xoá?';
            this.modalRef = this.modalService.show(template);
        }
        else {
            this.fileList.removeFile(file);
        }
    }
    /**
     * @return {?}
     */
    removeFile() {
        if (!this._file) {
            return;
        }
        this.modalRef.hide();
        this.calendarService.deleteAttachment({ fileID: this._file.id, element: this.calEvent.uuid }, (/**
         * @return {?}
         */
        () => { }));
        this.fileList.removeFile(this._file);
        this._file = null;
    }
    /**
     * @private
     * @return {?}
     */
    _attachmentToFiles() {
        if (this.calEvent && this.calEvent.attachments) {
            this.files = this.calEvent.attachments.map((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                return {
                    name: item.file,
                    type: item.mimetype,
                    size: item.size,
                    id: item.gridfsUUID,
                    data: item
                };
            }));
        }
        else {
            this.files = [];
        }
    }
}
AttachmentListComponent.decorators = [
    { type: Component, args: [{
                selector: 'calendar-attachment-list',
                template: "<file-list [template]=\"fileItem\" [files]=\"files\" [fileEl]=\"fileEl\" [removable]=\"removable\"></file-list>\n<ng-template #fileItem let-file=\"file\">\n    <div class=\"file-item\" [title]=\"file.name\">\n        <img [src]=\"file.name | fileIcon\" class=\"file-item__image\">\n        <div class=\"file-item__name text-primary\" (click)=\"viewAttachment(file, $event)\">{{file.name}}</div>\n        <i *ngIf=\"removable\" (click)=\"removeAttachment(file, template)\"\n           class=\"fa fa-trash file-item__remove text-danger\"></i>\n    </div>\n</ng-template>\n<ng-template #template>\n    <div class=\"modal-header\">\n        <h4 class=\"modal-title pull-left\">Th\u00F4ng b\u00E1o</h4>\n        <button type=\"button\" class=\"close pull-right\" aria-label=\"Close\" (click)=\"modalRef.hide()\">\n            <span aria-hidden=\"true\">&times;</span>\n        </button>\n    </div>\n    <div class=\"modal-body\" [innerHTML]=\"confirmMessage\"></div>\n    <div class=\"modal-footer\">\n        <button type=\"button\" class=\"btn btn-danger\" (click)=\"removeFile()\">X\u00F3a</button>\n        <button type=\"button\" class=\"btn btn-secondary\" (click)=\"modalRef.hide()\">\u0110\u00F3ng</button>\n    </div>\n</ng-template>",
                styles: [".file-item{height:24px;line-height:24px;overflow:hidden;font-size:13px;cursor:pointer;margin-bottom:5px}.file-item__image{float:left;width:24px;height:24px;margin-right:10px}.file-item__name{float:left;max-width:80%;white-space:nowrap;overflow:hidden}.file-item__remove{width:24px;height:24px;line-height:24px;text-align:center}"]
            }] }
];
/** @nocollapse */
AttachmentListComponent.ctorParameters = () => [
    { type: CalendarService },
    { type: BsModalService }
];
AttachmentListComponent.propDecorators = {
    calEvent: [{ type: Input }],
    fileEl: [{ type: Input }],
    removable: [{ type: Input }],
    fileList: [{ type: ViewChild, args: [FileListComponent,] }]
};
if (false) {
    /** @type {?} */
    AttachmentListComponent.prototype.calEvent;
    /** @type {?} */
    AttachmentListComponent.prototype.fileEl;
    /** @type {?} */
    AttachmentListComponent.prototype.removable;
    /** @type {?} */
    AttachmentListComponent.prototype.fileList;
    /** @type {?} */
    AttachmentListComponent.prototype.files;
    /** @type {?} */
    AttachmentListComponent.prototype.modalRef;
    /** @type {?} */
    AttachmentListComponent.prototype.confirmMessage;
    /**
     * @type {?}
     * @private
     */
    AttachmentListComponent.prototype._file;
    /**
     * @type {?}
     * @private
     */
    AttachmentListComponent.prototype.calendarService;
    /**
     * @type {?}
     * @private
     */
    AttachmentListComponent.prototype.modalService;
}
