// inet-social/esm5/src/social-friend/group-create/group-create.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Output, ViewChild } from '@angular/core';
import { ListUserComponent } from '../list-user/list-user.component';
import { SocialService } from '../../social.service';
var GroupCreateComponent = /** @class */ (function () {
    function GroupCreateComponent(socialService) {
        this.socialService = socialService;
        this.onCreate = new EventEmitter();
    }
    /**
     * @return {?}
     */
    GroupCreateComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._input = this.inputGroup.nativeElement;
        document.body.appendChild(this.modal._element.nativeElement);
    };
    /**
     * @return {?}
     */
    GroupCreateComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.modal._element.nativeElement.remove();
    };
    /**
     * @return {?}
     */
    GroupCreateComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this._input.value = '';
        this.listUser.resetData();
        this.modal.show();
    };
    /**
     * @return {?}
     */
    GroupCreateComponent.prototype.onShown = /**
     * @return {?}
     */
    function () {
        this._input.focus();
        this.listUser.onSearch();
    };
    /**
     * @return {?}
     */
    GroupCreateComponent.prototype.onSave = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var params = {
            category: this._input.value.trim(),
            member: this.listUser.selectMembers.map((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.username; })).join(',')
        };
        if (!params.category) {
            this._input.focus();
            return;
        }
        if (!params.member) {
            this.listUser.focusInput();
            return;
        }
        this.socialService.groupMemberCreate(params, (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.onCreate.emit(data);
        }));
        this.modal.hide();
    };
    /**
     * @return {?}
     */
    GroupCreateComponent.prototype.onCancel = /**
     * @return {?}
     */
    function () {
        this.modal.hide();
    };
    GroupCreateComponent.decorators = [
        { type: Component, args: [{
                    selector: '[socialGroupCreate]',
                    template: "<div class=\"modal fade\" bsModal #modal=\"bs-modal\" (onShown)=\"onShown()\" tabindex=\"-1\" role=\"dialog\">\n  <div class=\"modal-dialog\">\n    <div class=\"modal-content\">\n      <div class=\"modal-header\">\n        <h5 class=\"modal-title pull-left\">T\u1EA1o nh\u00F3m</h5>\n        <button type=\"button\" class=\"close pull-right\" (click)=\"modal.hide()\" tabindex=\"-1\">\n          <span aria-hidden=\"true\">&times;</span>\n        </button>\n      </div>\n      <div class=\"modal-body\">\n        <div class=\"form-group\">\n          <label>T\u00EAn nh\u00F3m</label>\n          <input #inputGroup type=\"text\" class=\"form-control\">\n        </div>\n        <div class=\"form-group\">\n          <label>Th\u00EAm th\u00E0nh vi\u00EAn</label>\n          <div socialListUser #listUser></div>\n        </div>\n      </div>\n      <div class=\"modal-footer\">\n        <button type=\"button\" class=\"btn btn-primary\" (click)=\"onSave()\">T\u1EA1o nh\u00F3m</button>\n        <button type=\"button\" class=\"btn btn-secondary\" (click)=\"onCancel()\">\u0110\u00F3ng</button>\n      </div>\n    </div>\n  </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    GroupCreateComponent.ctorParameters = function () { return [
        { type: SocialService }
    ]; };
    GroupCreateComponent.propDecorators = {
        onCreate: [{ type: Output }],
        modal: [{ type: ViewChild, args: ['modal',] }],
        inputGroup: [{ type: ViewChild, args: ['inputGroup',] }],
        listUser: [{ type: ViewChild, args: ['listUser',] }]
    };
    return GroupCreateComponent;
}());
export { GroupCreateComponent };
if (false) {
    /** @type {?} */
    GroupCreateComponent.prototype.onCreate;
    /** @type {?} */
    GroupCreateComponent.prototype.modal;
    /** @type {?} */
    GroupCreateComponent.prototype.inputGroup;
    /** @type {?} */
    GroupCreateComponent.prototype.listUser;
    /**
     * @type {?}
     * @private
     */
    GroupCreateComponent.prototype._input;
    /**
     * @type {?}
     * @private
     */
    GroupCreateComponent.prototype.socialService;
}
