// inet-ui/esm5/src/user-profile/access-role/access-role.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { GlobalContactService } from "../global-contact.service";
import { ErrorMessage } from "inet-core";
var AccessRoleComponent = /** @class */ (function () {
    function AccessRoleComponent(glService) {
        this.glService = glService;
        this.arrayRoleUser = [];
    }
    /**
     * @return {?}
     */
    AccessRoleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.loadAccessRole();
    };
    /**
     * @return {?}
     */
    AccessRoleComponent.prototype.loadAccessRole = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.glService.getAccessRoles().subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            if (result.type !== ErrorMessage.TYPE) {
                _this.arrayRoleUser = result.items;
            }
        }));
    };
    AccessRoleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-access-role',
                    template: "<data-table [items]=\"arrayRoleUser\"\n            [itemCount]=\"arrayRoleUser.length\"\n            [sortAsc]=\"true\"\n            [limit]=\"9999\"\n            [header]=\"false\"\n            [selectColumn]=\"false\"\n            [multiSelect]=\"false\"\n            (reload)=\"loadAccessRole()\"\n            [expandableRows]=\"false\"\n            [pagination]=\"false\"\n            [pagination_input]=\"true\"\n            [pagination_range]=\"true\"\n            [indexColumnHeader]=\"'#'\"\n            [indexColumn]=\"true\"\n            [selectOnRowClick]=\"true\"\n            [showReloading]=\"false\">\n  <data-table-column\n      [property]=\"'name'\"\n      [header]=\"'COMMON.MODULE.USER_PROFILE.GROUP_NAME' | translate\"\n      [width]=\"200\">\n  </data-table-column>\n  <data-table-column\n      [property]=\"'description'\"\n      [header]=\"'COMMON.MODULE.USER_PROFILE.DESCRIPTION' | translate\">\n  </data-table-column>\n</data-table>"
                }] }
    ];
    /** @nocollapse */
    AccessRoleComponent.ctorParameters = function () { return [
        { type: GlobalContactService }
    ]; };
    return AccessRoleComponent;
}());
export { AccessRoleComponent };
if (false) {
    /** @type {?} */
    AccessRoleComponent.prototype.arrayRoleUser;
    /**
     * @type {?}
     * @private
     */
    AccessRoleComponent.prototype.glService;
}
