// inet-ui/esm2015/src/interceptor/interceptor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgModule } from '@angular/core';
import { HttpResponse, HttpErrorResponse } from '@angular/common/http';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import 'rxjs/add/operator/do';
import { CoreService, HttpClientService, NotificationService, SecurityService } from "inet-core";
export class HttpsRequestInterceptor {
    /**
     * @param {?} securityService
     * @param {?} coreService
     * @param {?} httpClientService
     * @param {?} notifyService
     */
    constructor(securityService, coreService, httpClientService, notifyService) {
        this.securityService = securityService;
        this.coreService = coreService;
        this.httpClientService = httpClientService;
        this.notifyService = notifyService;
        return HttpsRequestInterceptor.instance = HttpsRequestInterceptor.instance || this;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        return next.handle(req).do((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            if (event instanceof HttpResponse) {
                // do stuff with response if you want
                this.httpClientService.hideLoading();
                /*
                if (event.status == 200) {
                }
                */
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        (err) => {
            if (err instanceof HttpErrorResponse) {
                //console.log('[HttpsRequestInterceptor][error]', err);
                this.httpClientService.hideLoading();
                if (this.coreService.getEnvironment().production) {
                    this.securityService.ping();
                }
                this.notifyService.showMessage(err.message, 'error', err.statusText);
                //let statusCode = err.status ;
                //console.log('[HttpInterceptor][ErrorResponse]=', err);
                //console.log('[HttpInterceptor][statusCode]=', statusCode);
                /*
                switch (statusCode) {
                  case 0:
                  case 401:
                     this.securityService.ping();
                    break;
                  default:
                    return;
                }
                */
            }
        }));
    }
}
HttpsRequestInterceptor.instance = null;
HttpsRequestInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
HttpsRequestInterceptor.ctorParameters = () => [
    { type: SecurityService },
    { type: CoreService },
    { type: HttpClientService },
    { type: NotificationService }
];
if (false) {
    /** @type {?} */
    HttpsRequestInterceptor.instance;
    /**
     * @type {?}
     * @private
     */
    HttpsRequestInterceptor.prototype.securityService;
    /**
     * @type {?}
     * @private
     */
    HttpsRequestInterceptor.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    HttpsRequestInterceptor.prototype.httpClientService;
    /**
     * @type {?}
     * @private
     */
    HttpsRequestInterceptor.prototype.notifyService;
}
export class InterceptorModule {
}
InterceptorModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    { provide: HTTP_INTERCEPTORS, useClass: HttpsRequestInterceptor, multi: true }
                ]
            },] }
];
