// inet-ui/esm2015/src/front-view/front-view.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
export class FrontViewDirective {
    /**
     * @param {?} _el
     */
    constructor(_el) {
        this._el = _el;
        this.onShow = new EventEmitter();
        this.onHide = new EventEmitter();
        this.isShow = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._$el = $(this._el.nativeElement);
        this._$el.click((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            if (e.target.isSameNode(this._el.nativeElement)) {
                this.hide();
            }
        }));
        document.body.appendChild(this._el.nativeElement);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._el.nativeElement.remove();
    }
    /**
     * @return {?}
     */
    show() {
        this.isShow = true;
        this._$el.show();
        this.onShow.emit();
    }
    /**
     * @return {?}
     */
    hide() {
        this.isShow = false;
        this._$el.hide();
        this.onHide.emit();
    }
}
FrontViewDirective.decorators = [
    { type: Directive, args: [{
                selector: '[appFrontView]',
                exportAs: 'appFrontView'
            },] }
];
/** @nocollapse */
FrontViewDirective.ctorParameters = () => [
    { type: ElementRef }
];
FrontViewDirective.propDecorators = {
    onShow: [{ type: Output }],
    onHide: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FrontViewDirective.prototype.onShow;
    /** @type {?} */
    FrontViewDirective.prototype.onHide;
    /** @type {?} */
    FrontViewDirective.prototype.isShow;
    /**
     * @type {?}
     * @private
     */
    FrontViewDirective.prototype._$el;
    /** @type {?} */
    FrontViewDirective.prototype._el;
}
