// inet-ui/esm2015/src/select2/app-select2.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef } from '@angular/core';
import { Input, Output } from '@angular/core';
import { EventEmitter } from '@angular/core';
import { Directive } from '@angular/core';
export class AppSelect2Directive {
    /**
     * @param {?} _el
     */
    constructor(_el) {
        this._el = _el;
        this.onChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.element = $(this._el.nativeElement).select2(this.options).change((/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            this.onChange.emit(e.target.value);
        }));
        this.updateSelect();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.options) {
            this.updateSelect();
        }
        if (changes.initValue) {
            this.updateValue();
        }
    }
    /**
     * @return {?}
     */
    updateSelect() {
        if (this.element && this.options) {
            this.element.empty().select2(this.options);
            this.updateValue();
        }
    }
    /**
     * @return {?}
     */
    updateValue() {
        if (this.element && this.initValue) {
            this.element.val(this.initValue).trigger('change');
        }
    }
}
AppSelect2Directive.decorators = [
    { type: Directive, args: [{
                selector: 'select[appSelect2]'
            },] }
];
/** @nocollapse */
AppSelect2Directive.ctorParameters = () => [
    { type: ElementRef }
];
AppSelect2Directive.propDecorators = {
    initValue: [{ type: Input }],
    options: [{ type: Input }],
    onChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    AppSelect2Directive.prototype.initValue;
    /** @type {?} */
    AppSelect2Directive.prototype.options;
    /** @type {?} */
    AppSelect2Directive.prototype.onChange;
    /** @type {?} */
    AppSelect2Directive.prototype.element;
    /**
     * @type {?}
     * @private
     */
    AppSelect2Directive.prototype._el;
}
