// inet-core/esm5/src/autosize.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Inspired from: https://stevepapa.com/ng-autosize/
 */
import { Input, ElementRef, HostListener, Directive } from '@angular/core';
var AutoSizeDirective = /** @class */ (function () {
    function AutoSizeDirective(element) {
        this.element = element;
        this.el = element.nativeElement;
        this._clientWidth = this.el.clientWidth;
    }
    Object.defineProperty(AutoSizeDirective.prototype, "minHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this._minHeight;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._minHeight = val;
            this.updateMinHeight();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AutoSizeDirective.prototype, "maxHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this._maxHeight;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._maxHeight = val;
            this.updateMaxHeight();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AutoSizeDirective.prototype.onResize = /**
     * @return {?}
     */
    function () {
        //Only apply adjustment if element width had changed.
        if (this.el.clientWidth === this._clientWidth)
            return;
        this._clientWidth = this.element.nativeElement.clientWidth;
        this.adjust();
    };
    /**
     * @return {?}
     */
    AutoSizeDirective.prototype.onInput = /**
     * @return {?}
     */
    function () {
        this.adjust();
    };
    /**
     * @return {?}
     */
    AutoSizeDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // set element resize allowed manually by user
        /** @type {?} */
        var style = window.getComputedStyle(this.el, null);
        if (style.resize === 'both') {
            this.el.style.resize = 'horizontal';
        }
        else if (style.resize === 'vertical') {
            this.el.style.resize = 'none';
        }
        // run first adjust
        this.adjust();
    };
    /**
     * @return {?}
     */
    AutoSizeDirective.prototype.adjust = /**
     * @return {?}
     */
    function () {
        // perform height adjustments after input changes, if height is different
        // if (this.el.style.height == this.element.nativeElement.scrollHeight + "px") return;
        this.el.style.overflow = 'hidden';
        this.el.style.height = 'auto';
        this.el.style.height = this.el.scrollHeight + "px";
    };
    /**
     * @return {?}
     */
    AutoSizeDirective.prototype.updateMinHeight = /**
     * @return {?}
     */
    function () {
        // Set textarea min height if input defined
        this.el.style.minHeight = this._minHeight + 'px';
    };
    /**
     * @return {?}
     */
    AutoSizeDirective.prototype.updateMaxHeight = /**
     * @return {?}
     */
    function () {
        // Set textarea max height if input defined
        this.el.style.maxHeight = this._maxHeight + 'px';
    };
    AutoSizeDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'textarea[appAutoSize]',
                    exportAs: 'appAutoSize'
                },] }
    ];
    /** @nocollapse */
    AutoSizeDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AutoSizeDirective.propDecorators = {
        minHeight: [{ type: Input, args: ['minHeight',] }],
        maxHeight: [{ type: Input, args: ['maxHeight',] }],
        onResize: [{ type: HostListener, args: ['window:resize',] }],
        onInput: [{ type: HostListener, args: ['input',] }]
    };
    return AutoSizeDirective;
}());
export { AutoSizeDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AutoSizeDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    AutoSizeDirective.prototype._minHeight;
    /**
     * @type {?}
     * @private
     */
    AutoSizeDirective.prototype._maxHeight;
    /**
     * @type {?}
     * @private
     */
    AutoSizeDirective.prototype._lastHeight;
    /**
     * @type {?}
     * @private
     */
    AutoSizeDirective.prototype._clientWidth;
    /** @type {?} */
    AutoSizeDirective.prototype.element;
}
