// inet-core/esm5/src/core.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import { HttpClientService } from './http-client.service';
import { map, shareReplay, tap } from 'rxjs/operators';
var CoreService = /** @class */ (function () {
    function CoreService(http) {
        this.http = http;
        return CoreService.instance = CoreService.instance || this;
    }
    /**
     * @return {?}
     */
    CoreService.prototype.logout = /**
     * @return {?}
     */
    function () {
        this.http.post(iNet.getPUrl('system/logout'), { 'username': iNet.username }).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var iFrame = document.createElement('iframe');
            iFrame.setAttribute('src', data.uuid);
            iFrame.addEventListener('load', (/**
             * @return {?}
             */
            function () {
                location.reload();
            }));
            document.body.appendChild(iFrame);
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
        }));
    };
    /*
    * Gets the avatar URL with the username
    * @deprecated Use UserProfileService.getAvatarUrlByUsername() instead
    */
    /*
        * Gets the avatar URL with the username
        * @deprecated Use UserProfileService.getAvatarUrlByUsername() instead
        */
    /**
     * @param {?} usercode
     * @param {?=} thumbnail
     * @return {?}
     */
    CoreService.prototype.getAvatar = /*
        * Gets the avatar URL with the username
        * @deprecated Use UserProfileService.getAvatarUrlByUsername() instead
        */
    /**
     * @param {?} usercode
     * @param {?=} thumbnail
     * @return {?}
     */
    function (usercode, thumbnail) {
        thumbnail = thumbnail || 50;
        return iNet.getPUrl('system/userprofile/photo') +
            ("?usercode=" + usercode + "&thumbnail=" + thumbnail);
    };
    /**
     * @return {?}
     */
    CoreService.prototype.isExternalUser = /**
     * @return {?}
     */
    function () {
        return (iNet.externalUser == 'true');
    };
    /**
     * @return {?}
     */
    CoreService.prototype.isCommunity = /**
     * @return {?}
     */
    function () {
        return (this.isExternalUser() || iNet.firmPrefix == 'smartcloud' || iNet.firmPrefix == 'community');
    };
    /**
     * @return {?}
     */
    CoreService.prototype.getOrg = /**
     * @return {?}
     */
    function () {
        return this.http.get(iNet.getPUrl('plugin/organization/list'))
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res.elements; })));
    };
    /**
     * @param {?=} params
     * @return {?}
     */
    CoreService.prototype.searchOrgan = /**
     * @param {?=} params
     * @return {?}
     */
    function (params) {
        if (params === void 0) { params = {}; }
        return this.http.postJSON(iNet.getPUrl('plugin/organization/search'), params)
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res.elements; })));
    };
    /**
     * @return {?}
     */
    CoreService.prototype.getSystemApplication = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.applications) {
            return of(this.applications);
        }
        if (!this.$appCache) {
            this.$appCache = this.http.postJSON(iNet.getPUrl('system/application/list'))
                .pipe(tap((/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return _this.applications = res.elements; })))
                .pipe(map((/**
             * @param {?} res
             * @return {?}
             */
            function (res) { return res.elements; })))
                .pipe(shareReplay(1));
        }
        return this.$appCache;
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    CoreService.prototype.searchFirmAccount = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        this.http.postJSON(iNet.getPUrl('plugin/firmaccount/list'), params).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            callback(data);
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            callback(null, err);
        }));
    };
    /**
     * @param {?} lang
     * @param {?=} callback
     * @return {?}
     */
    CoreService.prototype.updateLanguage = /**
     * @param {?} lang
     * @param {?=} callback
     * @return {?}
     */
    function (lang, callback) {
        this.http.postJSON(iNet.getPUrl('system/userprofile/languageupdate'), { intl: lang }).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (callback) {
                callback(lang);
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            if (callback) {
                callback(lang);
            }
        }));
    };
    /**
     * Get the Image URL for the current app
     * @param fileName - the file name
     */
    /**
     * Get the Image URL for the current app
     * @param {?} fileName - the file name
     * @return {?}
     */
    CoreService.prototype.getImageUrl = /**
     * Get the Image URL for the current app
     * @param {?} fileName - the file name
     * @return {?}
     */
    function (fileName) {
        if (!fileName) {
            return '';
        }
        if (fileName.indexOf('/') === 0) {
            return this.getAssetsPath() + fileName;
        }
        return this.getAssetsPath() + "/" + fileName;
    };
    /**
     * Build the file URL for the file server request
     * @param fileName - the given file name
     */
    /**
     * Build the file URL for the file server request
     * @param {?} fileName - the given file name
     * @return {?}
     */
    CoreService.prototype.getFileUrl = /**
     * Build the file URL for the file server request
     * @param {?} fileName - the given file name
     * @return {?}
     */
    function (fileName) {
        if (!fileName) {
            return '';
        }
        if (fileName.indexOf('/') === 0) {
            return this.getFileServerPath() + fileName;
        }
        return this.getFileServerPath() + "/" + fileName;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    CoreService.prototype.getSsoRedirect = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return iNet.getPUrl('ssorequest/redirect') + '?' + $.param(data);
    };
    /**
     * @return {?}
     */
    CoreService.prototype.convergePlugins = /**
     * @return {?}
     */
    function () {
        return this.convergeSearch({ converge: 'plugins' });
    };
    /**
     * @param {?} params
     * @return {?}
     */
    CoreService.prototype.convergeSearch = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return this.http.postJSON(iNet.getPUrl('converge/search'), params);
    };
    /**
     * @param {?=} v
     * @return {?}
     */
    CoreService.prototype.setEnvironment = /**
     * @param {?=} v
     * @return {?}
     */
    function (v) {
        if (v === void 0) { v = { production: false }; }
        CoreService.environment = v;
    };
    /**
     * @return {?}
     */
    CoreService.prototype.getEnvironment = /**
     * @return {?}
     */
    function () {
        return CoreService.environment;
    };
    /**
     * @private
     * @param {?=} path
     * @return {?}
     */
    CoreService.prototype.getPath = /**
     * @private
     * @param {?=} path
     * @return {?}
     */
    function (path) {
        if (path === void 0) { path = ''; }
        /** @type {?} */
        var lastIndex = path.lastIndexOf('/');
        if (lastIndex === path.length - 1) {
            return path.substring(0, lastIndex);
        }
        return path;
    };
    /**
     * Gets File server path
     */
    /**
     * Gets File server path
     * @return {?}
     */
    CoreService.prototype.getFileServerPath = /**
     * Gets File server path
     * @return {?}
     */
    function () {
        return this.getPath(iNet.fileServer);
    };
    /**
     * Gets assets path of current app
     */
    /**
     * Gets assets path of current app
     * @return {?}
     */
    CoreService.prototype.getAssetsPath = /**
     * Gets assets path of current app
     * @return {?}
     */
    function () {
        return this.getPath(iNet.imgFolder);
    };
    CoreService.instance = null; // Create a singleton service
    // Create a singleton service
    CoreService.environment = { production: false };
    CoreService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CoreService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    return CoreService;
}());
export { CoreService };
if (false) {
    /** @type {?} */
    CoreService.instance;
    /**
     * @type {?}
     * @private
     */
    CoreService.environment;
    /**
     * @type {?}
     * @private
     */
    CoreService.prototype.$appCache;
    /**
     * @type {?}
     * @private
     */
    CoreService.prototype.applications;
    /**
     * @type {?}
     * @private
     */
    CoreService.prototype.http;
}
/**
 * @record
 */
export function LogoutData() { }
if (false) {
    /** @type {?} */
    LogoutData.prototype.uuid;
}
/**
 * @record
 */
function CustomResponse() { }
if (false) {
    /** @type {?} */
    CustomResponse.prototype.elements;
}
/**
 * @record
 */
export function SsoRedirectData() { }
if (false) {
    /** @type {?} */
    SsoRedirectData.prototype.url;
    /** @type {?} */
    SsoRedirectData.prototype.application;
    /** @type {?|undefined} */
    SsoRedirectData.prototype.redirect;
}
/**
 * @record
 */
export function PathRedirectData() { }
if (false) {
    /** @type {?} */
    PathRedirectData.prototype.path;
    /** @type {?} */
    PathRedirectData.prototype.application;
    /** @type {?|undefined} */
    PathRedirectData.prototype.organId;
    /** @type {?|undefined} */
    PathRedirectData.prototype.firmPrefix;
}
