// inet-core/esm5/src/infinite-scroller.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
import { fromEvent } from "rxjs";
/**
 * @record
 */
function ScrollPosition() { }
if (false) {
    /** @type {?} */
    ScrollPosition.prototype.sH;
    /** @type {?} */
    ScrollPosition.prototype.sT;
    /** @type {?} */
    ScrollPosition.prototype.cH;
}
/** @type {?} */
var DEFAULT_SCROLL_POSITION = {
    sH: 0,
    sT: 0,
    cH: 0
};
var InfiniteScrollerDirective = /** @class */ (function () {
    function InfiniteScrollerDirective(elm) {
        var _this = this;
        this.elm = elm;
        this.scrollPercent = 70;
        this.isSubmit = false;
        this.isUserScrollingDown = (/**
         * @param {?} positions
         * @return {?}
         */
        function (positions) {
            return positions[0].sT < positions[1].sT;
        });
        this.isScrollExpectedPercent = (/**
         * @param {?} position
         * @return {?}
         */
        function (position) {
            return ((position.sT + position.cH) / position.sH) > (_this.scrollPercent / 100);
        });
    }
    /**
     * @return {?}
     */
    InfiniteScrollerDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.registerScrollEvent();
        this.streamScrollEvents();
        this.requestCallbackOnScroll();
    };
    /**
     * @private
     * @return {?}
     */
    InfiniteScrollerDirective.prototype.registerScrollEvent = /**
     * @private
     * @return {?}
     */
    function () {
        this.scrollEvent$ = fromEvent(this.elm.nativeElement, 'scroll');
    };
    /**
     * @private
     * @return {?}
     */
    InfiniteScrollerDirective.prototype.streamScrollEvents = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.userScrolledDown$ = this.scrollEvent$
            .map((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return ({
            sH: e.target.scrollHeight,
            sT: e.target.scrollTop,
            cH: e.target.clientHeight
        }); }))
            .pairwise()
            .filter((/**
         * @param {?} positions
         * @return {?}
         */
        function (positions) { return _this.isUserScrollingDown(positions) && _this.isScrollExpectedPercent(positions[1]); }));
    };
    /**
     * @private
     * @return {?}
     */
    InfiniteScrollerDirective.prototype.requestCallbackOnScroll = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.requestOnScroll$ = this.userScrolledDown$;
        if (this.immediateCallback) {
            this.requestOnScroll$ = this.requestOnScroll$
                .startWith([DEFAULT_SCROLL_POSITION, DEFAULT_SCROLL_POSITION]);
        }
        this.requestOnScroll$
            .exhaustMap((/**
         * @return {?}
         */
        function () {
            return _this.scrollCallback();
        }))
            .subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) { console.log(data); }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) { return console.log(err); }));
    };
    /**
     * @private
     * @return {?}
     */
    InfiniteScrollerDirective.prototype.isScrollSubmit = /**
     * @private
     * @return {?}
     */
    function () {
        return this.isSubmit;
    };
    ;
    InfiniteScrollerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[appInfiniteScroller]'
                },] }
    ];
    /** @nocollapse */
    InfiniteScrollerDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    InfiniteScrollerDirective.propDecorators = {
        scrollCallback: [{ type: Input }],
        immediateCallback: [{ type: Input }],
        scrollPercent: [{ type: Input }],
        isSubmit: [{ type: Input }]
    };
    return InfiniteScrollerDirective;
}());
export { InfiniteScrollerDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollerDirective.prototype.scrollEvent$;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollerDirective.prototype.userScrolledDown$;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollerDirective.prototype.requestOnScroll$;
    /** @type {?} */
    InfiniteScrollerDirective.prototype.scrollCallback;
    /** @type {?} */
    InfiniteScrollerDirective.prototype.immediateCallback;
    /** @type {?} */
    InfiniteScrollerDirective.prototype.scrollPercent;
    /** @type {?} */
    InfiniteScrollerDirective.prototype.isSubmit;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollerDirective.prototype.isUserScrollingDown;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollerDirective.prototype.isScrollExpectedPercent;
    /**
     * @type {?}
     * @private
     */
    InfiniteScrollerDirective.prototype.elm;
    /* Skipping unhandled member: ;*/
}
