// inet-core/esm5/src/resource-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var ResourceLoaderService = /** @class */ (function () {
    function ResourceLoaderService() {
        this._loaded = [];
        this._version = '';
        // this._initScriptVersion();
    }
    /**
     * @param {?} resources
     * @param {?} callback
     * @return {?}
     */
    ResourceLoaderService.prototype.load = /**
     * @param {?} resources
     * @param {?} callback
     * @return {?}
     */
    function (resources, callback) {
        if (resources.length < 1) {
            return callback();
        }
        this._loadResources(resources, callback);
    };
    /**
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    ResourceLoaderService.prototype.loadJS = /**
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    function (url, callback) {
        this._loadWithTagName('script', {
            type: 'text/javascript',
            src: this.getFullUrlJS(url)
        }, callback);
    };
    /**
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    ResourceLoaderService.prototype.loadCSS = /**
     * @param {?} url
     * @param {?} callback
     * @return {?}
     */
    function (url, callback) {
        this._loadWithTagName('link', {
            type: 'text/css',
            rel: 'stylesheet',
            href: url
        }, callback);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    ResourceLoaderService.prototype.getFullUrlJS = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        if (url.indexOf('//') < 0 && iNet.fileServer) {
            if (url.startsWith('/')) {
                url = url.substr(1, url.length);
            }
            return iNet.fileServer + url;
        }
        return url;
    };
    /**
     * @param {?} url
     * @return {?}
     */
    ResourceLoaderService.prototype.isLoaded = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return this._loaded.indexOf(url) > -1;
    };
    /**
     * @private
     * @param {?} resources
     * @param {?} callback
     * @return {?}
     */
    ResourceLoaderService.prototype._loadResources = /**
     * @private
     * @param {?} resources
     * @param {?} callback
     * @return {?}
     */
    function (resources, callback) {
        var _this = this;
        /** @type {?} */
        var splitGroups = this._groupResources(resources);
        /** @type {?} */
        var fn = (/**
         * @return {?}
         */
        function () {
            if (splitGroups.length < 1) {
                return callback();
            }
            _this._loadGroups(splitGroups.splice(0, 1)[0], fn);
        });
        fn();
    };
    //  Split group resources by sync/async
    //  Split group resources by sync/async
    /**
     * @private
     * @param {?} resources
     * @return {?}
     */
    ResourceLoaderService.prototype._groupResources = 
    //  Split group resources by sync/async
    /**
     * @private
     * @param {?} resources
     * @return {?}
     */
    function (resources) {
        /** @type {?} */
        var splitGroups = [];
        /** @type {?} */
        var groups;
        //  Split group resources by sync/async
        for (var i = 0; i < resources.length; i++) {
            if (!groups) {
                groups = [];
                splitGroups.push(groups);
            }
            groups.push(resources[i]);
            if (resources[i].sync) {
                groups = null;
            }
        }
        return splitGroups;
    };
    /**
     * @private
     * @param {?} resources
     * @param {?} callback
     * @return {?}
     */
    ResourceLoaderService.prototype._loadGroups = /**
     * @private
     * @param {?} resources
     * @param {?} callback
     * @return {?}
     */
    function (resources, callback) {
        var _this = this;
        /** @type {?} */
        var promises = [];
        resources.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return promises.push(_this._load(item)); }));
        Promise.all(promises).then((/**
         * @return {?}
         */
        function () {
            callback();
        }));
    };
    /**
     * @private
     * @param {?} resource
     * @return {?}
     */
    ResourceLoaderService.prototype._load = /**
     * @private
     * @param {?} resource
     * @return {?}
     */
    function (resource) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            if (resource.type === 'css') {
                _this.loadCSS(resource.url, resolve);
            }
            else {
                _this.loadJS(resource.url, resolve);
            }
        }));
    };
    /**
     * @private
     * @param {?} tagName
     * @param {?} properties
     * @param {?} callback
     * @return {?}
     */
    ResourceLoaderService.prototype._loadWithTagName = /**
     * @private
     * @param {?} tagName
     * @param {?} properties
     * @param {?} callback
     * @return {?}
     */
    function (tagName, properties, callback) {
        var _this = this;
        /** @type {?} */
        var url = properties.src || properties.href;
        // Loaded
        if (this.isLoaded(url)) {
            return callback();
        }
        /** @type {?} */
        var tagEl = document.createElement(tagName);
        for (var k in properties) {
            tagEl[k] = properties[k];
        }
        tagEl.onload = (/**
         * @return {?}
         */
        function () {
            _this._loaded.push(url);
            callback();
        });
        tagEl.onerror = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            console.warn("Load resource error: " + url);
            console.warn(e);
            _this._loaded.push(url);
            callback();
        });
        document.head.appendChild(tagEl);
    };
    /**
     * @private
     * @return {?}
     */
    ResourceLoaderService.prototype._initScriptVersion = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var $script = $('script');
        for (var i = 0; i < $script.length; i++) {
            /** @type {?} */
            var src = $script[i]['src'];
            /** @type {?} */
            var index = src.indexOf('?version=');
            if (index > -1) {
                this._version = src.substr(index + 1, src.length);
                break;
            }
        }
    };
    ResourceLoaderService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ResourceLoaderService.ctorParameters = function () { return []; };
    return ResourceLoaderService;
}());
export { ResourceLoaderService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ResourceLoaderService.prototype._loaded;
    /**
     * @type {?}
     * @private
     */
    ResourceLoaderService.prototype._version;
}
/**
 * @record
 */
export function Resource() { }
if (false) {
    /** @type {?} */
    Resource.prototype.url;
    /** @type {?|undefined} */
    Resource.prototype.sync;
    /** @type {?|undefined} */
    Resource.prototype.loaded;
    /** @type {?|undefined} */
    Resource.prototype.type;
}
