// inet-core/esm5/src/role-access.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { SecurityService } from './security.service';
var RoleAccessDirective = /** @class */ (function () {
    function RoleAccessDirective(templateRef, viewContainer, securityService) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.securityService = securityService;
    }
    /**
     * @return {?}
     */
    RoleAccessDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.applyPermission();
    };
    /**
     * @private
     * @return {?}
     */
    RoleAccessDirective.prototype.applyPermission = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.securityService.hasRole(this.roleAccess)) {
            this.viewContainer.createEmbeddedView(this.templateRef);
        }
        else {
            this.viewContainer.clear();
        }
    };
    RoleAccessDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[roleAccess]'
                },] }
    ];
    /** @nocollapse */
    RoleAccessDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef },
        { type: SecurityService }
    ]; };
    RoleAccessDirective.propDecorators = {
        roleAccess: [{ type: Input, args: ['roleAccess',] }]
    };
    return RoleAccessDirective;
}());
export { RoleAccessDirective };
if (false) {
    /** @type {?} */
    RoleAccessDirective.prototype.roleAccess;
    /**
     * @type {?}
     * @private
     */
    RoleAccessDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    RoleAccessDirective.prototype.viewContainer;
    /**
     * @type {?}
     * @private
     */
    RoleAccessDirective.prototype.securityService;
}
