// inet-core/esm5/src/security.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from "./http-client.service";
var SecurityService = /** @class */ (function () {
    function SecurityService(http) {
        this.http = http;
        this.roles = [];
        if (!SecurityService.instance) {
            this.roles = this.parserRoleFrom(iNet.permission || {});
        }
        return SecurityService.instance = SecurityService.instance || this;
    }
    /**
     * @return {?}
     */
    SecurityService.prototype.load = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            if (!window['iNet'] || iNet.isEmpty(iNet.prefix) || !iNet.enableLayout) {
                resolve();
            }
            else {
                _this.http.postJSON(iNet.getPUrl('system/userrole/list'))
                    .subscribe((/**
                 * @param {?} obj
                 * @return {?}
                 */
                function (obj) {
                    for (var key in obj) {
                        _this.append(obj[key]);
                    }
                    resolve(obj);
                }), (/**
                 * @param {?} err
                 * @return {?}
                 */
                function (err) {
                    resolve();
                }));
            }
        }));
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    SecurityService.prototype.parserRoleFrom = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var roles = [];
        for (var key in obj) {
            if (obj[key] === "true" || obj[key] === true) {
                roles.push(key);
            }
        }
        return roles;
    };
    /**
     * @private
     * @param {?} arrArg
     * @return {?}
     */
    SecurityService.prototype.uniqueArray = /**
     * @private
     * @param {?} arrArg
     * @return {?}
     */
    function (arrArg) {
        return arrArg.filter((/**
         * @param {?} elem
         * @param {?} pos
         * @param {?} arr
         * @return {?}
         */
        function (elem, pos, arr) {
            return arr.indexOf(elem) == pos;
        }));
    };
    ;
    /**
     * @private
     * @param {?} v
     * @return {?}
     */
    SecurityService.prototype.append = /**
     * @private
     * @param {?} v
     * @return {?}
     */
    function (v) {
        //console.log('[append]roles', v);
        this.roles = this.uniqueArray(this.roles.concat(v));
        return this.roles;
    };
    /**
     * @param {?} v
     * @return {?}
     */
    SecurityService.prototype.hasRole = /**
     * @param {?} v
     * @return {?}
     */
    function (v) {
        var _this = this;
        //console.log('[checkRole]=', v);
        /** @type {?} */
        var roles = v.split(',');
        if (roles.length > 1) {
            /** @type {?} */
            var __items = roles.filter((/**
             * @param {?} role
             * @return {?}
             */
            function (role) { return _this.roles.indexOf(role) > -1; }));
            return (__items.length > 0);
        }
        return (this.roles.indexOf(v) > -1);
    };
    /**
     * Check session timeout when system is idle
     */
    /**
     * Check session timeout when system is idle
     * @return {?}
     */
    SecurityService.prototype.ping = /**
     * Check session timeout when system is idle
     * @return {?}
     */
    function () {
        if (iNet.isEmpty(iNet.username) || !iNet.enableLayout) {
            return;
        }
        /** @type {?} */
        var iframe = document.createElement('iframe');
        iframe.style.display = "none";
        iframe.setAttribute("src", iNet.getPUrl('common/page/ping'));
        iframe.onload = (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var __loginUrl = 'cas/login';
            try {
                /** @type {?} */
                var __path = iframe.contentWindow.location.pathname || '';
                /** @type {?} */
                var __isLogin = (__loginUrl.match(__path) || []).length > 0;
                if (__isLogin) {
                    window.location.reload();
                }
            }
            catch (ex) {
                window.location.reload();
            }
            document.body.removeChild(iframe);
        });
        document.body.appendChild(iframe);
    };
    SecurityService.instance = null;
    SecurityService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SecurityService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    return SecurityService;
}());
export { SecurityService };
if (false) {
    /** @type {?} */
    SecurityService.instance;
    /**
     * @type {?}
     * @private
     */
    SecurityService.prototype.roles;
    /**
     * @type {?}
     * @private
     */
    SecurityService.prototype.http;
    /* Skipping unhandled member: ;*/
}
