// inet-core/esm5/src/suggestion.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from './http-client.service';
var SuggestionService = /** @class */ (function () {
    function SuggestionService(http) {
        this.http = http;
        return SuggestionService.instance = SuggestionService.instance || this;
    }
    /**
     * @param {?} params
     * @param {?=} callback
     * @return {?}
     */
    SuggestionService.prototype.saveSuggestion = /**
     * @param {?} params
     * @param {?=} callback
     * @return {?}
     */
    function (params, callback) {
        if (!params.keyword) {
            return;
        }
        return this.http.postJSON(iNet.getPUrl('system/suggestion/update'), params).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (callback) {
                callback(data);
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            if (callback) {
                callback(null, err);
            }
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    SuggestionService.prototype.loadSuggestion = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        return this.http.postJSON(iNet.getPUrl('system/suggestion/hint'), params).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            callback(data && data.items || []);
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            callback([], err);
        }));
    };
    SuggestionService.instance = null; // Create a singleton service
    SuggestionService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SuggestionService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    return SuggestionService;
}());
export { SuggestionService };
if (false) {
    /** @type {?} */
    SuggestionService.instance;
    /**
     * @type {?}
     * @private
     */
    SuggestionService.prototype.http;
}
/**
 * @record
 */
export function SuggestionParamData() { }
if (false) {
    /** @type {?} */
    SuggestionParamData.prototype.content;
    /** @type {?} */
    SuggestionParamData.prototype.keyword;
}
/**
 * @record
 */
export function SuggestionResponse() { }
if (false) {
    /** @type {?} */
    SuggestionResponse.prototype.items;
    /** @type {?} */
    SuggestionResponse.prototype.total;
}
/**
 * @record
 */
export function SuggestionItem() { }
if (false) {
    /** @type {?} */
    SuggestionItem.prototype.content;
    /** @type {?} */
    SuggestionItem.prototype.keyword;
    /** @type {?} */
    SuggestionItem.prototype.ownercode;
    /** @type {?} */
    SuggestionItem.prototype.usage;
    /** @type {?} */
    SuggestionItem.prototype.uuid;
}
