// inet-ui/esm5/src/sharing-information/sharing-information.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from "inet-core";
import * as i0 from "@angular/core";
import * as i1 from "inet-core";
var SharingInformationService = /** @class */ (function () {
    function SharingInformationService(http) {
        this.http = http;
        this.url = {
            application_list: iNet.getPUrl("system/application/list"),
            //firmContext (appname)
            data_rights: iNet.getPUrl("datarights/available"),
            //params: appname
            user_rights_list: iNet.getPUrl("datarights/list"),
            //params: appname, group, myshare: boolean
            data_rights_update: iNet.getPUrl("datarights/update"),
            //params: {member: string, group: string, right: string},
            data_rights_delete: iNet.getPUrl("datarights/delete"),
            //params: appname
            users_suggest: iNet.getPUrl("cloud/profileacct/list")
        };
    }
    /**
     * @param {?} params
     * @return {?}
     */
    SharingInformationService.prototype.getApplicationList = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.getJSON(this.url.application_list, params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SharingInformationService.prototype.getApplicationRights = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.getJSON(this.url.data_rights, params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SharingInformationService.prototype.getUserRights = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.getJSON(this.url.user_rights_list, params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SharingInformationService.prototype.updateUserRight = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.postJSON(this.url.data_rights_update, params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SharingInformationService.prototype.deleteUserRight = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.postJSON(this.url.data_rights_delete, params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    SharingInformationService.prototype.getSuggestUser = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return this.http.postJSON(this.url.users_suggest, params);
    };
    SharingInformationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    SharingInformationService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    /** @nocollapse */ SharingInformationService.ngInjectableDef = i0.defineInjectable({ factory: function SharingInformationService_Factory() { return new SharingInformationService(i0.inject(i1.HttpClientService)); }, token: SharingInformationService, providedIn: "root" });
    return SharingInformationService;
}());
export { SharingInformationService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SharingInformationService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    SharingInformationService.prototype.http;
}
