// inet-calendar/esm5/src/utils/subfirm/Dictionary.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SubFirmDictionary = /** @class */ (function () {
    function SubFirmDictionary() {
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    SubFirmDictionary.prototype.create = /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    function (data, callback) {
        $.ajax({
            type: 'post',
            url: this.getUrl('subfirm/dictionary/create'),
            data: data,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(data); }),
            error: (/**
             * @return {?}
             */
            function () { return callback(null); })
        });
    };
    /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    SubFirmDictionary.prototype.update = /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    function (data, callback) {
        data.dictID = data.dictID || data.uuid;
        $.ajax({
            type: 'post',
            url: this.getUrl('subfirm/dictionary/update'),
            data: data,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(data); }),
            error: (/**
             * @return {?}
             */
            function () { return callback(null); })
        });
    };
    /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    SubFirmDictionary.prototype.remove = /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    function (data, callback) {
        data.dictID = data.dictID || data.uuid;
        $.ajax({
            url: this.getUrl('subfirm/dictionary/remove'),
            data: data,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(data); }),
            error: (/**
             * @return {?}
             */
            function () { return callback(null); })
        });
    };
    /**
     * @param {?} reference
     * @param {?} callback
     * @return {?}
     */
    SubFirmDictionary.prototype.list = /**
     * @param {?} reference
     * @param {?} callback
     * @return {?}
     */
    function (reference, callback) {
        $.ajax({
            url: this.getUrl('subfirm/dictionary/list'),
            data: {
                reference: reference
            },
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(data && data.items || []); }),
            error: (/**
             * @return {?}
             */
            function () { return callback([]); })
        });
    };
    /**
     * @param {?} url
     * @return {?}
     */
    SubFirmDictionary.prototype.getUrl = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return iNet.getPUrl(url);
    };
    return SubFirmDictionary;
}());
export { SubFirmDictionary };
/**
 * @record
 */
export function DictionaryData() { }
if (false) {
    /** @type {?|undefined} */
    DictionaryData.prototype.dictID;
    /** @type {?|undefined} */
    DictionaryData.prototype.name;
    /** @type {?|undefined} */
    DictionaryData.prototype.value;
    /** @type {?|undefined} */
    DictionaryData.prototype.reference;
    /** @type {?|undefined} */
    DictionaryData.prototype.uuid;
    /** @type {?|undefined} */
    DictionaryData.prototype.cache;
    /** @type {?|undefined} */
    DictionaryData.prototype.order;
}
