// inet-social/esm5/src/routing/social-routing.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { SocialHomeComponent } from "./home/social-home.component";
import { SocialActivityViewComponent } from "./activity-view/social-activity-view.component";
/** @type {?} */
var routes = [
    {
        path: 'activity/social',
        component: SocialHomeComponent
    },
    {
        path: 'activity/social/activity/:id',
        component: SocialActivityViewComponent
    },
    {
        path: 'activity/social/**',
        redirectTo: 'activity/social',
        pathMatch: 'full'
    },
];
var SocialRoutingModule = /** @class */ (function () {
    function SocialRoutingModule() {
    }
    SocialRoutingModule.decorators = [
        { type: NgModule, args: [{
                    imports: [RouterModule.forChild(routes)],
                    exports: [RouterModule]
                },] }
    ];
    return SocialRoutingModule;
}());
export { SocialRoutingModule };
