// inet-ui/esm2015/src/web-token/web-token.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { WebTokenSigner } from "./lib/WebTokenSigner";
import { TokenSimulator } from "./lib/TokenSimulator";
import { EXT_HASH_ALG } from "./lib/Constants";
import { Observable } from "rxjs";
export class WebTokenService {
    constructor() {
        this._signer = new WebTokenSigner();
    }
    /**
     * @return {?}
     */
    initForDevMode() {
        console.warn("Init WebTokenSigner for dev mode. Not used it in production mode");
        this._signer.appInstalled = true;
        if (!window["_ExtTokenBridge"]) {
            window["_ExtTokenBridge"] = TokenSimulator;
        }
    }
    /**
     * @return {?}
     */
    ready() {
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            this._signer.checkReady().then((/**
             * @param {?} supported
             * @return {?}
             */
            (supported) => subscriber.next(supported)));
        }));
    }
    /**
     * @param {?=} forceSelect
     * @return {?}
     */
    selectCert(forceSelect) {
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            this._signer.selectCert(forceSelect).then((/**
             * @param {?} cert
             * @return {?}
             */
            (cert) => subscriber.next(cert)));
        }));
    }
    /**
     * @param {?} serial
     * @return {?}
     */
    getCertBySerial(serial) {
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            this.loadCerts().subscribe((/**
             * @param {?} certs
             * @return {?}
             */
            (certs) => {
                for (let i = 0; i < certs.length; i++) {
                    if (certs[i].serial === serial) {
                        return subscriber.next(certs[i]);
                    }
                }
                subscriber.next(null);
            }));
        }));
    }
    /**
     * @param {?=} serial
     * @return {?}
     */
    selectCertUnlock(serial) {
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            if (serial) {
                this.getCertBySerial(serial).subscribe((/**
                 * @param {?} cert
                 * @return {?}
                 */
                (cert) => this._unlockCert(cert, subscriber)));
            }
            else {
                this.selectCert().subscribe((/**
                 * @param {?} cert
                 * @return {?}
                 */
                (cert) => this._unlockCert(cert, subscriber)));
            }
        }));
    }
    /**
     * @private
     * @param {?} cert
     * @param {?} subscriber
     * @return {?}
     */
    _unlockCert(cert, subscriber) {
        if (cert) {
            this.unlockCert(cert).subscribe((/**
             * @param {?} unlocked
             * @return {?}
             */
            (unlocked) => {
                cert.unlocked = unlocked;
                subscriber.next(cert);
            }));
        }
        else {
            subscriber.next(cert);
        }
    }
    /**
     * @param {?} cert
     * @return {?}
     */
    unlockCert(cert) {
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            // Sign fake data to unlock token
            /** @type {?} */
            const fakeData = {
                hash: "cA/metLrX7NUOfviDVsGYeRxAIo=",
                hashAlg: EXT_HASH_ALG.SHA1,
                serial: cert.serial
            };
            this.signHash(fakeData).subscribe((/**
             * @param {?} response
             * @return {?}
             */
            (response) => {
                if (response && response.data && response.data.signature) {
                    subscriber.next(true);
                }
                else {
                    subscriber.next(false);
                }
            }));
        }));
    }
    /**
     * @return {?}
     */
    loadCerts() {
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            this._signer.loadCerts().then((/**
             * @param {?} data
             * @return {?}
             */
            (data) => subscriber.next(data)));
        }));
    }
    /**
     * @param {?} data
     * @return {?}
     */
    signHash(data) {
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        (subscriber) => {
            this._signer.signHash(data).then((/**
             * @param {?} response
             * @return {?}
             */
            (response) => subscriber.next(response)));
        }));
    }
}
WebTokenService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    WebTokenService.prototype._signer;
}
