// inet-ui/esm2015/src/time-picker/time-picker-modal/time-picker-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { TimepickerComponent } from "ngx-bootstrap";
export class TimePickerModalComponent {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.arrowkeys = true;
        this.hourStep = 1;
        this.minuteStep = 15;
        this.mousewheel = true;
        this.readonlyInput = false;
        this.secondsStep = 10;
        this.showMinutes = true;
        this.showSpinners = true;
        this.onChange = new EventEmitter();
    }
    /**
     * @param {?} date
     * @return {?}
     */
    set time(date) {
        this._timeDate = date;
        this.timeStr = TimePickerModalComponent._formatDate(date);
        this.onChange.emit(this.timeStr);
    }
    /**
     * @return {?}
     */
    get time() {
        return this._timeDate;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.$modal = $(this.el.nativeElement).detach().addClass('modal');
        this.$modalContent = this.$modal.find('.time-picker-modal');
        // Touched close
        this.$modal.on('mousedown click', (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            if (e.target.isSameNode(this.$modal[0])) {
                this.close();
            }
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.$modal.remove();
    }
    /**
     * @return {?}
     */
    close() {
        this.$modal.detach();
    }
    /**
     * @param {?} css
     * @return {?}
     */
    open(css) {
        this.$modalContent.css(css);
        document.body.appendChild(this.$modal[0]);
    }
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    static _formatDate(date) {
        if (!date || !date.getTime()) {
            return '';
        }
        else {
            return ('0' + date.getHours()).slice(-2) + ':' + ('0' + date.getMinutes()).slice(-2);
        }
    }
}
TimePickerModalComponent.decorators = [
    { type: Component, args: [{
                selector: 'time-picker-modal',
                template: "<div class=\"time-picker-modal\">\n    <timepicker #timePicker [(ngModel)]=\"time\"\n                [arrowkeys]=\"arrowkeys\"\n                [max]=\"max\"\n                [min]=\"min\"\n                [minuteStep]=\"minuteStep\"\n                [mousewheel]=\"mousewheel\"\n                [readonlyInput]=\"readonlyInput\"\n                [secondsStep]=\"secondsStep\"\n                [showMeridian]=\"showMeridian\"\n                [showMinutes]=\"showMinutes\"\n                [showSeconds]=\"showSeconds\"\n                [showSpinners]=\"showSpinners\">\n\n    </timepicker>\n</div>",
                styles: [":host(.modal){display:block;z-index:1101}.time-picker-modal{position:absolute;border:1px solid rgba(0,0,0,.2);background:#fff;border-radius:4px;padding:10px}"]
            }] }
];
/** @nocollapse */
TimePickerModalComponent.ctorParameters = () => [
    { type: ElementRef }
];
TimePickerModalComponent.propDecorators = {
    arrowkeys: [{ type: Input }],
    hourStep: [{ type: Input }],
    max: [{ type: Input }],
    min: [{ type: Input }],
    minuteStep: [{ type: Input }],
    mousewheel: [{ type: Input }],
    readonlyInput: [{ type: Input }],
    secondsStep: [{ type: Input }],
    showMeridian: [{ type: Input }],
    showMinutes: [{ type: Input }],
    showSeconds: [{ type: Input }],
    showSpinners: [{ type: Input }],
    time: [{ type: Input }],
    onChange: [{ type: Output }],
    timePicker: [{ type: ViewChild, args: ['timePicker',] }]
};
if (false) {
    /** @type {?} */
    TimePickerModalComponent.prototype.arrowkeys;
    /** @type {?} */
    TimePickerModalComponent.prototype.hourStep;
    /** @type {?} */
    TimePickerModalComponent.prototype.max;
    /** @type {?} */
    TimePickerModalComponent.prototype.min;
    /** @type {?} */
    TimePickerModalComponent.prototype.minuteStep;
    /** @type {?} */
    TimePickerModalComponent.prototype.mousewheel;
    /** @type {?} */
    TimePickerModalComponent.prototype.readonlyInput;
    /** @type {?} */
    TimePickerModalComponent.prototype.secondsStep;
    /** @type {?} */
    TimePickerModalComponent.prototype.showMeridian;
    /** @type {?} */
    TimePickerModalComponent.prototype.showMinutes;
    /** @type {?} */
    TimePickerModalComponent.prototype.showSeconds;
    /** @type {?} */
    TimePickerModalComponent.prototype.showSpinners;
    /** @type {?} */
    TimePickerModalComponent.prototype.onChange;
    /** @type {?} */
    TimePickerModalComponent.prototype.timePicker;
    /** @type {?} */
    TimePickerModalComponent.prototype.timeStr;
    /** @type {?} */
    TimePickerModalComponent.prototype.$modal;
    /** @type {?} */
    TimePickerModalComponent.prototype.$modalContent;
    /**
     * @type {?}
     * @private
     */
    TimePickerModalComponent.prototype._timeDate;
    /** @type {?} */
    TimePickerModalComponent.prototype.el;
}
