// inet-chat/esm2015/src/open-graph/open-graph.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class OpenGraphService {
    constructor() {
        this.regex = /(https?:\/\/[^\s]+)/;
    }
    /**
     * @param {?} link
     * @param {?} callback
     * @return {?}
     */
    loadPreviewLink(link, callback) {
        if (!link) {
            callback(null, {
                error: 'Link is empty'
            });
            return;
        }
        if (!this.isLink(link)) {
            callback(null, {
                error: 'Link invalid'
            });
            return;
        }
        $.ajax({
            type: 'post',
            url: iNet.getUrl('opengraph/fetch'),
            data: {
                url: link
            },
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                callback(this.convertOpenData(data, link));
            }),
            error: (/**
             * @param {?} xhr
             * @return {?}
             */
            function (xhr) {
                callback(null, xhr);
            })
        });
    }
    /**
     * @param {?} link
     * @return {?}
     */
    isLink(link) {
        return this.regex.test(link);
    }
    /**
     * @param {?} src
     * @param {?} callback
     * @return {?}
     */
    loadImageInfo(src, callback) {
        /** @type {?} */
        let img = new Image();
        img.onload = img.onerror = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            /** @type {?} */
            let imageInfo = {
                width: img.width,
                height: img.height
            };
            if (imageInfo.width && imageInfo.height) {
                imageInfo.available = true;
                imageInfo.landscape = imageInfo.width / imageInfo.height > 1.1;
            }
            callback(imageInfo);
        });
        img.src = src;
    }
    /**
     * @private
     * @param {?} properties
     * @param {?} url
     * @return {?}
     */
    convertOpenData(properties, url) {
        if (properties.length < 1) {
            return;
        }
        /** @type {?} */
        let openData = (/** @type {?} */ ({}));
        properties.forEach((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            openData[item.property] = item.content;
        }));
        if (!openData.url) {
            openData.url = url;
        }
        return openData;
    }
}
OpenGraphService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
OpenGraphService.ctorParameters = () => [];
if (false) {
    /**
     * @type {?}
     * @private
     */
    OpenGraphService.prototype.regex;
}
