// inet-ui/esm5/src/select-user/modal-select/modal-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { BsModalService } from "ngx-bootstrap";
import { CoreService } from "inet-core";
var ModalSelectComponent = /** @class */ (function () {
    function ModalSelectComponent(coreService, modalService) {
        this.coreService = coreService;
        this.modalService = modalService;
        this.selectUsers = [];
        this.users = [];
    }
    /**
     * @return {?}
     */
    ModalSelectComponent.prototype.setDefault = /**
     * @return {?}
     */
    function () {
        this.selectUsers = [];
    };
    /**
     * @return {?}
     */
    ModalSelectComponent.prototype.searchUser = /**
     * @return {?}
     */
    function () {
        var _this = this;
        clearTimeout(this.timer);
        /** @type {?} */
        var value = this.key || '';
        this.timer = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.coreService.searchFirmAccount({ keyword: value, pageSize: 10 }, (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                /** @type {?} */
                var users = data && data['items'] || [];
                _this.setUsers(users);
            }));
        }), 300);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    ModalSelectComponent.prototype.addUser = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (this.isSingle) {
            if (this.selectUsers[0]) {
                this.users.push(this.selectUsers[0]);
            }
            this.selectUsers[0] = item;
        }
        else {
            for (var i = 0; i < this.selectUsers.length; i++) {
                if (this.selectUsers[i].username === item.username) {
                    return;
                }
            }
            this.selectUsers.push(item);
        }
        this.setUsers(this.users);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    ModalSelectComponent.prototype.removeUser = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        for (var i = 0; i < this.selectUsers.length; i++) {
            if (this.selectUsers[i].username === item.username) {
                this.users.push(item);
                this.selectUsers.splice(i, 1);
                return;
            }
        }
    };
    /**
     * @return {?}
     */
    ModalSelectComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        this.searchUser();
        this.modalRef = this.modalService.show(this.selectModal);
    };
    /**
     * @return {?}
     */
    ModalSelectComponent.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.modalRef.hide();
    };
    /**
     * @private
     * @param {?} user
     * @return {?}
     */
    ModalSelectComponent.prototype.hasUser = /**
     * @private
     * @param {?} user
     * @return {?}
     */
    function (user) {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.selectUsers), _c = _b.next(); !_c.done; _c = _b.next()) {
                var item = _c.value;
                if (item.username === user.username) {
                    return true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return false;
    };
    /**
     * @private
     * @param {?} users
     * @return {?}
     */
    ModalSelectComponent.prototype.setUsers = /**
     * @private
     * @param {?} users
     * @return {?}
     */
    function (users) {
        var _this = this;
        this.users = users.filter((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return !_this.hasUser(item); }));
    };
    ModalSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-modal-select',
                    template: "<ng-template #selectModal>\n    <div class=\"modal-header border-0\">\n        <h5 class=\"modal-title pull-left\"><i class=\"fa fa-users\"></i> Ch\u1ECDn ng\u01B0\u1EDDi d\u00F9ng </h5>\n        <button type=\"button\" class=\"close pull-right\" aria-label=\"Close\" (click)=\"hide()\">\n            <span aria-hidden=\"true\">&times;</span>\n        </button>\n    </div>\n    <div class=\"modal-body p-0 font-size\">\n        <div class=\"col-sm-12 border-bottom\">\n            <div class=\"input-group input-group-sm\" style=\"padding-bottom: 10px;\">\n                <input (input)=\"searchUser()\"\n                       [(ngModel)]=\"key\" type=\"text\" class=\"form-control\" placeholder=\"T\u00ECm ki\u1EBFm\">\n                <div class=\"input-group-append\">\n                    <span class=\"input-group-text\" (click)=\"searchUser()\">\n                        <i class=\"fa fa-search\"></i>\n                    </span>\n                </div>\n            </div>\n        </div>\n        <nav class=\"nav nav-tabs\">\n            <span class=\"nav-item nav-link active rounded-0\" href=\"#nav-book\" data-toggle=\"tab\"  role=\"tab\">\n                <i class=\"fa fa-group\"></i> T\u1EA5t c\u1EA3\n            </span>\n            <span class=\"nav-item nav-link rounded-0\" href=\"#nav-book-select\" data-toggle=\"tab\" role=\"tab\">\n                <i class=\"fa fa-check-circle\"></i> \u0110\u00E3 ch\u1ECDn\n                <span [hidden]=\"selectUsers?.length<1\" class=\"ml-1 badge badge-primary rounded-circle\">{{ selectUsers?.length }}</span>\n            </span>\n        </nav>\n        <div class=\"tab-content\">\n            <div id=\"nav-book\" class=\"tab-pane fade active show\" role=\"tabpanel\" aria-labelledby=\"nav-home-tab\" aria-expanded=\"true\">\n                <ul class=\"list-group rounded-0\"\n                    [ngClass]=\"{'p-3':users?.length > 0}\">\n                    <li *ngFor=\"let user of users\"\n                        (click)=\"addUser(user)\"\n                        class=\"list-group-item rounded-0\">\n                        <div class=\"row\">\n                            <div class=\"col-sm-2\">\n                                <img userAvatar [usercode]=\"user?.username\" class=\"rounded-circle\">\n                            </div>\n                            <div class=\"col-sm-10\">\n                                <span class=\"name\">{{user.fullname}}</span>\n                                <br>\n                                <span class=\"email text-muted\"><i class=\"fa fa-envelope\"></i>&nbsp;{{user.username}}</span>\n                            </div>\n                        </div>\n                        <div class=\"clearfix\"></div>\n                    </li>\n                    <li *ngIf=\"users?.length<1\" class=\"list-group-item rounded-0 border-0\">\n                        <i>Kh\u00F4ng t\u00ECm th\u1EA5y d\u1EEF li\u1EC7u</i>\n                    </li>\n                </ul>\n            </div>\n            <div id=\"nav-book-select\"\n                 class=\"tab-pane fade\" role=\"tabpanel\" aria-labelledby=\"nav-profile-tab\" aria-expanded=\"false\">\n                <ul class=\"list-group rounded-0\"\n                    [ngClass]=\"{'p-3':selectUsers?.length > 0}\">\n                    <li *ngFor=\"let user of selectUsers\"\n                        (click)=\"removeUser(user)\"\n                        class=\"list-group-item rounded-0 selected\">\n                        <div class=\"row\">\n                            <div class=\"col-sm-2\">\n                                <img userAvatar [usercode]=\"user?.username\" class=\"rounded-circle\">\n                            </div>\n                            <div class=\"col-sm-10\">\n                                <span class=\"name\">{{user.fullname}}</span>\n                                <br>\n                                <span class=\"email\"><i class=\"fa fa-envelope\"></i>&nbsp;{{user.username}}</span>\n                            </div>\n                        </div>\n                        <div class=\"clearfix\"></div>\n                    </li>\n                    <li *ngIf=\"selectUsers?.length<1\" class=\"list-group-item rounded-0 border-0\">\n                        <i>Kh\u00F4ng t\u00ECm th\u1EA5y d\u1EEF li\u1EC7u</i>\n                    </li>\n                </ul>\n            </div>\n        </div>\n    </div>\n    <div class=\"modal-footer text-right\">\n        <button type=\"button\" class=\"btn btn-danger btn-sm\" (click)=\"hide()\"><i class=\"fa fa-remove\"></i> \u0110\u00F3ng</button>\n    </div>\n</ng-template>",
                    styles: [".tab-content{height:391px;overflow-x:hidden}.font-size{font-size:13px}.nav-tabs .nav-link.active{border-top:2px solid #4c8fbd!important;background:#fff}.nav-tabs .nav-link{background-color:#f9f9f9;color:#6b6b6b;cursor:pointer;border-bottom:1px solid #e9ecef}.list-group-item .name{font-weight:700}.list-group-item .email{white-space:nowrap}.list-group-item:hover{background-color:#f5f5f5;cursor:pointer}.list-group-item img{max-width:50px!important;height:50px;width:50px;padding:1px;border:1px solid #ebebeb;background-size:cover;background-repeat:no-repeat;background-position:center;content:\"\";display:block}.list-group-item.selected{background-color:#63abf7!important;border:1px solid #779!important;color:#fff}"]
                }] }
    ];
    /** @nocollapse */
    ModalSelectComponent.ctorParameters = function () { return [
        { type: CoreService },
        { type: BsModalService }
    ]; };
    ModalSelectComponent.propDecorators = {
        selectUsers: [{ type: Input, args: ['selectUsers',] }],
        isSingle: [{ type: Input, args: ['isSingle',] }],
        selectModal: [{ type: ViewChild, args: ['selectModal',] }]
    };
    return ModalSelectComponent;
}());
export { ModalSelectComponent };
if (false) {
    /** @type {?} */
    ModalSelectComponent.prototype.selectUsers;
    /** @type {?} */
    ModalSelectComponent.prototype.isSingle;
    /** @type {?} */
    ModalSelectComponent.prototype.selectModal;
    /** @type {?} */
    ModalSelectComponent.prototype.modalRef;
    /** @type {?} */
    ModalSelectComponent.prototype.users;
    /** @type {?} */
    ModalSelectComponent.prototype.key;
    /**
     * @type {?}
     * @private
     */
    ModalSelectComponent.prototype.timer;
    /**
     * @type {?}
     * @private
     */
    ModalSelectComponent.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    ModalSelectComponent.prototype.modalService;
}
