// inet-core/esm2015/src/websocket/websocket-jboss.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { WebSocketAbstract } from './websocket-abstract';
export class WebSocketJboss extends WebSocketAbstract {
    /**
     * @return {?}
     */
    get readyState() {
        return this._socket ? this._socket.readyState : null;
    }
    /**
     * @return {?}
     */
    stateChange() {
        if (this.readyState === WebSocket.OPEN) {
            this._ping();
        }
        super.stateChange();
    }
    /**
     * @param {?} options
     * @param {?} username
     * @return {?}
     */
    connect(options, username) {
        this._socket = new WebSocket(options);
        this._socket.onmessage = (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            // const envelop: WebSocketEnvelop = JSON.parse(event.data);
            // if (envelop.body) {
            //     this.onMessage.next(envelop.body);
            // }
            /** @type {?} */
            const envelopBody = JSON.parse(event.data);
            if (envelopBody.sender) {
                this.onMessage.next(envelopBody);
            }
        });
        this._socket.onopen = (/**
         * @return {?}
         */
        () => this.stateChange());
        this._socket.onerror = (/**
         * @return {?}
         */
        () => this.stateChange());
        this._socket.onclose = (/**
         * @return {?}
         */
        () => this.stateChange());
    }
    /**
     * @return {?}
     */
    close() {
        this._socket.close();
    }
    /**
     * @param {?} envelop
     * @return {?}
     */
    send(envelop) {
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        resolve => {
            try {
                this._socket.send(JSON.stringify(envelop));
                resolve(true);
            }
            catch (e) {
                resolve(false);
                console.warn('Send message failed', e);
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    _ping() {
        if (this.readyState === WebSocket.OPEN) {
            this._socket.send(JSON.stringify({ type: 'ping' }));
            setTimeout((/**
             * @return {?}
             */
            () => this._ping()), 10000);
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    WebSocketJboss.prototype._socket;
}
