// inet-ui/esm2015/src/collapse/collapse.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from '@angular/core';
export class CollapseDirective {
    /**
     * @param {?} _el
     */
    constructor(_el) {
        this._el = _el;
        this.expanded = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._element = $(this._el.nativeElement);
        // Click toggle
        if (this.toggleEl) {
            this.toggleEl.addEventListener('click', this.toggle.bind(this));
        }
        this.expanded = !this.expanded;
        this.toggle();
    }
    /**
     * @return {?}
     */
    show() {
        this.expanded = true;
        this._element.collapse('show').parent().addClass('collapse-expanded');
    }
    /**
     * @return {?}
     */
    hide() {
        this.expanded = false;
        this._element.collapse('hide').parent().removeClass('collapse-expanded');
    }
    /**
     * @return {?}
     */
    toggle() {
        if (this.expanded) {
            this.hide();
        }
        else {
            this.show();
        }
    }
}
CollapseDirective.decorators = [
    { type: Directive, args: [{
                selector: '[appCollapse]'
            },] }
];
/** @nocollapse */
CollapseDirective.ctorParameters = () => [
    { type: ElementRef }
];
CollapseDirective.propDecorators = {
    expanded: [{ type: Input }],
    toggleEl: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CollapseDirective.prototype.expanded;
    /** @type {?} */
    CollapseDirective.prototype.toggleEl;
    /**
     * @type {?}
     * @private
     */
    CollapseDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    CollapseDirective.prototype._el;
}
