// inet-chat/esm5/src/component/message-chat/message-chat.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild, ViewEncapsulation } from '@angular/core';
import { APP_CHAT } from "../../model/MessageConstants";
import { MessageChat } from "../../model/MessageChat";
import { ChatMessageService } from "../../message.service";
import { MessageChatHolder } from "../../model/MessageHolder";
import { FileListComponent } from "../../filelist/list/file-list.component";
import { VideoConferenceService } from "../../video-conference.service";
import { OpenGraphService } from '../../open-graph/open-graph.service';
// declare let $: any;
// import {ifvisible} from "ifvisible.js"
var MessageChatComponent = /** @class */ (function () {
    function MessageChatComponent(openGraphService, messageService, videoService, element) {
        var _this = this;
        this.openGraphService = openGraphService;
        this.messageService = messageService;
        this.videoService = videoService;
        this.element = element;
        this.audioCall = false;
        this.videoCall = false;
        this.closable = false;
        this.onAudioCall = new EventEmitter();
        this.onVideoCall = new EventEmitter();
        this.onClose = new EventEmitter();
        this.videoPanelVisible = false;
        this.linkRegex = /(https?:\/\/[^\s]+)/;
        this.holders = [];
        this.files = [];
        this.loading = true;
        this.profile = iNet.username;
        this.messageJoinCall = {};
        this._hasMoreMsg = false;
        this._params = {
            pageSize: 20,
            pageNumber: 0
        };
        this._messageObserver = this.messageService.messageReceived.subscribe((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            if (message.message.sender !== iNet.username) {
                if (message.message.sender === _this.username) {
                    if (message.message.message.alert) {
                        /** @type {?} */
                        var typeMessage = message.message.message.alert.split(":");
                        if (typeMessage[0] === 'uri_message') {
                            if (typeMessage[1] === 'video_conference' && !iNet.isEmpty(typeMessage[2])) {
                                _this.__messageChat = JSON.parse(_this.decodeBase64(typeMessage[2])) || {};
                                if (_this.__messageChat.action === 'END') {
                                    if (message.message.message.joins.split(',').length === 1) {
                                        message.displayText = _this.messageMissCall();
                                        _this._sendSuccess(message);
                                    }
                                    else {
                                        message.displayText = _this.messageJoinedCall(message.message.message.duration);
                                        _this._sendSuccess(message);
                                    }
                                }
                                if (_this.__messageChat.action === 'LEFT' && message.message.message.joins.split(',')[0] === iNet.username) {
                                    if (message.message.message.joins.split(',').length === 1) {
                                        message.displayText = _this.messageMissCall();
                                        message.message.sender = message.message.message.joins.split(',')[0];
                                        message.message.message.joins.split(',')[0] === iNet.username ? message.isSent = true : message.isSent = false;
                                        _this._sendSuccess(message);
                                    }
                                    else {
                                        message.displayText = _this.messageJoinedCallTo(message.message.message.duration);
                                        message.message.sender = message.message.message.joins.split(',')[0];
                                        message.message.message.joins.split(',')[0] === iNet.username ? message.isSent = true : message.isSent = false;
                                        _this._sendSuccess(message);
                                    }
                                }
                            }
                        }
                        else {
                            _this._sendSuccess(message);
                        }
                    }
                }
                else {
                    _this.messageService.updateLastMessage(_this.username, message.message);
                    _this.messageService.resetUnreadMessage(_this.username);
                }
            }
            else {
                /** @type {?} */
                var typeMessage = message.message.message.alert.split(":");
                if (typeMessage[0] === 'uri_message') {
                    if (typeMessage[1] === 'video_conference' && !iNet.isEmpty(typeMessage[2])) {
                        _this.__messageChat = JSON.parse(_this.decodeBase64(typeMessage[2])) || {};
                        if (_this.__messageChat.action === 'JOIN') {
                        }
                        else {
                            if (_this.__messageChat.action === 'LEFT') {
                                _this.videoService.dispose();
                                document.getElementById("meet").style.visibility = "hidden";
                                _this._sendSuccess(message);
                            }
                        }
                    }
                }
            }
            // if (message.message.sender === this.username) {
            // }
        }));
        this.element.nativeElement.addEventListener('drop', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.element.nativeElement.style.opacity = '1';
            _this._onDropFile(e);
        }));
        this.element.nativeElement.addEventListener('dragleave', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.element.nativeElement.style.opacity = '1';
        }));
        this.element.nativeElement.addEventListener('dragover', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.element.nativeElement.style.opacity = '.5';
            e.preventDefault();
        }));
    }
    Object.defineProperty(MessageChatComponent.prototype, "canSend", {
        // @ViewChild(LinkPreviewComponent) linkPreview: LinkPreviewComponent;
        get: 
        // @ViewChild(LinkPreviewComponent) linkPreview: LinkPreviewComponent;
        /**
         * @return {?}
         */
        function () {
            return this.files.length > 0 || this.message.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessageChatComponent.prototype, "message", {
        get: /**
         * @return {?}
         */
        function () {
            return this.input && this.input.nativeElement.value.trim() || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessageChatComponent.prototype, "scrollTop", {
        get: /**
         * @return {?}
         */
        function () {
            return this.messageBody.nativeElement.scrollTop;
        },
        set: /**
         * @param {?} scrollTop
         * @return {?}
         */
        function (scrollTop) {
            this.messageBody.nativeElement.scrollTop = scrollTop;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessageChatComponent.prototype, "scrollHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.messageBody.nativeElement.scrollHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessageChatComponent.prototype, "scrollBottom", {
        get: /**
         * @return {?}
         */
        function () {
            return this.scrollHeight - this.messageBody.nativeElement.scrollTop;
        },
        set: /**
         * @param {?} scrollBottom
         * @return {?}
         */
        function (scrollBottom) {
            this.scrollTop = this.scrollHeight - scrollBottom;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @param {?} ev
     * @return {?}
     */
    MessageChatComponent.prototype._onDropFile = /**
     * @private
     * @param {?} ev
     * @return {?}
     */
    function (ev) {
        // Prevent default behavior (Prevent file from being opened)
        ev.preventDefault();
        /** @type {?} */
        var files;
        if (ev.dataTransfer.items) {
            files = [];
            // Use DataTransferItemList interface to access the file(s)
            for (var i = 0; i < ev.dataTransfer.items.length; i++) {
                // If dropped items aren't files, reject them
                if (ev.dataTransfer.items[i].kind === 'file') {
                    files.push(ev.dataTransfer.items[i].getAsFile());
                }
            }
        }
        else {
            files = ev.dataTransfer.files;
        }
        if (files.length > 0) {
            this.fileList.addFiles(files);
        }
    };
    /**
     * @return {?}
     */
    MessageChatComponent.prototype.ngAfterContentChecked = /**
     * @return {?}
     */
    function () {
        // Collect scroll position before render new messages
        if (this._scrollNeedUpdate) {
            this._lastScrollBottom = this.scrollBottom;
        }
    };
    /**
     * @return {?}
     */
    MessageChatComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        // Restore scroll position after render new messages if not match
        if (this._scrollNeedUpdate) {
            /** @type {?} */
            var scrollBottom = this.scrollBottom;
            this._scrollNeedUpdate = false;
            if (this._lastScrollBottom != scrollBottom) {
                // Restore last scrollBottom
                this.scrollBottom = this._lastScrollBottom;
            }
        }
    };
    /**
     * @return {?}
     */
    MessageChatComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._onLoad();
        // ifvisible.on("focus", () => {
        //     if (ifvisible.now()) {
        //         this._params.pageNumber = 0;
        //         this._onLoad();
        //         // this.messageService.reconnetSocket();
        //     }
        // });
        this.messageBody.nativeElement.addEventListener('scroll', (/**
         * @return {?}
         */
        function () { return _this._onBodyScroll(); }));
        if (!this.videoService.isExistMeet()) {
            this.videoPanelVisible = false;
        }
    };
    /**
     * @private
     * @return {?}
     */
    MessageChatComponent.prototype._onBodyScroll = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var scrollBottom = this.scrollBottom;
        this._scrollingTop = this._scrollBottom <= scrollBottom;
        this._scrollBottom = scrollBottom;
        if (this.loading || !this._hasMoreMsg) {
            return;
        }
        // scroll to top
        if (this._scrollingTop && this.scrollTop < 150) {
            this._loadMessages();
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    MessageChatComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.username && !changes.username.firstChange) {
            this._onLoad();
            this.messageJoinCall['joined'] = false;
        }
    };
    /**
     * @return {?}
     */
    MessageChatComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._messageObserver.unsubscribe();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    MessageChatComponent.prototype.viewAttachment = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        window.open(e.file.url, '_blank');
    };
    /**
     * @return {?}
     */
    MessageChatComponent.prototype.sendMessage = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var value = this.message.trim();
        /** @type {?} */
        var __mes = this.messageService.revertMessage(value) || '';
        // let __mesDisplay = this.messageService.revertMessageDisplay(value) || '';
        /** @type {?} */
        var links = this.linkRegex.exec(__mes);
        if (links && links.length > 0) {
            this.openGraphService.loadPreviewLink(this.message, (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                value = _this.encodeBase64(JSON.stringify(_this.convertLinks(data)));
                _this.sendMessageSocket('uri_message:link_preview:' + value);
            }));
        }
        else {
            this.sendMessageSocket(value);
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    MessageChatComponent.prototype.sendMessageSocket = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value || this.files.length > 0) {
            if (this.files.length > 0) {
                this._sendPush(value);
            }
            else {
                this._sendSocket(value);
            }
            this._clearInput();
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageChatComponent.prototype.convertLinks = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var param = {};
        param['title'] = data['title'];
        param['image'] = data['image'];
        param['description'] = data['description'];
        param['url'] = data['url'];
        param['finalUrl'] = data['site_name'].toUpperCase();
        return param;
    };
    /**
     * @private
     * @return {?}
     */
    MessageChatComponent.prototype._loadMessages = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.loading = true;
        this._hasMoreMsg = false;
        this.messageService.loadMessages(this._params, (/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            // this.messages = result.items.concat(this.messages);
            _this._addMessages(result.items);
            if (_this._params.pageNumber === 0) {
                _this._scrollToBottom();
            }
            _this._scrollNeedUpdate = true;
            _this._params.pageNumber += 1;
            _this._hasMoreMsg = _this._params.pageSize * _this._params.pageNumber < result.total;
            setTimeout((/**
             * @return {?}
             */
            function () {
                _this.loading = false;
                _this._onBodyScroll();
            }), 500);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    MessageChatComponent.prototype._focus = /**
     * @private
     * @return {?}
     */
    function () {
        this.input.nativeElement.focus();
    };
    /**
     * @private
     * @return {?}
     */
    MessageChatComponent.prototype._scrollToBottom = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.scrollTop = _this.scrollHeight;
        }), 100);
    };
    /**
     * @private
     * @return {?}
     */
    MessageChatComponent.prototype._clearInput = /**
     * @private
     * @return {?}
     */
    function () {
        this.files.length = 0;
        this.input.nativeElement.value = '';
        this._focus();
    };
    /**
     * @private
     * @param {?} messagesArr
     * @param {?=} insertLast
     * @return {?}
     */
    MessageChatComponent.prototype._addMessages = /**
     * @private
     * @param {?} messagesArr
     * @param {?=} insertLast
     * @return {?}
     */
    function (messagesArr, insertLast) {
        var _this = this;
        if (insertLast === void 0) { insertLast = false; }
        /** @type {?} */
        var messages = this.convertArrayMessage(JSON.stringify(messagesArr));
        messages.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) { return b.date.getTime() - a.date.getTime(); }));
        messages.forEach((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            /** @type {?} */
            var holder;
            for (var i = 0; i < _this.holders.length; i++) {
                if (_this.holders[i].isSameDate(message.date)) {
                    holder = _this.holders[i];
                    break;
                }
            }
            if (!holder) {
                holder = new MessageChatHolder(message.date);
                if (insertLast) {
                    _this.holders.push(holder);
                }
                else {
                    _this.holders.unshift(holder);
                }
            }
            if (insertLast) {
                holder.messages.push(message);
            }
            else {
                holder.messages.unshift(message);
            }
        }));
    };
    /**
     * @private
     * @param {?} messages
     * @return {?}
     */
    MessageChatComponent.prototype.convertMessages = /**
     * @private
     * @param {?} messages
     * @return {?}
     */
    function (messages) {
        if (!iNet.isEmpty(messages)) {
            /** @type {?} */
            var __mes = this.messageService.revertMessage(messages.displayText) || '';
            /** @type {?} */
            var __mesDisplay = this.messageService.revertMessageDisplay(messages.displayText) || '';
            /** @type {?} */
            var typeMessages = __mes.split(":");
            if (typeMessages.length >= 2) {
                if (typeMessages[0] === 'uri_message' && typeMessages[1] === 'video_conference') {
                    /** @type {?} */
                    var __messageChat = JSON.parse(this.decodeBase64(typeMessages[2])) || {};
                    if (messages.message.sender === this.username) {
                        if (__messageChat.action === 'END') {
                            if (messages.message.message.joins.split(',').length === 1) {
                                messages.displayText = this.messageMissCall();
                                return messages;
                            }
                            else {
                                messages.displayText = this.messageJoinedCall(messages.message.message.duration);
                                return messages;
                            }
                        }
                        if (__messageChat.action === 'LEFT' && messages.message.message.joins.split(',')[0] === messages.message.sender) {
                            if (messages.message.message.joins.split(',').length === 1) {
                                messages.displayText = this.messageMissCall();
                                return messages;
                            }
                            else {
                                messages.displayText = this.messageJoinedCall(messages.message.message.duration);
                                return messages;
                            }
                        }
                        else {
                            return messages;
                        }
                    }
                    else {
                        if (__messageChat.action === 'END') {
                            if (messages.message.message.joins.split(',').length === 1) {
                                messages.displayText = this.messageMissCall();
                                return messages;
                            }
                            else {
                                messages.displayText = this.messageJoinedCallTo(messages.message.message.duration);
                                return messages;
                            }
                        }
                        else {
                            if (this.messageJoinCall['joined']) {
                                messages.displayText = this.messageJoinedCallTo(messages.message.message.duration);
                                return messages;
                            }
                            else {
                                messages.displayText = this.messageMissCall();
                                return messages;
                            }
                        }
                    }
                }
                else {
                    if (typeMessages[0] === 'uri_message' && typeMessages[1] === 'link_preview') {
                        /** @type {?} */
                        var __messageLink = JSON.parse(this.decodeBase64(typeMessages[2])) || {};
                        messages['_openGraph'] = __messageLink;
                        messages.displayText = __mesDisplay;
                        return messages;
                    }
                    else {
                        return messages;
                    }
                }
            }
            else {
                return messages;
            }
        }
    };
    /**
     * @return {?}
     */
    MessageChatComponent.prototype.messageMissCall = /**
     * @return {?}
     */
    function () {
        // $('.message-chat__miss').on('click',()=>{
        //     console.log('click')
        //     this.joinRoom(true);
        // })
        return '<div class="contact d-flex my-1" role="button" style="">\n' +
            '    <i class="message-chat__miss fa fa-phone ml-0"></i>\n' +
            '    <div class="contact-content pl-2">\n' +
            '        <div class="contact-name mb-1">Cuộc gọi bị nhỡ</div>\n' +
            '        <div class="contact-detail">Gọi lại</div>\n' +
            '    </div>\n' +
            '</div>';
    };
    /**
     * @param {?} duration
     * @return {?}
     */
    MessageChatComponent.prototype.messageJoinedCall = /**
     * @param {?} duration
     * @return {?}
     */
    function (duration) {
        return '<div class="contact d-flex my-1" role="button" style="">\n' +
            '    <i class="message-chat__action fa fa-phone ml-0"></i>\n' +
            '    <div class="contact-content pl-2">\n' +
            '        <div class="contact-name mb-1">Cuộc gọi thoại đến</div>\n' +
            '        <div class="contact-detail">' + duration + ' giây' + '</div>\n' +
            '    </div>\n' +
            '</div>';
    };
    /**
     * @param {?} duration
     * @return {?}
     */
    MessageChatComponent.prototype.messageJoinedCallTo = /**
     * @param {?} duration
     * @return {?}
     */
    function (duration) {
        return '<div class="contact d-flex my-1" role="button" style="">\n' +
            '    <i class="message-chat__action fa fa-phone ml-0"></i>\n' +
            '    <div class="contact-content pl-2">\n' +
            '        <div class="contact-name mb-1">Cuộc gọi thoại đi</div>\n' +
            '        <div class="contact-detail">' + duration + ' giây' + '</div>\n' +
            '    </div>\n' +
            '</div>';
    };
    /**
     * @private
     * @param {?} arrMessage
     * @return {?}
     */
    MessageChatComponent.prototype.convertArrayMessage = /**
     * @private
     * @param {?} arrMessage
     * @return {?}
     */
    function (arrMessage) {
        var _this = this;
        /** @type {?} */
        var arr = JSON.parse(arrMessage);
        arr.forEach((/**
         * @param {?} ele
         * @return {?}
         */
        function (ele) {
            ele = _this.convertMessages(ele);
            ele.date = new Date(ele.date);
            return ele;
        }));
        return arr;
    };
    /**
     * @param {?} str
     * @return {?}
     */
    MessageChatComponent.prototype.encodeBase64 = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return window.btoa(unescape(encodeURIComponent(str)));
    };
    /**
     * @param {?} str
     * @return {?}
     */
    MessageChatComponent.prototype.decodeBase64 = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return decodeURIComponent(escape(window.atob(str)));
    };
    /**
     * @private
     * @return {?}
     */
    MessageChatComponent.prototype._onLoad = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.disposeRoom();
        this._params.sender = this.username;
        this._params.pageNumber = 0;
        this._hasMoreMsg = false;
        this.holders = [];
        this.loading = true;
        this.files.length = 0;
        this.messageService.getContactByUserCode(this.username, (/**
         * @param {?} contact
         * @return {?}
         */
        function (contact) { return _this.contact = contact; }));
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this._loadMessages();
            _this._focus();
        }), 500);
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    MessageChatComponent.prototype._sendSocket = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var envelop = this.messageService.buildEnvelop(this.username, value);
        this.messageService.sendEnvelop(envelop);
        this._sendSuccess(new MessageChat(envelop.body));
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    MessageChatComponent.prototype._sendPush = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        /** @type {?} */
        var params = {
            content: value,
            alert: value,
            application: APP_CHAT,
            alias: this.username,
        };
        this.files.forEach((/**
         * @param {?} file
         * @param {?} i
         * @return {?}
         */
        function (file, i) {
            params['file-' + i] = file;
        }));
        this.messageService.pushMessage(params, (/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            if (message) {
                _this._sendSuccess(new MessageChat(message));
            }
        }));
    };
    /**
     * @private
     * @param {?} message
     * @return {?}
     */
    MessageChatComponent.prototype._sendSuccess = /**
     * @private
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this._addMessages([this.convertMessages(message)], true);
        this.messageService.updateLastMessage(this.username, message.message);
        this.messageService.resetUnreadMessage(this.username);
        this._scrollToBottom();
    };
    /**
     * @param {?} audioOnly
     * @param {?=} usercode
     * @return {?}
     */
    MessageChatComponent.prototype.joinRoom = /**
     * @param {?} audioOnly
     * @param {?=} usercode
     * @return {?}
     */
    function (audioOnly, usercode) {
        var _this = this;
        this.videoPanelVisible = false;
        this.messageJoinCall['joined'] = false;
        if (!this.videoService.isExistMeet()) {
            this.messageJoinCall['joined'] = false;
            return;
        }
        /** @type {?} */
        var element = document.querySelector('#meet');
        document.getElementById("meet").style.visibility = "visible";
        /** @type {?} */
        var __user = '';
        if (!!usercode) {
            __user = usercode;
        }
        else {
            __user = this.contact.usercode;
        }
        this.videoService.callTo(element, __user, audioOnly, 350, (/**
         * @param {?} api
         * @param {?} messages
         * @return {?}
         */
        function (api, messages) {
            api.addEventListener('readyToClose', (/**
             * @return {?}
             */
            function () {
                _this.videoService.dispose();
                document.getElementById("meet").style.visibility = "hidden";
                // this.messageService.sendEnvelop(envelop);
                if (!_this.messageJoinCall['joined']) {
                    /** @type {?} */
                    var envelop = _this.messageService.buildEnvelop(_this.username, messages);
                    _this._sendSuccess(new MessageChat(envelop.body));
                }
                // else{
                //     let __joins = [iNet.username,this.username];
                //     const envelop = this.messageService.buildEnvelop(this.username, messages,__joins.join(',') )
                //     console.log('envelop',envelop,__joins.join(','))
                //     this._sendSuccess(new MessageChat(envelop.body));
                // }
            }));
            api.addEventListener('videoConferenceJoined', (/**
             * @return {?}
             */
            function () {
                _this.messageJoinCall = _this.videoService.getMessageJoin();
                _this.messageJoinCall['joined'] = true;
            }));
        }));
    };
    /**
     * @return {?}
     */
    MessageChatComponent.prototype.disposeRoom = /**
     * @return {?}
     */
    function () {
        this.videoPanelVisible = true;
        this.videoService.dispose();
    };
    MessageChatComponent.decorators = [
        { type: Component, args: [{
                    selector: 'message-chat',
                    template: "<div class=\"messenger-contact\" *ngIf=\"contact\">\n    <div class=\"messenger-contact-name\">{{contact.fullname || contact.usercode}}</div>\n    <div style=\"white-space: nowrap\">\n        <i class=\"message-chat__action fa fa-phone\" *ngIf=\"audioCall && profile !== contact.usercode\" (click)=\"joinRoom(true)\"></i>\n<!--        <i class=\"message-chat__action fa fa-video-camera\" *ngIf=\"videoCall && profile !== contact.usercode\" (click)=\"joinRoom(false)\"></i>-->\n        <i class=\"message-chat__action fa fa-times\" *ngIf=\"closable\" (click)=\"onClose.emit(contact)\"></i>\n    </div>\n</div>\n<div #messageBody class=\"messenger-body\">\n    <div class=\"message-video\">\n        <!--<div *ngIf=\"videoPanelVisible\" class=\"container text-center mb-2\" style=\"background: #000000;min-height: 200px;\">-->\n            <!--<h5 class=\"text-white\" style=\"padding: 40px;\">\u0110ang c\u00F3 cu\u1ED9c tr\u00F2 chuy\u1EC7n</h5>-->\n            <!--<button class=\"btn btn-success\" (click)=\"joinRoom(false)\">Tham gia</button>-->\n        <!--</div>-->\n        <div id=\"meet\" class=\"container-video-meet\" [ngClass]=\"{'hide': videoPanelVisible}\"></div>\n    </div>\n    <div class=\"message-content\">\n        <div *ngFor=\"let holder of holders\">\n            <div class=\"messenger-date-group\">{{holder.display}}</div>\n            <div *ngFor=\"let message of holder.messages\" class=\"messenger-chat\" [ngClass]=\"{'sent': message.isSent}\" >\n                <img *ngIf=\"!message.isSent && message.displayText!==''\" class=\"messenger-chat-image\" userAvatar [usercode]=\"message.message.sender\">\n                <div class=\"messenger-chat-content\" >\n                    <file-list [files]=\"message.attachments\" [removable]=\"false\" (onClick)=\"viewAttachment($event)\" class=\"messenger-chat__file\"></file-list>\n                    <app-link-preview *ngIf=\"message._openGraph\" [openGraphData]=\"message._openGraph\"\n                                      [removable]=\"false\" class=\"social-block\"></app-link-preview>\n                    <span *ngIf=\"!message._openGraph\" [innerHTML]=\"message.displayText\"></span>\n                    <div class=\"messenger-chat-time\" [title]=\"message.longTime\">{{message.shortTime}}</div>\n                </div>\n            </div>\n        </div>\n    </div>\n</div>\n<div class=\"messenger-footer\">\n    <file-list #fileList fileList [fileEl]=\"fileEl\" [files]=\"files\"></file-list>\n    <div class=\"messenger-footer-inner\">\n        <div class=\"messenger-footer-input\">\n            <textarea type=\"text\" #input autosize  (keyup.enter)=\"sendMessage()\" placeholder=\"Nh\u1EADp n\u1ED9i dung ...\"></textarea>\n        </div>\n        <label class=\"message-chat__action fa fa-image\" style=\"margin-bottom: 0;\">\n            <input #fileEl type=\"file\"  (change)=\"input.focus()\" multiple style=\"display:none\">\n        </label>\n        <i *ngIf=\"canSend\" (click)=\"sendMessage()\" class=\"messenger-footer-send message-chat__action fa fa-arrow-right\" aria-hidden=\"true\"></i>\n    </div>\n</div>\n<i *ngIf=\"loading\" class=\"messenger-spinner fa fa-spinner fa-spin\"></i>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [":host{position:relative;display:block;width:100%;height:100%}.messenger-contact{top:0;left:0;width:100%;height:100px;display:flex;align-items:center;justify-content:space-between;padding:0 15px;border-bottom:1px solid #d3d3d3}.messenger-contact-name{max-height:36px;line-height:18px;overflow:hidden;color:#1c9dea;font-weight:700}.message-chat__miss{min-width:34px;width:34px;height:34px;line-height:34px;font-size:17px;text-align:center;border-radius:50%;background:red;color:#fff;margin-left:10px;cursor:pointer}.message-chat__miss:active,.message-chat__miss:hover{background:#8b1313;color:#fff}.message-chat__action{min-width:34px;width:34px;height:34px;line-height:34px;font-size:17px;text-align:center;border-radius:50%;background:#eff7fe;color:#1c9dea;margin-left:10px;cursor:pointer}.message-chat__action:active,.message-chat__action:hover{background:#1c9dea;color:#fff}.message-chat__action .sent_action{background:#046094!important}.messenger-body{background:#eff7fe;position:absolute;top:100px;bottom:65px;overflow-x:hidden;overflow-y:auto;width:100%}.messenger-body::-webkit-scrollbar{width:5px}.messenger-body::-webkit-scrollbar-track{background:#f1f1f1}.messenger-body::-webkit-scrollbar-thumb{background:#d6eaf7}.messenger-body::-webkit-scrollbar-thumb:hover{background:#d6eaf7}.message-video{display:flex;position:-webkit-sticky;position:sticky;top:0;left:0;right:0;z-index:10}.message-content{padding:15px}.messenger-date-group{width:100%;border-bottom:1px solid #ddd;margin:10px 0;padding:5px 0;text-align:center}.messenger-chat{margin-bottom:10px;overflow:hidden;display:flex}.messenger-chat.sent{flex-direction:row-reverse}.messenger-chat.sent .messenger-chat-content{background:#e5edf5;text-align:right;font-weight:700;color:#223645}.messenger-chat-image{min-width:32px;width:32px;height:32px;line-height:32px;font-size:16px;text-align:center;border-radius:50%;background:rgba(0,0,0,.1);margin-right:10px}.messenger-chat-content{background:#1c9dea;border-radius:25px;padding:7px 15px;overflow:hidden;max-width:80%;font-size:13px;line-height:1.2;word-break:break-word;font-weight:700;color:#fff}.messenger-chat-time{font-size:12px;color:#fff;margin-top:2px}.sent .messenger-chat-time{font-size:12px;color:#223645!important;margin-top:2px}.messenger-chat__file{margin-right:-5px}.messenger-chat__file /deep/ .file-item{width:76px;height:76px}.messenger-spinner{width:40px;height:40px;text-align:center;line-height:40px;position:absolute;top:50%;left:50%;font-size:40px;color:#666;margin-top:-20px;margin-left:-20px}.messenger-footer{position:absolute;left:0;bottom:0;width:100%;padding:10px 15px;background:#fff}.messenger-footer-inner{width:100%;display:flex;align-items:center}.messenger-footer-input{flex-grow:1;padding-top:6px}.messenger-footer-input textarea{border:0;outline:0;padding:10px 40px 0 10px;width:100%;font-size:14px;line-height:12px;border-radius:20px;background:#eff7fe;resize:none}.messenger-footer-send{color:#fff;background:#1c9dea}.messenger-footer-send:active,.messenger-footer-send:hover{background:#438eb9}.container-video-meet{height:100vh;width:100%}iframe{height:100vh!important}"]
                }] }
    ];
    /** @nocollapse */
    MessageChatComponent.ctorParameters = function () { return [
        { type: OpenGraphService },
        { type: ChatMessageService },
        { type: VideoConferenceService },
        { type: ElementRef }
    ]; };
    MessageChatComponent.propDecorators = {
        username: [{ type: Input }],
        audioCall: [{ type: Input }],
        videoCall: [{ type: Input }],
        closable: [{ type: Input }],
        onAudioCall: [{ type: Output }],
        onVideoCall: [{ type: Output }],
        onClose: [{ type: Output }],
        input: [{ type: ViewChild, args: ['input',] }],
        messageBody: [{ type: ViewChild, args: ['messageBody',] }],
        fileList: [{ type: ViewChild, args: ['fileList',] }]
    };
    return MessageChatComponent;
}());
export { MessageChatComponent };
if (false) {
    /** @type {?} */
    MessageChatComponent.prototype.username;
    /** @type {?} */
    MessageChatComponent.prototype.audioCall;
    /** @type {?} */
    MessageChatComponent.prototype.videoCall;
    /** @type {?} */
    MessageChatComponent.prototype.closable;
    /** @type {?} */
    MessageChatComponent.prototype.onAudioCall;
    /** @type {?} */
    MessageChatComponent.prototype.onVideoCall;
    /** @type {?} */
    MessageChatComponent.prototype.onClose;
    /** @type {?} */
    MessageChatComponent.prototype.input;
    /** @type {?} */
    MessageChatComponent.prototype.messageBody;
    /** @type {?} */
    MessageChatComponent.prototype.fileList;
    /** @type {?} */
    MessageChatComponent.prototype.videoPanelVisible;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype.linkRegex;
    /** @type {?} */
    MessageChatComponent.prototype.contact;
    /** @type {?} */
    MessageChatComponent.prototype.holders;
    /** @type {?} */
    MessageChatComponent.prototype.files;
    /** @type {?} */
    MessageChatComponent.prototype.loading;
    /** @type {?} */
    MessageChatComponent.prototype.profile;
    /** @type {?} */
    MessageChatComponent.prototype.messageJoinCall;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype._scrollingTop;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype._scrollBottom;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype._lastScrollBottom;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype._scrollNeedUpdate;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype._hasMoreMsg;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype._messageObserver;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype.__messageChat;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype._params;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype.openGraphService;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype.videoService;
    /**
     * @type {?}
     * @private
     */
    MessageChatComponent.prototype.element;
}
