// inet-social/esm5/src/routing/friend/social-friend.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, ViewContainerRef } from '@angular/core';
import { FormControl } from '@angular/forms';
import { Router } from "@angular/router";
import { SocialGroupComponent } from "../../social-friend/group/group.component";
import { FriendInvitationComponent } from "../../social-friend/invitation/friend-invitation.component";
import { FriendBrowseComponent } from "../../social-friend/browse/friend-browse.component";
var SocialFriendComponent = /** @class */ (function () {
    function SocialFriendComponent(router) {
        this.router = router;
        this.tabs = [
            {
                name: 'Trang chủ',
                component: '',
                path: ''
            },
            {
                name: 'Bạn bè',
                component: SocialGroupComponent,
                path: '/friend'
            },
            {
                name: 'Lời mời',
                component: FriendInvitationComponent,
                path: '/friend/invitation'
            },
            {
                name: 'Tìm bạn',
                component: FriendBrowseComponent,
                path: '/friend/browse'
            }
        ];
        this.searchValue = '';
        this.search = new FormControl();
        this.prefix = this.router.url.substring(0, this.router.url.indexOf('/friend'));
    }
    /**
     * @return {?}
     */
    SocialFriendComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._searchEl = $(this.searchEl.nativeElement);
        this.search.valueChanges.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (_this._searchable) {
                _this._activeComponent.instance.onSearch(value.trim());
            }
        }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SocialFriendComponent.prototype.changeTab = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.searchValue = '';
    };
    /**
     * @param {?} componentRef
     * @return {?}
     */
    SocialFriendComponent.prototype.loadTabComponent = /**
     * @param {?} componentRef
     * @return {?}
     */
    function (componentRef) {
        var _this = this;
        this._activeComponent = componentRef;
        this._searchable = componentRef.instance.onSearch;
        this.placeholder = componentRef.instance.placeholder || 'Tìm kiếm...';
        if (this._searchable) {
            this._searchEl.show();
            componentRef.instance.clearSearchValue = (/**
             * @return {?}
             */
            function () {
                _this.searchValue = '';
            });
        }
        else {
            this._searchEl.hide();
        }
    };
    SocialFriendComponent.decorators = [
        { type: Component, args: [{
                    template: "<div class=\"friend-toolbar\">\n    <div navigationTab\n         [tabs]=\"tabs\"\n         [viewContainerRef]=\"_tabViewContainer\"\n         [prefix]=\"prefix\"\n         (onChange)=\"changeTab($event)\"\n         (onLoad)=\"loadTabComponent($event)\" class=\"friend-tab\"></div>\n\n    <div #searchEl class=\"input-group search-input\">\n        <div class=\"input-group-prepend\">\n            <i class=\"input-group-text fa fa-search\"></i>\n        </div>\n        <input type=\"text\"\n               [formControl]=\"search\"\n               [(ngModel)]=\"searchValue\"\n               class=\"form-control\" [placeholder]=\"placeholder\">\n    </div>\n</div>\n<div class=\"friend-content\">\n    <ng-container #tabViewContainer></ng-container>\n</div>\n",
                    styles: [".friend-toolbar{border-bottom:1px solid rgba(0,0,0,.12);background:#e5e9ee;margin-bottom:20px;padding:10px 15px;min-height:58px}.friend-tab{float:left;margin-right:25px}.friend-content{padding:0 15px}.search-input{width:300px}@media (max-width:767px){.friend-content{padding:0}.friend-tab{margin:0 0 10px}.search-input{width:100%}}"]
                }] }
    ];
    /** @nocollapse */
    SocialFriendComponent.ctorParameters = function () { return [
        { type: Router }
    ]; };
    SocialFriendComponent.propDecorators = {
        _tabViewContainer: [{ type: ViewChild, args: ['tabViewContainer', { read: ViewContainerRef },] }],
        searchEl: [{ type: ViewChild, args: ['searchEl',] }]
    };
    return SocialFriendComponent;
}());
export { SocialFriendComponent };
if (false) {
    /** @type {?} */
    SocialFriendComponent.prototype._tabViewContainer;
    /** @type {?} */
    SocialFriendComponent.prototype.searchEl;
    /** @type {?} */
    SocialFriendComponent.prototype.tabs;
    /** @type {?} */
    SocialFriendComponent.prototype.searchValue;
    /** @type {?} */
    SocialFriendComponent.prototype.search;
    /** @type {?} */
    SocialFriendComponent.prototype.placeholder;
    /** @type {?} */
    SocialFriendComponent.prototype.prefix;
    /**
     * @type {?}
     * @private
     */
    SocialFriendComponent.prototype._activeComponent;
    /**
     * @type {?}
     * @private
     */
    SocialFriendComponent.prototype._searchEl;
    /**
     * @type {?}
     * @private
     */
    SocialFriendComponent.prototype._searchable;
    /**
     * @type {?}
     * @private
     */
    SocialFriendComponent.prototype.router;
}
