// inet-social/esm5/src/filelist/file-format.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
/** @type {?} */
var FILES_FORMAT = [
    'unknown', 'text', 'archive', 'audio', 'video',
    'png', 'jpg', 'gif', 'tiff', 'svg', 'ai', 'psd', 'dwg',
    'avi', 'fla', 'mp2', 'mp3', 'mp4', 'aac', 'flac', 'wma', 'wav', 'mxf',
    'iso', 'mdf', 'nrg',
    'zip', '7z', 'arj', 'rar',
    'pdf', 'doc', 'rtf', 'txt', 'xls', 'ppt',
    'css', 'csv', 'html', 'json', 'js', 'xml',
    'dbf', 'exe'
];
/** @type {?} */
var FILES_EDIT_FORMAT = [
    'doc', 'docx', 'odt', 'ods', 'xls', 'xlsx',
    'ppt', 'pps', 'pptm', 'pptx', 'pot', 'potx',
    'dot', 'dotx', 'docm', 'dotx', 'csv', 'sxw'
];
/** @type {?} */
var FILES_VIEW_FORMAT = ['pdf', 'png', 'jpg', 'jpeg', 'gif'];
/** @type {?} */
var UNKNOWN = 'unknown';
var FileFormatService = /** @class */ (function () {
    function FileFormatService() {
    }
    /**
     * @return {?}
     */
    FileFormatService.prototype.getImagePath = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var path = iNet.commonImageFolder;
        /** @type {?} */
        var lastIndex = path.lastIndexOf('/');
        if (lastIndex === path.length - 1) {
            return path.substring(0, lastIndex);
        }
        return path;
    };
    /**
     * @return {?}
     */
    FileFormatService.prototype.getFileFormatPath = /**
     * @return {?}
     */
    function () {
        return this.getImagePath() + "/format";
    };
    /**
     * @param {?} ext
     * @return {?}
     */
    FileFormatService.prototype.getUrlByExt = /**
     * @param {?} ext
     * @return {?}
     */
    function (ext) {
        /** @type {?} */
        var path = this.getFileFormatPath();
        if (FILES_FORMAT.indexOf(ext) > -1) {
            return path + "/" + ext.toLowerCase() + ".svg";
        }
        return path + "/" + UNKNOWN + ".svg";
    };
    /**
     * @param {?} name
     * @return {?}
     */
    FileFormatService.prototype.getUrlByName = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return this.getUrlByExt(this.getExtByName(name));
    };
    /**
     * @param {?} name
     * @return {?}
     */
    FileFormatService.prototype.getExtByName = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return name.split('.').pop().toLowerCase() || UNKNOWN;
    };
    /**
     * @return {?}
     */
    FileFormatService.prototype.getEditFormats = /**
     * @return {?}
     */
    function () {
        return FILES_EDIT_FORMAT;
    };
    /**
     * @return {?}
     */
    FileFormatService.prototype.getViewFormats = /**
     * @return {?}
     */
    function () {
        return FILES_VIEW_FORMAT;
    };
    FileFormatService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FileFormatService.ctorParameters = function () { return []; };
    return FileFormatService;
}());
export { FileFormatService };
