// inet-ui/esm5/src/plugin-manager/plugin-toolbar.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { PluginManagerService } from "./plugin-manager.service";
import { CoreService, NotificationService } from "inet-core";
var PluginToolbarDirective = /** @class */ (function () {
    function PluginToolbarDirective(el, renderer, coreService, notifyService, pluginManagerService) {
        this.el = el;
        this.renderer = renderer;
        this.coreService = coreService;
        this.notifyService = notifyService;
        this.pluginManagerService = pluginManagerService;
        this.idProperty = 'toolbar'; // The name of the field, Defaults to 'toolbar'.
        // console.log('[PluginToolbarDirective]constructor');
    }
    /**
     * @return {?}
     */
    PluginToolbarDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // console.log('[PluginToolbarDirective]ngOnInit');
        // subscribe messages
        this.subscription = this.pluginManagerService.getMessage().subscribe((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            if (!message) {
                return;
            }
            _this.message = message || {};
            /** @type {?} */
            var data = _this.message['data'] || {};
            // console.log('data', data, this.getIdProperty(), data.hasOwnProperty(this.getIdProperty()));
            if (data.hasOwnProperty(_this.getIdProperty())) {
                /** @type {?} */
                var toolbar_1 = data[_this.getIdProperty()] || [];
                /** @type {?} */
                var pluginId = data['pluginId'];
                // console.log('[PluginToolbarDirective]--pluginId--', pluginId);
                /** @type {?} */
                var container = void 0;
                if (_this.coreService.getEnvironment().production && _this.message.source) { // production
                    if (_this.message.source.frameElement) {
                        pluginId = _this.message.source.frameElement.id;
                        if (pluginId) {
                            container = _this.initContainerById(pluginId);
                            if (window.origin === _this.message.source.origin && _this.pluginManagerService.hasPluginId(pluginId)) {
                                _this.generateToolbar(container, toolbar_1);
                                _this.notificationDetectByData(data);
                            }
                        }
                        else {
                            console.warn("Could not find plugin id");
                        }
                    }
                }
                else { // For testing with localhost
                    if (pluginId && _this.pluginManagerService.hasPluginId(pluginId)) {
                        container = _this.initContainerById(pluginId);
                        _this.generateToolbar(container, toolbar_1);
                    }
                    _this.notificationDetectByData(data);
                }
            }
        }));
    };
    /**
     * @private
     * @param {?=} data
     * @return {?}
     */
    PluginToolbarDirective.prototype.notificationDetectByData = /**
     * @private
     * @param {?=} data
     * @return {?}
     */
    function (data) {
        if (data === void 0) { data = {}; }
        if (data.hasOwnProperty('notify')) {
            var _a = data['notify'], title = _a.title, type = _a.type, msg = _a.msg;
            this.notifyService.showMessage(msg, type, title);
        }
    };
    /**
     * @return {?}
     */
    PluginToolbarDirective.prototype.getIdProperty = /**
     * @return {?}
     */
    function () {
        return this.idProperty;
    };
    /**
     * @return {?}
     */
    PluginToolbarDirective.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.renderer.setProperty(this.el.nativeElement, 'innerHTML', '');
    };
    /**
     * @param {?} pluginId
     * @return {?}
     */
    PluginToolbarDirective.prototype.initContainerById = /**
     * @param {?} pluginId
     * @return {?}
     */
    function (pluginId) {
        /** @type {?} */
        var prefix = "plugin-" + this.getIdProperty() + "-" + pluginId + "-";
        /** @type {?} */
        var containerId = prefix + 'container';
        /** @type {?} */
        var element = document.getElementById(containerId);
        /** @type {?} */
        var exists = false;
        if (element) {
            element.innerHTML = '';
            exists = true;
        }
        else {
            element = this.renderer.createElement('span');
            this.renderer.setProperty(element, 'id', containerId);
            // this.renderer.setProperty(element, 'innerHTML', '');
        }
        return new ToolbarContainer(prefix, pluginId, element, exists);
    };
    /**
     * @param {?} container
     * @param {?=} buttons
     * @return {?}
     */
    PluginToolbarDirective.prototype.generateToolbar = /**
     * @param {?} container
     * @param {?=} buttons
     * @return {?}
     */
    function (container, buttons) {
        var _this = this;
        if (buttons === void 0) { buttons = []; }
        var e_1, _a;
        /** @type {?} */
        var pluginId = container.pluginId;
        /** @type {?} */
        var plugin = this.pluginManagerService.getPluginById(pluginId);
        if (plugin && plugin.resize) {
            plugin.resize();
        }
        var _loop_1 = function (button) {
            var e_2, _a, e_3, _b;
            if (button) {
                /** @type {?} */
                var buttonEl = this_1.renderer.createElement('button');
                /** @type {?} */
                var buttonId_1 = container.prefix + button['id'];
                this_1.renderer.setProperty(buttonEl, 'id', buttonId_1);
                this_1.renderer.addClass(buttonEl, 'btn');
                this_1.renderer.addClass(buttonEl, 'btn-sm');
                this_1.renderer.addClass(buttonEl, 'ml-1');
                /** @type {?} */
                var classes = (button['cls'] || '').split(' ');
                try {
                    for (var classes_1 = tslib_1.__values(classes), classes_1_1 = classes_1.next(); !classes_1_1.done; classes_1_1 = classes_1.next()) {
                        var cls = classes_1_1.value;
                        this_1.renderer.addClass(buttonEl, cls);
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (classes_1_1 && !classes_1_1.done && (_a = classes_1.return)) _a.call(classes_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
                /** @type {?} */
                var iconEl = this_1.renderer.createElement('i');
                /** @type {?} */
                var iconClasses = (button['iconCls'] || '').split(' ');
                try {
                    for (var iconClasses_1 = tslib_1.__values(iconClasses), iconClasses_1_1 = iconClasses_1.next(); !iconClasses_1_1.done; iconClasses_1_1 = iconClasses_1.next()) {
                        var iconCls = iconClasses_1_1.value;
                        this_1.renderer.addClass(iconEl, iconCls);
                    }
                }
                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                finally {
                    try {
                        if (iconClasses_1_1 && !iconClasses_1_1.done && (_b = iconClasses_1.return)) _b.call(iconClasses_1);
                    }
                    finally { if (e_3) throw e_3.error; }
                }
                /** @type {?} */
                var text = this_1.renderer.createText(' ' + button['text'] || '');
                this_1.renderer.listen(buttonEl, 'click', (/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) {
                    // console.log('[' + pluginId + '] plugin call ' + button['eventName']);
                    _this.pluginManagerService.getContentWindowById(pluginId).then((/**
                     * @param {?} contentWindow
                     * @return {?}
                     */
                    function (contentWindow) {
                        if (contentWindow) {
                            _this.pluginManagerService.sendMessageTo({ id: buttonId_1, pluginId: pluginId, eventName: button['eventName'] }, contentWindow);
                        }
                    }));
                }));
                this_1.renderer.appendChild(buttonEl, iconEl);
                this_1.renderer.appendChild(buttonEl, text);
                this_1.renderer.appendChild(container.element, buttonEl);
            }
        };
        var this_1 = this;
        try {
            for (var buttons_1 = tslib_1.__values(buttons), buttons_1_1 = buttons_1.next(); !buttons_1_1.done; buttons_1_1 = buttons_1.next()) {
                var button = buttons_1_1.value;
                _loop_1(button);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (buttons_1_1 && !buttons_1_1.done && (_a = buttons_1.return)) _a.call(buttons_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (!container.exists) {
            this.renderer.appendChild(this.el.nativeElement, container.element);
        }
    };
    /**
     * @return {?}
     */
    PluginToolbarDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.clear();
        // console.log('[PluginToolbarDirective]destroy');
        // unsubscribe to ensure no memory leaks
        this.subscription.unsubscribe();
    };
    PluginToolbarDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[appPluginToolbar]'
                },] }
    ];
    /** @nocollapse */
    PluginToolbarDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: CoreService },
        { type: NotificationService },
        { type: PluginManagerService }
    ]; };
    PluginToolbarDirective.propDecorators = {
        idProperty: [{ type: Input }]
    };
    return PluginToolbarDirective;
}());
export { PluginToolbarDirective };
if (false) {
    /** @type {?} */
    PluginToolbarDirective.prototype.message;
    /** @type {?} */
    PluginToolbarDirective.prototype.subscription;
    /** @type {?} */
    PluginToolbarDirective.prototype.idProperty;
    /**
     * @type {?}
     * @private
     */
    PluginToolbarDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    PluginToolbarDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    PluginToolbarDirective.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    PluginToolbarDirective.prototype.notifyService;
    /**
     * @type {?}
     * @private
     */
    PluginToolbarDirective.prototype.pluginManagerService;
}
var ToolbarContainer = /** @class */ (function () {
    function ToolbarContainer(prefix, pluginId, element, exists) {
        this.prefix = prefix;
        this.pluginId = pluginId;
        this.element = element;
        this.exists = exists;
    }
    return ToolbarContainer;
}());
export { ToolbarContainer };
if (false) {
    /** @type {?} */
    ToolbarContainer.prototype.pluginId;
    /** @type {?} */
    ToolbarContainer.prototype.element;
    /** @type {?} */
    ToolbarContainer.prototype.prefix;
    /** @type {?} */
    ToolbarContainer.prototype.exists;
}
