// inet-ui/esm5/src/report-template/report-template.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from "inet-core";
import * as i0 from "@angular/core";
import * as i1 from "inet-core";
var ReportTemplateService = /** @class */ (function () {
    function ReportTemplateService(http) {
        this.http = http;
        this.url = {
            view: iNet.getPUrl('gl/exceltmpl/view'),
            delete: iNet.getPUrl('gl/exceltmpl/delete'),
            update: iNet.getPUrl('gl/exceltmpl/update'),
            add: iNet.getPUrl('gl/exceltmpl/create'),
            list: iNet.getPUrl('gl/exceltmpl/list'),
            application: iNet.getPUrl('cloud/subfirm/application'),
            organ: iNet.getPUrl('plugin/firmprofile/view'),
            modules: iNet.getPUrl('gl/exceltmpl/modules'),
            download: iNet.getPUrl('gl/filecontents/download')
        };
    }
    /**
     * @param {?} params
     * @return {?}
     */
    ReportTemplateService.prototype.list = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.postJSON(this.url.list, params);
    };
    /**
     * @param {?} uuid
     * @return {?}
     */
    ReportTemplateService.prototype.delete = /**
     * @param {?} uuid
     * @return {?}
     */
    function (uuid) {
        this.http.showLoading();
        return this.http.getJSON(this.url.delete, { uuid: uuid });
    };
    /**
     * @param {?} uuid
     * @return {?}
     */
    ReportTemplateService.prototype.view = /**
     * @param {?} uuid
     * @return {?}
     */
    function (uuid) {
        this.http.showLoading();
        return this.http.postJSON(this.url.view, { uuid: uuid });
    };
    /**
     * @param {?} params
     * @return {?}
     */
    ReportTemplateService.prototype.add = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.post(this.url.add, params);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    ReportTemplateService.prototype.update = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        this.http.showLoading();
        return this.http.post(this.url.update, params);
    };
    /**
     * @return {?}
     */
    ReportTemplateService.prototype.application = /**
     * @return {?}
     */
    function () {
        this.http.showLoading();
        return this.http.getJSON(this.url.application);
    };
    /**
     * @return {?}
     */
    ReportTemplateService.prototype.organId = /**
     * @return {?}
     */
    function () {
        this.http.showLoading();
        return this.http.postJSON(this.url.organ);
    };
    /**
     * @param {?} application
     * @return {?}
     */
    ReportTemplateService.prototype.modules = /**
     * @param {?} application
     * @return {?}
     */
    function (application) {
        this.http.showLoading();
        return this.http.getJSON(this.url.modules, { application: application });
    };
    /**
     * @param {?} contentUuid
     * @return {?}
     */
    ReportTemplateService.prototype.downloadFile = /**
     * @param {?} contentUuid
     * @return {?}
     */
    function (contentUuid) {
        return this.http.downloadFile(this.url.download, { docID: contentUuid });
    };
    ReportTemplateService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ReportTemplateService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    /** @nocollapse */ ReportTemplateService.ngInjectableDef = i0.defineInjectable({ factory: function ReportTemplateService_Factory() { return new ReportTemplateService(i0.inject(i1.HttpClientService)); }, token: ReportTemplateService, providedIn: "root" });
    return ReportTemplateService;
}());
export { ReportTemplateService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ReportTemplateService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateService.prototype.http;
}
