// inet-core/esm5/src/vertx/VertxSocket.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as EventBus from 'vertx3-eventbus-client';
import { Subject } from "rxjs";
var VertxSocket = /** @class */ (function () {
    function VertxSocket(username, url, headers) {
        this.onStateChange = new Subject();
        this.onMessage = new Subject();
        this.onMessageOnline = new Subject();
        this.readyState = VertxSocketState.UNSET;
        this.message = '';
        this.url = url;
        this.headers = headers || {};
        this.username = username;
        this.connect();
    }
    Object.defineProperty(VertxSocket.prototype, "unset", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === VertxSocketState.UNSET;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VertxSocket.prototype, "closed", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === VertxSocketState.CLOSED;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VertxSocket.prototype, "connecting", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === VertxSocketState.CONNECTING;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VertxSocket.prototype, "connected", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === VertxSocketState.CONNECTED;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    VertxSocket.prototype.connect = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.unset) {
            console.warn('Socket is initialed');
            return;
        }
        this._stateChange(VertxSocketState.CONNECTING);
        this.socket = new EventBus(this.url, {
            transports: 'websocket'
        });
        this.socket.onopen = (/**
         * @return {?}
         */
        function () {
            // Channel receive chat message
            _this.socket.registerHandler(_this.username, _this.headers, (/**
             * @param {?} err
             * @param {?} msg
             * @return {?}
             */
            function (err, msg) {
                if (msg) {
                    _this.onMessage.next(msg.body);
                }
            }));
            // Channel receive online message
            _this.socket.registerHandler('news-feed', _this.headers, (/**
             * @param {?} err
             * @param {?} msg
             * @return {?}
             */
            function (err, msg) {
                if (msg) {
                    _this.onMessageOnline.next((/** @type {?} */ (msg.body)));
                }
            }));
            // Connected
            setTimeout((/**
             * @return {?}
             */
            function () {
                if (!_this.closed) {
                    _this._stateChange(VertxSocketState.CONNECTED);
                }
            }), 1000);
        });
        this.socket.onclose = (/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return _this._close(e); });
    };
    /**
     * @param {?} address
     * @param {?} message
     * @return {?}
     */
    VertxSocket.prototype.send = /**
     * @param {?} address
     * @param {?} message
     * @return {?}
     */
    function (address, message) {
        if (!this.connected) {
            throw Error('Socket is not connected');
        }
        this.socket.publish(address, message || '');
    };
    /**
     * @return {?}
     */
    VertxSocket.prototype.close = /**
     * @return {?}
     */
    function () {
        this._close();
    };
    /**
     * @private
     * @param {?} state
     * @return {?}
     */
    VertxSocket.prototype._stateChange = /**
     * @private
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.readyState = state;
        this.onStateChange.next();
    };
    /**
     * @private
     * @param {?=} message
     * @return {?}
     */
    VertxSocket.prototype._close = /**
     * @private
     * @param {?=} message
     * @return {?}
     */
    function (message) {
        if (this.closed) {
            console.warn('Socket is closed');
            return;
        }
        this._stateChange(VertxSocketState.CLOSED);
        this.message = message;
        if (this.socket) {
            try {
                this.socket.close();
            }
            catch (e) { }
            this.socket = null;
        }
        this.onStateChange.unsubscribe();
        this.onMessage.unsubscribe();
        this.onMessageOnline.unsubscribe();
    };
    return VertxSocket;
}());
export { VertxSocket };
if (false) {
    /** @type {?} */
    VertxSocket.prototype.url;
    /** @type {?} */
    VertxSocket.prototype.headers;
    /** @type {?} */
    VertxSocket.prototype.username;
    /** @type {?} */
    VertxSocket.prototype.socket;
    /** @type {?} */
    VertxSocket.prototype.onStateChange;
    /** @type {?} */
    VertxSocket.prototype.onMessage;
    /** @type {?} */
    VertxSocket.prototype.onMessageOnline;
    /** @type {?} */
    VertxSocket.prototype.readyState;
    /** @type {?} */
    VertxSocket.prototype.message;
}
/** @enum {string} */
var VertxSocketState = {
    UNAVAILABLE: 'UNAVAILABLE',
    UNSET: 'UNSET',
    CONNECTING: 'CONNECTING',
    CONNECTED: 'CONNECTED',
    CLOSED: 'CLOSED',
    RE_CONNECTING: 'RE_CONNECTING',
};
export { VertxSocketState };
/**
 * @record
 */
export function VertxReceiveData() { }
if (false) {
    /** @type {?} */
    VertxReceiveData.prototype.type;
    /** @type {?} */
    VertxReceiveData.prototype.address;
    /** @type {?} */
    VertxReceiveData.prototype.body;
}
/**
 * @record
 */
export function VertxMessage() { }
if (false) {
    /** @type {?|undefined} */
    VertxMessage.prototype.uuid;
    /** @type {?|undefined} */
    VertxMessage.prototype.application;
    /** @type {?|undefined} */
    VertxMessage.prototype.message;
    /** @type {?|undefined} */
    VertxMessage.prototype.sender;
    /** @type {?|undefined} */
    VertxMessage.prototype.sent;
    /** @type {?|undefined} */
    VertxMessage.prototype.type;
}
/**
 * @record
 */
export function VertxMessageContent() { }
if (false) {
    /** @type {?|undefined} */
    VertxMessageContent.prototype.application;
    /** @type {?|undefined} */
    VertxMessageContent.prototype.alert;
    /** @type {?|undefined} */
    VertxMessageContent.prototype.sender;
    /** @type {?|undefined} */
    VertxMessageContent.prototype.content;
}
/**
 * @record
 */
export function VertxOnlineData() { }
if (false) {
    /** @type {?} */
    VertxOnlineData.prototype.type;
    /** @type {?} */
    VertxOnlineData.prototype.address;
}
/** @enum {string} */
var VertxOnlineType = {
    register: 'register',
    unregister: 'unregister',
};
export { VertxOnlineType };
