// inet-ui/esm5/src/front-view/front-view.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FrontViewDirective } from './front-view.directive';
import { FrontViewComponent } from "./component/viewer.component";
import { FrontViewService } from "./front-view.service";
var FrontViewModule = /** @class */ (function () {
    function FrontViewModule() {
    }
    FrontViewModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        FrontViewDirective,
                        FrontViewComponent
                    ],
                    declarations: [
                        FrontViewDirective,
                        FrontViewComponent
                    ],
                    entryComponents: [
                        FrontViewComponent
                    ],
                    providers: [FrontViewService]
                },] }
    ];
    return FrontViewModule;
}());
export { FrontViewModule };
