// inet-ui/esm5/src/web-token/lib/WebTokenSigner.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EXT_ACTION, EXT_CHROME_URL, EXT_FIREFOX_URL, EXT_HASH_ALG, NATIVE_APP_URL } from "./Constants";
import { CertUtils } from "./CertUtils";
var WebTokenSigner = /** @class */ (function () {
    function WebTokenSigner() {
        this.isChrome = /Google Inc/.test(navigator.vendor);
        this.isFirefox = /Firefox/.test(navigator.userAgent);
        this.isWindows = navigator.platform.indexOf('Win') > -1;
        this.osSupported = this.isWindows;
        this.browserSupported = this.isChrome || this.isFirefox;
        this.appInstalled = false;
        this.appInfo = {
            version: null,
            nativeVersion: null,
            nativeLatestVersion: null
        };
        this.textOsUnSupport = 'Tính năng đang hỗ trợ cho máy tính dùng Windows';
        this.textBrowserUnSupport = 'Tính năng đang hỗ trợ cho Chrome và FireFox';
        this.textExtNotInstalled = "\n    C\u1EA7n c\u00E0i \u0111\u1EB7t extension <a href=\"" + (this.isChrome ? EXT_CHROME_URL : EXT_FIREFOX_URL) + "\" target=\"_blank\">TokenSigner</a> tr\u00EAn tr\u00ECnh duy\u1EC7t Web\n    ";
        this.textAppNotInstalled = "\n    C\u1EA7n c\u00E0i \u0111\u1EB7t \u1EE9ng d\u1EE5ng k\u00FD <a href=\"" + NATIVE_APP_URL + "\" target=\"_blank\">TokenSigner</a> tr\u00EAn m\u00E1y t\u00EDnh\n    ";
        this.title = "Chứng thư số";
        this.description = "Chọn chứng thư số để ký";
        this.textOk = 'Đồng ý';
        this.textCancel = 'Đóng';
        this.certNotFound = 'Không tìm thấy chứng thư số';
        this.textIssuer = 'Nhà cung cấp';
        this.textValid = 'Hiệu lực';
        this.zIndex = 1100;
        this._onEscapeKey = this._onEscapeKey.bind(this);
    }
    Object.defineProperty(WebTokenSigner.prototype, "extInstalled", {
        get: /**
         * @return {?}
         */
        function () {
            return !!window['_ExtTokenBridge'];
        },
        enumerable: true,
        configurable: true
    });
    ;
    /**
     * @param {?=} silent
     * @return {?}
     */
    WebTokenSigner.prototype.checkReady = /**
     * @param {?=} silent
     * @return {?}
     */
    function (silent) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            if (_this.appInstalled) {
                return resolve(true);
            }
            if (!_this.osSupported) {
                if (!silent) {
                    _this._createCertDialog(null, _this.textOsUnSupport);
                }
                return resolve(false);
            }
            if (!_this.browserSupported) {
                if (!silent) {
                    _this._createCertDialog(null, _this.textBrowserUnSupport);
                }
                return resolve(false);
            }
            if (!_this.extInstalled) {
                if (!silent) {
                    _this._createCertDialog(null, _this.textExtNotInstalled);
                }
                return resolve(false);
            }
            _this._loadAppInfo().then((/**
             * @return {?}
             */
            function () {
                if (!_this.appInstalled && !silent) {
                    _this._createCertDialog(null, _this.textAppNotInstalled);
                }
                resolve(_this.appInstalled);
            }));
        }));
    };
    /**
     * @return {?}
     */
    WebTokenSigner.prototype.checkNativeUpdate = /**
     * @return {?}
     */
    function () {
        return this.appInstalled && this.appInfo.nativeVersion < this.appInfo.nativeLatestVersion;
    };
    /**
     * @return {?}
     */
    WebTokenSigner.prototype.loadCerts = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            _this.checkReady(true).then((/**
             * @param {?} ready
             * @return {?}
             */
            function (ready) {
                if (ready) {
                    _this._postExt(EXT_ACTION.GET_CERTS).then((/**
                     * @param {?} data
                     * @return {?}
                     */
                    function (data) {
                        /** @type {?} */
                        var certs = data && data.data || [];
                        certs.forEach((/**
                         * @param {?} cert
                         * @return {?}
                         */
                        function (cert) {
                            cert.subjectData = CertUtils.parseCertString(cert.subject);
                            cert.issuerData = CertUtils.parseCertString(cert.issuer);
                        }));
                        resolve(certs);
                    }));
                }
                else {
                    resolve([]);
                }
            }));
        }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    WebTokenSigner.prototype.signHash = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            if (data && (data.hash || data.hashs)) {
                _this.checkReady().then((/**
                 * @param {?} ready
                 * @return {?}
                 */
                function (ready) {
                    if (ready) {
                        data.hashAlg = data.hashAlg || EXT_HASH_ALG.SHA1;
                        _this._postExt(EXT_ACTION.SIGN_HASH, data).then(resolve);
                    }
                    else {
                        resolve(null);
                    }
                }));
            }
            else {
                resolve({
                    error: "HASH_MISSING"
                });
            }
        }));
    };
    /**
     * @param {?=} forceSelect
     * @return {?}
     */
    WebTokenSigner.prototype.selectCert = /**
     * @param {?=} forceSelect
     * @return {?}
     */
    function (forceSelect) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            _this.checkReady().then((/**
             * @param {?} ready
             * @return {?}
             */
            function (ready) {
                if (ready) {
                    _this.loadCerts().then((/**
                     * @param {?} certs
                     * @return {?}
                     */
                    function (certs) {
                        if (certs.length > 1 || forceSelect) {
                            _this._certCallback = resolve;
                            _this._createCertDialog(certs);
                        }
                        else {
                            resolve(certs[0]);
                        }
                    }));
                }
                else {
                    resolve(null);
                }
            }));
        }));
    };
    /**
     * @param {?} dateStr
     * @return {?}
     */
    WebTokenSigner.prototype.dateDisplay = /**
     * @param {?} dateStr
     * @return {?}
     */
    function (dateStr) {
        /** @type {?} */
        var date = new Date(dateStr);
        return [date.getDate(), date.getMonth() + 1, date.getFullYear()].join('/');
    };
    /**
     * @private
     * @param {?} action
     * @param {?=} data
     * @return {?}
     */
    WebTokenSigner.prototype._postExt = /**
     * @private
     * @param {?} action
     * @param {?=} data
     * @return {?}
     */
    function (action, data) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            if (_this.extInstalled) {
                window["_ExtTokenBridge"].postToExtension(action, data).then((/**
                 * @param {?} response
                 * @return {?}
                 */
                function (response) {
                    resolve(response);
                }));
            }
            else {
                resolve({
                    error: "EXTENSION_NOT_INSTALL"
                });
            }
        }));
    };
    /**
     * @private
     * @return {?}
     */
    WebTokenSigner.prototype._loadAppInfo = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            _this._postExt(EXT_ACTION.GET_INFO).then((/**
             * @param {?} reply
             * @return {?}
             */
            function (reply) {
                if (reply && reply.data) {
                    Object.assign(_this.appInfo, reply.data);
                }
                _this.appInstalled = !!_this.appInfo.nativeVersion;
                resolve();
            }));
        }));
    };
    /**
     * @private
     * @param {?} certs
     * @param {?=} message
     * @return {?}
     */
    WebTokenSigner.prototype._createCertDialog = /**
     * @private
     * @param {?} certs
     * @param {?=} message
     * @return {?}
     */
    function (certs, message) {
        var _this = this;
        if (this._dialog) {
            return;
        }
        this._dialog = document.createElement('div');
        this._dialog.classList.add('webtoken-dialog');
        this._dialog.style.zIndex = this.zIndex;
        this._certs = certs;
        /** @type {?} */
        var msg;
        if (message) {
            msg = '<div class="webtoken-dialog__message">' + message + '</div>';
        }
        else {
            msg = certs.length < 1 ? "<div class=\"webtoken-dialog__message\">" + this.certNotFound + "</div>" : this._buildCertHtml(certs);
        }
        this._dialog.innerHTML = "\n        <style>\n        .webtoken-dialog {\n            position: fixed;\n            top: 0;\n            right: 0;\n            bottom: 0;\n            left: 0;\n            background: rgba(0, 0, 0, .3);\n        }\n        .webtoken-dialog__container {\n            width: 400px;\n            margin: auto;\n            background: #fff;\n            overflow: hidden;\n            color: #333;\n            position: absolute;\n            transform: translate(-50%, -50%);\n            top: 50%;\n            left: 50%;\n            border-radius: 4px;\n        }\n        .webtoken-dialog__header {\n            padding: 10px 15px;\n            border-bottom: 1px solid #ddd;\n        }\n        .webtoken-dialog__title {\n            font-size: 20px;\n            font-weight: bold;\n            margin-bottom: 5px;\n        }\n        .webtoken-dialog__desc {\n            font-size: 14px;\n            line-height: 20px;\n        }\n        .webtoken-dialog__close {\n            float: right;\n            width: 30px;\n            cursor: pointer;\n            text-align: center;\n            line-height: 30px;\n            margin: -5px 0;\n            font-size: 20px;\n        }\n        .webtoken-dialog__body {\n            max-height: 250px;\n            overflow: auto;\n            font-size: 14px;\n            line-height: 20px;\n        }\n        .webtoken-dialog__message {\n            padding: 20px;            \n        }\n        .webtoken-cert__item {\n            padding: 10px 20px;\n            cursor: pointer;\n            line-height: 20px;\n        }\n        .webtoken-cert__item.active {\n            background: #ddd !important;\n        }\n        .webtoken-cert__item:hover {\n            background: #f5f5f5;\n        }\n        .webtoken-cert__title {\n            font-size: 18px;\n            margin-bottom: 2px;\n        }\n        .webtoken-cert__text {\n            font-size: 14px;\n        }\n        .webtoken-dialog__footer {\n            padding: 10px 20px;\n            border-top: 1px solid #ddd;\n            text-align: right;\n        }\n        .webtoken-dialog__button {\n            height: 30px;\n            line-height: 30px;\n            padding: 0 10px;\n            min-width: 60px;\n            margin-left: 10px;\n            cursor: pointer;\n            border-radius: 4px;\n            font-weight: bold;\n            background: #fff;\n            border: 1px solid #ddd;\n        }\n        .webtoken-dialog__button:hover {\n            opacity: 8;\n        }\n        .btn-ok {\n            background: #007bff;\n            border-color: #007bff;\n            color: #fff;\n        }\n        </style>\n        <div class=\"webtoken-dialog__container\">\n        <div class=\"webtoken-dialog__header\">\n        <div class=\"webtoken-dialog__close\">\u00D7</div>\n        <div class=\"webtoken-dialog__title\">" + this.title + "</div>\n        " + (message ? '' : "<div class=\"webtoken-dialog__desc\">" + this.description + "</div>") + "\n        </div>\n        <div class=\"webtoken-dialog__body\">" + msg + "</div>\n        <div class=\"webtoken-dialog__footer\">\n        <button class=\"btn-close webtoken-dialog__button\">" + this.textCancel + "</button>\n        " + (message ? '' : "<button class=\"btn-ok webtoken-dialog__button\">" + this.textOk + "</button>") + "\n        </div>\n        </div>\n        ";
        /** @type {?} */
        var certEl = this._dialog.getElementsByClassName('webtoken-cert__item');
        // Close dialog
        this._dialog.addEventListener('click', (/**
         * @return {?}
         */
        function () { return _this._closeDialog(); }));
        this._dialog.getElementsByClassName('webtoken-dialog__close')[0].addEventListener('click', (/**
         * @return {?}
         */
        function () { return _this._closeDialog(); }));
        // Prevent close dialog
        this._dialog.getElementsByClassName('webtoken-dialog__container')[0].addEventListener('click', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            e.stopPropagation();
        }));
        this._dialog.getElementsByClassName('btn-close')[0].addEventListener('click', (/**
         * @return {?}
         */
        function () { return _this._closeDialog(); }));
        if (certs && certs.length > 0) {
            // Ok select
            this._dialog.getElementsByClassName('btn-ok')[0].addEventListener('click', (/**
             * @return {?}
             */
            function () {
                if (_this._certEl) {
                    /** @type {?} */
                    var index = Number(_this._certEl.getAttribute('index'));
                    if (_this._certs[index]) {
                        _this._certCallback(_this._certs[index]);
                    }
                }
                else {
                    _this._certCallback(null);
                }
                _this._closeDialog();
            }));
            // On click cert item
            for (var i = 0; i < certEl.length; i++) {
                certEl[i].addEventListener('click', (/**
                 * @param {?} e
                 * @return {?}
                 */
                function (e) { return _this._onClickCertEl(e); }));
            }
            // Select first cert
            certEl[0].click();
        }
        // Esc key close dialog
        document.addEventListener('keydown', this._onEscapeKey);
        document.body.appendChild(this._dialog);
    };
    /**
     * @private
     * @param {?} certs
     * @return {?}
     */
    WebTokenSigner.prototype._buildCertHtml = /**
     * @private
     * @param {?} certs
     * @return {?}
     */
    function (certs) {
        var _this = this;
        /** @type {?} */
        var html = '';
        certs.forEach((/**
         * @param {?} cert
         * @param {?} index
         * @return {?}
         */
        function (cert, index) {
            html += "\n            <div class=\"webtoken-cert__item\" index=\"" + index + "\">\n            <div class=\"webtoken-cert__title\">" + cert.subjectData.CN + "</div>\n            <div class=\"webtoken-cert__text\">" + _this.textIssuer + ": " + cert.issuerData.CN + "</div>\n            <div class=\"webtoken-cert__text\">" + _this.textValid + ": " + _this.dateDisplay(cert.validFrom) + " - " + _this.dateDisplay(cert.validTo) + "</div>\n            </div>\n            ";
        }));
        return html;
    };
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    WebTokenSigner.prototype._onClickCertEl = /**
     * @private
     * @param {?} e
     * @return {?}
     */
    function (e) {
        /** @type {?} */
        var certEl = this._findCerEl(e.target);
        if (certEl.classList.contains('active')) {
            return;
        }
        if (this._certEl) {
            this._certEl.classList.remove('active');
        }
        certEl.classList.add('active');
        this._certEl = certEl;
    };
    /**
     * @private
     * @param {?} certEl
     * @return {?}
     */
    WebTokenSigner.prototype._findCerEl = /**
     * @private
     * @param {?} certEl
     * @return {?}
     */
    function (certEl) {
        if (!certEl.classList.contains('webtoken-cert__item')) {
            return this._findCerEl(certEl.parentNode);
        }
        return certEl;
    };
    /**
     * @private
     * @return {?}
     */
    WebTokenSigner.prototype._closeDialog = /**
     * @private
     * @return {?}
     */
    function () {
        document.body.removeChild(this._dialog);
        document.removeEventListener('keydown', this._onEscapeKey);
        this._dialog = null;
        this._certEl = null;
        this._certCallback = null;
    };
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    WebTokenSigner.prototype._onEscapeKey = /**
     * @private
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.which === 27) {
            this._closeDialog();
        }
    };
    return WebTokenSigner;
}());
export { WebTokenSigner };
if (false) {
    /** @type {?} */
    WebTokenSigner.prototype.isChrome;
    /** @type {?} */
    WebTokenSigner.prototype.isFirefox;
    /** @type {?} */
    WebTokenSigner.prototype.isWindows;
    /** @type {?} */
    WebTokenSigner.prototype.osSupported;
    /** @type {?} */
    WebTokenSigner.prototype.browserSupported;
    /** @type {?} */
    WebTokenSigner.prototype.appInstalled;
    /** @type {?} */
    WebTokenSigner.prototype.appInfo;
    /** @type {?} */
    WebTokenSigner.prototype.textOsUnSupport;
    /** @type {?} */
    WebTokenSigner.prototype.textBrowserUnSupport;
    /** @type {?} */
    WebTokenSigner.prototype.textExtNotInstalled;
    /** @type {?} */
    WebTokenSigner.prototype.textAppNotInstalled;
    /** @type {?} */
    WebTokenSigner.prototype.title;
    /** @type {?} */
    WebTokenSigner.prototype.description;
    /** @type {?} */
    WebTokenSigner.prototype.textOk;
    /** @type {?} */
    WebTokenSigner.prototype.textCancel;
    /** @type {?} */
    WebTokenSigner.prototype.certNotFound;
    /** @type {?} */
    WebTokenSigner.prototype.textIssuer;
    /** @type {?} */
    WebTokenSigner.prototype.textValid;
    /** @type {?} */
    WebTokenSigner.prototype.zIndex;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._dialog;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._certEl;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._certs;
    /**
     * @type {?}
     * @private
     */
    WebTokenSigner.prototype._certCallback;
    /* Skipping unhandled member: ;*/
}
