// inet-ui/esm2015/src/translate/cloud-trans-loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Observable } from 'rxjs';
import 'rxjs/add/observable/of';
import { defineLocale } from 'ngx-bootstrap/chronos';
import { systemViLocale } from "./system-vi";
import { systemEnLocale } from "./system-en";
export class CloudTransLoader {
    constructor() {
        this.store = {};
        this.defineLocale(systemViLocale.abbr, systemViLocale);
        this.defineLocale(systemEnLocale.abbr, systemEnLocale);
    }
    /**
     * @param {?} lang
     * @param {?} data
     * @return {?}
     */
    assign(lang, data) {
        return Object.assign(this.getResourceByLang(lang), data);
    }
    /**
     * @param {?} lang
     * @param {?} data
     * @return {?}
     */
    defineLocale(lang, data) {
        defineLocale(lang, data);
        this.store[lang] = data;
    }
    /**
     * @param {?} lang
     * @return {?}
     */
    getTranslation(lang) {
        return Observable.of(this.getResourceByLang(lang));
    }
    /**
     * @param {?} lang
     * @return {?}
     */
    getResourceByLang(lang) {
        return this.store[lang] || systemEnLocale;
    }
    /**
     * Convert complex js object to dot notation js object
     * @private
     * @param {?} message
     * @param {?} value
     * @param {?} obj
     * @return {?}
     */
    dotizeStringToJSON(message, value, obj) {
        /** @type {?} */
        const parts = message.split(".");
        /** @type {?} */
        const last = parts.pop();
        /** @type {?} */
        let part;
        while (part = parts.shift()) {
            if (typeof obj[part] !== "object") {
                obj[part] = {};
            }
            obj = obj[part];
        }
        obj[last] = value;
    }
    /**
     * Convert Java .properties files data to JSON
     * @private
     * @param {?} str
     * @return {?}
     */
    propertiesToJSON(str) {
        // Concat lines that end with '\'.
        // Split by line breaks.
        // Remove commented lines
        // Create the JSON
        return str.replace(/\\\n/, "").split("\n")
            .filter((/**
         * @param {?} line
         * @return {?}
         */
        (line) => /(\#|\!)/.test(line.replace(/\s/g, "").slice(0, 1)) ? false : line))
            .reduce((/**
         * @param {?} obj
         * @param {?} line
         * @return {?}
         */
        (obj, line) => {
            /** @type {?} */
            const colonifiedLine = line.replace(/(\=)/, ":");
            /** @type {?} */
            const key = colonifiedLine.substring(0, colonifiedLine.indexOf(":")).trim();
            /** @type {?} */
            const value = this.asciiToNative(colonifiedLine.substring(colonifiedLine.indexOf(":") + 1).trim());
            obj[key] = value;
            return obj;
        }), {});
    }
    /*
        * Convert ASCII value to Native Character
         */
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    asciiToNative(value) {
        /** @type {?} */
        let character = value.split("\\u");
        /** @type {?} */
        let nativeStr = character[0];
        for (let i = 1; i < character.length; i++) {
            /** @type {?} */
            let code = character[i];
            nativeStr += String.fromCharCode(parseInt("0x" + code.substring(0, 4)));
            if (code.length > 4) {
                nativeStr += code.substring(4, code.length);
            }
        }
        return nativeStr;
    }
    /*
        * Convert Native value to ASCII
         */
    /**
     * @private
     * @param {?} value
     * @param {?=} ignoreLatin
     * @return {?}
     */
    native2Ascii(value, ignoreLatin = false) {
        /** @type {?} */
        let character = value.split("");
        /** @type {?} */
        let asciiStr = "";
        for (let i = 0; i < character.length; i++) {
            /** @type {?} */
            const code = Number(character[i].charCodeAt(0));
            if (!ignoreLatin || code > 127) {
                /** @type {?} */
                let charAscii = code.toString(16);
                charAscii = new String("0000").substring(charAscii.length, 4) + charAscii;
                asciiStr += "\\u" + charAscii;
            }
            else {
                asciiStr += character[i];
            }
        }
        return asciiStr;
    }
    /**
     * Convert items from the server to Translation
     * @param {?} items
     * @return {?}
     */
    convertResourceToObject(items) {
        /** @type {?} */
        let translation = {};
        for (const item of items) {
            this.dotizeStringToJSON(item['messageKey'], item['text'], translation);
        }
        return translation;
    }
    /**
     * Convert Java .properties files data to Translation
     * @param {?} str
     * @return {?}
     */
    convertPropertiesToObject(str) {
        /** @type {?} */
        let translation = {};
        /** @type {?} */
        const json = this.propertiesToJSON(str);
        for (const key in json) {
            this.dotizeStringToJSON(key, json[key], translation);
        }
        return translation;
    }
    /**
     * @param {?} lang
     * @param {?} items
     * @return {?}
     */
    setResources(lang, items) {
        this.assign(lang, this.convertResourceToObject(items));
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    CloudTransLoader.prototype.store;
}
