// inet-ui/esm5/src/navigation-tab/navigation-tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ComponentFactoryResolver, EventEmitter, Input, Output, ViewContainerRef } from '@angular/core';
import { NavigationEnd, Router } from "@angular/router";
import { filter } from "rxjs/operators";
var NavigationTab = /** @class */ (function () {
    function NavigationTab(componentFactoryResolver, router) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.router = router;
        this.prefix = '';
        this.onChange = new EventEmitter();
        this.onLoad = new EventEmitter();
    }
    /**
     * @return {?}
     */
    NavigationTab.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._router = this.router.events
            .pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return event instanceof NavigationEnd; })))
            .subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this._activeTab();
        }));
        this._activeTab();
    };
    /**
     * @return {?}
     */
    NavigationTab.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._router) {
            this._router.unsubscribe();
        }
    };
    /**
     * @param {?} tab
     * @return {?}
     */
    NavigationTab.prototype.activeTab = /**
     * @param {?} tab
     * @return {?}
     */
    function (tab) {
        if (tab.active) {
            return;
        }
        this.router.navigate([this.getFullPath(tab)]);
        this.onChange.emit(tab);
    };
    /**
     * @param {?} path
     * @return {?}
     */
    NavigationTab.prototype.getTabByPath = /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        // get tab with path matching
        for (var i = 0; i < this.tabs.length; i++) {
            if (this.getFullPath(this.tabs[i]) === path) {
                return this.tabs[i];
            }
        }
        return this.tabs[0];
    };
    /**
     * @param {?} tab
     * @return {?}
     */
    NavigationTab.prototype.getFullPath = /**
     * @param {?} tab
     * @return {?}
     */
    function (tab) {
        return this.prefix + tab.path;
    };
    /**
     * @private
     * @param {?} tab
     * @return {?}
     */
    NavigationTab.prototype._loadComponent = /**
     * @private
     * @param {?} tab
     * @return {?}
     */
    function (tab) {
        if (this.tabActive === tab) {
            return;
        }
        this.tabActive = tab;
        this.viewContainerRef.clear();
        if (tab.component) {
            /** @type {?} */
            var componentFactory = this.componentFactoryResolver.resolveComponentFactory(tab.component);
            this.componentRef = this.viewContainerRef.createComponent(componentFactory);
            this.onLoad.emit(this.componentRef);
        }
    };
    /**
     * @private
     * @return {?}
     */
    NavigationTab.prototype._activeTab = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var tab = this.getTabByPath(this.router.url);
        if (tab) {
            this.tabs.forEach((/**
             * @param {?} _tab
             * @return {?}
             */
            function (_tab) { return _tab.active = false; }));
            this._loadComponent(tab);
            tab.active = true;
        }
    };
    NavigationTab.decorators = [
        { type: Component, args: [{
                    selector: '[navigationTab]',
                    template: "<div class=\"navigation-tab\">\n  <ul class=\"nav navigation-tab__content\">\n    <li *ngFor=\"let tab of tabs\" class=\"nav-item\" [ngClass]=\"{'active': tab.active}\">\n      <span class=\"nav-link\" (click)=\"activeTab(tab)\">{{tab.name}}</span>\n    </li>\n  </ul>\n</div>\n",
                    styles: [".navigation-tab{padding-top:3px;overflow:auto}.nav-item,.navigation-tab__content{white-space:nowrap;display:inline-block}.nav-item{display:inline-block;color:rgba(0,0,0,.3);text-transform:uppercase;cursor:pointer;font-size:14px;font-weight:700}.nav-item:hover{color:rgba(0,0,0,.7)}.nav-item.active{color:#0062cc;border-bottom:2px solid #0062cc}"]
                }] }
    ];
    /** @nocollapse */
    NavigationTab.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: Router }
    ]; };
    NavigationTab.propDecorators = {
        tabs: [{ type: Input }],
        viewContainerRef: [{ type: Input }],
        prefix: [{ type: Input }],
        onChange: [{ type: Output }],
        onLoad: [{ type: Output }]
    };
    return NavigationTab;
}());
export { NavigationTab };
if (false) {
    /** @type {?} */
    NavigationTab.prototype.tabs;
    /** @type {?} */
    NavigationTab.prototype.viewContainerRef;
    /** @type {?} */
    NavigationTab.prototype.prefix;
    /** @type {?} */
    NavigationTab.prototype.onChange;
    /** @type {?} */
    NavigationTab.prototype.onLoad;
    /** @type {?} */
    NavigationTab.prototype.tabActive;
    /** @type {?} */
    NavigationTab.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    NavigationTab.prototype._router;
    /**
     * @type {?}
     * @private
     */
    NavigationTab.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    NavigationTab.prototype.router;
}
/**
 * @record
 */
export function NavigationData() { }
if (false) {
    /** @type {?} */
    NavigationData.prototype.name;
    /** @type {?} */
    NavigationData.prototype.component;
    /** @type {?} */
    NavigationData.prototype.path;
    /** @type {?|undefined} */
    NavigationData.prototype.active;
}
