// inet-ui/esm5/src/grid/components/header/header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, forwardRef } from '@angular/core';
import { DataTable } from '../table/table.component';
var DataTableHeader = /** @class */ (function () {
    function DataTableHeader(dataTable) {
        this.dataTable = dataTable;
    }
    DataTableHeader.decorators = [
        { type: Component, args: [{
                    moduleId: 'grid-module',
                    selector: 'data-table-header',
                    template: "<div class=\"data-table-header\">\n    <h4 class=\"title\" [textContent]=\"dataTable.headerTitle\"></h4>\n</div>",
                    styles: [".data-table-header{min-height:25px;margin-bottom:10px}.title{display:inline-block;margin:5px 0 0 5px}"]
                }] }
    ];
    /** @nocollapse */
    DataTableHeader.ctorParameters = function () { return [
        { type: DataTable, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return DataTable; })),] }] }
    ]; };
    return DataTableHeader;
}());
export { DataTableHeader };
if (false) {
    /** @type {?} */
    DataTableHeader.prototype.dataTable;
}
