// inet-ui/esm5/src/email-template/email-template-edit/email-template-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { FormBuilder, Validators } from "@angular/forms";
import { of, Subject } from "rxjs";
import { ErrorMessage, NotificationService } from "inet-core";
import { EmailTemplateService } from "../email-template.service";
import { TranslateService } from "@ngx-translate/core";
import { EmailTemplate } from "../email-template";
var EmailTemplateEditComponent = /** @class */ (function () {
    function EmailTemplateEditComponent(fb, notification, translate, emailTemplateService) {
        var _this = this;
        this.fb = fb;
        this.notification = notification;
        this.translate = translate;
        this.emailTemplateService = emailTemplateService;
        this.email = new EmailTemplate();
        this.visibleProperty = false; //False to hide the Property Panel.
        this.emailParams = [];
        this.editorConfig = {
            "editable": true,
            "spellcheck": true,
            "height": "calc(100vh - 205px)",
            "minHeight": "100%",
            "width": "100%",
            "minWidth": "0",
            "translate": "yes",
            "enableToolbar": true,
            "showToolbar": true,
            "placeholder": "",
            "imageEndPoint": "",
            "toolbar": [
                ["bold", "italic", "underline", "strikeThrough"],
                ["fontName", "fontSize", "color"],
                ["justifyLeft", "justifyCenter", "justifyRight", "justifyFull", "indent", "outdent"],
                ["emailParam"]
            ]
        };
        this.onValidate = new EventEmitter();
        this.onUpdate = new EventEmitter();
        this.onToggleProperty = new EventEmitter();
        this.formChanged = new Subject();
        this.translateSubscription = translate.get(['COMMON.MODULE.EMAIL_TEMPLATE']).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.columnsTranslations = data['COMMON.MODULE.EMAIL_TEMPLATE'];
        }));
    }
    /**
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initForm();
    };
    /**
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.translateSubscription.unsubscribe();
    };
    /**
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.initForm = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.emailTemplateForm = this.fb.group({
            templateId: [null],
            emailCode: [null, Validators.required],
            emailName: [null, Validators.required],
            emailSubject: [null],
            emailContent: [null],
            inactive: [null]
        });
        this.formChanged.debounceTime(250).distinctUntilChanged().subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            _this.onValidate.emit(v);
        }));
        this.emailTemplateForm.valueChanges.subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            _this.formChanged.next(_this.emailTemplateForm.valid);
        }));
    };
    /**
     * @param {?} emailId
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.loadById = /**
     * @param {?} emailId
     * @return {?}
     */
    function (emailId) {
        var _this = this;
        if (emailId) {
            this.emailTemplateService.loadById(emailId)
                .switchMap((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (data.type !== ErrorMessage.TYPE) {
                    _this.setData(tslib_1.__assign({}, _this.email, data));
                    return _this.emailTemplateService.listParams(_this.email.application, _this.email.srvVariable);
                }
                else {
                    return of(null);
                }
            }))
                .subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                _this.emailParams = result.items || [];
            }));
        }
    };
    /**
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.toggleProperty = /**
     * @return {?}
     */
    function () {
        this.visibleProperty = !this.visibleProperty;
        this.onToggleProperty.emit(this.visibleProperty);
        return this.visibleProperty;
    };
    /**
     * @param {?} emailTemplate
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.setData = /**
     * @param {?} emailTemplate
     * @return {?}
     */
    function (emailTemplate) {
        this.email = emailTemplate;
        this.email.templateId = this.email.uuid;
        this.email.inactive = !this.email.inactive; //active
        this.emailTemplateForm.patchValue(emailTemplate);
    };
    /**
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.clearData = /**
     * @return {?}
     */
    function () {
        this.emailTemplateForm.reset();
    };
    /**
     * @param {?} placeholderContent
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.setEditorPlaceholder = /**
     * @param {?} placeholderContent
     * @return {?}
     */
    function (placeholderContent) {
        this.editorConfig.placeholder = placeholderContent;
    };
    /**
     * @private
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.getData = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var data = this.emailTemplateForm.getRawValue();
        data.inactive = !data.inactive; //active
        return data;
    };
    /**
     * @return {?}
     */
    EmailTemplateEditComponent.prototype.update = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.emailTemplateForm.valid && this.email && this.email.uuid) {
            var _a = this.columnsTranslations, UPDATE_SUCCESS_1 = _a.UPDATE_SUCCESS, UPDATE_ERROR_1 = _a.UPDATE_ERROR, EMAIL_1 = _a.EMAIL;
            this.emailTemplateService.update(this.getData()).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (data.type !== ErrorMessage.TYPE) {
                    _this.setData(tslib_1.__assign({}, _this.email, data));
                    _this.notification.showMessage(UPDATE_SUCCESS_1, 'success', EMAIL_1);
                }
                else {
                    _this.notification.showMessage(UPDATE_ERROR_1, 'error', EMAIL_1);
                }
                _this.onUpdate.emit(data);
            }));
        }
    };
    EmailTemplateEditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-email-template-edit',
                    template: "<div class=\"row center-body m-0\">\n    <div class=\"row justify-content-center mt-0 mb-0 reponsive-card h-100\" style=\"width: 100%\">\n        <div class=\" card-body-shadow box col-sm-12 p-0\">\n            <div class=\"row ml-0 mr-0 h-100\">\n                <form [formGroup]=\"emailTemplateForm\" class=\"w-100\">\n                    <div class=\" form-group row m-0 w-100\">\n                        <div class=\"pl-0 pr-0 mt-1 mb-0\"\n                             [ngClass]=\"{'col-sm-9':visibleProperty,'col-sm-12':!visibleProperty}\">\n                            <div class=\"col-sm-12 p-0\">\n                                <div class=\"col-sm-12 p-0\">\n                                    <app-ngx-editor [emailParams]=\"emailParams\"\n                                                    [config]=\"editorConfig\" [formControl]=\"emailTemplateForm.controls['emailContent']\"\n                                                    [spellcheck]=\"true\" formControlName=\"emailContent\"></app-ngx-editor>\n                                </div>\n                            </div>\n                        </div>\n                        <div class=\"col-sm-3 px-0 pl-3 mt-2 mb-0\" [hidden]=\"!visibleProperty\">\n                            <div class=\"form-group row m-0 col-sm-12 mb-2 pl-0\">\n                                <label for=\"emailCode\" class=\"col-sm-4 text-left p-0 col-form-label \">\n                                    {{'COMMON.MODULE.EMAIL_TEMPLATE.EMAIL_CODE'| translate}}<span class=\"required\"> :</span>\n                                </label>\n                                <div class=\"col-sm-12 p-0 pl-1\">\n                                    <input type=\"text\"\n                                           class=\"form-control form-control-sm\"\n                                           name=\"emailCode\" [formControl]=\"emailTemplateForm.controls['emailCode']\"\n                                           [ngClass]=\"{'is-invalid': emailTemplateForm.controls['emailCode'].hasError('required') && emailTemplateForm.controls['emailCode'].touched}\"\n                                           required id=\"emailCode\" value=\"\" readonly>\n                                    <div *ngIf=\"emailTemplateForm.controls['emailCode'].hasError('required') && emailTemplateForm.controls['emailCode'].touched\"\n                                         class=\"text-danger mt-2 text-df\">\n                                        <div>\n                                            <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                            {{'COMMON.MODULE.EMAIL_TEMPLATE.EMAIL_CODE_REQUIRED' | translate}}\n                                        </div>\n                                    </div>\n                                </div>\n                            </div>\n                            <div class=\"col-sm-12 form-group row m-0 mb-2 pl-0\">\n                                <label for=\"emailName\"\n                                       class=\"col-sm-4 text-left p-0 col-form-label \">\n                                    {{'COMMON.MODULE.EMAIL_TEMPLATE.EMAIL_NAME' | translate}}<span class=\"required\"> :</span></label>\n                                <div class=\"col-sm-12 p-0\">\n                                    <input type=\"text\" class=\"form-control form-control-sm \"\n                                           formControlName=\"emailName\" [formControl]=\"emailTemplateForm.controls['emailName']\"\n                                           [ngClass]=\"{'is-invalid': emailTemplateForm.controls['emailName'].hasError('required') && emailTemplateForm.controls['emailName'].touched}\"\n                                           id=\"emailName\" name=\"emailName\" value=\"\" required>\n                                    <div *ngIf=\"emailTemplateForm.controls['emailName'].hasError('required') && emailTemplateForm.controls['emailName'].touched\"\n                                         class=\"text-danger mt-2 text-df\">\n                                        <div>\n                                            <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                            {{'COMMON.MODULE.EMAIL_TEMPLATE.EMAIL_NAME_REQUIRED' | translate}}\n                                        </div>\n                                    </div>\n                                </div>\n                            </div>\n\n                            <div class=\"form-group row m-0 col-sm-12 pl-0 pr-0 mb-2\">\n                                <div class=\"col-sm-12 form-group row m-0 pl-0\">\n                                    <label class=\"col-sm-4 text-left p-0 col-form-label \">\n                                        {{'COMMON.MODULE.EMAIL_TEMPLATE.EMAIL_SUBJECT' | translate}}:</label>\n                                    <div class=\"col-sm-12 p-0\">\n                                    <textarea type=\"text\" rows=\"5\" style=\"resize: none;\"\n                                              class=\"form-control form-control-sm \"\n                                              [formControl]=\"emailTemplateForm.controls['emailSubject']\"\n                                              name=\"emailSubject\" id=\"emailSubject\"></textarea>\n                                    </div>\n                                </div>\n                            </div>\n                            <div class=\"form-group row m-0 col-sm-12 pl-0 pr-0 mb-2\">\n                                <label class=\"control-label col-form-label col-sm-4 pr-0 pl-0 pt-0\">{{'COMMON.MODULE.EMAIL_TEMPLATE.ACTIVE'\n                                    |translate}} : </label>\n                                <div class=\"col-sm-8\">\n                                    <input id=\"inactive\" name=\"inactive\"\n                                           [formControl]=\"emailTemplateForm.controls['inactive']\"\n                                           class=\"switch switch-success\" type=\"checkbox\"/>\n                                    <span class=\"lbl\"></span>\n                                </div>\n                            </div>\n                        </div>\n                    </div>\n                </form>\n            </div>\n        </div>\n    </div>\n</div>\n",
                    styles: ["input[type=checkbox].switch{width:55px;height:25px}input[type=checkbox].switch+.lbl{margin:0 4px;min-height:24px}input[type=checkbox].switch+.lbl::before{font-family:'Open Sans';content:\"ON\\a0\\a0\\a0\\a0\\a0\\a0\\a0\\a0\\a0OFF\";color:#999;text-shadow:0 0 0 #999;font-weight:400;font-size:11px;line-height:17px;height:20px;overflow:hidden;border-radius:12px;background-color:#f5f5f5;box-shadow:inset 0 1px 1px 0 rgba(0,0,0,.15);border:1px solid #ccc;text-align:left;padding:0;width:52px;text-indent:-21px;margin-right:0;transition:text-indent .3s;top:auto}input[type=checkbox].switch+.lbl::after{font-family:'Open Sans';content:'III';font-size:12px;font-weight:400;letter-spacing:0;color:#aaa;text-shadow:none;background-color:#fff;border-radius:100%;width:22px;height:22px;line-height:22px;text-align:center;position:absolute;top:-2px;left:-3px;box-shadow:0 1px 1px 1px rgba(0,0,0,.3);transition:left .3s}input[type=checkbox].switch:checked+.lbl::before{text-indent:8px;color:#fff;text-shadow:0 0 0 #fff;background-color:#8ab2c9;border-color:#6a8ca8}input[type=checkbox].switch:checked+.lbl::after{left:34px;background-color:#fff;color:#98a0a5}input[type=checkbox].switch.switch-success+.lbl{position:relative}input[type=checkbox].switch.switch-success+.lbl::before{font-family:FontAwesome;content:\"\\f00d\";text-shadow:0 -1px 0 rgba(0,0,0,.25);box-shadow:none;border-width:0;font-weight:lighter;font-size:16px;border-radius:12px;display:inline-block;background-color:#888;color:#f2f2f2;width:52px;height:22px;line-height:21px;text-indent:32px;transition:background .1s}input[type=checkbox].switch.switch-success+.lbl::after{content:'';text-shadow:0 -1px 0 rgba(0,0,0,.25);position:absolute;top:2px;left:3px;border-radius:12px;box-shadow:0 -1px 0 rgba(0,0,0,.25);width:18px;height:18px;text-align:center;background-color:#f2f2f2;border:4px solid #f2f2f2;transition:left .2s}input[type=checkbox].switch.switch-success:checked+.lbl::before{content:\"\\f00c\";text-indent:6px;color:#fff;border-color:#b7d3e5;background-color:#2abb2a}input[type=checkbox].switch.switch-success:checked+.lbl::after{left:32px;background-color:#fff;border:4px solid #fff;text-shadow:0 -1px 0 rgba(0,200,0,.25)}@media (max-width:700px){:host .reponsive-card{padding:0 30px}}:host .pointer{pointer-events:none}:host .custom-center-span{display:flex;align-items:center}:host .center-body{flex-direction:row;justify-content:center;align-content:start}:host .cursor{cursor:pointer}:host .cursor:hover{color:#f22858}:host .custom-textarea{overflow-y:auto;max-width:100%;min-width:0}:host .custom-textarea::-webkit-scrollbar{width:6px}:host .custom-textarea::-webkit-scrollbar-track{background:#fff}:host .custom-textarea::-webkit-scrollbar-thumb{background:#888;border-radius:3px}:host .custom-textarea::-webkit-scrollbar-thumb:hover{background:#555}:host .list-height{max-height:390px}:host .custom-textarea ul li{background:#f7f7f7}"]
                }] }
    ];
    /** @nocollapse */
    EmailTemplateEditComponent.ctorParameters = function () { return [
        { type: FormBuilder },
        { type: NotificationService },
        { type: TranslateService },
        { type: EmailTemplateService }
    ]; };
    EmailTemplateEditComponent.propDecorators = {
        visibleProperty: [{ type: Input, args: ['visibleProperty',] }],
        emailId: [{ type: Input }],
        onValidate: [{ type: Output }],
        onUpdate: [{ type: Output }],
        onToggleProperty: [{ type: Output }]
    };
    return EmailTemplateEditComponent;
}());
export { EmailTemplateEditComponent };
if (false) {
    /** @type {?} */
    EmailTemplateEditComponent.prototype.email;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.emailTemplateForm;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.visibleProperty;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.emailId;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.emailParams;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.editorConfig;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.onValidate;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.onUpdate;
    /** @type {?} */
    EmailTemplateEditComponent.prototype.onToggleProperty;
    /**
     * @type {?}
     * @private
     */
    EmailTemplateEditComponent.prototype.formChanged;
    /**
     * @type {?}
     * @private
     */
    EmailTemplateEditComponent.prototype.translateSubscription;
    /**
     * @type {?}
     * @private
     */
    EmailTemplateEditComponent.prototype.columnsTranslations;
    /**
     * @type {?}
     * @private
     */
    EmailTemplateEditComponent.prototype.fb;
    /**
     * @type {?}
     * @private
     */
    EmailTemplateEditComponent.prototype.notification;
    /**
     * @type {?}
     * @private
     */
    EmailTemplateEditComponent.prototype.translate;
    /**
     * @type {?}
     * @private
     */
    EmailTemplateEditComponent.prototype.emailTemplateService;
}
