// inet-ui/esm5/src/web-token/web-token.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { WebTokenSigner } from "./lib/WebTokenSigner";
import { TokenSimulator } from "./lib/TokenSimulator";
import { EXT_HASH_ALG } from "./lib/Constants";
import { Observable } from "rxjs";
var WebTokenService = /** @class */ (function () {
    function WebTokenService() {
        this._signer = new WebTokenSigner();
    }
    /**
     * @return {?}
     */
    WebTokenService.prototype.initForDevMode = /**
     * @return {?}
     */
    function () {
        console.warn("Init WebTokenSigner for dev mode. Not used it in production mode");
        this._signer.appInstalled = true;
        if (!window["_ExtTokenBridge"]) {
            window["_ExtTokenBridge"] = TokenSimulator;
        }
    };
    /**
     * @return {?}
     */
    WebTokenService.prototype.ready = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            _this._signer.checkReady().then((/**
             * @param {?} supported
             * @return {?}
             */
            function (supported) { return subscriber.next(supported); }));
        }));
    };
    /**
     * @param {?=} forceSelect
     * @return {?}
     */
    WebTokenService.prototype.selectCert = /**
     * @param {?=} forceSelect
     * @return {?}
     */
    function (forceSelect) {
        var _this = this;
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            _this._signer.selectCert(forceSelect).then((/**
             * @param {?} cert
             * @return {?}
             */
            function (cert) { return subscriber.next(cert); }));
        }));
    };
    /**
     * @param {?} serial
     * @return {?}
     */
    WebTokenService.prototype.getCertBySerial = /**
     * @param {?} serial
     * @return {?}
     */
    function (serial) {
        var _this = this;
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            _this.loadCerts().subscribe((/**
             * @param {?} certs
             * @return {?}
             */
            function (certs) {
                for (var i = 0; i < certs.length; i++) {
                    if (certs[i].serial === serial) {
                        return subscriber.next(certs[i]);
                    }
                }
                subscriber.next(null);
            }));
        }));
    };
    /**
     * @param {?=} serial
     * @return {?}
     */
    WebTokenService.prototype.selectCertUnlock = /**
     * @param {?=} serial
     * @return {?}
     */
    function (serial) {
        var _this = this;
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            if (serial) {
                _this.getCertBySerial(serial).subscribe((/**
                 * @param {?} cert
                 * @return {?}
                 */
                function (cert) { return _this._unlockCert(cert, subscriber); }));
            }
            else {
                _this.selectCert().subscribe((/**
                 * @param {?} cert
                 * @return {?}
                 */
                function (cert) { return _this._unlockCert(cert, subscriber); }));
            }
        }));
    };
    /**
     * @private
     * @param {?} cert
     * @param {?} subscriber
     * @return {?}
     */
    WebTokenService.prototype._unlockCert = /**
     * @private
     * @param {?} cert
     * @param {?} subscriber
     * @return {?}
     */
    function (cert, subscriber) {
        if (cert) {
            this.unlockCert(cert).subscribe((/**
             * @param {?} unlocked
             * @return {?}
             */
            function (unlocked) {
                cert.unlocked = unlocked;
                subscriber.next(cert);
            }));
        }
        else {
            subscriber.next(cert);
        }
    };
    /**
     * @param {?} cert
     * @return {?}
     */
    WebTokenService.prototype.unlockCert = /**
     * @param {?} cert
     * @return {?}
     */
    function (cert) {
        var _this = this;
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            // Sign fake data to unlock token
            /** @type {?} */
            var fakeData = {
                hash: "cA/metLrX7NUOfviDVsGYeRxAIo=",
                hashAlg: EXT_HASH_ALG.SHA1,
                serial: cert.serial
            };
            _this.signHash(fakeData).subscribe((/**
             * @param {?} response
             * @return {?}
             */
            function (response) {
                if (response && response.data && response.data.signature) {
                    subscriber.next(true);
                }
                else {
                    subscriber.next(false);
                }
            }));
        }));
    };
    /**
     * @return {?}
     */
    WebTokenService.prototype.loadCerts = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            _this._signer.loadCerts().then((/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return subscriber.next(data); }));
        }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    WebTokenService.prototype.signHash = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        return Observable.create((/**
         * @param {?} subscriber
         * @return {?}
         */
        function (subscriber) {
            _this._signer.signHash(data).then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) { return subscriber.next(response); }));
        }));
    };
    WebTokenService.decorators = [
        { type: Injectable }
    ];
    return WebTokenService;
}());
export { WebTokenService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    WebTokenService.prototype._signer;
}
