// inet-ui/esm5/src/layout/bootstrap/app-side-nav/app-side-nav.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef } from '@angular/core';
import { AbstractSideNavComponent } from "../abstract-side-nav.component";
import { Router } from "@angular/router";
import { CoreService } from "inet-core";
var AppSideNavComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AppSideNavComponent, _super);
    function AppSideNavComponent(coreService, router, elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.coreService = coreService;
        _this.router = router;
        _this.elementRef = elementRef;
        _this.applications = [];
        _this.widgetApps = [];
        _this.systemUrl = 'about:blank';
        _this.SYSTEM_APPS = ['social', 'calendar'];
        _this.listApp();
        return _this;
    }
    /**
     * @return {?}
     */
    AppSideNavComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        window.addEventListener('message', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            /** @type {?} */
            var __data = e.data || {};
            /** @type {?} */
            var __eventName = __data['eventName'];
            /** @type {?} */
            var __result = {};
            if (!iNet.isEmpty(__data.result)) {
                if (typeof (__data.result) === 'string') {
                    __result = JSON.parse(__data.result);
                }
                else if (typeof (__data.result) === 'object') {
                    __result = __data.result || {};
                }
                //console.log('[message]', __result);
                switch (__eventName) {
                    case 'open':
                    case 'openApp':
                        _this.openApp(__result['application']);
                        _this.close();
                        break;
                    case 'openUrl':
                        window.location.href = iNet.getSSOUrl(__result['url']);
                        break;
                    case 'loaded':
                        /** @type {?} */
                        var iFrame = document.querySelectorAll("iframe[src=\"" + __result['src'] + "\"]")[0];
                        if (iFrame) {
                            iFrame['height'] = __result['height'];
                            iFrame['valid'] = true;
                        }
                        break;
                }
            }
        }), false);
    };
    /**
     * @return {?}
     */
    AppSideNavComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.close();
        if (this._router) {
            this._router.unsubscribe();
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    AppSideNavComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        /** @type {?} */
        var currentValue = changes['opened']["currentValue"];
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (currentValue) {
                _this.systemUrl = iNet.getPUrl('common/page/apps') + '?appname=system';
                _this.open();
            }
            else {
                _this.close();
            }
        }), 250);
    };
    /**
     * @private
     * @return {?}
     */
    AppSideNavComponent.prototype.listApp = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.coreService.getSystemApplication().subscribe((/**
         * @param {?} apps
         * @return {?}
         */
        function (apps) {
            _this.applications = apps
                .filter((/**
             * @param {?} app
             * @return {?}
             */
            function (app) { return (app.organId !== 'community' && app.firmContext); }))
                .map((/**
             * @param {?} application
             * @return {?}
             */
            function (application) {
                if (application.homepage.indexOf('http') < 0) {
                    application.homepage = iNet.getPUrl(application.module + "/" + application.homepage);
                    application.icon = _this.coreService.getFileServerPath() + "/images/" + application.module + "/" + application.icon;
                }
                if (!!application.widget) {
                    application.widget = iNet.getPUrl('common/page/apps') + "?appname=" + application.firmContext; //rebuild the widget url
                }
                return application;
            }));
            _this.widgetApps = _this.applications.filter((/**
             * @param {?} app
             * @return {?}
             */
            function (app) { return (!!app.widget); }));
        }));
    };
    /**
     * @param {?} app
     * @return {?}
     */
    AppSideNavComponent.prototype.openApp = /**
     * @param {?} app
     * @return {?}
     */
    function (app) {
        if (this.SYSTEM_APPS.includes(app)) { //system app
            this.router.navigate([app]);
        }
        else {
            /** @type {?} */
            var url = iNet.getSSOUrl(this.getUrlByContext(app));
            if (url) {
                window.location.href = url;
            }
        }
    };
    /**
     * @private
     * @param {?} content
     * @return {?}
     */
    AppSideNavComponent.prototype.getUrlByContext = /**
     * @private
     * @param {?} content
     * @return {?}
     */
    function (content) {
        /** @type {?} */
        var app = this.applications.find((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var webContext = _a.webContext;
            return webContext === content;
        }));
        return app ? app.homepage : null;
    };
    AppSideNavComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-admin-app-side-nav',
                    template: "<div class=\"slide-nav\" [ngClass]=\"{'opened': opened}\">\n    <ul class=\"list-box list-app\" *ngIf=\"opened\">\n        <iframe id=\"system-app-iframe\" [src]=\"systemUrl | safe\" scrolling=\"no\"></iframe>\n        <iframe *ngFor=\"let widgetApp of widgetApps\" [src]=\"widgetApp.widget | safe\" scrolling=\"no\"></iframe>\n    </ul>\n</div>\n",
                    styles: ["\n        iframe {\n            width: 100%;\n            border: none;\n            overflow: hidden;\n        }\n    "]
                }] }
    ];
    /** @nocollapse */
    AppSideNavComponent.ctorParameters = function () { return [
        { type: CoreService },
        { type: Router },
        { type: ElementRef }
    ]; };
    return AppSideNavComponent;
}(AbstractSideNavComponent));
export { AppSideNavComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AppSideNavComponent.prototype._router;
    /** @type {?} */
    AppSideNavComponent.prototype.applications;
    /** @type {?} */
    AppSideNavComponent.prototype.widgetApps;
    /** @type {?} */
    AppSideNavComponent.prototype.systemUrl;
    /** @type {?} */
    AppSideNavComponent.prototype.SYSTEM_APPS;
    /**
     * @type {?}
     * @private
     */
    AppSideNavComponent.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    AppSideNavComponent.prototype.router;
    /** @type {?} */
    AppSideNavComponent.prototype.elementRef;
}
