// inet-ui/esm2015/src/organization/organization.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CoreService, HttpClientService } from 'inet-core';
import { Organization } from '../model/organization';
import { shareReplay } from "rxjs/operators";
export class OrganizationService {
    /**
     * @param {?} http
     * @param {?} coreService
     */
    constructor(http, coreService) {
        this.http = http;
        this.coreService = coreService;
        this.url = {
            firm_load: iNet.getPUrl('cloud/firmorganid/load'),
            firm_update: iNet.getPUrl('cloud/firmprofile/update'),
            update_logo: iNet.getPUrl('cloud/firmprofile/logoupdate'),
            view_logo: iNet.getPUrl('plugin/firmlogo/view'),
            // For Essentials
            //load_info: iNet.getPUrl('gl/organ/profile/load'),
            load_info: iNet.getPUrl('cloud/firmprofile/load'),
            update_info: iNet.getPUrl('cloud/firmprofile/update'),
            search_org: iNet.getPUrl('plugin/organization/search'),
            view_org: iNet.getPUrl('plugin/firmprofile/view')
            //update_logo: iNet.getPUrl('gl/organ/profile/logo')
        };
    }
    /**
     * @return {?}
     */
    firmLoad() {
        this.http.showLoading();
        return this.http.getJSON(this.url.firm_load, { organId: iNet.organId })
            .map((/**
         * @param {?} res
         * @return {?}
         */
        (res) => res));
    }
    // saveLogoOrganization(params: any): Observable<any> {
    //     return this.http.postJSON(this.url.save_logo, params);
    // }
    /**
     * @param {?} orgInfo
     * @return {?}
     */
    firmUpdate(orgInfo) {
        this.http.showLoading();
        return this.http.postJSON(this.url.firm_update, orgInfo);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    updateLogo(data) {
        this.http.showLoading();
        return this.http.post(this.url.update_logo, data);
    }
    /**
     * @param {?} params
     * @return {?}
     */
    removeLogo(params) {
        this.http.showLoading();
        return this.http.postJSON(this.url.update_logo, params);
    }
    /**
     * @return {?}
     */
    load() {
        if (!this.$orgCache) {
            this.http.showLoading();
            this.$orgCache = this.http.getJSON(this.url.view_org)
                .pipe(shareReplay(1));
        }
        return this.$orgCache;
    }
    /**
     * @return {?}
     */
    loadInfo() {
        this.http.showLoading();
        return this.http.postJSON(this.url.load_info);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    update(data) {
        this.http.showLoading();
        // Clone data for update
        /** @type {?} */
        const org = Object.assign(new Organization(), data);
        org.officeAddressStr = JSON.stringify(org.officeAddress);
        delete org.officeAddress;
        delete org.createdDate;
        delete org.createdBy;
        delete org.banks;
        delete org.busAces;
        this.$orgCache = null;
        return this.http.postJSON(this.url.update_info, org);
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    searchOrganization(params = {}) {
        return this.http.postJSON(this.url.search_org, params);
    }
    /**
     * @param {?} orgInfo
     * @return {?}
     */
    getLogoUrlByOrganization(orgInfo) {
        /*
        if (!orgInfo.logo) {
            return iNet.BLANK_IMAGE_URL;
        }
         */
        return `${this.url.view_logo}?version=${orgInfo.modifiedDate}`;
        //return this.coreService.getFileUrl(orgInfo.logo + `?version=${orgInfo.modifiedDate}`);
    }
    /**
     * @return {?}
     */
    getLogoUrl() {
        return this.url.view_logo;
    }
    /**
     * @return {?}
     */
    viewInfo() {
        this.http.showLoading();
        return this.http.postJSON(this.url.view_org);
    }
}
OrganizationService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
OrganizationService.ctorParameters = () => [
    { type: HttpClientService },
    { type: CoreService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    OrganizationService.prototype.$orgCache;
    /**
     * @type {?}
     * @private
     */
    OrganizationService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    OrganizationService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    OrganizationService.prototype.coreService;
}
