// inet-ui/esm2015/src/report-template/report-template-edit/report-template-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { Location } from "@angular/common";
import { ActivatedRoute } from "@angular/router";
import { ErrorMessage, NotificationService } from "inet-core";
import { FormBuilder, Validators } from "@angular/forms";
import { TranslateService } from "@ngx-translate/core";
import { ReportTemplateService } from "../report-template.service";
export class ReportTemplateEditComponent {
    /**
     * @param {?} location
     * @param {?} fb
     * @param {?} notification
     * @param {?} service
     * @param {?} translate
     * @param {?} route
     */
    constructor(location, fb, notification, service, translate, route) {
        this.location = location;
        this.fb = fb;
        this.notification = notification;
        this.service = service;
        this.translate = translate;
        this.route = route;
        this.validForm = new EventEmitter();
        this.applicationList = [];
        this.typeList = [];
        this.moduleList = [];
        this.versionList = [{ name: 1, description: 'Phiên bản 1' }, { name: 2, description: 'Phiên bản 2' }];
        this.attachments = [];
        this.translateSubscription = this.translate.get(['COMMON.MODULE.REPORT_TEMPLATE']).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            this.formTranslations = res['COMMON.MODULE.REPORT_TEMPLATE'];
        }));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.subRouter = this.route.queryParamMap.subscribe((/**
         * @param {?} params
         * @return {?}
         */
        params => {
            this.applicationParams = params.get('application');
            this.moduleParams = params.get('module');
            this.typeParams = params.get('type');
        }));
        this.initForm();
        this.initData();
        this.loadApplication();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.translateSubscription) {
            this.translateSubscription.unsubscribe();
        }
        if (this.subRouter) {
            this.subRouter.unsubscribe();
        }
    }
    /**
     * @param {?} templateId
     * @return {?}
     */
    loadById(templateId) {
        if (templateId) {
            this.service.view(templateId).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data.type !== ErrorMessage.TYPE) {
                    this.templateId = templateId;
                    this.applicationParams = data.application;
                    this.templateFormGroup.patchValue(data);
                    this.fileName = data.name;
                    this.validForm.emit(true);
                }
            }));
        }
    }
    /**
     * @private
     * @return {?}
     */
    initData() {
        this.version.setValue(2);
        this.application.setValue(this.applicationParams);
        this.module.setValue(this.moduleParams);
        this.type.setValue(this.typeParams);
    }
    /**
     * @return {?}
     */
    reset() {
        this.templateFormGroup.reset();
        this.templateId = '';
        this.fileName = '';
        this.initData();
        this.loadApplication();
        this.changeFormValid();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeFile($event) {
        /** @type {?} */
        const files = $event.target.files;
        if (files.length > 0) {
            /** @type {?} */
            const upload = files[0];
            if (upload) {
                this.name.setValue(upload['name']);
                this.fileName = upload['name'];
            }
        }
        else {
        }
        this.changeFormValid();
    }
    /**
     * @return {?}
     */
    clearForm() {
        this.fileName = '';
        if (this.fileUpload.value) {
            this.fileUpload.setValue('');
        }
        this.changeFormValid();
    }
    /**
     * @private
     * @return {?}
     */
    loadApplication() {
        if (!this.appName) {
            this.service.application().subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data.type !== ErrorMessage.TYPE) {
                    this.applicationList = data.items || [];
                    this.application.setValue(this.applicationParams || this.application.value);
                    this.loadModules(this.applicationParams || this.application.value);
                }
            }));
        }
        else {
            this.applicationParams = this.appName;
            /** @type {?} */
            const app = { id: this.appName, description: this.appDesc || this.appName, name: this.appName };
            this.applicationList = [app];
            this.application.setValue(this.appName);
            this.templateFormGroup.controls['application'].disable();
            this.loadModules(this.appName);
        }
        this.service.organId().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            this.organId.setValue(data.organId);
        }));
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeApps($event) {
        this.loadModules($event.name);
    }
    /**
     * @private
     * @param {?} $event
     * @return {?}
     */
    loadModules($event) {
        /** @type {?} */
        const __modulesList = [];
        this.service.modules($event).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            Object.keys(data).forEach((/**
             * @param {?} key
             * @return {?}
             */
            key => {
                /** @type {?} */
                let item = Object.assign({ type: key }, data[key]);
                __modulesList.push(item);
            }));
            this.moduleList = __modulesList;
            if (this.moduleList.length > 0) {
                this.module.setValue(this.module.value || this.moduleList[0].type);
                this.changeModules($event);
            }
            else {
                this.module.setValue('');
                this.description.setValue('');
                this.clearForm();
            }
        }));
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeModules($event) {
        /** @type {?} */
        let __typesList = [];
        /** @type {?} */
        const types = this.moduleList.find((/**
         * @param {?} ele
         * @return {?}
         */
        ele => {
            return ele.type === this.module.value;
        }));
        Object.keys(types).forEach((/**
         * @param {?} ele
         * @return {?}
         */
        ele => {
            __typesList = types[ele];
        }));
        this.typeList = __typesList;
        if (this.typeList.length > 0) {
            this.type.setValue(this.type.value || this.typeList[0].type);
        }
    }
    /**
     * @return {?}
     */
    save() {
        /** @type {?} */
        const formData = new FormData(this.templateFormElementRef.nativeElement);
        formData.append('application', this.application.value);
        formData.append('module', this.module.value);
        formData.append('type', this.type.value);
        formData.append('version', this.version.value);
        formData.append('organId', this.organId.value);
        if (!iNet.isEmpty(this.templateId)) {
            formData.append('uuid', this.uuid.value);
            this.service.update(formData).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data.type !== ErrorMessage.TYPE) {
                    this.templateFormGroup.patchValue(data);
                    this.templateId = data.uuid;
                    this.fileName = data.name;
                    this.validForm.emit(true);
                    this.notification.showMessage(this.formTranslations['UPDATE_SUCCESS'], 'success', this.formTranslations['TITLE']);
                }
                else {
                    this.notification.showMessage(this.formTranslations['UPDATE_ERROR'], 'error', this.formTranslations['TITLE']);
                }
            }));
        }
        else {
            this.service.add(formData).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data.type !== ErrorMessage.TYPE) {
                    this.templateFormGroup.patchValue(data);
                    this.templateId = data.uuid;
                    this.fileName = data.name;
                    this.validForm.emit(true);
                    this.notification.showMessage(this.formTranslations['ADD_SUCCESS'], 'success', this.formTranslations['TITLE']);
                }
                else {
                    this.notification.showMessage(this.formTranslations['ADD_ERROR'], 'error', this.formTranslations['TITLE']);
                }
            }));
        }
    }
    /**
     * @return {?}
     */
    initForm() {
        this.templateFormGroup = this.fb.group({
            uuid: [],
            application: ['', Validators.required],
            module: ['', Validators.required],
            type: ['', Validators.required],
            name: ['', Validators.required],
            description: ['', Validators.required],
            version: ['', Validators.required],
            organId: [],
            fileUpload: [],
        });
    }
    /**
     * @return {?}
     */
    get uuid() {
        return this.templateFormGroup.get('uuid');
    }
    /**
     * @return {?}
     */
    get application() {
        return this.templateFormGroup.get('application');
    }
    /**
     * @return {?}
     */
    get module() {
        return this.templateFormGroup.get('module');
    }
    /**
     * @return {?}
     */
    get type() {
        return this.templateFormGroup.get('type');
    }
    /**
     * @return {?}
     */
    get name() {
        return this.templateFormGroup.get('name');
    }
    /**
     * @return {?}
     */
    get description() {
        return this.templateFormGroup.get('description');
    }
    /**
     * @return {?}
     */
    get version() {
        return this.templateFormGroup.get('version');
    }
    /**
     * @return {?}
     */
    get organId() {
        return this.templateFormGroup.get('organId');
    }
    /**
     * @return {?}
     */
    get fileUpload() {
        return this.templateFormGroup.get('fileUpload');
    }
    /**
     * @return {?}
     */
    changeFormValid() {
        /** @type {?} */
        let __valid;
        this.fileName === '' ? __valid = true : __valid = false;
        __valid ? this.validForm.emit(false) : this.validForm.emit(this.templateFormGroup.valid);
    }
}
ReportTemplateEditComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-report-template-edit',
                template: "<div id=\"report-template-content\" class=\"row center-body m-0\">\n    <div class=\"row justify-content-center mt-1 mb-0 reponsive-card \" style=\"width: 65%\">\n        <div class=\"card shadow box col-sm-12 mt-4\">\n            <div class=\"row ml-0 mr-0 \">\n                <form #templateForm [formGroup]=\"templateFormGroup\" class=\"w-100\">\n                    <div class=\"form-group row m-0 col-sm-12 mb-3 pl-0 mt-4\">\n                        <div class=\"col-sm-6 row m-0 p-0\">\n                            <label class=\"col-sm-4 text-left p-0 col-form-label \"> {{ 'COMMON.MODULE.REPORT_TEMPLATE.APPLICATION' | translate }} <span\n                                    class=\"required\"> :</span></label>\n                            <div class=\"col-sm-8 p-0\">\n                                <ng-select #application\n                                           class=\"col -sm-12 pl-0 pr-0\"\n                                           required\n                                           [clearable]=\"false\"\n                                           id=\"application\"\n                                           bindLabel=\"description\"\n                                           bindValue=\"name\"\n                                           [items]=\"applicationList\"\n                                           formControlName=\"application\"\n                                           name=\"application\"\n                                           (changeEvent)=\"changeFormValid()\"\n                                           (change)=\"changeApps($event)\"\n                                >\n                                </ng-select>\n                                <!-- validate -->\n                                <div *ngIf=\"application['invalid'] && (application['dirty'] || application['touched'])\"\n                                     class=\"text-danger mt-2 text-df\">\n                                    <div *ngIf=\"application['errors']['required']\">\n                                        <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                        {{ 'COMMON.MODULE.REPORT_TEMPLATE.APPLICATION_REQUIRED' | translate }}\n                                    </div>\n                                </div>\n                            </div>\n                        </div>\n                        <div class=\"col-sm-6 row m-0 p-0\">\n                            <label class=\"col-sm-4 text-left  col-form-label \"> {{ 'COMMON.MODULE.REPORT_TEMPLATE.MODULE' | translate }} <span\n                                    class=\"required\"> :</span></label>\n                            <div class=\"col-sm-8 p-0\">\n                                <ng-select\n                                        class=\"col-sm-12 pl-0 pr-0\"\n                                        required\n                                        [clearable]=\"false\"\n                                        id=\"module\"\n                                        [items]=\"moduleList\"\n                                        formControlName=\"module\"\n                                        bindValue=\"type\"\n                                        bindLabel=\"name\"\n                                        name=\"module\"\n                                        (change)=\"changeModules($event)\"\n                                        (changeEvent)=\"changeFormValid()\"\n                                >\n                                </ng-select>\n                                <!-- validate -->\n                                <div *ngIf=\"module.invalid && (module.dirty || module.touched)\"\n                                     class=\"text-danger mt-2 text-df\">\n                                    <div *ngIf=\"module.errors.required\">\n                                        <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                        {{ 'COMMON.MODULE.REPORT_TEMPLATE.MODULE_REQUIRED' | translate }}\n                                    </div>\n                                </div>\n                            </div>\n                        </div>\n\n                    </div>\n                    <div class=\"form-group row m-0 col-sm-12 mb-3 pl-0\">\n                        <div class=\"col-sm-6 row m-0 p-0\">\n                            <label class=\"col-sm-4 text-left p-0 col-form-label \"> {{ 'COMMON.MODULE.REPORT_TEMPLATE.TYPE' | translate }} <span\n                                    class=\"required\"> :</span></label>\n                            <div class=\"col-sm-8 p-0\">\n                                <ng-select\n                                        class=\"col-sm-12 pl-0 pr-0\"\n                                        required\n                                        [clearable]=\"false\"\n                                        id=\"type\"\n                                        [items]=\"typeList\"\n                                        formControlName=\"type\"\n                                        name=\"type\"\n                                        bindValue=\"type\"\n                                        bindLabel=\"name\"\n                                        (changeEvent)=\"changeFormValid()\">\n                                    <ng-template ng-label-tmp let-item=\"item\">\n                                        {{item['name'] | translate}}\n                                    </ng-template>\n                                    <ng-template ng-option-tmp let-item=\"item\">\n                                        {{item['name'] | translate}}\n                                    </ng-template>\n                                </ng-select>\n                                <!-- validate -->\n                                <div *ngIf=\"type.invalid && (type.dirty || type.touched)\"\n                                     class=\"text-danger mt-2 text-df\">\n                                    <div *ngIf=\"type.errors.required\">\n                                        <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                        {{ 'COMMON.MODULE.REPORT_TEMPLATE.TYPE_REQUIRED' | translate }}\n                                    </div>\n                                </div>\n                            </div>\n                        </div>\n\n                        <div class=\"col-sm-6 row m-0 p-0\">\n                            <label class=\"col-sm-4 text-left  col-form-label \"> {{ 'COMMON.MODULE.REPORT_TEMPLATE.NAME' | translate }} <span\n                                    class=\"required\"> :</span></label>\n                            <div class=\"col-sm-8 p-0\">\n                                <input type=\"text\"\n                                       class=\"form-control form-control-sm \"\n                                       formControlName=\"name\" readonly\n                                       [ngClass]=\"{'is-invalid': name.errors && name.touched}\"\n                                        (keyup)=\"changeFormValid()\"\n                                       id=\"name\" name=\"name\" value=\"\">\n                                <!-- validate -->\n                                <div *ngIf=\"name.invalid && (name.dirty || name.touched)\"\n                                     class=\"text-danger mt-2 text-df\">\n                                    <div *ngIf=\"name.errors.required\">\n                                        <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                        {{ 'COMMON.MODULE.REPORT_TEMPLATE.NAME_REQUIRED' | translate }}\n                                    </div>\n                                </div>\n                            </div>\n                        </div>\n                    </div>\n                    <div class=\"form-group row m-0 col-sm-12 mb-3 pl-0\">\n                        <div class=\"col-sm-6 row m-0 p-0\">\n                            <label class=\"col-sm-4 text-left p-0 col-form-label \"> {{ 'COMMON.MODULE.REPORT_TEMPLATE.VERSION' | translate }} <span\n                                    class=\"required\"> :</span></label>\n                            <div class=\"col-sm-8 p-0\">\n                                <ng-select\n                                        class=\"col -sm-12 pl-0 pr-0\"\n                                        required\n                                        [clearable]=\"false\"\n                                        id=\"version\"\n                                        bindLabel=\"description\"\n                                        bindValue=\"name\"\n                                        [items]=\"versionList\"\n                                        formControlName=\"version\"\n                                        name=\"version\"\n                                        (changeEvent)=\"changeFormValid()\"\n                                >\n                                </ng-select>\n                                <!-- validate -->\n                                <div *ngIf=\"version.invalid && (version.dirty || version.touched)\"\n                                     class=\"text-danger mt-2 text-df\">\n                                    <div *ngIf=\"version.errors.required\">\n                                        <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                        {{ 'COMMON.MODULE.REPORT_TEMPLATE.VERSION_REQUIRED' | translate }}\n                                    </div>\n                                </div>\n                            </div>\n                        </div>\n                    </div>\n\n                    <div class=\"col-sm-12 form-group row m-0 mb-3 pl-0\">\n                        <label for=\"description\"\n                               class=\"col-sm-2 text-left p-0 col-form-label \"> {{ 'COMMON.MODULE.REPORT_TEMPLATE.DESCRIPTION' | translate }} <span\n                                class=\"required\"> :</span></label>\n                        <div class=\"col-sm-10 p-0\">\n                  <textarea #descr type=\"text\" (keyup)=\"changeFormValid()\" style=\"resize: none\" rows=\"3\"\n                            class=\"form-control form-control-sm \"\n                            formControlName=\"description\"\n                            [ngClass]=\"{'is-invalid': description.errors && description.touched}\"\n                            id=\"description\" name=\"description\" value=\"\" required></textarea>\n                            <!-- validate -->\n                            <div *ngIf=\"description.invalid && (description.dirty || description.touched)\"\n                                 class=\"text-danger mt-2 text-df\">\n                                <div *ngIf=\"description.errors.required\">\n                                    <i class=\"fa fa-exclamation-triangle mr-1\"></i>\n                                    {{ 'COMMON.MODULE.REPORT_TEMPLATE.DESCRIPTION_REQUIRED' | translate }}\n                                </div>\n                            </div>\n                        </div>\n                    </div>\n\n                    <div class=\"col-sm-12 form-group row m-0 mb-3 pl-0\">\n                        <label class=\"control-label col-form-label col-sm-2 pl-0 \">\n                            {{ 'COMMON.MODULE.REPORT_TEMPLATE.FILE_UPLOAD' | translate }} <span\n                                class=\"required\"> :</span>\n                        </label>\n                        <div class=\"col-sm-10 p-0\">\n                            <label class=\"i-file-input\">\n                                <input #file type=\"file\" (change)=\"changeFile($event);changeFormValid()\" formControlName=\"fileUpload\"\n                                       name=\"fileUpload\"\n                                       class=\"form-control form-control-sm col-xs-12 col-sm-12\">\n                                <span class=\"i-file-container\" [ngClass]=\"{'selected': fileName}\"\n                                      [attr.data-title]=\"'COMMON.MODULE.REPORT_TEMPLATE.CHOOSE' | translate\">\n                                                            <span class=\"i-file-name\"\n                                                                  [attr.data-title]=\"fileName || ('COMMON.MODULE.REPORT_TEMPLATE.NO_FILE' | translate)\">\n                                                                <i class=\"file-icon fa fa-upload\"></i>\n                                                            </span>\n                                                        </span>\n                                <a class=\"remove\" *ngIf=\"fileName\" href=\"javascript:;\" (click)=\"clearForm()\">\n\n                                    <i class=\"fa fa-times\"></i>\n                                </a>\n                            </label>\n                        </div>\n                    </div>\n                </form>\n            </div>\n        </div>\n    </div>\n</div>\n\n",
                styles: [":host #report-template-content{height:calc(100vh - 115px)}@media (max-width:700px){:host .reponsive-card{padding:0 30px}}:host .pointer{pointer-events:none}:host .custom-center-span{display:flex;align-items:center}:host .center-body{flex-direction:row;justify-content:center;align-content:start}:host .shadow{box-shadow:0 .5rem 1rem rgba(0,0,0,.15)!important}"]
            }] }
];
/** @nocollapse */
ReportTemplateEditComponent.ctorParameters = () => [
    { type: Location },
    { type: FormBuilder },
    { type: NotificationService },
    { type: ReportTemplateService },
    { type: TranslateService },
    { type: ActivatedRoute }
];
ReportTemplateEditComponent.propDecorators = {
    templateId: [{ type: Input }],
    validForm: [{ type: Output }],
    appName: [{ type: Input }],
    appDesc: [{ type: Input }],
    templateFormElementRef: [{ type: ViewChild, args: ['templateForm',] }]
};
if (false) {
    /** @type {?} */
    ReportTemplateEditComponent.prototype.templateId;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.validForm;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.appName;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.appDesc;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.applicationList;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.typeList;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.moduleList;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.versionList;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.templateFormGroup;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.attachments;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.fileName;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.applicationParams;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.moduleParams;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.typeParams;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.subRouter;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.translateSubscription;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.formTranslations;
    /** @type {?} */
    ReportTemplateEditComponent.prototype.templateFormElementRef;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.location;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.fb;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.notification;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.service;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.translate;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateEditComponent.prototype.route;
}
