// inet-calendar/esm2015/src/dialog-view/dialog-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Output, ViewChild } from '@angular/core';
import { Router } from '@angular/router';
import { CalendarDialogDelete } from "../dialog-delete/dialog-delete.component";
import { CalendarEvent } from "../model/CalendarEvent";
import { CalendarUtils } from "../CalendarUtils";
export class CalendarDialogViewComponent {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
        this.onChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this._initModalView();
        this._addScrollUpdatePos();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._destroyModalView();
        this._removeScrollUpdatePos();
    }
    /**
     * @return {?}
     */
    editEvent() {
        this.hide();
        this.router.navigate(['/calendar/event', this.calendarEvent.event.uuid || '']);
    }
    /**
     * @return {?}
     */
    changeEvent() {
        this.hide();
        this.onChange.emit();
        this._onChange();
    }
    /**
     * @param {?} event
     * @param {?} jsEvent
     * @param {?} target
     * @param {?} onChange
     * @return {?}
     */
    viewEvent(event, jsEvent, target, onChange) {
        this._onChange = onChange;
        this.calendarEvent = new CalendarEvent(event);
        this.calEvent = event;
        this._targetEl = target;
        this._jsEvent = jsEvent;
        this._syncBodyHeight();
        this._preventScrolling();
        jsEvent.stopPropagation();
        setTimeout((/**
         * @return {?}
         */
        () => this._openModal(jsEvent, target)), 100);
    }
    /**
     * @return {?}
     */
    hide() {
        this._showModal = false;
        this._$modalView.hide();
        this._$modalView.css('opacity', 0);
    }
    /**
     * @return {?}
     */
    show() {
        this._showModal = true;
        this._$modalView.show();
    }
    /**
     * @return {?}
     */
    removeEvent() {
        this.viewDeleteModal.show();
    }
    /**
     * @private
     * @return {?}
     */
    _preventScrolling() {
        $(document.scrollingElement).css('overflow', 'hidden');
        clearTimeout(this._scrollTimer);
        this._scrollTimer = setTimeout((/**
         * @return {?}
         */
        () => {
            $(document.scrollingElement).css('overflow', '');
        }), 600);
    }
    /**
     * @private
     * @return {?}
     */
    _syncBodyHeight() {
        this.maxHeight = innerHeight - 150;
    }
    /**
     * @private
     * @return {?}
     */
    _initModalView() {
        this._$modalView = $(this.modalView.nativeElement).hide();
        // Click outside to close
        this.hide = this.hide.bind(this);
        document.body.addEventListener('click', this.hide);
        // Prevent close
        this._$modalView.on('click', (/**
         * @param {?} e
         * @return {?}
         */
        (e) => e.stopPropagation()));
        // Append modal to body
        $(document.body).append(this._$modalView);
    }
    /**
     * @private
     * @return {?}
     */
    _destroyModalView() {
        document.body.removeEventListener('click', this.hide);
        this._$modalView.remove();
    }
    /**
     * @private
     * @return {?}
     */
    _addScrollUpdatePos() {
        this._scrollUpdatePos = this._scrollUpdatePos.bind(this);
        document.addEventListener('scroll', this._scrollUpdatePos, true);
    }
    /**
     * @private
     * @return {?}
     */
    _removeScrollUpdatePos() {
        document.removeEventListener('scroll', this._scrollUpdatePos);
    }
    /**
     * @private
     * @return {?}
     */
    _scrollUpdatePos() {
        // Delay update pos for performance
        if (this._showModal && this._jsEvent && this._targetEl) {
            clearTimeout(this._timer);
            this._timer = setTimeout((/**
             * @return {?}
             */
            () => this._openModal(this._jsEvent, this._targetEl)), 100);
        }
    }
    /**
     * @private
     * @param {?} jsEvent
     * @param {?} target
     * @return {?}
     */
    _openModal(jsEvent, target) {
        // Show to get real size
        this.show();
        /** @type {?} */
        let pos = CalendarUtils.computePosByTargetEl(this.modalView.nativeElement, target, jsEvent);
        /** @type {?} */
        let transformCss = 'translate3d(' + pos.left + 'px,' + pos.top + 'px,0)';
        // Use transform translate3d to boot performance
        this._$modalView.css({
            '-webkit-transition-duration': '300ms',
            'transition-duration': '300ms',
            'opacity': '1',
            '-webkit-transform': transformCss,
            'transform': transformCss,
        });
    }
}
CalendarDialogViewComponent.decorators = [
    { type: Component, args: [{
                template: "<div #modalView class=\"modal-dialog popover\" role=\"document\">\n    <div class=\"modal-content\" *ngIf=\"calEvent\">\n            <div class=\"header-dialog\" [ngClass]=\"{'overlay': titleOverlay}\">\n                <i class=\"fa fa-pencil\" *ngIf=\"calEvent.isCreator || !!calendarEvent.myAttendee\" (click)=\"editEvent()\" title=\"Ch\u1EC9nh s\u1EEDa\"></i>\n                <i class=\"fa fa-trash\" *ngIf=\"calEvent.isCreator\" (click)=\"removeEvent()\" title=\"X\u00F3a\"></i>\n                <i (click)=\"hide()\" class=\"fa fa-times\" title=\"\u0110\u00F3ng\"></i>\n            </div>\n        <div class=\"content-dialog\" [ngStyle]=\"{'max-height.px': maxHeight}\">\n                <div class=\"field-group\">\n                    <i class=\"field-icon fa fa-calendar\"></i>\n                    <div class=\"field-content\">\n                        <div style=\"font-size:18px\" [innerHTML]=\"calendarEvent.subject\"></div>\n                        <div>{{calendarEvent.timeDisplay}}</div>\n                        <div>{{calendarEvent.repeatText}}</div>\n                    </div>\n                </div>\n                <div class=\"field-group\" *ngIf=\"calEvent.memberStr\">\n                    <i class=\"field-icon fa fa-users\" title=\"Kh\u00E1ch m\u1EDDi tham d\u1EF1\"></i>\n                    <div class=\"field-content\">\n                        <calendar-attendee-list [attendees]=\"calEvent.attendees\" [expandable]=\"true\"></calendar-attendee-list>\n                    </div>\n                </div>\n                <div class=\"field-group\" *ngIf=\"calEvent.location\">\n                    <i class=\"field-icon fa fa-map-marker\" title=\"\u0110\u1ECBa \u0111i\u1EC3m\"></i>\n                    <div class=\"field-content\">{{ calEvent.location }}</div>\n                </div>\n                <div class=\"field-group\" *ngIf=\"calendarEvent.summary\">\n                    <i class=\"field-icon fa fa-align-left\" title=\"M\u00F4 t\u1EA3 s\u1EF1 ki\u1EC7n\"></i>\n                    <div class=\"field-content\" [innerHTML]=\"calendarEvent.summary\"></div>\n                </div>\n                <div class=\"field-group\" *ngIf=\"calEvent['fullname']\">\n                    <i class=\"field-icon fa fa-user-circle\" title=\"Ng\u01B0\u1EDDi t\u1EA1o\"></i>\n                    <div class=\"field-content\">{{ calEvent['fullname'] }}</div>\n                </div>\n                <div class=\"field-group\" *ngIf=\"calEvent?.attachments?.length\">\n                    <i class=\"field-icon fa fa-paperclip\" title=\"T\u1EADp tin \u0111\u00EDnh k\u00E8m\"></i>\n                    <div class=\"field-content\">\n                        <calendar-attachment-list [calEvent]=\"calEvent\"></calendar-attachment-list>\n                    </div>\n                </div>\n            </div>\n    </div>\n</div>\n<calendar-dialog-delete #dialogRemove hidden [calEvent]=\"calEvent\" (onDelete)=\"changeEvent()\"></calendar-dialog-delete>\n",
                styles: [".modal-dialog{min-width:450px;margin:0;z-index:1040!important}.modal-content{padding:0;border-radius:4px;border:0;box-shadow:1px 1px 7px 3px rgba(0,0,0,.14)}.header-dialog{padding:10px 25px;text-align:right}.header-dialog .overlay{border-bottom:1px solid #eee}.header-dialog>i{cursor:pointer;width:34px;height:34px;line-height:34px;text-align:center;font-size:18px;border-radius:50%;color:#555;margin-left:10px}.header-dialog>i:hover{background:rgba(0,0,0,.1)}.field-group{margin-bottom:20px;padding-left:14px;padding-right:5px}.field-icon{width:30px;float:left;color:#666;font-size:20px;text-align:center;margin-right:15px;margin-top:2px}.field-content{overflow:hidden;line-height:1.6}.file-attach-layout{float:left;max-width:144px;border:1px solid #fff;padding-right:5px}.file-attach-content{height:36px;line-height:36px;border-radius:2px;cursor:pointer;color:#2a6496;margin-bottom:7px}.file-attach-img{float:left;width:20px;height:20px;margin-top:6px;margin-right:10px}.file-attach-text{display:block;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}a{color:#428bca;text-decoration:none;background:0 0}a:not([href]):hover{text-decoration:underline}.content-dialog{overflow-y:auto}"]
            }] }
];
/** @nocollapse */
CalendarDialogViewComponent.ctorParameters = () => [
    { type: Router }
];
CalendarDialogViewComponent.propDecorators = {
    onChange: [{ type: Output }],
    modalView: [{ type: ViewChild, args: ['modalView',] }],
    viewDeleteModal: [{ type: ViewChild, args: ['dialogRemove',] }]
};
if (false) {
    /** @type {?} */
    CalendarDialogViewComponent.prototype.onChange;
    /** @type {?} */
    CalendarDialogViewComponent.prototype.modalView;
    /** @type {?} */
    CalendarDialogViewComponent.prototype.viewDeleteModal;
    /** @type {?} */
    CalendarDialogViewComponent.prototype.calendarEvent;
    /** @type {?} */
    CalendarDialogViewComponent.prototype.calEvent;
    /** @type {?} */
    CalendarDialogViewComponent.prototype.maxHeight;
    /** @type {?} */
    CalendarDialogViewComponent.prototype.titleOverlay;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype._showModal;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype._targetEl;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype._jsEvent;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype._timer;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype._scrollTimer;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype._$modalView;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype._onChange;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogViewComponent.prototype.router;
}
