// inet-ui/esm5/src/viewer/error-page/error-page.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { Location } from '@angular/common';
import { HttpClientService } from "inet-core";
import { ViewerService } from "../../file/viewer.service";
var ErrorPageComponent = /** @class */ (function () {
    function ErrorPageComponent(route, http, viewerService, location) {
        this.route = route;
        this.http = http;
        this.viewerService = viewerService;
        this.location = location;
    }
    /**
     * @return {?}
     */
    ErrorPageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.sub = this.route.params.subscribe((/**
         * @param {?} params
         * @return {?}
         */
        function (params) {
            _this.docId = params.id;
        }));
    };
    /**
     * @return {?}
     */
    ErrorPageComponent.prototype.download = /**
     * @return {?}
     */
    function () {
        this.viewerService.downloadById(this.docId);
    };
    /**
     * @return {?}
     */
    ErrorPageComponent.prototype.back = /**
     * @return {?}
     */
    function () {
        this.location.back();
    };
    ErrorPageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-error-page',
                    template: "<div class=\"error-page\">\n  <div class=\"error-content text-center\">\n    <div>\n      <div class=\"error-subtitle\" [innerHTML]=\"'COMMON.MODULE.VIEWER.UNSUPPORTED' | translate\"></div>\n      <button class=\"btn btn-primary\" (click)=\"download()\"> <i class=\"fa fa-download\"></i> {{'TOOLBAR.DOWNLOAD' | translate}}</button>\n      <button class=\"btn btn-danger ml-1\" (click)=\"back()\"> <i class=\"fa fa-remove\"></i> {{'TOOLBAR.CLOSE' | translate}}</button>\n    </div>\n  </div>\n</div>",
                    styles: [".error-page{position:relative}.error-content{padding:40px 1rem}.error-subtitle{font-size:20px;font-weight:300;padding-bottom:10px}"]
                }] }
    ];
    /** @nocollapse */
    ErrorPageComponent.ctorParameters = function () { return [
        { type: ActivatedRoute },
        { type: HttpClientService },
        { type: ViewerService },
        { type: Location }
    ]; };
    return ErrorPageComponent;
}());
export { ErrorPageComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.sub;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.docId;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.viewerService;
    /**
     * @type {?}
     * @private
     */
    ErrorPageComponent.prototype.location;
}
