// inet-calendar/esm2015/src/utils/CalBuilder.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { vnToLatin } from "./common/Accent";
import { DateUtils } from "./common/DateUtils";
import { AjaxAPI } from "./common/AjaxAPI";
import { CalAttendeeType } from "./model/CalAttendeeType";
import { CalConfigs } from "./CalConfigs";
import { CalType } from "./model/CalType";
import { CalCategory } from "./model/CalCategory";
import { CalCategoryIcon } from "./model/CalCategoryIcon";
import { CalAttendeeRole } from "./model/CalAttendeeRole";
import { CalSubType } from "./model/CalSubType";
import { CalMode } from "./model/CalMode";
import { CalendarCar } from "./CalendarCar";
import { SubFirmDictionary } from "./subfirm/Dictionary";
import { CalAttendeeState } from "./model/CalAttendeeState";
import { LEADER_REFERENCE } from "./CalConstants";
import * as moment_ from 'moment';
/** @type {?} */
const moment = moment_;
window['moment'] = moment;
export class CalBuilder {
    constructor() {
        this.TIME_NEW_UPDATE = 12;
        this.roleUrl = 'xschedule/page/role';
        this.MODE = CalMode;
        this.ROLE = CalAttendeeRole;
        this.calSubType = CalSubType;
        this.calCategory = CalCategory;
        this.vnToLatin = vnToLatin;
        this.dateUtils = DateUtils;
        this.calAttendeeType = CalAttendeeType;
        this.calAttendeeState = CalAttendeeState;
        this.calType = CalType;
        this.subFirmDic = new SubFirmDictionary();
        this.carUtils = new CalendarCar(this.subFirmDic);
        this.configs = CalConfigs;
        this.resources = {
            days: [
                'Chủ nhật',
                'Thứ hai',
                'Thứ ba',
                'Thứ tư',
                'Thứ năm',
                'Thứ sáu',
                'Thứ bảy'
            ],
            statuses: [
                'Mới tạo',
                'Ban hành',
                'Chờ duyệt',
                'Đã hủy',
                'Thông qua',
                'Lịch đơn vị khác'
            ],
            allDay: 'Cả ngày',
            hour: 'giờ',
            minute: 'phút'
        };
        this._userRoles = {};
        this._mapEvents = {};
        this._queue = {};
        this._organs = [];
        this._myOrgans = [];
        this.typeIsDepartment = (/**
         * @param {?} type
         * @return {?}
         */
        (type) => type === CalType.DEPARTMENT);
        this.typeIsCommunity = (/**
         * @param {?} type
         * @return {?}
         */
        (type) => type === CalType.COMMUNITY);
        this.typeIsOrganization = (/**
         * @param {?} type
         * @return {?}
         */
        (type) => type === CalType.ORGANIZATION);
        this.catIsInvitation = (/**
         * @param {?} category
         * @return {?}
         */
        (category) => CalCategory.INVITATION === category);
        this.catIsPersonal = (/**
         * @param {?} category
         * @return {?}
         */
        (category) => CalCategory.PERSONAL === category);
        this.catIsCar = (/**
         * @param {?} category
         * @return {?}
         */
        (category) => CalCategory.CAR === category);
        this.catIsImportant = (/**
         * @param {?} category
         * @return {?}
         */
        (category) => CalCategory.IMPORTANT === category);
        this.formatDateStr = (/**
         * @param {?} date
         * @return {?}
         */
        (date) => this.formatDate(date, this.getConfigValue('DATE_FORMAT')));
        this.formatTimeStr = (/**
         * @param {?} date
         * @return {?}
         */
        (date) => this.formatDate(date, this.getConfigValue('TIME_FORMAT')));
        this.formatDateTimeStr = (/**
         * @param {?} date
         * @return {?}
         */
        (date) => this.formatDate(date, this.getConfigValue('DATE_TIME_FORMAT')));
        this.formatDate = (/**
         * @param {?} date
         * @param {?} format
         * @return {?}
         */
        (date, format) => moment(date).format(format));
        this.getEventStatus = (/**
         * @param {?} mode
         * @return {?}
         */
        (mode) => this.resources.statuses[mode]);
        this.isCreate = (/**
         * @param {?} mode
         * @return {?}
         */
        (mode) => mode === CalMode.CREATE);
        this.isPublish = (/**
         * @param {?} mode
         * @return {?}
         */
        (mode) => mode === CalMode.PUBLISH);
        this.isSent = (/**
         * @param {?} mode
         * @return {?}
         */
        (mode) => mode === CalMode.SENT);
        this.isCancel = (/**
         * @param {?} mode
         * @return {?}
         */
        (mode) => mode === CalMode.CANCEL);
        this.isOther = (/**
         * @param {?} mode
         * @return {?}
         */
        (mode) => mode === CalMode.OTHER);
    }
    /**
     * @return {?}
     */
    _applyConfig() {
        iNet.apply(this, DateUtils);
        // Apply config
        if (window['xCalendarConfigList'] && window['xCalendarConfigList'].length > 0) {
            window['xCalendarConfigList'].forEach((/**
             * @param {?} config
             * @return {?}
             */
            (config) => {
                /** @type {?} */
                let obj = this.getConfigObj(config.name);
                if (obj) {
                    $.extend(obj, config);
                }
                else {
                    this.configs.push(config);
                }
            }));
        }
        if (window['xCalendarConfigCustoms']) {
            for (let k in window['xCalendarConfigCustoms']) {
                /** @type {?} */
                let obj = this.getConfigObj(k);
                if (obj) {
                    obj.value = window['xCalendarConfigCustoms'][k];
                }
            }
        }
        // Apply resources
        if (window['xcalendarResources']) {
            this.resources = window['xcalendarResources'];
        }
        this.configs.forEach((/**
         * @param {?} obj
         * @return {?}
         */
        function (obj) {
            if (obj.value === 'false') {
                obj.value = false;
            }
        }));
    }
    /**
     * @param {?} name
     * @return {?}
     */
    getConfigValue(name) {
        /** @type {?} */
        let obj = this.getConfigObj(name);
        return obj && obj.value;
    }
    /**
     * @param {?} name
     * @return {?}
     */
    getConfigObj(name) {
        for (let i = 0; i < this.configs.length; i++) {
            if (this.configs[i].name === name) {
                return this.configs[i];
            }
        }
    }
    /**
     * @param {?} name
     * @param {?} executor
     * @param {?} callback
     * @return {?}
     */
    addQueue(name, executor, callback) {
        this._queue[name] = this._queue[name] || [];
        this._queue[name].push(callback);
        if (this._queue[name].length === 1) {
            new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            (resolve) => {
                executor(resolve);
            })).then((/**
             * @param {?} value
             * @return {?}
             */
            (value) => {
                this._queue[name].forEach((/**
                 * @param {?} fn
                 * @return {?}
                 */
                (fn) => {
                    fn && fn(value);
                }));
                delete this._queue[name];
            }));
        }
    }
    /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    ready(callback, options) {
        this._loadOrgans((/**
         * @param {?} organs
         * @return {?}
         */
        (organs) => {
            callback(organs, this._getOrganOnInit());
        }), options);
    }
    // Cache all org in a minute for performance
    /**
     * @param {?} key
     * @param {?=} value
     * @return {?}
     */
    cacheSessionStorage(key, value) {
        try {
            if (!value) {
                // Cache data in a minute
                /** @type {?} */
                let timeCache = 5 * 60000;
                /** @type {?} */
                let cache = sessionStorage.getItem(key);
                if (cache) {
                    cache = JSON.parse(cache);
                    if (new Date().getTime() - cache.time < timeCache) {
                        return cache.data;
                    }
                    else {
                        sessionStorage.removeItem(key);
                    }
                }
                return false;
            }
            else {
                sessionStorage.setItem(key, JSON.stringify({
                    time: new Date().getTime(),
                    data: value
                }));
            }
            // Fixed prev version
            localStorage.removeItem(key);
        }
        catch (e) { }
    }
    // Cache data on global window, refresh tab will reset
    /**
     * @param {?} key
     * @param {?=} data
     * @return {?}
     */
    cacheDataTemp(key, data) {
        this._cacheStore = this._cacheStore || {};
        if (data === undefined) {
            return this._cacheStore[key];
        }
        else {
            this._cacheStore[key] = data;
        }
    }
    /**
     * @param {?} orgid
     * @return {?}
     */
    setOrgId(orgid) {
        this.orgid = orgid;
        this._cacheOrgId(orgid);
    }
    /**
     * @return {?}
     */
    getOrgId() {
        return this.orgid || this._cacheOrgId() || iNet.organId;
    }
    /**
     * @param {?=} orgid
     * @return {?}
     */
    _cacheOrgId(orgid) {
        /** @type {?} */
        let id = 'xcalendar_orgid';
        try {
            if (!orgid) {
                return localStorage.getItem(id);
            }
            else {
                localStorage.setItem(id, orgid);
            }
        }
        catch (e) { }
    }
    /**
     * @param {?=} date
     * @return {?}
     */
    getHourMinute(date) {
        if (date.getTime) {
            // date object
            return date.getHours() * 60 + date.getMinutes();
        }
        else {
            // string hour
            date = date.split(':');
            return Number(date[0]) * 60 + Number(date[1]);
        }
    }
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    create(form, success, options) {
        return this.postForm({
            url: 'calbuilder/element/create',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (event && event.uuid)
                    this.indexEvent(event);
                success(event);
            })
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    updateEventByRole(params, success, options) {
        /** @type {?} */
        let organId = params && params.orgid || this.getOrgId();
        this.loadRole(organId, (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            let event = this.getEventByObjectId(params && params.element);
            if (this.isReviewer(organId)) {
                this.reviewUpdate(params, success, options);
            }
            else if (this.isOwnReviewer(organId) && event && this.meIsCreator(event)) {
                this.ownerReviewUpdate(params, success, options);
            }
            else {
                this.update(params, success, options);
            }
        }), options);
    }
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    update(form, success, options) {
        return this.postForm({
            url: 'calbuilder/element/update',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (event && event.uuid)
                    this.indexEvent(event);
                success && success(event);
            })
        }, options);
    }
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    reviewUpdate(form, success, options) {
        return this.postForm({
            url: 'calbuilder/reviewupdate',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (event && event.uuid)
                    this.indexEvent(event);
                success && success(event);
            })
        }, options);
    }
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    ownerReviewUpdate(form, success, options) {
        return this.postForm({
            url: 'calbuilder/owner/reviewupdate',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (event && event.uuid)
                    this.indexEvent(event);
                success && success(event);
            })
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    remove(params, callback, options) {
        return this.sendRequest({
            url: 'calbuilder/element/delete',
            data: params,
            success: callback
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    cancel(params, success, options) {
        /** @type {?} */
        let organId = params && params.orgid || this.getOrgId();
        this.loadRole(organId, (/**
         * @return {?}
         */
        () => {
            if (this.isReviewer(organId)) {
                this.reviewCancel(params, success, options);
            }
            else if (this.isOwnReviewer(organId)) {
                this.ownerReviewCancel(params, success, options);
            }
        }), options);
    }
    // element, members
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    reviewNotify(params, success, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/reviewnotify',
            data: params,
            success: success
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    reviewCancel(params, success, options) {
        return this.sendRequest({
            url: 'calbuilder/reviewcancel',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (event && event.uuid)
                    this.indexEvent(event);
                success && success(event);
            })
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    ownerReviewCancel(params, success, options) {
        return this.sendRequest({
            url: 'calbuilder/owner/reviewcancel',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (event && event.uuid)
                    this.indexEvent(event);
                success && success(event);
            })
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    deleteAttachment(params, success, options) {
        /**
         * fileID: 'id1,id2'
         * element: 'uuid'
         */
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/file/delete',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                success && success(this.indexEvent(event));
            })
        }, options);
    }
    /**
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadOrganMembers(organId, callback, options) {
        organId = organId || this.getOrgId();
        return this.searchAttendee({ orgid: organId }, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    searchAttendee(params, success, options) {
        this._searchAttendee(params, (/**
         * @param {?} members
         * @return {?}
         */
        function (members) {
            members.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            (a, b) => a.fullname.localeCompare(b.fullname)));
            success(members);
        }), options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    _searchAttendee(params, success, options) {
        return this.sendRequest({
            url: 'calbuilder/attendee/search',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                /** @type {?} */
                let members = data && data.elements || [];
                members = members.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                (item) => !item.alias));
                members.forEach((/**
                 * @param {?} member
                 * @return {?}
                 */
                (member) => this._indexAccount(member)));
                success(members);
            }),
            error: (/**
             * @return {?}
             */
            function () {
                success([]);
            })
        }, options, true);
    }
    /**
     * @param {?} account
     * @return {?}
     */
    _indexAccount(account) {
        // Remove suffix email before search user "binh@binhdinh.gov.vn"
        // don't search on suffix "@binhdinh.gov.vn"
        account['_index'] = vnToLatin(account.fullname.toLowerCase()) + ' ' +
            account.username.split('@')[0];
        account.display = this.insertPrefixAttendee(account.fullname || account.username);
        return account;
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadDepartments(params, callback, options) {
        return this.sendRequest({
            url: 'calbuilder/department/list',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                data = data && data.items || [];
                data.sort((/**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */
                (a, b) => {
                    return a.order - b.order;
                }));
                callback(data);
            }),
            error: (/**
             * @return {?}
             */
            () => {
                callback([]);
            })
        }, options, true);
    }
    /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    getMyOrg(callback, options) {
        /** @type {?} */
        let keyCache = 'cal_my_organs';
        // Remove data unused prev version
        localStorage.removeItem(keyCache);
        keyCache += iNet.usercode;
        this._myOrgans = this._myOrgans.length && this._myOrgans || this.cacheSessionStorage(keyCache) || [];
        if (this._myOrgans.length > 0) {
            return callback(this._myOrgans);
        }
        this.sendRequest({
            url: 'plugin/organization/list',
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                /** @type {?} */
                let organs = this._convertOrg(data && data.elements || []);
                if (organs.length > 0) {
                    organs = this.visibleMyOrgans(organs);
                    this._myOrgans = organs;
                    this.cacheSessionStorage(keyCache, organs);
                }
                callback(organs);
            }),
            error: (/**
             * @return {?}
             */
            () => callback([]))
        }, options, true);
    }
    /**
     * @param {?} organs
     * @return {?}
     */
    visibleMyOrgans(organs) {
        return organs.filter((/**
         * @param {?} organ
         * @return {?}
         */
        (organ) => {
            /** @type {?} */
            let _organ = this.getOrgById(organ.firmUUID);
            return _organ && !!_organ.groupType;
        }));
    }
    /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    getAllOrg(callback, options) {
        /** @type {?} */
        let keyCache = 'cal_organ_list';
        // Remove data unused prev version
        localStorage.removeItem(keyCache);
        keyCache += iNet.usercode;
        this._organs = this._organs.length && this._organs || this.cacheSessionStorage(keyCache) || [];
        // Check all org is cached
        if (this._organs.length > 0) {
            return callback(this._organs);
        }
        return this.sendRequest({
            data: { pageSize: -1 },
            url: 'plugin/organization/search',
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                /** @type {?} */
                let organs = this._convertOrg(data && data.items || []);
                if (organs.length > 0) {
                    // Only get organ has groupType
                    organs = this.visibleOrgans(organs);
                    this._organs = organs;
                    this.cacheSessionStorage(keyCache, this._organs);
                }
                callback(organs);
            }),
            error: (/**
             * @return {?}
             */
            () => callback([]))
        }, options, true);
    }
    /**
     * @param {?} organs
     * @return {?}
     */
    visibleOrgans(organs) {
        return organs.filter((/**
         * @param {?} organ
         * @return {?}
         */
        (organ) => !!organ.groupType));
    }
    /**
     * @private
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    _loadOrgans(callback, options) {
        if (this._organs && this._organs.length > 0) {
            callback(this._organs.slice());
        }
        else {
            this.getAllOrg((/**
             * @param {?} organs
             * @return {?}
             */
            (organs) => {
                // Detect user login
                this.loadRole('', (/**
                 * @param {?} role
                 * @return {?}
                 */
                (role) => {
                    this.isAuth = !!role.isAuth;
                    if (role.isAuth) {
                        this.getMyOrg((/**
                         * @param {?} myOrgans
                         * @return {?}
                         */
                        (myOrgans) => {
                            // Move my organs to first
                            for (let i = myOrgans.length - 1; i >= 0; i--) {
                                /** @type {?} */
                                let myOrgan = myOrgans[i];
                                for (let j = 0; j < organs.length; j++) {
                                    /** @type {?} */
                                    let organ = organs[j];
                                    if (organ.firmUUID === myOrgan.firmUUID) {
                                        organs.splice(j, 1);
                                        organ.myOrg = true;
                                        organs.unshift(organ);
                                        break;
                                    }
                                }
                            }
                            callback(organs);
                        }), options);
                    }
                    else {
                        callback(organs);
                    }
                }), options);
            }), options);
        }
    }
    /**
     * @param {?} callback
     * @param {?=} domain
     * @return {?}
     */
    getTreeOrg(callback, domain) {
        this._loadOrgans((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            /** @type {?} */
            let tree = [];
            // Set children is empty
            data.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                item.children = [];
            }));
            // Build tree org list
            data.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                if (item.parentID) {
                    /** @type {?} */
                    let parentOrgan = this.getOrgById(item.parentID);
                    if (parentOrgan) {
                        parentOrgan.children.push(item);
                    }
                }
                else {
                    tree.push(item);
                }
            }));
            tree.data = data;
            callback && callback(tree);
        }), domain);
    }
    /**
     * @param {?=} item
     * @return {?}
     */
    indexOrgSearch(item) {
        item._index = vnToLatin([item.firmName, item.firmUUID, item.firmPrefix].join(' ').toLowerCase());
    }
    /**
     * @param {?=} orgid
     * @return {?}
     */
    isMyOrg(orgid) {
        /** @type {?} */
        let org = this.getOrgById(orgid);
        return org && org.myOrg;
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadEvent(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/element/load',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                if (!data || !data.uuid) {
                    // Event not found, try to load review event
                    this.loadRole(params.orgid, (/**
                     * @return {?}
                     */
                    () => {
                        if (this.isReviewer(params.orgid)) {
                            this.loadEventReview(params, callback, options);
                        }
                        else {
                            callback && callback(data);
                        }
                    }));
                    return;
                }
                this.indexEvent(data);
                callback && callback(data);
            })
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    updateState(params, callback, options) {
        if (params.notes) {
            params.notes = this.formatTextUpdate(params.notes);
        }
        return this.loadEvent(params, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    loadEventReview(params, success, options) {
        return this.sendRequest({
            url: 'calbuilder/reviewload',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                if (event && event.uuid)
                    this.indexEvent(event);
                success && success(event);
            })
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    loadGroupEvent(params, success, options) {
        return this.sendRequest({
            url: 'calbuilder/element/group',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => success(this.modifyEvent(data && data.items || [])))
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    loadOrgEvents(params, callback, options) {
        this._loadOnDifferYear(this._loadOrgEvents, params, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _loadOrgEvents(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/mainboard/view',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(this.modifyEvent(data && data.items || [])))
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    loadCommunityEvents(params, callback, options) {
        this._loadOnDifferYear(this._loadCommunityEvents, params, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _loadCommunityEvents(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/mainboard/published',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(this.modifyEvent(data && data.items || []))),
            useOrgId: true
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    loadAllPublishEvents(params, callback, options) {
        this._loadOnDifferYear(this._loadAllPublishEvents, params, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _loadAllPublishEvents(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/mainboard/published',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(this.modifyEvent(data && data.items || []))),
            removeOrgId: true
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadPublishEvents(params, callback, options) {
        // Load role
        this.loadRole(params.orgid, (/**
         * @return {?}
         */
        () => {
            // Reviewer
            if (this.isMyOrg(params.orgid)) {
                this.loadOrgEvents(params, callback, options);
            }
            else {
                this.loadCommunityEvents(params, callback, options);
            }
        }), options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    getReviewEvents(params, callback, options) {
        this._loadOnDifferYear(this._getReviewEvents, params, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _getReviewEvents(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/reviewlist',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(this.modifyEvent(data && data.items || [])))
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    getViewEvents(params, callback, options) {
        this._loadOnDifferYear(this._getViewEvents, params, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _getViewEvents(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/viewlist',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(this.modifyEvent(data && data.items || [])))
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    reviewSearchEvents(params, callback, options) {
        this._loadOnDifferYear(this._reviewSearchEvents, params, callback, options);
    }
    // members: ",,"
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _reviewSearchEvents(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/reviewsearch',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(this.modifyEvent(data && data.items || [])))
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    getOwnEvents(params, callback, options) {
        this._loadOnDifferYear(this._getOwnEvents, params, callback, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _getOwnEvents(params, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/element/list',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(this.modifyEvent(data && data.items || [])))
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    getListEvents(params, callback, options) {
        /** @type {?} */
        let promises = [];
        params.orgid = params.orgid || this.getOrgId();
        // Load review event
        this.loadRole(params.orgid, (/**
         * @return {?}
         */
        () => {
            // My events
            promises.push(new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            (resolve) => {
                /** @type {?} */
                let _options = $.extend({}, options);
                _options.error = (/**
                 * @return {?}
                 */
                function () {
                    resolve([]);
                });
                this.getOwnEvents(params, resolve, _options);
            })));
            if (this.isReviewer(params.orgid)) {
                // Reviewer
                promises.push(new Promise((/**
                 * @param {?} resolve
                 * @return {?}
                 */
                (resolve) => {
                    /** @type {?} */
                    let _options = $.extend({}, options);
                    _options.error = (/**
                     * @return {?}
                     */
                    function () {
                        resolve([]);
                    });
                    this.getReviewEvents(params, resolve, _options);
                })));
            }
            else if (this.isViewer(params.orgid)) {
                // Viewer
                promises.push(new Promise((/**
                 * @param {?} resolve
                 * @return {?}
                 */
                (resolve) => {
                    /** @type {?} */
                    let _options = $.extend({}, options);
                    _options.error = (/**
                     * @return {?}
                     */
                    function () {
                        resolve([]);
                    });
                    this.getViewEvents(params, resolve, _options);
                })));
            }
            // Load all promises
            Promise.all(promises).then((/**
             * @param {?} values
             * @return {?}
             */
            (values) => {
                /** @type {?} */
                let events = [];
                values.forEach((/**
                 * @param {?} value
                 * @return {?}
                 */
                (value) => events = events.concat(value || [])));
                events = this.removeEventDuplicate(events);
                callback(this.modifyEvent(events));
            }));
        }), options);
    }
    /**
     * @param {?} events
     * @return {?}
     */
    modifyEvent(events) {
        if (!events)
            return;
        this.removeEventDuplicate(events);
        events.forEach((/**
         * @param {?} item
         * @return {?}
         */
        (item) => this.indexEvent(item)));
        this.sortEvents(events);
        return events;
    }
    /**
     * @param {?} events
     * @return {?}
     */
    removeEventDuplicate(events) {
        if (events.length > 1) {
            for (let i = 1; i < events.length; i++) {
                for (let j = i - 1; j >= 0; j--) {
                    if (events[i].uuid === events[j].uuid && events[i].day === events[j].day) {
                        events.splice(i, 1);
                        --i;
                        break;
                    }
                }
            }
        }
        return events;
    }
    /**
     * @param {?} events
     * @return {?}
     */
    sortEvents(events) {
        // sort by start date
        this.sortEventStart(events);
        this.sortEventPriority(events);
        return events;
    }
    /**
     * @param {?} events
     * @return {?}
     */
    sortEventStart(events) {
        this._sort(events, 'from');
        return events;
    }
    /**
     * @param {?} events
     * @return {?}
     */
    sortEventPriority(events) {
        // sort by priority on day
        for (let i = 0; i < events.length; i++) {
            /** @type {?} */
            let item = events[i];
            item._order = Number(item._order) || 0;
            if (item._order > 0) {
                moveFirst(item);
                i = events.indexOf(item);
            }
        }
        /**
         * @param {?} item
         * @return {?}
         */
        function moveFirst(item) {
            if (item._order !== 0) {
                /** @type {?} */
                let index = events.indexOf(item);
                if (index > 0) {
                    /** @type {?} */
                    let tmp = events[index - 1];
                    if (item.day === tmp.day && (tmp._order === 0 || item._order < tmp._order)) {
                        events[index] = tmp;
                        events[index - 1] = item;
                        moveFirst(item);
                    }
                }
            }
        }
        return events;
    }
    /**
     * @param {?} arr
     * @param {?=} orderBy
     * @return {?}
     */
    _sort(arr, orderBy) {
        for (let i = 1; i < arr.length; i++) {
            if (i === 0)
                continue;
            /** @type {?} */
            let item = arr[i];
            /** @type {?} */
            let prev = arr[i - 1];
            if (item[orderBy] < prev[orderBy]) {
                arr[i] = prev;
                arr[i - 1] = item;
                i -= 2;
            }
        }
    }
    // keywords: string[]
    /**
     * @param {?} events
     * @param {?} keywords
     * @return {?}
     */
    searchEventsByKeywords(events, keywords) {
        if (keywords && keywords.length > 0) {
            return events.filter((/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                for (let i = 0; i < keywords.length; i++) {
                    if (event._index.indexOf(keywords[i]) < 0) {
                        return false;
                    }
                }
                return true;
            }));
        }
        return events.slice();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    indexEvent(event) {
        // Save event with link to get later
        event.objectId = event.uuid + event.mode + event.day;
        this._mapEvents[event.objectId] = event;
        this._mapEvents[event.uuid] = event;
        if (!event._indexed) {
            this.indexEventMembers(event);
            if (window['beforeIndexEvent']) {
                window['beforeIndexEvent'](event);
            }
            event.subjectDisplay = this.formatTextDisplay(event.subject || '');
            event.locationDisplay = this.formatTextDisplay(event.location || '');
            this.indexRepeat(event);
            this.indexEventAttribute(event);
            this.indexEventDate(event);
            this.indexEventSearch(event);
            this.indexCustom(event);
            event.category = event.category || '';
            event.categoryIcon = CalCategoryIcon[event.category];
            event.department = event.departID;
            event._indexed = true;
        }
        return event;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    indexEventMembers(event) {
        /** @type {?} */
        let obj = {};
        for (let k in CalAttendeeRole) {
            obj[k] = [];
        }
        event.members = event.members || [];
        event.attendees = [];
        event.members.forEach((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            item.name = this.attendeeDisplay(item);
            item.value = this.attendeeValue(item);
            // Type attendee
            if (item.alias) {
                item.type = CalAttendeeType.ALIAS;
            }
            else if (item.username) {
                item.type = CalAttendeeType.USER;
            }
            else {
                item.type = CalAttendeeType.ORGAN;
            }
            obj[item.role] = obj[item.role] || [];
            obj[item.role].push(item);
            if (item.role === CalAttendeeRole.CREATOR) {
                event.creator = item.username;
                event.fullname = item.fullname;
                event.isCreator = this.usernameIsMe(event.creator);
                event.creatorStr = event.fullname + '(' + event.creator + ')';
            }
            else {
                event.attendees.push(item);
            }
        }));
        event.chairmanStr = this._groupMember2String(obj.CHAIRMAN);
        event.memberStr = this._groupMember2String(obj.MEMBER);
        // Todo: test prepare str
        event.prepareStr = this._groupMember2String(obj.PREPARER);
        event.watcherStr = event.observerStr = this._groupMember2String(obj.OBSERVER);
        event.membersDecode = obj;
    }
    /**
     * @param {?} members
     * @return {?}
     */
    _groupMember2String(members) {
        /** @type {?} */
        let str = '';
        /** @type {?} */
        let data = this.groupAttendeeByOrgan(members);
        for (let k in data) {
            /** @type {?} */
            let items = data[k];
            /** @type {?} */
            let organ = items.filter((/**
             * @param {?} item
             * @return {?}
             */
            (item) => item.type === CalAttendeeType.ORGAN))[0];
            if (organ) {
                str += '<b>' + organ.organName + '</b>: ';
            }
            // Todo: consider to use this.buildMemberShow
            items.forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                if (item.type !== CalAttendeeType.ORGAN) {
                    str += item.name + ', ';
                }
            }));
            str = str.slice(0, -2);
            str += '<br>';
        }
        return str.slice(0, -4);
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    indexEventAttribute(event) {
        for (let k in event.attribute) {
            event['_' + k] = event.attribute[k];
        }
        event._order = Number(event._order) || 0;
        if (event._subType) {
            event.type = event._subType;
        }
    }
    /**
     * @param {?} calEvent
     * @return {?}
     */
    initEvent(calEvent) {
        calEvent = $.extend({}, calEvent);
        this._initEventDate(calEvent);
        calEvent.startTime = calEvent.from.format('H:i');
        calEvent.toTime = calEvent.to.format('H:i');
        return calEvent;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _initEventDate(event) {
        if (event) {
            /** @type {?} */
            let date;
            if (event && event.uuid) {
                date = new Date(event.year, 0, event.day);
                if (date.getTime()) {
                    event.from = new Date(date.getTime() + event.startTime * 60 * 1000);
                    event.to = new Date(date.getTime() + event.toTime * 60 * 1000);
                }
            }
            if (!date || !date.getTime()) {
                date = new Date();
                date.setHours(date.getHours() + 1, 0, 0, 0);
                event.from = date;
                date = new Date(date);
                date.setHours(date.getHours() + 1);
                event.to = date;
            }
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    indexEventDate(event) {
        this._initEventDate(event);
        // all day event
        event.allDay = event._allday === 'true';
        event.timeStr = this.formatDateStr(event.from);
        if (event.allDay) {
            event.startTimeStr = event.toTimeStr = this.resources.allDay;
            event.hourStr = this.resources.allDay;
        }
        else {
            event.startTimeStr = this.formatTimeStr(event.from);
            event.toTimeStr = this.formatTimeStr(event.to);
            event.hourStr = event.startTimeStr + ' - ' + event.toTimeStr;
        }
        event.timeStr += '; ' + event.hourStr;
        event.start = event.from;
        event.end = event.to;
        // Time create and update
        if (event.createTime) {
            event.createDate = new Date(event.createTime);
            event.createDateStr = this.formatDateTimeStr(event.createDate);
        }
        if (event.updateTime) {
            event.updateDate = new Date(event.updateTime);
            event.updateDateStr = this.formatDateTimeStr(event.updateDate);
        }
        /** @type {?} */
        let hourUpdated = new Date().getTime() - (event.updateTime || event.createTime);
        hourUpdated /= (60 * 60 * 1000);
        if (hourUpdated < 0) {
            hourUpdated = 0;
        }
        if (hourUpdated <= this.TIME_NEW_UPDATE) {
            event.recentUpdated = true;
            event.lastUpdateStr = this.getTextForMinutes(hourUpdated * 60);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    indexEventSearch(event) {
        // Index data to search
        event._index = '';
        event._members = [];
        if (event.subject) {
            event._index = vnToLatin(event.subject.toLowerCase());
        }
        // Index location to search
        if (event.location) {
            event._index += ' ' + vnToLatin(event.location.toLowerCase());
        }
        for (let i = 0; i < event.members.length; i++) {
            /** @type {?} */
            let member = event.members[i];
            if (member.fullname)
                event._index += ' ' + vnToLatin(member.fullname.toLowerCase());
            if (member.username) {
                event._index += ' ' + member.username.toLowerCase();
                // attendee
                if (member.role !== CalAttendeeRole.CREATOR) {
                    event._members.push(member.username);
                }
            }
        }
        // Attachments
        if (event.attachments) {
            event.attachments.forEach((/**
             * @param {?} file
             * @return {?}
             */
            (file) => {
                file.orgid = event.firmUUID;
            }));
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    indexRepeat(event) {
        //"FREQ=WEEKLY;UNTIL=20170707;INTERVAL=1;BYDAY=MO"
        if (event.rrule) {
            /** @type {?} */
            let obj = {};
            event.rrule.split(';').forEach((/**
             * @param {?} item
             * @return {?}
             */
            (item) => {
                item = item.split('=');
                if (item[0] === 'FREQ') {
                    obj.type = obj.rrule = obj.rrmode = item[1];
                }
                else if (item[0] === 'BYDAY') {
                    obj.wkdays = item[1].split(',');
                }
            }));
            if (event.lstart) {
                obj.lstart = new Date(event.lstart);
            }
            if (event.until) {
                obj.until = new Date(event.until);
            }
            event.repeat = obj;
        }
    }
    // Custom index
    /**
     * @param {?} event
     * @return {?}
     */
    indexCustom(event) {
    }
    /**
     * @param {?=} orgid
     * @return {?}
     */
    isReviewer(orgid) {
        orgid = orgid || this.getOrgId();
        return this._userRoles[orgid] && !!this._userRoles[orgid].reviewer;
    }
    /**
     * @param {?=} orgid
     * @return {?}
     */
    isOwnReviewer(orgid) {
        orgid = orgid || this.getOrgId();
        return this._userRoles[orgid] && !!this._userRoles[orgid].own_reviewer;
    }
    /**
     * @param {?=} orgid
     * @return {?}
     */
    isViewer(orgid) {
        orgid = orgid || this.getOrgId();
        return this._userRoles[orgid] && !!this._userRoles[orgid].viewer;
    }
    /**
     * @param {?=} query
     * @return {?}
     */
    getQueryParams(query) {
        /** @type {?} */
        let params = {};
        query = query || location.search;
        query.replace('?', '').split('&').forEach((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            if (item) {
                /** @type {?} */
                let equalIndex = item.indexOf('=');
                if (equalIndex > -1)
                    params[item.slice(0, equalIndex)] = item.slice(equalIndex + 1);
            }
        }));
        return params;
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    buildQueryString(params) {
        /** @type {?} */
        let query = '';
        for (let k in params) {
            query += k + '=' + encodeURIComponent(params[k]) + '&';
        }
        return query.slice(0, -1);
    }
    /**
     * @param {?} usercode
     * @param {?=} domain
     * @return {?}
     */
    getUserAvatar(usercode, domain) {
        return AjaxAPI.getUrl('system/userprofile/photo', domain) +
            '?usercode=' + usercode + '&thumbnail=50';
    }
    /**
     * @param {?} event
     * @param {?} callback
     * @param {?=} currentOrganId
     * @param {?=} options
     * @return {?}
     */
    getPermissionOnEvent(event, callback, currentOrganId, options) {
        /** @type {?} */
        let mode = event.mode;
        if (this.isOther(mode)) {
            callback();
            return;
        }
        if (!event.uuid) {
            return callback({
                save: true,
                saveCreate: true
            });
        }
        currentOrganId = currentOrganId || this.getOrgId() || event.firmUUID;
        this.loadRole(currentOrganId, (/**
         * @return {?}
         */
        () => {
            this.indexEvent(event);
            /** @type {?} */
            let permissions;
            /** @type {?} */
            const canReview = this.isReviewer(currentOrganId) ||
                (this.isOwnReviewer(currentOrganId) && this.canOwnReviewEvent(event));
            /** @type {?} */
            const canReviewUpdate = !this.isCreate(mode) && !this.isCancel(mode) && canReview;
            if (event.firmUUID === currentOrganId) {
                /** @type {?} */
                const isEditMyEvent = this.canUpdateOwnEvent(event);
                if (event.type === CalSubType.PERSONAL && !this.getConfigValue('PERSONAL_CAN_PUBLISH')) {
                    // Personal calendar
                    permissions = {
                        save: isEditMyEvent,
                        delete: isEditMyEvent
                    };
                }
                else {
                    permissions = {
                        save: isEditMyEvent || canReviewUpdate,
                        approve: canReview && (this.isSent(mode) || isEditMyEvent),
                        reject: canReview && this.isSent(mode),
                        send: isEditMyEvent && !canReview,
                        revert: (this.isSent(mode) && event.isCreator) ||
                            ((this.isCancel(mode) || this.isPublish(mode)) && canReview),
                        cancel: canReview && this.isPublish(mode),
                        delete: isEditMyEvent
                    };
                }
            }
            else {
                // External update attendees
                permissions = {};
                if (this.isPublish(event.mode) && this.hasFirmAttendee(event, currentOrganId) &&
                    this.canExternalUpdateEvent(event, currentOrganId)) {
                    permissions = { updateAttendee: true };
                }
            }
            permissions.print = true;
            permissions.exportInvitation = this.catIsInvitation(event.category) && event._templateId;
            callback(permissions);
        }), options);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    canUpdateOwnEvent(event) {
        return event.isCreator && this.isCreate(event.mode);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    canReviewEvent(event) {
        return this.isReviewer(event.firmUUID);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    canOwnReviewEvent(event) {
        return this.isOwnReviewer(event.firmUUID) && event.isCreator;
    }
    /**
     * @param {?} event
     * @param {?} ownOrganId
     * @return {?}
     */
    canUpdateAttendee(event, ownOrganId) {
        return this.isFirmEvent(event, ownOrganId) && (this.canUpdateOwnEvent(event) ||
            this.canOwnReviewEvent(event) ||
            this.canReviewEvent(event));
    }
    /**
     * @param {?} event
     * @param {?} organId
     * @return {?}
     */
    canExternalUpdateEvent(event, organId) {
        return this.isPublish(event.mode) && !this.isFirmEvent(event, organId) && this.isReviewer(organId);
    }
    /**
     * @param {?} event
     * @param {?} ownOrganId
     * @return {?}
     */
    isFirmEvent(event, ownOrganId) {
        return event.firmUUID === ownOrganId;
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    searchAnyAttendee(params, callback, options) {
        this.searchUserAsAttendee(params, (/**
         * @param {?} users
         * @return {?}
         */
        (users) => {
            this.searchOrganAsAttendee(params, (/**
             * @param {?} orgs
             * @return {?}
             */
            (orgs) => {
                orgs.forEach((/**
                 * @param {?} org
                 * @return {?}
                 */
                (org) => users.push(org)));
                callback(users);
            }));
        }), options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    searchUserAsAttendee(params, callback, options) {
        /** @type {?} */
        let organId = params.orgid || this.getOrgId();
        /** @type {?} */
        let organName = this.getOrgNameById(organId);
        this.loadOrganMembers(organId, (/**
         * @param {?} members
         * @return {?}
         */
        (members) => {
            /** @type {?} */
            let keyword = params.keyword || '';
            /** @type {?} */
            let limit = params.limit || -1;
            /** @type {?} */
            let attendees = [];
            keyword = vnToLatin(keyword).toLowerCase();
            for (let i = 0; i < members.length; i++) {
                /** @type {?} */
                let member = members[i];
                if (!keyword || member._index.indexOf(keyword) > -1) {
                    /** @type {?} */
                    let attendee = this.memberToAttendee(member);
                    attendee.organId = organId;
                    attendee.organName = organName;
                    attendees.push(attendee);
                    if (limit > 0 && attendees.length >= limit) {
                        break;
                    }
                }
            }
            callback(attendees);
        }), options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    searchOrganAsAttendee(params, callback, options) {
        this.loadOrgByKeyword(params, (/**
         * @param {?} organs
         * @return {?}
         */
        (organs) => {
            callback(organs.map((/**
             * @param {?} organ
             * @return {?}
             */
            (organ) => this.organToAttendee(organ))));
        }), options);
    }
    /**
     * @param {?} fullname
     * @return {?}
     */
    insertPrefixAttendee(fullname) {
        if (this.getConfigValue('PREFIX_ATTENDEE_NAME')) {
            return this.getConfigValue('PREFIX_ATTENDEE_NAME') + fullname;
        }
        return fullname;
    }
    /**
     * @param {?} member
     * @return {?}
     */
    memberToAttendee(member) {
        return {
            username: member.username,
            fullname: member.fullname,
            display: this.insertPrefixAttendee(member.fullname),
            type: CalAttendeeType.USER
        };
    }
    /**
     * @param {?} organ
     * @return {?}
     */
    organToAttendee(organ) {
        return {
            organId: organ.firmUUID,
            organName: organ.firmName,
            type: CalAttendeeType.ORGAN
        };
    }
    /**
     * @param {?} alias
     * @param {?} organId
     * @return {?}
     */
    aliasToAttendee(alias, organId) {
        organId = organId || this.getOrgId();
        return {
            alias: alias,
            organId: organId,
            organName: this.getOrgNameById(organId),
            type: CalAttendeeType.ALIAS
        };
    }
    // Search user and orgs to select when create calendar
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadAttendeeByName(params, callback, options) {
        if (!params || !params.keyword)
            return callback([]);
        // Limit user search
        params.limit = params.limit || 5;
        this.loadUserByKeyword(params, (/**
         * @param {?} users
         * @return {?}
         */
        (users) => {
            this.loadOrgByKeyword(params, (/**
             * @param {?} orgs
             * @return {?}
             */
            (orgs) => {
                callback(users.concat(orgs));
            }));
        }), options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadUserByKeyword(params, callback, options) {
        // Load all users and search on client
        this.searchAttendee({
            orgid: params.orgid
        }, (/**
         * @param {?} members
         * @return {?}
         */
        (members) => {
            /** @type {?} */
            let users = [];
            /** @type {?} */
            let keyword = vnToLatin(params.keyword.toLowerCase());
            for (let i = 0; i < members.length; i++) {
                /** @type {?} */
                let user = members[i];
                if (user._index.indexOf(keyword) > -1) {
                    users.push({
                        name: user.fullname,
                        value: user.username,
                        type: CalAttendeeType.USER,
                        member: user
                    });
                    if (users.length >= params.limit) {
                        break;
                    }
                }
            }
            callback(users);
        }), options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadOrgByKeyword(params, callback, options) {
        /** @type {?} */
        let data = [];
        this._loadOrgans((/**
         * @param {?} orgs
         * @return {?}
         */
        (orgs) => {
            // Filter orgs match
            /** @type {?} */
            let key = vnToLatin(params.keyword).toLowerCase();
            for (let i = 0; i < orgs.length; i++) {
                /** @type {?} */
                let item = orgs[i];
                if (item._index.indexOf(key) > -1) {
                    data.push(item);
                    if (data.length >= params.limit)
                        break;
                }
            }
            callback(data);
        }), options);
    }
    /**
     * @param {?} item
     * @param {?=} domain
     * @param {?=} event
     * @return {?}
     */
    getFileUrl(item, domain, event) {
        if (!item)
            return '';
        /** @type {?} */
        let orgid = item.orgid || this.getOrgId();
        /** @type {?} */
        let _role = this._userRoles[orgid];
        /** @type {?} */
        let url;
        if (_role && _role.isAuth && this.isMyOrg(orgid)) {
            url = 'calbuilder/binary/view';
        }
        else if (this.hasFirmAttendee(event, this.getOrgId())) {
            orgid = this.getOrgId();
            url = 'calbuilder/binary/view';
        }
        else {
            url = 'calbuilder/mainboard/binary/view';
        }
        return this.getUrlByOrgId(url, orgid, domain) + '?fileID=' + item.uuid + '&element=' + item.folder + '&orgid=' + orgid;
    }
    /**
     * @param {?} item
     * @param {?=} domain
     * @param {?=} event
     * @return {?}
     */
    getFileViewUrl(item, domain, event) {
        /** @type {?} */
        let url = this.getFileUrl(item, domain, event);
        if (item.file && item.file.slice(-3).toLowerCase() === "pdf") {
            url += '&view=pdf&t=' + new Date().getTime();
        }
        return url;
    }
    /**
     * @param {?} event
     * @param {?} organId
     * @return {?}
     */
    hasFirmAttendee(event, organId) {
        if (event && organId) {
            for (let i = 0; i < event.members.length; i++) {
                /** @type {?} */
                let member = event.members[i];
                if (member.type === CalAttendeeType.ORGAN && member.organId === organId) {
                    return true;
                }
            }
        }
    }
    /**
     * @param {?} uuid
     * @return {?}
     */
    getOrgByUid(uuid) {
        for (let i = 0; i < this._organs.length; i++) {
            if (this._organs[i].uuid === uuid) {
                return this._organs[i];
            }
        }
    }
    /**
     * @param {?} orgId
     * @return {?}
     */
    getOrgById(orgId) {
        for (let i = 0; i < this._organs.length; i++) {
            if (this._organs[i].firmUUID === orgId) {
                return this._organs[i];
            }
        }
    }
    /**
     * @param {?} orgId
     * @return {?}
     */
    getOrgNameById(orgId) {
        /** @type {?} */
        let organ = this.getOrgById(orgId);
        return organ && organ.firmName || '';
    }
    /**
     * @param {?} prefix
     * @return {?}
     */
    getOrgByPrefix(prefix) {
        for (let i = 0; i < this._organs.length; i++) {
            if (this._organs[i].firmPrefix === prefix) {
                return this._organs[i];
            }
        }
    }
    /**
     * @param {?} url
     * @param {?} orgid
     * @param {?=} domain
     * @return {?}
     */
    getUrlByOrgId(url, orgid, domain) {
        /** @type {?} */
        let item = this.getOrgById(orgid);
        return AjaxAPI.getUrl(url, domain, item && item.firmPrefix);
    }
    /**
     * @param {?} orgid
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadRole(orgid, callback, options) {
        orgid = orgid || this.getOrgId();
        if (this._userRoles[orgid]) {
            return callback(this._userRoles[orgid]);
        }
        this.sendRequest({
            url: this.roleUrl,
            data: { orgid: orgid },
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                try {
                    data = JSON.parse(data);
                    if (data.isAuth) {
                        this._userRoles[orgid] = data;
                        iNet.usercode = data.usercode;
                    }
                    callback(data);
                }
                catch (e) {
                    callback({});
                }
            }),
            error: (/**
             * @return {?}
             */
            () => {
                callback({});
            })
        }, options, true);
    }
    /**
     * @param {?} organIds
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadRoles(organIds, callback, options) {
        /** @type {?} */
        let promises = [];
        organIds.forEach((/**
         * @param {?} organId
         * @return {?}
         */
        (organId) => {
            promises.push(new Promise(((/**
             * @param {?} resolve
             * @return {?}
             */
            resolve => {
                this.loadRole(organId, resolve, options);
            }))));
        }));
        Promise.all(promises).then((/**
         * @param {?} values
         * @return {?}
         */
        (values) => callback(values)));
    }
    /**
     * @param {?} events
     * @param {?=} date
     * @param {?=} firstDay
     * @return {?}
     */
    generateEventWeekView(events, date, firstDay) {
        /** @type {?} */
        let segEvents = [];
        /** @type {?} */
        let startWeek = DateUtils.getDateStartWeek(date, firstDay);
        for (let i = 0; i < 7; i++) {
            /** @type {?} */
            let dayOfYear = DateUtils.dayOfYear(startWeek);
            segEvents.push(this.generateEventView(events.filter((/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                return event.day === dayOfYear;
            })), startWeek));
            startWeek.setDate(startWeek.getDate() + 1);
        }
        return segEvents;
    }
    /**
     * @param {?} segEvents
     * @return {?}
     */
    getLastUpdateSegEvents(segEvents) {
        /** @type {?} */
        let lastUpdate;
        segEvents.forEach((/**
         * @param {?} segEvent
         * @return {?}
         */
        (segEvent) => {
            segEvent.items.forEach((/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                /** @type {?} */
                let lastUpdateEvent = event.updateDate || event.createDate;
                if (!lastUpdate || lastUpdate < lastUpdateEvent) {
                    lastUpdate = lastUpdateEvent;
                }
            }));
        }));
        return lastUpdate;
    }
    /**
     * @param {?} events
     * @param {?} date
     * @return {?}
     */
    generateEventView(events, date) {
        date = new Date(date);
        return {
            day: DateUtils.dayOfYear(date),
            display: this.getDisplayByDay(date.getDay()),
            dayStr: this.formatDateStr(date),
            isToday: DateUtils.dateIsToday(date),
            date: date,
            items: events
        };
    }
    /**
     * @param {?=} day
     * @return {?}
     */
    getDisplayByDay(day) {
        return this.resources.days[day];
    }
    /**
     * @param {?} members
     * @param {?=} role
     * @return {?}
     */
    buildMembers(members, role) {
        members.forEach((/**
         * @param {?} member
         * @return {?}
         */
        (member) => member.role = role || CalAttendeeRole.MEMBER));
        return members;
    }
    /**
     * @param {?} attendees
     * @return {?}
     */
    groupAttendeeByOrgan(attendees) {
        /** @type {?} */
        let data = {};
        attendees.forEach((/**
         * @param {?} attendee
         * @return {?}
         */
        function (attendee) {
            data[attendee.organId] = data[attendee.organId] || [];
            data[attendee.organId].push(attendee);
        }));
        return data;
    }
    /**
     * @param {?} event
     * @param {?} attendees
     * @param {?} ownOrganId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    groupAttendeeByOrganRole(event, attendees, ownOrganId, callback, options) {
        if (attendees.length < 1) {
            return callback([]);
        }
        /** @type {?} */
        let groupByOrgans = this.groupAttendeeByOrgan(attendees);
        /** @type {?} */
        let organIds = Object.keys(groupByOrgans);
        if (organIds.indexOf(ownOrganId) < 0) {
            organIds.push(ownOrganId);
        }
        this.loadRoles(organIds, (/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            let internalUpdate = !event.uuid || this.canUpdateAttendee(event, ownOrganId);
            /** @type {?} */
            let data = [];
            Object.keys(groupByOrgans).forEach((/**
             * @param {?} organId
             * @return {?}
             */
            (organId) => {
                /** @type {?} */
                let isMyOrgan = organId === ownOrganId;
                /** @type {?} */
                let externalUpdate = isMyOrgan && this.canExternalUpdateEvent(event, organId);
                /** @type {?} */
                let attendees = groupByOrgans[organId];
                data.push({
                    organId: organId,
                    organName: this.getOrgNameById(organId),
                    internalUpdate: internalUpdate,
                    externalUpdate: externalUpdate,
                    updateAttendee: internalUpdate || externalUpdate,
                    attendees: attendees,
                    canRemoveAttendee: (/**
                     * @param {?} attendee
                     * @return {?}
                     */
                    function (attendee) {
                        return internalUpdate || (externalUpdate && attendee.type !== CalAttendeeType.ORGAN);
                    }),
                    canUpdateState: (/**
                     * @param {?} attendee
                     * @return {?}
                     */
                    (attendee) => {
                        return attendee.readTime !== undefined && isMyOrgan && this.attendeeIsMe(attendee);
                    })
                });
            }));
            callback(data);
        }), options);
    }
    /**
     * @param {?=} item
     * @return {?}
     */
    buildMemberStr(item) {
        item.members = item.members || [];
        return item.members.map((/**
         * @param {?} item
         * @return {?}
         */
        (item) => item.name)).join(',');
    }
    /**
     * @param {?} members
     * @return {?}
     */
    buildMemberShow(members) {
        return (members || []).map((/**
         * @param {?} item
         * @return {?}
         */
        (item) => item.name)).join(';');
    }
    /**
     * @param {?} data
     * @param {?=} event
     * @return {?}
     */
    modifyDataUpdate(data, event) {
        /** @type {?} */
        let attendeeJson = [];
        // Remove all members and rebuild
        if (data.attendee) {
            /** @type {?} */
            let newAttendees = data.attendee.filter((/**
             * @param {?} attendee
             * @return {?}
             */
            function (attendee) {
                return attendee.readTime === undefined;
            }));
            if (event && event.uuid) {
                /** @type {?} */
                let remainAttendees = data.attendee.filter((/**
                 * @param {?} attendee
                 * @return {?}
                 */
                function (attendee) {
                    return attendee.readTime !== undefined;
                }));
                event.members.forEach((/**
                 * @param {?} member
                 * @return {?}
                 */
                (member) => {
                    if (member.role !== CalAttendeeRole.CREATOR) {
                        if (remainAttendees.indexOf(member) < 0) {
                            attendeeJson.push(this.attendeeRemoveJson(member));
                        }
                    }
                }));
            }
            newAttendees.forEach((/**
             * @param {?} attendee
             * @return {?}
             */
            (attendee) => attendeeJson.push(this.attendeeAddJson(attendee))));
        }
        // Secure text update
        if (data.subject) {
            data.subject = this.formatTextUpdate(data.subject);
        }
        if (data.location) {
            data.location = this.formatTextUpdate(data.location);
        }
        data.attendee = JSON.stringify(attendeeJson);
        // Personal calendar
        if (!this.getConfigValue('PERSONAL_CAN_PUBLISH') && data.type === CalSubType.PERSONAL) {
            data.mode = CalMode.CREATE;
        }
        if (data.type === CalSubType.MEET_PEOPLE || data.type === CalSubType.PERSONAL) {
            data._subType = data.type;
            data.type = data.type.split(';')[0];
        }
        else {
            data._subType = '';
        }
        return data;
    }
    /**
     * @param {?} attendee
     * @return {?}
     */
    attendeeAddJson(attendee) {
        /** @type {?} */
        let obj = {
            organId: attendee.organId,
            organName: attendee.organName,
            role: attendee.role || CalAttendeeRole.MEMBER
        };
        switch (attendee.type) {
            case CalAttendeeType.USER:
                obj.members = attendee.username + ':' + attendee.fullname;
                obj.display = attendee.display;
                break;
            case CalAttendeeType.ORGAN:
                if (this.getConfigValue('EVENT_BETWEEN_UNIT')) {
                    obj.firms = attendee.organId + ':' + attendee.organName;
                }
                break;
            default:
                obj.alias = attendee.alias || attendee.organName;
        }
        return obj;
    }
    /**
     * @param {?} attendee
     * @return {?}
     */
    attendeeRemoveJson(attendee) {
        /** @type {?} */
        let obj = {
            organId: attendee.organId
        };
        switch (attendee.type) {
            case CalAttendeeType.USER:
                obj.members = attendee.username + ':REMOVE';
                break;
            case CalAttendeeType.ORGAN:
                obj.firms = attendee.organId + ':REMOVE';
                break;
            default:
                obj.alias = attendee.alias + ':REMOVE';
        }
        return obj;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    _convertOrg(data) {
        data = data || [];
        for (let i = 0; i < data.length; i++) {
            /** @type {?} */
            let item = data[i];
            item.firmName = item.orgName || item.name;
            item.firmPrefix = item.uriPrefix || item.prefix;
            item.firmUUID = item.organId || item.organiId || '';
            this.indexOrgSearch(item);
        }
        // sort by name
        data.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        (a, b) => {
            return a.firmName.localeCompare(b.firmName);
        }));
        return data;
    }
    /**
     * @param {?} objectId
     * @return {?}
     */
    viewEventOnModal(objectId) {
        if (this._mapEvents[objectId]) {
            iNet.plugins.ready((/**
             * @param {?} plugins
             * @return {?}
             */
            (plugins) => {
                if (plugins.xcalendar) {
                    plugins.xcalendar.openView(this._mapEvents[objectId]);
                }
            }));
        }
    }
    // Save and get event
    /**
     * @param {?} objectId
     * @return {?}
     */
    getEventByObjectId(objectId) {
        return this._mapEvents[objectId];
    }
    /**
     * @param {?=} minutes
     * @return {?}
     */
    getTextForMinutes(minutes) {
        minutes = parseInt(minutes) || 1;
        if (minutes >= 60) {
            return Math.round(minutes / 60) + ' ' + this.resources.hour;
        }
        else {
            return minutes + ' ' + this.resources.minute;
        }
    }
    /**
     * @param {?} events
     * @return {?}
     */
    groupEventBySession(events) {
        /** @type {?} */
        let data = {
            allday: [],
            morning: [],
            afternoon: [],
            evening: []
        };
        events.forEach((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            if (item.allDay) {
                data.allday.push(item);
            }
            else if (item.from.getHours() < 12) {
                data.morning.push(item);
            }
            else if (item.from.getHours() < 18) {
                data.afternoon.push(item);
            }
            else {
                data.evening.push(item);
            }
        }));
        for (let k in data) {
            this.sortEvents(data[k]);
        }
        return data;
    }
    /**
     * @param {?} events
     * @param {?} persons
     * @param {?} date
     * @return {?}
     */
    groupEventByPerson(events, persons, date) {
        /** @type {?} */
        let data = {
            persons: [],
            segEvents: this.generateEventWeekView(events, date)
        };
        /** @type {?} */
        let eventByDays = data.segEvents.map((/**
         * @param {?} segEvent
         * @return {?}
         */
        function (segEvent) {
            return segEvent.items;
        }));
        persons.forEach((/**
         * @param {?} person
         * @return {?}
         */
        (person) => {
            /** @type {?} */
            let eventDays = eventByDays.map((/**
             * @param {?} events
             * @return {?}
             */
            (events) => events.filter((/**
             * @param {?} event
             * @return {?}
             */
            event => this.memberIsAttendee(event, person.name)))));
            data.persons.push({
                person: person,
                eventDays: eventDays
            });
        }));
        return data;
    }
    /**
     * @param {?} events
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    groupEventByLeaders(events, organId, callback, options) {
        this.loadLeaders(organId, (/**
         * @param {?} leaders
         * @return {?}
         */
        (leaders) => {
            /** @type {?} */
            let dataLeaders = [];
            /** @type {?} */
            let otherEvents = [];
            leaders.forEach((/**
             * @param {?} leader
             * @return {?}
             */
            (leader) => dataLeaders.push({
                username: leader.name,
                fullname: leader.value,
                events: []
            })));
            events.forEach((/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                /** @type {?} */
                let isFound;
                for (let i = 0; i < dataLeaders.length; i++) {
                    /** @type {?} */
                    let item = dataLeaders[i];
                    if (this.memberIsAttendee(event, item.username)) {
                        isFound = true;
                        item.events.push(event);
                    }
                }
                if (!isFound) {
                    otherEvents.push(event);
                }
            }));
            callback(dataLeaders, otherEvents);
        }), options, true);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    meIsCreator(event) {
        for (let i = event.members.length - 1; i >= 0; i--) {
            if (event.members[i].role === CalAttendeeRole.CREATOR && this.attendeeIsMe(event.members[i])) {
                return true;
            }
        }
        return false;
    }
    /**
     * @param {?} attendde
     * @return {?}
     */
    attendeeIsMe(attendde) {
        return this.usernameIsMe(attendde.username);
    }
    /**
     * @param {?} attendde
     * @return {?}
     */
    attendeeDisplay(attendde) {
        return attendde.alias || attendde.display || attendde.fullname || attendde.username || attendde.organName;
    }
    /**
     * @param {?} attendde
     * @return {?}
     */
    attendeeValue(attendde) {
        return attendde.alias || attendde.username || attendde.organId;
    }
    /**
     * @param {?} text
     * @return {?}
     */
    formatTextDisplay(text) {
        /** @type {?} */
        let display = this.formatTextUpdate(text);
        // break line
        display = display.replace(/\n/g, '<br>');
        return display;
    }
    /**
     * @param {?} text
     * @return {?}
     */
    formatTextUpdate(text) {
        /** @type {?} */
        let $fragment = $('<div>').html(text);
        $fragment.find('style,script').remove();
        return $fragment.html();
    }
    /**
     * @param {?} event
     * @param {?} member
     * @return {?}
     */
    memberIsAttendee(event, member) {
        // for (let i = 0; i < event.members.length; i++) {
        //     if (event.members[i].role !== CalAttendeeRole.CREATOR && event.members[i].username === member) {
        //         return true;
        //     }
        // }
        for (let i = 0; i < event.members.length; i++) {
            if (event.members[i].username === member) {
                return true;
            }
        }
    }
    /**
     * @param {?} event
     * @param {?} username
     * @return {?}
     */
    getAttendeeByUsername(event, username) {
        for (let i = 0; i < event.members.length; i++) {
            if (event.members[i].username === username) {
                return event.members[i];
            }
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    getAttendeeIsMe(event) {
        for (let i = 0; i < event.members.length; i++) {
            /** @type {?} */
            let member = event.members[i];
            if (member.role !== CalAttendeeRole.CREATOR && this.usernameIsMe(member.username)) {
                return member;
            }
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    eventIsViewed(event) {
        /** @type {?} */
        let attendee = this.getAttendeeIsMe(event);
        return !attendee || attendee.readTime;
    }
    /**
     * @param {?} username
     * @return {?}
     */
    usernameIsMe(username) {
        return username === iNet.usercode;
    }
    /**
     * @param {?} data
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    sendRequest(data, options, cacheResponse) {
        /** @type {?} */
        let option = this._buildAjaxOptions(data, options);
        if (cacheResponse) {
            /** @type {?} */
            let keyCache = option.url;
            if (option.data && Object.keys(option.data).length > 0) {
                keyCache += JSON.stringify(option.data);
            }
            /** @type {?} */
            let dataCached = this.cacheDataTemp(keyCache);
            if (dataCached) {
                return option.success(dataCached);
            }
            this.addQueue(keyCache, (/**
             * @param {?} resolve
             * @return {?}
             */
            (resolve) => {
                /** @type {?} */
                let _option = $.extend({}, option);
                _option.success = (/**
                 * @param {?} data
                 * @return {?}
                 */
                (data) => {
                    this.cacheDataTemp(keyCache, data);
                    resolve(data);
                });
                _option.error = (/**
                 * @param {?} error
                 * @return {?}
                 */
                (error) => {
                    resolve(null, error);
                });
                AjaxAPI.sendRequest(_option);
            }), (/**
             * @param {?} data
             * @param {?} error
             * @return {?}
             */
            (data, error) => {
                if (error) {
                    option.error && option.error(error);
                }
                else {
                    option.success && option.success(data);
                }
            }));
        }
        else {
            return AjaxAPI.sendRequest(option);
        }
    }
    /**
     * @param {?} data
     * @param {?=} options
     * @return {?}
     */
    postForm(data, options) {
        return AjaxAPI.postForm(this._buildAjaxOptions(data, options));
    }
    /**
     * @param {?} data
     * @param {?=} options
     * @return {?}
     */
    _buildAjaxOptions(data, options) {
        /** @type {?} */
        let option = $.extend({}, data, options);
        /** @type {?} */
        let orgid = option.data && option.data.orgid || this.getOrgId();
        /** @type {?} */
        let org = this.getOrgById(orgid);
        /** @type {?} */
        let isLocalFirm = org && org.firmPrefix === iNet.path.slice(1);
        /** @type {?} */
        let isUseOrgId = !option.removeOrgId && (isLocalFirm || option.useOrgId);
        if (isUseOrgId) {
            option.data = option.data || {};
            option.data.orgid = orgid;
        }
        else if (option.data) {
            delete option.data.orgid;
        }
        option.url = this.getUrlByOrgId(option.url, orgid, option.domain);
        return option;
    }
    /**
     * @return {?}
     */
    _getOrganOnInit() {
        /** @type {?} */
        let organ;
        /** @type {?} */
        let params = this.getQueryParams();
        // Get by url query
        if (params['orgid']) {
            this.setOrgId(params['orgid']);
            organ = this.getOrgById(params['orgid']);
        }
        // get organ by firm prefix
        if (!organ && iNet.firmPrefix !== 'smartcloud') {
            organ = this.getOrgByPrefix(iNet.firmPrefix);
        }
        // Get organ active recently
        if (!organ) {
            /** @type {?} */
            let organId = this.getOrgId() || iNet.organId;
            organ = this.getOrgById(organId);
        }
        // Get organ local
        if (!organ) {
            organ = this.getOrgByPrefix(iNet.path.slice(1));
        }
        // get organ first
        if (!organ) {
            organ = this._myOrgans && this._myOrgans[0] || this._organs && this._organs[0];
        }
        if (organ) {
            this.setOrgId(organ.firmUUID);
        }
        return organ;
    }
    /**
     * @param {?} cars
     * @param {?} events
     * @return {?}
     */
    groupEventByCar(cars, events) {
        /** @type {?} */
        let data = [];
        cars.forEach((/**
         * @param {?} car
         * @return {?}
         */
        function (car) {
            data.push({
                car: car,
                events: []
            });
        }));
        events.forEach((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            /** @type {?} */
            let carData = data[0];
            if (event._carUUID) {
                for (let i = 0; i < data.length; i++) {
                    if (data[i].car.uuid === event._carUUID) {
                        carData = data[i];
                    }
                }
            }
            carData.events.push(event);
        }));
        return data;
    }
    /**
     * @param {?} events
     * @return {?}
     */
    filterMeetPeopleEvents(events) {
        return events.filter((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            return event._subType === CalSubType.MEET_PEOPLE;
        }));
    }
    /**
     * @param {?} executor
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    _loadOnDifferYear(executor, params, callback, options) {
        if (!params.from || !params.to || params.from <= params.to) {
            executor.apply(this, [params, callback, options]);
        }
        else {
            // Load in differ year
            /** @type {?} */
            let promises = [];
            /** @type {?} */
            let endOfYear = new Date(params.year + 1, 0);
            endOfYear.setTime(endOfYear.getTime() - 1);
            /** @type {?} */
            let params1 = $.extend({}, params);
            /** @type {?} */
            let options1 = $.extend({}, options);
            params1.to = DateUtils.dayOfYear(endOfYear);
            promises.push(new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            (resolve) => {
                executor.apply(this, [params1, resolve, options1]);
            })));
            /** @type {?} */
            let params2 = $.extend({}, params);
            /** @type {?} */
            let options2 = $.extend({}, options);
            params2.year = params.year + 1;
            params2.from = 1;
            promises.push(new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            (resolve) => {
                executor.apply(this, [params2, resolve, options2]);
            })));
            Promise.all(promises).then((/**
             * @param {?} values
             * @return {?}
             */
            (values) => {
                /** @type {?} */
                let events = [].concat.apply([], values);
                this.removeEventDuplicate(events);
                callback(events);
            }));
        }
    }
    /**
     * @param {?} params
     * @param {?} date
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    loadMonthEvents(params, date, callback, options) {
        /** @type {?} */
        let from;
        /** @type {?} */
        let to;
        date = new Date(date);
        // Start month
        date.setDate(1);
        // Start date of week
        from = DateUtils.getDateStartWeek(date);
        // End month
        date.setMonth(date.getMonth() + 1);
        date.setTime(date.getTime() - 1);
        // End date of week
        to = DateUtils.getDateEndWeek(date);
        params = $.extend({}, params, this.getRangeParams(from, to));
        if (this.isMyOrg(params.orgid)) {
            this.getListEvents(params, callback, options);
        }
        else {
            this.loadPublishEvents(params, callback, options);
        }
    }
    /**
     * @param {?} calEvent
     * @param {?} domain
     * @return {?}
     */
    getExportUrl(calEvent, domain) {
        if (calEvent && calEvent.uuid && calEvent._templateId) {
            return this.getUrlByOrgId('calbuilder/invitation/export', calEvent.firmUUID, domain) +
                '?templateID=' + calEvent._templateId +
                '&element=' + calEvent.uuid;
        }
    }
    /**
     * @param {?} date
     * @return {?}
     */
    getWeekParams(date) {
        /** @type {?} */
        let weekRange = DateUtils.getWeekRange(date);
        return this.getRangeParams(weekRange.startWeek, weekRange.endWeek);
    }
    /**
     * @param {?} from
     * @param {?} to
     * @return {?}
     */
    getRangeParams(from, to) {
        return {
            year: from.getFullYear(),
            from: DateUtils.dayOfYear(from),
            to: DateUtils.dayOfYear(to)
        };
    }
    /**
     * @param {?} date
     * @param {?} weekParams
     * @return {?}
     */
    dateEqualWeekParams(date, weekParams) {
        /** @type {?} */
        let params = this.getWeekParams(date);
        return date && weekParams && params.year === weekParams.year &&
            params.from === weekParams.from &&
            params.to === weekParams.to;
    }
    /**
     * @param {?} event
     * @param {?=} params
     * @return {?}
     */
    eventIsOnRange(event, params) {
        return (params.startTime >= event.startTime && params.startTime < event.toTime) ||
            (params.toTime <= event.toTime && params.toTime > event.startTime) ||
            (params.startTime <= event.startTime && params.toTime >= event.toTime);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    verifyDuplicateEvents(params, callback, options) {
        /** @type {?} */
        let duplicateEvents = [];
        this.loadPublishEvents({
            year: params.year,
            day: params.day,
            orgid: params.orgid || this.getOrgId()
        }, (/**
         * @param {?} events
         * @return {?}
         */
        (events) => {
            /** @type {?} */
            let allEvents = [];
            if (this.catIsCar(params.category)) {
                events = events.filter((/**
                 * @param {?} event
                 * @return {?}
                 */
                (event) => this.catIsCar(event.category)));
            }
            else {
                events = events.filter((/**
                 * @param {?} event
                 * @return {?}
                 */
                (event) => !this.catIsCar(event.category)));
            }
            /** @type {?} */
            const detector = (/**
             * @param {?} username
             * @return {?}
             */
            (username) => {
                /** @type {?} */
                let obj = {
                    username: username,
                    events: []
                };
                events.forEach((/**
                 * @param {?} event
                 * @return {?}
                 */
                (event) => {
                    /** @type {?} */
                    let member = this.getAttendeeByUsername(event, username);
                    if (member) {
                        obj.fullname = member.fullname;
                        obj.events.push(event);
                        allEvents.push(event);
                    }
                }));
                if (obj.events.length > 0) {
                    duplicateEvents.push(obj);
                }
            });
            events = events.filter((/**
             * @param {?} event
             * @return {?}
             */
            (event) => {
                return event.mode === CalMode.PUBLISH && event.uuid !== params.element &&
                    this.eventIsOnRange(event, params);
            }));
            if (params.attendee) {
                params.attendee.forEach((/**
                 * @param {?} attendee
                 * @return {?}
                 */
                (attendee) => detector(attendee.username)));
            }
            // Todo: creator is attendee
            /** @type {?} */
            let isExist = false;
            for (let i = 0; i < duplicateEvents.length; i++) {
                if (duplicateEvents[i].username === iNet.usercode) {
                    isExist = true;
                    break;
                }
            }
            if (!isExist) {
                detector(iNet.usercode);
            }
            duplicateEvents['allEvents'] = this.removeEventDuplicate(allEvents);
            callback(duplicateEvents);
        }), options);
    }
    /**
     * @param {?} params
     * @param {?} event
     * @return {?}
     */
    needVerifyDuplicate(params, event) {
        /** @type {?} */
        var isEventPublished = event.mode === CalMode.PUBLISH;
        /** @type {?} */
        var actionPublish = !isEventPublished && Number(params.mode) === CalMode.PUBLISH;
        /** @type {?} */
        var isTimeChanged = event.uuid && ((params.startTime && event.startTime !== params.startTime) ||
            (params.toTime && event.toTime !== params.toTime) ||
            (params.day && event.day !== params.day)) && isEventPublished;
        return actionPublish || isTimeChanged;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    subjectDisplayWithMaxLength(event) {
        // Subject max length
        /** @type {?} */
        var maxLength = Number(this.getConfigValue('SUBJECT_DISPLAY_MAX_LENGTH')) || 0;
        // Display view more
        if (maxLength > 0 && maxLength < event.subject.length) {
            /** @type {?} */
            var subjectDisplay = event.subject.substr(0, maxLength);
            subjectDisplay = this.formatTextDisplay(subjectDisplay);
            subjectDisplay += '...';
            return this.formatTextDisplay(subjectDisplay);
        }
        return event.subjectDisplay;
    }
    // SubFirm Dictionary Utils
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    dictCreate(data, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/create',
            data: data,
            success: callback
        }, options);
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    dictUpdate(data, callback, options) {
        data.dictID = data.dictID || data.uuid;
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/update',
            data: data,
            success: callback
        }, options);
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    dictRemove(data, callback, options) {
        data.dictID = data.dictID || data.uuid;
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/remove',
            data: data,
            success: callback
        }, options);
    }
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    dictList(params, callback, options, cacheResponse) {
        // reference: required
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/list',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            (data) => callback(data && data.items || [])),
            error: (/**
             * @return {?}
             */
            () => callback([]))
        }, options, cacheResponse);
    }
    // Leader
    /**
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    loadLeaders(organId, callback, options, cacheResponse) {
        /** @type {?} */
        let params = {
            orgid: organId || this.getOrgId(),
            reference: LEADER_REFERENCE
        };
        this.dictList(params, (/**
         * @param {?} leaders
         * @return {?}
         */
        function (leaders) {
            leaders.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                return b.order - a.order;
            }));
            callback(leaders);
        }), options, cacheResponse);
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    leaderCreateOrUpdate(data, callback, options) {
        if (!data.uuid) {
            data.reference = LEADER_REFERENCE;
            this.dictCreate(data, callback, options);
        }
        else {
            this.dictUpdate(data, callback, options);
        }
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    leaderRemove(data, callback, options) {
        this.dictRemove(data, callback, options);
    }
}
if (false) {
    /** @type {?} */
    CalBuilder.prototype.isAuth;
    /** @type {?} */
    CalBuilder.prototype.orgid;
    /** @type {?} */
    CalBuilder.prototype.TIME_NEW_UPDATE;
    /** @type {?} */
    CalBuilder.prototype.roleUrl;
    /** @type {?} */
    CalBuilder.prototype.MODE;
    /** @type {?} */
    CalBuilder.prototype.ROLE;
    /** @type {?} */
    CalBuilder.prototype.calSubType;
    /** @type {?} */
    CalBuilder.prototype.calCategory;
    /** @type {?} */
    CalBuilder.prototype.vnToLatin;
    /** @type {?} */
    CalBuilder.prototype.dateUtils;
    /** @type {?} */
    CalBuilder.prototype.calAttendeeType;
    /** @type {?} */
    CalBuilder.prototype.calAttendeeState;
    /** @type {?} */
    CalBuilder.prototype.calType;
    /** @type {?} */
    CalBuilder.prototype.subFirmDic;
    /** @type {?} */
    CalBuilder.prototype.carUtils;
    /** @type {?} */
    CalBuilder.prototype.configs;
    /** @type {?} */
    CalBuilder.prototype.resources;
    /** @type {?} */
    CalBuilder.prototype._userRoles;
    /** @type {?} */
    CalBuilder.prototype._mapEvents;
    /** @type {?} */
    CalBuilder.prototype._cacheStore;
    /** @type {?} */
    CalBuilder.prototype._queue;
    /** @type {?} */
    CalBuilder.prototype._organs;
    /** @type {?} */
    CalBuilder.prototype._myOrgans;
    /** @type {?} */
    CalBuilder.prototype.typeIsDepartment;
    /** @type {?} */
    CalBuilder.prototype.typeIsCommunity;
    /** @type {?} */
    CalBuilder.prototype.typeIsOrganization;
    /** @type {?} */
    CalBuilder.prototype.catIsInvitation;
    /** @type {?} */
    CalBuilder.prototype.catIsPersonal;
    /** @type {?} */
    CalBuilder.prototype.catIsCar;
    /** @type {?} */
    CalBuilder.prototype.catIsImportant;
    /** @type {?} */
    CalBuilder.prototype.formatDateStr;
    /** @type {?} */
    CalBuilder.prototype.formatTimeStr;
    /** @type {?} */
    CalBuilder.prototype.formatDateTimeStr;
    /** @type {?} */
    CalBuilder.prototype.formatDate;
    /** @type {?} */
    CalBuilder.prototype.getEventStatus;
    /** @type {?} */
    CalBuilder.prototype.isCreate;
    /** @type {?} */
    CalBuilder.prototype.isPublish;
    /** @type {?} */
    CalBuilder.prototype.isSent;
    /** @type {?} */
    CalBuilder.prototype.isCancel;
    /** @type {?} */
    CalBuilder.prototype.isOther;
}
