// inet-calendar/esm2015/src/utils/CalendarCar.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class CalendarCar {
    /**
     * @param {?} subFirmDict
     */
    constructor(subFirmDict) {
        this.carDefault = {
            name: 'Mặc định',
            uuid: ''
        };
        this.carRef = 'CAL_BUILDER_CAR';
        this.driverRef = 'CAL_BUILDER_DRIVER';
        this.subFirmDict = subFirmDict;
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    loadCars(callback) {
        this._loadList(this.carRef, callback);
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    createOrUpdateCars(data, callback) {
        if (!data.uuid) {
            data.name = this.carRef;
            data.reference = this.carRef;
            this.subFirmDict.create(data, (/**
             * @param {?} data
             * @return {?}
             */
            (data) => this.parseJsonValue(data, callback)));
        }
        else {
            this.subFirmDict.update(data, (/**
             * @param {?} data
             * @return {?}
             */
            (data) => this.parseJsonValue(data, callback)));
        }
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    loadDrivers(callback) {
        this._loadList(this.driverRef, callback);
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    createOrUpdateDrivers(data, callback) {
        if (!data.uuid) {
            data.name = this.driverRef;
            data.reference = this.driverRef;
            this.subFirmDict.create(data, (/**
             * @param {?} data
             * @return {?}
             */
            (data) => this.parseJsonValue(data, callback)));
        }
        else {
            this.subFirmDict.update(data, (/**
             * @param {?} data
             * @return {?}
             */
            (data) => this.parseJsonValue(data, callback)));
        }
    }
    /**
     * @param {?} ref
     * @param {?} callback
     * @return {?}
     */
    _loadList(ref, callback) {
        this.subFirmDict.list(ref, (/**
         * @param {?} results
         * @return {?}
         */
        (results) => this.parseJsonValue(results[0], callback)));
    }
    /**
     * @param {?} data
     * @param {?} callback
     * @return {?}
     */
    parseJsonValue(data, callback) {
        /** @type {?} */
        let values = [];
        try {
            values = JSON.parse(data.value);
        }
        catch (e) { }
        callback(values, data);
    }
}
if (false) {
    /** @type {?} */
    CalendarCar.prototype.carDefault;
    /** @type {?} */
    CalendarCar.prototype.carRef;
    /** @type {?} */
    CalendarCar.prototype.driverRef;
    /** @type {?} */
    CalendarCar.prototype.subFirmDict;
}
