// inet-calendar/esm2015/src/calendar-today/calendar-today.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter } from '@angular/core';
import { CalendarService } from "../calendar.service";
import { DateUtils } from "../utils/common/DateUtils";
export class CalendarTodayComponent {
    /**
     * @param {?} calendarService
     */
    constructor(calendarService) {
        this.calendarService = calendarService;
        this.onChange = new EventEmitter();
        this.events = [];
        this.colors = [
            // '#438EB9',
            '#007bff',
            '#3CA54A',
            '#FF9800',
            '#EF5B49',
            '#77901B',
            '#7D5392',
            '#808080'
        ];
        this._date = new Date();
        this.calendarService.ready((/**
         * @return {?}
         */
        () => this.loadTodayEvents()));
    }
    /**
     * @return {?}
     */
    loadTodayEvents() {
        if (!this._date) {
            return;
        }
        this.calendarService.loadEvents({
            day: DateUtils.dayOfYear(this._date),
            year: this._date.getFullYear()
        }, (/**
         * @param {?} events
         * @return {?}
         */
        (events) => {
            events.forEach((/**
             * @param {?} e
             * @return {?}
             */
            (e) => this._initEventStatus(e)));
            this.events = events.slice(0, 3); // Maximum 3 events
        }));
    }
    /**
     * @param {?} event
     * @param {?} jsEvent
     * @param {?} targetEl
     * @return {?}
     */
    onViewEvent(event, jsEvent, targetEl) {
        this.calendarService.viewEvent(event, jsEvent, targetEl, (/**
         * @return {?}
         */
        () => {
            this.loadTodayEvents();
            this.onChange.emit(true);
        }));
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    _initEventStatus(event) {
        /** @type {?} */
        let now = new Date();
        event.progressing = event.start < now && event.end > now;
        event.ended = event.end < now;
        event.pending = event.start > now;
        if (event.progressing) {
            event.statusText = 'Đang diễn ra';
            event.statusColor = this.colors[0];
        }
        else if (event.pending) {
            event.statusText = 'Chưa diễn ra';
            event.statusColor = this.colors[5];
        }
        else {
            event.statusText = 'Kết thúc';
            event.statusColor = this.colors[6];
        }
        event.style = {
            'background-color': this.colors[event.mode]
        };
    }
}
CalendarTodayComponent.decorators = [
    { type: Component, args: [{
                selector: 'calendar-widget-today',
                template: "<div #viewToggle class=\"calendar-body\">\n    <div *ngFor=\"let event of events\" #targetEl (click)=\"onViewEvent(event, $event, targetEl)\"\n         class=\"calendar-event\">\n        <div class=\"calendar-event__state\" [ngStyle]=\"event['style']\"></div>\n        <div class=\"calendar-event__content\">\n            <div class=\"calendar-event__title\">\n                {{event.subject}}\n                <div class=\"calendar-event__progress\" [style.color]=\"event['statusColor']\">{{event['statusText']}}</div>\n            </div>\n            <div class=\"calendar-event__text\">\n                <i class=\"fa fa-clock-o calendar-event__icon\"></i>\n                {{event['hourStr']}}\n            </div>\n            <div *ngIf=\"event.location\" class=\"calendar-event__text\">\n                <i class=\"fa fa-map-marker calendar-event__icon\"></i>\n                {{event.location}}\n            </div>\n        </div>\n    </div>\n</div>\n",
                styles: [".calendar-body{max-height:300px;overflow:hidden auto}.calendar-event{padding:5px 15px;color:#212121;cursor:pointer}.calendar-event:hover{background:rgba(0,0,0,.05)}.calendar-event__state{width:8px;height:8px;border-radius:50%;margin:5px 10px 0 0;float:left}.calendar-event__content{overflow:hidden}.calendar-event__title{line-height:18px;margin-bottom:3px}.calendar-event__progress{font-size:80%;text-transform:uppercase;font-weight:700}.calendar-event__text{font-size:90%;line-height:18px;color:#757575}.calendar-event__icon{width:12px}"]
            }] }
];
/** @nocollapse */
CalendarTodayComponent.ctorParameters = () => [
    { type: CalendarService }
];
CalendarTodayComponent.propDecorators = {
    onChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CalendarTodayComponent.prototype.onChange;
    /** @type {?} */
    CalendarTodayComponent.prototype.events;
    /** @type {?} */
    CalendarTodayComponent.prototype.colors;
    /**
     * @type {?}
     * @private
     */
    CalendarTodayComponent.prototype._date;
    /**
     * @type {?}
     * @private
     */
    CalendarTodayComponent.prototype.calendarService;
}
