// inet-core/esm2015/src/vertx/vertx-socket.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { VertxSocketState } from "./VertxSocket";
import { VertxSocketManager } from "./VertxSocketManager";
export class SocketService {
    constructor() {
        this.onStateChange = new Subject();
        this.onMessage = new Subject();
        this.onMemberStateChange = new Subject();
        this.init();
    }
    /**
     * @return {?}
     */
    get readyState() {
        return this.socket && this.socket.readyState || VertxSocketState.UNSET;
    }
    /**
     * @return {?}
     */
    get status() {
        return this.readyState;
    }
    /**
     * @return {?}
     */
    get connecting() {
        return this.socket ? this.socket.connecting : false;
    }
    /**
     * @return {?}
     */
    get connected() {
        return this.socket ? this.socket.connected : false;
    }
    /**
     * @return {?}
     */
    get closed() {
        return this.socket ? this.socket.closed : false;
    }
    /**
     * @return {?}
     */
    get reconnecting() {
        return this.socket ? this.socket.reconnecting : false;
    }
    /**
     * @return {?}
     */
    get unavailable() {
        return this.socket ? this.socket.unavailable : false;
    }
    /**
     * @return {?}
     */
    init() {
        if (!this.socket) {
            this.socket = new VertxSocketManager();
            this.socket.onStateChange.subscribe((/**
             * @return {?}
             */
            () => this.onStateChange.next(this.readyState)));
            this.socket.onMessage.subscribe((/**
             * @param {?} message
             * @return {?}
             */
            (message) => this.onMessage.next(message)));
            this.socket.onMessageOnline.subscribe((/**
             * @param {?} message
             * @return {?}
             */
            (message) => this.onMemberStateChange.next(message)));
        }
    }
    /**
     * @return {?}
     */
    connect() {
        this.init();
    }
    /**
     * @return {?}
     */
    disconnect() {
        if (this.socket) {
            this.socket.close();
            this.socket = null;
        }
    }
    /**
     * @param {?} address
     * @param {?} message
     * @return {?}
     */
    publish(address, message) {
        return this.send(address, message);
    }
    /**
     * @param {?} address
     * @param {?} message
     * @return {?}
     */
    send(address, message) {
        this.init();
        return this.socket.send(address, message);
    }
    /**
     * @param {?} app
     * @param {?} message
     * @return {?}
     */
    createMessagePackage(app, message) {
        return {
            message: {
                application: app,
                alert: message,
                sender: iNet.username,
                content: message
            },
            application: app,
            sender: iNet.username,
            sent: new Date().getTime()
        };
    }
    // @deprecated
    /**
     * @param {?} groupCode
     * @param {?} message
     * @return {?}
     */
    sendGroup(groupCode, message) {
        // Prepend # into groupCode
        this.send('#' + groupCode, message);
    }
}
SocketService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SocketService.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    SocketService.prototype.onStateChange;
    /** @type {?} */
    SocketService.prototype.onMessage;
    /** @type {?} */
    SocketService.prototype.onMemberStateChange;
    /**
     * @type {?}
     * @private
     */
    SocketService.prototype.socket;
}
