// inet-ui/esm2015/src/notify/notify.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ViewChild } from '@angular/core';
import { HttpClientService } from "inet-core";
import { ToastrService } from "./toastr/toastr.service";
import { ToastContainerDirective } from "./toastr/toast.directive";
export class NotifyService {
    /**
     * @param {?} http
     * @param {?} toastService
     */
    constructor(http, toastService) {
        this.http = http;
        this.toastService = toastService;
        this.url = {
            list: iNet.getPUrl('social/activity/notify')
        };
        this.toastService.overlayContainer = this.toastContainer;
    }
    /**
     * @return {?}
     */
    count() {
        return this.http.getJSON(this.url.list);
    }
    /**
     * @param {?} params
     * @return {?}
     */
    getMessage(params) {
        return this.http.getJSON(this.url.list, params);
    }
    /**
     * @param {?} msg
     * @param {?=} type
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    showMessage(msg, type, title = 'Thông báo', config = {}) {
        switch (type) {
            case 'info':
                this.toastService.info(msg, title, config);
                break;
            case 'error':
                this.toastService.error(msg, title, config);
                break;
            case 'warning':
                this.toastService.warning(msg, title, config);
                break;
            case 'success':
            default:
                this.toastService.success(msg, title, config);
        }
    }
    /**
     * @return {?}
     */
    getNotifyService() {
        return this.toastService;
    }
}
NotifyService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
NotifyService.ctorParameters = () => [
    { type: HttpClientService },
    { type: ToastrService }
];
NotifyService.propDecorators = {
    toastContainer: [{ type: ViewChild, args: [ToastContainerDirective,] }]
};
if (false) {
    /** @type {?} */
    NotifyService.prototype.toastContainer;
    /**
     * @type {?}
     * @private
     */
    NotifyService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    NotifyService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    NotifyService.prototype.toastService;
}
