// inet-social/esm5/src/modal-view/modal-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewChild } from '@angular/core';
var SocialModalViewComponent = /** @class */ (function () {
    function SocialModalViewComponent() {
    }
    /**
     * @return {?}
     */
    SocialModalViewComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.$modal = $(this.viewModal.nativeElement);
        this.$modal.on('hidden.bs.modal', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            delete _this.activity;
        }));
        $(document.body).append(this.$modal);
    };
    /**
     * @return {?}
     */
    SocialModalViewComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.$modal.remove();
    };
    /**
     * @param {?} activity
     * @return {?}
     */
    SocialModalViewComponent.prototype.viewActivity = /**
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        var _this = this;
        this.activity = activity;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.$modal.modal('show');
        }), 100);
    };
    SocialModalViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-social-modal-view',
                    template: "<div #viewModal class=\"modal fade\" tabindex=\"-1\" role=\"dialog\">\n    <div class=\"modal-dialog modal-lg\" role=\"document\">\n        <div class=\"modal-content\">\n            <div class=\"modal-header\">\n                <h5 class=\"modal-title\">Th\u1EA3o lu\u1EADn</h5>\n                <button type=\"button\" class=\"close\" data-dismiss=\"modal\" aria-label=\"Close\">\n                    <span aria-hidden=\"true\">&times;</span>\n                </button>\n            </div>\n            <div class=\"modal-body\">\n                <div *ngIf=\"activity\" socialActivity [activity]=\"activity\"></div>\n            </div>\n        </div>\n    </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SocialModalViewComponent.ctorParameters = function () { return []; };
    SocialModalViewComponent.propDecorators = {
        viewModal: [{ type: ViewChild, args: ['viewModal',] }]
    };
    return SocialModalViewComponent;
}());
export { SocialModalViewComponent };
if (false) {
    /** @type {?} */
    SocialModalViewComponent.prototype.viewModal;
    /** @type {?} */
    SocialModalViewComponent.prototype.activity;
    /** @type {?} */
    SocialModalViewComponent.prototype.$modal;
}
