// inet-ui/esm2015/src/autocomplete/autocomplete-input.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, HostListener, Input, Output } from '@angular/core';
import { AutocompleteListDirective } from "./autocomplete-list.directive";
export class AutocompleteInputDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.onBackSpaceRemove = new EventEmitter();
    }
    /**
     * @return {?}
     */
    onFocus() {
        this.autoList.open();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onInput(e) {
        switch (e.which) {
            case 8:
                if (!e.target.value) {
                    this.onBackSpaceRemove.emit(e);
                }
                break;
        }
        this.autoList.onInput(e);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.hideList = this.hideList.bind(this);
        // Click out hide list
        document.addEventListener('click', this.hideList);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        document.removeEventListener('click', this.hideList);
    }
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    hideList(e) {
        if (!e.target.isSameNode(this.el.nativeElement)) {
            this.autoList.hide();
        }
    }
}
AutocompleteInputDirective.decorators = [
    { type: Directive, args: [{
                selector: '[appAutocompleteInput]'
            },] }
];
/** @nocollapse */
AutocompleteInputDirective.ctorParameters = () => [
    { type: ElementRef }
];
AutocompleteInputDirective.propDecorators = {
    autoList: [{ type: Input }],
    onBackSpaceRemove: [{ type: Output }],
    onFocus: [{ type: HostListener, args: ['focus',] }],
    onInput: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    AutocompleteInputDirective.prototype.autoList;
    /** @type {?} */
    AutocompleteInputDirective.prototype.onBackSpaceRemove;
    /**
     * @type {?}
     * @protected
     */
    AutocompleteInputDirective.prototype.el;
}
