// inet-ui/esm2015/src/user-profile/global-contact.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService, UserProfileService } from 'inet-core';
import { Contact } from "../model/contact";
export class GlobalContactService {
    /**
     * @param {?} http
     * @param {?} userProfileService
     */
    constructor(http, userProfileService) {
        this.http = http;
        this.userProfileService = userProfileService;
        this.url = {
            // update: iNet.getPUrl('gl/contact/update'),
            update: iNet.getPUrl('system/userprofile/update'),
            upload_photo: iNet.getPUrl('system/userprofile/photoupdate'),
            delete_photo: iNet.getPUrl('system/userprofile/delphoto'),
            update_signature: iNet.getPUrl('system/userprofile/updatesign'),
            remove_signature: iNet.getPUrl('system/userprofile/delsign'),
            view_signature_photo: iNet.getPUrl('usersignature/photo'),
            access_roles: iNet.getPUrl('cloud/subfirmrole/group')
        };
    }
    /**
     * @return {?}
     */
    getProfile() {
        return this.userProfileService.getProfile();
    }
    /**
     * @param {?} data
     * @return {?}
     */
    update(data) {
        this.showLoading();
        /** @type {?} */
        const contact = Object.assign(new Contact(), data);
        if (contact.address) {
            contact.addressStr = JSON.stringify(contact.address);
        }
        if (contact.fullDateOfBirth) {
            contact.dateOfBirth = contact.fullDateOfBirth.getTime();
        }
        delete contact.address;
        delete contact.fullName;
        return this.http.postJSON(this.url.update, contact);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    uploadPhoto(data) {
        this.http.showLoading();
        return this.http.post(this.url.upload_photo, data);
    }
    /**
     * @return {?}
     */
    deleteAvatar() {
        this.http.showLoading();
        return this.http.post(this.url.delete_photo, null);
    }
    /**
     * @return {?}
     */
    showLoading() {
        this.http.showLoading();
    }
    /**
     * @param {?} data
     * @return {?}
     */
    uploadSignaturePhoto(data) {
        this.http.showLoading();
        return this.http.post(this.url.update_signature, data);
    }
    /**
     * @return {?}
     */
    getSignatureImageUrl() {
        return this.url.view_signature_photo;
    }
    /**
     * @return {?}
     */
    getAccessRoles() {
        return this.http.postJSON(this.url.access_roles);
    }
}
GlobalContactService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
GlobalContactService.ctorParameters = () => [
    { type: HttpClientService },
    { type: UserProfileService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    GlobalContactService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    GlobalContactService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    GlobalContactService.prototype.userProfileService;
}
