// inet-ui/esm5/src/user-profile/global-contact.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService, UserProfileService } from 'inet-core';
import { Contact } from "../model/contact";
var GlobalContactService = /** @class */ (function () {
    function GlobalContactService(http, userProfileService) {
        this.http = http;
        this.userProfileService = userProfileService;
        this.url = {
            // update: iNet.getPUrl('gl/contact/update'),
            update: iNet.getPUrl('system/userprofile/update'),
            upload_photo: iNet.getPUrl('system/userprofile/photoupdate'),
            delete_photo: iNet.getPUrl('system/userprofile/delphoto'),
            update_signature: iNet.getPUrl('system/userprofile/updatesign'),
            remove_signature: iNet.getPUrl('system/userprofile/delsign'),
            view_signature_photo: iNet.getPUrl('usersignature/photo'),
            access_roles: iNet.getPUrl('cloud/subfirmrole/group')
        };
    }
    /**
     * @return {?}
     */
    GlobalContactService.prototype.getProfile = /**
     * @return {?}
     */
    function () {
        return this.userProfileService.getProfile();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    GlobalContactService.prototype.update = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.showLoading();
        /** @type {?} */
        var contact = Object.assign(new Contact(), data);
        if (contact.address) {
            contact.addressStr = JSON.stringify(contact.address);
        }
        if (contact.fullDateOfBirth) {
            contact.dateOfBirth = contact.fullDateOfBirth.getTime();
        }
        delete contact.address;
        delete contact.fullName;
        return this.http.postJSON(this.url.update, contact);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    GlobalContactService.prototype.uploadPhoto = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.http.showLoading();
        return this.http.post(this.url.upload_photo, data);
    };
    /**
     * @return {?}
     */
    GlobalContactService.prototype.deleteAvatar = /**
     * @return {?}
     */
    function () {
        this.http.showLoading();
        return this.http.post(this.url.delete_photo, null);
    };
    /**
     * @return {?}
     */
    GlobalContactService.prototype.showLoading = /**
     * @return {?}
     */
    function () {
        this.http.showLoading();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    GlobalContactService.prototype.uploadSignaturePhoto = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.http.showLoading();
        return this.http.post(this.url.update_signature, data);
    };
    /**
     * @return {?}
     */
    GlobalContactService.prototype.getSignatureImageUrl = /**
     * @return {?}
     */
    function () {
        return this.url.view_signature_photo;
    };
    /**
     * @return {?}
     */
    GlobalContactService.prototype.getAccessRoles = /**
     * @return {?}
     */
    function () {
        return this.http.postJSON(this.url.access_roles);
    };
    GlobalContactService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    GlobalContactService.ctorParameters = function () { return [
        { type: HttpClientService },
        { type: UserProfileService }
    ]; };
    return GlobalContactService;
}());
export { GlobalContactService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GlobalContactService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    GlobalContactService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    GlobalContactService.prototype.userProfileService;
}
