// inet-social/esm5/src/embed/social-embed.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { SocialActivity } from "../model/Activity";
import { SocialService } from "../social.service";
var SocialEmbedComponent = /** @class */ (function () {
    function SocialEmbedComponent(socialService) {
        this.socialService = socialService;
        this.allowComment = true;
        this.onLoad = new EventEmitter();
        this.onError = new EventEmitter();
        this.totalComments = 0;
    }
    /**
     * @return {?}
     */
    SocialEmbedComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.activity) {
            this._loaded();
        }
        else {
            this._loadActivity();
        }
    };
    /**
     * @private
     * @return {?}
     */
    SocialEmbedComponent.prototype._loadActivity = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var params;
        if (this.activityID) {
            params = {
                activity: this.activityID
            };
        }
        else if (this.contextID && this.application) {
            params = {
                contextID: this.contextID,
                application: this.application
            };
        }
        if (params) {
            this.socialService.activityLoad(params, (/**
             * @param {?} activity
             * @param {?} err
             * @return {?}
             */
            function (activity, err) {
                if (err) {
                    _this.onError.emit(err);
                }
                else {
                    if (activity.errors) {
                        _this.onError.emit(activity);
                    }
                    else {
                        _this.activity = activity;
                        _this._loaded();
                    }
                }
            }));
        }
        else {
            this.onError.emit({
                message: 'NOT_FOUND'
            });
        }
    };
    /**
     * @private
     * @return {?}
     */
    SocialEmbedComponent.prototype._loaded = /**
     * @private
     * @return {?}
     */
    function () {
        this.onLoad.emit(this.activity);
    };
    SocialEmbedComponent.decorators = [
        { type: Component, args: [{
                    selector: 'social-discussion',
                    template: "<div *ngIf=\"activity\" socialCommentList\n     [activity]=\"activity\"\n     [allowComment]=\"allowComment\"\n     (commentLoad)=\"totalComments = $event.total\"></div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SocialEmbedComponent.ctorParameters = function () { return [
        { type: SocialService }
    ]; };
    SocialEmbedComponent.propDecorators = {
        activityID: [{ type: Input }],
        contextID: [{ type: Input }],
        application: [{ type: Input }],
        activity: [{ type: Input }],
        allowComment: [{ type: Input }],
        onLoad: [{ type: Output }],
        onError: [{ type: Output }]
    };
    return SocialEmbedComponent;
}());
export { SocialEmbedComponent };
if (false) {
    /** @type {?} */
    SocialEmbedComponent.prototype.activityID;
    /** @type {?} */
    SocialEmbedComponent.prototype.contextID;
    /** @type {?} */
    SocialEmbedComponent.prototype.application;
    /** @type {?} */
    SocialEmbedComponent.prototype.activity;
    /** @type {?} */
    SocialEmbedComponent.prototype.allowComment;
    /** @type {?} */
    SocialEmbedComponent.prototype.onLoad;
    /** @type {?} */
    SocialEmbedComponent.prototype.onError;
    /** @type {?} */
    SocialEmbedComponent.prototype.totalComments;
    /**
     * @type {?}
     * @private
     */
    SocialEmbedComponent.prototype.socialService;
}
