// inet-social/esm5/src/activity-list/activity-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, Input } from '@angular/core';
import { SocialService } from '../social.service';
import { ActivityPostComponent } from '../activity-post/activity-post.component';
var ActivityListComponent = /** @class */ (function () {
    function ActivityListComponent(socialService) {
        var _this = this;
        this.socialService = socialService;
        this.postEnable = true;
        this.scroller = document;
        this.type = '';
        this.application = '';
        this.loader = (/**
         * @param {?} params
         * @param {?} callback
         * @return {?}
         */
        function (params, callback) { return _this.doLoadWall(params, callback); });
        this.activities = [];
        this.showLoading = true;
        this.hasMore = false;
        this.params = {
            pageSize: 5,
            pageNumber: 0
        };
        // Receive activity notify real time
        this._activityChange = this.socialService.activityChange.subscribe((/**
         * @param {?} activityId
         * @return {?}
         */
        function (activityId) {
            if (activityId) {
                _this.socialService.activityLoad({
                    activity: activityId
                }, (/**
                 * @param {?} activity
                 * @return {?}
                 */
                function (activity) {
                    if (activity && activity.uuid) {
                        _this._addActivity(activity, true);
                    }
                }));
            }
        }));
    }
    /**
     * @return {?}
     */
    ActivityListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initScroller();
        this.loadWall(true);
    };
    /**
     * @return {?}
     */
    ActivityListComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.removeEventScroller();
        this._activityChange.unsubscribe();
    };
    /**
     * @param {?} id
     * @return {?}
     */
    ActivityListComponent.prototype.removeActivityById = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        for (var i = 0; i < this.activities.length; i++) {
            /** @type {?} */
            var activity = this.activities[i];
            if (activity.uuid === id) {
                this.activities.splice(i, 1);
                return;
            }
        }
    };
    /**
     * @param {?=} init
     * @return {?}
     */
    ActivityListComponent.prototype.loadWall = /**
     * @param {?=} init
     * @return {?}
     */
    function (init) {
        var _this = this;
        if (init) {
            this.clearWall();
            this.params.pageNumber = 0;
            this.showLoading = true;
            if (this.type) {
                this.params.type = this.type;
            }
            if (this.application) {
                this.params.application = this.application;
            }
        }
        this._loading = true;
        this.loader(this.params, (/**
         * @param {?} data
         * @param {?} err
         * @return {?}
         */
        function (data, err) { return _this._renderWall(data); }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    ActivityListComponent.prototype.doLoadWall = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        this.socialService.activityWall(params, callback);
    };
    /**
     * @return {?}
     */
    ActivityListComponent.prototype.clearWall = /**
     * @return {?}
     */
    function () {
        this.activities.length = 0;
    };
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    ActivityListComponent.prototype._renderWall = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        if (data) {
            /** @type {?} */
            var activities = data['items'] || [];
            activities.forEach((/**
             * @param {?} activity
             * @return {?}
             */
            function (activity) { return _this._addActivity(activity); }));
            this.hasMore = data['total'] > this.activities.length;
        }
        this.showLoading = false;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this._loading = false;
            _this._scrollBottomListener();
        }), 1000);
    };
    /**
     * @param {?} params
     * @return {?}
     */
    ActivityListComponent.prototype.activityCreate = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        if (params.activity) {
            this.socialService.activityUpdate(params, this._activityPosted.bind(this));
        }
        else {
            this.socialService.activityPost(params, this._activityPosted.bind(this));
        }
    };
    /**
     * @param {?} activity
     * @return {?}
     */
    ActivityListComponent.prototype.activityUpdated = /**
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        // Replace by activity updated
        for (var i = 0; i < this.activities.length; i++) {
            if (this.activities[i].uuid === activity.uuid) {
                this.activities[i] = activity;
                return;
            }
        }
        this.activities.unshift(activity);
    };
    /**
     * @param {?} activity
     * @return {?}
     */
    ActivityListComponent.prototype.activityDeleted = /**
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        this.removeActivityById(activity.uuid);
    };
    /**
     * @return {?}
     */
    ActivityListComponent.prototype.initScroller = /**
     * @return {?}
     */
    function () {
        this._scrollBottomListener = this._scrollBottomListener.bind(this);
        if (this.scroller) {
            this.initEventScroller(this.scroller);
        }
    };
    /**
     * @return {?}
     */
    ActivityListComponent.prototype.scrollToTop = /**
     * @return {?}
     */
    function () {
        if (this.scroller) {
            this.scroller.scrollTop = 0;
        }
    };
    /**
     * @param {?} scroller
     * @return {?}
     */
    ActivityListComponent.prototype.initEventScroller = /**
     * @param {?} scroller
     * @return {?}
     */
    function (scroller) {
        this.scroller = scroller;
        this.scroller.addEventListener('scroll', this._scrollBottomListener);
    };
    /**
     * @return {?}
     */
    ActivityListComponent.prototype.removeEventScroller = /**
     * @return {?}
     */
    function () {
        if (this.scroller) {
            this.scroller.removeEventListener('scroll', this._scrollBottomListener);
        }
    };
    /**
     * @private
     * @param {?=} e
     * @return {?}
     */
    ActivityListComponent.prototype._scrollBottomListener = /**
     * @private
     * @param {?=} e
     * @return {?}
     */
    function (e) {
        if (this._loading || !this.hasMore) {
            return;
        }
        /** @type {?} */
        var scroller = e && e.target || this.scroller;
        /** @type {?} */
        var target = scroller.scrollingElement || scroller;
        if (target) {
            // Scroll bottom to load more
            /** @type {?} */
            var bottom = target.scrollHeight - target.scrollTop - target.clientHeight;
            if (bottom < 150) {
                this.params.pageNumber++;
                this.loadWall();
            }
        }
    };
    /**
     * @private
     * @param {?} activity
     * @param {?} err
     * @return {?}
     */
    ActivityListComponent.prototype._activityPosted = /**
     * @private
     * @param {?} activity
     * @param {?} err
     * @return {?}
     */
    function (activity, err) {
        if (activity && activity.uuid) {
            this.activityPost.resetForm();
            this.activityPost.hideForm(true);
            this._addActivity(activity, true);
        }
    };
    /**
     * @private
     * @param {?} activity
     * @param {?=} insertAtFirst
     * @return {?}
     */
    ActivityListComponent.prototype._addActivity = /**
     * @private
     * @param {?} activity
     * @param {?=} insertAtFirst
     * @return {?}
     */
    function (activity, insertAtFirst) {
        this.removeActivityById(activity.uuid);
        if (insertAtFirst) {
            this.activities.unshift(activity);
        }
        else {
            this.activities.push(activity);
        }
    };
    ActivityListComponent.decorators = [
        { type: Component, args: [{
                    selector: '[socialActivityList]',
                    template: "<div *ngIf=\"postEnable\" class=\"social-card\"\n     activityPost\n     #activityPost\n     [shareMemberEnable]=\"true\"\n     [linkPreviewEnable]=\"true\"\n     (onPost)=\"activityCreate($event)\">\n</div>\n\n<div class=\"social-welcome\" *ngIf=\"!showLoading && activities.length < 1\">\n    <i class=\"fa fa-bell-o\"></i>\n    <div>Ch\u01B0a c\u00F3 n\u1ED9i dung n\u00E0o</div>\n</div>\n\n<div class=\"social-loading\" *ngIf=\"showLoading\">\n    <i class=\"fa fa-spinner fa-spin\"></i>\n</div>\n\n<div *ngFor=\"let activity of activities\"\n     socialActivity\n     [activity]=\"activity\"\n     (onDeleted)=\"activityDeleted($event)\"\n     (onUpdated)=\"activityUpdated($event)\">\n</div>\n\n<div class=\"card\" *ngIf=\"hasMore\">\n    <div class=\"card-body text-center\">\n        \u0110ang t\u1EA3i d\u1EEF li\u1EC7u..\n    </div>\n</div>\n",
                    styles: [".social-block{display:block;margin-top:10px}.social-card{position:relative;border-radius:2px;padding:15px;background:#fff;font-size:14px;margin-bottom:15px}.social-mark{position:absolute;top:-2px;left:-2px;padding:7px 9px;color:#fff}.social-mark:before{position:absolute;z-index:1}.social-mark:after{content:'';position:absolute;top:0;left:0;border-bottom:45px solid transparent;border-left:45px solid #faa807}.social-welcome{text-align:center;color:rgba(0,0,0,.54);margin-top:80px}.social-welcome i{font-size:60px;margin-bottom:20px;width:100px;height:100px;line-height:100px;border-radius:50%;background:rgba(0,0,0,.1)}.social-loading{text-align:center;margin-top:40px;color:rgba(0,0,0,.54)}.social-loading i{font-size:40px}.activity-column-active{margin:-7.5px}.activity-column-active .activity-column{float:left;width:50%;padding:0 7.5px}"]
                }] }
    ];
    /** @nocollapse */
    ActivityListComponent.ctorParameters = function () { return [
        { type: SocialService }
    ]; };
    ActivityListComponent.propDecorators = {
        activityPost: [{ type: ViewChild, args: ['activityPost',] }],
        postEnable: [{ type: Input }],
        scroller: [{ type: Input }],
        type: [{ type: Input }],
        application: [{ type: Input }],
        loader: [{ type: Input }]
    };
    return ActivityListComponent;
}());
export { ActivityListComponent };
if (false) {
    /** @type {?} */
    ActivityListComponent.prototype.activityPost;
    /** @type {?} */
    ActivityListComponent.prototype.postEnable;
    /** @type {?} */
    ActivityListComponent.prototype.scroller;
    /** @type {?} */
    ActivityListComponent.prototype.type;
    /** @type {?} */
    ActivityListComponent.prototype.application;
    /** @type {?} */
    ActivityListComponent.prototype.loader;
    /** @type {?} */
    ActivityListComponent.prototype.activities;
    /** @type {?} */
    ActivityListComponent.prototype.showLoading;
    /** @type {?} */
    ActivityListComponent.prototype.hasMore;
    /** @type {?} */
    ActivityListComponent.prototype.params;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype._activityChange;
    /**
     * @type {?}
     * @private
     */
    ActivityListComponent.prototype._loading;
    /** @type {?} */
    ActivityListComponent.prototype.socialService;
}
