// inet-ui/esm2015/src/sharing-information/sharing-information-template/sharing-information-template.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, ViewChild } from '@angular/core';
import { Observable, of, Subject } from "rxjs";
import { SharingInformationService } from "../sharing-information.service";
import { NotificationService } from "inet-core";
export class SharingInformationTemplateComponent {
    /**
     * @param {?} sharingInformationService
     * @param {?} notification
     */
    constructor(sharingInformationService, notification) {
        this.sharingInformationService = sharingInformationService;
        this.notification = notification;
        this.page = 0;
        this.limit = 10;
        this.finishLoading = false;
        this.dataTables = [];
        this.searchData = [];
        this.searchTerm$ = new Subject();
        this.destroy$ = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.onSearchUsers();
        this.loadTableData();
        this.onDocumentClick = this.onDocumentClick.bind(this);
        document.addEventListener('click', this.onDocumentClick);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        document.removeEventListener('click', this.onDocumentClick);
    }
    /**
     * @param {?} val
     * @return {?}
     */
    mapApplicationData(val) {
        if (val.organId !== 'community' && val.pattern !== 'calista' && val.firmContext && val.pattern) {
            return {
                "appName": val.firmContext,
                "appIcon": val.icon
            };
        }
    }
    /**
     * @return {?}
     */
    loadTableData() {
        this.sharingInformationService.getApplicationList({}).concatMap((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data && data.elements) {
                return of(data.elements.map((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => this.mapApplicationData(item))).filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item)));
            }
            return of(null);
        }))
            .concatMap((/**
         * @param {?} appInfo
         * @return {?}
         */
        (appInfo) => {
            this.dataTableInfo = appInfo || [];
            /** @type {?} */
            let appNames = appInfo.map((/**
             * @param {?} item
             * @return {?}
             */
            item => item.appName));
            if (appNames.length) {
                /** @type {?} */
                let __sources = appNames.map((/**
                 * @param {?} appname
                 * @param {?} index
                 * @return {?}
                 */
                (appname, index) => this.sharingInformationService.getApplicationRights({ appname: appname })));
                return Observable.concat(...__sources);
            }
            return of(null);
        }))
            .concatMap((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            if (data && data.length) {
                this.dataTables.push(data);
                return this.sharingInformationService.getUserRights({ appname: data[0].application });
            }
            else {
                this.dataTableInfo = this.dataTableInfo.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item.appName !== data.appname));
            }
            return of(null);
        }))
            .subscribe((/**
         * @param {?} data
         * @return {?}
         */
        data => {
            if (data && data.total) {
                /** @type {?} */
                let __userData = data.items || [];
                this.dataTables.forEach((/**
                 * @param {?} items
                 * @return {?}
                 */
                (items) => {
                    items.forEach((/**
                     * @param {?} table
                     * @return {?}
                     */
                    table => {
                        table.onSearch = false;
                        table.hidden = false;
                        /** @type {?} */
                        let __users = __userData.filter((/**
                         * @param {?} __user
                         * @return {?}
                         */
                        __user => __user.group == table.category && __user.application == table.application));
                        if (__users.length && __users[0].application == table.application) {
                            table.data = __users || [];
                            table.data = this.mergeValueInArray(table.data);
                        }
                    }));
                }));
            }
        }), (/**
         * @return {?}
         */
        () => {
        }), (/**
         * @return {?}
         */
        () => {
            this.finishLoading = true;
            // console.log(this.dataTables);
        }));
    }
    /**
     * @param {?} subItem
     * @return {?}
     */
    onToggleTable(subItem) {
        if (iNet.isEmpty(subItem.hidden)) {
            subItem.hidden = false;
        }
        else {
            subItem.hidden = !subItem.hidden;
        }
    }
    /**
     * @param {?} subItem
     * @return {?}
     */
    onSetSubItem(subItem) {
        this.subItem = subItem;
    }
    /**
     * @param {?} arr
     * @return {?}
     */
    mergeValueInArray(arr) {
        arr.forEach((/**
         * @param {?} item
         * @return {?}
         */
        (item) => {
            // console.log(arr);
            /** @type {?} */
            let __users = arr.filter((/**
             * @param {?} data
             * @return {?}
             */
            data => data.member === item.member)) || [];
            if (__users.length > 1) {
                arr = arr.filter((/**
                 * @param {?} val
                 * @return {?}
                 */
                val => val.member !== item.member));
                /** @type {?} */
                let __right = __users.map((/**
                 * @param {?} user
                 * @return {?}
                 */
                user => user.right)).join(",");
                /** @type {?} */
                let __uuids = __users.map((/**
                 * @param {?} user
                 * @return {?}
                 */
                user => [user.right, user.uuid]));
                // let __user: any={};
                arr.push(Object.assign({}, __users[0], { right: __right, uuids: __uuids }));
                // console.log(arr);
            }
            arr.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                if (!item.uuids) {
                    item.uuids = [[item.right, item.uuid]];
                }
            }));
        }));
        return arr;
    }
    /**
     * @param {?} user
     * @param {?} subItem
     * @param {?} $event
     * @return {?}
     */
    onCheckUserRight(user, subItem, $event) {
        /** @type {?} */
        let __checked = $event.target.checked;
        /** @type {?} */
        let __value = $event.target.value;
        if (__checked) {
            this.sharingInformationService.updateUserRight({
                appname: user.application,
                member: user.member,
                right: __value,
                group: user.group,
            }).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data && data.member) {
                    subItem.data.push(data);
                    /** @type {?} */
                    let __users = subItem.data.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => item.member === data.member && item.uuid)) || [];
                    if (__users.length > 1) {
                        /** @type {?} */
                        let __right = __users.map((/**
                         * @param {?} user
                         * @return {?}
                         */
                        user => user.right)).join(",");
                        subItem.data.forEach((/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => {
                            if (item.member == data.member && item.uuids) {
                                item.right = __right;
                                item.uuids.push([__value, data.uuid]);
                            }
                        }));
                    }
                    if (__users.length === 1) {
                        /** @type {?} */
                        let __right = __users.map((/**
                         * @param {?} user
                         * @return {?}
                         */
                        user => user.right)).join(",");
                        /** @type {?} */
                        let __uuids = __users.map((/**
                         * @param {?} user
                         * @return {?}
                         */
                        user => [user.right, user.uuid]));
                        subItem.data.push(Object.assign({}, __users[0], { right: __right, uuids: __uuids }));
                    }
                    subItem.data = subItem.data.filter((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => item.uuids));
                    // console.log("----->", subItem.data);
                    this.notification.showMessage("Đã chia sẻ dữ liệu cho " + data.member, "success", "Chia sẻ dữ liệu");
                }
                else {
                    this.notification.showMessage("Có lỗi xảy ra khi chia sẻ dữ liệu cho " + user.member, "error", "Chia sẻ dữ liệu");
                }
            }));
        }
        else {
            /** @type {?} */
            let __item = user.uuids.filter((/**
             * @param {?} item
             * @return {?}
             */
            item => item[0] == __value))[0] || [];
            this.sharingInformationService.deleteUserRight({
                appname: user.application,
                uuid: __item[1] || user.uuid
            }).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data && data.uuid) {
                    /** @type {?} */
                    let __index = user.uuids.findIndex((/**
                     * @param {?} item
                     * @return {?}
                     */
                    item => item[1] == data.uuid));
                    user.uuids.splice(__index, 1);
                    /** @type {?} */
                    let __rights = user.right.split(",").filter((/**
                     * @param {?} name
                     * @return {?}
                     */
                    name => name !== __value));
                    user.right = __rights.join(",");
                    if (!user.uuids.length) {
                        // console.log("innnn");
                        /** @type {?} */
                        let index = subItem.data.findIndex((/**
                         * @param {?} item
                         * @return {?}
                         */
                        item => item.member === user.member));
                        index > -1 ? subItem.data.splice(index, 1) : '';
                    }
                    this.notification.showMessage("Đã hủy chia sẻ dữ liệu cho " + user.member, "success", "Chia sẻ dữ liệu");
                }
                else {
                    this.notification.showMessage("Có lỗi xảy ra khi chia dữ liệu cho " + user.member, "error", "Chia sẻ dữ liệu");
                }
            }));
        }
    }
    /**
     * @param {?} $event
     * @param {?} subItem
     * @return {?}
     */
    onKeyUp($event, subItem) {
        /** @type {?} */
        let __keyword = $event.target.value || "";
        this.searchTerm$.next(__keyword);
    }
    /**
     * @return {?}
     */
    onSearchUsers() {
        this.searchTerm$
            .delay(50)
            .debounceTime(200)
            .switchMap((/**
         * @param {?} keyword
         * @return {?}
         */
        (keyword) => {
            this.keyword = keyword;
            return keyword ? this.sharingInformationService.getSuggestUser({ keyword: keyword }) : of(null);
        }))
            .subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            /** @type {?} */
            let __data = data ? data.items || [] : [];
            if (__data.length) {
                __data = __data.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item.username !== iNet.username));
            }
            this.searchData = __data;
            if (this.subItem.data) {
                /** @type {?} */
                let __temp = this.subItem.data.map((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => item.member));
                this.searchData = this.searchData.filter((/**
                 * @param {?} data
                 * @return {?}
                 */
                data => !__temp.includes(data.username)));
            }
        }));
    }
    /**
     * @param {?} subItem
     * @param {?} user
     * @return {?}
     */
    onChooseUser(subItem, user) {
        this.keyword = '';
        this.searchTerm$.next("");
        this.searchData = [];
        /** @type {?} */
        let __user = {
            application: subItem.application,
            member: user.username,
            right: "",
            group: subItem.category,
        };
        this.__userOnAdd = __user;
        if (!subItem.data) {
            subItem.data = [];
        }
        subItem.data.push(__user);
    }
    /**
     * @param {?} subItem
     * @return {?}
     */
    onFocusSearchOut(subItem) {
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.searchData = [];
            subItem.onSearch = false;
        }), 300);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onDocumentClick(event) {
        if (this.suggestElement && this.suggestElement.nativeElement.contains(event.target)) {
            return;
        }
        this.searchData = [];
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    scrollMouse($event) {
        /** @type {?} */
        const target = $event.target;
        if (this.searchData.length !== this.total && (target.offsetHeight + target.scrollTop >= target.scrollHeight)) {
            this.page++;
            this.sharingInformationService.getSuggestUser({
                pageSize: this.limit,
                pageNumber: this.page,
                keyword: this.keyword
            }).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                /** @type {?} */
                let newData = data.items || [];
                this.searchData = this.searchData.concat(newData);
            }));
        }
    }
}
SharingInformationTemplateComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-sharing-information-template',
                template: "<div *ngIf=\"finishLoading\" class=\"sharing-table-wrapper custom-toolbar\">\n    <div *ngFor=\"let tableInfo of dataTableInfo; let index = index\" class=\"sharing-table-container\">\n        <div class=\"sharing-table-header\">\n            <div class=\"sharing-table-header-wrapper\">\n                <div class=\"d-flex\">\n                    <div class=\"sharing-table-popicon\">\n                        <img class=\"p-1 plugin-icon float-left application-icon\"\n                             [src]=\"tableInfo.appIcon\" onerror=\"this.onerror=null; this.src='https://media.inetcloud.vn/images/unicorn/app.png'\">\n                    </div>\n                    <div class=\"sharing-table-application\">\n                        <span class=\"mr-2\" style=\"text-transform: capitalize\">{{tableInfo.appName}}</span>\n                    </div>\n                </div>\n            </div>\n            <div *ngFor=\"let item of dataTables; let  index=index\" class=\"row m-0 pr-2 pl-2\">\n                <div *ngFor=\"let subItem of item; let subIndex= index\" class=\"col-lg-12\">\n                    <div *ngIf=\"subItem.application == tableInfo.appName\">\n                        <div class=\"right-name\" [style.marginBottom]=\"subItem.hidden?'6px':''\" (click)=\"onToggleTable(subItem)\"\n                             [style.background]=\"subItem.hidden==false?'#ededed':'#ededed'\">\n                            <div class=\"right-icon\">\n                                <i class=\"fa fa-angle-right\" [ngClass]=\"subItem.hidden==false?'fa-angle-down':''\"\n                                   aria-hidden=\"true\"></i>\n                            </div>\n                            {{subItem.category || \"\"}}\n                        </div>\n                        <div [hidden]=\"subItem.hidden\" class=\"table-wrapper col-lg-12 p-0\">\n                            <div class=\"table-responsive custom-toolbar\">\n                                <table class=\"table\">\n                                    <thead>\n                                    <tr>\n                                        <th id=\"user-name\" scope=\"col\">T\u00E0i kho\u1EA3n</th>\n                                        <th *ngFor=\"let right of subItem.rights; let rightIndex = index\" scope=\"col\"\n                                            style=\"text-transform: capitalize\">{{right}}</th>\n                                        <!--                                        <th></th>-->\n                                    </tr>\n                                    </thead>\n                                    <tbody style=\"position:relative\">\n                                    <tr class=\"user-search p-0 d-flex align-items-center\" style=\"margin-bottom: 36px\">\n                                        <td style=\"width: 100%; border:none;padding: 6px 0 0\">\n                                            <div class=\"input-group input-group-sm p-0 position-absolute\">\n                                                <div class=\"input-group-prepend\">\n                                                    <span class=\"input-group-text\" id=\"basic-addon1\" style=\"height: 30px\">\n                                                        <i class=\"fa fa-search\" aria-hidden=\"true\"></i></span>\n                                                </div>\n                                                <input (keyup)=\"onKeyUp($event, subItem)\" type=\"text\"\n                                                       (focusin)=\"subItem.onSearch = true; onSetSubItem(subItem)\"\n                                                       (focusout)=\"onFocusSearchOut(subItem)\"\n                                                       class=\"form-control form-control-sm\"\n                                                       placeholder=\"T\u00ECm t\u00E0i kho\u1EA3n\" aria-label=\"Username\"\n                                                       aria-describedby=\"basic-addon1\" style=\"height: 30px\">\n                                                <div #suggestEle (scroll)=\"scrollMouse($event)\"\n                                                     *ngIf=\"subItem.onSearch && searchData.length\"\n                                                     class=\"form-invitation-popover custom-toolbar\">\n                                                    <div *ngFor=\"let user of searchData; let i=index\"\n                                                         (click)=\"onChooseUser(subItem, user)\"\n                                                         class=\"popover-data d-flex\">\n                                                        <div class=\"user-avatar mr-2\">\n                                                            <img class=\"user-avatar-custom\" userAvatar\n                                                                 [usercode]=\"user.username\">\n                                                        </div>\n                                                        <div class=\"user-info-wrapper\">\n                                                            <div class=\"user-name\">\n                                                                {{user.fullname || \"\"}}\n                                                            </div>\n                                                            <div class=\"user-email\">\n                                                                {{user.username}}\n                                                            </div>\n                                                        </div>\n                                                    </div>\n                                                </div>\n                                            </div>\n                                        </td>\n                                    </tr>\n                                    <tr *ngFor=\"let user of subItem.data\">\n                                        <td>{{user.member}}</td>\n                                        <td *ngFor=\"let right of subItem.rights;\">\n                                            <div class=\"custom-control custom-checkbox ml-2\">\n                                                <input (click)=\"onCheckUserRight(user, subItem, $event)\"\n                                                       [checked]=\"user.right.includes(right)\" [value]=\"right\"\n                                                       type=\"checkbox\" class=\"custom-control-input\"\n                                                       [id]=\"user.application+user.member+subItem.category+right\">\n                                                <label class=\"custom-control-label\"\n                                                       [for]=\"user.application+user.member+subItem.category+right\"><span\n                                                        class=\"hidden-span\">a</span></label>\n                                            </div>\n                                        </td>\n                                    </tr>\n                                    </tbody>\n                                </table>\n                            </div>\n                        </div>\n                    </div>\n                </div>\n            </div>\n        </div>\n    </div>\n</div>\n",
                styles: [".sharing-table-wrapper{background:#fff;padding:0 20px}.sharing-table-wrapper .hidden-span{visibility:hidden}.sharing-table-wrapper table tr td:not(:first-child),.sharing-table-wrapper table tr th:not(:first-child){text-align:center}.sharing-table-wrapper table tbody tr td{padding:6px}.sharing-table-wrapper #user-name{font-weight:400;width:180px}.sharing-table-wrapper .sharing-table-container .sharing-table-header .sharing-table-header-wrapper{display:flex;justify-content:space-between;padding-bottom:4px;color:#366dd0}.sharing-table-wrapper .sharing-table-container .sharing-table-header .sharing-table-header-wrapper .sharing-table-popicon{margin-right:6px;font-size:21px}.sharing-table-wrapper .sharing-table-container .sharing-table-header .sharing-table-header-wrapper .sharing-table-popicon .application-icon{height:40px}.sharing-table-wrapper .sharing-table-container .sharing-table-header .sharing-table-header-wrapper .sharing-table-application{font-size:1.4em;font-weight:700;display:flex;align-items:center}.sharing-table-wrapper .sharing-table-container .sharing-table-header .user-search .form-invitation-popover{max-height:150px;width:100%;z-index:5;background:#fff;border:.5px solid #d3d3d3;border-radius:.1rem;overflow-x:hidden;overflow-y:scroll;top:32px}.sharing-table-wrapper .sharing-table-container .sharing-table-header .user-search .form-invitation-popover .popover-data{padding:6px;cursor:pointer;border-bottom:1px solid #f0f0f0}.sharing-table-wrapper .sharing-table-container .sharing-table-header .user-search .form-invitation-popover .popover-data:hover{background:#d5f1ff}.sharing-table-wrapper .sharing-table-container .sharing-table-header .user-search .form-invitation-popover .popover-data .user-avatar-custom{width:33px;height:33px;border:1px solid #e3e3e3}.sharing-table-wrapper .sharing-table-container .sharing-table-header .user-search .form-invitation-popover .popover-data .user-info-wrapper{display:flex;flex-direction:column}.sharing-table-wrapper .sharing-table-container .sharing-table-header .user-search .form-invitation-popover .popover-data .user-info-wrapper .user-email{font-size:80%}.right-name{font-size:16px;padding:6px 3px;border-radius:.2rem;font-weight:700;display:flex;cursor:pointer}.right-name:hover{background:#ededed!important}.right-icon{margin-right:10px}@-webkit-keyframes slidein{0%{margin-top:-12px}100%{margin-top:0}}@keyframes slidein{0%{margin-top:-12px}100%{margin-top:0}}.table-wrapper{-webkit-animation-name:slidein;animation-name:slidein;-webkit-animation-duration:.25s;animation-duration:.25s;display:block}"]
            }] }
];
/** @nocollapse */
SharingInformationTemplateComponent.ctorParameters = () => [
    { type: SharingInformationService },
    { type: NotificationService }
];
SharingInformationTemplateComponent.propDecorators = {
    suggestElement: [{ type: ViewChild, args: ['suggestEle',] }]
};
if (false) {
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.total;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.page;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.limit;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.keyword;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.finishLoading;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.subItem;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.dataTables;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.searchData;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.dataTableInfo;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.searchTerm$;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.destroy$;
    /**
     * @type {?}
     * @private
     */
    SharingInformationTemplateComponent.prototype.__userOnAdd;
    /** @type {?} */
    SharingInformationTemplateComponent.prototype.suggestElement;
    /**
     * @type {?}
     * @private
     */
    SharingInformationTemplateComponent.prototype.sharingInformationService;
    /**
     * @type {?}
     * @private
     */
    SharingInformationTemplateComponent.prototype.notification;
}
