// inet-ui/esm2015/src/layout/bootstrap/app-side-nav/app-side-nav.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef } from '@angular/core';
import { AbstractSideNavComponent } from "../abstract-side-nav.component";
import { Router } from "@angular/router";
import { CoreService } from "inet-core";
export class AppSideNavComponent extends AbstractSideNavComponent {
    /**
     * @param {?} coreService
     * @param {?} router
     * @param {?} elementRef
     */
    constructor(coreService, router, elementRef) {
        super(elementRef);
        this.coreService = coreService;
        this.router = router;
        this.elementRef = elementRef;
        this.applications = [];
        this.widgetApps = [];
        this.systemUrl = 'about:blank';
        this.SYSTEM_APPS = ['social', 'calendar'];
        this.listApp();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        window.addEventListener('message', (/**
         * @param {?} e
         * @return {?}
         */
        (e) => {
            /** @type {?} */
            var __data = e.data || {};
            /** @type {?} */
            var __eventName = __data['eventName'];
            /** @type {?} */
            var __result = {};
            if (!iNet.isEmpty(__data.result)) {
                if (typeof (__data.result) === 'string') {
                    __result = JSON.parse(__data.result);
                }
                else if (typeof (__data.result) === 'object') {
                    __result = __data.result || {};
                }
                //console.log('[message]', __result);
                switch (__eventName) {
                    case 'open':
                    case 'openApp':
                        this.openApp(__result['application']);
                        this.close();
                        break;
                    case 'openUrl':
                        window.location.href = iNet.getSSOUrl(__result['url']);
                        break;
                    case 'loaded':
                        /** @type {?} */
                        const iFrame = document.querySelectorAll(`iframe[src="${__result['src']}"]`)[0];
                        if (iFrame) {
                            iFrame['height'] = __result['height'];
                            iFrame['valid'] = true;
                        }
                        break;
                }
            }
        }), false);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.close();
        if (this._router) {
            this._router.unsubscribe();
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const currentValue = changes['opened']["currentValue"];
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (currentValue) {
                this.systemUrl = iNet.getPUrl('common/page/apps') + '?appname=system';
                this.open();
            }
            else {
                this.close();
            }
        }), 250);
    }
    /**
     * @private
     * @return {?}
     */
    listApp() {
        this.coreService.getSystemApplication().subscribe((/**
         * @param {?} apps
         * @return {?}
         */
        (apps) => {
            this.applications = apps
                .filter((/**
             * @param {?} app
             * @return {?}
             */
            (app) => (app.organId !== 'community' && app.firmContext)))
                .map((/**
             * @param {?} application
             * @return {?}
             */
            (application) => {
                if (application.homepage.indexOf('http') < 0) {
                    application.homepage = iNet.getPUrl(`${application.module}/${application.homepage}`);
                    application.icon = `${this.coreService.getFileServerPath()}/images/${application.module}/${application.icon}`;
                }
                if (!!application.widget) {
                    application.widget = `${iNet.getPUrl('common/page/apps')}?appname=${application.firmContext}`; //rebuild the widget url
                }
                return application;
            }));
            this.widgetApps = this.applications.filter((/**
             * @param {?} app
             * @return {?}
             */
            (app) => (!!app.widget)));
        }));
    }
    /**
     * @param {?} app
     * @return {?}
     */
    openApp(app) {
        if (this.SYSTEM_APPS.includes(app)) { //system app
            this.router.navigate([app]);
        }
        else {
            /** @type {?} */
            const url = iNet.getSSOUrl(this.getUrlByContext(app));
            if (url) {
                window.location.href = url;
            }
        }
    }
    /**
     * @private
     * @param {?} content
     * @return {?}
     */
    getUrlByContext(content) {
        /** @type {?} */
        const app = this.applications.find((/**
         * @param {?} __0
         * @return {?}
         */
        ({ webContext }) => webContext === content));
        return app ? app.homepage : null;
    }
}
AppSideNavComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-admin-app-side-nav',
                template: "<div class=\"slide-nav\" [ngClass]=\"{'opened': opened}\">\n    <ul class=\"list-box list-app\" *ngIf=\"opened\">\n        <iframe id=\"system-app-iframe\" [src]=\"systemUrl | safe\" scrolling=\"no\"></iframe>\n        <iframe *ngFor=\"let widgetApp of widgetApps\" [src]=\"widgetApp.widget | safe\" scrolling=\"no\"></iframe>\n    </ul>\n</div>\n",
                styles: [`
        iframe {
            width: 100%;
            border: none;
            overflow: hidden;
        }
    `]
            }] }
];
/** @nocollapse */
AppSideNavComponent.ctorParameters = () => [
    { type: CoreService },
    { type: Router },
    { type: ElementRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AppSideNavComponent.prototype._router;
    /** @type {?} */
    AppSideNavComponent.prototype.applications;
    /** @type {?} */
    AppSideNavComponent.prototype.widgetApps;
    /** @type {?} */
    AppSideNavComponent.prototype.systemUrl;
    /** @type {?} */
    AppSideNavComponent.prototype.SYSTEM_APPS;
    /**
     * @type {?}
     * @private
     */
    AppSideNavComponent.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    AppSideNavComponent.prototype.router;
    /** @type {?} */
    AppSideNavComponent.prototype.elementRef;
}
