// inet-ui/esm5/src/dialog/confirm-dialog/confirm-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { DialogAction } from "../dialog-action";
import { TranslateService } from "@ngx-translate/core";
var ConfirmDialogComponent = /** @class */ (function () {
    function ConfirmDialogComponent(modalService, translate) {
        var _this = this;
        this.modalService = modalService;
        this.translate = translate;
        this.id = iNet.generateId();
        this.iconCls = '';
        this.title = '';
        this.content = '';
        this.cls = '';
        this.actions = [];
        this.translateSubscription = translate.get(['TOOLBAR']).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            var TOOLBAR = (res || { 'TOOLBAR': { 'CLOSE': 'Close' } }).TOOLBAR;
            _this.toolbarTranslations = TOOLBAR;
        }));
    }
    /**
     * @return {?}
     */
    ConfirmDialogComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.getActions().length < 1) {
            var CLOSE = this.toolbarTranslations.CLOSE;
            /** @type {?} */
            var closeAction = new DialogAction(CLOSE, 'btn-danger', 'fa fa-check', this.hide);
            this.setActions([closeAction]);
        }
    };
    /**
     * @return {?}
     */
    ConfirmDialogComponent.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.modalRef.hide();
    };
    /**
     * @return {?}
     */
    ConfirmDialogComponent.prototype.show = /**
     * @return {?}
     */
    function () {
        this.modalRef = this.modalService.show(this.confirmModal);
    };
    /**
     * @param {?} action
     * @return {?}
     */
    ConfirmDialogComponent.prototype.pushAction = /**
     * @param {?} action
     * @return {?}
     */
    function (action) {
        this.actions.push(action);
    };
    /**
     * @param {?} actions
     * @return {?}
     */
    ConfirmDialogComponent.prototype.setActions = /**
     * @param {?} actions
     * @return {?}
     */
    function (actions) {
        this.actions = actions;
    };
    /**
     * @return {?}
     */
    ConfirmDialogComponent.prototype.getActions = /**
     * @return {?}
     */
    function () {
        return this.actions;
    };
    /**
     * @param {?} v
     * @return {?}
     */
    ConfirmDialogComponent.prototype.setData = /**
     * @param {?} v
     * @return {?}
     */
    function (v) {
        this.data = v;
    };
    /**
     * @return {?}
     */
    ConfirmDialogComponent.prototype.getData = /**
     * @return {?}
     */
    function () {
        return this.data;
    };
    /**
     * @return {?}
     */
    ConfirmDialogComponent.prototype.getId = /**
     * @return {?}
     */
    function () {
        return this.id;
    };
    ConfirmDialogComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-confirm-dialog',
                    template: "<ng-template #confirmModal>\n    <div class=\"modal-header\">\n        <h4 class=\"modal-title pull-left\"><i class=\"{{iconCls}}\"></i> {{title}}</h4>\n        <button type=\"button\" class=\"close pull-right\" (click)=\"hide()\">\n            <span aria-hidden=\"true\">&times;</span>\n        </button>\n    </div>\n    <div class=\"modal-body\" [innerHTML]=\"content\"></div>\n    <div class=\"modal-footer text-right\">\n        <button type=\"button\" *ngFor=\"let action of actions\" class=\"btn btn-sm {{action.cls}}\"\n                (click)=\"action?.fn.bind(this)()\"><i class=\"{{action.iconCls}}\"></i> {{action.title}}</button>\n    </div>\n</ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    ConfirmDialogComponent.ctorParameters = function () { return [
        { type: BsModalService },
        { type: TranslateService }
    ]; };
    ConfirmDialogComponent.propDecorators = {
        confirmModal: [{ type: ViewChild, args: ['confirmModal',] }],
        id: [{ type: Input, args: ['id',] }],
        iconCls: [{ type: Input, args: ['iconCls',] }],
        title: [{ type: Input, args: ['title',] }],
        content: [{ type: Input, args: ['content',] }],
        cls: [{ type: Input, args: ['cls',] }],
        actions: [{ type: Input, args: ['actions',] }]
    };
    return ConfirmDialogComponent;
}());
export { ConfirmDialogComponent };
if (false) {
    /** @type {?} */
    ConfirmDialogComponent.prototype.confirmModal;
    /** @type {?} */
    ConfirmDialogComponent.prototype.id;
    /** @type {?} */
    ConfirmDialogComponent.prototype.iconCls;
    /** @type {?} */
    ConfirmDialogComponent.prototype.title;
    /** @type {?} */
    ConfirmDialogComponent.prototype.content;
    /** @type {?} */
    ConfirmDialogComponent.prototype.cls;
    /** @type {?} */
    ConfirmDialogComponent.prototype.actions;
    /**
     * @type {?}
     * @private
     */
    ConfirmDialogComponent.prototype.modalRef;
    /**
     * @type {?}
     * @private
     */
    ConfirmDialogComponent.prototype.data;
    /**
     * @type {?}
     * @private
     */
    ConfirmDialogComponent.prototype.toolbarTranslations;
    /**
     * @type {?}
     * @private
     */
    ConfirmDialogComponent.prototype.translateSubscription;
    /**
     * @type {?}
     * @private
     */
    ConfirmDialogComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ConfirmDialogComponent.prototype.translate;
}
