// inet-social/esm5/src/social-friend/list-user/list-user.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { SocialService } from '../../social.service';
import { AccentService, CacheStorageService } from 'inet-core';
var ListUserComponent = /** @class */ (function () {
    function ListUserComponent(socialService, accentService, cacheStorage) {
        var _this = this;
        this.socialService = socialService;
        this.accentService = accentService;
        this.cacheStorage = cacheStorage;
        this.placeholder = 'Tìm kiếm...';
        this.height = '240px';
        this.members = [];
        this.selectMembers = [];
        this.showMembers = [];
        this.matchMembers = [];
        this._allEmployees = 'social_employees';
        this._allFriends = 'social_friends';
        /** @type {?} */
        var employees = this.cacheStorage.getData(this._allEmployees);
        // Cache employees on 5 minutes
        if (employees) {
            this._initMembers(JSON.parse(employees));
        }
        else {
            this.cacheStorage.promiseQueue(this._allEmployees, (/**
             * @param {?} key
             * @return {?}
             */
            function (key) {
                _this.socialService.searchMember({ pageSize: -1 }, (/**
                 * @param {?} users
                 * @return {?}
                 */
                function (users) {
                    if (users && users.length > 0) {
                        _this.cacheStorage.setData(key, JSON.stringify(users));
                    }
                    _this.cacheStorage.resolveQueue(key, _this, [users]);
                }));
            }), (/**
             * @param {?} users
             * @return {?}
             */
            function (users) { return _this._initMembers(users); }));
        }
        this.cacheStorage.promiseQueue(this._allFriends, (/**
         * @param {?} key
         * @return {?}
         */
        function (key) {
            _this.socialService.listFriends({ pageSize: -1 }, (/**
             * @param {?} users
             * @return {?}
             */
            function (users) {
                _this.cacheStorage.resolveQueue(key, _this, [users]);
            }));
        }), (/**
         * @param {?} users
         * @return {?}
         */
        function (users) { return _this._initMembers(users); }));
    }
    /**
     * @return {?}
     */
    ListUserComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._search = this.search.nativeElement;
    };
    /**
     * @return {?}
     */
    ListUserComponent.prototype.resetData = /**
     * @return {?}
     */
    function () {
        this.selectMembers = [];
        this.showMembers = [];
        this.matchMembers = [];
        this._search.value = '';
        this.members.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return delete item.selected; }));
    };
    /**
     * @return {?}
     */
    ListUserComponent.prototype.focusInput = /**
     * @return {?}
     */
    function () {
        this._search.focus();
    };
    /**
     * @return {?}
     */
    ListUserComponent.prototype.onSearch = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var value = this._search.value.trim();
        if (value) {
            value = this.accentService.viToEn(value.toLowerCase());
            this.matchMembers = this.members.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item._index.indexOf(value) > -1; }));
        }
        else {
            this.matchMembers = this.members;
        }
        this.showMembers = this.matchMembers;
    };
    /**
     * @return {?}
     */
    ListUserComponent.prototype.onScroll = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} member
     * @return {?}
     */
    ListUserComponent.prototype.addMember = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        member.selected = !member.selected;
        if (member.selected) {
            this.selectMembers.push(member);
        }
        else {
            this.removeMember(member);
        }
    };
    /**
     * @param {?} member
     * @return {?}
     */
    ListUserComponent.prototype.removeMember = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        this.selectMembers.splice(this.selectMembers.indexOf(member), 1);
    };
    /**
     * @private
     * @param {?} users
     * @return {?}
     */
    ListUserComponent.prototype._initMembers = /**
     * @private
     * @param {?} users
     * @return {?}
     */
    function (users) {
        var _this = this;
        users.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return _this._addMember(item); }));
        this._sortMembers();
    };
    /**
     * @private
     * @param {?} member
     * @return {?}
     */
    ListUserComponent.prototype._addMember = /**
     * @private
     * @param {?} member
     * @return {?}
     */
    function (member) {
        member.username = member.username || member.usercode;
        if (this._hasMember(member)) {
            return;
        }
        member._index = this.accentService.viToEn(member.fullname.toLowerCase()) + ' ' + member.username;
        this.members.push(member);
    };
    /**
     * @private
     * @return {?}
     */
    ListUserComponent.prototype._sortMembers = /**
     * @private
     * @return {?}
     */
    function () {
        this.members.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) { return a.fullname.localeCompare(b.fullname); }));
    };
    /**
     * @private
     * @param {?} member
     * @return {?}
     */
    ListUserComponent.prototype._hasMember = /**
     * @private
     * @param {?} member
     * @return {?}
     */
    function (member) {
        for (var i = 0; i < this.members.length; i++) {
            if (this.members[i].username === member.username) {
                return true;
            }
        }
    };
    ListUserComponent.decorators = [
        { type: Component, args: [{
                    selector: '[socialListUser]',
                    template: "<div class=\"auto-tag form-control\" style=\"max-height: 140px;overflow: auto;\">\n  <div *ngFor=\"let member of selectMembers\" class=\"auto-tag-item\">\n    <span>{{member.fullname}}</span>\n    <i class=\"fa fa-times\" (click)=\"removeMember(member)\"></i>\n  </div>\n  <input #search (keyup)=\"onSearch()\" type=\"text\" class=\"auto-tag-input\" [placeholder]=\"placeholder\">\n</div>\n<div #contactList (scroll)=\"onScroll()\"\n     style=\"overflow:auto;height: 240px;margin-top:10px\">\n  <div class=\"list-item list-avatar\" *ngFor=\"let member of showMembers\" (click)=\"addMember(member)\">\n    <img userAvatar [usercode]=\"member.username || member.usercode\" class=\"list-item__image\">\n    <div class=\"list-item__content\">{{member.fullname}}</div>\n    <label class=\"checkbox-container\">\n      <input type=\"checkbox\" [checked]=\"member.selected\" disabled>\n      <span class=\"checkmark\"></span>\n    </label>\n  </div>\n</div>",
                    styles: [".auto-tag{display:flex;flex-wrap:wrap;height:auto;padding-bottom:0}.auto-tag .auto-tag-item{display:flex;align-items:center;border-radius:4px;background:#eee;padding:0 10px;white-space:nowrap;margin:0 5px 5px 0}.auto-tag .auto-tag-item i{cursor:pointer;padding:5px;margin-right:-5px}.auto-tag .auto-tag-input{border:0;outline:0}.checkbox-container{display:block;position:relative;cursor:pointer;font-size:22px;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;min-width:20px;min-height:20px;margin:0}.checkbox-container input{position:absolute;opacity:0;cursor:pointer}.checkbox-container .checkmark{position:absolute;top:0;left:0;height:20px;width:20px;background-color:#eee;border-radius:4px}.checkbox-container .checkmark:after{content:\"\";position:absolute;display:none;left:8px;top:4px;width:5px;height:10px;border:solid #fff;border-width:0 3px 3px 0;transform:rotate(45deg)}.checkbox-container:hover .checkmark{background-color:#ccc}.checkbox-container input:checked+.checkmark{background-color:#2196f3}.checkbox-container input:checked+.checkmark:after{display:block}.list-avatar{padding:5px 10px;overflow:hidden;cursor:pointer;margin-bottom:5px;display:flex;align-items:center}.list-item__icon,.list-item__image{width:32px;height:32px;border-radius:50%}.list-item__icon{line-height:32px;text-align:center;background:rgba(0,0,0,.05);color:rgba(0,0,0,.5)}.list-item__content{overflow:hidden;text-overflow:ellipsis;white-space:nowrap;padding:0 15px;line-height:32px;flex-grow:1}"]
                }] }
    ];
    /** @nocollapse */
    ListUserComponent.ctorParameters = function () { return [
        { type: SocialService },
        { type: AccentService },
        { type: CacheStorageService }
    ]; };
    ListUserComponent.propDecorators = {
        placeholder: [{ type: Input }],
        height: [{ type: Input }],
        search: [{ type: ViewChild, args: ['search',] }]
    };
    return ListUserComponent;
}());
export { ListUserComponent };
if (false) {
    /** @type {?} */
    ListUserComponent.prototype.placeholder;
    /** @type {?} */
    ListUserComponent.prototype.height;
    /** @type {?} */
    ListUserComponent.prototype.search;
    /** @type {?} */
    ListUserComponent.prototype.members;
    /** @type {?} */
    ListUserComponent.prototype.selectMembers;
    /** @type {?} */
    ListUserComponent.prototype.showMembers;
    /** @type {?} */
    ListUserComponent.prototype.matchMembers;
    /**
     * @type {?}
     * @private
     */
    ListUserComponent.prototype._search;
    /**
     * @type {?}
     * @private
     */
    ListUserComponent.prototype._allEmployees;
    /**
     * @type {?}
     * @private
     */
    ListUserComponent.prototype._allFriends;
    /**
     * @type {?}
     * @private
     */
    ListUserComponent.prototype.socialService;
    /**
     * @type {?}
     * @private
     */
    ListUserComponent.prototype.accentService;
    /**
     * @type {?}
     * @private
     */
    ListUserComponent.prototype.cacheStorage;
}
