// inet-ui/esm5/src/common/location.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService, ResponseData } from 'inet-core';
import { Observable } from 'rxjs/Rx';
import 'rxjs/add/operator/map';
import 'rxjs/add/observable/of';
import 'rxjs/add/operator/do';
import { shareReplay } from 'rxjs/operators';
var LocationService = /** @class */ (function () {
    function LocationService(http) {
        this.http = http;
        this.url = {
            // list_country: iNet.getPUrl('cloudapp/country/list'),
            // list_city: iNet.getPUrl('cloudapp/city/list'),
            // list_district: iNet.getPUrl('cloudapp/district/list'),
            list_country: iNet.getPUrl('gl/country/list'),
            list_city: iNet.getPUrl('gl/city/list'),
            list_district: iNet.getPUrl('gl/district/list')
        };
        this.$provinceCache = {};
        this.$districtCache = {};
        this.provinceData = {}; // store province by country code
        // store province by country code
        this.districtData = {}; // store district by province code
        return LocationService.instance = LocationService.instance || this;
    }
    /**
     * @private
     * @return {?}
     */
    LocationService.prototype.getCountry = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this.http.getJSON(this.url.list_country).do((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return _this.countries = res['items']; }));
    };
    /**
     * @private
     * @param {?} countryCode
     * @return {?}
     */
    LocationService.prototype._getProvinceByCountryCode = /**
     * @private
     * @param {?} countryCode
     * @return {?}
     */
    function (countryCode) {
        var _this = this;
        return this.http.getJSON(this.url.list_city, { countryCode: countryCode })
            .do((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return _this.provinceData[countryCode] = res['items']; }));
    };
    /**
     * @private
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    LocationService.prototype._getDistrictByCode = /**
     * @private
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    function (countryCode, provinceCode) {
        var _this = this;
        /** @type {?} */
        var key = countryCode + '_' + provinceCode;
        return this.http.getJSON(this.url.list_district, { countryCode: countryCode, cityCode: provinceCode })
            .do((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return _this.districtData[key] = res['items']; }));
    };
    /**
     * @return {?}
     */
    LocationService.prototype.listCountry = /**
     * @return {?}
     */
    function () {
        if (this.countries) {
            /** @type {?} */
            var responseData = new ResponseData();
            responseData.items = this.countries;
            return Observable.of(responseData);
        }
        if (!this.$countryCache) {
            this.$countryCache = this.getCountry().pipe(shareReplay(1));
        }
        return this.$countryCache;
    };
    /*
     * @deprecated Use getProvinceByCountryCode(countryCode) instead
     */
    /*
         * @deprecated Use getProvinceByCountryCode(countryCode) instead
         */
    /**
     * @param {?} countryId
     * @return {?}
     */
    LocationService.prototype.getProvinceByCountryId = /*
         * @deprecated Use getProvinceByCountryCode(countryCode) instead
         */
    /**
     * @param {?} countryId
     * @return {?}
     */
    function (countryId) {
        return this.http.getJSON(this.url.list_city, { country: countryId });
    };
    /*
     * @deprecated Use getDistrictByCode(countryCode, provinceCode) instead
     */
    /*
         * @deprecated Use getDistrictByCode(countryCode, provinceCode) instead
         */
    /**
     * @param {?} uuid
     * @return {?}
     */
    LocationService.prototype.getDistrictByProvinceId = /*
         * @deprecated Use getDistrictByCode(countryCode, provinceCode) instead
         */
    /**
     * @param {?} uuid
     * @return {?}
     */
    function (uuid) {
        return this.http.getJSON(this.url.list_district, { city: uuid });
    };
    /**
     * @param {?} countryCode
     * @return {?}
     */
    LocationService.prototype.getProvinceByCountryCode = /**
     * @param {?} countryCode
     * @return {?}
     */
    function (countryCode) {
        if (this.provinceData[countryCode]) {
            /** @type {?} */
            var responseData = new ResponseData();
            responseData.items = this.provinceData[countryCode];
            return Observable.of(responseData);
        }
        if (!this.$provinceCache[countryCode]) {
            this.$provinceCache[countryCode] = this._getProvinceByCountryCode(countryCode).pipe(shareReplay(1));
        }
        return this.$provinceCache[countryCode];
    };
    /**
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    LocationService.prototype.getDistrictByCode = /**
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    function (countryCode, provinceCode) {
        /** @type {?} */
        var key = countryCode + '_' + provinceCode;
        if (this.districtData[key]) {
            /** @type {?} */
            var responseData = new ResponseData();
            responseData.items = this.districtData[key];
            return Observable.of(responseData);
        }
        if (!this.$districtCache[key]) {
            this.$districtCache[key] = this._getDistrictByCode(countryCode, provinceCode).pipe(shareReplay(1));
        }
        return this.$districtCache[key];
    };
    LocationService.instance = null;
    LocationService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LocationService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    return LocationService;
}());
export { LocationService };
if (false) {
    /** @type {?} */
    LocationService.instance;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.$countryCache;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.$provinceCache;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.$districtCache;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.countries;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.provinceData;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.districtData;
    /**
     * @type {?}
     * @private
     */
    LocationService.prototype.http;
}
