// inet-ui/esm5/src/common/notify-message.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from 'inet-core';
var NotifyMessageService = /** @class */ (function () {
    function NotifyMessageService(http) {
        this.http = http;
        this.url = {
            list: iNet.getPUrl('social/activity/notify'),
            clear: iNet.getPUrl('social/activity/clearnotify'),
            load_notify: iNet.getPUrl('social/activity/loadnotify')
        };
        return NotifyMessageService.instance = NotifyMessageService.instance || this;
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    NotifyMessageService.prototype.getMessages = /**
     * @param {?=} params
     * @return {?}
     */
    function (params) {
        if (params === void 0) { params = {}; }
        return this.http.getJSON(this.url.list, params);
    };
    /**
     * @param {?=} params
     * @return {?}
     */
    NotifyMessageService.prototype.count = /**
     * @param {?=} params
     * @return {?}
     */
    function (params) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @return {?}
         */
        function (resolve) {
            _this.http.getJSON(_this.url.list, params).subscribe((/**
             * @param {?} v
             * @return {?}
             */
            function (v) {
                resolve(v);
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                resolve();
            }));
        }));
    };
    /**
     * @param {?=} params
     * @return {?}
     */
    NotifyMessageService.prototype.clearAll = /**
     * @param {?=} params
     * @return {?}
     */
    function (params) {
        if (params === void 0) { params = {}; }
        return this.http.postJSON(this.url.clear, params);
    };
    /**
     * @param {?} app
     * @param {?} activityId
     * @return {?}
     */
    NotifyMessageService.prototype.loadNotify = /**
     * @param {?} app
     * @param {?} activityId
     * @return {?}
     */
    function (app, activityId) {
        return this.http.getJSON(this.url.load_notify, { activity: activityId, application: app });
    };
    /**
     * @param {?} activityId
     * @return {?}
     */
    NotifyMessageService.prototype.clearByActivityId = /**
     * @param {?} activityId
     * @return {?}
     */
    function (activityId) {
        return this.http.postJSON(this.url.clear, { activity: activityId });
    };
    NotifyMessageService.instance = null;
    NotifyMessageService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    NotifyMessageService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    return NotifyMessageService;
}());
export { NotifyMessageService };
if (false) {
    /** @type {?} */
    NotifyMessageService.instance;
    /**
     * @type {?}
     * @private
     */
    NotifyMessageService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    NotifyMessageService.prototype.http;
}
