// inet-chat/esm2015/src/component/room-chat/room-chat.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, } from '@angular/core';
import { VideoConferenceService } from "../../video-conference.service";
export class RoomChatComponent {
    /**
     * @param {?} videoService
     */
    constructor(videoService) {
        this.videoService = videoService;
        this.videoPanelVisible = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.joinRoom();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.disposeRoom();
    }
    /**
     * @return {?}
     */
    joinRoom() {
        this.videoPanelVisible = false;
        if (!this.videoService.isExistMeet()) {
            return;
        }
        this.disposeRoom();
        this.videoService.join(document.querySelector('#meetRoom'), this.roomName, false, 600);
        this.videoService.getApi().addEventListener('readyToClose', (/**
         * @return {?}
         */
        () => {
            this.disposeRoom();
        }));
    }
    /**
     * @return {?}
     */
    disposeRoom() {
        this.videoPanelVisible = true;
        this.videoService.dispose();
    }
}
RoomChatComponent.decorators = [
    { type: Component, args: [{
                selector: 'room-chat',
                template: "<div class=\"message-video\">\n    <div *ngIf=\"videoPanelVisible\" class=\"container text-center mb-2\" style=\"background: #000000;min-height: 250px;\">\n        <h5 class=\"text-white\" style=\"padding: 40px;\">Cu\u1ED9c h\u1ECDp c\u00F4ng ty , ai c\u0169ng c\u00F3 th\u1EC3 tham gia</h5>\n        <button class=\"btn btn-success\" (click)=\"joinRoom()\">Tham gia</button>\n    </div>\n    <div id=\"meetRoom\" style=\"width: 100%\" [ngClass]=\"{'hide': videoPanelVisible}\"></div>\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
RoomChatComponent.ctorParameters = () => [
    { type: VideoConferenceService }
];
RoomChatComponent.propDecorators = {
    roomName: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    RoomChatComponent.prototype.roomName;
    /** @type {?} */
    RoomChatComponent.prototype.videoPanelVisible;
    /**
     * @type {?}
     * @private
     */
    RoomChatComponent.prototype.videoService;
}
