// inet-calendar/esm2015/src/calendar-routing.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CalendarComponent } from "./calendar.component";
import { CalendarFormComponent } from "./form/calendar-form.component";
/** @type {?} */
const routes = [
    {
        path: 'calendar/v',
        children: [{
                path: '**',
                component: CalendarComponent
            }]
    },
    {
        path: 'calendar/event',
        children: [
            {
                path: ':id',
                component: CalendarFormComponent
            },
            {
                path: '**',
                redirectTo: 'create',
                pathMatch: 'full'
            }
        ]
    },
    {
        path: 'calendar/create',
        children: [
            {
                path: ':data',
                component: CalendarFormComponent
            },
            {
                path: '**',
                component: CalendarFormComponent
            }
        ]
    },
    {
        path: 'calendar',
        redirectTo: 'calendar/v',
        pathMatch: 'full'
    }
];
export class CalendarRoutingModule {
}
CalendarRoutingModule.decorators = [
    { type: NgModule, args: [{
                imports: [RouterModule.forChild(routes)],
                exports: [RouterModule]
            },] }
];
