// inet-calendar/esm2015/src/calendar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { CalendarService } from './calendar.service';
import { NotificationService, ResourceLoaderService } from 'inet-core';
import { xCalendar } from "./utils/xCalendar";
import { DateUtils } from "./utils/common/DateUtils";
import { Calendar } from '@fullcalendar/core';
import dayGridPlugin from '@fullcalendar/daygrid';
import timeGridPlugin from '@fullcalendar/timegrid';
import interactionPlugin from '@fullcalendar/interaction';
import * as moment_ from "moment";
import { filter } from "rxjs/operators";
import { CalendarTodayComponent } from "./calendar-today/calendar-today.component";
import { CalendarUtils } from "./CalendarUtils";
/** @type {?} */
const moment = moment_;
export class CalendarComponent {
    /**
     * @param {?} calendarService
     * @param {?} route
     * @param {?} router
     * @param {?} resourceLoader
     * @param {?} notify
     */
    constructor(calendarService, route, router, resourceLoader, notify) {
        this.calendarService = calendarService;
        this.route = route;
        this.router = router;
        this.resourceLoader = resourceLoader;
        this.notify = notify;
        this.viewTypes = [
            new CalendarViewMode('Ngày', CalendarMode.timeGridDay, {
                columnHeaderText: (/**
                 * @param {?} date
                 * @return {?}
                 */
                (date) => {
                    /** @type {?} */
                    let str = moment(date).format('dddd, D/M/Y');
                    return str.slice(0, 1).toUpperCase() + str.slice(1);
                }),
                titleFormat: (/**
                 * @return {?}
                 */
                () => CalendarUtils.displayDate(this.calendar.getDate()))
            }),
            new CalendarViewMode('Tuần', CalendarMode.timeGridWeek, {
                titleFormat: (/**
                 * @return {?}
                 */
                () => CalendarUtils.displayWeek(this.calendar.getDate()))
            }),
            new CalendarViewMode('Tháng', CalendarMode.dayGridMonth, {
                titleFormat: (/**
                 * @return {?}
                 */
                () => CalendarUtils.displayMonth(this.calendar.getDate()))
            })
        ];
        this.defaultMode = CalendarMode.dayGridMonth;
        this.calendarConfig = {
            plugins: [dayGridPlugin, timeGridPlugin, interactionPlugin],
            themeSystem: 'bootstrap',
            firstDay: 1,
            locale: 'vi',
            allDaySlot: true,
            allDayText: 'Cả ngày',
            header: false,
            nowIndicator: true,
            navLinks: true,
            // can click day/week names to navigate views
            editable: true,
            droppable: true,
            eventLimit: true,
            // allow "more" link when too many events
            selectable: true,
            selectMirror: true,
            weekends: true,
            views: {},
            businessHours: {
                daysOfWeek: [1, 2, 3, 4, 5],
                // Monday - Friday
                startTime: '00:01',
                endTime: '23:59'
            },
            eventDrop: (/**
             * @param {?} arg
             * @return {?}
             */
            (arg) => this._updateEventTime(arg)),
            eventResize: (/**
             * @param {?} arg
             * @return {?}
             */
            (arg) => this._updateEventTime(arg)),
            select: (/**
             * @param {?} arg
             * @return {?}
             */
            (arg) => this._selectRange(arg)),
            eventClick: (/**
             * @param {?} arg
             * @return {?}
             */
            (arg) => this._viewEvent(arg)),
            eventDragStart: (/**
             * @return {?}
             */
            () => this._hideViewDialog()),
            datesRender: (/**
             * @return {?}
             */
            () => this._calendarChange()),
            height: (/**
             * @return {?}
             */
            function () {
                return innerHeight - 59;
            })
        };
        this._modeCache = '__calendar_mode__';
        this._events = [];
        // Apply cache config
        Object.assign(this.calendarConfig, this._getDataView());
        // Merge view config
        this.viewTypes.forEach((/**
         * @param {?} view
         * @return {?}
         */
        (view) => {
            if (view.options) {
                this.calendarConfig.views[view.value] = view.options;
            }
        }));
        // Use locale
        moment.locale(this.calendarConfig.locale.toString());
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this._init();
        this._hideViewDialog();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.calendar.destroy();
        if (this._routerObserver) {
            this._routerObserver.unsubscribe();
        }
    }
    /**
     * @private
     * @return {?}
     */
    _init() {
        this.calendar = new Calendar(this.calendarRef.nativeElement, this.calendarConfig);
        this.calendar.render();
        // Change date view
        this._routerObserver = this.router.events.pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        event => event instanceof NavigationEnd)))
            .subscribe((/**
         * @param {?} event
         * @return {?}
         */
        (event) => this._onNavigated(event)));
    }
    /**
     * @private
     * @param {?} args
     * @return {?}
     */
    _updateEventTime(args) {
        /** @type {?} */
        let event;
        if (args.oldEvent) {
            event = this._events[Number(args.oldEvent.id)];
        }
        else {
            event = this._events[Number(args.event.id)];
        }
        if (event.mode !== 1) {
            /** @type {?} */
            let toTime;
            if (args.event.allDay) {
                toTime = 23 * 60 + 59;
            }
            else {
                toTime = args.event.end.getHours() * 60 + args.event.end.getMinutes();
            }
            /** @type {?} */
            let params = {
                element: event.uuid,
                year: args.event.start.getFullYear(),
                day: DateUtils.dayOfYear(args.event.start),
                startTime: args.event.start.getHours() * 60 + args.event.start.getMinutes(),
                toTime: toTime,
                _allday: args.event.allDay
            };
            if (event.repeat) {
                Object.assign(params, event.repeat);
            }
            this.calendarService.updateEvent(params, (/**
             * @param {?} data
             * @param {?} err
             * @return {?}
             */
            (data, err) => {
                if (!data) {
                    this.notify.showMessage('Cập nhật lịch không thành công: ' + err.status, 'error');
                    args.revert();
                    return;
                }
                this.reloadEvents();
            }));
        }
        else {
            // Can't change event
            args.revert();
        }
    }
    /**
     * @return {?}
     */
    createEvent() {
        /** @type {?} */
        let date = new Date();
        date.setHours(date.getHours() + 1, 0);
        this._createEvent(date, new Date(date.getTime() + 60 * 60 * 1000));
    }
    /**
     * @private
     * @param {?=} forceChange
     * @return {?}
     */
    _loadEvents(forceChange = false) {
        this._hideViewDialog();
        clearTimeout(this._eventTimer);
        this._eventTimer = setTimeout((/**
         * @return {?}
         */
        () => {
            this._loadRangeEvents(forceChange);
        }), 500);
    }
    /**
     * @private
     * @param {?} start
     * @param {?} end
     * @param {?=} allDay
     * @return {?}
     */
    _createEvent(start, end, allDay = false) {
        /** @type {?} */
        const dataBase64 = window.btoa(JSON.stringify({
            start: start.valueOf(),
            end: end.valueOf(),
            attribute: {
                allday: allDay + ''
            }
        }));
        this.router.navigate(['/calendar/create', dataBase64]);
        this._hideViewDialog();
    }
    /**
     * @private
     * @param {?} info
     * @return {?}
     */
    _selectRange(info) {
        this._createEvent(info.start, info.end, info.start.getTime() === (info.end.getTime() - 24 * 60 * 60 * 1000));
    }
    /**
     * @private
     * @return {?}
     */
    _calendarChange() {
        if (this.calendar.view.type === CalendarMode.dayGridDay) {
            this.calendar.changeView(CalendarMode.timeGridDay);
            // Trigger change
            setTimeout((/**
             * @return {?}
             */
            () => this._calendarChange()));
            return;
        }
        setTimeout((/**
         * @return {?}
         */
        () => {
            this._updateInfo();
            this._loadEvents();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    _updateInfo() {
        this.viewTitle = this.calendar.view.title;
        this.viewTypes.forEach((/**
         * @param {?} view
         * @return {?}
         */
        (view) => view.active = view.value === this.calendar.view.type));
        this._checkIsToday();
        this._updateViewParams();
    }
    /**
     * @private
     * @return {?}
     */
    _checkIsToday() {
        /** @type {?} */
        const range = {
            start: this.calendar.view.activeStart,
            end: this.calendar.view.activeEnd,
        };
        this.isToday = this._dateInRange(new Date(), range);
    }
    /**
     * @private
     * @param {?=} forceChange
     * @return {?}
     */
    _loadRangeEvents(forceChange) {
        if (forceChange || this._outOfRange()) {
            this._currentRange = this._getRange();
            this.calendarService.loadEvents(xCalendar.getRangeParams(this._currentRange.start, this._currentRange.end), (/**
             * @param {?} events
             * @return {?}
             */
            (events) => this._renderEvents(events || [])));
            this.calendar.removeAllEventSources();
        }
    }
    // Range view default is 6 week
    /**
     * @private
     * @return {?}
     */
    _getRange() {
        /** @type {?} */
        let start = new Date(this.calendar.getDate());
        /** @type {?} */
        let end;
        start.setDate(1);
        start.setHours(0, 0, 0, 0);
        while (start.getDay() !== this.calendarConfig.firstDay) {
            start.setDate(start.getDate() - 1);
        }
        end = new Date(start.getTime() + 6 * 7 * 24 * 60 * 60 * 1000);
        return {
            start: start,
            end: end
        };
    }
    /**
     * @private
     * @return {?}
     */
    _outOfRange() {
        if (!this._currentRange) {
            return true;
        }
        return !(this._dateInRange(this.calendar.view.currentStart) && this._dateInRange(this.calendar.view.currentEnd));
    }
    /**
     * @private
     * @param {?} date
     * @param {?=} range
     * @return {?}
     */
    _dateInRange(date, range) {
        range = range || this._currentRange;
        return date >= range.start && date <= range.end;
    }
    /**
     * @private
     * @param {?} events
     * @return {?}
     */
    _renderEvents(events) {
        this._events = events;
        /** @type {?} */
        let eventInputs = events.map((/**
         * @param {?} event
         * @param {?} index
         * @return {?}
         */
        (event, index) => {
            /** @type {?} */
            let eventInput = {
                id: index,
                uuid: event.uuid,
                allDay: event.allDay,
                title: event.subject,
                start: event.start,
                end: event.end,
            };
            if (event.end < new Date()) {
                // Event on the past
                eventInput.textColor = '#999';
                eventInput.color = '#ddd';
            }
            else if (event.isCreator) {
                // Owner event
                eventInput.textColor = '#fff';
                eventInput.color = '#4da0d8';
            }
            else {
                // Other event
                eventInput.textColor = '#333';
                eventInput.color = '#99ed96';
            }
            return eventInput;
        }));
        this._addEventSource(eventInputs);
    }
    /**
     * @private
     * @param {?} args
     * @return {?}
     */
    _viewEvent(args) {
        /** @type {?} */
        let event = this._events[Number(args.event.id)];
        this.calendarService.viewEvent(event, args.jsEvent, args.jsEvent.currentTarget, (/**
         * @return {?}
         */
        () => this.reloadEvents()));
    }
    /**
     * @private
     * @param {?} events
     * @return {?}
     */
    _addEventSource(events) {
        this.calendar.addEventSource({
            events: events
        });
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    _onNavigated(event) {
        /** @type {?} */
        const viewData = this._getDataView();
        this.calendar.changeView(viewData.defaultView, viewData.defaultDate);
    }
    /**
     * @param {?} type
     * @return {?}
     */
    changeView(type) {
        this.calendar.changeView(type.value);
    }
    /**
     * @return {?}
     */
    onPrev() {
        this.calendar.prev();
    }
    /**
     * @return {?}
     */
    onNext() {
        this.calendar.next();
    }
    /**
     * @return {?}
     */
    onToday() {
        this.calendar.changeView(CalendarMode.timeGridDay, new Date());
    }
    /**
     * @private
     * @param {?} mode
     * @return {?}
     */
    _getCalendarMode(mode) {
        for (let i = 0; i < this.viewTypes.length; i++) {
            if (this.viewTypes[i].value === mode) {
                return this.viewTypes[i];
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    _updateViewParams() {
        /** @type {?} */
        let params = [
            this.calendar.view.type,
            moment(this.calendar.getDate()).format('YYYY-M-D')
        ];
        this.router.navigate(['/calendar/v/' + params.join('/')]);
    }
    /**
     * @private
     * @return {?}
     */
    _getDataView() {
        /** @type {?} */
        let segments = this.route.snapshot.url;
        /** @type {?} */
        let modeView = this._cacheModeView();
        /** @type {?} */
        let date;
        if (segments.length > 1) {
            modeView = segments[0].path;
            date = new Date(segments[1].path);
        }
        if (!date || !date.getTime()) {
            date = new Date();
        }
        if (!this._getCalendarMode(modeView)) {
            modeView = this._getCalendarMode(this.defaultMode).value;
        }
        this._cacheModeView(modeView);
        return {
            defaultView: modeView,
            defaultDate: date,
        };
    }
    /**
     * @private
     * @param {?=} modeView
     * @return {?}
     */
    _cacheModeView(modeView) {
        if (modeView === undefined) {
            return localStorage.getItem(this._modeCache);
        }
        else {
            localStorage.setItem(this._modeCache, modeView);
        }
    }
    /**
     * @return {?}
     */
    reloadEvents() {
        this._hideViewDialog();
        this._loadEvents(true);
        this.calendarWidget.loadTodayEvents();
    }
    /**
     * @private
     * @return {?}
     */
    _hideViewDialog() {
        this.calendarService.hideViewModal();
    }
}
CalendarComponent.decorators = [
    { type: Component, args: [{
                selector: 'inet-calendar',
                template: "<div class=\"calendar-app\">\n    <div class=\"calendar-app__nav p-3\">\n        <div class=\"d-flex align-items-center\">\n            <div class=\"flex-grow-1\" style=\"font-size: 16px;font-weight: 500;line-height: 18px\">{{viewTitle}}</div>\n            <div class=\"btn-group\" role=\"group\">\n                <button (click)=\"onPrev()\" type=\"button\" class=\"btn btn-light\"><i class=\"fa fa-chevron-left\"></i></button>\n                <button (click)=\"onNext()\" type=\"button\" class=\"btn btn-light\"><i class=\"fa fa-chevron-right\"></i></button>\n            </div>\n        </div>\n        <div class=\"btn-group d-flex mt-3\" role=\"group\">\n            <button *ngFor=\"let type of viewTypes\" (click)=\"changeView(type)\"\n                    [ngClass]=\"{'btn-light': !type.active}\" type=\"button\" class=\"btn btn-primary flex-grow-1\">\n                {{type.name}}\n            </button>\n        </div>\n        <button (click)=\"onToday()\" [ngClass]=\"{'btn-light': !isToday}\" type=\"button\" class=\"btn btn-primary w-100 mt-3\">H\u00F4m nay</button>\n        <button (click)=\"createEvent()\" type=\"button\" class=\"btn btn-primary w-100 mt-3\">T\u1EA1o s\u1EF1 ki\u1EC7n</button>\n        <!-- <div (click)=\"createEvent()\" class=\"calendar-add-btn mt-3 pt-2 pb-2 text-primary\">\n            <i class=\"fa fa-plus mr-1\"></i> Th\u00EAm s\u1EF1 ki\u1EC7n\n        </div> -->\n        <div class=\"mt-3\" style=\"margin-left: -15px;margin-right: -15px;\">\n            <div class=\"ml-3 mr-3 pb-1 mb-1\" style=\"border-bottom: 2px solid #007bff;font-weight: bold;\">S\u1EF1 ki\u1EC7n h\u00F4m nay</div>\n            <calendar-widget-today (onChange)=\"reloadEvents()\"></calendar-widget-today>\n        </div>\n    </div>\n    <div class=\"calendar-app__body pr-1\">\n        <div #calendarRef></div>\n    </div>\n</div>\n",
                styles: [".calendar-app{width:100%;height:calc(100vh - 60px);background:#fff;padding:0}.calendar-app__nav{height:100%;float:left;width:300px}.calendar-app__body{height:100%;overflow:hidden;background:rgba(0,0,0,.01)}.calendar-add-btn{cursor:pointer;font-weight:700}.calendar-add-btn:hover{opacity:.8}"]
            }] }
];
/** @nocollapse */
CalendarComponent.ctorParameters = () => [
    { type: CalendarService },
    { type: ActivatedRoute },
    { type: Router },
    { type: ResourceLoaderService },
    { type: NotificationService }
];
CalendarComponent.propDecorators = {
    calendarRef: [{ type: ViewChild, args: ['calendarRef',] }],
    calendarWidget: [{ type: ViewChild, args: [CalendarTodayComponent,] }]
};
if (false) {
    /** @type {?} */
    CalendarComponent.prototype.calendarRef;
    /** @type {?} */
    CalendarComponent.prototype.calendarWidget;
    /** @type {?} */
    CalendarComponent.prototype.viewTypes;
    /** @type {?} */
    CalendarComponent.prototype.defaultMode;
    /** @type {?} */
    CalendarComponent.prototype.viewTitle;
    /** @type {?} */
    CalendarComponent.prototype.calendarConfig;
    /** @type {?} */
    CalendarComponent.prototype.calendar;
    /** @type {?} */
    CalendarComponent.prototype.isToday;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype._eventTimer;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype._modeCache;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype._routerObserver;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype._events;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype._currentRange;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype.calendarService;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype.route;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype.resourceLoader;
    /**
     * @type {?}
     * @private
     */
    CalendarComponent.prototype.notify;
}
export class CalendarViewMode {
    /**
     * @param {?} name
     * @param {?} value
     * @param {?=} options
     */
    constructor(name, value, options) {
        this.active = false;
        this.name = name;
        this.value = value;
        this.options = options;
    }
}
if (false) {
    /** @type {?} */
    CalendarViewMode.prototype.name;
    /** @type {?} */
    CalendarViewMode.prototype.value;
    /** @type {?} */
    CalendarViewMode.prototype.options;
    /** @type {?} */
    CalendarViewMode.prototype.active;
}
/** @enum {string} */
const CalendarMode = {
    dayGridDay: 'dayGridDay',
    timeGridDay: 'timeGridDay',
    timeGridWeek: 'timeGridWeek',
    dayGridMonth: 'dayGridMonth',
};
export { CalendarMode };
