// inet-social/esm5/inet-social.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { SocialModule, SocialService, ActivityComponent, ActivityListComponent, ActivityPostComponent, ActivityViewComponent, SocialEmbedComponent, SocialGroupComponent, FriendBrowseComponent, FriendInvitationComponent, SocialComponent, SocialLatestComponent, SocialAppModule } from './index';
export { ActivityActionMoreComponent as ɵn } from './src/activity-action-more/activity-action-more.component';
export { AttachmentListComponent as ɵh } from './src/attachment-list/attachment-list.component';
export { CommentListComponent as ɵk } from './src/comment-list/comment-list.component';
export { CommentPostComponent as ɵj } from './src/comment-post/comment-post.component';
export { CommentComponent as ɵi } from './src/comment/comment.component';
export { FileExtPipe as ɵd } from './src/filelist/file-ext.pipe';
export { FileFormatService as ɵc } from './src/filelist/file-format.service';
export { FileIconPipe as ɵb } from './src/filelist/file-icon.pipe';
export { FileListModule as ɵa } from './src/filelist/file-list.module';
export { FileListComponent as ɵe } from './src/filelist/list/file-list.component';
export { LocalNews as ɵm } from './src/localnews/localnews.component';
export { MessageBodyComponent as ɵl } from './src/message-body/message-body.component';
export { SocialModalViewComponent as ɵo } from './src/modal-view/modal-view.component';
export { ProgressComponent as ɵf } from './src/progress/progress.component';
export { SocialActivityViewComponent as ɵu } from './src/routing/activity-view/social-activity-view.component';
export { SocialFriendComponent as ɵv } from './src/routing/friend/social-friend.component';
export { SocialHomeComponent as ɵt } from './src/routing/home/social-home.component';
export { SocialRightLayoutComponent as ɵw } from './src/routing/right-layout/social-right-layout.component';
export { SocialRoutingModule as ɵs } from './src/routing/social-routing.module';
export { ShareMemberComponent as ɵg } from './src/share-member/share-member.component';
export { GroupCreateComponent as ɵq } from './src/social-friend/group-create/group-create.component';
export { GroupMemberComponent as ɵr } from './src/social-friend/group-member/group-member.component';
export { ListUserComponent as ɵp } from './src/social-friend/list-user/list-user.component';
