// inet-chat/esm5/src/DateFormatter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as _moment from 'moment';
/** @type {?} */
var moment = _moment;
var DateFormatter = /** @class */ (function () {
    function DateFormatter(date) {
        this.date = new Date(date);
    }
    /**
     * @return {?}
     */
    DateFormatter.prototype.formatFullYear = /**
     * @return {?}
     */
    function () {
        return moment(this.date).format('DD/MM/YYYY, H:mm');
    };
    /**
     * @return {?}
     */
    DateFormatter.prototype.formatYear = /**
     * @return {?}
     */
    function () {
        return moment(this.date).format('DD/MM/YYYY');
    };
    /**
     * @return {?}
     */
    DateFormatter.prototype.formatMonth = /**
     * @return {?}
     */
    function () {
        return moment(this.date).format('DD/MM, H:mm');
    };
    /**
     * @return {?}
     */
    DateFormatter.prototype.formatDate = /**
     * @return {?}
     */
    function () {
        return moment(this.date).format('DD/MM, H:mm');
    };
    /**
     * @return {?}
     */
    DateFormatter.prototype.formatTime = /**
     * @return {?}
     */
    function () {
        return moment(this.date).format('H:mm');
    };
    /**
     * @return {?}
     */
    DateFormatter.prototype.format = /**
     * @return {?}
     */
    function () {
        if (!this.date.getTime()) {
            return '';
        }
        /** @type {?} */
        var now = new Date();
        /** @type {?} */
        var conditions = {
            onYear: now.getFullYear() === this.date.getFullYear(),
            onMonth: now.getMonth() === this.date.getMonth(),
            onDate: now.getDate() === this.date.getDate(),
        };
        if (conditions.onYear && conditions.onMonth && conditions.onDate) {
            return this.formatTime();
        }
        if (conditions.onYear && conditions.onMonth) {
            return this.formatDate();
        }
        if (conditions.onYear) {
            return this.formatMonth();
        }
        return this.formatYear();
    };
    return DateFormatter;
}());
export { DateFormatter };
if (false) {
    /** @type {?} */
    DateFormatter.prototype.date;
}
