// inet-chat/esm5/src/message.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService, WebSocketClientService } from "inet-core";
import { APP_CHAT } from "./model/MessageConstants";
import { MessageContact } from "./model/MessageContact";
import { Observable, Subject } from "rxjs";
import { MessageChat } from "./model/MessageChat";
var ChatMessageService = /** @class */ (function () {
    function ChatMessageService(httpService, socketService) {
        var _this = this;
        this.httpService = httpService;
        this.socketService = socketService;
        this._messageSubject = new Subject();
        this._unreadSubject = new Subject();
        this.contacts = [];
        this.messageReceived = this._messageSubject.asObservable();
        this.unreadMessageChange = this._unreadSubject.asObservable();
        if (ChatMessageService._instance) {
            return ChatMessageService._instance;
        }
        ChatMessageService._instance = this;
        this._socketObserver = this.socketService.onMessage.subscribe((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            if (message.application === APP_CHAT) {
                _this.updateLastMessage(message.sender, message);
                _this._messageSubject.next(new MessageChat(message));
            }
        }));
    }
    /**
     * @return {?}
     */
    ChatMessageService.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this._socketObserver.unsubscribe();
        this._messageSubject.unsubscribe();
        this._unreadSubject.unsubscribe();
    };
    /**
     * @return {?}
     */
    ChatMessageService.prototype.loadFriends = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return new Observable((/**
         * @param {?} observer
         * @return {?}
         */
        function (observer) {
            _this.httpService.getJSON(iNet.getPUrl('unifiedpush/list')).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                observer.next((data || []).map((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return new MessageContact(item); })));
                observer.complete();
            }));
        }));
    };
    /**
     * @param {?} callback
     * @return {?}
     */
    ChatMessageService.prototype.loadContacts = /**
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        var _this = this;
        if (this.contacts.length > 0) {
            callback(this.contacts);
        }
        else {
            this.loadFriends().subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this.contacts = data.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item.usercode !== iNet.usercode; }));
                _this.updateUnreadMessage();
                callback(_this.contacts);
            }), (/**
             * @return {?}
             */
            function () { return callback([]); }));
        }
    };
    /**
     * @return {?}
     */
    ChatMessageService.prototype.updateUnreadMessage = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var unread = 0;
        this.contacts.forEach((/**
         * @param {?} friend
         * @return {?}
         */
        function (friend) { return unread += friend.unread; }));
        this._unreadSubject.next(unread);
    };
    /**
     * @param {?} usercode
     * @param {?} callback
     * @return {?}
     */
    ChatMessageService.prototype.getContactByUserCode = /**
     * @param {?} usercode
     * @param {?} callback
     * @return {?}
     */
    function (usercode, callback) {
        var _this = this;
        if (usercode === 'news-feed' || usercode === iNet.usercode) {
            return;
        }
        this.loadContacts((/**
         * @param {?} contacts
         * @return {?}
         */
        function (contacts) {
            for (var i = 0; i < contacts.length; i++) {
                if (contacts[i].usercode === usercode) {
                    return callback(contacts[i]);
                }
            }
            // contact not exist
            /** @type {?} */
            var contact = new MessageContact({ usercode: usercode, fullname: usercode, unread: 1 });
            _this.contacts.unshift(contact);
            callback(contact);
        }));
    };
    /**
     * @param {?} usercode
     * @return {?}
     */
    ChatMessageService.prototype.hasContact = /**
     * @param {?} usercode
     * @return {?}
     */
    function (usercode) {
        for (var i = 0; i < this.contacts.length; i++) {
            if (this.contacts[i].usercode === usercode) {
                return true;
            }
        }
        return false;
    };
    /**
     * @param {?} username
     * @param {?} message
     * @return {?}
     */
    ChatMessageService.prototype.send = /**
     * @param {?} username
     * @param {?} message
     * @return {?}
     */
    function (username, message) {
        return this.socketService.send(username, message, APP_CHAT);
    };
    /**
     * @param {?} envelop
     * @return {?}
     */
    ChatMessageService.prototype.sendEnvelop = /**
     * @param {?} envelop
     * @return {?}
     */
    function (envelop) {
        return this.socketService.sendEnvelop(envelop);
    };
    /**
     * @param {?} address
     * @param {?} message
     * @param {?=} join
     * @return {?}
     */
    ChatMessageService.prototype.buildEnvelop = /**
     * @param {?} address
     * @param {?} message
     * @param {?=} join
     * @return {?}
     */
    function (address, message, join) {
        if (join) {
            return this.socketService.buildEnvelop(address, message, APP_CHAT, join);
        }
        else {
            return this.socketService.buildEnvelop(address, message, APP_CHAT);
        }
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    ChatMessageService.prototype.pushMessage = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        /** @type {?} */
        var formData = this.httpService.convertToFormData(params);
        this.httpService.post(iNet.getPUrl('message/upload'), formData).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var envelop = _this.socketService.buildEnvelop(params.alias, params.content, APP_CHAT);
            envelop.body.message = data.message;
            envelop.body.message.content = envelop.body.message.alert = params.content;
            envelop.body.uuid = data.uuid;
            _this.socketService.sendEnvelop(envelop);
            callback(envelop.body);
        }));
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    ChatMessageService.prototype.loadMessages = /**
     * @param {?} params
     * @param {?} callback
     * @return {?}
     */
    function (params, callback) {
        var _this = this;
        this.httpService.postJSON(iNet.getPUrl('message/query'), params).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            data = data || {};
            data.items = data.items || [];
            data.total = data.total || 0;
            data.items.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) { return a.sent - b.sent; }));
            data.items = data.items.map((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return new MessageChat(item); }));
            _this.resetUnreadMessage(params['sender']);
            callback(data);
        }));
    };
    /**
     * @param {?} usercode
     * @param {?} message
     * @return {?}
     */
    ChatMessageService.prototype.updateLastMessage = /**
     * @param {?} usercode
     * @param {?} message
     * @return {?}
     */
    function (usercode, message) {
        var _this = this;
        // Update last message, move contact to first
        this.getContactByUserCode(usercode, (/**
         * @param {?} contact
         * @return {?}
         */
        function (contact) {
            /** @type {?} */
            var index = _this.contacts.indexOf(contact);
            if (index > -1) {
                _this.contacts.splice(index, 1);
            }
            _this.contacts.unshift(contact);
            /** @type {?} */
            var typeMessages = message.message.alert.split(":");
            if (typeMessages.length > 2) {
                if (typeMessages[0] === 'uri_message' && typeMessages[1] === 'video_conference') {
                    contact.message = '<i class="fa fa-phone pr-1" aria-hidden="true"></i> Cuộc gọi thoại';
                }
                else {
                    /** @type {?} */
                    var __mesDisplay = _this.revertMessageDisplay(message.message.alert) || '';
                    if (typeMessages[0] === 'uri_message' && typeMessages[1] === 'link_preview') {
                        contact.message = __mesDisplay;
                    }
                }
            }
            else {
                contact.message = message.message.alert;
            }
            // contact.message = message.message.alert;
            // Not me send
            if (message.sender !== iNet.usercode) {
                contact.unread++;
                _this.updateUnreadMessage();
            }
        }));
    };
    /**
     * @param {?} str
     * @return {?}
     */
    ChatMessageService.prototype.decodeBase64 = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return decodeURIComponent(escape(window.atob(str)));
    };
    /**
     * @param {?} username
     * @return {?}
     */
    ChatMessageService.prototype.resetUnreadMessage = /**
     * @param {?} username
     * @return {?}
     */
    function (username) {
        var _this = this;
        this.getContactByUserCode(username, (/**
         * @param {?} contact
         * @return {?}
         */
        function (contact) {
            contact.unread = 0;
            _this.updateUnreadMessage();
        }));
    };
    /**
     * @param {?} keyword
     * @param {?} callback
     * @return {?}
     */
    ChatMessageService.prototype.searchContacts = /**
     * @param {?} keyword
     * @param {?} callback
     * @return {?}
     */
    function (keyword, callback) {
        var _this = this;
        this.loadContacts((/**
         * @param {?} contacts
         * @return {?}
         */
        function (contacts) { return callback(_this._filterUsers(keyword, contacts)); }));
    };
    /**
     * @private
     * @param {?} keyword
     * @param {?} contacts
     * @return {?}
     */
    ChatMessageService.prototype._filterUsers = /**
     * @private
     * @param {?} keyword
     * @param {?} contacts
     * @return {?}
     */
    function (keyword, contacts) {
        if (!keyword) {
            return contacts;
        }
        return contacts.filter((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.isMatch(keyword); }));
    };
    /**
     * @param {?} message
     * @return {?}
     */
    ChatMessageService.prototype.revertMessage = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        /** @type {?} */
        var __index = message.indexOf("uri_message") || 0;
        return message.substring(__index);
    };
    /**
     * @param {?} message
     * @return {?}
     */
    ChatMessageService.prototype.revertMessageDisplay = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        /** @type {?} */
        var __index = message.indexOf("uri_message") || 0;
        return message.substring(0, __index);
    };
    /**
     * @return {?}
     */
    ChatMessageService.prototype.reconnetSocket = /**
     * @return {?}
     */
    function () {
        this.socketService.close();
    };
    /**
     * @param {?} message
     * @return {?}
     */
    ChatMessageService.prototype.checkJoinVideoCall = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
    };
    ChatMessageService._instance = null;
    ChatMessageService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ChatMessageService.ctorParameters = function () { return [
        { type: HttpClientService },
        { type: WebSocketClientService }
    ]; };
    return ChatMessageService;
}());
export { ChatMessageService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ChatMessageService._instance;
    /**
     * @type {?}
     * @private
     */
    ChatMessageService.prototype._socketObserver;
    /**
     * @type {?}
     * @private
     */
    ChatMessageService.prototype._messageSubject;
    /**
     * @type {?}
     * @private
     */
    ChatMessageService.prototype._unreadSubject;
    /** @type {?} */
    ChatMessageService.prototype.contacts;
    /** @type {?} */
    ChatMessageService.prototype.messageReceived;
    /** @type {?} */
    ChatMessageService.prototype.unreadMessageChange;
    /**
     * @type {?}
     * @private
     */
    ChatMessageService.prototype.httpService;
    /** @type {?} */
    ChatMessageService.prototype.socketService;
}
