// inet-chat/esm5/src/video-conference.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ErrorMessage, HttpClientService } from "inet-core";
var VideoConferenceService = /** @class */ (function () {
    function VideoConferenceService(httpService) {
        this.httpService = httpService;
        this.domain = 'jitsi.inetcloud.vn';
        this.messageJoin = { app: 'iChat', conferenceID: '', sender: '', joined: '', inAccess: '' };
        return VideoConferenceService.instance = VideoConferenceService.instance || this;
    }
    /**
     * @param {?} parentNode
     * @param {?} roomName
     * @param {?=} audioOnly
     * @param {?=} height
     * @param {?=} callback
     * @return {?}
     */
    VideoConferenceService.prototype.join = /**
     * @param {?} parentNode
     * @param {?} roomName
     * @param {?=} audioOnly
     * @param {?=} height
     * @param {?=} callback
     * @return {?}
     */
    function (parentNode, roomName, audioOnly, height, callback) {
        var _this = this;
        if (audioOnly === void 0) { audioOnly = false; }
        if (height === void 0) { height = 350; }
        if (!JitsiMeetExternalAPI) {
            return;
        }
        /** @type {?} */
        var tokenId = '';
        this.httpService.postJSON(iNet.getPUrl('conference/authorized'), { roomID: roomName }).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            tokenId = data.webdata || '';
            _this.dispose();
            /** @type {?} */
            var options = {
                roomName: roomName,
                width: '100%',
                height: '100%',
                parentNode: parentNode,
                jwt: tokenId
            };
            if (audioOnly) {
                options.height = '100%';
                options['configOverwrite'] = {
                    startAudioOnly: true,
                    startWithVideoMuted: true
                };
            }
            _this.api = new JitsiMeetExternalAPI(_this.domain, options);
            if (callback) {
                callback(_this.api);
            }
            _this.api.executeCommand('sendTones', {
                tones: '12345#',
                // The dial pad touch tones to play. For example, '12345#'.
                duration: 200,
                // Optional. The number of milliseconds each tone should play. The default is 200.
                pause: 200 // Optional. The number of milliseconds between each tone. The default is 200.
            });
            // this.api.executeCommand('displayName', iNet.displayName);
            // this.api.executeCommand('email', iNet.username);
            // this.api.executeCommand('setVideoQuality', 720);
        }));
        /*
        this.api.addEventListener('readyToClose',  ()=> {
            this.dispose();
        });

         */
    };
    /**
     * @param {?} parentNode
     * @param {?} userName
     * @param {?=} audioOnly
     * @param {?=} height
     * @param {?=} callback
     * @return {?}
     */
    VideoConferenceService.prototype.callTo = /**
     * @param {?} parentNode
     * @param {?} userName
     * @param {?=} audioOnly
     * @param {?=} height
     * @param {?=} callback
     * @return {?}
     */
    function (parentNode, userName, audioOnly, height, callback) {
        var _this = this;
        if (audioOnly === void 0) { audioOnly = false; }
        if (height === void 0) { height = 350; }
        /** @type {?} */
        var conferenceID = '';
        this.httpService.postJSON(iNet.getPUrl('conference/end2end'), { callTo: userName }).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (data.type !== ErrorMessage.TYPE) {
                conferenceID = data.conferenceID || '';
                if (!iNet.isEmpty(conferenceID)) {
                    if (!JitsiMeetExternalAPI) {
                        return;
                    }
                    /** @type {?} */
                    var tokenId_1 = '';
                    /** @type {?} */
                    var messageJSON = { sender: iNet.username, conferenceID: conferenceID };
                    /** @type {?} */
                    var __messageStr_1 = 'uri_message:video_conference:' + _this.encodeBase64(JSON.stringify(messageJSON)) || '';
                    // this.chatMessageService.send(userName,__messageStr);
                    _this.httpService.postJSON(iNet.getPUrl('conference/authorized'), { roomID: conferenceID }).subscribe((/**
                     * @param {?} data
                     * @return {?}
                     */
                    function (data) {
                        tokenId_1 = data.webdata || '';
                        _this.dispose();
                        /** @type {?} */
                        var options = {
                            roomName: conferenceID,
                            width: '100%',
                            height: '100%',
                            parentNode: parentNode,
                            jwt: tokenId_1
                        };
                        if (audioOnly) {
                            options.height = '100%';
                            options['configOverwrite'] = {
                                startAudioOnly: true,
                                startWithVideoMuted: true
                            };
                        }
                        _this.api = new JitsiMeetExternalAPI(_this.domain, options);
                        _this.messageJoin.sender = iNet.username;
                        _this.messageJoin.conferenceID = conferenceID;
                        if (callback) {
                            callback(_this.api, __messageStr_1);
                        }
                        _this.api.executeCommand('sendTones', {
                            tones: '12345#',
                            // The dial pad touch tones to play. For example, '12345#'.
                            duration: 200,
                            // Optional. The number of milliseconds each tone should play. The default is 200.
                            pause: 200 // Optional. The number of milliseconds between each tone. The default is 200.
                        });
                        // this.api.executeCommand('displayName', iNet.displayName);
                        // this.api.executeCommand('email', iNet.username);
                        // this.api.executeCommand('setVideoQuality', 720);
                    }));
                }
            }
        }));
        /*
        this.api.addEventListener('readyToClose',  ()=> {
            this.dispose();
        });

         */
    };
    /**
     * @return {?}
     */
    VideoConferenceService.prototype.getApi = /**
     * @return {?}
     */
    function () {
        return this.api;
    };
    /**
     * @return {?}
     */
    VideoConferenceService.prototype.dispose = /**
     * @return {?}
     */
    function () {
        if (this.api) {
            this.api.dispose();
            delete this.api;
        }
    };
    /**
     * @return {?}
     */
    VideoConferenceService.prototype.getMessageJoin = /**
     * @return {?}
     */
    function () {
        return this.messageJoin;
    };
    /**
     * @param {?} receiver
     * @return {?}
     */
    VideoConferenceService.prototype.getRoomNameByReceiver = /**
     * @param {?} receiver
     * @return {?}
     */
    function (receiver) {
        return iNet.username.split('@')[0] + '_' + receiver.split('@')[0];
    };
    /**
     * @return {?}
     */
    VideoConferenceService.prototype.isExistMeet = /**
     * @return {?}
     */
    function () {
        return !!JitsiMeetExternalAPI;
    };
    /**
     * @param {?} str
     * @return {?}
     */
    VideoConferenceService.prototype.encodeBase64 = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return window.btoa(unescape(encodeURIComponent(str)));
    };
    VideoConferenceService.instance = null;
    VideoConferenceService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    VideoConferenceService.ctorParameters = function () { return [
        { type: HttpClientService }
    ]; };
    return VideoConferenceService;
}());
export { VideoConferenceService };
if (false) {
    /** @type {?} */
    VideoConferenceService.instance;
    /** @type {?} */
    VideoConferenceService.prototype.api;
    /** @type {?} */
    VideoConferenceService.prototype.domain;
    /** @type {?} */
    VideoConferenceService.prototype.messageJoin;
    /**
     * @type {?}
     * @private
     */
    VideoConferenceService.prototype.httpService;
}
