// inet-social/esm5/src/activity-view/activity-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { SocialService } from '../social.service';
var ActivityViewComponent = /** @class */ (function () {
    function ActivityViewComponent(socialService) {
        this.socialService = socialService;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ActivityViewComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.activityId.currentValue !== changes.activityId.previousValue) {
            this._load();
        }
    };
    /**
     * @param {?} activity
     * @return {?}
     */
    ActivityViewComponent.prototype.activityUpdated = /**
     * @param {?} activity
     * @return {?}
     */
    function (activity) {
        this.activity = activity;
    };
    /**
     * @private
     * @return {?}
     */
    ActivityViewComponent.prototype._load = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.notFound = false;
        this.socialService.activityLoad({
            activity: this.activityId
        }, (/**
         * @param {?} activity
         * @return {?}
         */
        function (activity) {
            if (activity && activity.uuid) {
                _this.activity = activity;
            }
            else {
                _this.notFound = true;
            }
        }));
    };
    ActivityViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'social-activity-view',
                    template: "<div *ngIf=\"activity\" socialActivity [activity]=\"activity\" (onUpdated)=\"activityUpdated($event)\"></div>\n<h4 *ngIf=\"notFound\" class=\"text-center\" style=\"margin-top:70px\">Kh\u00F4ng t\u00ECm th\u1EA5y tin</h4>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ActivityViewComponent.ctorParameters = function () { return [
        { type: SocialService }
    ]; };
    ActivityViewComponent.propDecorators = {
        activityId: [{ type: Input }]
    };
    return ActivityViewComponent;
}());
export { ActivityViewComponent };
if (false) {
    /** @type {?} */
    ActivityViewComponent.prototype.activityId;
    /** @type {?} */
    ActivityViewComponent.prototype.activity;
    /** @type {?} */
    ActivityViewComponent.prototype.notFound;
    /**
     * @type {?}
     * @private
     */
    ActivityViewComponent.prototype.socialService;
}
