// inet-calendar/esm2015/src/dialog-delete/dialog-delete.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { CalendarService } from "../calendar.service";
import { ModalDirective } from "ngx-bootstrap";
export class CalendarDialogDelete {
    /**
     * @param {?} calendarService
     */
    constructor(calendarService) {
        this.calendarService = calendarService;
        this.onDelete = new EventEmitter();
        this.dialogTitle = 'Lịch công tác';
        this.dialogContent = '';
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this._initDialogDelete();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._destroyDialogDelete();
    }
    /**
     * @return {?}
     */
    hide() {
        this.confirmDialog.hide();
    }
    /**
     * @return {?}
     */
    delete() {
        this.calendarService.removeEvent({
            element: this.calEvent.uuid,
            orgid: this.calEvent.firmUUID
        }, (/**
         * @return {?}
         */
        () => {
            this.hide();
            this.onDelete.emit();
        }));
    }
    /**
     * @return {?}
     */
    show() {
        this.dialogContent = 'Bạn có đồng ý xóa sự kiện <b>"' + this.calEvent.subject + '"</b> không?';
        this.confirmDialog.show();
    }
    /**
     * @private
     * @return {?}
     */
    _initDialogDelete() {
        this._$dialogDelete = $(this.dialogDelete.nativeElement).hide();
        // Click outside to close
        this.hide = this.hide.bind(this);
        document.body.addEventListener('click', this.hide);
        // Prevent close
        this._$dialogDelete.on('click', (/**
         * @param {?} e
         * @return {?}
         */
        (e) => e.stopPropagation()));
        // Append modal to body
        $(document.body).append(this._$dialogDelete);
    }
    /**
     * @private
     * @return {?}
     */
    _destroyDialogDelete() {
        document.body.removeEventListener('click', this.hide);
        this._$dialogDelete.remove();
    }
}
CalendarDialogDelete.decorators = [
    { type: Component, args: [{
                selector: 'calendar-dialog-delete',
                template: "<div #dialogDelete class=\"modal fade\" bsModal #confirmDialog=\"bs-modal\" tabindex=\"-1\" role=\"dialog\" aria-hidden=\"true\">\n    <div class=\"modal-dialog\" role=\"document\">\n        <div class=\"modal-content\">\n            <div class=\"modal-header\">\n                <h4 class=\"modal-title pull-left\"> {{dialogTitle}}</h4>\n                <button type=\"button\" class=\"close pull-right\" aria-label=\"Close\" (click)=\"hide()\">\n                    <span aria-hidden=\"true\">&times;</span>\n                </button>\n            </div>\n            <div class=\"modal-body\" [innerHTML]=\"dialogContent\">\n            </div>\n            <div class=\"modal-footer\">\n                <div class=\"text-right\">\n                    <button type=\"button\" class=\"btn btn-lg btn-danger btn-sm\" (click)=\"delete()\">\n                        <i class=\"fa fa-trash\"></i> X\u00F3a\n                    </button>\n                    <button type=\"button\" class=\"btn btn-lg btn-default ml-2 btn-sm\" (click)=\"hide()\">\n                        <i class=\"fa fa-times\"></i> \u0110\u00F3ng\n                    </button>\n                </div>\n            </div>\n        </div>\n    </div>\n</div>",
                styles: [""]
            }] }
];
/** @nocollapse */
CalendarDialogDelete.ctorParameters = () => [
    { type: CalendarService }
];
CalendarDialogDelete.propDecorators = {
    calEvent: [{ type: Input }],
    onDelete: [{ type: Output }],
    confirmDialog: [{ type: ViewChild, args: [ModalDirective,] }],
    dialogDelete: [{ type: ViewChild, args: ['dialogDelete',] }]
};
if (false) {
    /** @type {?} */
    CalendarDialogDelete.prototype.calEvent;
    /** @type {?} */
    CalendarDialogDelete.prototype.onDelete;
    /** @type {?} */
    CalendarDialogDelete.prototype.confirmDialog;
    /** @type {?} */
    CalendarDialogDelete.prototype.dialogDelete;
    /** @type {?} */
    CalendarDialogDelete.prototype.dialogTitle;
    /** @type {?} */
    CalendarDialogDelete.prototype.dialogContent;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogDelete.prototype._$dialogDelete;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogDelete.prototype.calendarService;
}
