// inet-social/esm2015/src/social-friend/group/group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { SocialService } from "../../social.service";
import { NavigationEnd, Router } from "@angular/router";
import { GroupMemberComponent } from "../group-member/group-member.component";
import { GroupCreateComponent } from "../group-create/group-create.component";
import { BsModalService } from 'ngx-bootstrap';
import { AccentService } from 'inet-core';
import { filter } from "rxjs/operators";
export class SocialGroupComponent {
    /**
     * @param {?} socialService
     * @param {?} accentService
     * @param {?} router
     * @param {?} modalService
     */
    constructor(socialService, accentService, router, modalService) {
        this.socialService = socialService;
        this.accentService = accentService;
        this.router = router;
        this.modalService = modalService;
        this.friends = [];
        this.members = [];
        this._prefix = '/social/friend';
        this.placeholder = 'Tìm kiếm...';
        this._loadAllFriends();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onSearch(value) {
        this.searching = !!value;
        if (value) {
            value = this.accentService.viToEn(value);
            if (this.viewGroups) {
                this.friends.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    item.hidden = item._index.indexOf(value) < 0;
                }));
            }
            else {
                this.members.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => {
                    item.hidden = item._index.indexOf(value) < 0;
                }));
            }
        }
        else {
            if (this.viewGroups) {
                this.friends.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => delete item.hidden));
            }
            else {
                this.members.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                item => delete item.hidden));
            }
        }
    }
    /**
     * @return {?}
     */
    clearSearchValue() { }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._router = this.router.events
            .pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        event => event instanceof NavigationEnd)))
            .subscribe((/**
         * @return {?}
         */
        () => {
            this._actionView();
        }));
        this._actionView();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._router.unsubscribe();
    }
    /**
     * @return {?}
     */
    createGroup() {
        this.modalCreateMember.open();
    }
    /**
     * @return {?}
     */
    groupCreated() {
        this._loadGroups();
        this._loadAllFriends();
    }
    /**
     * @param {?} group
     * @return {?}
     */
    viewGroup(group) {
        this.router.navigate([this._prefix + '/' + group]);
    }
    /**
     * @param {?} group
     * @param {?} template
     * @return {?}
     */
    removeGroup(group, template) {
        this.groupActive = group;
        this.modalGroupDeleteRef = this.modalService.show(template);
    }
    // Todo: improve remove group
    /**
     * @return {?}
     */
    onDeleteGroup() {
        this.modalGroupDeleteRef.hide();
        this._loadFriends(this.groupActive, (/**
         * @param {?} members
         * @return {?}
         */
        (members) => {
            this.socialService.groupMemberDelete({
                category: this.groupActive,
                member: this._getMemberStr(members)
            }, (/**
             * @return {?}
             */
            () => {
                this._loadGroups();
                this._loadAllFriends();
                if (!this.viewGroups) {
                    this.toGroups();
                }
            }));
        }));
        this.groups.splice(this.groups.indexOf(this.groupActive), 1);
    }
    /**
     * @param {?} friend
     * @param {?} template
     * @return {?}
     */
    removeFriend(friend, template) {
        this.modalMemberDeleteRef = this.modalService.show(template);
        this.memberActive = friend;
    }
    /**
     * @return {?}
     */
    onDeleteMember() {
        /** @type {?} */
        const promises = [];
        if (this.memberActive.categories) {
            // Member on groups
            this.memberActive.categories.forEach((/**
             * @param {?} category
             * @return {?}
             */
            category => {
                promises.push(this._deleteMemberPromise(this.memberActive.usercode, category));
            }));
        }
        else {
            promises.push(this._deleteMemberPromise(this.memberActive.usercode, this.memberActive.category));
        }
        Promise.all(promises).then((/**
         * @param {?} values
         * @return {?}
         */
        values => {
            if (this.groupView) {
                this._loadMembers();
            }
            this._loadAllFriends();
        }));
        this.modalMemberDeleteRef.hide();
    }
    /**
     * @private
     * @param {?} usercode
     * @param {?} category
     * @return {?}
     */
    _deleteMemberPromise(usercode, category) {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => {
            this.socialService.groupMemberDelete({
                category: category,
                member: usercode
            }, resolve);
        }));
    }
    /**
     * @return {?}
     */
    addMemberGroup() {
        this.modalAddMember.open();
    }
    /**
     * @param {?} members
     * @return {?}
     */
    saveMemberGroup(members) {
        this.socialService.groupMemberCreate({
            category: this.groupView,
            member: this._getMemberStr(members)
        }, (/**
         * @return {?}
         */
        () => {
            this._loadMembers();
            this._loadAllFriends();
        }));
    }
    /**
     * @return {?}
     */
    toGroups() {
        this.router.navigate([this._prefix]);
    }
    /**
     * @private
     * @return {?}
     */
    _viewGroup() {
        this.viewGroups = true;
        this.groupView = '';
        if (!this.groups) {
            this._loadGroups();
        }
    }
    /**
     * @private
     * @return {?}
     */
    _loadGroups() {
        this.noGroups = false;
        this.socialService.listGroups((/**
         * @param {?} groups
         * @return {?}
         */
        groups => {
            this.groups = groups;
            this.noGroups = groups.length < 1;
        }));
    }
    /**
     * @private
     * @param {?} group
     * @return {?}
     */
    _viewGroupMember(group) {
        this.viewMembers = true;
        this.members = [];
        this.groupView = group;
        this._loadMembers();
    }
    /**
     * @private
     * @return {?}
     */
    _loadMembers() {
        this._loadFriends(this.groupView, (/**
         * @param {?} users
         * @return {?}
         */
        users => this.members = users));
    }
    /**
     * @private
     * @return {?}
     */
    _loadAllFriends() {
        this.friends = [];
        this.noFriends = false;
        this._loadFriends('', (/**
         * @param {?} users
         * @return {?}
         */
        users => {
            /** @type {?} */
            const tmp = [];
            users.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                /** @type {?} */
                const index = tmp.indexOf(item.usercode);
                if (index < 0) {
                    tmp.push(item.usercode);
                    item.categories = [item.category];
                    this.friends.push(item);
                }
                else {
                    this.friends[index].categories.push(item.category);
                }
            }));
            this.noFriends = this.friends.length < 1;
        }));
    }
    /**
     * @private
     * @param {?} category
     * @param {?} callback
     * @return {?}
     */
    _loadFriends(category, callback) {
        this.socialService.listFriends({
            pageSize: -1,
            category: category
        }, (/**
         * @param {?} users
         * @return {?}
         */
        (users) => {
            users.forEach((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                item._index = this._indexSearch(item.fullname) + ' ' + item.usercode;
            }));
            callback(users);
        }));
    }
    /**
     * @private
     * @param {?} members
     * @return {?}
     */
    _getMemberStr(members) {
        return members.map((/**
         * @param {?} item
         * @return {?}
         */
        item => item.usercode || item.username)).join(',');
    }
    /**
     * @private
     * @return {?}
     */
    _actionView() {
        this.viewGroups = this.viewMembers = false;
        /** @type {?} */
        const group = this.router.url.split('/')[3];
        this.clearSearchValue();
        if (!group) {
            this._viewGroup();
        }
        else {
            this._viewGroupMember(decodeURIComponent(group));
        }
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    _indexSearch(value) {
        return this.accentService.viToEn(value.toLowerCase());
    }
}
SocialGroupComponent.decorators = [
    { type: Component, args: [{
                template: "<div *ngIf=\"viewGroups; else memberTemplate\">\n  <div class=\"group-toolbar\">\n    <button *ngIf=\"viewGroups\" class=\"btn btn-primary btn-sm\" (click)=\"createGroup()\">\n      <i class=\"fa fa-plus\"></i>\n      <span>T\u1EA1o nh\u00F3m</span>\n    </button>\n  </div>\n  <div *ngIf=\"!searching\">\n    <div class=\"grid-title\">Nh\u00F3m</div>\n    <div class=\"grid-container\">\n      <div class=\"grid-item\" *ngFor=\"let group of groups\" (click)=\"viewGroup(group)\">\n        <i class=\"grid-item__icon fa fa-users\"></i>\n        <div class=\"grid-item__content\" [title]=\"group\">\n          <div class=\"grid-content__ellipsis\">{{group}}</div>\n        </div>\n        <button (click)=\"$event.stopPropagation(); removeGroup(group, templateConfirmGroup)\"\n                class=\"btn btn-sm btn-danger grid-item__btn\"\n                title=\"X\u00F3a nh\u00F3m\">\n          <i class=\"fa fa-trash\"></i>\n        </button>\n      </div>\n      <div class=\"grid-item\" *ngIf=\"noGroups\">\n          <i class=\"grid-item__icon fa fa-bell\"></i>\n          <div class=\"grid-item__content\">Ch\u01B0a c\u00F3 nh\u00F3m \u0111\u01B0\u1EE3c t\u1EA1o</div>\n      </div>\n    </div>\n  </div>\n  <div class=\"grid-title\">B\u1EA1n b\u00E8</div>\n  <div class=\"grid-container\">\n    <div class=\"grid-item\" *ngFor=\"let friend of friends\" [hidden]=\"friend.hidden\">\n      <img userAvatar [usercode]=\"friend.usercode\" class=\"grid-item__icon\">\n      <div class=\"grid-item__content\">\n        <div class=\"grid-content__ellipsis\">{{friend.fullname}}</div>\n        <div class=\"grid-content__brief\">{{friend.usercode}}</div>\n        <!--<div>-->\n          <!--<div *ngFor=\"let group of friend.categories\" class=\"group-item\">{{group}}</div>-->\n        <!--</div>-->\n      </div>\n      <button (click)=\"removeFriend(friend, templateConfirmMember)\"\n              class=\"btn btn-sm btn-danger grid-item__btn\"\n              title=\"X\u00F3a b\u1EA1n\">\n        <i class=\"fa fa-trash\"></i>\n      </button>\n    </div>\n    <div class=\"grid-item grid-item__sm\" *ngIf=\"noFriends\">\n      <i class=\"grid-item__icon fa fa-bell\"></i>\n      <div class=\"grid-item__content\">Ch\u01B0a c\u00F3 b\u1EA1n b\u00E8</div>\n    </div>\n  </div>\n</div>\n\n<ng-template #memberTemplate>\n  <div class=\"group-toolbar\">\n    <button class=\"btn btn-primary btn-sm\" (click)=\"toGroups()\">\n      <i class=\"fa fa-arrow-left\"></i>\n      <span>Tr\u1EDF v\u1EC1</span>\n    </button>\n    <button class=\"btn btn-primary btn-sm\" (click)=\"addMemberGroup()\">\n      <i class=\"fa fa-plus\"></i>\n      <span>Th\u00EAm th\u00E0nh vi\u00EAn</span>\n    </button>\n    <button class=\"btn btn-danger btn-sm\" (click)=\"removeGroup(groupView, templateConfirmGroup)\">\n      <i class=\"fa fa-trash\"></i>\n      <span>X\u00F3a nh\u00F3m</span>\n    </button>\n  </div>\n\n  <div class=\"grid-title\">{{groupView}}</div>\n\n  <div class=\"grid-container\">\n    <div class=\"grid-item\" *ngFor=\"let friend of members\" [hidden]=\"friend.hidden\">\n      <img userAvatar [usercode]=\"friend.usercode\" class=\"grid-item__icon\">\n      <div class=\"grid-item__content\">\n        <div class=\"grid-content__ellipsis\">{{friend.fullname}}</div>\n        <div class=\"grid-content__brief\">{{friend.usercode}}</div>\n      </div>\n      <button (click)=\"removeFriend(friend, templateConfirmMember)\"\n              class=\"btn btn-sm btn-danger grid-item__btn\"\n              title=\"X\u00F3a b\u1EA1n\">\n        <i class=\"fa fa-trash\"></i>\n      </button>\n    </div>\n    <div class=\"grid-item grid-item__sm\" *ngIf=\"members && !members.length\">\n        <i class=\"grid-item__icon fa fa-bell\"></i>\n        <div class=\"grid-item__content\">Ch\u01B0a c\u00F3 th\u00E0nh vi\u00EAn trong nh\u00F3m</div>\n    </div>\n  </div>\n\n</ng-template>\n\n<div socialGroupCreate #modalCreateMember (onCreate)=\"groupCreated()\"></div>\n<div socialGroupMember #modalAddMember (onSelect)=\"saveMemberGroup($event)\"></div>\n\n<ng-template #templateConfirmGroup>\n  <div class=\"modal-header\">\n    <h4 class=\"modal-title pull-left\">X\u00F3a nh\u00F3m</h4>\n    <button type=\"button\" class=\"close pull-right\" (click)=\"modalGroupDeleteRef.hide()\">\n      <span aria-hidden=\"true\">&times;</span>\n    </button>\n  </div>\n  <div class=\"modal-body\">\n    X\u00F3a <b>{{groupActive}}</b> kh\u1ECFi danh s\u00E1ch nh\u00F3m?\n  </div>\n  <div class=\"modal-footer\">\n    <button type=\"button\" class=\"btn btn-danger\" (click)=\"onDeleteGroup()\">X\u00F3a nh\u00F3m</button>\n    <button type=\"button\" class=\"btn btn-secondary\" (click)=\"modalGroupDeleteRef.hide()\">H\u1EE7y</button>\n  </div>\n</ng-template>\n<ng-template #templateConfirmMember>\n  <div class=\"modal-header\">\n    <h4 class=\"modal-title pull-left\">X\u00F3a b\u1EA1n b\u00E8</h4>\n    <button type=\"button\" class=\"close pull-right\" (click)=\"modalMemberDeleteRef.hide()\">\n      <span aria-hidden=\"true\">&times;</span>\n    </button>\n  </div>\n  <div class=\"modal-body\">\n    X\u00F3a <b>{{memberActive.fullname}}</b> kh\u1ECFi nh\u00F3m\n    <b>{{memberActive.categories ? memberActive.categories.join(', ') : memberActive.category}}</b>?\n  </div>\n  <div class=\"modal-footer\">\n    <button type=\"button\" class=\"btn btn-danger\" (click)=\"onDeleteMember()\">X\u00F3a b\u1EA1n</button>\n    <button type=\"button\" class=\"btn btn-secondary\" (click)=\"modalMemberDeleteRef.hide()\">H\u1EE7y</button>\n  </div>\n</ng-template>",
                styles: [".grid-title{color:rgba(0,0,0,.54);font-size:14px;margin:8px;font-weight:700}.grid-container{padding-bottom:20px;overflow:hidden}.grid-item{float:left;width:50%;padding:0 15px;margin-right:-1px;margin-bottom:-1px;border:1px solid rgba(0,0,0,.05);cursor:pointer;display:flex;height:90px;align-items:center}.grid-item__sm{height:80px}.grid-item__icon{width:60px;height:60px;line-height:60px;text-align:center;color:rgba(0,0,0,.54);border-radius:50%;font-size:26px;background:rgba(0,0,0,.05)}.grid-item__btn{line-height:20px;margin-left:5px}.grid-item__content{flex-grow:1;padding:0 10px 0 15px;line-height:20px;overflow:hidden;font-size:16px}.grid-content__brief,.grid-content__ellipsis{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.grid-content__brief{font-size:14px;color:rgba(0,0,0,.4)}@media (max-width:767px){.grid-item{width:100%}.grid-item__icon{width:40px;height:40px;line-height:40px}}", ".group-toolbar{margin:-20px -15px 20px;padding:7px 15px;background:rgba(0,0,0,.03)}.group-item{display:inline-block;margin-top:5px;font-size:12px;line-height:20px;padding:0 5px;overflow:hidden;border-radius:2px;color:#fff;background:#007bff;margin-right:5px}"]
            }] }
];
/** @nocollapse */
SocialGroupComponent.ctorParameters = () => [
    { type: SocialService },
    { type: AccentService },
    { type: Router },
    { type: BsModalService }
];
SocialGroupComponent.propDecorators = {
    modalCreateMember: [{ type: ViewChild, args: ['modalCreateMember',] }],
    modalAddMember: [{ type: ViewChild, args: ['modalAddMember',] }]
};
if (false) {
    /** @type {?} */
    SocialGroupComponent.prototype.modalCreateMember;
    /** @type {?} */
    SocialGroupComponent.prototype.modalAddMember;
    /** @type {?} */
    SocialGroupComponent.prototype.viewGroups;
    /** @type {?} */
    SocialGroupComponent.prototype.groups;
    /** @type {?} */
    SocialGroupComponent.prototype.friends;
    /** @type {?} */
    SocialGroupComponent.prototype.viewMembers;
    /** @type {?} */
    SocialGroupComponent.prototype.members;
    /** @type {?} */
    SocialGroupComponent.prototype.groupView;
    /** @type {?} */
    SocialGroupComponent.prototype.groupActive;
    /** @type {?} */
    SocialGroupComponent.prototype.memberActive;
    /** @type {?} */
    SocialGroupComponent.prototype.modalGroupDeleteRef;
    /** @type {?} */
    SocialGroupComponent.prototype.modalMemberDeleteRef;
    /** @type {?} */
    SocialGroupComponent.prototype.noGroups;
    /** @type {?} */
    SocialGroupComponent.prototype.noFriends;
    /** @type {?} */
    SocialGroupComponent.prototype.searching;
    /**
     * @type {?}
     * @private
     */
    SocialGroupComponent.prototype._prefix;
    /**
     * @type {?}
     * @private
     */
    SocialGroupComponent.prototype._router;
    /** @type {?} */
    SocialGroupComponent.prototype.placeholder;
    /**
     * @type {?}
     * @private
     */
    SocialGroupComponent.prototype.socialService;
    /**
     * @type {?}
     * @private
     */
    SocialGroupComponent.prototype.accentService;
    /**
     * @type {?}
     * @private
     */
    SocialGroupComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    SocialGroupComponent.prototype.modalService;
}
