// inet-ui/esm2015/src/layout/bootstrap/abstract-side-nav.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostListener, Input, Output } from "@angular/core";
export class AbstractSideNavComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.opened = false;
        this.onClose = new EventEmitter();
        this.onClear = new EventEmitter();
        this.onLoad = new EventEmitter();
        this.overflowCls = 'overflow-hidden';
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const currentValue = changes['opened']["currentValue"];
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (currentValue) {
                this.open();
            }
            else if (changes['opened']['previousValue'] != currentValue) {
                this.close();
            }
        }), 1);
    }
    /**
     * @return {?}
     */
    open() {
        if (!document.body.classList.contains(this.overflowCls)) {
            document.body.classList.add(this.overflowCls);
        }
    }
    /**
     * @return {?}
     */
    close() {
        if (document.body.classList.contains(this.overflowCls)) {
            document.body.classList.remove(this.overflowCls);
        }
        this.opened = false;
        this.onClose.emit(this.opened);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clickOut(event) {
        if (!this.elementRef.nativeElement.contains(event.target)
            && this.opened && !event.target.closest('.nav-item') &&
            !event.target.closest('.search-input__icon')) {
            this.close();
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clickInside(event) {
        if (this.elementRef.nativeElement.contains(event.target)) {
        }
    }
}
AbstractSideNavComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-admin-side-nav',
                template: ``
            }] }
];
/** @nocollapse */
AbstractSideNavComponent.ctorParameters = () => [
    { type: ElementRef }
];
AbstractSideNavComponent.propDecorators = {
    opened: [{ type: Input }],
    onClose: [{ type: Output }],
    onClear: [{ type: Output }],
    onLoad: [{ type: Output }],
    clickOut: [{ type: HostListener, args: ['document:click', ['$event'],] }],
    clickInside: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    AbstractSideNavComponent.prototype.opened;
    /** @type {?} */
    AbstractSideNavComponent.prototype.onClose;
    /** @type {?} */
    AbstractSideNavComponent.prototype.onClear;
    /** @type {?} */
    AbstractSideNavComponent.prototype.onLoad;
    /**
     * @type {?}
     * @private
     */
    AbstractSideNavComponent.prototype.overflowCls;
    /** @type {?} */
    AbstractSideNavComponent.prototype.elementRef;
}
