// inet-ui/esm2015/src/layout/bootstrap/bootstrap-layout.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { LayoutComponent } from './layout.component';
import { AdminNavbarComponent } from './navbar/navbar.component';
import { AuthenticateGuard } from "../../common/authenticate.guard";
import { NotifyModule } from "../../notify/notify.module";
import { BootstrapLayoutRoutingModule } from "./bootstrap-layout-routing.module";
import { CommonModule } from "@angular/common";
import { HttpClientModule } from "@angular/common/http";
import { CloudTranslateModule } from "../../translate/cloud-translate.module";
import { ModalModule } from "ngx-bootstrap";
import { CoreModule, HttpClientService, LoadingIndicatorService } from "inet-core";
import { UserProfileModule } from "../../user-profile/user-profile.module";
import { BootstrapUserProfileComponent } from "./user-profile/bootstrap-user-profile.component";
import { InterceptorModule } from "../../interceptor/interceptor.module";
import { OrganizationService } from "../../organization/organization.service";
import { BootstrapCompanyProfileComponent } from "./company-profile/bootstrap-company-profile.component";
import { OrganizationModule } from "../../organization/organization.module";
import { NavigationCancel, NavigationEnd, NavigationError, NavigationStart, Router } from "@angular/router";
import { NotifyMessageService } from "../../common/notify-message.service";
import { MessageSideNavComponent } from "./message-side-nav/message-side-nav.component";
import { AppSideNavComponent } from "./app-side-nav/app-side-nav.component";
import { SocialSideNavComponent } from "./social-side-nav/social-side-nav.component";
import { AbstractSideNavComponent } from "./abstract-side-nav.component";
import { CommonToolbarComponent } from "./common-toolbar/common-toolbar.component";
import { SafePipeModule } from "../../pipes/safe-pipe.module";
import { SharingInformationModule } from "../../sharing-information/sharing-information.module";
export class BootstrapLayoutModule {
    /**
     * @param {?} router
     * @param {?} loadingService
     */
    constructor(router, loadingService) {
        /*
        * True to enable layout for controlling the menu,header,etc....
        * This option should generally be enabled when layout are being used inside the framework.
         */
        iNet.enableLayout = true;
        router.events.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        event => {
            if (event instanceof NavigationStart) {
                loadingService.showLoading();
            }
            if ((event instanceof NavigationError || event instanceof NavigationEnd || event instanceof NavigationCancel)) {
                loadingService.hideLoading();
            }
        }));
    }
}
BootstrapLayoutModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    CoreModule,
                    BootstrapLayoutRoutingModule,
                    HttpClientModule,
                    InterceptorModule,
                    ModalModule.forRoot(),
                    CloudTranslateModule,
                    NotifyModule,
                    UserProfileModule,
                    OrganizationModule,
                    SafePipeModule,
                    SharingInformationModule,
                ],
                declarations: [
                    AdminNavbarComponent,
                    LayoutComponent,
                    BootstrapUserProfileComponent,
                    BootstrapCompanyProfileComponent,
                    MessageSideNavComponent,
                    AppSideNavComponent,
                    SocialSideNavComponent,
                    AbstractSideNavComponent,
                    CommonToolbarComponent
                ],
                exports: [InterceptorModule, CloudTranslateModule, LayoutComponent, AbstractSideNavComponent],
                schemas: [CUSTOM_ELEMENTS_SCHEMA],
                providers: [AuthenticateGuard, OrganizationService, HttpClientService, NotifyMessageService]
            },] }
];
/** @nocollapse */
BootstrapLayoutModule.ctorParameters = () => [
    { type: Router },
    { type: LoadingIndicatorService }
];
