// inet-ui/esm2015/src/layout/bootstrap/layout.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef, ViewChild } from '@angular/core';
import { NavigationEnd, Router } from "@angular/router";
import { AdminNavbarComponent } from "./navbar/navbar.component";
export class LayoutComponent {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
        this.themeClass = 'theme-silver';
        this.hideToggler = false;
        this.hideTogglerButton = true;
        this.visibleMenu = true;
        this.expandedMenu = false;
        this.expandedSlider = false;
        this.expandedMessageSlider = false;
        this.expandedAppSlider = false;
        this.expandedSocialSlider = false;
        this.expandedChatSlider = false;
        this.hideSearch = false;
        this.hideCompanyMenu = false;
        this.hideProfileMenu = false;
        this.hideBrandName = true;
        this.hideLogo = false;
        this.visibleOverlay = false;
        this.fullScreenLayout = false;
        this.fullLayoutWithUrls = [];
        this.routeChange = new EventEmitter();
        this.onToggle = new EventEmitter();
        this.onClickLogo = new EventEmitter();
        this.visibleToolbar = false;
        this.visibleChatIcon = true;
        this.visibleMessageIcon = true;
        this.visibleAppIcon = true;
        this.visibleNoteIcon = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (window.localStorage) {
            this.expandedMenu = (window.localStorage.getItem('expandedMenu') === '1');
        }
        // this.fullLayoutWithUrls.push('/user-profile');
        this._router = this.router.events.filter((/**
         * @param {?} event
         * @return {?}
         */
        event => event instanceof NavigationEnd))
            .subscribe((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            /** @type {?} */
            const url = event.url;
            this.onChangeUrl(url);
        }));
    }
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    onChangeUrl(url) {
        if (!this.fullScreenLayout) {
            /** @type {?} */
            let __visibleMenu = !!this.menu;
            if (__visibleMenu) {
                for (let i = 0; i < this.fullLayoutWithUrls.length; i++) {
                    if ((url !== '/' && this.fullLayoutWithUrls[i] !== '/' && url.indexOf(this.fullLayoutWithUrls[i]) === 0)
                        || (url === this.fullLayoutWithUrls[i])) {
                        __visibleMenu = false;
                        break;
                    }
                }
            }
            this.visibleMenu = __visibleMenu;
        }
        else {
            this.visibleMenu = false;
        }
        this.hideToggler = !this.visibleMenu;
        this.routeChange.emit({
            url: url,
            visibleMenu: this.visibleMenu,
            hideToggler: this.hideToggler,
            expandedMenu: this.expandedMenu,
            layout: this
        });
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        //Using jQuery to handle menu
        $(document).ready((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            const $menu = $('#mainnav-menu');
            /** @type {?} */
            const activeCls = 'active';
            /** @type {?} */
            const $elements = $menu.find('li');
            $elements.on('hide.bs.collapse', (/**
             * @return {?}
             */
            function () {
                $(this).removeClass(activeCls);
            }));
            $elements.on('show.bs.collapse', (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                const $currentActive = $menu.find('li.active');
                $currentActive.find('a[data-toggle]').trigger("click");
                $(this).addClass(activeCls);
            }));
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._router.unsubscribe();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onToggleMenu($event) {
        this.expandedMenu = !this.expandedMenu;
        window.localStorage.setItem('expandedMenu', this.expandedMenu ? '1' : '0');
        window.dispatchEvent(new Event('resize'));
        this.onToggle.emit(this.expandedMenu);
    }
    /**
     * @private
     * @param {?} v
     * @return {?}
     */
    updateSlider(v) {
        this.expandedSlider = v;
        this.visibleOverlay = v;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onToggleMessageSlider($event) {
        this.expandedAppSlider = false;
        this.expandedSocialSlider = false;
        this.expandedChatSlider = false;
        this.expandedMessageSlider = !this.expandedMessageSlider;
        this.updateSlider(this.expandedMessageSlider);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onToggleAppSlider($event) {
        this.expandedChatSlider = false;
        this.expandedSocialSlider = false;
        this.expandedMessageSlider = false;
        this.expandedAppSlider = !this.expandedAppSlider;
        this.updateSlider(this.expandedAppSlider);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onToggleSocialSlider($event) {
        this.expandedAppSlider = false;
        this.expandedChatSlider = false;
        this.expandedMessageSlider = false;
        this.expandedSocialSlider = !this.expandedSocialSlider;
        this.updateSlider(this.expandedSocialSlider);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onToggleNoteSlider($event) {
        this.expandedAppSlider = false;
        this.expandedSocialSlider = false;
        this.expandedChatSlider = false;
        this.expandedMessageSlider = false;
        this.updateSlider(this.expandedMessageSlider);
        this.router.navigate(['note', 'list']);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onCloseSlider($event) {
        this.expandedMessageSlider = false;
        this.expandedAppSlider = false;
        this.expandedChatSlider = false;
        this.expandedSocialSlider = false;
        this.expandedSlider = false;
        this.visibleOverlay = false;
    }
    //====================Message count=================
    /**
     * @param {?} $event
     * @return {?}
     */
    onClearMessage($event) {
        if (this.navbarComponent) {
            this.navbarComponent.setMessageCount('');
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onLoadMessage($event) {
        if (this.navbarComponent) {
            this.navbarComponent.countMessage();
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    clickLogo($event) {
        this.onClickLogo.emit(this.homeRouterLink);
    }
}
LayoutComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-layout',
                template: "<div id=\"container\" [class]=\"themeClass\" [ngClass]=\"{'mainnav-lg': expandedMenu, 'mainnav-sm': !expandedMenu}\">\n    <div toastContainer></div>\n    <app-admin-navbar (onToggleMenu)=\"onToggleMenu($event)\"\n                      (onToggleMessageSideNav)=\"onToggleMessageSlider($event)\"\n                      (onToggleAppSideNav)=\"onToggleAppSlider($event)\"\n                      (onToggleSocialSideNav)=\"onToggleSocialSlider($event)\"\n                      (onToggleNoteSideNav)=\"onToggleNoteSlider($event)\"\n                      (onClickLogo)=\"clickLogo($event)\"\n                      [hideSearch]=\"hideSearch\" [applicationName]=\"appName\"\n                      [homeRouterLink]=\"homeRouterLink\" [navbarItemTpl]=\"navbarItemTpl\" [profileItemTpl]=\"profileItemTpl\"\n                      [navbarMenuTpl]=\"navbarMenu\" [visibleAppIcon]=\"visibleAppIcon\" [visibleChatIcon]=\"visibleChatIcon\"\n                      [visibleMessageIcon]=\"visibleMessageIcon\" [visibleNoteIcon]=\"visibleNoteIcon\"\n                      [hideToggler]=\"hideToggler\" [hideTogglerButton]=\"hideTogglerButton\"\n                      [hideBrandName]=\"hideBrandName\" [hideLogo]=\"hideLogo\"\n                      [hideCompanyMenu]=\"hideCompanyMenu\" [hideProfileMenu]=\"hideProfileMenu\">\n    </app-admin-navbar>\n    <app-admin-message-side-nav [opened]=\"expandedMessageSlider\" (onClear)=\"onClearMessage($event)\"\n                          (onClose)=\"onCloseSlider($event)\"\n                          (onLoad)=\"onLoadMessage($event)\"></app-admin-message-side-nav>\n    <app-admin-app-side-nav [opened]=\"expandedAppSlider\" (onClose)=\"onCloseSlider($event)\"></app-admin-app-side-nav>\n    <app-admin-social-side-nav [opened]=\"expandedSocialSlider\" (onClose)=\"onCloseSlider($event)\"></app-admin-social-side-nav>\n    <div class=\"container-fluid\">\n        <div class=\"row\">\n            <nav id=\"mainnav-container\" *ngIf=\"visibleMenu\">\n                <div id=\"mainnav\">\n                    <ng-template [ngTemplateOutlet]=\"menu\"></ng-template>\n                </div>\n            </nav>\n            <div id=\"content-container\" [ngClass]=\"{'pl-0': !visibleMenu , 'tb-visible': !!toolbar && visibleToolbar}\">\n                <ng-template [ngTemplateOutlet]=\"toolbar\"></ng-template>\n                <router-outlet></router-outlet>\n                <div id=\"loading-indicator\" title=\"{{'INDICATOR.PROCESSING' | translate}}\">\n                    <span>{{'INDICATOR.PROCESSING' | translate}}</span>\n                </div>\n            </div>\n        </div>\n    </div>\n    <div class=\"overlay\" [ngClass]=\"{'d-block': expandedSlider || visibleOverlay}\"></div>\n</div>\n"
            }] }
];
/** @nocollapse */
LayoutComponent.ctorParameters = () => [
    { type: Router }
];
LayoutComponent.propDecorators = {
    themeClass: [{ type: Input, args: ['theme',] }],
    menu: [{ type: ContentChild, args: ['menu',] }],
    navbarItemTpl: [{ type: ContentChild, args: ['navbarItem',] }],
    profileItemTpl: [{ type: ContentChild, args: ['profileItem',] }],
    navbarMenu: [{ type: ContentChild, args: ['navbarMenu',] }],
    toolbar: [{ type: ContentChild, args: ['toolbar',] }],
    homeRouterLink: [{ type: Input, args: ['homeRouterLink',] }],
    appName: [{ type: Input }],
    hideToggler: [{ type: Input }],
    hideTogglerButton: [{ type: Input }],
    visibleMenu: [{ type: Input }],
    expandedMenu: [{ type: Input }],
    expandedSlider: [{ type: Input }],
    expandedMessageSlider: [{ type: Input }],
    expandedAppSlider: [{ type: Input }],
    expandedSocialSlider: [{ type: Input }],
    expandedChatSlider: [{ type: Input }],
    hideSearch: [{ type: Input }],
    hideCompanyMenu: [{ type: Input }],
    hideProfileMenu: [{ type: Input }],
    hideBrandName: [{ type: Input }],
    hideLogo: [{ type: Input }],
    visibleOverlay: [{ type: Input }],
    fullScreenLayout: [{ type: Input }],
    fullLayoutWithUrls: [{ type: Input }],
    routeChange: [{ type: Output }],
    onToggle: [{ type: Output }],
    onClickLogo: [{ type: Output }],
    visibleToolbar: [{ type: Input }],
    visibleChatIcon: [{ type: Input }],
    visibleMessageIcon: [{ type: Input }],
    visibleAppIcon: [{ type: Input }],
    visibleNoteIcon: [{ type: Input }],
    navbarComponent: [{ type: ViewChild, args: [AdminNavbarComponent,] }]
};
if (false) {
    /** @type {?} */
    LayoutComponent.prototype.themeClass;
    /** @type {?} */
    LayoutComponent.prototype.menu;
    /** @type {?} */
    LayoutComponent.prototype.navbarItemTpl;
    /** @type {?} */
    LayoutComponent.prototype.profileItemTpl;
    /** @type {?} */
    LayoutComponent.prototype.navbarMenu;
    /** @type {?} */
    LayoutComponent.prototype.toolbar;
    /** @type {?} */
    LayoutComponent.prototype.homeRouterLink;
    /** @type {?} */
    LayoutComponent.prototype.appName;
    /** @type {?} */
    LayoutComponent.prototype.hideToggler;
    /** @type {?} */
    LayoutComponent.prototype.hideTogglerButton;
    /** @type {?} */
    LayoutComponent.prototype.visibleMenu;
    /** @type {?} */
    LayoutComponent.prototype.expandedMenu;
    /** @type {?} */
    LayoutComponent.prototype.expandedSlider;
    /** @type {?} */
    LayoutComponent.prototype.expandedMessageSlider;
    /** @type {?} */
    LayoutComponent.prototype.expandedAppSlider;
    /** @type {?} */
    LayoutComponent.prototype.expandedSocialSlider;
    /** @type {?} */
    LayoutComponent.prototype.expandedChatSlider;
    /** @type {?} */
    LayoutComponent.prototype.hideSearch;
    /** @type {?} */
    LayoutComponent.prototype.hideCompanyMenu;
    /** @type {?} */
    LayoutComponent.prototype.hideProfileMenu;
    /** @type {?} */
    LayoutComponent.prototype.hideBrandName;
    /** @type {?} */
    LayoutComponent.prototype.hideLogo;
    /** @type {?} */
    LayoutComponent.prototype.visibleOverlay;
    /** @type {?} */
    LayoutComponent.prototype.fullScreenLayout;
    /** @type {?} */
    LayoutComponent.prototype.fullLayoutWithUrls;
    /** @type {?} */
    LayoutComponent.prototype.routeChange;
    /** @type {?} */
    LayoutComponent.prototype.onToggle;
    /** @type {?} */
    LayoutComponent.prototype.onClickLogo;
    /** @type {?} */
    LayoutComponent.prototype.visibleToolbar;
    /** @type {?} */
    LayoutComponent.prototype.visibleChatIcon;
    /** @type {?} */
    LayoutComponent.prototype.visibleMessageIcon;
    /** @type {?} */
    LayoutComponent.prototype.visibleAppIcon;
    /** @type {?} */
    LayoutComponent.prototype.visibleNoteIcon;
    /**
     * @type {?}
     * @private
     */
    LayoutComponent.prototype._router;
    /** @type {?} */
    LayoutComponent.prototype.navbarComponent;
    /**
     * @type {?}
     * @private
     */
    LayoutComponent.prototype.router;
}
