// inet-ui/esm5/src/user-profile/user-profile-info/user-profile-info.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { ErrorMessage } from "inet-core";
import { Contact } from "../../model/contact";
import { LocationService } from "../../common/location.service";
import { FormBuilder, Validators } from "@angular/forms";
import { Subject } from "rxjs/Subject";
import { CustomValidators } from "ng2-validation";
import { DictionaryService } from "../../common/dictionary.service";
import * as moment from 'moment';
window['moment'] = window['moment'] || moment;
var UserProfileInfoComponent = /** @class */ (function () {
    function UserProfileInfoComponent(dictionaryService, locationService, fb) {
        this.dictionaryService = dictionaryService;
        this.locationService = locationService;
        this.fb = fb;
        this.profile = new Contact();
        this.countries = [];
        this.provinces = [];
        this.districts = [];
        this.genders = [
            { code: "M", key: "COMMON.MODULE.USER_PROFILE.MALE" },
            { code: "FM", key: "COMMON.MODULE.USER_PROFILE.FEMALE" },
            { code: "OT", key: "COMMON.MODULE.USER_PROFILE.OTHER" }
        ];
        this.countryCode = 'VN';
        this.onValidate = new EventEmitter();
        this.editable = false;
        this.formChanged = new Subject();
    }
    /**
     * @return {?}
     */
    UserProfileInfoComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.formChanged.debounceTime(250).distinctUntilChanged().subscribe((/**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            _this.onValidate.emit(v);
        }));
        this.profileForm = this.fb.group({
            lastName: [null, Validators.required],
            middleName: [null, null],
            firstName: [null, Validators.required],
            email: [null, Validators.compose([CustomValidators.email])],
            fullDateOfBirth: [null, null],
            sex: [null, null],
            phone: [null, null],
            country: [null, null],
            street: [null, null],
            district: [null, null],
            province: [null, null]
        });
        this.onChanges();
        this.loadAddress();
    };
    /*
    private loadDictionary() {
        this.dictionaryService.findByKeys('SEX').subscribe((response: ResponseData) => {
            if (ErrorMessage.TYPE !== response.type) {
                // const {data} = response.data.items||[];
                // const {SEX} = data;
                // this.genders = SEX;
                this.genders = response.items || [];
            }
        });
    }
    */
    //
    // private loadGender(){
    //     this.dictionaryService.getSex({}).subscribe((response: ResponseData) => {
    //         if (ErrorMessage.TYPE !== response.type) {
    //             this.genders = response.items || [];
    //         }
    //     });
    // }
    /*
        private loadDictionary() {
            this.dictionaryService.findByKeys('SEX').subscribe((response: ResponseData) => {
                if (ErrorMessage.TYPE !== response.type) {
                    // const {data} = response.data.items||[];
                    // const {SEX} = data;
                    // this.genders = SEX;
                    this.genders = response.items || [];
                }
            });
        }
        */
    //
    // private loadGender(){
    //     this.dictionaryService.getSex({}).subscribe((response: ResponseData) => {
    //         if (ErrorMessage.TYPE !== response.type) {
    //             this.genders = response.items || [];
    //         }
    //     });
    // }
    /**
     * @private
     * @return {?}
     */
    UserProfileInfoComponent.prototype.loadAddress = /*
        private loadDictionary() {
            this.dictionaryService.findByKeys('SEX').subscribe((response: ResponseData) => {
                if (ErrorMessage.TYPE !== response.type) {
                    // const {data} = response.data.items||[];
                    // const {SEX} = data;
                    // this.genders = SEX;
                    this.genders = response.items || [];
                }
            });
        }
        */
    //
    // private loadGender(){
    //     this.dictionaryService.getSex({}).subscribe((response: ResponseData) => {
    //         if (ErrorMessage.TYPE !== response.type) {
    //             this.genders = response.items || [];
    //         }
    //     });
    // }
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.locationService.listCountry().subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (ErrorMessage.TYPE !== data.type) {
                _this.countries = data.items || [];
                if (_this.profile.address.countryCode) {
                    _this.loadProvinceByCountryCode(_this.profile.address.countryCode);
                    if (_this.profile.address.provinceCode) {
                        _this.loadDistrictByCode(_this.profile.address.countryCode, _this.profile.address.provinceCode);
                    }
                }
            }
        }));
    };
    /**
     * @private
     * @param {?} countryCode
     * @return {?}
     */
    UserProfileInfoComponent.prototype.loadProvinceByCountryCode = /**
     * @private
     * @param {?} countryCode
     * @return {?}
     */
    function (countryCode) {
        var _this = this;
        this.countryCode = countryCode;
        this.provinceCode = null;
        this.districts = [];
        this.locationService.getProvinceByCountryCode(countryCode).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (ErrorMessage.TYPE !== data.type) {
                _this.provinces = data.items || [];
            }
        }));
    };
    /**
     * @private
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    UserProfileInfoComponent.prototype.loadDistrictByCode = /**
     * @private
     * @param {?} countryCode
     * @param {?} provinceCode
     * @return {?}
     */
    function (countryCode, provinceCode) {
        var _this = this;
        this.provinceCode = provinceCode;
        this.locationService.getDistrictByCode(countryCode, provinceCode).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            if (ErrorMessage.TYPE !== data.type) {
                _this.districts = data.items || [];
            }
        }));
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    UserProfileInfoComponent.prototype.onChangeCountry = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var target$ = $event.target;
        //const countryId = target$.options[target$.selectedIndex].getAttribute('data-uuid');
        this.loadProvinceByCountryCode(target$.value);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    UserProfileInfoComponent.prototype.onChangeProvince = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        /** @type {?} */
        var target$ = $event.target;
        // const countryId = target$.options[target$.selectedIndex].getAttribute('data-uuid');
        this.loadDistrictByCode(this.countryCode, target$.value);
    };
    /**
     * @param {?=} event
     * @return {?}
     */
    UserProfileInfoComponent.prototype.onChangeDate = /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        if (window['moment']) {
            /** @type {?} */
            var date = window['moment'](new Date(event));
            if ((!date.isValid())) {
                setTimeout((/**
                 * @return {?}
                 */
                function () { return _this.profile.fullDateOfBirth = new Date(); }));
            }
        }
    };
    /**
     * @private
     * @return {?}
     */
    UserProfileInfoComponent.prototype.onChanges = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.profileForm.valueChanges.subscribe((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            _this.formChanged.next(_this.profileForm.valid);
        }));
        if (this.editable) {
            this.profileForm.enable();
        }
        else {
            this.profileForm.disable();
        }
    };
    UserProfileInfoComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-user-profile-info',
                    template: "<form [formGroup]=\"profileForm\">\n    <div class=\"row\">\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">\n                    {{'COMMON.MODULE.USER_PROFILE.LAST_NAME' | translate}} : <i class=\"required\"></i>\n                </label>\n                <div class=\"col-sm-9 pl-0\">\n                    <input [(ngModel)]=\"profile.lastName\" [formControl]=\"profileForm.controls['lastName']\" required name=\"lastName\"\n                           [ngClass]=\"{'is-invalid':profileForm.controls['lastName'].hasError('required') && profileForm.controls['lastName'].touched}\"\n                           class=\"form-control form-control-sm col-xs-12 col-sm-12\" maxlength=\"256\"/>\n                    <div *ngIf=\"profileForm.controls['lastName'].hasError('required') && profileForm.controls['lastName'].touched\" class=\"text-danger col-sm-12 pl-0\"><i\n                            class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.LAST_NAME_REQUIRED_MSG' | translate}}\n                    </div>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">{{'COMMON.MODULE.USER_PROFILE.MIDDLE_NAME' | translate}} : </label>\n                <div class=\"col-sm-9 pl-0\">\n                    <input [(ngModel)]=\"profile.middleName\" [formControl]=\"profileForm.controls['middleName']\"\n                           name=\"middleName\" class=\"form-control form-control-sm col-xs-12 col-sm-12\" maxlength=\"64\"/>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">\n                    {{'COMMON.MODULE.USER_PROFILE.FIRST_NAME' | translate}} : <i class=\"required\"></i></label>\n                <div class=\"col-sm-9 pl-0\">\n                    <input [(ngModel)]=\"profile.firstName\" [formControl]=\"profileForm.controls['firstName']\" required name=\"firstName\"\n                           [ngClass]=\"{'is-invalid':profileForm.controls['firstName'].hasError('required') && profileForm.controls['firstName'].touched}\"\n                           class=\"form-control form-control-sm col-xs-12 col-sm-12 \"/>\n                    <div *ngIf=\"profileForm.controls['firstName'].hasError('required') && profileForm.controls['firstName'].touched\"\n                         class=\"text-danger col-sm-12 pl-0\">\n                        <i class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.FIRST_NAME_REQUIRED_MSG' | translate}}\n                    </div>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">{{'COMMON.MODULE.USER_PROFILE.BIRTHDAY' | translate}} :</label>\n                <div class=\"input-group input-group-sm col-sm-3 pl-0\">\n                    <input [(ngModel)]=\"profile.fullDateOfBirth\" [formControl]=\"profileForm.controls['fullDateOfBirth']\"\n                           type=\"text\" name=\"fullDateOfBirth\" class=\"form-control form-control-sm col-xs-12 col-sm-12\"\n                           bsDatepicker #fullDateOfBirth=\"bsDatepicker\"\n                           (bsValueChange)=\"onChangeDate($event)\">\n                    <div (click)=\"fullDateOfBirth.toggle()\" class=\"input-group-append input-group-sm\">\n                        <div class=\"input-group-text\"><i class=\"fa fa-calendar\"></i></div>\n                    </div>\n                </div>\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">{{'COMMON.MODULE.USER_PROFILE.GENDER' | translate}} : </label>\n                <div class=\"col-sm-3 pl-0\">\n                    <select [(ngModel)]=\"profile.sex\" name=\"sex\" [formControl]=\"profileForm.controls['sex']\" class=\"form-control form-control-sm col-xs-12 col-sm-12 \">\n                        <option [ngValue]=\"!profile.sex ? profile.sex : null\">{{'COMMON.SELECTION.DEFAULT_VALUE' | translate}}</option>\n                        <option *ngFor=\"let gender of genders\" [value]=\"gender.code\">\n                            {{gender.key | translate}}\n                        </option>\n                    </select>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">{{'COMMON.MODULE.USER_PROFILE.EMAIL' | translate}} :</label>\n                <div class=\"col-sm-9 pl-0 input-group input-group-sm\">\n                    <div class=\"input-group-prepend input-group-sm\">\n                        <div class=\"input-group-text icon-div\"><i class=\"fa fa-envelope\"></i></div>\n                    </div>\n                    <input [(ngModel)]=\"profile.primaryEmail\" name=\"primaryEmail\" class=\"form-control form-control-sm col-xs-12 col-sm-12\"\n                           type=\"email\" [formControl]=\"profileForm.controls['email']\"\n                           [ngClass]=\"{'is-invalid': profileForm.controls['email'].hasError('pattern')\n                           && profileForm.controls['email'].touched}\"/>\n                    <!--div *ngIf=\"profileForm.controls['email'].hasError('required') && profileForm.controls['email'].touched\"\n                         class=\"pl-0 ml-0 col-xs-12 col-sm-12 text-danger\">\n                        <i class=\"fa fa-exclamation-triangle\"></i> {{'MODULE.USER_MANAGER.EMAIL_REQUIRED_MSG' | translate}}\n                    </div-->\n                    <div *ngIf=\"profileForm.controls['email'].hasError('email') && profileForm.controls['email'].touched\"\n                         class=\"pl-0 ml-0 col-xs-12 col-sm-12 text-danger\">\n                        <i class=\"fa fa-exclamation-triangle\"></i> {{'MODULE.USER_MANAGER.EMAIL_INVALID_MSG' | translate}}\n                    </div>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">{{'COMMON.MODULE.USER_PROFILE.MOBILE' | translate}}: </label>\n                <div class=\"col-sm-9 pl-0 input-group input-group-sm\">\n                    <div class=\"input-group-prepend input-group-sm\">\n                        <div class=\"input-group-text icon-div\"><i class=\"fa fa-phone icon-custom\"></i>\n                        </div>\n                    </div>\n                    <input [(ngModel)]=\"profile.mobilePhone\" [formControl]=\"profileForm.controls['phone']\" name=\"mobilePhone\" class=\"form-control form-control-sm col-xs-12 col-sm-12\" type=\"text\"/>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">\n                    {{'COMMON.MODULE.ADDRESS.STREET' | translate}} : </label>\n                <div class=\"col-sm-9 pl-0\">\n                    <input [(ngModel)]=\"profile.address.address\" name=\"street\" [formControl]=\"profileForm.controls['street']\" class=\"form-control form-control-sm col-xs-12 col-sm-12 \"/>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\">{{'COMMON.MODULE.ADDRESS.COUNTRY' | translate}}:</label>\n                <div class=\"col-sm-9 pl-0\">\n                    <select [(ngModel)]=\"profile.address.countryCode\" name=\"country\" [formControl]=\"profileForm.controls['country']\" (change)=\"onChangeCountry($event)\"\n                            class=\"form-control form-control-sm col-xs-12 col-sm-12 \">\n                        <option [ngValue]=\"!profile.address.countryCode ? profile.address.countryCode : null\">{{'COMMON.SELECTION.DEFAULT_VALUE' | translate}}</option>\n                        <option *ngFor=\"let country of countries\" [value]=\"country.code\">\n                            {{country.name}}\n                        </option>\n                    </select>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\"> {{'COMMON.MODULE.ADDRESS.CITY' | translate}} :</label>\n                <div class=\"col-sm-9 pl-0\">\n                    <select [(ngModel)]=\"profile.address.provinceCode\" name=\"province\" [formControl]=\"profileForm.controls['province']\" (change)=\"onChangeProvince($event)\"\n                            class=\"form-control form-control-sm col-xs-12 col-sm-12 \">\n                        <option [ngValue]=\"!profile.address.provinceCode ? profile.address.provinceCode : null\">{{'COMMON.SELECTION.DEFAULT_VALUE' | translate}}</option>\n                        <option *ngFor=\"let province of provinces\" [value]=\"province.code\">\n                            {{province.name}}\n                        </option>\n                    </select>\n                </div>\n            </div>\n        </div>\n        <div class=\"col-sm-12\">\n            <div class=\"form-group margin-b-form-group row\">\n                <label class=\"control-label col-form-label col-sm-3 lbl-bold\"> {{'COMMON.MODULE.ADDRESS.DISTRICT' | translate}} :</label>\n                <div class=\"col-sm-9 pl-0\">\n                    <select [(ngModel)]=\"profile.address.districtCode\" name=\"district\" [formControl]=\"profileForm.controls['district']\"\n                            class=\"form-control form-control-sm col-xs-12 col-sm-12 \">\n                        <option [ngValue]=\"!profile.address.districtCode ? profile.address.districtCode : null\">{{'COMMON.SELECTION.DEFAULT_VALUE' | translate}}</option>\n                        <option *ngFor=\"let district of districts\" [value]=\"district.code\">\n                            {{district.name}}\n                        </option>\n                    </select>\n                </div>\n            </div>\n        </div>\n    </div>\n</form>\n"
                }] }
    ];
    /** @nocollapse */
    UserProfileInfoComponent.ctorParameters = function () { return [
        { type: DictionaryService },
        { type: LocationService },
        { type: FormBuilder }
    ]; };
    UserProfileInfoComponent.propDecorators = {
        profile: [{ type: Input }],
        onValidate: [{ type: Output }],
        editable: [{ type: Input }]
    };
    return UserProfileInfoComponent;
}());
export { UserProfileInfoComponent };
if (false) {
    /** @type {?} */
    UserProfileInfoComponent.prototype.profile;
    /** @type {?} */
    UserProfileInfoComponent.prototype.profileForm;
    /** @type {?} */
    UserProfileInfoComponent.prototype.countries;
    /** @type {?} */
    UserProfileInfoComponent.prototype.provinces;
    /** @type {?} */
    UserProfileInfoComponent.prototype.districts;
    /** @type {?} */
    UserProfileInfoComponent.prototype.genders;
    /**
     * @type {?}
     * @private
     */
    UserProfileInfoComponent.prototype.countryCode;
    /**
     * @type {?}
     * @private
     */
    UserProfileInfoComponent.prototype.provinceCode;
    /** @type {?} */
    UserProfileInfoComponent.prototype.onValidate;
    /** @type {?} */
    UserProfileInfoComponent.prototype.editable;
    /**
     * @type {?}
     * @private
     */
    UserProfileInfoComponent.prototype.formChanged;
    /**
     * @type {?}
     * @private
     */
    UserProfileInfoComponent.prototype.dictionaryService;
    /**
     * @type {?}
     * @private
     */
    UserProfileInfoComponent.prototype.locationService;
    /**
     * @type {?}
     * @private
     */
    UserProfileInfoComponent.prototype.fb;
}
