// inet-chat/esm5/src/component/contact-list/contact-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Output, ViewChild } from '@angular/core';
import { ChatMessageService } from "../../message.service";
import { WebSocketClientService } from "inet-core";
var ContactListComponent = /** @class */ (function () {
    function ContactListComponent(messageService, webSocketService) {
        this.messageService = messageService;
        this.webSocketService = webSocketService;
        this.contactSelected = new EventEmitter();
        this.roomSelected = new EventEmitter();
        this.callTo = new EventEmitter();
        this.contacts = [];
        this.userInteract = false;
        this.lastPage = false;
        this.username = iNet.username;
        this.fullname = iNet.displayName;
        this._pageNumber = 0;
        this._pageSize = 20;
        this._loading = false;
    }
    /**
     * @return {?}
     */
    ContactListComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._contactEl = $(this.contactRef.nativeElement);
        this._detectUserInteract();
        this.searchContacts();
        this._statusObserver = this.messageService.socketService.onStateChange.subscribe((/**
         * @return {?}
         */
        function () {
            if (_this.messageService.socketService.closed) {
                _this.statusText = 'Kết nối không thành công';
            }
            else if (_this.messageService.socketService.connecting) {
                _this.statusText = 'Đang kết nối';
            }
            else if (_this.messageService.socketService.connected) {
                _this.statusText = '';
            }
        }));
        // Scroll bottom load more
        this._contactEl.on('scroll', (/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if (_this._loading || _this.lastPage) {
                return;
            }
            if (e.target.scrollTop + e.target.clientHeight > e.target.scrollHeight - 100) {
                _this.loadContacts();
            }
        }));
    };
    /**
     * @return {?}
     */
    ContactListComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this._statusObserver) {
            this._statusObserver.unsubscribe();
            this._statusObserver = null;
        }
    };
    /**
     * @param {?} contact
     * @return {?}
     */
    ContactListComponent.prototype.selectContact = /**
     * @param {?} contact
     * @return {?}
     */
    function (contact) {
        this.contactSelected.emit(contact);
    };
    /**
     * @param {?=} value
     * @return {?}
     */
    ContactListComponent.prototype.searchContacts = /**
     * @param {?=} value
     * @return {?}
     */
    function (value) {
        var _this = this;
        if (value === void 0) { value = ''; }
        value = value.toLowerCase().trim();
        if (value === this._keyword) {
            return;
        }
        this._keyword = value;
        this.messageService.searchContacts(value, (/**
         * @param {?} contacts
         * @return {?}
         */
        function (contacts) {
            _this._pageNumber = 0;
            _this._loading = false;
            _this._contacts = contacts;
            _this.contacts.length = 0;
            _this.loadContacts();
        }));
    };
    /**
     * @return {?}
     */
    ContactListComponent.prototype.loadContacts = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.userInteract || this._loading || !this._contacts) {
            return;
        }
        /** @type {?} */
        var begin = this._pageNumber * this._pageSize;
        this.lastPage = this._contacts && this._contacts.length < begin;
        if (this.lastPage) {
            return;
        }
        this._loading = true;
        this.contacts = this.contacts.concat(this._contacts.slice(begin, begin + this._pageSize));
        this.contacts.forEach((/**
         * @param {?} ele
         * @return {?}
         */
        function (ele) {
            /** @type {?} */
            var __mes = _this.messageService.revertMessage(ele.message) || '';
            /** @type {?} */
            var __mesDisplay = _this.messageService.revertMessageDisplay(ele.message) || '';
            /** @type {?} */
            var typeMessages = __mes.split(":");
            if (typeMessages.length > 0) {
                if (typeMessages[0] === 'uri_message' && (typeMessages[1] === 'video_conference...' || typeMessages[1] === 'video_conference')) {
                    ele.message = '<i class="fa fa-phone pr-1" aria-hidden="true"></i> Cuộc gọi thoại';
                    return ele;
                }
                else {
                    if (typeMessages[0] === 'uri_message' && typeMessages[1] === 'link_preview...') {
                        ele.message = __mesDisplay;
                        return ele;
                    }
                    else {
                        return ele;
                    }
                }
            }
        }));
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this._loading = false;
            _this._pageNumber++;
        }), 300);
        this.selectContact(this.contacts[0]);
    };
    /**
     * @param {?} str
     * @return {?}
     */
    ContactListComponent.prototype.decodeBase64 = /**
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return decodeURIComponent(escape(window.atob(str)));
    };
    /**
     * @private
     * @return {?}
     */
    ContactListComponent.prototype._detectUserInteract = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this._timer = setInterval((/**
         * @return {?}
         */
        function () {
            if (_this._contactEl.is(':visible')) {
                _this.userInteract = true;
                _this.loadContacts();
                clearInterval(_this._timer);
                _this._timer = null;
            }
        }), 100);
    };
    /**
     * @param {?} roomName
     * @return {?}
     */
    ContactListComponent.prototype.selectRoom = /**
     * @param {?} roomName
     * @return {?}
     */
    function (roomName) {
        this.roomSelected.emit(roomName);
    };
    /**
     * @param {?} usercode
     * @return {?}
     */
    ContactListComponent.prototype.callToContact = /**
     * @param {?} usercode
     * @return {?}
     */
    function (usercode) {
        this.callTo.emit(usercode);
    };
    ContactListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'message-contact-list',
                    template: "<div  class=\"contact bg-white\" role=\"button\">\n    <div  class=\"contact-online online\" >\n        <img class=\"contact-image\" userAvatar [usercode]=\"username\">\n    </div>\n    <div  class=\"contact-content\">\n        <div  class=\"contact-name\">{{fullname}}</div>\n    </div>\n    <div  style=\"white-space: nowrap\"><i  class=\"message-chat__action fa fa-ellipsis-v\"></i>\n    </div>\n</div>\n\n<div class=\"contact-list__search\">\n    <search-input [placeholder]=\"'T\u00ECm ki\u1EBFm ...'\" (onChanged)=\"searchContacts($event)\"></search-input>\n    <!--    <div class=\"text-center m-1\">-->\n    <!--        <button (click)=\"selectRoom('company')\" type=\"button\" class=\"btn btn-block btn-light\">-->\n    <!--            <i class=\"fa fa-video-camera\"></i> Cu\u1ED9c h\u1ECDp c\u00F4ng ty</button>-->\n    <!--    </div>-->\n</div>\n<div #contactRef class=\"contact-list__body\">\n    <div *ngFor=\"let contact of contacts\" class=\"contact\" role=\"button\" (click)=\"selectContact(contact)\">\n        <div class=\"contact-online\" [ngClass]=\"{'online': contact.online,'offline': !contact.online }\">\n            <img class=\"contact-image\" userAvatar [usercode]=\"contact.usercode\">\n        </div>\n        <div class=\"contact-content\">\n            <div class=\"contact-name\">{{contact.fullname || contact.usercode}}</div>\n            <div class=\"contact-detail\" [innerHTML]=\"contact.message\"></div>\n        </div>\n        <div *ngIf=\"contact.unread\" class=\"contact-icon badge badge-danger\" style=\"margin-right: 5px\">\n            {{contact.unread}}\n        </div>\n        <div *ngIf=\"contact.usercode !== username\" (click)=\"callToContact(contact)\" style=\"white-space: nowrap\"><i\n                class=\"message-chat__action fa fa-phone\"></i></div>\n        <!--        <i class=\"contact-icon fa fa-circle text-secondary\" [ngClass]=\"{'text-success': contact.online}\"></i>-->\n    </div>\n</div>\n<!--<ul class=\"chat-main\">-->\n<!--    <li data-to=\"blank\">-->\n<!--        <div class=\"chat-box\">-->\n<!--            <div class=\"profile offline bg-size\" style=\"background-image: url(&quot;../assets/images/contact/1.jpg&quot;); background-size: cover; background-position: center center; display: block;\"><img class=\"bg-img\" src=\"../assets/images/contact/1.jpg\" alt=\"Avatar\" style=\"display: none;\"></div>-->\n<!--            <div class=\"details\">-->\n<!--                <h5>Josephin water</h5>-->\n<!--                <h6>Hi, i am josephin. How are you.. ! There are many variations of passages.</h6>-->\n<!--            </div>-->\n<!--            <div class=\"date-status\"><i class=\"ti-pin2\"></i>-->\n<!--                <h6>22/10/19</h6>-->\n<!--                <h6 class=\"font-success status\"> Seen</h6>-->\n<!--            </div>-->\n<!--        </div>-->\n<!--    </li>-->\n<!--    <li class=\"active\" data-to=\"chating\">-->\n<!--        <div class=\"chat-box\">-->\n<!--            <div class=\"profile online bg-size\" style=\"background-image: url(&quot;../assets/images/contact/2.jpg&quot;); background-size: cover; background-position: center center; display: block;\"><img class=\"bg-img\" src=\"../assets/images/contact/2.jpg\" alt=\"Avatar\" style=\"display: none;\"></div>-->\n<!--            <div class=\"details\">-->\n<!--                <h5>Jony Lynetin</h5>-->\n<!--                <h6>Typing................</h6>-->\n<!--            </div>-->\n<!--            <div class=\"date-status\"><i class=\"ti-pin2\"></i>-->\n<!--                <h6>30/11/19</h6>-->\n<!--                <div class=\"badge badge-primary sm\">8</div>-->\n<!--            </div>-->\n<!--        </div>-->\n<!--    </li>-->\n<!--    <li data-to=\"blank\">-->\n<!--        <div class=\"chat-box\">-->\n<!--            <div class=\"profile unreachable bg-size\" style=\"background-image: url(&quot;../assets/images/contact/3.jpg&quot;); background-size: cover; background-position: center center; display: block;\"><img class=\"bg-img\" src=\"../assets/images/contact/3.jpg\" alt=\"Avatar\" style=\"display: none;\"></div>-->\n<!--            <div class=\"details\">-->\n<!--                <h5>Sufiya Elija</h5>-->\n<!--                <h6>I need job, please help me.</h6>-->\n<!--            </div>-->\n<!--            <div class=\"date-status\"><i class=\"ti-pin2\"></i>-->\n<!--                <h6>15/06/19</h6>-->\n<!--                <h6 class=\"font-dark status\"> Sending</h6>-->\n<!--            </div>-->\n<!--        </div>-->\n<!--    </li>-->\n<!--    <li data-to=\"blank\">-->\n<!--        <div class=\"chat-box\">-->\n<!--            <div class=\"profile busy bg-size\" style=\"background-image: url(&quot;../assets/images/contact/4.jpg&quot;); background-size: cover; background-position: center center; display: block;\"><img class=\"bg-img\" src=\"../assets/images/contact/4.jpg\" alt=\"Avatar\" style=\"display: none;\"></div>-->\n<!--            <div class=\"details\">-->\n<!--                <h5>Mukrani Pabelo</h5>-->\n<!--                <h6>Hi, i am josephin. How are you.. ! There are many variations of passages.</h6>-->\n<!--            </div>-->\n<!--            <div class=\"date-status\"><i class=\"ti-pin2\"></i>-->\n<!--                <h6>04/06/19</h6>-->\n<!--                <h6 class=\"font-danger status\"> Failed</h6>-->\n<!--            </div>-->\n<!--        </div>-->\n<!--    </li>-->\n<!--    <li data-to=\"blank\">-->\n<!--        <div class=\"chat-box\">-->\n<!--            <div class=\"profile busy bg-size\" style=\"background-image: url(&quot;../assets/images/contact/2.jpg&quot;); background-size: cover; background-position: center center; display: block;\"><img class=\"bg-img\" src=\"../assets/images/contact/2.jpg\" alt=\"Avatar\" style=\"display: none;\"></div>-->\n<!--            <div class=\"details\">-->\n<!--                <h5>Jhon Deo</h5>-->\n<!--                <h6>Hi, i am josephin. How are you.. ! There are many variations of passages.</h6>-->\n<!--            </div>-->\n<!--            <div class=\"date-status\"><i class=\"ti-pin2\"></i>-->\n<!--                <h6>04/06/19</h6>-->\n<!--                <h6 class=\"font-danger status\"> Failed</h6>-->\n<!--            </div>-->\n<!--        </div>-->\n<!--    </li>-->\n<!--    <li data-to=\"blank\">-->\n<!--        <div class=\"chat-box\">-->\n<!--            <div class=\"profile busy bg-size\" style=\"background-image: url(&quot;../assets/images/contact/1.jpg&quot;); background-size: cover; background-position: center center; display: block;\"><img class=\"bg-img\" src=\"../assets/images/contact/1.jpg\" alt=\"Avatar\" style=\"display: none;\"></div>-->\n<!--            <div class=\"details\">-->\n<!--                <h5>Pabelo Mukrani</h5>-->\n<!--                <h6>Hi, i am josephin. How are you.. ! There are many variations of passages.</h6>-->\n<!--            </div>-->\n<!--            <div class=\"date-status\"><i class=\"ti-pin2\"></i>-->\n<!--                <h6>04/06/19</h6>-->\n<!--                <h6 class=\"font-danger status\"> Failed</h6>-->\n<!--            </div>-->\n<!--        </div>-->\n<!--    </li>-->\n<!--</ul>-->",
                    styles: [".message-chat__action{min-width:34px;width:34px;height:34px;line-height:34px;font-size:17px;text-align:center;border-radius:50%;background:#eff7fe;color:#1c9dea;margin-left:10px;cursor:pointer}.message-chat__action:hover{background:#1c9dea;color:#fff}.subject-text,.watermark .leftwatermark{display:none!important}:host{display:block;position:relative;height:100%}.contact-list__search{position:absolute;top:50px;left:0;width:100%;height:50px;padding:10px;background:#fff;border-bottom:1px solid #d3d3d3}.contact-list__body{position:absolute;top:100px;left:0;bottom:0;width:100%;overflow:auto;background:#fff}.contact-list__body::-webkit-scrollbar{width:5px}.contact-list__body::-webkit-scrollbar-track{background:#f1f1f1}.contact-list__body::-webkit-scrollbar-thumb{background:#d6eaf7}.contact-list__body::-webkit-scrollbar-thumb:hover{background:#d6eaf7}.contact{padding:10px;cursor:pointer;display:flex;align-items:center}.contact:active,.contact:hover{background:#eff7fe}.contact:last-child{border-bottom:1px solid #ddd}.contact-image{min-width:40px;width:40px;height:40px;line-height:40px;font-size:20px;text-align:center;border-radius:50%;margin-right:10px;background:rgba(0,0,0,.1)}.contact-content{flex-grow:1;overflow:hidden;font-size:16px;line-height:20px}.contact-name{margin-bottom:2px;color:rgba(0,0,0,.7);font-weight:700;font-size:14px}.contact-detail{font-size:13px;color:rgba(0,0,0,.5);height:14px;line-height:14px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.contact-icon{font-size:9px}.contact-online.offline::before,.contact-online.online::before{content:\"\";position:absolute;height:13px;width:13px;border-radius:100%;border:2px solid #fff;left:44px}.contact-online.online::before{background:#28a745}.contact-online.offline::before{background:#6c757d}"]
                }] }
    ];
    /** @nocollapse */
    ContactListComponent.ctorParameters = function () { return [
        { type: ChatMessageService },
        { type: WebSocketClientService }
    ]; };
    ContactListComponent.propDecorators = {
        contactSelected: [{ type: Output }],
        roomSelected: [{ type: Output }],
        callTo: [{ type: Output }],
        contactRef: [{ type: ViewChild, args: ['contactRef',] }]
    };
    return ContactListComponent;
}());
export { ContactListComponent };
if (false) {
    /** @type {?} */
    ContactListComponent.prototype.contactSelected;
    /** @type {?} */
    ContactListComponent.prototype.roomSelected;
    /** @type {?} */
    ContactListComponent.prototype.callTo;
    /** @type {?} */
    ContactListComponent.prototype.contactRef;
    /** @type {?} */
    ContactListComponent.prototype.contacts;
    /** @type {?} */
    ContactListComponent.prototype.statusText;
    /** @type {?} */
    ContactListComponent.prototype.userInteract;
    /** @type {?} */
    ContactListComponent.prototype.lastPage;
    /** @type {?} */
    ContactListComponent.prototype.username;
    /** @type {?} */
    ContactListComponent.prototype.fullname;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._contacts;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._statusObserver;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._pageNumber;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._pageSize;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._loading;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._timer;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._contactEl;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._keyword;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype._messageObserver;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    ContactListComponent.prototype.webSocketService;
}
