// inet-ui/esm2015/src/layout/bootstrap/user-profile/bootstrap-user-profile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { Location } from "@angular/common";
import { CoreService, ErrorMessage, NewPassword, NotificationService, UserProfile, UserProfileService } from "inet-core";
import { FormBuilder, FormControl, Validators } from "@angular/forms";
import { CustomValidators } from "ng2-validation";
import { Contact } from "../../../model/contact";
import { Address } from "../../../model/address";
import { GlobalContactService } from "../../../user-profile/global-contact.service";
import { TranslateService } from "@ngx-translate/core";
/** @type {?} */
const newPassword = new FormControl('', Validators.required);
/** @type {?} */
const confirmNewPassword = new FormControl('', CustomValidators.equalTo(newPassword));
export class BootstrapUserProfileComponent {
    /**
     * @param {?} location
     * @param {?} fb
     * @param {?} translate
     * @param {?} notificationService
     * @param {?} contactService
     * @param {?} userProfileService
     * @param {?} coreService
     */
    constructor(location, fb, translate, notificationService, contactService, userProfileService, coreService) {
        this.location = location;
        this.fb = fb;
        this.translate = translate;
        this.notificationService = notificationService;
        this.contactService = contactService;
        this.userProfileService = userProfileService;
        this.coreService = coreService;
        this.generalFormValid = false;
        this.password = new NewPassword();
        this.tabName = 'profile-general-info'; //default tab
        this.userProfile = new UserProfile();
        this.translateSubscription = translate.get(['COMMON.MODULE.USER_PROFILE']).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            this.userTranslations = res['COMMON.MODULE.USER_PROFILE'];
        }));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.loadProfile();
        this.frmChangePassword = this.fb.group({
            password: [null, Validators.required],
            newPassword: newPassword,
            confirmNewPassword: confirmNewPassword
        });
        this.frmSignature = this.fb.group({
            signVerify: [null, Validators.required],
            fileUpload: [null, Validators.required]
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.translateSubscription) {
            this.translateSubscription.unsubscribe();
        }
    }
    /**
     * @private
     * @return {?}
     */
    loadProfile() {
        this.userProfileService.getProfile().subscribe((/**
         * @param {?} profile
         * @return {?}
         */
        (profile) => {
            this.userProfile = Object.assign(new UserProfile(), profile);
            this.profile = Object.assign(new Contact(), profile);
            // console.log("profile on load: ",this.profile);
            this.profile.lastName = this.userProfile.lname;
            this.profile.middleName = this.userProfile.mname;
            this.profile.firstName = this.userProfile.fname;
            if (this.userProfile.birthday) {
                this.profile.fullDateOfBirth = new Date(this.userProfile.birthday);
            }
            if (this.userProfile.phone) {
                this.profile.mobilePhone = this.userProfile.phone;
            }
            if (this.userProfile.email) {
                this.profile.primaryEmail = this.userProfile.email;
            }
            // if (this.profile.dateOfBirth) {
            //     this.profile.fullDateOfBirth = new Date(this.profile.dateOfBirth);
            // }
            this.profile.fullName = [this.profile.lastName, this.profile.middleName, this.profile.firstName].join(' ');
            if (!this.profile.address) {
                this.profile.address = new Address();
            }
            /*
            if (this.profile.avatar) {
                let avatarUrl: string;
                avatarUrl = this.coreService.getFileUrl(this.profile.avatar);
                if (this.userProfileService.getAvatarVersion() > 0) {
                    avatarUrl = `${avatarUrl}?version=${this.userProfileService.getAvatarVersion()}`;
                }
                this.avatarUrl = avatarUrl;
            }
             */
        }));
        this.userProfileService.getAvatarUrl().then((/**
         * @param {?} url
         * @return {?}
         */
        url => {
            this.avatarUrl = url;
        }));
        this.userProfileService.getSignPictureId().then((/**
         * @param {?} pictureId
         * @return {?}
         */
        pictureId => {
            if (pictureId) { //get signature image from contact service
                this.signatureUrl = this.contactService.getSignatureImageUrl();
            }
        }));
        this.userProfileService.getSignVerifyNumber().then((/**
         * @param {?} signVerifyNumber
         * @return {?}
         */
        signVerifyNumber => {
            if (signVerifyNumber) { //get signature image from contact service
                this.signVerifyNumber = signVerifyNumber;
            }
        }));
        /*
        this.contactService.getProfile().subscribe((data: Contact) => {
            this.profile = Object.assign(new Contact(), data);
            if (this.profile.dateOfBirth) {
                this.profile.fullDateOfBirth = new Date(this.profile.dateOfBirth);
            }
            this.profile.fullName = [this.profile.lastName, this.profile.middleName, this.profile.firstName].join(' ');
            if (!this.profile.address) {
                this.profile.address = new Address();
            }
            if(this.profile.avatar) {
                this.avatarUrl = this.coreService.getFileUrl(this.profile.avatar);
            }
        });
        */
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onChange($event) {
        this.generalFormValid = $event;
    }
    /**
     * @return {?}
     */
    back() {
        this.location.back();
    }
    /**
     * @param {?=} $event
     * @return {?}
     */
    updateProfile($event) {
        if (this.generalFormValid) {
            /** @type {?} */
            let userProfile = Object.assign(new UserProfile(), this.profile);
            // console.log(userProfile, this.profile);
            userProfile.fname = this.profile.firstName;
            userProfile.mname = this.profile.middleName;
            userProfile.lname = this.profile.lastName;
            if (this.profile.fullDateOfBirth) {
                userProfile.birthday = this.profile.fullDateOfBirth.valueOf();
            }
            if (this.profile.mobilePhone) {
                userProfile.phone = this.profile.mobilePhone;
            }
            if (this.profile.primaryEmail) {
                userProfile.email = this.profile.primaryEmail;
            }
            if (userProfile.address) {
                userProfile.addressStr = JSON.stringify(userProfile.address);
            }
            delete userProfile.address;
            delete userProfile.fullName;
            this.userProfileService.update(userProfile).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                const { TITLE, UPDATE_PROFILE_SUCCESSFUL_MSG, UPDATE_PROFILE_ERROR_MSG } = this.userTranslations;
                if (ErrorMessage.TYPE !== data.type) {
                    this.notificationService.showMessage(UPDATE_PROFILE_SUCCESSFUL_MSG, 'success', TITLE);
                }
                else {
                    this.notificationService.showMessage(UPDATE_PROFILE_ERROR_MSG, 'error', TITLE);
                }
            }));
        }
    }
    /**
     * @param {?=} $event
     * @return {?}
     */
    changePassword($event) {
        if (this.frmChangePassword.valid) {
            this.contactService.showLoading();
            this.userProfileService.changePassword(this.password).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                const { USERNAME, CHANGE_PASSWORD_SUCCESSFUL_MSG, CURRENT_PASSWORD_INCORRECT, CHANGE_PASSWORD_ERROR_MSG } = this.userTranslations;
                if (ErrorMessage.TYPE !== data.type) {
                    if ("SUCCESS" === data.uuid) {
                        this.notificationService.showMessage(CHANGE_PASSWORD_SUCCESSFUL_MSG, 'sucsess', USERNAME);
                        this.frmChangePassword.reset();
                    }
                    else {
                        this.notificationService.showMessage(CURRENT_PASSWORD_INCORRECT, 'error', USERNAME);
                    }
                }
                else {
                    this.notificationService.showMessage(CHANGE_PASSWORD_ERROR_MSG, 'error', USERNAME);
                }
            }));
        }
    }
    /**
     * @param {?=} $event
     * @return {?}
     */
    updateSignature($event) {
        if (this.frmSignature.valid) {
            /** @type {?} */
            const formData = new FormData(this.frmSignaturePhotoElementRef.nativeElement);
            // Upload Signature
            this.contactService.uploadSignaturePhoto(formData).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                const { SIGNATURE_INFORMATION, UPDATE_SIGNATURE_SUCCESSFUL_MSG, UPDATE_SIGNATURE_ERROR_MSG } = this.userTranslations;
                if (ErrorMessage.TYPE !== data.type) {
                    this.notificationService.showMessage(UPDATE_SIGNATURE_SUCCESSFUL_MSG, 'success', SIGNATURE_INFORMATION);
                }
                else {
                    this.notificationService.showMessage(UPDATE_SIGNATURE_ERROR_MSG, 'error', SIGNATURE_INFORMATION);
                }
            }));
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    changeSignatureFile($event) {
        /** @type {?} */
        const files = $event.target.files;
        if (files.length > 0) {
            /** @type {?} */
            const image = files[0];
            if (image) {
                this.fileName = image['name'];
                /** @type {?} */
                const reader = new FileReader();
                reader.onload = (/**
                 * @param {?} e
                 * @return {?}
                 */
                (e) => {
                    image.url = e.target.result;
                    this.signatureUrl = image.url;
                });
                reader.readAsDataURL(image);
            }
        }
        else {
            this.signatureUrl = null;
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onSelectSignatureFile($event) {
        if (this.signatureFileUpload && this.signatureFileUpload.nativeElement) {
            this.signatureFileUpload.nativeElement.click();
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    updateBlankImage($event) {
        if ($event) {
            this.signatureUrl = iNet.BLANK_IMAGE_URL;
        }
    }
    /**
     * @return {?}
     */
    clearSignatureForm() {
        if (this.signatureFileUpload.nativeElement.value) {
            this.signatureFileUpload.nativeElement.value = '';
            this.fileName = '';
            if (this.profile.avatar) {
                this.avatarUrl = this.coreService.getFileUrl(this.profile.avatar);
            }
        }
    }
    /**
     * @param {?} tabName
     * @return {?}
     */
    onSelectTab(tabName) {
        this.tabName = tabName;
    }
}
BootstrapUserProfileComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-bootstrap-user-profile',
                template: "<div class=\"default-layout tb-visible\">\n    <app-common-toolbar></app-common-toolbar>\n    <div *ngIf=\"tabName!=='sharing-information'\" class=\"container-fluid nav-fixed-top cp-toolbar\">\n        <button *ngIf=\"tabName=='profile-general-info'\" [title]=\"'C\u1EADp nh\u1EADt th\u00F4ng tin c\u00E1 nh\u00E2n'\" (click)=\"updateProfile($event)\" class=\"btn btn-primary btn-sm\" [disabled]=\"!generalFormValid\" type=\"button\">\n            <i class=\"fa fa-save\"></i>\n        </button>\n        <button *ngIf=\"tabName==='profile-signature-info'\" [title]=\"'C\u1EADp nh\u1EADt ch\u1EEF k\u00FD s\u1ED1'\" class=\"btn btn-primary btn-sm\" [disabled]=\"!frmSignature.valid\" type=\"button\"\n                (click)=\"updateSignature($event)\">\n            <i class=\"fa fa-save\"></i> <!--{{'COMMON.MODULE.USER_PROFILE.UPDATE_SIGNATURE' | translate}}-->\n        </button>\n        <button *ngIf=\"tabName==='profile-security-info'\" [title]=\"'\u0110\u1ED5i m\u1EADt kh\u1EA9u'\" class=\"btn btn-primary btn-sm\" [disabled]=\"!frmChangePassword.valid\" type=\"button\"\n                (click)=\"changePassword($event)\">\n                <i class=\"fa fa-key\"></i> <!--{{'COMMON.MODULE.USER_PROFILE.CHANGE_PASSWORD' | translate}}-->\n        </button>\n    </div>\n    <div class=\"cp-content\" [ngClass]=\"tabName==='sharing-information'?'toolbar-none':''\">\n        <div class=\"container-fluid p-1\">\n            <app-user-profile *ngIf=\"profile\" [profile]=\"profile\" [imageAvatarUrl]=\"avatarUrl\" [upload]=\"true\">\n                <ng-template #tabContent>\n                    <tabset>\n                        <tab (select)=\"onSelectTab('profile-general-info')\" id=\"profile-general-info\">\n                            <ng-template tabHeading>\n                                <b><i class=\"fa fa-id-card margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.GENERAL_INFORMATION' | translate}}</b>\n                            </ng-template>\n                            <div class=\"mt-3\">\n                                <app-user-profile-info [profile]=\"profile\" [editable]=\"true\" (onValidate)=\"onChange($event)\"></app-user-profile-info>\n                            </div>\n<!--                            <div class=\"form-group margin-b-form-group row\">-->\n<!--                                <label class=\"control-label col-form-label col-sm-3 font-weight-bold\"></label>-->\n<!--                                <div class=\"col-sm-9 pl-0\">-->\n<!--                                    <button class=\"btn btn-primary btn-sm\" [disabled]=\"!generalFormValid\" type=\"button\" (click)=\"updateProfile($event)\">-->\n<!--                                        <i class=\"fa fa-save\"></i> {{'TOOLBAR.UPDATE' | translate}}-->\n<!--                                    </button>-->\n<!--                                </div>-->\n<!--                            </div>-->\n                        </tab>\n                        <tab (select)=\"onSelectTab('profile-signature-info')\" id=\"profile-signature-info\">\n                            <ng-template tabHeading>\n                                <b><i class=\"fa fa-gear margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.SIGNATURE_INFORMATION' | translate}}</b>\n                            </ng-template>\n                            <div class=\"row mt-3\">\n                                <form #frmSignaturePhoto [formGroup]=\"frmSignature\" (ngSubmit)=\"updateSignature($event)\" class=\"col-sm-12\">\n                                    <div class=\"col-sm-12\">\n                                        <div class=\"form-group margin-b-form-group row\">\n                                            <label class=\"control-label col-form-label col-sm-3 font-weight-bold\">\n                                                {{'COMMON.MODULE.USER_PROFILE.SIGNATURE_VERIFY' | translate}}  <i class=\"required\"></i>:</label>\n                                            <div class=\"col-sm-4 pl-0\">\n                                                <div class=\"input-group\">\n                                                    <div class=\"input-group-prepend\">\n                                                        <div class=\"input-group-text\"><i class=\"fa fa-phone\"></i></div>\n                                                    </div>\n                                                    <input [formControl]=\"frmSignature.controls['signVerify']\" [(ngModel)]=\"signVerifyNumber\" name=\"phone\"\n                                                           class=\"form-control form-control-sm col-xs-12 col-sm-12\" autocomplete=\"off\"\n                                                           [ngClass]=\"{'is-invalid':frmSignature.controls['signVerify'].hasError('required')\n                                                            && frmSignature.controls['signVerify'].touched}\"\n                                                           type=\"text\" maxlength=\"64\"/>\n                                                </div>\n                                                <div *ngIf=\"frmSignature.controls['signVerify'].hasError('required') && frmSignature.controls['signVerify'].touched\" class=\"text-danger mt-1\">\n                                                    <i class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.SIGNATURE_VERIFY_REQUIRED_MSG' | translate}}\n                                                </div>\n\n                                            </div>\n                                        </div>\n                                    </div>\n                                    <div class=\"col-sm-12\">\n                                        <div class=\"form-group margin-b-form-group row\">\n                                            <label class=\"control-label col-form-label col-sm-3 font-weight-bold\">\n                                                {{'COMMON.MODULE.USER_PROFILE.SIGNATURE_FILE' | translate}}  <i class=\"required\"></i>:\n                                            </label>\n                                            <div class=\"col-sm-4 pl-0\">\n                                                <label class=\"i-file-input\">\n                                                    <input #signatureFileUpload [formControl]=\"frmSignature.controls['fileUpload']\" required\n                                                           name=\"fileUpload\" type=\"file\" accept=\"image/*\" (change)=\"changeSignatureFile($event)\"\n                                                           class=\"form-control form-control-sm col-xs-12 col-sm-12\">\n                                                        <span class=\"i-file-container\" [ngClass]=\"{'selected': fileName}\" [attr.data-title]=\"'COMMON.MODULE.USER_PROFILE.CHOOSE' | translate\">\n                                                            <span class=\"i-file-name\" [attr.data-title]=\"fileName || ('COMMON.MODULE.USER_PROFILE.NO_FILE' | translate)\">\n                                                                <i class=\"file-icon fa fa-upload\"></i>\n                                                            </span>\n                                                        </span>\n                                                    <a class=\"remove\" *ngIf=\"fileName\" href=\"javascript:;\" (click)=\"clearSignatureForm()\">\n                                                        <i class=\"fa fa-times\"></i>\n                                                    </a>\n                                                </label>\n                                                <!--input #signatureFileUpload [formControl]=\"frmSignature.controls['fileUpload']\" required\n                                                       name=\"fileUpload\" type=\"file\" accept=\"image/*\" (change)=\"changeSignatureFile($event)\"\n                                                       class=\"form-control form-control-sm col-xs-12 col-sm-12\"-->\n                                                <div *ngIf=\"!signatureUrl || frmSignature.controls['fileUpload'].hasError('required') && frmSignature.controls['fileUpload'].touched\" class=\"text-danger\">\n                                                    <i class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.SIGNATURE_FILE_REQUIRED_MSG' | translate}}\n                                                </div>\n                                                <div class=\"thumb-xl member-thumb mt-3 m-b-10 center-block\">\n                                                    <img *ngIf=\"signatureUrl\" (click)=\"onSelectSignatureFile($event)\" [src]=\"signatureUrl\" (error)=\"updateBlankImage($event)\"\n                                                         class=\"signature-thumbnail border\" alt=\"Signature\">\n                                                    <div *ngIf=\"!signatureUrl\" class=\"text-center upload-signature-photo mt-3\" (click)=\"onSelectSignatureFile($event)\">\n                                                        <i aria-hidden=\"true\" class=\"fa fa-picture-o\"></i>\n                                                    </div>\n                                                </div>\n\n                                            </div>\n                                        </div>\n                                    </div>\n<!--                                    <div class=\"col-sm-12 mt-3\">-->\n<!--                                        <div class=\"form-group margin-b-form-group row\">-->\n<!--                                            <label class=\"control-label col-form-label col-sm-3 font-weight-bold\"></label>-->\n<!--                                            <div class=\"col-sm-9 pl-0\">-->\n<!--                                                <button class=\"btn btn-primary btn-sm\" [disabled]=\"!frmSignature.valid\" type=\"button\" (click)=\"updateSignature($event)\">-->\n<!--                                                    <i class=\"fa fa-save\"></i> {{'COMMON.MODULE.USER_PROFILE.UPDATE_SIGNATURE' | translate}}-->\n<!--                                                </button>-->\n<!--                                            </div>-->\n<!--                                        </div>-->\n<!--                                    </div>-->\n                                </form>\n\n                            </div>\n                        </tab>\n                        <!--tab id=\"profile-access-role\">\n                            <ng-template tabHeading>\n                                <b><i class=\"fa fa-user-secret margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.ACCESS_ROLE' | translate}}</b>\n                            </ng-template>\n                            <div class=\"mt-3\">\n                                <app-access-role></app-access-role>\n                            </div>\n                        </tab-->\n                        <tab (select)=\"onSelectTab('profile-security-info')\" id=\"profile-security-info\">\n                            <ng-template tabHeading>\n                                <b><i class=\"fa fa-lock margin-r-5\"></i> {{'COMMON.MODULE.USER_PROFILE.SECURITY_INFORMATION' | translate}}</b>\n                            </ng-template>\n                            <div class=\"row mt-3\">\n                                <form [formGroup]=\"frmChangePassword\" (ngSubmit)=\"changePassword($event)\" class=\"col-sm-12\">\n                                    <div class=\"col-sm-12 alert alert-info\">\n                                        {{'COMMON.MODULE.USER_PROFILE.CHANGE_PASSWORD_INFO_MSG' | translate}}\n                                    </div>\n                                    <div class=\"col-sm-12\">\n                                        <div class=\"form-group margin-b-form-group row\">\n                                            <label class=\"control-label col-form-label col-sm-3 font-weight-bold\">\n                                                {{'COMMON.MODULE.USER_PROFILE.CURRENT_PASSWORD' | translate}} <i class=\"required\"></i>: </label>\n                                            <div class=\"col-sm-9 pl-0\">\n                                                <input type=\"password\" [(ngModel)]=\"password.oldpass\" name=\"password\"\n                                                       [formControl]=\"frmChangePassword.controls['password']\" required maxlength=\"64\"\n                                                       [ngClass]=\"{'is-invalid':frmChangePassword.controls['password'].hasError('required')\n                                                        && frmChangePassword.controls['password'].touched}\"\n                                                       class=\"form-control col-xs-12 col-sm-12\" />\n\n                                                <div *ngIf=\"frmChangePassword.controls['password'].hasError('required')\n                                                 && frmChangePassword.controls['password'].touched\" class=\"text-danger\">\n                                                    <i class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.CURRENT_PASSWORD_REQUIRED_MSG' | translate}}\n                                                </div>\n                                            </div>\n                                        </div>\n                                    </div>\n                                    <div class=\"col-sm-12\">\n                                        <div class=\"form-group margin-b-form-group row\">\n                                            <label class=\"control-label col-form-label col-sm-3 font-weight-bold\">\n                                                {{'COMMON.MODULE.USER_PROFILE.NEW_PASSWORD' | translate}} <i class=\"required\"></i>: </label>\n                                            <div class=\"col-sm-9 pl-0\">\n                                                <input type=\"password\" [(ngModel)]=\"password.newpass\" name=\"newPassword\"\n                                                       [formControl]=\"frmChangePassword.controls['newPassword']\" required maxlength=\"64\"\n                                                       [ngClass]=\"{'is-invalid':frmChangePassword.controls['newPassword'].hasError('required')\n                                                        && frmChangePassword.controls['newPassword'].touched}\"\n                                                       class=\"form-control col-xs-12 col-sm-12\" />\n\n                                                <div *ngIf=\"frmChangePassword.controls['newPassword'].hasError('required')\n                                                 && frmChangePassword.controls['newPassword'].touched\" class=\"text-danger\">\n                                                    <i class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.NEW_PASSWORD_REQUIRED_MSG' | translate}}\n                                                </div>\n                                            </div>\n                                        </div>\n                                    </div>\n                                    <div class=\"col-sm-12\">\n                                        <div class=\"form-group margin-b-form-group row\">\n                                            <label class=\"control-label col-form-label col-sm-3 font-weight-bold\">\n                                                {{'COMMON.MODULE.USER_PROFILE.CONFIRM_NEW_PASSWORD' | translate}} <i class=\"required\"></i>: </label>\n                                            <div class=\"col-sm-9 pl-0\">\n                                                <input type=\"password\" [(ngModel)]=\"password.confirmpass\" name=\"confirmNewPassword\"\n                                                       [formControl]=\"frmChangePassword.controls['confirmNewPassword']\" required maxlength=\"64\"\n                                                       [ngClass]=\"{'is-invalid': (frmChangePassword.controls['confirmNewPassword'].hasError('required')\n                                                        || frmChangePassword.controls['confirmNewPassword'].hasError('equalTo'))\n                                                        && frmChangePassword.controls['confirmNewPassword'].touched}\"\n                                                       class=\"form-control col-xs-12 col-sm-12\" />\n\n                                                <div *ngIf=\"frmChangePassword.controls['confirmNewPassword'].hasError('required')\n                                                 && frmChangePassword.controls['confirmNewPassword'].touched\" class=\"text-danger\">\n                                                    <i class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.CONFIRM_NEW_PASSWORD_REQUIRED_MSG' | translate}}\n                                                </div>\n\n                                                <div *ngIf=\"frmChangePassword.controls['confirmNewPassword'].hasError('equalTo')\n                                                 && frmChangePassword.controls['confirmNewPassword'].touched\"  class=\"text-danger\">\n                                                    <i class=\"fa fa-exclamation-triangle\"></i> {{'COMMON.MODULE.USER_PROFILE.PASSWORD_DO_NOT_MATCH' | translate}}\n                                                </div>\n\n                                            </div>\n                                        </div>\n                                    </div>\n<!--                                    <div class=\"col-sm-12\">-->\n<!--                                        <div class=\"form-group margin-b-form-group row\">-->\n<!--                                        <label class=\"control-label col-form-label col-sm-3 font-weight-bold\"></label>-->\n<!--                                            <div class=\"col-sm-9 pl-0\">-->\n<!--                                                <button class=\"btn btn-primary btn-sm\" [disabled]=\"!frmChangePassword.valid\" type=\"button\" (click)=\"changePassword($event)\">-->\n<!--                                                    <i class=\"fa fa-key\"></i> {{'COMMON.MODULE.USER_PROFILE.CHANGE_PASSWORD' | translate}}-->\n<!--                                                </button>-->\n<!--                                            </div>-->\n<!--                                        </div>-->\n<!--                                    </div>-->\n                                </form>\n                            </div>\n                        </tab>\n                        <tab (select)=\"onSelectTab('sharing-information')\" id=\"sharing-information\">\n                            <ng-template tabHeading>\n                                <b><i class=\"fa fa-share-alt-square margin-r-5\"></i> {{'Chia s\u1EBB th\u00F4ng tin' | translate}}</b>\n                            </ng-template>\n                            <div class=\"row mt-3\">\n                                <div class=\"col-lg-12 p-0\">\n                                    <app-sharing-information-template></app-sharing-information-template>\n                                </div>\n                            </div>\n                        </tab>\n                    </tabset>\n                </ng-template>\n            </app-user-profile>\n        </div>\n    </div>\n</div>\n",
                styles: [':host { width: 100%; }']
            }] }
];
/** @nocollapse */
BootstrapUserProfileComponent.ctorParameters = () => [
    { type: Location },
    { type: FormBuilder },
    { type: TranslateService },
    { type: NotificationService },
    { type: GlobalContactService },
    { type: UserProfileService },
    { type: CoreService }
];
BootstrapUserProfileComponent.propDecorators = {
    profile: [{ type: Input }],
    frmSignaturePhotoElementRef: [{ type: ViewChild, args: ['frmSignaturePhoto',] }],
    signatureFileUpload: [{ type: ViewChild, args: ['signatureFileUpload',] }]
};
if (false) {
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.profile;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.generalFormValid;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.password;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.frmChangePassword;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.frmSignature;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.avatarUrl;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.signatureUrl;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.tabName;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.navButtonTitle;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.userTranslations;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.translateSubscription;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.userProfile;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.fileName;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.signVerifyNumber;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.frmSignaturePhotoElementRef;
    /** @type {?} */
    BootstrapUserProfileComponent.prototype.signatureFileUpload;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.location;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.fb;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.translate;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.notificationService;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.contactService;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.userProfileService;
    /**
     * @type {?}
     * @private
     */
    BootstrapUserProfileComponent.prototype.coreService;
}
