// inet-calendar/esm5/src/attendee-list/attendee-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, IterableDiffers, Output } from '@angular/core';
import { CalAttendeeState } from "../utils/model/CalAttendeeState";
var CalendarAttendeeListComponent = /** @class */ (function () {
    function CalendarAttendeeListComponent(differs) {
        this.attendees = [];
        this.editable = false;
        this.expandable = false;
        this.onDelete = new EventEmitter();
        this.expandEnable = true;
        this.expanded = true;
        this.differ = differs.find([]).create(null);
    }
    /**
     * @return {?}
     */
    CalendarAttendeeListComponent.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var change = this.differ.diff(this.attendees);
        // Detect push item on array
        if (change) {
            this.generateDetail();
        }
    };
    /**
     * @param {?} attendee
     * @return {?}
     */
    CalendarAttendeeListComponent.prototype.removeAttendee = /**
     * @param {?} attendee
     * @return {?}
     */
    function (attendee) {
        /** @type {?} */
        var index = this.attendees.indexOf(attendee);
        if (index > -1) {
            this.attendees.splice(index, 1);
            this.onDelete.emit();
        }
    };
    /**
     * @return {?}
     */
    CalendarAttendeeListComponent.prototype.generateDetail = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var accept = 0;
        /** @type {?} */
        var reject = 0;
        /** @type {?} */
        var unknown = 0;
        /** @type {?} */
        var arr = [];
        this.attendees.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item.state === CalAttendeeState.ACCEPTED) {
                item._stateText = 'Tham dự';
                accept++;
            }
            else if (item.state === CalAttendeeState.DECLINED) {
                item._stateText = 'Không tham dự';
                reject++;
            }
            else if (item.state === CalAttendeeState.NEEDS_ACTION) {
                item._stateText = 'Chưa quyết định';
                unknown++;
            }
        }));
        if (accept) {
            arr.push(accept + ' có');
        }
        if (reject) {
            arr.push(reject + ' không');
        }
        if (unknown) {
            arr.push(unknown + ' chưa quyết định');
        }
        this.stateText = arr.join(', ');
    };
    /**
     * @return {?}
     */
    CalendarAttendeeListComponent.prototype.expandAttendees = /**
     * @return {?}
     */
    function () {
        if (this.expandEnable) {
            this.expanded = !this.expanded;
        }
    };
    /**
     * @param {?} attendee
     * @return {?}
     */
    CalendarAttendeeListComponent.prototype.getStateIcon = /**
     * @param {?} attendee
     * @return {?}
     */
    function (attendee) {
        /** @type {?} */
        var cls;
        switch (attendee.state) {
            case 'ACCEPTED':
                cls = 'fa-check accept';
                break;
            case 'DECLINED':
                cls = 'fa-check accept';
                break;
            default:
                cls = 'fa-check unknown';
        }
        return 'attendee-info fa ' + cls;
    };
    CalendarAttendeeListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'calendar-attendee-list',
                    template: "<div *ngIf=\"attendees?.length\" (click)=\"expandAttendees()\" [ngClass]=\"expandEnable ? 'expandable' : ''\"\n     class=\"attendee-detail\">\n    <span class=\"attendee-number\">{{attendees.length}} kh\u00E1ch</span>\n    <i *ngIf=\"expandEnable\" class=\"fa fa-angle-{{expanded ? 'up' : 'down'}}\" style=\"padding-left: 10px\"></i>\n    <div class=\"attendee-desc\">{{stateText}}</div>\n</div>\n<div [ngClass]=\"{'hide': !expanded}\">\n    <div class=\"attendee-item\" *ngFor=\"let attendee of attendees\">\n        <div>\n            <img *ngIf=\"attendee.username; else aliasAvatar\"\n                 class=\"attendee-avatar\"\n                 userAvatar [usercode]=\"attendee.username\">\n            <ng-template #aliasAvatar>\n                <i class=\"attendee-avatar fa fa-question\"></i>\n            </ng-template>\n        </div>\n        <i *ngIf=\"attendee.state\" [title]=\"attendee._stateText\" [ngClass]=\"getStateIcon(attendee)\"></i>\n        <div class=\"attendee-content\">\n            <b>{{attendee.fullname || attendee.alias}} </b>\n            <span> {{attendee.notes}}</span>\n        </div>\n        <div *ngIf=\"editable\">\n            <i class=\"attendee-remove fa fa-times\" (click)=\"removeAttendee(attendee)\"></i>\n        </div>\n    </div>\n</div>",
                    styles: [".attendee-detail{margin-bottom:2px;line-height:16px;display:inline-block}.attendee-detail.expandable{cursor:pointer}.attendee-detail.expandable:hover{background:rgba(0,0,0,.04)}.attendee-number{font-size:14px}.attendee-desc{font-size:13px;color:rgba(0,0,0,.5)}.attendee-item{position:relative;padding:4px 0;overflow:hidden;display:flex;align-items:center}.attendee-item:hover{background:rgba(0,0,0,.04)}.attendee-item:hover .attendee-remove{display:block}.attendee-item .attendee-remove{flex-grow:0;color:rgba(0,0,0,.3);width:24px;height:24px;line-height:24px;cursor:pointer;text-align:center;display:none;font-size:12px}.attendee-avatar{flex-grow:0;width:24px;height:24px;border-radius:50%;margin-right:15px}i.attendee-avatar{text-align:center;line-height:24px;font-size:20px;background:#6c757d;color:#fff}.attendee-content{flex-grow:1;overflow:hidden;font-size:13px;line-height:16px;color:rgba(0,0,0,.5)}.attendee-content b{color:rgba(0,0,0,.7)}.attendee-info{position:absolute;top:50%;left:18px;font-size:8px;width:14px;height:14px;line-height:12px;border-radius:50%;border:1px solid #fff;color:#fff;text-align:center}.attendee-info.accept{background:#28a745}.attendee-info.reject{background:#dc3545}.attendee-info.unknown{background:#6c757d}"]
                }] }
    ];
    /** @nocollapse */
    CalendarAttendeeListComponent.ctorParameters = function () { return [
        { type: IterableDiffers }
    ]; };
    CalendarAttendeeListComponent.propDecorators = {
        attendees: [{ type: Input }],
        editable: [{ type: Input }],
        expandable: [{ type: Input }],
        onDelete: [{ type: Output }]
    };
    return CalendarAttendeeListComponent;
}());
export { CalendarAttendeeListComponent };
if (false) {
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.attendees;
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.editable;
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.expandable;
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.onDelete;
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.differ;
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.stateText;
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.expandEnable;
    /** @type {?} */
    CalendarAttendeeListComponent.prototype.expanded;
}
