// inet-social/esm5/src/comment-list/comment-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { SocialService } from '../social.service';
import { SocialActivity } from '../model/Activity';
import { CommentPostComponent } from "../comment-post/comment-post.component";
var CommentListComponent = /** @class */ (function () {
    function CommentListComponent(socialService) {
        var _this = this;
        this.socialService = socialService;
        this.allowComment = true;
        this.commentLoad = new EventEmitter();
        this.commentParams = {
            pageSize: 5,
            pageNumber: 0
        };
        this.viewDate = this.socialService.viewDate;
        // Receive comment notify real time
        this._commentChange = this.socialService.commentChange.subscribe((/**
         * @param {?} activityId
         * @return {?}
         */
        function (activityId) {
            if (activityId === _this.activity.uuid) {
                _this.load(true);
            }
        }));
    }
    /**
     * @return {?}
     */
    CommentListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // this.load(true);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    CommentListComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.load(true);
    };
    /**
     * @return {?}
     */
    CommentListComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._commentChange.unsubscribe();
    };
    /**
     * @return {?}
     */
    CommentListComponent.prototype.focusForm = /**
     * @return {?}
     */
    function () {
        if (this.commentPost) {
            this.commentPost.focusMessage();
        }
    };
    /**
     * @param {?} params
     * @return {?}
     */
    CommentListComponent.prototype.commentCreate = /**
     * @param {?} params
     * @return {?}
     */
    function (params) {
        params.activity = this.activity.uuid;
        this.socialService.commentPost(params, this.commentPosted.bind(this));
    };
    /**
     * @param {?=} isInit
     * @return {?}
     */
    CommentListComponent.prototype.load = /**
     * @param {?=} isInit
     * @return {?}
     */
    function (isInit) {
        var _this = this;
        if (isInit) {
            this.commentParams.pageNumber = 0;
        }
        else {
            this.commentParams.pageNumber++;
        }
        this.commentParams.activity = this.activity.uuid;
        if (!this.activity.comments) {
            this.activity.comments = [];
        }
        this.socialService.commentWall(this.commentParams, (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var comments = data && data['items'] || [];
            comments.forEach((/**
             * @param {?} comment
             * @return {?}
             */
            function (comment) { return _this.replaceOrAddComment(comment); }));
            _this.activity.comments.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                return b.posted - a.posted;
            }));
            _this.remainComments =
                (data.total - (_this.commentParams.pageNumber + 1) * _this.commentParams.pageSize) || 0;
            _this.commentLoad.emit(data);
        }));
    };
    /**
     * @param {?} comment
     * @return {?}
     */
    CommentListComponent.prototype.commentUpdated = /**
     * @param {?} comment
     * @return {?}
     */
    function (comment) {
        this.replaceOrAddComment(comment);
    };
    /**
     * @param {?} comment
     * @return {?}
     */
    CommentListComponent.prototype.commentDeleted = /**
     * @param {?} comment
     * @return {?}
     */
    function (comment) {
        this.removeComment(comment.uuid);
    };
    /**
     * @private
     * @param {?} comment
     * @param {?} err
     * @return {?}
     */
    CommentListComponent.prototype.commentPosted = /**
     * @private
     * @param {?} comment
     * @param {?} err
     * @return {?}
     */
    function (comment, err) {
        if (comment && comment.uuid && this.commentPost) {
            this.commentPost.resetForm();
            this.replaceOrAddComment(comment, true);
        }
    };
    /**
     * @private
     * @param {?} uuid
     * @return {?}
     */
    CommentListComponent.prototype.getIndexComment = /**
     * @private
     * @param {?} uuid
     * @return {?}
     */
    function (uuid) {
        for (var i = 0; i < this.activity.comments.length; i++) {
            if (this.activity.comments[i].uuid === uuid) {
                return i;
            }
        }
        return -1;
    };
    /**
     * @private
     * @param {?} comment
     * @param {?=} insertAtFirst
     * @return {?}
     */
    CommentListComponent.prototype.replaceOrAddComment = /**
     * @private
     * @param {?} comment
     * @param {?=} insertAtFirst
     * @return {?}
     */
    function (comment, insertAtFirst) {
        /** @type {?} */
        var index = this.getIndexComment(comment.uuid);
        if (index > -1) {
            this.activity.comments[index] = comment;
        }
        else {
            this.activity.comments[insertAtFirst ? 'unshift' : 'push'](comment);
        }
    };
    /**
     * @protected
     * @param {?} uuid
     * @return {?}
     */
    CommentListComponent.prototype.removeComment = /**
     * @protected
     * @param {?} uuid
     * @return {?}
     */
    function (uuid) {
        /** @type {?} */
        var index = this.getIndexComment(uuid);
        if (index > -1) {
            this.activity.comments.splice(index, 1);
        }
    };
    CommentListComponent.decorators = [
        { type: Component, args: [{
                    selector: '[socialCommentList]',
                    template: "<div *ngIf=\"allowComment\" class=\"social-comment\">\n    <img userAvatar class=\"social-comment__avatar\">\n    <div commentPost (onPost)=\"commentCreate($event)\" class=\"social-comment__content\"></div>\n</div>\n\n<div socialComment\n     *ngFor=\"let comment of activity.comments\"\n     [comment]=\"comment\"\n     (onUpdated)=\"commentUpdated($event)\"\n     (onDeleted)=\"commentDeleted($event)\"\n>\n</div>\n\n<div *ngIf=\"remainComments > 0\" (click)=\"load()\" class=\"social-comment__more\">Xem th\u00EAm..</div>\n",
                    styles: [".social-block{display:block;margin-top:10px}.social-action-more{float:right;margin:-5px -10px 0 0}.social-flat-btn{line-height:26px;text-align:center;color:rgba(0,0,0,.5);margin-right:15px;display:inline-block;outline:0!important;border:0;padding:0 7px;background:0 0}.social-flat-btn i{line-height:inherit;font-size:16px}.social-flat-btn span{margin-left:5px}.social-flat-btn.active,.social-flat-btn:hover{color:#2067b0}.social-comment{padding:7px 0;color:#212121;font-size:13px}.social-comment__avatar{width:30px;height:30px;border-radius:50%;margin-right:10px;float:left}.social-comment__content{overflow:hidden;line-height:16px}.social-comment__name{color:#2067b0}.social-comment__action{color:#6c757d;font-size:12px}.social-comment__more{cursor:pointer;color:#2067b0;margin-left:40px;padding-bottom:10px}"]
                }] }
    ];
    /** @nocollapse */
    CommentListComponent.ctorParameters = function () { return [
        { type: SocialService }
    ]; };
    CommentListComponent.propDecorators = {
        activity: [{ type: Input }],
        allowComment: [{ type: Input }],
        commentLoad: [{ type: Output }],
        commentPost: [{ type: ViewChild, args: [CommentPostComponent,] }]
    };
    return CommentListComponent;
}());
export { CommentListComponent };
if (false) {
    /** @type {?} */
    CommentListComponent.prototype.activity;
    /** @type {?} */
    CommentListComponent.prototype.allowComment;
    /** @type {?} */
    CommentListComponent.prototype.commentLoad;
    /** @type {?} */
    CommentListComponent.prototype.commentPost;
    /** @type {?} */
    CommentListComponent.prototype.remainComments;
    /** @type {?} */
    CommentListComponent.prototype.viewDate;
    /**
     * @type {?}
     * @private
     */
    CommentListComponent.prototype.commentParams;
    /**
     * @type {?}
     * @private
     */
    CommentListComponent.prototype._commentChange;
    /**
     * @type {?}
     * @protected
     */
    CommentListComponent.prototype.socialService;
}
