// inet-chat/esm2015/src/open-graph/link-preview/link-preview.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { OpenGraphService } from "../open-graph.service";
export class LinkPreviewComponent {
    /**
     * @param {?} openGraphService
     */
    constructor(openGraphService) {
        this.openGraphService = openGraphService;
        this.removable = true;
        this.onLoad = new EventEmitter();
        this.onDelete = new EventEmitter();
        this._hidden = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.openGraphService.isLink(this.link)) {
            this.loadPreview();
        }
        else if (this.openGraphData) {
            this.loadImageInfo();
        }
        else {
            this.setHidden(true);
        }
    }
    /**
     * @return {?}
     */
    isPreview() {
        return !this.isHidden() && (!!this.link || !!this.openGraphData);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    setData(data) {
        if (!data || !data.title) {
            return;
        }
        this.openGraphData = data;
        this.loadImageInfo();
        this.setHidden(false);
    }
    /**
     * @return {?}
     */
    getData() {
        return this.openGraphData;
    }
    /**
     * @param {?} link
     * @return {?}
     */
    setLink(link) {
        if (this.link === link) {
            return;
        }
        this.link = link;
        this.loadPreview();
    }
    /**
     * @return {?}
     */
    removePreView() {
        this.onDelete.emit(this);
        this.clearData();
    }
    /**
     * @param {?} url
     * @return {?}
     */
    getDomain(url) {
        if (!url) {
            return '';
        }
        /** @type {?} */
        let domain = url.split('/')[2] || '';
        return domain.replace('www.', '');
    }
    /**
     * @return {?}
     */
    clearData() {
        this.link = null;
        this.openGraphData = null;
        this.setHidden(true);
    }
    /**
     * @private
     * @return {?}
     */
    loadPreview() {
        this.setHidden(false);
        if (this.openGraphService.isLink(this.link)) {
            this.openGraphService.loadPreviewLink(this.link, (/**
             * @param {?} data
             * @return {?}
             */
            (data) => {
                if (data) {
                    this.setData(data);
                }
                else {
                    this.setHidden(true);
                }
            }));
        }
        else {
            this.setHidden(true);
        }
    }
    /**
     * @private
     * @return {?}
     */
    loadImageInfo() {
        if (this.openGraphData && this.openGraphData.image) {
            this.openGraphService.loadImageInfo(this.openGraphData.image, (/**
             * @param {?} imageInfo
             * @return {?}
             */
            (imageInfo) => {
                this.openGraphData.imageInfo = imageInfo;
            }));
        }
    }
    /**
     * @private
     * @param {?} hidden
     * @return {?}
     */
    setHidden(hidden) {
        this._hidden = hidden;
    }
    /**
     * @private
     * @return {?}
     */
    isHidden() {
        return this._hidden;
    }
}
LinkPreviewComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-link-preview',
                template: "<div *ngIf=\"!_hidden\" class=\"link-preview\">\n    <div *ngIf=\"!openGraphData; else preview\">\n        <i class=\"fa fa-spinner fa-spin link-preview-spinner\"></i>\n        <div class=\"link-preview-content\">\u0110ang t\u1EA3i ch\u1EBF \u0111\u1ED9 xem tr\u01B0\u1EDBc...</div>\n    </div>\n    <ng-template #preview>\n        <div *ngIf=\"openGraphData?.imageInfo?.available\"\n             [ngClass]=\"openGraphData.imageInfo.landscape ? 'landscape mb-0' : 'portrait mb-0'\"\n             class=\"link-preview-image\">\n            <img [src]=\"openGraphData.image\">\n        </div>\n        <div class=\"link-preview-content bg-white\">\n            <div class=\"link-preview-title\">{{openGraphData.title}}</div>\n            <div class=\"link-preview-text\">{{openGraphData.description}}</div>\n            <div class=\"link-preview-url\">{{getDomain(openGraphData.url)}}</div>\n        </div>\n        <a [href]=\"openGraphData.url\" target=\"_blank\" rel=\"noopener nofollow\" class=\"link-preview-anchor\"></a>\n    </ng-template>\n    <i *ngIf=\"removable\" (click)=\"removePreView()\" class=\"fa fa-times link-preview-remove\"></i>\n</div>",
                styles: [".link-preview{margin-bottom:15px;display:block;overflow:hidden;border:1px solid rgba(0,0,0,.1);position:relative;color:#555;width:300px}.link-preview-spinner{float:left;width:44px;height:44px;font-size:20px;line-height:44px;text-align:center}.link-preview-anchor{position:absolute;top:0;right:0;bottom:0;left:0;z-index:1;opacity:0;font-size:10px}.link-preview-remove{position:absolute;top:8px;right:8px;width:24px;height:24px;line-height:24px;border-radius:50%;background:rgba(0,0,0,.2);color:#fff;text-align:center;z-index:2;cursor:pointer;font-size:14px}.link-preview-remove:hover{background:rgba(0,0,0,.4)}.link-preview-image{overflow:hidden}.link-preview-image.landscape{margin-bottom:5px;border-bottom:1px solid rgba(0,0,0,.1);max-height:250px;width:330px}.link-preview-image.landscape img{width:100%;-o-object-fit:contain;object-fit:contain;height:100%}.link-preview-image.portrait{float:left;width:160px;height:160px;position:relative}.link-preview-image.portrait img{position:absolute;min-width:100%;min-height:100%;top:50%;left:50%;transform:translate3d(-50%,-50%,0)}.link-preview-content{overflow:hidden;padding:10px 30px 10px 15px}.link-preview-title{color:#333;font-size:15px;line-height:22px}.link-preview-text{overflow:hidden;margin-top:7px;font-size:10px;line-height:18px}.link-preview-url{margin-top:7px;text-transform:uppercase;font-size:12px;color:#888}"]
            }] }
];
/** @nocollapse */
LinkPreviewComponent.ctorParameters = () => [
    { type: OpenGraphService }
];
LinkPreviewComponent.propDecorators = {
    link: [{ type: Input }],
    openGraphData: [{ type: Input }],
    removable: [{ type: Input }],
    onLoad: [{ type: Output }],
    onDelete: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LinkPreviewComponent.prototype.link;
    /** @type {?} */
    LinkPreviewComponent.prototype.openGraphData;
    /** @type {?} */
    LinkPreviewComponent.prototype.removable;
    /** @type {?} */
    LinkPreviewComponent.prototype.onLoad;
    /** @type {?} */
    LinkPreviewComponent.prototype.onDelete;
    /** @type {?} */
    LinkPreviewComponent.prototype._hidden;
    /**
     * @type {?}
     * @private
     */
    LinkPreviewComponent.prototype.openGraphService;
}
