// inet-ui/esm5/src/layout/bootstrap/abstract-side-nav.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostListener, Input, Output } from "@angular/core";
var AbstractSideNavComponent = /** @class */ (function () {
    function AbstractSideNavComponent(elementRef) {
        this.elementRef = elementRef;
        this.opened = false;
        this.onClose = new EventEmitter();
        this.onClear = new EventEmitter();
        this.onLoad = new EventEmitter();
        this.overflowCls = 'overflow-hidden';
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    AbstractSideNavComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        /** @type {?} */
        var currentValue = changes['opened']["currentValue"];
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (currentValue) {
                _this.open();
            }
            else if (changes['opened']['previousValue'] != currentValue) {
                _this.close();
            }
        }), 1);
    };
    /**
     * @return {?}
     */
    AbstractSideNavComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        if (!document.body.classList.contains(this.overflowCls)) {
            document.body.classList.add(this.overflowCls);
        }
    };
    /**
     * @return {?}
     */
    AbstractSideNavComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        if (document.body.classList.contains(this.overflowCls)) {
            document.body.classList.remove(this.overflowCls);
        }
        this.opened = false;
        this.onClose.emit(this.opened);
    };
    /**
     * @return {?}
     */
    AbstractSideNavComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} event
     * @return {?}
     */
    AbstractSideNavComponent.prototype.clickOut = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.elementRef.nativeElement.contains(event.target)
            && this.opened && !event.target.closest('.nav-item') &&
            !event.target.closest('.search-input__icon')) {
            this.close();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    AbstractSideNavComponent.prototype.clickInside = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.elementRef.nativeElement.contains(event.target)) {
        }
    };
    AbstractSideNavComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-admin-side-nav',
                    template: ""
                }] }
    ];
    /** @nocollapse */
    AbstractSideNavComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AbstractSideNavComponent.propDecorators = {
        opened: [{ type: Input }],
        onClose: [{ type: Output }],
        onClear: [{ type: Output }],
        onLoad: [{ type: Output }],
        clickOut: [{ type: HostListener, args: ['document:click', ['$event'],] }],
        clickInside: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return AbstractSideNavComponent;
}());
export { AbstractSideNavComponent };
if (false) {
    /** @type {?} */
    AbstractSideNavComponent.prototype.opened;
    /** @type {?} */
    AbstractSideNavComponent.prototype.onClose;
    /** @type {?} */
    AbstractSideNavComponent.prototype.onClear;
    /** @type {?} */
    AbstractSideNavComponent.prototype.onLoad;
    /**
     * @type {?}
     * @private
     */
    AbstractSideNavComponent.prototype.overflowCls;
    /** @type {?} */
    AbstractSideNavComponent.prototype.elementRef;
}
