// inet-calendar/esm5/src/utils/CalBuilder.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { vnToLatin } from "./common/Accent";
import { DateUtils } from "./common/DateUtils";
import { AjaxAPI } from "./common/AjaxAPI";
import { CalAttendeeType } from "./model/CalAttendeeType";
import { CalConfigs } from "./CalConfigs";
import { CalType } from "./model/CalType";
import { CalCategory } from "./model/CalCategory";
import { CalCategoryIcon } from "./model/CalCategoryIcon";
import { CalAttendeeRole } from "./model/CalAttendeeRole";
import { CalSubType } from "./model/CalSubType";
import { CalMode } from "./model/CalMode";
import { CalendarCar } from "./CalendarCar";
import { SubFirmDictionary } from "./subfirm/Dictionary";
import { CalAttendeeState } from "./model/CalAttendeeState";
import { LEADER_REFERENCE } from "./CalConstants";
import * as moment_ from 'moment';
/** @type {?} */
var moment = moment_;
window['moment'] = moment;
var CalBuilder = /** @class */ (function () {
    function CalBuilder() {
        var _this = this;
        this.TIME_NEW_UPDATE = 12;
        this.roleUrl = 'xschedule/page/role';
        this.MODE = CalMode;
        this.ROLE = CalAttendeeRole;
        this.calSubType = CalSubType;
        this.calCategory = CalCategory;
        this.vnToLatin = vnToLatin;
        this.dateUtils = DateUtils;
        this.calAttendeeType = CalAttendeeType;
        this.calAttendeeState = CalAttendeeState;
        this.calType = CalType;
        this.subFirmDic = new SubFirmDictionary();
        this.carUtils = new CalendarCar(this.subFirmDic);
        this.configs = CalConfigs;
        this.resources = {
            days: [
                'Chủ nhật',
                'Thứ hai',
                'Thứ ba',
                'Thứ tư',
                'Thứ năm',
                'Thứ sáu',
                'Thứ bảy'
            ],
            statuses: [
                'Mới tạo',
                'Ban hành',
                'Chờ duyệt',
                'Đã hủy',
                'Thông qua',
                'Lịch đơn vị khác'
            ],
            allDay: 'Cả ngày',
            hour: 'giờ',
            minute: 'phút'
        };
        this._userRoles = {};
        this._mapEvents = {};
        this._queue = {};
        this._organs = [];
        this._myOrgans = [];
        this.typeIsDepartment = (/**
         * @param {?} type
         * @return {?}
         */
        function (type) { return type === CalType.DEPARTMENT; });
        this.typeIsCommunity = (/**
         * @param {?} type
         * @return {?}
         */
        function (type) { return type === CalType.COMMUNITY; });
        this.typeIsOrganization = (/**
         * @param {?} type
         * @return {?}
         */
        function (type) { return type === CalType.ORGANIZATION; });
        this.catIsInvitation = (/**
         * @param {?} category
         * @return {?}
         */
        function (category) { return CalCategory.INVITATION === category; });
        this.catIsPersonal = (/**
         * @param {?} category
         * @return {?}
         */
        function (category) { return CalCategory.PERSONAL === category; });
        this.catIsCar = (/**
         * @param {?} category
         * @return {?}
         */
        function (category) { return CalCategory.CAR === category; });
        this.catIsImportant = (/**
         * @param {?} category
         * @return {?}
         */
        function (category) { return CalCategory.IMPORTANT === category; });
        this.formatDateStr = (/**
         * @param {?} date
         * @return {?}
         */
        function (date) { return _this.formatDate(date, _this.getConfigValue('DATE_FORMAT')); });
        this.formatTimeStr = (/**
         * @param {?} date
         * @return {?}
         */
        function (date) { return _this.formatDate(date, _this.getConfigValue('TIME_FORMAT')); });
        this.formatDateTimeStr = (/**
         * @param {?} date
         * @return {?}
         */
        function (date) { return _this.formatDate(date, _this.getConfigValue('DATE_TIME_FORMAT')); });
        this.formatDate = (/**
         * @param {?} date
         * @param {?} format
         * @return {?}
         */
        function (date, format) { return moment(date).format(format); });
        this.getEventStatus = (/**
         * @param {?} mode
         * @return {?}
         */
        function (mode) { return _this.resources.statuses[mode]; });
        this.isCreate = (/**
         * @param {?} mode
         * @return {?}
         */
        function (mode) { return mode === CalMode.CREATE; });
        this.isPublish = (/**
         * @param {?} mode
         * @return {?}
         */
        function (mode) { return mode === CalMode.PUBLISH; });
        this.isSent = (/**
         * @param {?} mode
         * @return {?}
         */
        function (mode) { return mode === CalMode.SENT; });
        this.isCancel = (/**
         * @param {?} mode
         * @return {?}
         */
        function (mode) { return mode === CalMode.CANCEL; });
        this.isOther = (/**
         * @param {?} mode
         * @return {?}
         */
        function (mode) { return mode === CalMode.OTHER; });
    }
    /**
     * @return {?}
     */
    CalBuilder.prototype._applyConfig = /**
     * @return {?}
     */
    function () {
        var _this = this;
        iNet.apply(this, DateUtils);
        // Apply config
        if (window['xCalendarConfigList'] && window['xCalendarConfigList'].length > 0) {
            window['xCalendarConfigList'].forEach((/**
             * @param {?} config
             * @return {?}
             */
            function (config) {
                /** @type {?} */
                var obj = _this.getConfigObj(config.name);
                if (obj) {
                    $.extend(obj, config);
                }
                else {
                    _this.configs.push(config);
                }
            }));
        }
        if (window['xCalendarConfigCustoms']) {
            for (var k in window['xCalendarConfigCustoms']) {
                /** @type {?} */
                var obj = this.getConfigObj(k);
                if (obj) {
                    obj.value = window['xCalendarConfigCustoms'][k];
                }
            }
        }
        // Apply resources
        if (window['xcalendarResources']) {
            this.resources = window['xcalendarResources'];
        }
        this.configs.forEach((/**
         * @param {?} obj
         * @return {?}
         */
        function (obj) {
            if (obj.value === 'false') {
                obj.value = false;
            }
        }));
    };
    /**
     * @param {?} name
     * @return {?}
     */
    CalBuilder.prototype.getConfigValue = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        /** @type {?} */
        var obj = this.getConfigObj(name);
        return obj && obj.value;
    };
    /**
     * @param {?} name
     * @return {?}
     */
    CalBuilder.prototype.getConfigObj = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        for (var i = 0; i < this.configs.length; i++) {
            if (this.configs[i].name === name) {
                return this.configs[i];
            }
        }
    };
    /**
     * @param {?} name
     * @param {?} executor
     * @param {?} callback
     * @return {?}
     */
    CalBuilder.prototype.addQueue = /**
     * @param {?} name
     * @param {?} executor
     * @param {?} callback
     * @return {?}
     */
    function (name, executor, callback) {
        var _this = this;
        this._queue[name] = this._queue[name] || [];
        this._queue[name].push(callback);
        if (this._queue[name].length === 1) {
            new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                executor(resolve);
            })).then((/**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                _this._queue[name].forEach((/**
                 * @param {?} fn
                 * @return {?}
                 */
                function (fn) {
                    fn && fn(value);
                }));
                delete _this._queue[name];
            }));
        }
    };
    /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.ready = /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (callback, options) {
        var _this = this;
        this._loadOrgans((/**
         * @param {?} organs
         * @return {?}
         */
        function (organs) {
            callback(organs, _this._getOrganOnInit());
        }), options);
    };
    // Cache all org in a minute for performance
    // Cache all org in a minute for performance
    /**
     * @param {?} key
     * @param {?=} value
     * @return {?}
     */
    CalBuilder.prototype.cacheSessionStorage = 
    // Cache all org in a minute for performance
    /**
     * @param {?} key
     * @param {?=} value
     * @return {?}
     */
    function (key, value) {
        try {
            if (!value) {
                // Cache data in a minute
                /** @type {?} */
                var timeCache = 5 * 60000;
                /** @type {?} */
                var cache = sessionStorage.getItem(key);
                if (cache) {
                    cache = JSON.parse(cache);
                    if (new Date().getTime() - cache.time < timeCache) {
                        return cache.data;
                    }
                    else {
                        sessionStorage.removeItem(key);
                    }
                }
                return false;
            }
            else {
                sessionStorage.setItem(key, JSON.stringify({
                    time: new Date().getTime(),
                    data: value
                }));
            }
            // Fixed prev version
            localStorage.removeItem(key);
        }
        catch (e) { }
    };
    // Cache data on global window, refresh tab will reset
    // Cache data on global window, refresh tab will reset
    /**
     * @param {?} key
     * @param {?=} data
     * @return {?}
     */
    CalBuilder.prototype.cacheDataTemp = 
    // Cache data on global window, refresh tab will reset
    /**
     * @param {?} key
     * @param {?=} data
     * @return {?}
     */
    function (key, data) {
        this._cacheStore = this._cacheStore || {};
        if (data === undefined) {
            return this._cacheStore[key];
        }
        else {
            this._cacheStore[key] = data;
        }
    };
    /**
     * @param {?} orgid
     * @return {?}
     */
    CalBuilder.prototype.setOrgId = /**
     * @param {?} orgid
     * @return {?}
     */
    function (orgid) {
        this.orgid = orgid;
        this._cacheOrgId(orgid);
    };
    /**
     * @return {?}
     */
    CalBuilder.prototype.getOrgId = /**
     * @return {?}
     */
    function () {
        return this.orgid || this._cacheOrgId() || iNet.organId;
    };
    /**
     * @param {?=} orgid
     * @return {?}
     */
    CalBuilder.prototype._cacheOrgId = /**
     * @param {?=} orgid
     * @return {?}
     */
    function (orgid) {
        /** @type {?} */
        var id = 'xcalendar_orgid';
        try {
            if (!orgid) {
                return localStorage.getItem(id);
            }
            else {
                localStorage.setItem(id, orgid);
            }
        }
        catch (e) { }
    };
    /**
     * @param {?=} date
     * @return {?}
     */
    CalBuilder.prototype.getHourMinute = /**
     * @param {?=} date
     * @return {?}
     */
    function (date) {
        if (date.getTime) {
            // date object
            return date.getHours() * 60 + date.getMinutes();
        }
        else {
            // string hour
            date = date.split(':');
            return Number(date[0]) * 60 + Number(date[1]);
        }
    };
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.create = /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (form, success, options) {
        var _this = this;
        return this.postForm({
            url: 'calbuilder/element/create',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event && event.uuid)
                    _this.indexEvent(event);
                success(event);
            })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.updateEventByRole = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        /** @type {?} */
        var organId = params && params.orgid || this.getOrgId();
        this.loadRole(organId, (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var event = _this.getEventByObjectId(params && params.element);
            if (_this.isReviewer(organId)) {
                _this.reviewUpdate(params, success, options);
            }
            else if (_this.isOwnReviewer(organId) && event && _this.meIsCreator(event)) {
                _this.ownerReviewUpdate(params, success, options);
            }
            else {
                _this.update(params, success, options);
            }
        }), options);
    };
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.update = /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (form, success, options) {
        var _this = this;
        return this.postForm({
            url: 'calbuilder/element/update',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event && event.uuid)
                    _this.indexEvent(event);
                success && success(event);
            })
        }, options);
    };
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.reviewUpdate = /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (form, success, options) {
        var _this = this;
        return this.postForm({
            url: 'calbuilder/reviewupdate',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event && event.uuid)
                    _this.indexEvent(event);
                success && success(event);
            })
        }, options);
    };
    /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.ownerReviewUpdate = /**
     * @param {?} form
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (form, success, options) {
        var _this = this;
        return this.postForm({
            url: 'calbuilder/owner/reviewupdate',
            data: form,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event && event.uuid)
                    _this.indexEvent(event);
                success && success(event);
            })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.remove = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        return this.sendRequest({
            url: 'calbuilder/element/delete',
            data: params,
            success: callback
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.cancel = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        /** @type {?} */
        var organId = params && params.orgid || this.getOrgId();
        this.loadRole(organId, (/**
         * @return {?}
         */
        function () {
            if (_this.isReviewer(organId)) {
                _this.reviewCancel(params, success, options);
            }
            else if (_this.isOwnReviewer(organId)) {
                _this.ownerReviewCancel(params, success, options);
            }
        }), options);
    };
    // element, members
    // element, members
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.reviewNotify = 
    // element, members
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/reviewnotify',
            data: params,
            success: success
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.reviewCancel = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        return this.sendRequest({
            url: 'calbuilder/reviewcancel',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event && event.uuid)
                    _this.indexEvent(event);
                success && success(event);
            })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.ownerReviewCancel = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        return this.sendRequest({
            url: 'calbuilder/owner/reviewcancel',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event && event.uuid)
                    _this.indexEvent(event);
                success && success(event);
            })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.deleteAttachment = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        /**
         * fileID: 'id1,id2'
         * element: 'uuid'
         */
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/file/delete',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                success && success(_this.indexEvent(event));
            })
        }, options);
    };
    /**
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadOrganMembers = /**
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (organId, callback, options) {
        organId = organId || this.getOrgId();
        return this.searchAttendee({ orgid: organId }, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.searchAttendee = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        this._searchAttendee(params, (/**
         * @param {?} members
         * @return {?}
         */
        function (members) {
            members.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) { return a.fullname.localeCompare(b.fullname); }));
            success(members);
        }), options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype._searchAttendee = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        return this.sendRequest({
            url: 'calbuilder/attendee/search',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                /** @type {?} */
                var members = data && data.elements || [];
                members = members.filter((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return !item.alias; }));
                members.forEach((/**
                 * @param {?} member
                 * @return {?}
                 */
                function (member) { return _this._indexAccount(member); }));
                success(members);
            }),
            error: (/**
             * @return {?}
             */
            function () {
                success([]);
            })
        }, options, true);
    };
    /**
     * @param {?} account
     * @return {?}
     */
    CalBuilder.prototype._indexAccount = /**
     * @param {?} account
     * @return {?}
     */
    function (account) {
        // Remove suffix email before search user "binh@binhdinh.gov.vn"
        // don't search on suffix "@binhdinh.gov.vn"
        account['_index'] = vnToLatin(account.fullname.toLowerCase()) + ' ' +
            account.username.split('@')[0];
        account.display = this.insertPrefixAttendee(account.fullname || account.username);
        return account;
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadDepartments = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        return this.sendRequest({
            url: 'calbuilder/department/list',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                data = data && data.items || [];
                data.sort((/**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */
                function (a, b) {
                    return a.order - b.order;
                }));
                callback(data);
            }),
            error: (/**
             * @return {?}
             */
            function () {
                callback([]);
            })
        }, options, true);
    };
    /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.getMyOrg = /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (callback, options) {
        var _this = this;
        /** @type {?} */
        var keyCache = 'cal_my_organs';
        // Remove data unused prev version
        localStorage.removeItem(keyCache);
        keyCache += iNet.usercode;
        this._myOrgans = this._myOrgans.length && this._myOrgans || this.cacheSessionStorage(keyCache) || [];
        if (this._myOrgans.length > 0) {
            return callback(this._myOrgans);
        }
        this.sendRequest({
            url: 'plugin/organization/list',
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                /** @type {?} */
                var organs = _this._convertOrg(data && data.elements || []);
                if (organs.length > 0) {
                    organs = _this.visibleMyOrgans(organs);
                    _this._myOrgans = organs;
                    _this.cacheSessionStorage(keyCache, organs);
                }
                callback(organs);
            }),
            error: (/**
             * @return {?}
             */
            function () { return callback([]); })
        }, options, true);
    };
    /**
     * @param {?} organs
     * @return {?}
     */
    CalBuilder.prototype.visibleMyOrgans = /**
     * @param {?} organs
     * @return {?}
     */
    function (organs) {
        var _this = this;
        return organs.filter((/**
         * @param {?} organ
         * @return {?}
         */
        function (organ) {
            /** @type {?} */
            var _organ = _this.getOrgById(organ.firmUUID);
            return _organ && !!_organ.groupType;
        }));
    };
    /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.getAllOrg = /**
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (callback, options) {
        var _this = this;
        /** @type {?} */
        var keyCache = 'cal_organ_list';
        // Remove data unused prev version
        localStorage.removeItem(keyCache);
        keyCache += iNet.usercode;
        this._organs = this._organs.length && this._organs || this.cacheSessionStorage(keyCache) || [];
        // Check all org is cached
        if (this._organs.length > 0) {
            return callback(this._organs);
        }
        return this.sendRequest({
            data: { pageSize: -1 },
            url: 'plugin/organization/search',
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                /** @type {?} */
                var organs = _this._convertOrg(data && data.items || []);
                if (organs.length > 0) {
                    // Only get organ has groupType
                    organs = _this.visibleOrgans(organs);
                    _this._organs = organs;
                    _this.cacheSessionStorage(keyCache, _this._organs);
                }
                callback(organs);
            }),
            error: (/**
             * @return {?}
             */
            function () { return callback([]); })
        }, options, true);
    };
    /**
     * @param {?} organs
     * @return {?}
     */
    CalBuilder.prototype.visibleOrgans = /**
     * @param {?} organs
     * @return {?}
     */
    function (organs) {
        return organs.filter((/**
         * @param {?} organ
         * @return {?}
         */
        function (organ) { return !!organ.groupType; }));
    };
    /**
     * @private
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype._loadOrgans = /**
     * @private
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (callback, options) {
        var _this = this;
        if (this._organs && this._organs.length > 0) {
            callback(this._organs.slice());
        }
        else {
            this.getAllOrg((/**
             * @param {?} organs
             * @return {?}
             */
            function (organs) {
                // Detect user login
                _this.loadRole('', (/**
                 * @param {?} role
                 * @return {?}
                 */
                function (role) {
                    _this.isAuth = !!role.isAuth;
                    if (role.isAuth) {
                        _this.getMyOrg((/**
                         * @param {?} myOrgans
                         * @return {?}
                         */
                        function (myOrgans) {
                            // Move my organs to first
                            for (var i = myOrgans.length - 1; i >= 0; i--) {
                                /** @type {?} */
                                var myOrgan = myOrgans[i];
                                for (var j = 0; j < organs.length; j++) {
                                    /** @type {?} */
                                    var organ = organs[j];
                                    if (organ.firmUUID === myOrgan.firmUUID) {
                                        organs.splice(j, 1);
                                        organ.myOrg = true;
                                        organs.unshift(organ);
                                        break;
                                    }
                                }
                            }
                            callback(organs);
                        }), options);
                    }
                    else {
                        callback(organs);
                    }
                }), options);
            }), options);
        }
    };
    /**
     * @param {?} callback
     * @param {?=} domain
     * @return {?}
     */
    CalBuilder.prototype.getTreeOrg = /**
     * @param {?} callback
     * @param {?=} domain
     * @return {?}
     */
    function (callback, domain) {
        var _this = this;
        this._loadOrgans((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            /** @type {?} */
            var tree = [];
            // Set children is empty
            data.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                item.children = [];
            }));
            // Build tree org list
            data.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.parentID) {
                    /** @type {?} */
                    var parentOrgan = _this.getOrgById(item.parentID);
                    if (parentOrgan) {
                        parentOrgan.children.push(item);
                    }
                }
                else {
                    tree.push(item);
                }
            }));
            tree.data = data;
            callback && callback(tree);
        }), domain);
    };
    /**
     * @param {?=} item
     * @return {?}
     */
    CalBuilder.prototype.indexOrgSearch = /**
     * @param {?=} item
     * @return {?}
     */
    function (item) {
        item._index = vnToLatin([item.firmName, item.firmUUID, item.firmPrefix].join(' ').toLowerCase());
    };
    /**
     * @param {?=} orgid
     * @return {?}
     */
    CalBuilder.prototype.isMyOrg = /**
     * @param {?=} orgid
     * @return {?}
     */
    function (orgid) {
        /** @type {?} */
        var org = this.getOrgById(orgid);
        return org && org.myOrg;
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadEvent = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/element/load',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (!data || !data.uuid) {
                    // Event not found, try to load review event
                    _this.loadRole(params.orgid, (/**
                     * @return {?}
                     */
                    function () {
                        if (_this.isReviewer(params.orgid)) {
                            _this.loadEventReview(params, callback, options);
                        }
                        else {
                            callback && callback(data);
                        }
                    }));
                    return;
                }
                _this.indexEvent(data);
                callback && callback(data);
            })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.updateState = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        if (params.notes) {
            params.notes = this.formatTextUpdate(params.notes);
        }
        return this.loadEvent(params, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadEventReview = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        return this.sendRequest({
            url: 'calbuilder/reviewload',
            data: params,
            success: (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event && event.uuid)
                    _this.indexEvent(event);
                success && success(event);
            })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadGroupEvent = /**
     * @param {?} params
     * @param {?} success
     * @param {?=} options
     * @return {?}
     */
    function (params, success, options) {
        var _this = this;
        return this.sendRequest({
            url: 'calbuilder/element/group',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return success(_this.modifyEvent(data && data.items || [])); })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype.loadOrgEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        this._loadOnDifferYear(this._loadOrgEvents, params, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._loadOrgEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/mainboard/view',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(_this.modifyEvent(data && data.items || [])); })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype.loadCommunityEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        this._loadOnDifferYear(this._loadCommunityEvents, params, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._loadCommunityEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/mainboard/published',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(_this.modifyEvent(data && data.items || [])); }),
            useOrgId: true
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype.loadAllPublishEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        this._loadOnDifferYear(this._loadAllPublishEvents, params, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._loadAllPublishEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/mainboard/published',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(_this.modifyEvent(data && data.items || [])); }),
            removeOrgId: true
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadPublishEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        // Load role
        this.loadRole(params.orgid, (/**
         * @return {?}
         */
        function () {
            // Reviewer
            if (_this.isMyOrg(params.orgid)) {
                _this.loadOrgEvents(params, callback, options);
            }
            else {
                _this.loadCommunityEvents(params, callback, options);
            }
        }), options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype.getReviewEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        this._loadOnDifferYear(this._getReviewEvents, params, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._getReviewEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/reviewlist',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(_this.modifyEvent(data && data.items || [])); })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype.getViewEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        this._loadOnDifferYear(this._getViewEvents, params, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._getViewEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/viewlist',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(_this.modifyEvent(data && data.items || [])); })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype.reviewSearchEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        this._loadOnDifferYear(this._reviewSearchEvents, params, callback, options);
    };
    // members: ",,"
    // members: ",,"
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._reviewSearchEvents = 
    // members: ",,"
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/reviewsearch',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(_this.modifyEvent(data && data.items || [])); })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.getOwnEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        this._loadOnDifferYear(this._getOwnEvents, params, callback, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._getOwnEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        return this.sendRequest({
            type: 'post',
            url: 'calbuilder/element/list',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(_this.modifyEvent(data && data.items || [])); })
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.getListEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        /** @type {?} */
        var promises = [];
        params.orgid = params.orgid || this.getOrgId();
        // Load review event
        this.loadRole(params.orgid, (/**
         * @return {?}
         */
        function () {
            // My events
            promises.push(new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                /** @type {?} */
                var _options = $.extend({}, options);
                _options.error = (/**
                 * @return {?}
                 */
                function () {
                    resolve([]);
                });
                _this.getOwnEvents(params, resolve, _options);
            })));
            if (_this.isReviewer(params.orgid)) {
                // Reviewer
                promises.push(new Promise((/**
                 * @param {?} resolve
                 * @return {?}
                 */
                function (resolve) {
                    /** @type {?} */
                    var _options = $.extend({}, options);
                    _options.error = (/**
                     * @return {?}
                     */
                    function () {
                        resolve([]);
                    });
                    _this.getReviewEvents(params, resolve, _options);
                })));
            }
            else if (_this.isViewer(params.orgid)) {
                // Viewer
                promises.push(new Promise((/**
                 * @param {?} resolve
                 * @return {?}
                 */
                function (resolve) {
                    /** @type {?} */
                    var _options = $.extend({}, options);
                    _options.error = (/**
                     * @return {?}
                     */
                    function () {
                        resolve([]);
                    });
                    _this.getViewEvents(params, resolve, _options);
                })));
            }
            // Load all promises
            Promise.all(promises).then((/**
             * @param {?} values
             * @return {?}
             */
            function (values) {
                /** @type {?} */
                var events = [];
                values.forEach((/**
                 * @param {?} value
                 * @return {?}
                 */
                function (value) { return events = events.concat(value || []); }));
                events = _this.removeEventDuplicate(events);
                callback(_this.modifyEvent(events));
            }));
        }), options);
    };
    /**
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.modifyEvent = /**
     * @param {?} events
     * @return {?}
     */
    function (events) {
        var _this = this;
        if (!events)
            return;
        this.removeEventDuplicate(events);
        events.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return _this.indexEvent(item); }));
        this.sortEvents(events);
        return events;
    };
    /**
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.removeEventDuplicate = /**
     * @param {?} events
     * @return {?}
     */
    function (events) {
        if (events.length > 1) {
            for (var i = 1; i < events.length; i++) {
                for (var j = i - 1; j >= 0; j--) {
                    if (events[i].uuid === events[j].uuid && events[i].day === events[j].day) {
                        events.splice(i, 1);
                        --i;
                        break;
                    }
                }
            }
        }
        return events;
    };
    /**
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.sortEvents = /**
     * @param {?} events
     * @return {?}
     */
    function (events) {
        // sort by start date
        this.sortEventStart(events);
        this.sortEventPriority(events);
        return events;
    };
    /**
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.sortEventStart = /**
     * @param {?} events
     * @return {?}
     */
    function (events) {
        this._sort(events, 'from');
        return events;
    };
    /**
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.sortEventPriority = /**
     * @param {?} events
     * @return {?}
     */
    function (events) {
        // sort by priority on day
        for (var i = 0; i < events.length; i++) {
            /** @type {?} */
            var item = events[i];
            item._order = Number(item._order) || 0;
            if (item._order > 0) {
                moveFirst(item);
                i = events.indexOf(item);
            }
        }
        /**
         * @param {?} item
         * @return {?}
         */
        function moveFirst(item) {
            if (item._order !== 0) {
                /** @type {?} */
                var index = events.indexOf(item);
                if (index > 0) {
                    /** @type {?} */
                    var tmp = events[index - 1];
                    if (item.day === tmp.day && (tmp._order === 0 || item._order < tmp._order)) {
                        events[index] = tmp;
                        events[index - 1] = item;
                        moveFirst(item);
                    }
                }
            }
        }
        return events;
    };
    /**
     * @param {?} arr
     * @param {?=} orderBy
     * @return {?}
     */
    CalBuilder.prototype._sort = /**
     * @param {?} arr
     * @param {?=} orderBy
     * @return {?}
     */
    function (arr, orderBy) {
        for (var i = 1; i < arr.length; i++) {
            if (i === 0)
                continue;
            /** @type {?} */
            var item = arr[i];
            /** @type {?} */
            var prev = arr[i - 1];
            if (item[orderBy] < prev[orderBy]) {
                arr[i] = prev;
                arr[i - 1] = item;
                i -= 2;
            }
        }
    };
    // keywords: string[]
    // keywords: string[]
    /**
     * @param {?} events
     * @param {?} keywords
     * @return {?}
     */
    CalBuilder.prototype.searchEventsByKeywords = 
    // keywords: string[]
    /**
     * @param {?} events
     * @param {?} keywords
     * @return {?}
     */
    function (events, keywords) {
        if (keywords && keywords.length > 0) {
            return events.filter((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                for (var i = 0; i < keywords.length; i++) {
                    if (event._index.indexOf(keywords[i]) < 0) {
                        return false;
                    }
                }
                return true;
            }));
        }
        return events.slice();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.indexEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // Save event with link to get later
        event.objectId = event.uuid + event.mode + event.day;
        this._mapEvents[event.objectId] = event;
        this._mapEvents[event.uuid] = event;
        if (!event._indexed) {
            this.indexEventMembers(event);
            if (window['beforeIndexEvent']) {
                window['beforeIndexEvent'](event);
            }
            event.subjectDisplay = this.formatTextDisplay(event.subject || '');
            event.locationDisplay = this.formatTextDisplay(event.location || '');
            this.indexRepeat(event);
            this.indexEventAttribute(event);
            this.indexEventDate(event);
            this.indexEventSearch(event);
            this.indexCustom(event);
            event.category = event.category || '';
            event.categoryIcon = CalCategoryIcon[event.category];
            event.department = event.departID;
            event._indexed = true;
        }
        return event;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.indexEventMembers = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        /** @type {?} */
        var obj = {};
        for (var k in CalAttendeeRole) {
            obj[k] = [];
        }
        event.members = event.members || [];
        event.attendees = [];
        event.members.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            item.name = _this.attendeeDisplay(item);
            item.value = _this.attendeeValue(item);
            // Type attendee
            if (item.alias) {
                item.type = CalAttendeeType.ALIAS;
            }
            else if (item.username) {
                item.type = CalAttendeeType.USER;
            }
            else {
                item.type = CalAttendeeType.ORGAN;
            }
            obj[item.role] = obj[item.role] || [];
            obj[item.role].push(item);
            if (item.role === CalAttendeeRole.CREATOR) {
                event.creator = item.username;
                event.fullname = item.fullname;
                event.isCreator = _this.usernameIsMe(event.creator);
                event.creatorStr = event.fullname + '(' + event.creator + ')';
            }
            else {
                event.attendees.push(item);
            }
        }));
        event.chairmanStr = this._groupMember2String(obj.CHAIRMAN);
        event.memberStr = this._groupMember2String(obj.MEMBER);
        // Todo: test prepare str
        event.prepareStr = this._groupMember2String(obj.PREPARER);
        event.watcherStr = event.observerStr = this._groupMember2String(obj.OBSERVER);
        event.membersDecode = obj;
    };
    /**
     * @param {?} members
     * @return {?}
     */
    CalBuilder.prototype._groupMember2String = /**
     * @param {?} members
     * @return {?}
     */
    function (members) {
        /** @type {?} */
        var str = '';
        /** @type {?} */
        var data = this.groupAttendeeByOrgan(members);
        for (var k in data) {
            /** @type {?} */
            var items = data[k];
            /** @type {?} */
            var organ = items.filter((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.type === CalAttendeeType.ORGAN; }))[0];
            if (organ) {
                str += '<b>' + organ.organName + '</b>: ';
            }
            // Todo: consider to use this.buildMemberShow
            items.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.type !== CalAttendeeType.ORGAN) {
                    str += item.name + ', ';
                }
            }));
            str = str.slice(0, -2);
            str += '<br>';
        }
        return str.slice(0, -4);
    };
    /**
     * @param {?=} event
     * @return {?}
     */
    CalBuilder.prototype.indexEventAttribute = /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        for (var k in event.attribute) {
            event['_' + k] = event.attribute[k];
        }
        event._order = Number(event._order) || 0;
        if (event._subType) {
            event.type = event._subType;
        }
    };
    /**
     * @param {?} calEvent
     * @return {?}
     */
    CalBuilder.prototype.initEvent = /**
     * @param {?} calEvent
     * @return {?}
     */
    function (calEvent) {
        calEvent = $.extend({}, calEvent);
        this._initEventDate(calEvent);
        calEvent.startTime = calEvent.from.format('H:i');
        calEvent.toTime = calEvent.to.format('H:i');
        return calEvent;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype._initEventDate = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event) {
            /** @type {?} */
            var date = void 0;
            if (event && event.uuid) {
                date = new Date(event.year, 0, event.day);
                if (date.getTime()) {
                    event.from = new Date(date.getTime() + event.startTime * 60 * 1000);
                    event.to = new Date(date.getTime() + event.toTime * 60 * 1000);
                }
            }
            if (!date || !date.getTime()) {
                date = new Date();
                date.setHours(date.getHours() + 1, 0, 0, 0);
                event.from = date;
                date = new Date(date);
                date.setHours(date.getHours() + 1);
                event.to = date;
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.indexEventDate = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this._initEventDate(event);
        // all day event
        event.allDay = event._allday === 'true';
        event.timeStr = this.formatDateStr(event.from);
        if (event.allDay) {
            event.startTimeStr = event.toTimeStr = this.resources.allDay;
            event.hourStr = this.resources.allDay;
        }
        else {
            event.startTimeStr = this.formatTimeStr(event.from);
            event.toTimeStr = this.formatTimeStr(event.to);
            event.hourStr = event.startTimeStr + ' - ' + event.toTimeStr;
        }
        event.timeStr += '; ' + event.hourStr;
        event.start = event.from;
        event.end = event.to;
        // Time create and update
        if (event.createTime) {
            event.createDate = new Date(event.createTime);
            event.createDateStr = this.formatDateTimeStr(event.createDate);
        }
        if (event.updateTime) {
            event.updateDate = new Date(event.updateTime);
            event.updateDateStr = this.formatDateTimeStr(event.updateDate);
        }
        /** @type {?} */
        var hourUpdated = new Date().getTime() - (event.updateTime || event.createTime);
        hourUpdated /= (60 * 60 * 1000);
        if (hourUpdated < 0) {
            hourUpdated = 0;
        }
        if (hourUpdated <= this.TIME_NEW_UPDATE) {
            event.recentUpdated = true;
            event.lastUpdateStr = this.getTextForMinutes(hourUpdated * 60);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.indexEventSearch = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // Index data to search
        event._index = '';
        event._members = [];
        if (event.subject) {
            event._index = vnToLatin(event.subject.toLowerCase());
        }
        // Index location to search
        if (event.location) {
            event._index += ' ' + vnToLatin(event.location.toLowerCase());
        }
        for (var i = 0; i < event.members.length; i++) {
            /** @type {?} */
            var member = event.members[i];
            if (member.fullname)
                event._index += ' ' + vnToLatin(member.fullname.toLowerCase());
            if (member.username) {
                event._index += ' ' + member.username.toLowerCase();
                // attendee
                if (member.role !== CalAttendeeRole.CREATOR) {
                    event._members.push(member.username);
                }
            }
        }
        // Attachments
        if (event.attachments) {
            event.attachments.forEach((/**
             * @param {?} file
             * @return {?}
             */
            function (file) {
                file.orgid = event.firmUUID;
            }));
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.indexRepeat = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        //"FREQ=WEEKLY;UNTIL=20170707;INTERVAL=1;BYDAY=MO"
        if (event.rrule) {
            /** @type {?} */
            var obj_1 = {};
            event.rrule.split(';').forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                item = item.split('=');
                if (item[0] === 'FREQ') {
                    obj_1.type = obj_1.rrule = obj_1.rrmode = item[1];
                }
                else if (item[0] === 'BYDAY') {
                    obj_1.wkdays = item[1].split(',');
                }
            }));
            if (event.lstart) {
                obj_1.lstart = new Date(event.lstart);
            }
            if (event.until) {
                obj_1.until = new Date(event.until);
            }
            event.repeat = obj_1;
        }
    };
    // Custom index
    // Custom index
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.indexCustom = 
    // Custom index
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
    };
    /**
     * @param {?=} orgid
     * @return {?}
     */
    CalBuilder.prototype.isReviewer = /**
     * @param {?=} orgid
     * @return {?}
     */
    function (orgid) {
        orgid = orgid || this.getOrgId();
        return this._userRoles[orgid] && !!this._userRoles[orgid].reviewer;
    };
    /**
     * @param {?=} orgid
     * @return {?}
     */
    CalBuilder.prototype.isOwnReviewer = /**
     * @param {?=} orgid
     * @return {?}
     */
    function (orgid) {
        orgid = orgid || this.getOrgId();
        return this._userRoles[orgid] && !!this._userRoles[orgid].own_reviewer;
    };
    /**
     * @param {?=} orgid
     * @return {?}
     */
    CalBuilder.prototype.isViewer = /**
     * @param {?=} orgid
     * @return {?}
     */
    function (orgid) {
        orgid = orgid || this.getOrgId();
        return this._userRoles[orgid] && !!this._userRoles[orgid].viewer;
    };
    /**
     * @param {?=} query
     * @return {?}
     */
    CalBuilder.prototype.getQueryParams = /**
     * @param {?=} query
     * @return {?}
     */
    function (query) {
        /** @type {?} */
        var params = {};
        query = query || location.search;
        query.replace('?', '').split('&').forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item) {
                /** @type {?} */
                var equalIndex = item.indexOf('=');
                if (equalIndex > -1)
                    params[item.slice(0, equalIndex)] = item.slice(equalIndex + 1);
            }
        }));
        return params;
    };
    /**
     * @param {?=} params
     * @return {?}
     */
    CalBuilder.prototype.buildQueryString = /**
     * @param {?=} params
     * @return {?}
     */
    function (params) {
        /** @type {?} */
        var query = '';
        for (var k in params) {
            query += k + '=' + encodeURIComponent(params[k]) + '&';
        }
        return query.slice(0, -1);
    };
    /**
     * @param {?} usercode
     * @param {?=} domain
     * @return {?}
     */
    CalBuilder.prototype.getUserAvatar = /**
     * @param {?} usercode
     * @param {?=} domain
     * @return {?}
     */
    function (usercode, domain) {
        return AjaxAPI.getUrl('system/userprofile/photo', domain) +
            '?usercode=' + usercode + '&thumbnail=50';
    };
    /**
     * @param {?} event
     * @param {?} callback
     * @param {?=} currentOrganId
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.getPermissionOnEvent = /**
     * @param {?} event
     * @param {?} callback
     * @param {?=} currentOrganId
     * @param {?=} options
     * @return {?}
     */
    function (event, callback, currentOrganId, options) {
        var _this = this;
        /** @type {?} */
        var mode = event.mode;
        if (this.isOther(mode)) {
            callback();
            return;
        }
        if (!event.uuid) {
            return callback({
                save: true,
                saveCreate: true
            });
        }
        currentOrganId = currentOrganId || this.getOrgId() || event.firmUUID;
        this.loadRole(currentOrganId, (/**
         * @return {?}
         */
        function () {
            _this.indexEvent(event);
            /** @type {?} */
            var permissions;
            /** @type {?} */
            var canReview = _this.isReviewer(currentOrganId) ||
                (_this.isOwnReviewer(currentOrganId) && _this.canOwnReviewEvent(event));
            /** @type {?} */
            var canReviewUpdate = !_this.isCreate(mode) && !_this.isCancel(mode) && canReview;
            if (event.firmUUID === currentOrganId) {
                /** @type {?} */
                var isEditMyEvent = _this.canUpdateOwnEvent(event);
                if (event.type === CalSubType.PERSONAL && !_this.getConfigValue('PERSONAL_CAN_PUBLISH')) {
                    // Personal calendar
                    permissions = {
                        save: isEditMyEvent,
                        delete: isEditMyEvent
                    };
                }
                else {
                    permissions = {
                        save: isEditMyEvent || canReviewUpdate,
                        approve: canReview && (_this.isSent(mode) || isEditMyEvent),
                        reject: canReview && _this.isSent(mode),
                        send: isEditMyEvent && !canReview,
                        revert: (_this.isSent(mode) && event.isCreator) ||
                            ((_this.isCancel(mode) || _this.isPublish(mode)) && canReview),
                        cancel: canReview && _this.isPublish(mode),
                        delete: isEditMyEvent
                    };
                }
            }
            else {
                // External update attendees
                permissions = {};
                if (_this.isPublish(event.mode) && _this.hasFirmAttendee(event, currentOrganId) &&
                    _this.canExternalUpdateEvent(event, currentOrganId)) {
                    permissions = { updateAttendee: true };
                }
            }
            permissions.print = true;
            permissions.exportInvitation = _this.catIsInvitation(event.category) && event._templateId;
            callback(permissions);
        }), options);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.canUpdateOwnEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return event.isCreator && this.isCreate(event.mode);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.canReviewEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.isReviewer(event.firmUUID);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.canOwnReviewEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.isOwnReviewer(event.firmUUID) && event.isCreator;
    };
    /**
     * @param {?} event
     * @param {?} ownOrganId
     * @return {?}
     */
    CalBuilder.prototype.canUpdateAttendee = /**
     * @param {?} event
     * @param {?} ownOrganId
     * @return {?}
     */
    function (event, ownOrganId) {
        return this.isFirmEvent(event, ownOrganId) && (this.canUpdateOwnEvent(event) ||
            this.canOwnReviewEvent(event) ||
            this.canReviewEvent(event));
    };
    /**
     * @param {?} event
     * @param {?} organId
     * @return {?}
     */
    CalBuilder.prototype.canExternalUpdateEvent = /**
     * @param {?} event
     * @param {?} organId
     * @return {?}
     */
    function (event, organId) {
        return this.isPublish(event.mode) && !this.isFirmEvent(event, organId) && this.isReviewer(organId);
    };
    /**
     * @param {?} event
     * @param {?} ownOrganId
     * @return {?}
     */
    CalBuilder.prototype.isFirmEvent = /**
     * @param {?} event
     * @param {?} ownOrganId
     * @return {?}
     */
    function (event, ownOrganId) {
        return event.firmUUID === ownOrganId;
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.searchAnyAttendee = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        this.searchUserAsAttendee(params, (/**
         * @param {?} users
         * @return {?}
         */
        function (users) {
            _this.searchOrganAsAttendee(params, (/**
             * @param {?} orgs
             * @return {?}
             */
            function (orgs) {
                orgs.forEach((/**
                 * @param {?} org
                 * @return {?}
                 */
                function (org) { return users.push(org); }));
                callback(users);
            }));
        }), options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.searchUserAsAttendee = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        /** @type {?} */
        var organId = params.orgid || this.getOrgId();
        /** @type {?} */
        var organName = this.getOrgNameById(organId);
        this.loadOrganMembers(organId, (/**
         * @param {?} members
         * @return {?}
         */
        function (members) {
            /** @type {?} */
            var keyword = params.keyword || '';
            /** @type {?} */
            var limit = params.limit || -1;
            /** @type {?} */
            var attendees = [];
            keyword = vnToLatin(keyword).toLowerCase();
            for (var i = 0; i < members.length; i++) {
                /** @type {?} */
                var member = members[i];
                if (!keyword || member._index.indexOf(keyword) > -1) {
                    /** @type {?} */
                    var attendee = _this.memberToAttendee(member);
                    attendee.organId = organId;
                    attendee.organName = organName;
                    attendees.push(attendee);
                    if (limit > 0 && attendees.length >= limit) {
                        break;
                    }
                }
            }
            callback(attendees);
        }), options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.searchOrganAsAttendee = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        this.loadOrgByKeyword(params, (/**
         * @param {?} organs
         * @return {?}
         */
        function (organs) {
            callback(organs.map((/**
             * @param {?} organ
             * @return {?}
             */
            function (organ) { return _this.organToAttendee(organ); })));
        }), options);
    };
    /**
     * @param {?} fullname
     * @return {?}
     */
    CalBuilder.prototype.insertPrefixAttendee = /**
     * @param {?} fullname
     * @return {?}
     */
    function (fullname) {
        if (this.getConfigValue('PREFIX_ATTENDEE_NAME')) {
            return this.getConfigValue('PREFIX_ATTENDEE_NAME') + fullname;
        }
        return fullname;
    };
    /**
     * @param {?} member
     * @return {?}
     */
    CalBuilder.prototype.memberToAttendee = /**
     * @param {?} member
     * @return {?}
     */
    function (member) {
        return {
            username: member.username,
            fullname: member.fullname,
            display: this.insertPrefixAttendee(member.fullname),
            type: CalAttendeeType.USER
        };
    };
    /**
     * @param {?} organ
     * @return {?}
     */
    CalBuilder.prototype.organToAttendee = /**
     * @param {?} organ
     * @return {?}
     */
    function (organ) {
        return {
            organId: organ.firmUUID,
            organName: organ.firmName,
            type: CalAttendeeType.ORGAN
        };
    };
    /**
     * @param {?} alias
     * @param {?} organId
     * @return {?}
     */
    CalBuilder.prototype.aliasToAttendee = /**
     * @param {?} alias
     * @param {?} organId
     * @return {?}
     */
    function (alias, organId) {
        organId = organId || this.getOrgId();
        return {
            alias: alias,
            organId: organId,
            organName: this.getOrgNameById(organId),
            type: CalAttendeeType.ALIAS
        };
    };
    // Search user and orgs to select when create calendar
    // Search user and orgs to select when create calendar
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadAttendeeByName = 
    // Search user and orgs to select when create calendar
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        if (!params || !params.keyword)
            return callback([]);
        // Limit user search
        params.limit = params.limit || 5;
        this.loadUserByKeyword(params, (/**
         * @param {?} users
         * @return {?}
         */
        function (users) {
            _this.loadOrgByKeyword(params, (/**
             * @param {?} orgs
             * @return {?}
             */
            function (orgs) {
                callback(users.concat(orgs));
            }));
        }), options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadUserByKeyword = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        // Load all users and search on client
        this.searchAttendee({
            orgid: params.orgid
        }, (/**
         * @param {?} members
         * @return {?}
         */
        function (members) {
            /** @type {?} */
            var users = [];
            /** @type {?} */
            var keyword = vnToLatin(params.keyword.toLowerCase());
            for (var i = 0; i < members.length; i++) {
                /** @type {?} */
                var user = members[i];
                if (user._index.indexOf(keyword) > -1) {
                    users.push({
                        name: user.fullname,
                        value: user.username,
                        type: CalAttendeeType.USER,
                        member: user
                    });
                    if (users.length >= params.limit) {
                        break;
                    }
                }
            }
            callback(users);
        }), options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadOrgByKeyword = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        /** @type {?} */
        var data = [];
        this._loadOrgans((/**
         * @param {?} orgs
         * @return {?}
         */
        function (orgs) {
            // Filter orgs match
            /** @type {?} */
            var key = vnToLatin(params.keyword).toLowerCase();
            for (var i = 0; i < orgs.length; i++) {
                /** @type {?} */
                var item = orgs[i];
                if (item._index.indexOf(key) > -1) {
                    data.push(item);
                    if (data.length >= params.limit)
                        break;
                }
            }
            callback(data);
        }), options);
    };
    /**
     * @param {?} item
     * @param {?=} domain
     * @param {?=} event
     * @return {?}
     */
    CalBuilder.prototype.getFileUrl = /**
     * @param {?} item
     * @param {?=} domain
     * @param {?=} event
     * @return {?}
     */
    function (item, domain, event) {
        if (!item)
            return '';
        /** @type {?} */
        var orgid = item.orgid || this.getOrgId();
        /** @type {?} */
        var _role = this._userRoles[orgid];
        /** @type {?} */
        var url;
        if (_role && _role.isAuth && this.isMyOrg(orgid)) {
            url = 'calbuilder/binary/view';
        }
        else if (this.hasFirmAttendee(event, this.getOrgId())) {
            orgid = this.getOrgId();
            url = 'calbuilder/binary/view';
        }
        else {
            url = 'calbuilder/mainboard/binary/view';
        }
        return this.getUrlByOrgId(url, orgid, domain) + '?fileID=' + item.uuid + '&element=' + item.folder + '&orgid=' + orgid;
    };
    /**
     * @param {?} item
     * @param {?=} domain
     * @param {?=} event
     * @return {?}
     */
    CalBuilder.prototype.getFileViewUrl = /**
     * @param {?} item
     * @param {?=} domain
     * @param {?=} event
     * @return {?}
     */
    function (item, domain, event) {
        /** @type {?} */
        var url = this.getFileUrl(item, domain, event);
        if (item.file && item.file.slice(-3).toLowerCase() === "pdf") {
            url += '&view=pdf&t=' + new Date().getTime();
        }
        return url;
    };
    /**
     * @param {?} event
     * @param {?} organId
     * @return {?}
     */
    CalBuilder.prototype.hasFirmAttendee = /**
     * @param {?} event
     * @param {?} organId
     * @return {?}
     */
    function (event, organId) {
        if (event && organId) {
            for (var i = 0; i < event.members.length; i++) {
                /** @type {?} */
                var member = event.members[i];
                if (member.type === CalAttendeeType.ORGAN && member.organId === organId) {
                    return true;
                }
            }
        }
    };
    /**
     * @param {?} uuid
     * @return {?}
     */
    CalBuilder.prototype.getOrgByUid = /**
     * @param {?} uuid
     * @return {?}
     */
    function (uuid) {
        for (var i = 0; i < this._organs.length; i++) {
            if (this._organs[i].uuid === uuid) {
                return this._organs[i];
            }
        }
    };
    /**
     * @param {?} orgId
     * @return {?}
     */
    CalBuilder.prototype.getOrgById = /**
     * @param {?} orgId
     * @return {?}
     */
    function (orgId) {
        for (var i = 0; i < this._organs.length; i++) {
            if (this._organs[i].firmUUID === orgId) {
                return this._organs[i];
            }
        }
    };
    /**
     * @param {?} orgId
     * @return {?}
     */
    CalBuilder.prototype.getOrgNameById = /**
     * @param {?} orgId
     * @return {?}
     */
    function (orgId) {
        /** @type {?} */
        var organ = this.getOrgById(orgId);
        return organ && organ.firmName || '';
    };
    /**
     * @param {?} prefix
     * @return {?}
     */
    CalBuilder.prototype.getOrgByPrefix = /**
     * @param {?} prefix
     * @return {?}
     */
    function (prefix) {
        for (var i = 0; i < this._organs.length; i++) {
            if (this._organs[i].firmPrefix === prefix) {
                return this._organs[i];
            }
        }
    };
    /**
     * @param {?} url
     * @param {?} orgid
     * @param {?=} domain
     * @return {?}
     */
    CalBuilder.prototype.getUrlByOrgId = /**
     * @param {?} url
     * @param {?} orgid
     * @param {?=} domain
     * @return {?}
     */
    function (url, orgid, domain) {
        /** @type {?} */
        var item = this.getOrgById(orgid);
        return AjaxAPI.getUrl(url, domain, item && item.firmPrefix);
    };
    /**
     * @param {?} orgid
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadRole = /**
     * @param {?} orgid
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (orgid, callback, options) {
        var _this = this;
        orgid = orgid || this.getOrgId();
        if (this._userRoles[orgid]) {
            return callback(this._userRoles[orgid]);
        }
        this.sendRequest({
            url: this.roleUrl,
            data: { orgid: orgid },
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                try {
                    data = JSON.parse(data);
                    if (data.isAuth) {
                        _this._userRoles[orgid] = data;
                        iNet.usercode = data.usercode;
                    }
                    callback(data);
                }
                catch (e) {
                    callback({});
                }
            }),
            error: (/**
             * @return {?}
             */
            function () {
                callback({});
            })
        }, options, true);
    };
    /**
     * @param {?} organIds
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadRoles = /**
     * @param {?} organIds
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (organIds, callback, options) {
        var _this = this;
        /** @type {?} */
        var promises = [];
        organIds.forEach((/**
         * @param {?} organId
         * @return {?}
         */
        function (organId) {
            promises.push(new Promise(((/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                _this.loadRole(organId, resolve, options);
            }))));
        }));
        Promise.all(promises).then((/**
         * @param {?} values
         * @return {?}
         */
        function (values) { return callback(values); }));
    };
    /**
     * @param {?} events
     * @param {?=} date
     * @param {?=} firstDay
     * @return {?}
     */
    CalBuilder.prototype.generateEventWeekView = /**
     * @param {?} events
     * @param {?=} date
     * @param {?=} firstDay
     * @return {?}
     */
    function (events, date, firstDay) {
        /** @type {?} */
        var segEvents = [];
        /** @type {?} */
        var startWeek = DateUtils.getDateStartWeek(date, firstDay);
        var _loop_1 = function (i) {
            /** @type {?} */
            var dayOfYear = DateUtils.dayOfYear(startWeek);
            segEvents.push(this_1.generateEventView(events.filter((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                return event.day === dayOfYear;
            })), startWeek));
            startWeek.setDate(startWeek.getDate() + 1);
        };
        var this_1 = this;
        for (var i = 0; i < 7; i++) {
            _loop_1(i);
        }
        return segEvents;
    };
    /**
     * @param {?} segEvents
     * @return {?}
     */
    CalBuilder.prototype.getLastUpdateSegEvents = /**
     * @param {?} segEvents
     * @return {?}
     */
    function (segEvents) {
        /** @type {?} */
        var lastUpdate;
        segEvents.forEach((/**
         * @param {?} segEvent
         * @return {?}
         */
        function (segEvent) {
            segEvent.items.forEach((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                /** @type {?} */
                var lastUpdateEvent = event.updateDate || event.createDate;
                if (!lastUpdate || lastUpdate < lastUpdateEvent) {
                    lastUpdate = lastUpdateEvent;
                }
            }));
        }));
        return lastUpdate;
    };
    /**
     * @param {?} events
     * @param {?} date
     * @return {?}
     */
    CalBuilder.prototype.generateEventView = /**
     * @param {?} events
     * @param {?} date
     * @return {?}
     */
    function (events, date) {
        date = new Date(date);
        return {
            day: DateUtils.dayOfYear(date),
            display: this.getDisplayByDay(date.getDay()),
            dayStr: this.formatDateStr(date),
            isToday: DateUtils.dateIsToday(date),
            date: date,
            items: events
        };
    };
    /**
     * @param {?=} day
     * @return {?}
     */
    CalBuilder.prototype.getDisplayByDay = /**
     * @param {?=} day
     * @return {?}
     */
    function (day) {
        return this.resources.days[day];
    };
    /**
     * @param {?} members
     * @param {?=} role
     * @return {?}
     */
    CalBuilder.prototype.buildMembers = /**
     * @param {?} members
     * @param {?=} role
     * @return {?}
     */
    function (members, role) {
        members.forEach((/**
         * @param {?} member
         * @return {?}
         */
        function (member) { return member.role = role || CalAttendeeRole.MEMBER; }));
        return members;
    };
    /**
     * @param {?} attendees
     * @return {?}
     */
    CalBuilder.prototype.groupAttendeeByOrgan = /**
     * @param {?} attendees
     * @return {?}
     */
    function (attendees) {
        /** @type {?} */
        var data = {};
        attendees.forEach((/**
         * @param {?} attendee
         * @return {?}
         */
        function (attendee) {
            data[attendee.organId] = data[attendee.organId] || [];
            data[attendee.organId].push(attendee);
        }));
        return data;
    };
    /**
     * @param {?} event
     * @param {?} attendees
     * @param {?} ownOrganId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.groupAttendeeByOrganRole = /**
     * @param {?} event
     * @param {?} attendees
     * @param {?} ownOrganId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (event, attendees, ownOrganId, callback, options) {
        var _this = this;
        if (attendees.length < 1) {
            return callback([]);
        }
        /** @type {?} */
        var groupByOrgans = this.groupAttendeeByOrgan(attendees);
        /** @type {?} */
        var organIds = Object.keys(groupByOrgans);
        if (organIds.indexOf(ownOrganId) < 0) {
            organIds.push(ownOrganId);
        }
        this.loadRoles(organIds, (/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var internalUpdate = !event.uuid || _this.canUpdateAttendee(event, ownOrganId);
            /** @type {?} */
            var data = [];
            Object.keys(groupByOrgans).forEach((/**
             * @param {?} organId
             * @return {?}
             */
            function (organId) {
                /** @type {?} */
                var isMyOrgan = organId === ownOrganId;
                /** @type {?} */
                var externalUpdate = isMyOrgan && _this.canExternalUpdateEvent(event, organId);
                /** @type {?} */
                var attendees = groupByOrgans[organId];
                data.push({
                    organId: organId,
                    organName: _this.getOrgNameById(organId),
                    internalUpdate: internalUpdate,
                    externalUpdate: externalUpdate,
                    updateAttendee: internalUpdate || externalUpdate,
                    attendees: attendees,
                    canRemoveAttendee: (/**
                     * @param {?} attendee
                     * @return {?}
                     */
                    function (attendee) {
                        return internalUpdate || (externalUpdate && attendee.type !== CalAttendeeType.ORGAN);
                    }),
                    canUpdateState: (/**
                     * @param {?} attendee
                     * @return {?}
                     */
                    function (attendee) {
                        return attendee.readTime !== undefined && isMyOrgan && _this.attendeeIsMe(attendee);
                    })
                });
            }));
            callback(data);
        }), options);
    };
    /**
     * @param {?=} item
     * @return {?}
     */
    CalBuilder.prototype.buildMemberStr = /**
     * @param {?=} item
     * @return {?}
     */
    function (item) {
        item.members = item.members || [];
        return item.members.map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.name; })).join(',');
    };
    /**
     * @param {?} members
     * @return {?}
     */
    CalBuilder.prototype.buildMemberShow = /**
     * @param {?} members
     * @return {?}
     */
    function (members) {
        return (members || []).map((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.name; })).join(';');
    };
    /**
     * @param {?} data
     * @param {?=} event
     * @return {?}
     */
    CalBuilder.prototype.modifyDataUpdate = /**
     * @param {?} data
     * @param {?=} event
     * @return {?}
     */
    function (data, event) {
        var _this = this;
        /** @type {?} */
        var attendeeJson = [];
        // Remove all members and rebuild
        if (data.attendee) {
            /** @type {?} */
            var newAttendees = data.attendee.filter((/**
             * @param {?} attendee
             * @return {?}
             */
            function (attendee) {
                return attendee.readTime === undefined;
            }));
            if (event && event.uuid) {
                /** @type {?} */
                var remainAttendees_1 = data.attendee.filter((/**
                 * @param {?} attendee
                 * @return {?}
                 */
                function (attendee) {
                    return attendee.readTime !== undefined;
                }));
                event.members.forEach((/**
                 * @param {?} member
                 * @return {?}
                 */
                function (member) {
                    if (member.role !== CalAttendeeRole.CREATOR) {
                        if (remainAttendees_1.indexOf(member) < 0) {
                            attendeeJson.push(_this.attendeeRemoveJson(member));
                        }
                    }
                }));
            }
            newAttendees.forEach((/**
             * @param {?} attendee
             * @return {?}
             */
            function (attendee) { return attendeeJson.push(_this.attendeeAddJson(attendee)); }));
        }
        // Secure text update
        if (data.subject) {
            data.subject = this.formatTextUpdate(data.subject);
        }
        if (data.location) {
            data.location = this.formatTextUpdate(data.location);
        }
        data.attendee = JSON.stringify(attendeeJson);
        // Personal calendar
        if (!this.getConfigValue('PERSONAL_CAN_PUBLISH') && data.type === CalSubType.PERSONAL) {
            data.mode = CalMode.CREATE;
        }
        if (data.type === CalSubType.MEET_PEOPLE || data.type === CalSubType.PERSONAL) {
            data._subType = data.type;
            data.type = data.type.split(';')[0];
        }
        else {
            data._subType = '';
        }
        return data;
    };
    /**
     * @param {?} attendee
     * @return {?}
     */
    CalBuilder.prototype.attendeeAddJson = /**
     * @param {?} attendee
     * @return {?}
     */
    function (attendee) {
        /** @type {?} */
        var obj = {
            organId: attendee.organId,
            organName: attendee.organName,
            role: attendee.role || CalAttendeeRole.MEMBER
        };
        switch (attendee.type) {
            case CalAttendeeType.USER:
                obj.members = attendee.username + ':' + attendee.fullname;
                obj.display = attendee.display;
                break;
            case CalAttendeeType.ORGAN:
                if (this.getConfigValue('EVENT_BETWEEN_UNIT')) {
                    obj.firms = attendee.organId + ':' + attendee.organName;
                }
                break;
            default:
                obj.alias = attendee.alias || attendee.organName;
        }
        return obj;
    };
    /**
     * @param {?} attendee
     * @return {?}
     */
    CalBuilder.prototype.attendeeRemoveJson = /**
     * @param {?} attendee
     * @return {?}
     */
    function (attendee) {
        /** @type {?} */
        var obj = {
            organId: attendee.organId
        };
        switch (attendee.type) {
            case CalAttendeeType.USER:
                obj.members = attendee.username + ':REMOVE';
                break;
            case CalAttendeeType.ORGAN:
                obj.firms = attendee.organId + ':REMOVE';
                break;
            default:
                obj.alias = attendee.alias + ':REMOVE';
        }
        return obj;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    CalBuilder.prototype._convertOrg = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        data = data || [];
        for (var i = 0; i < data.length; i++) {
            /** @type {?} */
            var item = data[i];
            item.firmName = item.orgName || item.name;
            item.firmPrefix = item.uriPrefix || item.prefix;
            item.firmUUID = item.organId || item.organiId || '';
            this.indexOrgSearch(item);
        }
        // sort by name
        data.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) {
            return a.firmName.localeCompare(b.firmName);
        }));
        return data;
    };
    /**
     * @param {?} objectId
     * @return {?}
     */
    CalBuilder.prototype.viewEventOnModal = /**
     * @param {?} objectId
     * @return {?}
     */
    function (objectId) {
        var _this = this;
        if (this._mapEvents[objectId]) {
            iNet.plugins.ready((/**
             * @param {?} plugins
             * @return {?}
             */
            function (plugins) {
                if (plugins.xcalendar) {
                    plugins.xcalendar.openView(_this._mapEvents[objectId]);
                }
            }));
        }
    };
    // Save and get event
    // Save and get event
    /**
     * @param {?} objectId
     * @return {?}
     */
    CalBuilder.prototype.getEventByObjectId = 
    // Save and get event
    /**
     * @param {?} objectId
     * @return {?}
     */
    function (objectId) {
        return this._mapEvents[objectId];
    };
    /**
     * @param {?=} minutes
     * @return {?}
     */
    CalBuilder.prototype.getTextForMinutes = /**
     * @param {?=} minutes
     * @return {?}
     */
    function (minutes) {
        minutes = parseInt(minutes) || 1;
        if (minutes >= 60) {
            return Math.round(minutes / 60) + ' ' + this.resources.hour;
        }
        else {
            return minutes + ' ' + this.resources.minute;
        }
    };
    /**
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.groupEventBySession = /**
     * @param {?} events
     * @return {?}
     */
    function (events) {
        /** @type {?} */
        var data = {
            allday: [],
            morning: [],
            afternoon: [],
            evening: []
        };
        events.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item.allDay) {
                data.allday.push(item);
            }
            else if (item.from.getHours() < 12) {
                data.morning.push(item);
            }
            else if (item.from.getHours() < 18) {
                data.afternoon.push(item);
            }
            else {
                data.evening.push(item);
            }
        }));
        for (var k in data) {
            this.sortEvents(data[k]);
        }
        return data;
    };
    /**
     * @param {?} events
     * @param {?} persons
     * @param {?} date
     * @return {?}
     */
    CalBuilder.prototype.groupEventByPerson = /**
     * @param {?} events
     * @param {?} persons
     * @param {?} date
     * @return {?}
     */
    function (events, persons, date) {
        var _this = this;
        /** @type {?} */
        var data = {
            persons: [],
            segEvents: this.generateEventWeekView(events, date)
        };
        /** @type {?} */
        var eventByDays = data.segEvents.map((/**
         * @param {?} segEvent
         * @return {?}
         */
        function (segEvent) {
            return segEvent.items;
        }));
        persons.forEach((/**
         * @param {?} person
         * @return {?}
         */
        function (person) {
            /** @type {?} */
            var eventDays = eventByDays.map((/**
             * @param {?} events
             * @return {?}
             */
            function (events) {
                return events.filter((/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) { return _this.memberIsAttendee(event, person.name); }));
            }));
            data.persons.push({
                person: person,
                eventDays: eventDays
            });
        }));
        return data;
    };
    /**
     * @param {?} events
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.groupEventByLeaders = /**
     * @param {?} events
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (events, organId, callback, options) {
        var _this = this;
        this.loadLeaders(organId, (/**
         * @param {?} leaders
         * @return {?}
         */
        function (leaders) {
            /** @type {?} */
            var dataLeaders = [];
            /** @type {?} */
            var otherEvents = [];
            leaders.forEach((/**
             * @param {?} leader
             * @return {?}
             */
            function (leader) { return dataLeaders.push({
                username: leader.name,
                fullname: leader.value,
                events: []
            }); }));
            events.forEach((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                /** @type {?} */
                var isFound;
                for (var i = 0; i < dataLeaders.length; i++) {
                    /** @type {?} */
                    var item = dataLeaders[i];
                    if (_this.memberIsAttendee(event, item.username)) {
                        isFound = true;
                        item.events.push(event);
                    }
                }
                if (!isFound) {
                    otherEvents.push(event);
                }
            }));
            callback(dataLeaders, otherEvents);
        }), options, true);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.meIsCreator = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        for (var i = event.members.length - 1; i >= 0; i--) {
            if (event.members[i].role === CalAttendeeRole.CREATOR && this.attendeeIsMe(event.members[i])) {
                return true;
            }
        }
        return false;
    };
    /**
     * @param {?} attendde
     * @return {?}
     */
    CalBuilder.prototype.attendeeIsMe = /**
     * @param {?} attendde
     * @return {?}
     */
    function (attendde) {
        return this.usernameIsMe(attendde.username);
    };
    /**
     * @param {?} attendde
     * @return {?}
     */
    CalBuilder.prototype.attendeeDisplay = /**
     * @param {?} attendde
     * @return {?}
     */
    function (attendde) {
        return attendde.alias || attendde.display || attendde.fullname || attendde.username || attendde.organName;
    };
    /**
     * @param {?} attendde
     * @return {?}
     */
    CalBuilder.prototype.attendeeValue = /**
     * @param {?} attendde
     * @return {?}
     */
    function (attendde) {
        return attendde.alias || attendde.username || attendde.organId;
    };
    /**
     * @param {?} text
     * @return {?}
     */
    CalBuilder.prototype.formatTextDisplay = /**
     * @param {?} text
     * @return {?}
     */
    function (text) {
        /** @type {?} */
        var display = this.formatTextUpdate(text);
        // break line
        display = display.replace(/\n/g, '<br>');
        return display;
    };
    /**
     * @param {?} text
     * @return {?}
     */
    CalBuilder.prototype.formatTextUpdate = /**
     * @param {?} text
     * @return {?}
     */
    function (text) {
        /** @type {?} */
        var $fragment = $('<div>').html(text);
        $fragment.find('style,script').remove();
        return $fragment.html();
    };
    /**
     * @param {?} event
     * @param {?} member
     * @return {?}
     */
    CalBuilder.prototype.memberIsAttendee = /**
     * @param {?} event
     * @param {?} member
     * @return {?}
     */
    function (event, member) {
        // for (let i = 0; i < event.members.length; i++) {
        //     if (event.members[i].role !== CalAttendeeRole.CREATOR && event.members[i].username === member) {
        //         return true;
        //     }
        // }
        for (var i = 0; i < event.members.length; i++) {
            if (event.members[i].username === member) {
                return true;
            }
        }
    };
    /**
     * @param {?} event
     * @param {?} username
     * @return {?}
     */
    CalBuilder.prototype.getAttendeeByUsername = /**
     * @param {?} event
     * @param {?} username
     * @return {?}
     */
    function (event, username) {
        for (var i = 0; i < event.members.length; i++) {
            if (event.members[i].username === username) {
                return event.members[i];
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.getAttendeeIsMe = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        for (var i = 0; i < event.members.length; i++) {
            /** @type {?} */
            var member = event.members[i];
            if (member.role !== CalAttendeeRole.CREATOR && this.usernameIsMe(member.username)) {
                return member;
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.eventIsViewed = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var attendee = this.getAttendeeIsMe(event);
        return !attendee || attendee.readTime;
    };
    /**
     * @param {?} username
     * @return {?}
     */
    CalBuilder.prototype.usernameIsMe = /**
     * @param {?} username
     * @return {?}
     */
    function (username) {
        return username === iNet.usercode;
    };
    /**
     * @param {?} data
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    CalBuilder.prototype.sendRequest = /**
     * @param {?} data
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    function (data, options, cacheResponse) {
        var _this = this;
        /** @type {?} */
        var option = this._buildAjaxOptions(data, options);
        if (cacheResponse) {
            /** @type {?} */
            var keyCache_1 = option.url;
            if (option.data && Object.keys(option.data).length > 0) {
                keyCache_1 += JSON.stringify(option.data);
            }
            /** @type {?} */
            var dataCached = this.cacheDataTemp(keyCache_1);
            if (dataCached) {
                return option.success(dataCached);
            }
            this.addQueue(keyCache_1, (/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                /** @type {?} */
                var _option = $.extend({}, option);
                _option.success = (/**
                 * @param {?} data
                 * @return {?}
                 */
                function (data) {
                    _this.cacheDataTemp(keyCache_1, data);
                    resolve(data);
                });
                _option.error = (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    resolve(null, error);
                });
                AjaxAPI.sendRequest(_option);
            }), (/**
             * @param {?} data
             * @param {?} error
             * @return {?}
             */
            function (data, error) {
                if (error) {
                    option.error && option.error(error);
                }
                else {
                    option.success && option.success(data);
                }
            }));
        }
        else {
            return AjaxAPI.sendRequest(option);
        }
    };
    /**
     * @param {?} data
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.postForm = /**
     * @param {?} data
     * @param {?=} options
     * @return {?}
     */
    function (data, options) {
        return AjaxAPI.postForm(this._buildAjaxOptions(data, options));
    };
    /**
     * @param {?} data
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype._buildAjaxOptions = /**
     * @param {?} data
     * @param {?=} options
     * @return {?}
     */
    function (data, options) {
        /** @type {?} */
        var option = $.extend({}, data, options);
        /** @type {?} */
        var orgid = option.data && option.data.orgid || this.getOrgId();
        /** @type {?} */
        var org = this.getOrgById(orgid);
        /** @type {?} */
        var isLocalFirm = org && org.firmPrefix === iNet.path.slice(1);
        /** @type {?} */
        var isUseOrgId = !option.removeOrgId && (isLocalFirm || option.useOrgId);
        if (isUseOrgId) {
            option.data = option.data || {};
            option.data.orgid = orgid;
        }
        else if (option.data) {
            delete option.data.orgid;
        }
        option.url = this.getUrlByOrgId(option.url, orgid, option.domain);
        return option;
    };
    /**
     * @return {?}
     */
    CalBuilder.prototype._getOrganOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var organ;
        /** @type {?} */
        var params = this.getQueryParams();
        // Get by url query
        if (params['orgid']) {
            this.setOrgId(params['orgid']);
            organ = this.getOrgById(params['orgid']);
        }
        // get organ by firm prefix
        if (!organ && iNet.firmPrefix !== 'smartcloud') {
            organ = this.getOrgByPrefix(iNet.firmPrefix);
        }
        // Get organ active recently
        if (!organ) {
            /** @type {?} */
            var organId = this.getOrgId() || iNet.organId;
            organ = this.getOrgById(organId);
        }
        // Get organ local
        if (!organ) {
            organ = this.getOrgByPrefix(iNet.path.slice(1));
        }
        // get organ first
        if (!organ) {
            organ = this._myOrgans && this._myOrgans[0] || this._organs && this._organs[0];
        }
        if (organ) {
            this.setOrgId(organ.firmUUID);
        }
        return organ;
    };
    /**
     * @param {?} cars
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.groupEventByCar = /**
     * @param {?} cars
     * @param {?} events
     * @return {?}
     */
    function (cars, events) {
        /** @type {?} */
        var data = [];
        cars.forEach((/**
         * @param {?} car
         * @return {?}
         */
        function (car) {
            data.push({
                car: car,
                events: []
            });
        }));
        events.forEach((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            /** @type {?} */
            var carData = data[0];
            if (event._carUUID) {
                for (var i = 0; i < data.length; i++) {
                    if (data[i].car.uuid === event._carUUID) {
                        carData = data[i];
                    }
                }
            }
            carData.events.push(event);
        }));
        return data;
    };
    /**
     * @param {?} events
     * @return {?}
     */
    CalBuilder.prototype.filterMeetPeopleEvents = /**
     * @param {?} events
     * @return {?}
     */
    function (events) {
        return events.filter((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            return event._subType === CalSubType.MEET_PEOPLE;
        }));
    };
    /**
     * @param {?} executor
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    CalBuilder.prototype._loadOnDifferYear = /**
     * @param {?} executor
     * @param {?} params
     * @param {?} callback
     * @param {?} options
     * @return {?}
     */
    function (executor, params, callback, options) {
        var _this = this;
        if (!params.from || !params.to || params.from <= params.to) {
            executor.apply(this, [params, callback, options]);
        }
        else {
            // Load in differ year
            /** @type {?} */
            var promises = [];
            /** @type {?} */
            var endOfYear = new Date(params.year + 1, 0);
            endOfYear.setTime(endOfYear.getTime() - 1);
            /** @type {?} */
            var params1_1 = $.extend({}, params);
            /** @type {?} */
            var options1_1 = $.extend({}, options);
            params1_1.to = DateUtils.dayOfYear(endOfYear);
            promises.push(new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                executor.apply(_this, [params1_1, resolve, options1_1]);
            })));
            /** @type {?} */
            var params2_1 = $.extend({}, params);
            /** @type {?} */
            var options2_1 = $.extend({}, options);
            params2_1.year = params.year + 1;
            params2_1.from = 1;
            promises.push(new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                executor.apply(_this, [params2_1, resolve, options2_1]);
            })));
            Promise.all(promises).then((/**
             * @param {?} values
             * @return {?}
             */
            function (values) {
                /** @type {?} */
                var events = [].concat.apply([], values);
                _this.removeEventDuplicate(events);
                callback(events);
            }));
        }
    };
    /**
     * @param {?} params
     * @param {?} date
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.loadMonthEvents = /**
     * @param {?} params
     * @param {?} date
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, date, callback, options) {
        /** @type {?} */
        var from;
        /** @type {?} */
        var to;
        date = new Date(date);
        // Start month
        date.setDate(1);
        // Start date of week
        from = DateUtils.getDateStartWeek(date);
        // End month
        date.setMonth(date.getMonth() + 1);
        date.setTime(date.getTime() - 1);
        // End date of week
        to = DateUtils.getDateEndWeek(date);
        params = $.extend({}, params, this.getRangeParams(from, to));
        if (this.isMyOrg(params.orgid)) {
            this.getListEvents(params, callback, options);
        }
        else {
            this.loadPublishEvents(params, callback, options);
        }
    };
    /**
     * @param {?} calEvent
     * @param {?} domain
     * @return {?}
     */
    CalBuilder.prototype.getExportUrl = /**
     * @param {?} calEvent
     * @param {?} domain
     * @return {?}
     */
    function (calEvent, domain) {
        if (calEvent && calEvent.uuid && calEvent._templateId) {
            return this.getUrlByOrgId('calbuilder/invitation/export', calEvent.firmUUID, domain) +
                '?templateID=' + calEvent._templateId +
                '&element=' + calEvent.uuid;
        }
    };
    /**
     * @param {?} date
     * @return {?}
     */
    CalBuilder.prototype.getWeekParams = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        /** @type {?} */
        var weekRange = DateUtils.getWeekRange(date);
        return this.getRangeParams(weekRange.startWeek, weekRange.endWeek);
    };
    /**
     * @param {?} from
     * @param {?} to
     * @return {?}
     */
    CalBuilder.prototype.getRangeParams = /**
     * @param {?} from
     * @param {?} to
     * @return {?}
     */
    function (from, to) {
        return {
            year: from.getFullYear(),
            from: DateUtils.dayOfYear(from),
            to: DateUtils.dayOfYear(to)
        };
    };
    /**
     * @param {?} date
     * @param {?} weekParams
     * @return {?}
     */
    CalBuilder.prototype.dateEqualWeekParams = /**
     * @param {?} date
     * @param {?} weekParams
     * @return {?}
     */
    function (date, weekParams) {
        /** @type {?} */
        var params = this.getWeekParams(date);
        return date && weekParams && params.year === weekParams.year &&
            params.from === weekParams.from &&
            params.to === weekParams.to;
    };
    /**
     * @param {?} event
     * @param {?=} params
     * @return {?}
     */
    CalBuilder.prototype.eventIsOnRange = /**
     * @param {?} event
     * @param {?=} params
     * @return {?}
     */
    function (event, params) {
        return (params.startTime >= event.startTime && params.startTime < event.toTime) ||
            (params.toTime <= event.toTime && params.toTime > event.startTime) ||
            (params.startTime <= event.startTime && params.toTime >= event.toTime);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.verifyDuplicateEvents = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (params, callback, options) {
        var _this = this;
        /** @type {?} */
        var duplicateEvents = [];
        this.loadPublishEvents({
            year: params.year,
            day: params.day,
            orgid: params.orgid || this.getOrgId()
        }, (/**
         * @param {?} events
         * @return {?}
         */
        function (events) {
            /** @type {?} */
            var allEvents = [];
            if (_this.catIsCar(params.category)) {
                events = events.filter((/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) { return _this.catIsCar(event.category); }));
            }
            else {
                events = events.filter((/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) { return !_this.catIsCar(event.category); }));
            }
            /** @type {?} */
            var detector = (/**
             * @param {?} username
             * @return {?}
             */
            function (username) {
                /** @type {?} */
                var obj = {
                    username: username,
                    events: []
                };
                events.forEach((/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) {
                    /** @type {?} */
                    var member = _this.getAttendeeByUsername(event, username);
                    if (member) {
                        obj.fullname = member.fullname;
                        obj.events.push(event);
                        allEvents.push(event);
                    }
                }));
                if (obj.events.length > 0) {
                    duplicateEvents.push(obj);
                }
            });
            events = events.filter((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                return event.mode === CalMode.PUBLISH && event.uuid !== params.element &&
                    _this.eventIsOnRange(event, params);
            }));
            if (params.attendee) {
                params.attendee.forEach((/**
                 * @param {?} attendee
                 * @return {?}
                 */
                function (attendee) { return detector(attendee.username); }));
            }
            // Todo: creator is attendee
            /** @type {?} */
            var isExist = false;
            for (var i = 0; i < duplicateEvents.length; i++) {
                if (duplicateEvents[i].username === iNet.usercode) {
                    isExist = true;
                    break;
                }
            }
            if (!isExist) {
                detector(iNet.usercode);
            }
            duplicateEvents['allEvents'] = _this.removeEventDuplicate(allEvents);
            callback(duplicateEvents);
        }), options);
    };
    /**
     * @param {?} params
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.needVerifyDuplicate = /**
     * @param {?} params
     * @param {?} event
     * @return {?}
     */
    function (params, event) {
        /** @type {?} */
        var isEventPublished = event.mode === CalMode.PUBLISH;
        /** @type {?} */
        var actionPublish = !isEventPublished && Number(params.mode) === CalMode.PUBLISH;
        /** @type {?} */
        var isTimeChanged = event.uuid && ((params.startTime && event.startTime !== params.startTime) ||
            (params.toTime && event.toTime !== params.toTime) ||
            (params.day && event.day !== params.day)) && isEventPublished;
        return actionPublish || isTimeChanged;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CalBuilder.prototype.subjectDisplayWithMaxLength = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // Subject max length
        /** @type {?} */
        var maxLength = Number(this.getConfigValue('SUBJECT_DISPLAY_MAX_LENGTH')) || 0;
        // Display view more
        if (maxLength > 0 && maxLength < event.subject.length) {
            /** @type {?} */
            var subjectDisplay = event.subject.substr(0, maxLength);
            subjectDisplay = this.formatTextDisplay(subjectDisplay);
            subjectDisplay += '...';
            return this.formatTextDisplay(subjectDisplay);
        }
        return event.subjectDisplay;
    };
    // SubFirm Dictionary Utils
    // SubFirm Dictionary Utils
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.dictCreate = 
    // SubFirm Dictionary Utils
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (data, callback, options) {
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/create',
            data: data,
            success: callback
        }, options);
    };
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.dictUpdate = /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (data, callback, options) {
        data.dictID = data.dictID || data.uuid;
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/update',
            data: data,
            success: callback
        }, options);
    };
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.dictRemove = /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (data, callback, options) {
        data.dictID = data.dictID || data.uuid;
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/remove',
            data: data,
            success: callback
        }, options);
    };
    /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    CalBuilder.prototype.dictList = /**
     * @param {?} params
     * @param {?} callback
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    function (params, callback, options, cacheResponse) {
        // reference: required
        return this.sendRequest({
            type: 'post',
            url: 'subfirm/dictionary/list',
            data: params,
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) { return callback(data && data.items || []); }),
            error: (/**
             * @return {?}
             */
            function () { return callback([]); })
        }, options, cacheResponse);
    };
    // Leader
    // Leader
    /**
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    CalBuilder.prototype.loadLeaders = 
    // Leader
    /**
     * @param {?} organId
     * @param {?} callback
     * @param {?=} options
     * @param {?=} cacheResponse
     * @return {?}
     */
    function (organId, callback, options, cacheResponse) {
        /** @type {?} */
        var params = {
            orgid: organId || this.getOrgId(),
            reference: LEADER_REFERENCE
        };
        this.dictList(params, (/**
         * @param {?} leaders
         * @return {?}
         */
        function (leaders) {
            leaders.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                return b.order - a.order;
            }));
            callback(leaders);
        }), options, cacheResponse);
    };
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.leaderCreateOrUpdate = /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (data, callback, options) {
        if (!data.uuid) {
            data.reference = LEADER_REFERENCE;
            this.dictCreate(data, callback, options);
        }
        else {
            this.dictUpdate(data, callback, options);
        }
    };
    /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    CalBuilder.prototype.leaderRemove = /**
     * @param {?} data
     * @param {?} callback
     * @param {?=} options
     * @return {?}
     */
    function (data, callback, options) {
        this.dictRemove(data, callback, options);
    };
    return CalBuilder;
}());
export { CalBuilder };
if (false) {
    /** @type {?} */
    CalBuilder.prototype.isAuth;
    /** @type {?} */
    CalBuilder.prototype.orgid;
    /** @type {?} */
    CalBuilder.prototype.TIME_NEW_UPDATE;
    /** @type {?} */
    CalBuilder.prototype.roleUrl;
    /** @type {?} */
    CalBuilder.prototype.MODE;
    /** @type {?} */
    CalBuilder.prototype.ROLE;
    /** @type {?} */
    CalBuilder.prototype.calSubType;
    /** @type {?} */
    CalBuilder.prototype.calCategory;
    /** @type {?} */
    CalBuilder.prototype.vnToLatin;
    /** @type {?} */
    CalBuilder.prototype.dateUtils;
    /** @type {?} */
    CalBuilder.prototype.calAttendeeType;
    /** @type {?} */
    CalBuilder.prototype.calAttendeeState;
    /** @type {?} */
    CalBuilder.prototype.calType;
    /** @type {?} */
    CalBuilder.prototype.subFirmDic;
    /** @type {?} */
    CalBuilder.prototype.carUtils;
    /** @type {?} */
    CalBuilder.prototype.configs;
    /** @type {?} */
    CalBuilder.prototype.resources;
    /** @type {?} */
    CalBuilder.prototype._userRoles;
    /** @type {?} */
    CalBuilder.prototype._mapEvents;
    /** @type {?} */
    CalBuilder.prototype._cacheStore;
    /** @type {?} */
    CalBuilder.prototype._queue;
    /** @type {?} */
    CalBuilder.prototype._organs;
    /** @type {?} */
    CalBuilder.prototype._myOrgans;
    /** @type {?} */
    CalBuilder.prototype.typeIsDepartment;
    /** @type {?} */
    CalBuilder.prototype.typeIsCommunity;
    /** @type {?} */
    CalBuilder.prototype.typeIsOrganization;
    /** @type {?} */
    CalBuilder.prototype.catIsInvitation;
    /** @type {?} */
    CalBuilder.prototype.catIsPersonal;
    /** @type {?} */
    CalBuilder.prototype.catIsCar;
    /** @type {?} */
    CalBuilder.prototype.catIsImportant;
    /** @type {?} */
    CalBuilder.prototype.formatDateStr;
    /** @type {?} */
    CalBuilder.prototype.formatTimeStr;
    /** @type {?} */
    CalBuilder.prototype.formatDateTimeStr;
    /** @type {?} */
    CalBuilder.prototype.formatDate;
    /** @type {?} */
    CalBuilder.prototype.getEventStatus;
    /** @type {?} */
    CalBuilder.prototype.isCreate;
    /** @type {?} */
    CalBuilder.prototype.isPublish;
    /** @type {?} */
    CalBuilder.prototype.isSent;
    /** @type {?} */
    CalBuilder.prototype.isCancel;
    /** @type {?} */
    CalBuilder.prototype.isOther;
}
