// angular-slickgrid/esm2015/app/modules/angular-slickgrid/modules/angular-slickgrid.module.js
import * as tslib_1 from "tslib";
var AngularSlickgridModule_1;
import { AngularUtilService } from './../services/angularUtilService';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';
import { AngularSlickgridComponent } from './../components/angular-slickgrid.component';
import { CollectionService } from './../services/collection.service';
import { FilterFactory } from '../filters/filterFactory';
import { GraphqlService } from './../services/graphql.service';
import { GridOdataService } from './../services/grid-odata.service';
import { SlickPaginationComponent } from './../components/slick-pagination.component';
let AngularSlickgridModule = AngularSlickgridModule_1 = class AngularSlickgridModule {
    static forRoot(config = {}) {
        return {
            ngModule: AngularSlickgridModule_1,
            providers: [
                { provide: 'config', useValue: config },
                AngularUtilService,
                CollectionService,
                FilterFactory,
                GraphqlService,
                GridOdataService
            ]
        };
    }
};
AngularSlickgridModule = AngularSlickgridModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [
            CommonModule,
            TranslateModule
        ],
        declarations: [
            AngularSlickgridComponent,
            SlickPaginationComponent
        ],
        exports: [
            AngularSlickgridComponent,
            SlickPaginationComponent
        ],
        entryComponents: [AngularSlickgridComponent]
    })
], AngularSlickgridModule);
export { AngularSlickgridModule };
