// inet-social/esm5/src/localnews/localnews.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var LocalNews = /** @class */ (function () {
    function LocalNews() {
    }
    LocalNews.decorators = [
        { type: Component, args: [{
                    selector: '[localNews]',
                    template: "<div *ngIf=\"data\" class=\"local-news\">\n    <h3 class=\"local-news__title\">\n        {{data.title}}\n    </h3>\n    <div class=\"local-news__image\">\n        <img [src]=\"data.image\">\n    </div>\n    <h4 class=\"local-news__description\">\n        {{data.description}}\n    </h4>\n    <a [href]=\"data.url\" target=\"_blank\" class=\"local-news__url\"></a>\n</div>",
                    styles: [".local-news{margin:10px 0;padding:10px 0;position:relative;overflow:hidden;border-top:1px solid rgba(0,0,0,.1);border-bottom:1px solid rgba(0,0,0,.1)}.local-news__title{font:700 14px arial;color:#333;margin-bottom:5px}.local-news__description{color:#333;font:400 13px/16px arial;margin-bottom:5px;overflow:hidden}.local-news__image{position:relative;margin-right:10px;float:left;width:140px}.local-news__image img{width:100%;min-height:70px;border:1px solid rgba(0,0,0,.01)}.local-news__url{position:absolute;top:0;right:0;bottom:0;left:0;z-index:100}"]
                }] }
    ];
    LocalNews.propDecorators = {
        data: [{ type: Input }]
    };
    return LocalNews;
}());
export { LocalNews };
if (false) {
    /** @type {?} */
    LocalNews.prototype.data;
}
/**
 * @record
 */
export function LocalNewsData() { }
if (false) {
    /** @type {?} */
    LocalNewsData.prototype.image;
    /** @type {?} */
    LocalNewsData.prototype.title;
    /** @type {?} */
    LocalNewsData.prototype.description;
    /** @type {?|undefined} */
    LocalNewsData.prototype.url;
}
