// inet-calendar/esm2015/src/remote-module/time-picker/time-picker-modal.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, forwardRef, HostListener, Input, Output } from '@angular/core';
import { TimePickerModalComponent } from "./time-picker-modal/time-picker-modal.component";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
/** @type {?} */
export const DEFAULT_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => TimePickerModalDirective)),
    multi: true
};
export class TimePickerModalDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
        this.valueChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    onFocus() {
        this.open();
        this.onChange();
    }
    /**
     * @return {?}
     */
    onChange() {
        this.timePickerModal.time = this.getDateInput();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.timePickerModal) {
            console.error('[TimePickerModalDirective] Missing timePickerModal');
        }
        this.timePickerModal.onChange.subscribe((/**
         * @param {?} timeStr
         * @return {?}
         */
        (timeStr) => {
            this.writeValue(timeStr);
            if (this.propagateChange) {
                this.propagateChange(timeStr);
            }
        }));
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.el.nativeElement.value = value;
        this.valueChange.emit(value);
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.propagateChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
    }
    /**
     * @return {?}
     */
    open() {
        /** @type {?} */
        let bounding = this.el.nativeElement.getBoundingClientRect();
        /** @type {?} */
        let css = {
            left: bounding.left
        };
        if (bounding.top + 150 < innerHeight) {
            // Show bottom
            css['top'] = bounding.top + bounding.height + 2;
        }
        else {
            // Show top
            css['bottom'] = bounding.top - 2;
        }
        this.timePickerModal.open(css);
    }
    /**
     * @return {?}
     */
    close() {
        this.timePickerModal.close();
    }
    /**
     * @private
     * @return {?}
     */
    getDateInput() {
        /** @type {?} */
        let value = this.getValue();
        if (!value) {
            return null;
        }
        /** @type {?} */
        let date = new Date();
        date.setMinutes(0);
        // HH:mm:ss format
        /** @type {?} */
        let times;
        if (value.indexOf(':') > -1) {
            times = value.split(':');
        }
        else {
            times = [];
            for (let i = 0; i < value.length + 1; i += 2) {
                times.push(value.substr(i, 2));
            }
        }
        times[0] = Number(times[0]);
        times[1] = Number(times[1]) || 0;
        // Invalid time
        if (!times[0] && times[0] !== 0) {
            return null;
        }
        date.setHours(times[0]);
        date.setMinutes(times[1]);
        return date;
    }
    /**
     * @private
     * @return {?}
     */
    getValue() {
        return this.el.nativeElement.value.trim();
    }
}
TimePickerModalDirective.decorators = [
    { type: Directive, args: [{
                selector: '[timePicker]',
                providers: [DEFAULT_VALUE_ACCESSOR]
            },] }
];
/** @nocollapse */
TimePickerModalDirective.ctorParameters = () => [
    { type: ElementRef }
];
TimePickerModalDirective.propDecorators = {
    valueChange: [{ type: Output }],
    timePickerModal: [{ type: Input }],
    onFocus: [{ type: HostListener, args: ['focus',] }],
    onChange: [{ type: HostListener, args: ['change',] }]
};
if (false) {
    /** @type {?} */
    TimePickerModalDirective.prototype.valueChange;
    /** @type {?} */
    TimePickerModalDirective.prototype.timePickerModal;
    /**
     * @type {?}
     * @private
     */
    TimePickerModalDirective.prototype.propagateChange;
    /**
     * @type {?}
     * @private
     */
    TimePickerModalDirective.prototype.el;
}
