// inet-core/esm5/src/websocket/websocket-client.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { WebSocketEnvelop, EnvelopBody, EnvelopMessage } from './websocket-abstract';
import { WebSocketJboss } from './websocket-jboss';
import { WebSocketVertx } from './websocket-vertx';
import { Subject } from 'rxjs';
var WebSocketClient = /** @class */ (function () {
    function WebSocketClient() {
        var _this = this;
        // message from socket server
        this.onMessage = new Subject();
        // socket state change
        this.onStateChange = new Subject();
        // envelop send will be timeout after 30s
        this.sendTimeout = 30000;
        this._autoconnectAfter = 5000; // 5s
        // 5s
        this._autoconnectRepeat = 4; // 5 times
        // 5 times
        this._autoconnectRetryAfter = 30000; // 30s
        this._envelopQueue = [];
        if (!WebSocketClient.instance) {
            this._getAccountInfo((/**
             * @return {?}
             */
            function () { return _this._register(); }));
        }
        return WebSocketClient.instance = WebSocketClient.instance || this;
    }
    Object.defineProperty(WebSocketClient.prototype, "readyState", {
        get: /**
         * @return {?}
         */
        function () {
            return this.socket ? this.socket.readyState : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSocketClient.prototype, "connecting", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === WebSocket.CONNECTING || this.readyState === null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSocketClient.prototype, "connected", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === WebSocket.OPEN;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSocketClient.prototype, "closing", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === WebSocket.CLOSING;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSocketClient.prototype, "closed", {
        get: /**
         * @return {?}
         */
        function () {
            return this.readyState === WebSocket.CLOSED;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSocketClient.prototype, "autoconnect", {
        get: /**
         * @return {?}
         */
        function () {
            return this._autoconnect;
        },
        enumerable: true,
        configurable: true
    });
    // close websocket client
    // close websocket client
    /**
     * @return {?}
     */
    WebSocketClient.prototype.close = 
    // close websocket client
    /**
     * @return {?}
     */
    function () {
        this.socket.close();
    };
    /**
     * @param {?} address
     * @param {?} message
     * @param {?=} application
     * @return {?}
     */
    WebSocketClient.prototype.send = /**
     * @param {?} address
     * @param {?} message
     * @param {?=} application
     * @return {?}
     */
    function (address, message, application) {
        if (application === void 0) { application = ''; }
        return this.sendEnvelop(this.buildEnvelop(address, message, application));
    };
    /**
     * @param {?} envelop
     * @return {?}
     */
    WebSocketClient.prototype.sendEnvelop = /**
     * @param {?} envelop
     * @return {?}
     */
    function (envelop) {
        var _this = this;
        if (this.connected) {
            return this.socket.send(envelop);
        }
        else {
            return new Promise((/**
             * @param {?} resolve
             * @return {?}
             */
            function (resolve) {
                // Add to queue and waiting socket connected to send
                /** @type {?} */
                var data = {
                    e: envelop,
                    f: resolve
                };
                data.timer = setTimeout((/**
                 * @return {?}
                 */
                function () { return _this._completeEnvelopQueue(data, false); }), _this.sendTimeout);
                _this._envelopQueue.push(data);
            }));
        }
    };
    /**
     * @param {?} address
     * @param {?} message
     * @param {?=} application
     * @param {?=} join
     * @return {?}
     */
    WebSocketClient.prototype.buildEnvelop = /**
     * @param {?} address
     * @param {?} message
     * @param {?=} application
     * @param {?=} join
     * @return {?}
     */
    function (address, message, application, join) {
        if (application === void 0) { application = ''; }
        /** @type {?} */
        var envelop = new WebSocketEnvelop();
        envelop.address = address;
        envelop.body = new EnvelopBody();
        envelop.body.application = application;
        envelop.body.sender = this._username;
        envelop.body.alias = [address];
        envelop.body.message = new EnvelopMessage(this._username, message, join || '');
        return envelop;
    };
    /**
     * @private
     * @return {?}
     */
    WebSocketClient.prototype._sendEnvelopQueue = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this._envelopQueue.forEach((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.sendEnvelop(data.e).then((/**
             * @param {?} status
             * @return {?}
             */
            function (status) { return _this._completeEnvelopQueue(data, status); }));
        }));
    };
    /**
     * @private
     * @param {?} data
     * @param {?} status
     * @return {?}
     */
    WebSocketClient.prototype._completeEnvelopQueue = /**
     * @private
     * @param {?} data
     * @param {?} status
     * @return {?}
     */
    function (data, status) {
        /** @type {?} */
        var index = this._envelopQueue.indexOf(data);
        if (index >= 0) {
            this._envelopQueue.splice(index, 1);
        }
        data.f(status);
    };
    /**
     * @private
     * @param {?=} callback
     * @return {?}
     */
    WebSocketClient.prototype._register = /**
     * @private
     * @param {?=} callback
     * @return {?}
     */
    function (callback) {
        var _this = this;
        $.ajax({
            url: iNet.getPUrl('account/register'),
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (typeof data === 'string') {
                    _this.socket = new WebSocketJboss(data, _this._username);
                }
                else {
                    _this.socket = new WebSocketVertx(data, _this._username);
                }
                _this.socket.onStateChange.subscribe((/**
                 * @return {?}
                 */
                function () { return _this._stateChange(); }));
                _this.socket.onMessage.subscribe((/**
                 * @param {?} body
                 * @return {?}
                 */
                function (body) {
                    _this.onMessage.next(body);
                }));
                callback && callback();
            }),
            error: (/**
             * @return {?}
             */
            function () {
                callback && callback();
                throw new Error("Account register error");
            })
        });
    };
    /**
     * @private
     * @param {?} callback
     * @return {?}
     */
    WebSocketClient.prototype._getAccountInfo = /**
     * @private
     * @param {?} callback
     * @return {?}
     */
    function (callback) {
        var _this = this;
        if (this._username) {
            return callback(this._username);
        }
        $.ajax({
            url: iNet.getPUrl('system/userprofile/view'),
            success: (/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this._username = data && data.user && (data.user.usercode || data.user.username) || '';
                callback(_this._username);
            }),
            error: (/**
             * @return {?}
             */
            function () { return callback(null); })
        });
    };
    /**
     * @private
     * @return {?}
     */
    WebSocketClient.prototype._stateChange = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.closed && !this.autoconnect) {
            this._runAutoConnect();
        }
        if (this._autoconnect) {
            return;
        }
        if (this.connected) {
            this._sendEnvelopQueue();
        }
        this._cancelAutoConnect();
        this.onStateChange.next();
    };
    /**
     * @private
     * @return {?}
     */
    WebSocketClient.prototype._runAutoConnect = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this._autoconnect = true;
        this._autoconectTimer = setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.connected) {
                _this._cancelAutoConnect();
                return;
            }
            if (_this._autoconectCount >= _this._autoconnectRepeat) {
                _this._cancelAutoConnect();
                // Sleep and retry connect after 30s
                setTimeout((/**
                 * @return {?}
                 */
                function () { return _this._runAutoConnect(); }), _this._autoconnectRetryAfter);
                return;
            }
            _this._register((/**
             * @return {?}
             */
            function () {
                _this._autoconectCount++;
                if (!_this.connected) {
                    _this._runAutoConnect();
                }
            }));
        }), this._autoconnectAfter);
    };
    /**
     * @private
     * @return {?}
     */
    WebSocketClient.prototype._cancelAutoConnect = /**
     * @private
     * @return {?}
     */
    function () {
        this._autoconnect = false;
        this._autoconectCount = 0;
        clearTimeout(this._autoconectTimer);
    };
    WebSocketClient.instance = null;
    return WebSocketClient;
}());
export { WebSocketClient };
if (false) {
    /** @type {?} */
    WebSocketClient.instance;
    /** @type {?} */
    WebSocketClient.prototype.socket;
    /** @type {?} */
    WebSocketClient.prototype.onMessage;
    /** @type {?} */
    WebSocketClient.prototype.onStateChange;
    /** @type {?} */
    WebSocketClient.prototype.sendTimeout;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._username;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._autoconnect;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._autoconnectAfter;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._autoconnectRepeat;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._autoconnectRetryAfter;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._autoconectCount;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._autoconectTimer;
    /**
     * @type {?}
     * @private
     */
    WebSocketClient.prototype._envelopQueue;
}
