// inet-core/esm5/src/websocket/websocket-abstract.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
/**
 * @abstract
 */
var /**
 * @abstract
 */
WebSocketAbstract = /** @class */ (function () {
    function WebSocketAbstract(options, username) {
        // receive message from socket server
        this.onMessage = new Subject();
        this.onStateChange = new Subject();
        this.connect(options, username);
    }
    // update readyState
    // update readyState
    /**
     * @return {?}
     */
    WebSocketAbstract.prototype.stateChange = 
    // update readyState
    /**
     * @return {?}
     */
    function () {
        this.onStateChange.next(this.readyState);
    };
    return WebSocketAbstract;
}());
/**
 * @abstract
 */
export { WebSocketAbstract };
if (false) {
    /** @type {?} */
    WebSocketAbstract.prototype.onMessage;
    /** @type {?} */
    WebSocketAbstract.prototype.onStateChange;
    /** @type {?} */
    WebSocketAbstract.prototype.readyState;
    /**
     * @abstract
     * @param {?} options
     * @param {?} username
     * @return {?}
     */
    WebSocketAbstract.prototype.connect = function (options, username) { };
    /**
     * @abstract
     * @return {?}
     */
    WebSocketAbstract.prototype.close = function () { };
    /**
     * @abstract
     * @param {?} envelop
     * @return {?}
     */
    WebSocketAbstract.prototype.send = function (envelop) { };
}
/** @enum {string} */
var WebSocketType = {
    // send to the first address
    SEND: 'send',
    // publish to the all address
    PUBLISH: 'publish',
};
export { WebSocketType };
var WebSocketEnvelop = /** @class */ (function () {
    function WebSocketEnvelop() {
        this.type = WebSocketType.PUBLISH;
    }
    return WebSocketEnvelop;
}());
export { WebSocketEnvelop };
if (false) {
    /** @type {?} */
    WebSocketEnvelop.prototype.type;
    /** @type {?} */
    WebSocketEnvelop.prototype.address;
    /** @type {?} */
    WebSocketEnvelop.prototype.body;
}
var EnvelopBody = /** @class */ (function () {
    function EnvelopBody() {
    }
    return EnvelopBody;
}());
export { EnvelopBody };
if (false) {
    /** @type {?} */
    EnvelopBody.prototype.uuid;
    /** @type {?} */
    EnvelopBody.prototype.application;
    /** @type {?} */
    EnvelopBody.prototype.message;
    /** @type {?} */
    EnvelopBody.prototype.sender;
    /** @type {?} */
    EnvelopBody.prototype.alias;
    /** @type {?} */
    EnvelopBody.prototype.sent;
}
var EnvelopMessage = /** @class */ (function () {
    // , joins?: string, duration?: string
    function EnvelopMessage(sender, content, joins, duration) {
        this.content = this.alert = content;
        this.sender = sender;
        if (!!joins) {
            this.joins = joins || sender;
        }
        this.duration = duration || "0";
    }
    return EnvelopMessage;
}());
export { EnvelopMessage };
if (false) {
    /** @type {?} */
    EnvelopMessage.prototype.sender;
    /** @type {?} */
    EnvelopMessage.prototype.alert;
    /** @type {?} */
    EnvelopMessage.prototype.content;
    /** @type {?} */
    EnvelopMessage.prototype.joins;
    /** @type {?} */
    EnvelopMessage.prototype.duration;
}
