// inet-ui/esm2015/src/image-cropper/cropper-avatar/cropper-avatar-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output, TemplateRef, ViewChild } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
export class CropperAvatarDialogComponent {
    /**
     * @param {?} modalService
     */
    constructor(modalService) {
        this.modalService = modalService;
        this.cropperReady = false;
        this.onCropped = new EventEmitter();
        this.onHide = new EventEmitter();
        this.onSelectFile = new EventEmitter();
        this.isModalShown = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.modalService.onShow.subscribe((/**
         * @return {?}
         */
        () => {
            this.isModalShown = true;
        }));
        this.modalService.onHide.subscribe((/**
         * @return {?}
         */
        () => {
            this.isModalShown = false;
        }));
    }
    /**
     * @param {?} event
     * @return {?}
     */
    imageCropped(event) {
        this.imageCroppedEvent = event;
    }
    /**
     * @return {?}
     */
    imageLoaded() {
        this.cropperReady = true;
    }
    /**
     * @return {?}
     */
    loadImageFailed() {
        //console.log('Load failed');
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    setImageChangedEvent($event) {
        this.imageChangedEvent = $event;
    }
    /**
     * @return {?}
     */
    hide() {
        this.modalRef.hide();
        this.setImageChangedEvent({});
        this.onHide.emit(this.imageCroppedEvent.base64);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    show($event) {
        this.setImageChangedEvent($event);
        if (!this.isModalShown) {
            this.modalRef = this.modalService.show(this.cropperModal);
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    select($event) {
        this.onSelectFile.emit($event);
    }
    /**
     * @return {?}
     */
    submit() {
        this.onCropped.emit(this.imageCroppedEvent.base64);
        this.hide();
    }
    /**
     * @return {?}
     */
    getImageBlob() {
        return this.imageCroppedEvent.file;
    }
}
CropperAvatarDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-cropper-avatar-dialog',
                template: "<ng-template #cropperAvatarModal>\n    <div class=\"modal-header\">\n        <h5 class=\"modal-title pull-left\"><i class=\"fa fa-crop\"></i> {{'COMMON.MODULE.CROPPER.TITLE' | translate }}</h5>\n        <button type=\"button\" class=\"close pull-right\" (click)=\"hide()\">\n            <span aria-hidden=\"true\">&times;</span>\n        </button>\n    </div>\n    <div class=\"modal-body p-0\">\n        <image-cropper [imageChangedEvent]=\"imageChangedEvent\"\n                       [maintainAspectRatio]=\"true\"\n                       [aspectRatio]=\"1\" [resizeToWidth]=\"150\"\n                       [roundCropper]=\"true\" format=\"png\" outputType=\"both\"\n                       (imageCropped)=\"imageCropped($event)\"\n                       (imageLoaded)=\"imageLoaded()\"\n                       (loadImageFailed)=\"loadImageFailed()\"\n                       [style.display]=\"cropperReady ? null : 'none'\"></image-cropper>\n    </div>\n    <div class=\"modal-footer text-right\">\n        <button type=\"button\" class=\"btn btn-primary btn-sm\" (click)=\"select($event)\">\n            <i class=\"fa fa-picture-o\"></i> {{ 'COMMON.MODULE.CROPPER.SELECT_NEW_IMAGE'  | translate}}\n        </button>\n\n        <button type=\"button\" class=\"btn btn-success btn-sm ml-1\" (click)=\"submit()\">\n            <i class=\"fa fa-crop\"></i> {{ 'TOOLBAR.OK' | translate }}\n        </button>\n        <button type=\"button\" class=\"btn btn-danger btn-sm ml-1\" (click)=\"hide()\">\n            <i class=\"fa fa-times\"></i> {{ 'TOOLBAR.CANCEL' | translate }}\n        </button>\n    </div>\n</ng-template>"
            }] }
];
/** @nocollapse */
CropperAvatarDialogComponent.ctorParameters = () => [
    { type: BsModalService }
];
CropperAvatarDialogComponent.propDecorators = {
    cropperModal: [{ type: ViewChild, args: ['cropperAvatarModal',] }],
    onCropped: [{ type: Output }],
    onHide: [{ type: Output }],
    onSelectFile: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CropperAvatarDialogComponent.prototype.cropperModal;
    /**
     * @type {?}
     * @private
     */
    CropperAvatarDialogComponent.prototype.modalRef;
    /** @type {?} */
    CropperAvatarDialogComponent.prototype.cropperReady;
    /** @type {?} */
    CropperAvatarDialogComponent.prototype.imageChangedEvent;
    /** @type {?} */
    CropperAvatarDialogComponent.prototype.imageCroppedEvent;
    /** @type {?} */
    CropperAvatarDialogComponent.prototype.onCropped;
    /** @type {?} */
    CropperAvatarDialogComponent.prototype.onHide;
    /** @type {?} */
    CropperAvatarDialogComponent.prototype.onSelectFile;
    /**
     * @type {?}
     * @private
     */
    CropperAvatarDialogComponent.prototype.isModalShown;
    /**
     * @type {?}
     * @private
     */
    CropperAvatarDialogComponent.prototype.modalService;
}
