// inet-ui/esm2015/src/layout/bootstrap/message-side-nav/message-side-nav.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef } from '@angular/core';
import { NotifyMessageService } from "../../../common/notify-message.service";
import { AbstractSideNavComponent } from "../abstract-side-nav.component";
export class MessageSideNavComponent extends AbstractSideNavComponent {
    /**
     * @param {?} service
     * @param {?} elementRef
     */
    constructor(service, elementRef) {
        super(elementRef);
        this.service = service;
        this.elementRef = elementRef;
        this.loaded = false;
        this.messages = [];
        this.pageNumber = 0;
        this.pageSize = 10;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const currentValue = changes['opened']["currentValue"];
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (currentValue) {
                this.open();
                if (!this.loaded) {
                    this.pageNumber = 0;
                    this.listMessage({ pageSize: this.pageSize, pageNumber: this.pageNumber, notifylist: true });
                }
            }
            else {
                this.close();
            }
        }), 1);
    }
    /**
     * @param {?} params
     * @return {?}
     */
    listMessage(params) {
        this.service.getMessages(params).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        (response) => {
            this.messages = (/** @type {?} */ (response.items));
            this.totalMessage = response.total;
            this.loaded = true;
        }));
    }
    /**
     * @return {?}
     */
    loadMessageMore() {
        this.pageNumber += 1;
        this.service.getMessages({ pageSize: this.pageSize, pageNumber: this.pageNumber, notifylist: true }).subscribe((/**
         * @param {?} response
         * @return {?}
         */
        (response) => {
            this.messages = [...this.messages, ...(/** @type {?} */ (response.items))];
        }));
    }
    /**
     * @return {?}
     */
    clearMessage() {
        this.service.clearAll().subscribe((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            this.messages = [];
            this.totalMessage = 0;
            this.onClear.emit(true);
        }));
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    openMessage(msg) {
        if (!msg.read) {
            this.service.loadNotify(msg.application, msg.activityID).subscribe((/**
             * @param {?} message
             * @return {?}
             */
            (message) => {
                msg.read = true;
                this.onLoad.emit(true);
            }));
        }
    }
}
MessageSideNavComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-admin-message-side-nav',
                template: "<div class=\"slide-nav\" [ngClass]=\"{'opened': opened}\">\n    <div class=\"list-header\">\n        <span class=\"font-weight-bold\"><i class=\"fa fa-bell\"></i> {{'COMMON.NOTIFICATION.TITLE' | translate}}</span>\n        <button (click)=\"close()\" class=\"close\" type=\"button\"><i class=\"fa fa-times\"></i></button>\n    </div>\n    <ul class=\"list-box\">\n        <li *ngFor=\"let item of messages\">\n            <a href=\"javascript:;\" (click)=\"openMessage(item)\">\n                <div class=\"list-item\">\n                    <div class=\"list-left\">\n                        <img userAvatar [usercode]=\"item.sender\" class=\"user-avatar mt-0 mr-2 rounded-circle\">\n                    </div>\n                    <div class=\"list-body\">\n                        <div class=\"media-body ml-2\">\n                            <p class=\"mb-1 message-item p-2\" [ngClass]=\"{'unread': !item.read}\">\n                                <b>{{item.fullname}}</b> {{item.message}}\n                            </p>\n                            <!--div class=\"px-2\">{{item.created | date}}</div-->\n                        </div>\n                    </div>\n                </div>\n            </a>\n        </li>\n        <li *ngIf=\"!messages?.length\">\n            <div class=\"empty-msg-container\">\n                <i>{{'COMMON.NOTIFICATION.EMPTY_MSG' | translate}}</i>\n            </div>\n        </li>\n        <div *ngIf=\"messages?.length && totalMessage!=messages?.length\" class=\"mt-1 text-center\">\n            <a (click)=\"loadMessageMore()\" href=\"javascript:;\">\n                {{'COMMON.NOTIFICATION.LOAD_MORE' | translate}} {{totalMessage-messages?.length}} {{'COMMON.NOTIFICATION.ITEMS' | translate}}\n            </a>\n        </div>\n    </ul>\n    <div class=\"list-footer\">\n        <a *ngIf=\"messages?.length\" (click)=\"clearMessage()\" href=\"javascript:;\"><i class=\"fa fa-eraser text-danger\"></i> {{'COMMON.NOTIFICATION.CLEAR_ALL' | translate}}</a>\n    </div>\n</div>\n",
                styles: [":host a{text-decoration:none}:host .list-body{padding:0 0 0 2rem}:host .list-left{height:2rem;width:2rem;position:absolute}:host .user-avatar{height:2rem;width:2rem}:host .message-item{background-color:#edf2f9;border-radius:.375rem!important}:host .unread{border:1px dotted #999}:host .pr-2,:host .px-2{padding-right:.5rem!important}:host .empty-msg-container{text-align:center;line-height:30px;padding-top:15%;color:#999}"]
            }] }
];
/** @nocollapse */
MessageSideNavComponent.ctorParameters = () => [
    { type: NotifyMessageService },
    { type: ElementRef }
];
if (false) {
    /** @type {?} */
    MessageSideNavComponent.prototype.keyword;
    /** @type {?} */
    MessageSideNavComponent.prototype.loaded;
    /** @type {?} */
    MessageSideNavComponent.prototype.messages;
    /** @type {?} */
    MessageSideNavComponent.prototype.totalMessage;
    /** @type {?} */
    MessageSideNavComponent.prototype.pageNumber;
    /** @type {?} */
    MessageSideNavComponent.prototype.pageSize;
    /**
     * @type {?}
     * @private
     */
    MessageSideNavComponent.prototype.service;
    /** @type {?} */
    MessageSideNavComponent.prototype.elementRef;
}
