// inet-ui/esm2015/src/custom-pagination/custom-pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
export class CustomPaginationComponent {
    constructor() {
        this.pageSize = 10;
        this.pageNumber = 1;
        this.totalItems = 0;
        this.firstLoad = true;
        this.params = {};
        this.pageCount = 0;
        this.onLoad = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.processOnTotalChanged();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.firstLoad) {
            this.load();
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changeToFirstPage(event) {
        this.pageNumber = 1;
        this.onPageChanged(event, this.pageNumber);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changeToLastPage(event) {
        this.pageNumber = this.pageCount;
        this.onPageChanged(event, this.pageNumber);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changeToNextPage(event) {
        if (this.pageNumber < this.pageCount) {
            this.pageNumber++;
            this.onPageChanged(event, this.pageNumber);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changeToPreviousPage(event) {
        if (this.pageNumber > 1) {
            this.pageNumber--;
            this.onPageChanged(event, this.pageNumber);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changeToCurrentPage(event) {
        this.pageNumber = event.currentTarget.value;
        if (this.pageNumber < 1) {
            this.pageNumber = 1;
        }
        else if (this.pageNumber > this.pageCount) {
            this.pageNumber = this.pageCount;
        }
        event.currentTarget.value = this.pageNumber;
        this.onPageChanged(event, this.pageNumber);
    }
    /**
     * @return {?}
     */
    get maxPage() {
        return Math.ceil(this.totalItems / this.pageSize);
    }
    /**
     * @return {?}
     */
    getQuery() {
        /** @type {?} */
        let query = {
            pageNumber: (this.pageNumber < 1) ? 0 : this.pageNumber - 1,
            pageSize: this.pageSize
        };
        query = Object.assign({}, this.getParams(), query);
        return query;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    setTotalItems(v) {
        this.totalItems = v;
        this.processOnTotalChanged();
    }
    /**
     * @param {?=} params
     * @return {?}
     */
    setParams(params = {}) {
        this.params = params;
    }
    /**
     * @return {?}
     */
    getParams() {
        return this.params || {};
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    reload(event) {
        this.onLoad.emit(this.getQuery());
    }
    /**
     * @return {?}
     */
    load() {
        /** @type {?} */
        let query = { newPage: 1, pageSize: this.pageSize };
        query = Object.assign({}, this.getParams(), query);
        this.processOnPaginationChanged(null, query);
    }
    /**
     * @param {?=} event
     * @param {?=} pageNumber
     * @return {?}
     */
    onPageChanged(event, pageNumber) {
        this.processOnPaginationChanged(event, { newPage: pageNumber, pageSize: -1 });
    }
    /**
     * @param {?} event
     * @param {?} args
     * @return {?}
     */
    processOnPaginationChanged(event, args) {
        this.pageNumber = args.newPage;
        this.onLoad.emit(this.getQuery());
        return 'onPaginationChanged';
    }
    /**
     * @return {?}
     */
    processOnTotalChanged() {
        this.pageCount = this.maxPage || 0;
        this.pageNumber = this.pageCount > 0 ? 1 : 0;
        this.query = this.getQuery();
    }
}
CustomPaginationComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-custom-pagination',
                template: "<div class=\"pagination-box\">\n    <div class=\"pagination-range\">\n        <ng-template [ngTemplateOutlet]=\"basicSearch\"></ng-template>\n    </div>\n    <div class=\"pagination-controllers\">\n        <div class=\"pagination-pages mr-1\" >\n            <button [disabled]=\"pageNumber === 1 || totalItems === 0\" class=\"btn btn-sm btn-default pagination-firstpage\"\n                    [title]=\"'grid.firstText' | translate\"\n                    (click)=\"changeToFirstPage($event)\">\n                <i class=\"fa fa-step-backward\"></i>\n            </button>\n            <button [disabled]=\"pageNumber === 1 || totalItems === 0\" class=\"btn btn-sm btn-default pagination-prevpage ml-1\"\n                    [title]=\"'grid.prevText' | translate\"\n                    (click)=\"changeToPreviousPage($event)\">\n                <i class=\"fa fa-chevron-left\"></i>\n            </button>\n            <div class=\"pagination-page ml-1\" >\n                <div class=\"input-group input-group-sm\">\n                    <input #pageInput value=\"{{pageNumber}}\" [disabled]=\"totalItems === 0\" (change)=\"changeToCurrentPage($event)\" type=\"number\" class=\"form-control\" min=\"0\" step=\"1\"/>\n                    <div class=\"input-group-append\">\n                        <span class=\"input-group-text\"><span>/</span><span>{{pageCount}}</span></span>\n                    </div>\n                </div>\n            </div>\n            <button [disabled]=\"pageNumber === pageCount || totalItems === 0\" class=\"btn btn-sm btn-default pagination-nextpage ml-1\"\n                    [title]=\"'grid.nextText' | translate\"\n                    (click)=\"changeToNextPage($event)\">\n                <i class=\"fa fa-chevron-right\"></i>\n            </button>\n            <button [disabled]=\"pageNumber === pageCount || totalItems === 0\"  (click)=\"changeToLastPage($event)\"\n                    [title]=\"'grid.lastText' | translate\"\n                    class=\"btn btn-sm btn-default pagination-lastpage ml-1\">\n                <i class=\"fa fa-step-forward\"></i>\n            </button>\n            <button   [title]=\"'grid.refreshText' | translate\"  (click)=\"reload($event)\" type=\"button\" class=\"btn btn-sm btn-default pagination-reload ml-1\">\n                <i class=\"fa fa-refresh\"></i>\n            </button>\n        </div>\n    </div>\n    <ng-template [ngTemplateOutlet]=\"advanceSearch\"></ng-template>\n</div>\n",
                styles: [".pagination-box{position:relative;display:block;min-height:37px;background-color:#e9ecef;border:1px solid #ced4da;padding:2px}.pagination-range{margin-left:3px;display:inline-block;max-width:calc(100vw - 35%)}.pagination-controllers{float:right}.pagination-controllers input{min-width:60px;text-align:right;max-width:80px}.pagination-limit{margin-right:10px;display:inline-table;width:180px;float:left}.pagination-pages{display:inline-block}.pagination-page{display:inline-table}.pagination-box button{outline:0!important}.column-selector-button,.pagination-firstpage,.pagination-lastpage,.pagination-nextpage,.pagination-prevpage,.pagination-reload{vertical-align:top}"]
            }] }
];
/** @nocollapse */
CustomPaginationComponent.ctorParameters = () => [];
CustomPaginationComponent.propDecorators = {
    pageSize: [{ type: Input, args: ["pageSize",] }],
    pageNumber: [{ type: Input, args: ["pageNumber",] }],
    totalItems: [{ type: Input, args: ["totalItems",] }],
    firstLoad: [{ type: Input, args: ["firstLoad",] }],
    params: [{ type: Input, args: ["params",] }],
    basicSearch: [{ type: ContentChild, args: ['basicSearch',] }],
    advanceSearch: [{ type: ContentChild, args: ['advanceSearch',] }],
    onLoad: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CustomPaginationComponent.prototype.pageSize;
    /** @type {?} */
    CustomPaginationComponent.prototype.pageNumber;
    /** @type {?} */
    CustomPaginationComponent.prototype.totalItems;
    /** @type {?} */
    CustomPaginationComponent.prototype.firstLoad;
    /** @type {?} */
    CustomPaginationComponent.prototype.params;
    /** @type {?} */
    CustomPaginationComponent.prototype.basicSearch;
    /** @type {?} */
    CustomPaginationComponent.prototype.advanceSearch;
    /** @type {?} */
    CustomPaginationComponent.prototype.query;
    /** @type {?} */
    CustomPaginationComponent.prototype.pageCount;
    /** @type {?} */
    CustomPaginationComponent.prototype.onLoad;
}
