// inet-ui/esm5/src/translate/cloud-translate.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from 'inet-core';
import { TranslateService } from "@ngx-translate/core";
var CloudTranslateService = /** @class */ (function () {
    function CloudTranslateService(http, translate) {
        this.http = http;
        this.translate = translate;
        this.url = {
            list: iNet.getPUrl('message/keys')
        };
    }
    /**
     * @return {?}
     */
    CloudTranslateService.prototype.getAllMessage = /**
     * @return {?}
     */
    function () {
        return this.http.getJSON(this.url.list);
    };
    /**
     * @param {?} appName
     * @return {?}
     */
    CloudTranslateService.prototype.getMessageByApp = /**
     * @param {?} appName
     * @return {?}
     */
    function (appName) {
        return this.http.getJSON(this.url.list, { group: appName });
    };
    /**
     * @return {?}
     */
    CloudTranslateService.prototype.getCurrentLang = /**
     * @return {?}
     */
    function () {
        if (window.localStorage) {
            return window.localStorage.getItem(CloudTranslateService.LANGUAGE_KEY) || this.translate.currentLang;
        }
        return this.translate.currentLang;
    };
    /**
     * @param {?} lang
     * @return {?}
     */
    CloudTranslateService.prototype.setCurrentLang = /**
     * @param {?} lang
     * @return {?}
     */
    function (lang) {
        if (window.localStorage) {
            localStorage.setItem(CloudTranslateService.LANGUAGE_KEY, lang);
        }
        this.translate.setDefaultLang(lang);
    };
    CloudTranslateService.LANGUAGE_KEY = 'language';
    CloudTranslateService.GRID_KEY = 'grid';
    CloudTranslateService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CloudTranslateService.ctorParameters = function () { return [
        { type: HttpClientService },
        { type: TranslateService }
    ]; };
    return CloudTranslateService;
}());
export { CloudTranslateService };
if (false) {
    /** @type {?} */
    CloudTranslateService.LANGUAGE_KEY;
    /** @type {?} */
    CloudTranslateService.GRID_KEY;
    /**
     * @type {?}
     * @private
     */
    CloudTranslateService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    CloudTranslateService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    CloudTranslateService.prototype.translate;
}
