// inet-ui/esm5/src/translate/cloud-translate.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BsLocaleService } from "ngx-bootstrap/datepicker";
import { TranslateLoader, TranslateModule, TranslateService } from "@ngx-translate/core";
import { listLocales } from "ngx-bootstrap";
import { CloudTransLoader } from "./cloud-trans-loader";
import { CoreService, HttpClientService } from "inet-core";
import { CloudTranslateService } from "./cloud-translate.service";
var CloudTranslateModule = /** @class */ (function () {
    function CloudTranslateModule(translate, localeService, coreService, cloudTranslate) {
        this.translate = translate;
        this.localeService = localeService;
        this.coreService = coreService;
        this.cloudTranslate = cloudTranslate;
        /** @type {?} */
        var languages = listLocales();
        translate.addLangs(languages);
        /** @type {?} */
        var lang = cloudTranslate.getCurrentLang() || languages[0];
        translate.use(lang);
        //Change a locale
        this.localeService.use(lang);
        translate.get([CloudTranslateService.GRID_KEY]).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            if (iNet.ui.grid.Grid && res) {
                iNet.applyIf(iNet.ui.grid.Grid.prototype, res[CloudTranslateService.GRID_KEY]);
            }
        }));
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    CloudTranslateModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = {}; }
        return TranslateModule.forRoot({
            loader: {
                provide: TranslateLoader,
                useClass: config.loader || CloudTransLoader,
                deps: [HttpClientService]
            }
        });
    };
    CloudTranslateModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        TranslateModule.forRoot({
                            loader: { provide: TranslateLoader, useClass: CloudTransLoader }
                        })
                    ],
                    declarations: [],
                    exports: [TranslateModule],
                    providers: [BsLocaleService, TranslateService, CloudTranslateService, HttpClientService]
                },] }
    ];
    /** @nocollapse */
    CloudTranslateModule.ctorParameters = function () { return [
        { type: TranslateService },
        { type: BsLocaleService },
        { type: CoreService },
        { type: CloudTranslateService }
    ]; };
    return CloudTranslateModule;
}());
export { CloudTranslateModule };
if (false) {
    /** @type {?} */
    CloudTranslateModule.prototype.translate;
    /** @type {?} */
    CloudTranslateModule.prototype.localeService;
    /** @type {?} */
    CloudTranslateModule.prototype.coreService;
    /**
     * @type {?}
     * @private
     */
    CloudTranslateModule.prototype.cloudTranslate;
}
