// inet-ui/esm5/src/translate/cloud-trans-loader.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Observable } from 'rxjs';
import 'rxjs/add/observable/of';
import { defineLocale } from 'ngx-bootstrap/chronos';
import { systemViLocale } from "./system-vi";
import { systemEnLocale } from "./system-en";
var CloudTransLoader = /** @class */ (function () {
    function CloudTransLoader() {
        this.store = {};
        this.defineLocale(systemViLocale.abbr, systemViLocale);
        this.defineLocale(systemEnLocale.abbr, systemEnLocale);
    }
    /**
     * @param {?} lang
     * @param {?} data
     * @return {?}
     */
    CloudTransLoader.prototype.assign = /**
     * @param {?} lang
     * @param {?} data
     * @return {?}
     */
    function (lang, data) {
        return Object.assign(this.getResourceByLang(lang), data);
    };
    /**
     * @param {?} lang
     * @param {?} data
     * @return {?}
     */
    CloudTransLoader.prototype.defineLocale = /**
     * @param {?} lang
     * @param {?} data
     * @return {?}
     */
    function (lang, data) {
        defineLocale(lang, data);
        this.store[lang] = data;
    };
    /**
     * @param {?} lang
     * @return {?}
     */
    CloudTransLoader.prototype.getTranslation = /**
     * @param {?} lang
     * @return {?}
     */
    function (lang) {
        return Observable.of(this.getResourceByLang(lang));
    };
    /**
     * @param {?} lang
     * @return {?}
     */
    CloudTransLoader.prototype.getResourceByLang = /**
     * @param {?} lang
     * @return {?}
     */
    function (lang) {
        return this.store[lang] || systemEnLocale;
    };
    /**
     * Convert complex js object to dot notation js object
     */
    /**
     * Convert complex js object to dot notation js object
     * @private
     * @param {?} message
     * @param {?} value
     * @param {?} obj
     * @return {?}
     */
    CloudTransLoader.prototype.dotizeStringToJSON = /**
     * Convert complex js object to dot notation js object
     * @private
     * @param {?} message
     * @param {?} value
     * @param {?} obj
     * @return {?}
     */
    function (message, value, obj) {
        /** @type {?} */
        var parts = message.split(".");
        /** @type {?} */
        var last = parts.pop();
        /** @type {?} */
        var part;
        while (part = parts.shift()) {
            if (typeof obj[part] !== "object") {
                obj[part] = {};
            }
            obj = obj[part];
        }
        obj[last] = value;
    };
    /**
     * Convert Java .properties files data to JSON
     */
    /**
     * Convert Java .properties files data to JSON
     * @private
     * @param {?} str
     * @return {?}
     */
    CloudTransLoader.prototype.propertiesToJSON = /**
     * Convert Java .properties files data to JSON
     * @private
     * @param {?} str
     * @return {?}
     */
    function (str) {
        var _this = this;
        // Concat lines that end with '\'.
        // Split by line breaks.
        // Remove commented lines
        // Create the JSON
        return str.replace(/\\\n/, "").split("\n")
            .filter((/**
         * @param {?} line
         * @return {?}
         */
        function (line) { return /(\#|\!)/.test(line.replace(/\s/g, "").slice(0, 1)) ? false : line; }))
            .reduce((/**
         * @param {?} obj
         * @param {?} line
         * @return {?}
         */
        function (obj, line) {
            /** @type {?} */
            var colonifiedLine = line.replace(/(\=)/, ":");
            /** @type {?} */
            var key = colonifiedLine.substring(0, colonifiedLine.indexOf(":")).trim();
            /** @type {?} */
            var value = _this.asciiToNative(colonifiedLine.substring(colonifiedLine.indexOf(":") + 1).trim());
            obj[key] = value;
            return obj;
        }), {});
    };
    /*
    * Convert ASCII value to Native Character
     */
    /*
        * Convert ASCII value to Native Character
         */
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    CloudTransLoader.prototype.asciiToNative = /*
        * Convert ASCII value to Native Character
         */
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var character = value.split("\\u");
        /** @type {?} */
        var nativeStr = character[0];
        for (var i = 1; i < character.length; i++) {
            /** @type {?} */
            var code = character[i];
            nativeStr += String.fromCharCode(parseInt("0x" + code.substring(0, 4)));
            if (code.length > 4) {
                nativeStr += code.substring(4, code.length);
            }
        }
        return nativeStr;
    };
    /*
    * Convert Native value to ASCII
     */
    /*
        * Convert Native value to ASCII
         */
    /**
     * @private
     * @param {?} value
     * @param {?=} ignoreLatin
     * @return {?}
     */
    CloudTransLoader.prototype.native2Ascii = /*
        * Convert Native value to ASCII
         */
    /**
     * @private
     * @param {?} value
     * @param {?=} ignoreLatin
     * @return {?}
     */
    function (value, ignoreLatin) {
        if (ignoreLatin === void 0) { ignoreLatin = false; }
        /** @type {?} */
        var character = value.split("");
        /** @type {?} */
        var asciiStr = "";
        for (var i = 0; i < character.length; i++) {
            /** @type {?} */
            var code = Number(character[i].charCodeAt(0));
            if (!ignoreLatin || code > 127) {
                /** @type {?} */
                var charAscii = code.toString(16);
                charAscii = new String("0000").substring(charAscii.length, 4) + charAscii;
                asciiStr += "\\u" + charAscii;
            }
            else {
                asciiStr += character[i];
            }
        }
        return asciiStr;
    };
    /**
     * Convert items from the server to Translation
     */
    /**
     * Convert items from the server to Translation
     * @param {?} items
     * @return {?}
     */
    CloudTransLoader.prototype.convertResourceToObject = /**
     * Convert items from the server to Translation
     * @param {?} items
     * @return {?}
     */
    function (items) {
        var e_1, _a;
        /** @type {?} */
        var translation = {};
        try {
            for (var items_1 = tslib_1.__values(items), items_1_1 = items_1.next(); !items_1_1.done; items_1_1 = items_1.next()) {
                var item = items_1_1.value;
                this.dotizeStringToJSON(item['messageKey'], item['text'], translation);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (items_1_1 && !items_1_1.done && (_a = items_1.return)) _a.call(items_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return translation;
    };
    /**
     * Convert Java .properties files data to Translation
     */
    /**
     * Convert Java .properties files data to Translation
     * @param {?} str
     * @return {?}
     */
    CloudTransLoader.prototype.convertPropertiesToObject = /**
     * Convert Java .properties files data to Translation
     * @param {?} str
     * @return {?}
     */
    function (str) {
        /** @type {?} */
        var translation = {};
        /** @type {?} */
        var json = this.propertiesToJSON(str);
        for (var key in json) {
            this.dotizeStringToJSON(key, json[key], translation);
        }
        return translation;
    };
    /**
     * @param {?} lang
     * @param {?} items
     * @return {?}
     */
    CloudTransLoader.prototype.setResources = /**
     * @param {?} lang
     * @param {?} items
     * @return {?}
     */
    function (lang, items) {
        this.assign(lang, this.convertResourceToObject(items));
    };
    return CloudTransLoader;
}());
export { CloudTransLoader };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CloudTransLoader.prototype.store;
}
