// inet-ui/esm5/src/translate/system-vi.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var systemViLocale = {
    abbr: 'vi',
    vi: 'Tiếng Việt',
    en: 'Tiếng Anh',
    months: 'Tháng 1_Tháng 2_Tháng 3_Tháng 4_Tháng 5_Tháng 6_Tháng 7_Tháng 8_Tháng 9_Tháng 10_Tháng 11_Tháng 12'.split('_'),
    monthsShort: 'TH1_TH2_TH3_TH4_TH5_TH6_TH7_TH8_TH9_TH10_TH11_TH12'.split('_'),
    weekdays: 'Chủ nhật_Thứ 2_Thứ 3_Thứ 4_Thứ 5_Thứ 6_Thứ 7'.split('_'),
    weekdaysShort: 'CN_T2_T3_T4_T5_T6_T7'.split('_'),
    weekdaysMin: 'CN_T2_T3_T4_T5_T6_T7'.split('_'),
    longDateFormat: {
        LT: 'HH:mm',
        LTS: 'HH:mm:ss',
        L: 'DD/MM/YYYY',
        LL: 'D MMMM YYYY',
        LLL: 'D MMMM YYYY HH:mm',
        LLLL: 'dddd, D MMMM YYYY HH:mm'
    },
    calendar: {
        sameDay: '[Hôm nay lúc] LT',
        nextDay: '[Ngày mai lúc] LT',
        nextWeek: 'dddd [lúc] LT',
        lastDay: '[Hôm qua lúc] LT',
        lastWeek: '[Tuần trước] dddd [lúc] LT',
        sameElse: 'L'
    },
    relativeTime: {
        future: 'trong %s',
        past: '%s trước',
        s: '1 vài giây',
        ss: '%d giây',
        m: '1 phút',
        mm: '%d phút',
        h: '1 giờ',
        hh: '%d giờ',
        d: '1 ngày',
        dd: '%d ngày',
        M: '1 tháng',
        MM: '%d tháng',
        y: '1 năm',
        yy: '%d năm'
    },
    dayOfMonthOrdinalParse: /\d{1,2}(st|nd|rd|th)/,
    ordinal: (/**
     * @param {?} _num
     * @return {?}
     */
    function (_num) {
        /** @type {?} */
        var num = Number(_num);
        /** @type {?} */
        var b = num % 10;
        /** @type {?} */
        var output = (~~(num % 100 / 10) === 1) ? 'th' :
            (b === 1) ? 'st' :
                (b === 2) ? 'nd' :
                    (b === 3) ? 'rd' : 'th';
        return num + output;
    }),
    week: {
        dow: 1,
        doy: 4 // The week that contains Jan 4th is the first week of the year.
    },
    grid: {
        displayMsg: ' Hiển thị <b>{0}</b> - <b>{1}</b> trong tổng số <b>{2}</b>',
        emptyMsg: 'Không có dữ liệu để hiển thị',
        beforePageText: 'Trang',
        afterPageText: 'của {0}',
        firstText: 'Trang đầu',
        prevText: 'Trang trước',
        nextText: 'Trang sau',
        lastText: 'Trang cuối',
        refreshText: 'Làm tươi',
        searchText: 'Tìm kiếm',
        searchPlaceholder: 'Từ khóa',
        downloadText: 'Tải về',
        translations: {
            headerReload: 'Nạp lại',
            headerColumnSelector: 'Chọn cột hiển thị',
            indexColumn: 'Thứ tự hiển thị',
            selectColumn: 'Cột chọn',
            selectRow: 'Chọn dòng',
            selectAllRows: 'Chọn',
            expandColumn: 'Mở rộng',
            expandRow: 'Xem nhiều hơn',
            paginationLimit: 'Phần tử',
            paginationRange: 'Hiển thị',
            firstText: 'Trang đầu',
            prevText: 'Trang trước',
            nextText: 'Trang sau',
            lastText: 'Trang cuối'
        }
    },
    TOOLBAR: {
        ADD: 'Thêm mới',
        EDIT: 'Sửa',
        SAVE: 'Lưu',
        UPDATE: 'Cập nhật',
        DELETE: 'Xóa',
        SEARCH: 'Tìm kiếm',
        OK: 'Đồng ý',
        CANCEL: 'Bỏ qua',
        RESET: 'Nhập lại',
        BACK: 'Quay lại',
        CLOSE: 'Đóng',
        DOWNLOAD: 'Tải về'
    },
    COMMON: {
        WELCOME: 'Xin chào',
        PLACEHOLDER_SEARCH: 'Tìm kiếm ...',
        NOTIFICATION: {
            TITLE: 'Thông báo',
            EMPTY_MSG: 'Không có dữ liệu để hiển thị',
            LOAD_MORE: 'Tải thêm',
            ITEMS: 'thông báo',
            CLEAR_ALL: 'Xóa tất cả thông báo'
        },
        APPLICATION: {
            TITLE: 'Ứng dụng',
            EVENTS_TODAY: 'Sự kiện hôm nay',
            NEW_ACTIVITY: 'Tương tác mới'
        },
        MENU: {
            LANGUAGE: 'Ngôn ngữ',
            HOME: 'Trang chủ',
            USER_PROFILE: 'Thông tin cá nhân',
            ORGANIZATION_INFORMATION: 'Thông tin công ty',
            CONFIG: 'Cấu hình',
            SIGNATURE: 'Chữ ký số',
            ABOUT: 'Thông tin hệ thống',
            BACK: 'Quay lại',
            LOGOUT: 'Đăng xuất'
        },
        SELECTION: {
            DEFAULT_VALUE: '- Chọn -'
        },
        MESSAGE: {
            CONFIRM_DELETE_TITLE: 'Xóa ?',
            CONFIRM_DELETE_MSG: 'Bạn có chắc là đồng ý muốn xóa dữ liệu đã chọn không ?'
        },
        MODULE: {
            ADDRESS: {
                CITY: 'Tỉnh/Thành phố',
                DISTRICT: 'Quận/Huyện',
                COUNTRY: 'Quốc gia',
                ADDRESS: 'Địa chỉ',
                STREET: 'Địa chỉ',
                ADDRESS_REQUIRED_MSG: 'Địa chỉ không được rỗng',
                COUNTRY_REQUIRED_MSG: 'Quốc gia không được rỗng'
            },
            ORGANIZATION: {
                TITLE: 'Công ty',
                CODE: 'Mã công ty',
                NAME: 'Tên công ty',
                FOREIGN_NAME: 'Tên quốc tế',
                ABBREVIATION_NAME: 'Tên viết tắt',
                TAX_CODE: 'Mã số thuế',
                BILLING_ADDRESS: 'Địa chỉ trên hóa đơn',
                LEGAL_REPRESENTATIVE: 'Đại diện pháp luật',
                WEBSITE: 'Website',
                INDUSTRY: 'Ngành nghề',
                COMPANY_SIZE: 'Quy mô công ty',
                DESCRIPTION: 'Mô tả',
                LINK: 'Địa chỉ truy cập',
                CODE_REQUIRED_MSG: 'Mã công ty không được rỗng',
                NAME_REQUIRED_MSG: 'Tên công ty không được rỗng',
                EMAIL: 'Thư điện tử',
                EMAIL_INVALID_MSG: 'Địa chỉ thư điện tử không hợp lệ',
                PHONE: 'Điện thoại',
                FAX: 'Fax',
                LANGUAGE: 'Ngôn ngữ',
                CHOSEN_IMAGE: 'Chọn Logo ...',
                STANDARD_IMAGE_SIZE: 'Kích thước chuẩn',
                UPDATE_LOGO_ERROR_MSG: 'Có lỗi xảy ra khi cập nhật Logo',
                EMPLOYEES: 'nhân viên',
                SELF_EMPLOYED: 'Tự làm chủ',
                REMOVE_LOGO: 'Xóa Logo',
                SAVE_LOGO: 'Lưu Logo',
                REMOVE_LOGO_SUCCESSFUL_MSG: 'Logo đã được xóa',
                REMOVE_LOGO_ERROR_MSG: 'Có lỗi xảy ra khi xóa Logo',
                UPDATE_SUCCESSFUL_MSG: 'Thông tin đã được cập nhật',
                UPDATE_ERROR_MSG: 'Có lỗi xảy ra khi cập nhật thông tin',
                TAXCODE_REQUIRED_MSG: 'Mã số thuế không được rỗng',
            },
            USER_PROFILE: {
                TITLE: 'Thông tin tài khoản',
                USERNAME: 'Tài khoản',
                LAST_NAME: 'Họ',
                MIDDLE_NAME: 'Tên đệm',
                FIRST_NAME: 'Tên',
                BIRTHDAY: 'Ngày sinh',
                PHONE: 'Điện thoại',
                MOBILE: 'Di động',
                FULL_NAME: 'Họ và tên',
                EMAIL: 'Thư điện tử',
                GROUP_NAME: 'Tên nhóm quyền',
                DESCRIPTION: 'Mô tả nhóm quyền',
                GENDER: 'Giới tính',
                MALE: 'Nam',
                FEMALE: 'Nữ',
                OTHER: 'Khác',
                FIRST_NAME_REQUIRED_MSG: 'Tên không được rỗng',
                LAST_NAME_REQUIRED_MSG: 'Họ không được rỗng',
                EMAIL_REQUIRED_MSG: 'Địa chỉ thư điện tử không được rỗng',
                EMAIL_INVALID_MSG: 'Địa chỉ thư điện tử không hợp lệ',
                GENERAL_INFORMATION: 'Thông tin tài khoản',
                CHOSEN_IMAGE: 'Chọn ảnh ...',
                SECURITY_INFORMATION: 'Thông tin bảo mật',
                CURRENT_PASSWORD: 'Mật khẩu',
                NEW_PASSWORD: 'Mật khẩu mới',
                CONFIRM_NEW_PASSWORD: ' Xác nhận mật khẩu',
                CHANGE_PASSWORD: 'Đổi mật khẩu',
                CURRENT_PASSWORD_REQUIRED_MSG: 'Vui lòng nhập mật khẩu hiện tại',
                NEW_PASSWORD_REQUIRED_MSG: 'Vui lòng nhập mật khẩu mới',
                CONFIRM_NEW_PASSWORD_REQUIRED_MSG: 'Vui lòng nhập xác nhận mật khẩu mới',
                PASSWORD_DO_NOT_MATCH: 'Xác nhận mật khẩu mới chưa đúng',
                CHANGE_PASSWORD_INFO_MSG: 'Để đổi mật khẩu, vui lòng nhập mật khẩu hiện tại của bạn, sau đó nhập mật khẩu mới của bạn hai lần',
                CHANGE_PASSWORD_SUCCESSFUL_MSG: 'Mật khẩu đã được thay đổi',
                CHANGE_PASSWORD_ERROR_MSG: 'Có lỗi xảy ra khi đổi mật khẩu',
                CURRENT_PASSWORD_INCORRECT: 'Mật khẩu hiện tại không đúng. Vui lòng nhập lại',
                UPDATE_PROFILE_SUCCESSFUL_MSG: 'Thông tin tài khoản đã được thay đổi',
                UPDATE_PROFILE_ERROR_MSG: 'Có lỗi xảy ra khi cập nhật thông tin tài khoản',
                AVATAR: 'Ảnh đại diện',
                SAVE_IMAGE: 'Lưu ảnh',
                REMOVE_IMAGE: 'Xóa ảnh',
                IMAGE_UPDATE_SUCCESSFUL_MSG: 'Ảnh đại diện đã được cập nhật',
                IMAGE_UPDATE_ERROR_MSG: 'Có lỗi xảy ra khi cập nhật ảnh đại diện',
                IMAGE_REMOVE_SUCCESSFUL_MSG: 'Ảnh đại diện đã được xóa',
                IMAGE_REMOVE_ERROR_MSG: 'Có lỗi xảy ra khi xóa ảnh đại diện',
                SIGNATURE_INFORMATION: 'Chữ ký số',
                ACCESS_ROLE: 'Quyền truy cập',
                SIGNATURE_VERIFY: 'Số điện thoại',
                SIGNATURE_FILE: 'Tệp ảnh chữ ký',
                SIGNATURE_VERIFY_REQUIRED_MSG: 'Số điện thoại không được để rỗng',
                UPDATE_SIGNATURE_SUCCESSFUL_MSG: 'Thông tin chữ ký số đã được cập nhật',
                UPDATE_SIGNATURE_ERROR_MSG: 'Có lỗi xảy ra khi cập nhật thông tin chữ ký số',
                SIGNATURE_FILE_REQUIRED_MSG: 'Vui lòng chọn ảnh chữ ký để cập nhật',
                UPDATE_SIGNATURE: 'Cập nhật chữ ký số',
                NO_FILE: 'Không có tệp tin ...',
                CHOOSE: 'Chọn tệp'
            },
            VIEWER: {
                UNSUPPORTED: 'Rất tiếc, định dạng không được hỗ trợ xem trực tuyến <br /> trên trình duyệt này'
            },
            CROPPER: {
                TITLE: 'Công cụ cắt ảnh',
                SELECT_NEW_IMAGE: 'Chọn ảnh khác ...'
            },
            SIGNATURE_MANAGE: {
                CREATE_SIGN: 'Thêm chữ ký số',
                TITLE: 'Chữ ký',
                NAME: 'Tên',
                NAME_REQUIRED_MSG: 'Tên chữ ký không được rỗng',
                MOBILE: 'Di động',
                ISSUED: 'Đơn vị phát hành',
                FILE: 'Tệp ảnh',
                FILE_REQUIRED_MSG: 'Chưa chọn ảnh',
                PHOTO: 'Ảnh',
                RESULTS: 'Hiển thị',
                DELETE_SUCCESSFUL_MSG: 'Chữ ký đã được xóa',
                DELETE_ERROR_MSG: 'Có lỗi xảy ra khi xóa chữ ký',
                SAVE_SUCCESSFUL_MSG: 'Chữ ký đã được tạo',
                SAVE_ERROR_MSG: 'Có lỗi xảy ra khi tạo chữ ký',
                NO_INFORMATION: 'Không tìm thấy kết quả phù hợp',
                KEY_SEARCH: 'Hãy chắc rằng từ khóa bạn cần tìm kiếm là chính xác.'
            },
            EMAIL_TEMPLATE: {
                "SEND_TEST": "Gửi test",
                "EMAIL": "Mẫu email",
                "LIST": "Danh sách mẫu email",
                "ADD_EMAIL": "Thêm mới mẫu email",
                "EDIT_EMAIL": "Chỉnh sửa mẫu email",
                "EMAIL_CODE": "Mã",
                "EMAIL_NAME": "Tên",
                "EMAIL_SUBJECT": "Tiêu đề",
                "EMAIL_CONTENT": "Nội dung email",
                "NOTE": "Ghi chú",
                "SAMPLE_INFORMATION": "Thông tin mẫu email",
                "SYSTEM_INFORMATION": "Thông tin tham số cấu hình",
                "TO": "Từ",
                "FROM": "Đến",
                "EMAIL_CODE_EXIST_MSG": "Mã đã tồn tại",
                "EMAIL_CODE_REQUIRED": "Vui lòng nhập mã",
                "EMAIL_NAME_REQUIRED": "Vui lòng nhập tên",
                "ADD_SUCCESS": "Mẫu email đã được thêm mới",
                "ADD_ERROR": "Có lỗi xảy ra trong quá trình thêm mới",
                "UPDATE_SUCCESS": "Mẫu email đã được cập nhật",
                "UPDATE_ERROR": "Có lỗi xảy ra trong quá trình cập nhật dữ liệu",
                "DELETE_SUCCESS": "Mẫu email đã được xóa.",
                "DELETE_ERROR": "Có lỗi xảy ra trong quá trình xóa dữ liệu",
                "DELETE_TITLE": "Xóa mẫu email?",
                "DELETE_CONTENT": "Bạn có chắc là muốn xóa mẫu email này không ?",
                "ACTIVE": "Sử dụng"
            },
            REPORT_TEMPLATE: {
                "TITLE": "Biểu mẫu",
                "LIST": 'Danh sách biểu mẫu',
                "ADD": "Thêm mới biểu mẫu",
                "APPLICATION": "Ứng dụng",
                "MODULE": "Mô-đun",
                "TYPE": "Loại",
                "VERSION": "Phiên bản",
                "NAME": "Tên",
                "DESCRIPTION": "Mô tả",
                "FILE_UPLOAD": "Tệp tin",
                "APPLICATION_REQUIRED": "Tên ứng dụng không được rỗng",
                "MODULE_REQUIRED": "Tên mô-đun không được rỗng",
                "TYPE_REQUIRED": "Tên loại không được rỗng",
                "VERSION_REQUIRED": "Phiên bản không được rỗng",
                "NAME_REQUIRED": "Tên biểu mẫu không được rỗng",
                "DESCRIPTION_REQUIRED": "Mô tả không được rỗng",
                "ADD_SUCCESS": "Biểu mẫu đã được thêm mới.",
                "ADD_ERROR": "Không thể thêm mới. Có lỗi xảy ra trong quá trình thêm mới biểu mẫu",
                "UPDATE_SUCCESS": "Biểu mẫu đã được chỉnh sửa",
                "UPDATE_ERROR": "Không thể chỉnh sửa. Có lỗi xảy ra trong quá trình chỉnh sửa biểu mẫu",
                "DELETE_SUCCESS": "Biểu mẫu đã được xóa",
                "DELETE_ERROR": "Không thể xóa. Có lỗi xảy ra trong quá trình xóa biểu mẫu",
                "DELETE_TITLE": "Xóa biểu mẫu?",
                "DELETE_CONTENT": "Bạn có chắc là muốn xóa biểu mẫu này không?",
                "NO_FILE": "Không có tệp tin ...",
                "CHOOSE": "Chọn tệp"
            }
        }
    },
    INDICATOR: {
        LOADING: 'Đang tải ...',
        PROCESSING: 'Đang xử lý ...',
        SAVING: 'Đang lưu ...',
        UPDATING: 'Đang cập nhật ...'
    },
    "ALL_SELECTED": "Tất cả đã được chọn",
    "CANCEL": "Bỏ qua",
    "CLEAR_ALL_FILTERS": "Xóa tất cả điều kiện lọc",
    "CLEAR_ALL_SORTING": "Xóa tất cả việc sắp xếp",
    "COLUMNS": "Cột",
    "COMMANDS": "Chức năng",
    "CONTAINS": "Có chứa",
    "ENDS_WITH": "Kết thúc với",
    "EQUALS": "Bằng",
    "EXPORT_TO_CSV": "Kết xuất định dạng CSV",
    "EXPORT_TO_TAB_DELIMITED": "Kết xuất định dạng Văn bản (phân cách bằng tab)",
    "FROM_TO_OF_TOTAL_ITEMS": "{{from}}-{{to}} của tổng số {{totalItems}} dòng",
    "FORCE_FIT_COLUMNS": "Hiển thị mặc định",
    "INVALID_FLOAT": "Dữ liệu số phải hợp lệ và có số thập phân tối đa là {{maxDecimal}}.",
    "GROUP_BY": "Nhóm theo",
    "HIDE_COLUMN": "Ẩn cột",
    "IN_COLLECTION_SEPERATED_BY_COMMA": "Tìm kiếm dữ liệu trong tập dữ liệu, các dữ liệu phải được phân tách bằng dấu phẩy (a,b)",
    "ITEMS": "Dòng",
    "ITEMS_PER_PAGE": "dòng trên mỗi trang",
    "NOT_IN_COLLECTION_SEPERATED_BY_COMMA": "Tìm kiếm dữ liệu không nằm trong tập dữ liệu, các dữ liệu phải được phân tách bằng dấu phẩy (a,b)",
    "OF": "của",
    "PAGE": "Trang",
    "PAGE_X_OF_Y": "trang {{x}} của {{y}}",
    "REFRESH_DATASET": "Làm mới dữ liệu",
    "REMOVE_FILTER": "Hủy lọc dữ liệu",
    "REMOVE_SORT": "Hủy sắp xếp dữ liệu",
    "SAVE": "Lưu dữ liệu",
    "SELECT_ALL": "Chọn tất cả",
    "SORT_ASCENDING": "Sắp xếp tăng dần",
    "SORT_DESCENDING": "Sắp xếp giảm dần",
    "STARTS_WITH": "Bắt đầu với",
    "SYNCHRONOUS_RESIZE": "Thay đổi kích thước đồng bộ",
    "TOGGLE_FILTER_ROW": "Ẩn/hiện dòng lọc dữ liệu",
    "TOGGLE_PRE_HEADER_ROW": "Chuyển đổi dòng trước tiêu đề",
    "X_OF_Y_SELECTED": "# của % đã chọn",
    "REPORT_INV_LIST": "Bảng kê hóa đơn",
    "REPORT_INV_USING_SITUATION": "Báo cáo tình hình sử dụng hóa đơn",
    "ADJUSTMENT": "Điều chỉnh hóa đơn",
    "REPLACEMENT": "Thay thế hóa đơn",
    "DESTROYED": "Xóa bỏ hóa đơn"
};
