// inet-calendar/esm5/src/dialog-repeat/dialog-repeat.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Output, ViewChild } from '@angular/core';
import { CalendarService } from '../calendar.service';
import { CalendarUtils } from "../CalendarUtils";
var CalendarDialogRepeatComponent = /** @class */ (function () {
    function CalendarDialogRepeatComponent(calendarService) {
        this.calendarService = calendarService;
        this.onRepeat = new EventEmitter();
        this.data = {};
        this.dateFormat = 'DD/MM/YYYY';
        this.repeatTypes = CalendarUtils.repeatTypes;
        this.weekDays = CalendarUtils.weekDays;
    }
    /**
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.$modal = $(this.repeatModal.nativeElement);
        document.body.appendChild(this.repeatModal.nativeElement);
    };
    /**
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.$modal.is(':visible')) {
            this.$modal.modal('hide').on('hidden.bs.modal', (/**
             * @return {?}
             */
            function () {
                $(this).remove();
            }));
        }
        else {
            this.$modal.remove();
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.openRepeat = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.data = data || {};
        this.initData();
        this.$modal.modal('show');
    };
    /**
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.saveRepeat = /**
     * @return {?}
     */
    function () {
        this.onRepeat.emit(this.getRepeatData());
    };
    /**
     * @param {?} item
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.changeWeekDay = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        item.checked = !item.checked;
        if (this.getWeekDayChecked().length < 1) {
            item.checked = true;
        }
    };
    /**
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.changeDate = /**
     * @return {?}
     */
    function () {
        if (!this.start || !this.start.getTime()) {
            this.start = new Date();
        }
        if (!this.until || !this.until.getTime() || this.start >= this.until) {
            this.until = new Date(this.start.getFullYear(), this.start.getMonth(), this.start.getDate());
            this.until.setMonth(this.until.getMonth() + 1);
        }
    };
    /**
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.getRepeatData = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var params = {
            rrmode: this.rrmode,
            lstart: this.start.getTime(),
            until: this.until.getTime(),
            wkdays: []
        };
        if (params.rrmode == 'WEEKLY') {
            // this.weekDays.forEach(item => {
            //   if (item.checked) {
            //     params.wkdays += item.value + ',';
            //   }
            // });
            // params.wkdays = params.wkdays.slice(0, -1);
            params.wkdays = this.getWeekDayChecked().map((/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return item.value; }));
        }
        return params;
    };
    /**
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.getWeekDayChecked = /**
     * @return {?}
     */
    function () {
        return this.weekDays.filter((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            return item.checked;
        }));
    };
    /**
     * @private
     * @return {?}
     */
    CalendarDialogRepeatComponent.prototype.initData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.rrmode = this.data.rrmode || 'DAILY';
        this.start = new Date(this.data.lstart);
        this.until = new Date(this.data.until);
        this.changeDate();
        if (this.data.wkdays) {
            this.weekDays.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (_this.data.wkdays.indexOf(item.value) > -1) {
                    item.checked = true;
                }
            }));
        }
        else {
            this.weekDays[(this.start.getDay() + 6) % 7].checked = true;
        }
    };
    CalendarDialogRepeatComponent.decorators = [
        { type: Component, args: [{
                    selector: 'calendar-repeat-modal',
                    template: "<div #repeatModal class=\"modal fade\" tabindex=\"-1\" role=\"dialog\">\n    <div class=\"modal-dialog\" role=\"document\">\n        <div class=\"modal-content\">\n            <div class=\"modal-body\">\n                <div class=\"form-group\">\n                    <label>L\u1EB7p l\u1ECBch</label>\n                    <select [(ngModel)]=\"rrmode\" class=\"form-control\">\n                        <option *ngFor=\"let item of repeatTypes\" [value]=\"item.value\">{{ item.label }}</option>\n                    </select>\n                </div>\n                <div *ngIf=\"rrmode === 'WEEKLY'\" class=\"form-group\" style=\"margin-bottom: 0;\">\n                    <div class=\"check-box-i\">\n                        <div *ngFor=\"let day of weekDays\"\n                             [title]=\"day.title\"\n                             (click)=\"changeWeekDay(day)\"\n                             [ngClass]=\"{'check-box-active':day.checked}\"\n                             class=\"check-box-item\">{{ day.label }}</div>\n                    </div>\n                </div>\n                <div class=\"form-group\">\n                    <label>B\u1EAFt \u0111\u1EA7u</label>\n                    <div class=\"input-group\">\n                        <input\n                            class=\"form-control\"\n                            #repeatStart=\"bsDatepicker\"\n                            bsDatepicker\n                            [bsConfig]=\"{ dateInputFormat: dateFormat }\"\n                            (ngModelChange)=\"changeDate()\"\n                            [(ngModel)]=\"start\">\n                        <div (click)=\"repeatStart.toggle()\" class=\"input-group-append\">\n                            <i class=\"fa fa-calendar input-group-text\"></i>\n                        </div>\n                    </div>\n                </div>\n                <div class=\"form-group\">\n                    <label>K\u1EBFt th\u00FAc</label>\n                    <div class=\"input-group\">\n                        <input\n                            class=\"form-control\"\n                            #repeatUntil=\"bsDatepicker\"\n                            bsDatepicker\n                            [bsConfig]=\"{ dateInputFormat: dateFormat }\"\n                            (ngModelChange)=\"changeDate()\"\n                            [(ngModel)]=\"until\">\n                        <div (click)=\"repeatUntil.toggle()\" class=\"input-group-append\">\n                            <i class=\"fa fa-calendar input-group-text\"></i>\n                        </div>\n                    </div>\n                </div>\n                <div class=\"form-action\">\n                    <button type=\"button\" class=\"btn btn-secondary btn-sm\" data-dismiss=\"modal\">\n                        \u0110\u00F3ng\n                    </button>\n                    <button type=\"button\" class=\"ml-1 btn btn-primary btn-sm\" data-dismiss=\"modal\" (click)=\"saveRepeat()\">\n                        Xong\n                    </button>\n                </div>\n            </div>\n        </div>\n    </div>\n</div>",
                    styles: [".modal-dialog{width:400px}.form-action{text-align:right;margin:15px 0}.check-box-i,.check-box-item{overflow:hidden}.check-box-item{float:left;width:30px;height:30px;line-height:30px;border-radius:50%;text-align:center;margin:0 10px 20px 0;font-size:12px;background:rgba(0,0,0,.1);cursor:pointer}.check-box-active{background:#438eb9;color:#fff}.check-box-disable{opacity:.5}"]
                }] }
    ];
    /** @nocollapse */
    CalendarDialogRepeatComponent.ctorParameters = function () { return [
        { type: CalendarService }
    ]; };
    CalendarDialogRepeatComponent.propDecorators = {
        onRepeat: [{ type: Output }],
        repeatModal: [{ type: ViewChild, args: ['repeatModal',] }],
        repeatStart: [{ type: ViewChild, args: ['repeatStart',] }],
        repeatUntil: [{ type: ViewChild, args: ['repeatUntil',] }]
    };
    return CalendarDialogRepeatComponent;
}());
export { CalendarDialogRepeatComponent };
if (false) {
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.onRepeat;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.repeatModal;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.repeatStart;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.repeatUntil;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.start;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.until;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.rrmode;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.repeatTypes;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.weekDays;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.data;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.$modal;
    /** @type {?} */
    CalendarDialogRepeatComponent.prototype.dateFormat;
    /**
     * @type {?}
     * @private
     */
    CalendarDialogRepeatComponent.prototype.calendarService;
}
