// inet-social/esm2015/src/attachment-list/attachment-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { SocialService } from '../social.service';
export class AttachmentListComponent {
    /**
     * @param {?} socialService
     */
    constructor(socialService) {
        this.socialService = socialService;
        this.attachments = [];
        this.files = [];
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.attachments) {
            this._indexFiles();
        }
    }
    /**
     * @param {?} file
     * @param {?} imageEl
     * @return {?}
     */
    viewImage(file, imageEl) {
        /** @type {?} */
        const options = {
            index: this._getImageIndex(file.id),
            clickElement: imageEl
        };
        this.socialService.viewImages(this._images, options);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    viewAttachment(data) {
        /** @type {?} */
        const file = data.file;
        /** @type {?} */
        const attachment = this._getAttachmentById(file.id);
        if (attachment) {
            if (file.image) {
                /** @type {?} */
                let imageEl = data.event.target;
                if (imageEl.tagName !== 'IMG') {
                    // click on image's parent
                    imageEl = imageEl.getElementsByTagName('img')[0];
                }
                this.viewImage(file, imageEl);
            }
            else {
                this.socialService.viewAttachment(attachment);
            }
        }
    }
    /**
     * @private
     * @param {?} id
     * @return {?}
     */
    _getAttachmentById(id) {
        for (let i = 0; i < this.attachments.length; i++) {
            if (this.attachments[i].gridfsUUID === id) {
                return this.attachments[i];
            }
        }
    }
    /**
     * @private
     * @param {?} id
     * @return {?}
     */
    _getImageIndex(id) {
        for (let i = 0; i < this._images.length; i++) {
            if (this._images[i].id === id) {
                return i;
            }
        }
        return 0;
    }
    /**
     * @private
     * @return {?}
     */
    _indexFiles() {
        this.files = this.socialService.attachmentToFiles(this.attachments);
        // Index images to view later
        this._images = [];
        for (let i = 0; i < this.files.length; i++) {
            if (this.files[i].image) {
                /** @type {?} */
                const attachment = this.attachments[i];
                this._images.push({
                    src: attachment.pathView,
                    id: attachment.gridfsUUID,
                    w: 1,
                    h: 1,
                    sizeDynamic: true
                });
            }
        }
    }
}
AttachmentListComponent.decorators = [
    { type: Component, args: [{
                selector: '[attachmentList]',
                template: "<file-list [files]=\"files\" [removable]=\"false\" (onClick)=\"viewAttachment($event)\"></file-list>",
                styles: [""]
            }] }
];
/** @nocollapse */
AttachmentListComponent.ctorParameters = () => [
    { type: SocialService }
];
AttachmentListComponent.propDecorators = {
    attachments: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AttachmentListComponent.prototype.attachments;
    /** @type {?} */
    AttachmentListComponent.prototype.files;
    /**
     * @type {?}
     * @private
     */
    AttachmentListComponent.prototype._images;
    /**
     * @type {?}
     * @private
     */
    AttachmentListComponent.prototype.socialService;
}
