// inet-ui/esm5/src/number-format/number-separator.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, HostListener, forwardRef, Attribute, ElementRef, EventEmitter, Output } from '@angular/core';
import { NumberUtilsService } from './number-utils.service';
import { NG_VALIDATORS } from '@angular/forms';
var NumberSeparatorDirective = /** @class */ (function () {
    function NumberSeparatorDirective(utilsService, elementRef, name) {
        this.utilsService = utilsService;
        this.elementRef = elementRef;
        this.name = name;
        this.separator = ',';
        this.ngModel = new EventEmitter();
        this.el = this.elementRef.nativeElement;
        if (this.el) {
            this.el.type = 'text';
        }
    }
    /**
     * @return {?}
     */
    NumberSeparatorDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} c
     * @return {?}
     */
    NumberSeparatorDirective.prototype.validate = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        // self value
        /** @type {?} */
        var v = c.value;
        if (v == 0 && !this.numberMin && !this.numberMax) {
            this.el.value = "0";
            return null;
        }
        if (v) {
            // console.log('validate[value]', v, this.el.value, v !== this.el.value);
            // first update
            if (this.el && (v !== this.el.value)) {
                this.el.value = this.addSeparator(this.el.value);
            }
            // control value
            v = parseInt(v.toString().replace(/[^0-9.,]/g, ''), 10);
            // number validator
            if (!this.isValidParam(v)) {
                this.el.value = null;
                return {
                    pattern: true
                };
            }
            // min validator
            if (v && this.isValidParam(this.numberMax) && v > parseInt(this.numberMax, 10)) {
                return {
                    max: false
                };
            }
            // max validator
            if (v && this.isValidParam(this.numberMin) && v < parseInt(this.numberMin, 10)) {
                return {
                    min: false
                };
            }
        }
        return null;
    };
    /**
     * @private
     * @param {?} param
     * @return {?}
     */
    NumberSeparatorDirective.prototype.isValidParam = /**
     * @private
     * @param {?} param
     * @return {?}
     */
    function (param) {
        return param && this.utilsService.isNumeric(param);
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    NumberSeparatorDirective.prototype.getNumber = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        value = value.replace(/[^0-9.,]/g, '');
        return value;
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    NumberSeparatorDirective.prototype.checkLength = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.isValidParam(this.numberLength) && value.length > parseInt(this.numberLength, 10)) {
            value = value.slice(0, parseInt(this.numberLength, 10));
        }
        return value;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    NumberSeparatorDirective.prototype.onInput = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        // console.log('onInput[data]', data.target.value);
        /** @type {?} */
        var value = this.utilsService.removeSeparator($event.target.value, this.separator);
        value = this.getNumber(value);
        if (this.utilsService.isNumeric(value)) {
            value = this.checkLength(value);
            $event.target.value = this.utilsService.addSeparator(value, this.separator);
        }
    };
    /*

    @HostListener("change", ["$event.target.value"])
    onChange(value) {
        this.el.value = this.addSeparator(value);
    }

    onInputChange(value) {
        const control = this.injector.get(NgControl);
        const newValue = this.utilsService.removeSeparator(value, this.separator);
        console.log('[newValue]', newValue);
        control.valueAccessor.writeValue(newValue);
    }

    */
    /*
    
        @HostListener("change", ["$event.target.value"])
        onChange(value) {
            this.el.value = this.addSeparator(value);
        }
    
        onInputChange(value) {
            const control = this.injector.get(NgControl);
            const newValue = this.utilsService.removeSeparator(value, this.separator);
            console.log('[newValue]', newValue);
            control.valueAccessor.writeValue(newValue);
        }
    
        */
    /**
     * @private
     * @param {?} v
     * @return {?}
     */
    NumberSeparatorDirective.prototype.addSeparator = /*
    
        @HostListener("change", ["$event.target.value"])
        onChange(value) {
            this.el.value = this.addSeparator(value);
        }
    
        onInputChange(value) {
            const control = this.injector.get(NgControl);
            const newValue = this.utilsService.removeSeparator(value, this.separator);
            console.log('[newValue]', newValue);
            control.valueAccessor.writeValue(newValue);
        }
    
        */
    /**
     * @private
     * @param {?} v
     * @return {?}
     */
    function (v) {
        /** @type {?} */
        var value = this.utilsService.removeSeparator(v, this.separator);
        value = this.getNumber(value);
        if (this.utilsService.isNumeric(value)) {
            value = this.checkLength(value);
            return this.utilsService.addSeparator(value, this.separator);
        }
        return v;
    };
    NumberSeparatorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[appNumberSeparator]',
                    providers: [
                        { provide: NG_VALIDATORS, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return NumberSeparatorDirective; })), multi: true }
                    ]
                    /*
                    host: {
                        '(ngModelChange)' : 'onInputChange($event)'
                    }
                    */
                },] }
    ];
    /** @nocollapse */
    NumberSeparatorDirective.ctorParameters = function () { return [
        { type: NumberUtilsService },
        { type: ElementRef },
        { type: String, decorators: [{ type: Attribute, args: ['name',] }] }
    ]; };
    NumberSeparatorDirective.propDecorators = {
        separator: [{ type: Input, args: ['separator',] }],
        numberMin: [{ type: Input, args: ['numberMin',] }],
        numberMax: [{ type: Input, args: ['numberMax',] }],
        numberLength: [{ type: Input, args: ['numberLength',] }],
        ngModel: [{ type: Output }],
        onInput: [{ type: HostListener, args: ['input', ['$event'],] }]
    };
    return NumberSeparatorDirective;
}());
export { NumberSeparatorDirective };
if (false) {
    /** @type {?} */
    NumberSeparatorDirective.prototype.separator;
    /** @type {?} */
    NumberSeparatorDirective.prototype.numberMin;
    /** @type {?} */
    NumberSeparatorDirective.prototype.numberMax;
    /** @type {?} */
    NumberSeparatorDirective.prototype.numberLength;
    /**
     * @type {?}
     * @private
     */
    NumberSeparatorDirective.prototype.el;
    /** @type {?} */
    NumberSeparatorDirective.prototype.ngModel;
    /**
     * @type {?}
     * @private
     */
    NumberSeparatorDirective.prototype.utilsService;
    /**
     * @type {?}
     * @private
     */
    NumberSeparatorDirective.prototype.elementRef;
    /** @type {?} */
    NumberSeparatorDirective.prototype.name;
}
