// inet-ui/esm5/src/photoswipe/photoswipe.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApplicationRef, ComponentFactoryResolver, Injectable, Injector } from '@angular/core';
import { PhotoSwipeComponent } from './photoswipe.component';
import { Router } from '@angular/router';
var PhotoSwipe = /** @class */ (function () {
    function PhotoSwipe(componentFactoryResolver, injector, appRef, router) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.injector = injector;
        this.appRef = appRef;
        this.router = router;
    }
    /**
     * @param {?} items
     * @param {?=} options
     * @param {?=} container
     * @return {?}
     */
    PhotoSwipe.prototype.open = /**
     * @param {?} items
     * @param {?=} options
     * @param {?=} container
     * @return {?}
     */
    function (items, options, container) {
        var _this = this;
        if (!items || items.length < 1) {
            return;
        }
        /** @type {?} */
        var containerEl = container && container.nativeElement || document.body;
        if (!this.photoSwipeFactory) {
            this.photoSwipeFactory = this.componentFactoryResolver.resolveComponentFactory(PhotoSwipeComponent);
        }
        this.photoSwipeComponentRef = this.photoSwipeFactory.create(this.injector);
        // Attach component to the appRef so that it's inside the ng component tree
        this.appRef.attachView(this.photoSwipeComponentRef.hostView);
        // Append DOM element
        containerEl.appendChild(this.photoSwipeComponentRef.hostView['rootNodes'][0]);
        this.photoSwipeComponentRef.instance.open(items, options);
        this.photoSwipeComponentRef.instance.onDestroy.subscribe((/**
         * @return {?}
         */
        function () {
            // Remove
            _this.appRef.detachView(_this.photoSwipeComponentRef.hostView);
            _this.photoSwipeComponentRef.destroy();
            _this.photoSwipeComponentRef = null;
            _this.routerSubscribe.unsubscribe();
            _this.routerSubscribe = null;
        }));
        this.subscribeRouterChange();
    };
    /**
     * @private
     * @return {?}
     */
    PhotoSwipe.prototype.subscribeRouterChange = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.routerSubscribe = this.router.events.subscribe((/**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            _this.closeOpening();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    PhotoSwipe.prototype.closeOpening = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.photoSwipeComponentRef) {
            this.photoSwipeComponentRef.instance.destroy();
        }
    };
    PhotoSwipe.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PhotoSwipe.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: ApplicationRef },
        { type: Router }
    ]; };
    return PhotoSwipe;
}());
export { PhotoSwipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PhotoSwipe.prototype.photoSwipeFactory;
    /**
     * @type {?}
     * @private
     */
    PhotoSwipe.prototype.photoSwipeComponentRef;
    /**
     * @type {?}
     * @private
     */
    PhotoSwipe.prototype.routerSubscribe;
    /**
     * @type {?}
     * @private
     */
    PhotoSwipe.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    PhotoSwipe.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    PhotoSwipe.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    PhotoSwipe.prototype.router;
}
