// inet-ui/esm2015/src/report-template/report-template.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClientService } from "inet-core";
import * as i0 from "@angular/core";
import * as i1 from "inet-core";
export class ReportTemplateService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.url = {
            view: iNet.getPUrl('gl/exceltmpl/view'),
            delete: iNet.getPUrl('gl/exceltmpl/delete'),
            update: iNet.getPUrl('gl/exceltmpl/update'),
            add: iNet.getPUrl('gl/exceltmpl/create'),
            list: iNet.getPUrl('gl/exceltmpl/list'),
            application: iNet.getPUrl('cloud/subfirm/application'),
            organ: iNet.getPUrl('plugin/firmprofile/view'),
            modules: iNet.getPUrl('gl/exceltmpl/modules'),
            download: iNet.getPUrl('gl/filecontents/download')
        };
    }
    /**
     * @param {?} params
     * @return {?}
     */
    list(params) {
        this.http.showLoading();
        return this.http.postJSON(this.url.list, params);
    }
    /**
     * @param {?} uuid
     * @return {?}
     */
    delete(uuid) {
        this.http.showLoading();
        return this.http.getJSON(this.url.delete, { uuid: uuid });
    }
    /**
     * @param {?} uuid
     * @return {?}
     */
    view(uuid) {
        this.http.showLoading();
        return this.http.postJSON(this.url.view, { uuid: uuid });
    }
    /**
     * @param {?} params
     * @return {?}
     */
    add(params) {
        this.http.showLoading();
        return this.http.post(this.url.add, params);
    }
    /**
     * @param {?} params
     * @return {?}
     */
    update(params) {
        this.http.showLoading();
        return this.http.post(this.url.update, params);
    }
    /**
     * @return {?}
     */
    application() {
        this.http.showLoading();
        return this.http.getJSON(this.url.application);
    }
    /**
     * @return {?}
     */
    organId() {
        this.http.showLoading();
        return this.http.postJSON(this.url.organ);
    }
    /**
     * @param {?} application
     * @return {?}
     */
    modules(application) {
        this.http.showLoading();
        return this.http.getJSON(this.url.modules, { application: application });
    }
    /**
     * @param {?} contentUuid
     * @return {?}
     */
    downloadFile(contentUuid) {
        return this.http.downloadFile(this.url.download, { docID: contentUuid });
    }
}
ReportTemplateService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ReportTemplateService.ctorParameters = () => [
    { type: HttpClientService }
];
/** @nocollapse */ ReportTemplateService.ngInjectableDef = i0.defineInjectable({ factory: function ReportTemplateService_Factory() { return new ReportTemplateService(i0.inject(i1.HttpClientService)); }, token: ReportTemplateService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ReportTemplateService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    ReportTemplateService.prototype.http;
}
