// inet-social/esm2015/src/social-friend/group-member/group-member.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ListUserComponent } from '../list-user/list-user.component';
export class GroupMemberComponent {
    constructor() {
        this.title = 'Thêm thành viên';
        this.onSelect = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        document.body.appendChild(this.modal._element.nativeElement);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.modal._element.nativeElement.remove();
    }
    /**
     * @return {?}
     */
    open() {
        this.listUser.resetData();
        this.modal.show();
    }
    /**
     * @return {?}
     */
    onShown() {
        this.listUser.focusInput();
        this.listUser.onSearch();
    }
    /**
     * @return {?}
     */
    onSave() {
        this.onSelect.emit(this.listUser.selectMembers);
        this.modal.hide();
    }
    /**
     * @return {?}
     */
    onCancel() {
        this.modal.hide();
    }
}
GroupMemberComponent.decorators = [
    { type: Component, args: [{
                selector: '[socialGroupMember]',
                template: "<div class=\"modal fade\" bsModal #modal=\"bs-modal\" (onShown)=\"onShown()\" tabindex=\"-1\" role=\"dialog\">\n  <div class=\"modal-dialog\">\n    <div class=\"modal-content\">\n      <div class=\"modal-header\">\n        <h5 class=\"modal-title pull-left\">Th\u00EAm th\u00E0nh vi\u00EAn</h5>\n        <button type=\"button\" class=\"close pull-right\" (click)=\"modal.hide()\" tabindex=\"-1\">\n          <span aria-hidden=\"true\">&times;</span>\n        </button>\n      </div>\n      <div class=\"modal-body\">\n        <div socialListUser #listUser></div>\n      </div>\n      <div class=\"modal-footer\">\n        <button type=\"button\" class=\"btn btn-primary\" (click)=\"onSave()\">L\u01B0u</button>\n        <button type=\"button\" class=\"btn btn-secondary\" (click)=\"onCancel()\">\u0110\u00F3ng</button>\n      </div>\n    </div>\n  </div>\n</div>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
GroupMemberComponent.ctorParameters = () => [];
GroupMemberComponent.propDecorators = {
    title: [{ type: Input }],
    onSelect: [{ type: Output }],
    modal: [{ type: ViewChild, args: ['modal',] }],
    listUser: [{ type: ViewChild, args: ['listUser',] }]
};
if (false) {
    /** @type {?} */
    GroupMemberComponent.prototype.title;
    /** @type {?} */
    GroupMemberComponent.prototype.onSelect;
    /** @type {?} */
    GroupMemberComponent.prototype.modal;
    /** @type {?} */
    GroupMemberComponent.prototype.listUser;
}
